--- build/generated/SDL3pp_mixer.h
+++ include/SDL3pp/SDL3pp_mixer.h
@@ -1,10 +1,13 @@
 #ifndef SDL3PP_MIXER_H_
 #define SDL3PP_MIXER_H_
 
-#include <SDL3/SDL_mixer.h>
 #include "SDL3pp_audio.h"
 #include "SDL3pp_version.h"
 
+#if defined(SDL3PP_ENABLE_MIXER) || defined(SDL3PP_DOC)
+
+#include <SDL3_mixer/SDL_mixer.h>
+
 namespace SDL {
 
 /**
@@ -562,9 +565,8 @@
    * @param predecode if true, data will be fully uncompressed before returning.
    * @param closeio true if SDL_mixer should close `io` before returning
    *                (success or failure).
-   * @returns an audio object that can be used to make sound on a mixer on
-   *          success.
-   * @throws Error on failure.
+   * @returns an audio object that can be used to make sound on a mixer, or
+   *          nullptr on failure; call GetError() for more information.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
@@ -582,8 +584,8 @@
    *
    * This is equivalent to calling:
    *
-   * ```c
-   * Mixer.LoadAudio_IO(mixer, IOStream.FromFile(path, "rb"), predecode, true);
+   * ```cpp
+   * mixer.LoadAudio_IO(mixer, IOStream.FromFile(path, "rb"), predecode, true);
    * ```
    *
    * This function loads data from a path on the filesystem. There is also a
@@ -592,9 +594,8 @@
    *
    * @param path the path on the filesystem to load data from.
    * @param predecode if true, data will be fully uncompressed before returning.
-   * @returns an audio object that can be used to make sound on a mixer on
-   *          success.
-   * @throws Error on failure.
+   * @returns an audio object that can be used to make sound on a mixer, or
+   *          nullptr on failure; call GetError() for more information.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
@@ -648,7 +649,6 @@
    * When done with a Audio, it can be freed with Audio.Destroy().
    *
    * @param data the buffer where the audio data lives.
-   * @param datalen the size, in bytes, of the buffer.
    * @param free_when_done if true, `data` will be given to free() when the
    *                       Audio is destroyed.
    * @returns an audio object that can be used to make sound on a mixer on
@@ -721,7 +721,6 @@
    * mixer may be specified.
    *
    * @param data the raw PCM data to load.
-   * @param datalen the size, in bytes, of the raw PCM data.
    * @param spec what format the raw data is in.
    * @returns an audio object that can be used to make sound on a mixer on
    *          success.
@@ -761,7 +760,6 @@
    * simply wants to avoid the extra copy.
    *
    * @param data the buffer where the raw PCM data lives.
-   * @param datalen the size, in bytes, of the buffer.
    * @param spec what format the raw data is in.
    * @param free_when_done if true, `data` will be given to free() when the
    *                       Audio is destroyed.
@@ -851,9 +849,7 @@
    * Tracks are not provided in any guaranteed order.
    *
    * @param tag the tag to search.
-   * @param count a pointer filled in with the number of tracks returned, can be
-   *              nullptr.
-   * @returns  nullptr-terminated on success.
+   * @returns an array of the tracks, nullptr-terminated on success.
    * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread.
@@ -1282,8 +1278,6 @@
    * callback.
    *
    * @param cb the function to call when the mixer mixes. May be nullptr.
-   * @param userdata an opaque pointer provided to the callback for its own
-   *                 personal use.
    * @throws Error on failure.
    *
    * @since This function is available since SDL_mixer 3.0.0.
@@ -1328,12 +1322,11 @@
    * initialized, if available tracks to mix run out, the end of the buffer will
    * be initialized with silence; this silence will not be counted in the return
    * value, so the caller has the option to identify how much of the buffer has
-   * legimitate contents vs appended silence. As such, any value >= 0 signifies
+   * legitimate contents vs appended silence. As such, any value >= 0 signifies
    * success. A return value of -1 means failure (out of memory, invalid
    * parameters, etc).
    *
    * @param buffer a pointer to a buffer to store audio in.
-   * @param buflen the number of bytes to store in buffer.
    * @returns The number of bytes of mixed audio, discounting appended silence,
    *          on success.
    * @throws Error on failure.
@@ -1454,8 +1447,6 @@
  *
  * Locking a nullptr mixer is a safe no-op.
  *
- * @param mixer the mixer to lock. May be nullptr.
- *
  * @threadsafety It is safe to call this function from any thread.
  *
  * @since This function is available since SDL_mixer 3.0.0.
@@ -1502,7 +1493,7 @@
    *
    * Locking a nullptr mixer is a safe no-op.
    *
-   * @param mixer the mixer to lock. May be nullptr.
+   * @param resource the mixer to lock. May be nullptr.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
@@ -1534,8 +1525,6 @@
    *
    * Unlocking a nullptr mixer is a safe no-op.
    *
-   * @param mixer the mixer to unlock. May be nullptr.
-   *
    * @threadsafety This call must be paired with a previous Mixer.Lock call on
    *               the same thread.
    *
@@ -1570,8 +1559,6 @@
    *
    * Unlocking a nullptr mixer is a safe no-op.
    *
-   * @param mixer the mixer to unlock. May be nullptr.
-   *
    * @threadsafety This call must be paired with a previous Mixer.Lock call on
    *               the same thread.
    *
@@ -1685,8 +1672,8 @@
    * @param predecode if true, data will be fully uncompressed before returning.
    * @param closeio true if SDL_mixer should close `io` before returning
    *                (success or failure).
-   * @post an audio object that can be used to make sound on a mixer on success.
-   * @throws Error on failure.
+   * @post an audio object that can be used to make sound on a mixer, or nullptr
+   *       on failure; call GetError() for more information.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
@@ -1716,8 +1703,8 @@
    *              nullptr.
    * @param path the path on the filesystem to load data from.
    * @param predecode if true, data will be fully uncompressed before returning.
-   * @post an audio object that can be used to make sound on a mixer on success.
-   * @throws Error on failure.
+   * @post an audio object that can be used to make sound on a mixer, or nullptr
+   *       on failure; call GetError() for more information.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
@@ -1837,7 +1824,6 @@
    * @param mixer a mixer this audio is intended to be used with. May be
    *              nullptr.
    * @param data the raw PCM data to load.
-   * @param datalen the size, in bytes, of the raw PCM data.
    * @param spec what format the raw data is in.
    * @post an audio object that can be used to make sound on a mixer on success.
    * @throws Error on failure.
@@ -2197,7 +2183,7 @@
  * A callback that fires when a Track is mixing at various stages.
  *
  * This callback is fired for different parts of the mixing pipeline, and gives
- * the app visbility into the audio data that is being generated at various
+ * the app visibility into the audio data that is being generated at various
  * stages.
  *
  * The audio data passed through here is _not_ const data; the app is permitted
@@ -2239,7 +2225,7 @@
  * A callback that fires when a Track is mixing at various stages.
  *
  * This callback is fired for different parts of the mixing pipeline, and gives
- * the app visbility into the audio data that is being generated at various
+ * the app visibility into the audio data that is being generated at various
  * stages.
  *
  * The audio data passed through here is _not_ const data; the app is permitted
@@ -2660,9 +2646,7 @@
    *
    * Tags are not provided in any guaranteed order.
    *
-   * @param count a pointer filled in with the number of tags returned, can be
-   *              nullptr.
-   * @returns  nullptr-terminated on success.
+   * @returns an array of the tags, nullptr-terminated on success.
    * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread.
@@ -3296,7 +3280,6 @@
    * array after this call.
    *
    * @param chmap the new channel map, nullptr to reset to default.
-   * @param count The number of channels in the map.
    * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread.
@@ -3386,7 +3369,7 @@
    * If 3D positioning isn't enabled for this track, through a call to
    * Track.Set3DPosition(), this will return (0,0,0).
    *
-   * @param position on successful return, will contain the track's position.
+   * @returns the he track's position on successful return.
    * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread.
@@ -3473,8 +3456,6 @@
    * without any gap in the audio.
    *
    * @param cb the function to call when the track stops. May be nullptr.
-   * @param userdata an opaque pointer provided to the callback for its own
-   *                 personal use.
    * @throws Error on failure.
    *
    * @since This function is available since SDL_mixer 3.0.0.
@@ -3534,8 +3515,6 @@
    * callback.
    *
    * @param cb the function to call when the track mixes. May be nullptr.
-   * @param userdata an opaque pointer provided to the callback for its own
-   *                 personal use.
    * @throws Error on failure.
    *
    * @since This function is available since SDL_mixer 3.0.0.
@@ -3602,8 +3581,6 @@
    * callback.
    *
    * @param cb the function to call when the track mixes. May be nullptr.
-   * @param userdata an opaque pointer provided to the callback for its own
-   *                 personal use.
    * @throws Error on failure.
    *
    * @since This function is available since SDL_mixer 3.0.0.
@@ -4021,6 +3998,8 @@
   constexpr operator GroupRaw() const noexcept { return get(); }
 };
 
+#ifdef SDL3PP_DOC
+
 /**
  * The current major version of SDL_mixer headers.
  *
@@ -4070,6 +4049,8 @@
    (SDL_MIXER_MAJOR_VERSION > X || SDL_MIXER_MINOR_VERSION > Y ||              \
     SDL_MIXER_MICRO_VERSION >= Z))
 
+#endif // SDL3PP_DOC
+
 namespace MIX {
 
 /**
@@ -4525,9 +4506,8 @@
  * @param predecode if true, data will be fully uncompressed before returning.
  * @param closeio true if SDL_mixer should close `io` before returning (success
  *                or failure).
- * @returns an audio object that can be used to make sound on a mixer on
- *          success.
- * @throws Error on failure.
+ * @returns an audio object that can be used to make sound on a mixer, or
+ *          nullptr on failure; call GetError() for more information.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -4555,12 +4535,12 @@
                     IOStreamRef io,
                     bool predecode,
                     bool closeio)
-  : m_resource(CheckError(MIX_LoadAudio_IO(mixer, io, predecode, closeio)))
+  : m_resource(MIX_LoadAudio_IO(mixer, io, predecode, closeio))
 {
 }
 
 inline Audio::Audio(MixerRef mixer, StringParam path, bool predecode)
-  : m_resource(CheckError(MIX_LoadAudio(mixer, path, predecode)))
+  : m_resource(MIX_LoadAudio(mixer, path, predecode))
 {
 }
 
@@ -4588,8 +4568,8 @@
  *
  * This is equivalent to calling:
  *
- * ```c
- * Mixer.LoadAudio_IO(mixer, IOStream.FromFile(path, "rb"), predecode, true);
+ * ```cpp
+ * mixer.LoadAudio_IO(mixer, IOStream.FromFile(path, "rb"), predecode, true);
  * ```
  *
  * This function loads data from a path on the filesystem. There is also a
@@ -4599,9 +4579,8 @@
  * @param mixer a mixer this audio is intended to be used with. May be nullptr.
  * @param path the path on the filesystem to load data from.
  * @param predecode if true, data will be fully uncompressed before returning.
- * @returns an audio object that can be used to make sound on a mixer on
- *          success.
- * @throws Error on failure.
+ * @returns an audio object that can be used to make sound on a mixer, or
+ *          nullptr on failure; call GetError() for more information.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -4663,7 +4642,6 @@
  *
  * @param mixer a mixer this audio is intended to be used with. May be nullptr.
  * @param data the buffer where the audio data lives.
- * @param datalen the size, in bytes, of the buffer.
  * @param free_when_done if true, `data` will be given to free() when the Audio
  *                       is destroyed.
  * @returns an audio object that can be used to make sound on a mixer on
@@ -4683,8 +4661,8 @@
                              SourceBytes data,
                              bool free_when_done)
 {
-  return CheckError(
-    MIX_LoadAudioNoCopy(mixer, data.data(), data.size_bytes(), free_when_done));
+  return Audio(CheckError(MIX_LoadAudioNoCopy(
+    mixer, data.data(), data.size_bytes(), free_when_done)));
 }
 
 inline Audio Mixer::LoadAudioNoCopy(SourceBytes data, bool free_when_done)
@@ -4829,7 +4807,6 @@
  *
  * @param mixer a mixer this audio is intended to be used with. May be nullptr.
  * @param data the raw PCM data to load.
- * @param datalen the size, in bytes, of the raw PCM data.
  * @param spec what format the raw data is in.
  * @returns an audio object that can be used to make sound on a mixer on
  *          success.
@@ -4880,7 +4857,6 @@
  *
  * @param mixer a mixer this audio is intended to be used with. May be nullptr.
  * @param data the buffer where the raw PCM data lives.
- * @param datalen the size, in bytes, of the buffer.
  * @param spec what format the raw data is in.
  * @param free_when_done if true, `data` will be given to free() when the Audio
  *                       is destroyed.
@@ -4903,8 +4879,8 @@
                                 const AudioSpec& spec,
                                 bool free_when_done)
 {
-  return CheckError(MIX_LoadRawAudioNoCopy(
-    mixer, data.data(), data.size_bytes(), &spec, free_when_done));
+  return Audio(CheckError(MIX_LoadRawAudioNoCopy(
+    mixer, data.data(), data.size_bytes(), &spec, free_when_done)));
 }
 
 inline Audio Mixer::LoadRawAudioNoCopy(SourceBytes data,
@@ -4956,7 +4932,7 @@
                                  float amplitude,
                                  Sint64 ms)
 {
-  return CheckError(MIX_CreateSineWaveAudio(mixer, hz, amplitude, ms));
+  return Audio(CheckError(MIX_CreateSineWaveAudio(mixer, hz, amplitude, ms)));
 }
 
 inline Audio Mixer::CreateSineWaveAudio(int hz, float amplitude, Sint64 ms)
@@ -5080,8 +5056,10 @@
 
 inline Sint64 Audio::GetDuration() { return SDL::GetAudioDuration(m_resource); }
 
+/// Unknown duration, when the length of the audio can't be determined.
 constexpr Sint64 DURATION_UNKNOWN = MIX_DURATION_UNKNOWN;
 
+/// Infinite duration, when the audio never runs out of sound to generate.
 constexpr Sint64 DURATION_INFINITE = MIX_DURATION_INFINITE;
 
 /**
@@ -5523,9 +5501,7 @@
  * Tags are not provided in any guaranteed order.
  *
  * @param track the track to query.
- * @param count a pointer filled in with the number of tags returned, can be
- *              nullptr.
- * @returns  nullptr-terminated on success.
+ * @returns an array of the tags, nullptr-terminated on success.
  * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
@@ -5534,7 +5510,9 @@
  */
 inline OwnArray<char*> GetTrackTags(TrackRef track)
 {
-  return CheckError(MIX_GetTrackTags(track));
+  int count;
+  auto result = CheckError(MIX_GetTrackTags(track, &count));
+  return OwnArray<char*>(result, count);
 }
 
 inline OwnArray<char*> Track::GetTags()
@@ -5549,9 +5527,7 @@
  *
  * @param mixer the mixer to query.
  * @param tag the tag to search.
- * @param count a pointer filled in with the number of tracks returned, can be
- *              nullptr.
- * @returns  nullptr-terminated on success.
+ * @returns an array of the tracks, nullptr-terminated on success.
  * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
@@ -5560,7 +5536,9 @@
  */
 inline OwnArray<TrackRef> GetTaggedTracks(MixerRef mixer, StringParam tag)
 {
-  return CheckError(MIX_GetTaggedTracks(mixer, tag));
+  int count;
+  auto result = CheckError(MIX_GetTaggedTracks(mixer, tag, &count));
+  return OwnArray<TrackRef>(reinterpret_cast<TrackRef*>(result), count);
 }
 
 inline OwnArray<TrackRef> Mixer::GetTaggedTracks(StringParam tag)
@@ -5863,7 +5841,7 @@
  */
 inline Sint64 TrackMSToFrames(TrackRef track, Milliseconds ms)
 {
-  return MIX_TrackMSToFrames(track, ms);
+  return MIX_TrackMSToFrames(track, ms.count());
 }
 
 inline Sint64 Track::MSToFrames(Milliseconds ms)
@@ -5898,7 +5876,7 @@
  */
 inline Milliseconds TrackFramesToMS(TrackRef track, Sint64 frames)
 {
-  return MIX_TrackFramesToMS(track, frames);
+  return Milliseconds(MIX_TrackFramesToMS(track, frames));
 }
 
 inline Milliseconds Track::FramesToMS(Sint64 frames)
@@ -5927,7 +5905,7 @@
  */
 inline Sint64 AudioMSToFrames(AudioRef audio, Milliseconds ms)
 {
-  return MIX_AudioMSToFrames(audio, ms);
+  return MIX_AudioMSToFrames(audio, ms.count());
 }
 
 inline Sint64 Audio::MSToFrames(Milliseconds ms)
@@ -5959,7 +5937,7 @@
  */
 inline Milliseconds AudioFramesToMS(AudioRef audio, Sint64 frames)
 {
-  return MIX_AudioFramesToMS(audio, frames);
+  return Milliseconds(MIX_AudioFramesToMS(audio, frames));
 }
 
 inline Milliseconds Audio::FramesToMS(Sint64 frames)
@@ -5985,7 +5963,7 @@
  */
 inline Sint64 MSToFrames(int sample_rate, Milliseconds ms)
 {
-  return MIX_MSToFrames(sample_rate, ms);
+  return MIX_MSToFrames(sample_rate, ms.count());
 }
 
 /**
@@ -6011,7 +5989,7 @@
  */
 inline Milliseconds FramesToMS(int sample_rate, Sint64 frames)
 {
-  return MIX_FramesToMS(sample_rate, frames);
+  return Milliseconds(MIX_FramesToMS(sample_rate, frames));
 }
 
 /**
@@ -6917,7 +6895,6 @@
  *
  * @param track the track to change.
  * @param chmap the new channel map, nullptr to reset to default.
- * @param count The number of channels in the map.
  * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
@@ -6926,7 +6903,8 @@
  */
 inline void SetTrackOutputChannelMap(TrackRef track, std::span<const int> chmap)
 {
-  CheckError(MIX_SetTrackOutputChannelMap(track, chmap));
+  CheckError(
+    MIX_SetTrackOutputChannelMap(track, chmap.data(), narrowS32(chmap.size())));
 }
 
 inline void Track::SetOutputChannelMap(std::span<const int> chmap)
@@ -7033,7 +7011,7 @@
  * Track.Set3DPosition(), this will return (0,0,0).
  *
  * @param track the track to query.
- * @param position on successful return, will contain the track's position.
+ * @returns the he track's position on successful return.
  * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
@@ -7044,7 +7022,9 @@
  */
 inline Point3D GetTrack3DPosition(TrackRef track)
 {
-  return CheckError(MIX_GetTrack3DPosition(track));
+  Point3D position;
+  CheckError(MIX_GetTrack3DPosition(track, &position));
+  return position;
 }
 
 inline Point3D Track::Get3DPosition()
@@ -7245,8 +7225,6 @@
  *
  * @param track the track to assign this callback to.
  * @param cb the function to call when the track stops. May be nullptr.
- * @param userdata an opaque pointer provided to the callback for its own
- *                 personal use.
  * @throws Error on failure.
  *
  * @since This function is available since SDL_mixer 3.0.0.
@@ -7255,7 +7233,7 @@
  */
 inline void SetTrackStoppedCallback(TrackRef track, TrackStoppedCB cb)
 {
-  static_assert(false, "Not implemented");
+  SetTrackStoppedCallback(track, cb.wrapper, cb.data);
 }
 
 inline void Track::SetStoppedCallback(TrackStoppedCallback cb, void* userdata)
@@ -7324,8 +7302,6 @@
  *
  * @param track the track to assign this callback to.
  * @param cb the function to call when the track mixes. May be nullptr.
- * @param userdata an opaque pointer provided to the callback for its own
- *                 personal use.
  * @throws Error on failure.
  *
  * @since This function is available since SDL_mixer 3.0.0.
@@ -7335,7 +7311,7 @@
  */
 inline void SetTrackRawCallback(TrackRef track, TrackMixCB cb)
 {
-  static_assert(false, "Not implemented");
+  SetTrackRawCallback(track, cb.wrapper, cb.data);
 }
 
 inline void Track::SetRawCallback(TrackMixCallback cb, void* userdata)
@@ -7410,8 +7386,6 @@
  *
  * @param track the track to assign this callback to.
  * @param cb the function to call when the track mixes. May be nullptr.
- * @param userdata an opaque pointer provided to the callback for its own
- *                 personal use.
  * @throws Error on failure.
  *
  * @since This function is available since SDL_mixer 3.0.0.
@@ -7421,7 +7395,7 @@
  */
 inline void SetTrackCookedCallback(TrackRef track, TrackMixCB cb)
 {
-  static_assert(false, "Not implemented");
+  SetTrackCookedCallback(track, cb.wrapper, cb.data);
 }
 
 inline void Track::SetCookedCallback(TrackMixCallback cb, void* userdata)
@@ -7483,8 +7457,6 @@
  *
  * @param group the mixing group to assign this callback to.
  * @param cb the function to call when the group mixes. May be nullptr.
- * @param userdata an opaque pointer provided to the callback for its own
- *                 personal use.
  * @throws Error on failure.
  *
  * @since This function is available since SDL_mixer 3.0.0.
@@ -7493,7 +7465,7 @@
  */
 inline void SetGroupPostMixCallback(GroupRef group, GroupMixCB cb)
 {
-  static_assert(false, "Not implemented");
+  SetGroupPostMixCallback(group, cb.wrapper, cb.data);
 }
 
 inline void Group::SetPostMixCallback(GroupMixCallback cb, void* userdata)
@@ -7546,8 +7518,6 @@
  *
  * @param mixer the mixer to assign this callback to.
  * @param cb the function to call when the mixer mixes. May be nullptr.
- * @param userdata an opaque pointer provided to the callback for its own
- *                 personal use.
  * @throws Error on failure.
  *
  * @since This function is available since SDL_mixer 3.0.0.
@@ -7556,7 +7526,7 @@
  */
 inline void SetPostMixCallback(MixerRef mixer, PostMixCB cb)
 {
-  static_assert(false, "Not implemented");
+  SetPostMixCallback(mixer, cb.wrapper, cb.data);
 }
 
 inline void Mixer::SetPostMixCallback(PostMixCallback cb, void* userdata)
@@ -7604,13 +7574,12 @@
  * initialized, if available tracks to mix run out, the end of the buffer will
  * be initialized with silence; this silence will not be counted in the return
  * value, so the caller has the option to identify how much of the buffer has
- * legimitate contents vs appended silence. As such, any value >= 0 signifies
+ * legitimate contents vs appended silence. As such, any value >= 0 signifies
  * success. A return value of -1 means failure (out of memory, invalid
  * parameters, etc).
  *
  * @param mixer the mixer for which to generate more audio.
  * @param buffer a pointer to a buffer to store audio in.
- * @param buflen the number of bytes to store in buffer.
  * @returns The number of bytes of mixed audio, discounting appended silence, on
  *          success.
  * @throws Error on failure.
@@ -7623,7 +7592,8 @@
  */
 inline int Generate(MixerRef mixer, TargetBytes buffer)
 {
-  return CheckError(MIX_Generate(mixer, buffer.data(), buffer.size_bytes()));
+  return CheckError(
+    MIX_Generate(mixer, buffer.data(), narrowS32(buffer.size_bytes())), -1);
 }
 
 inline int Mixer::Generate(TargetBytes buffer)
@@ -7691,7 +7661,7 @@
    *
    * This function allows properties to be specified. This is intended to supply
    * file-specific settings, such as where to find SoundFonts for a MIDI file,
-   * etc. In most cases, the caller should pass a zero to specify no extra
+   * etc. In most cases, the caller should pass a nullptr to specify no extra
    * properties.
    *
    * Properties are discussed in [SDL's
@@ -7704,7 +7674,7 @@
    * call to MIX.Init(), but does not need an actual Mixer to have been created.
    *
    * @param path the path on the filesystem from which to load data.
-   * @param props decoder-specific properties. May be zero.
+   * @param props decoder-specific properties. May be nullptr.
    * @post an audio decoder, ready to decode.
    *
    * @threadsafety It is safe to call this function from any thread.
@@ -7726,7 +7696,7 @@
    *
    * This function allows properties to be specified. This is intended to supply
    * file-specific settings, such as where to find SoundFonts for a MIDI file,
-   * etc. In most cases, the caller should pass a zero to specify no extra
+   * etc. In most cases, the caller should pass a nullptr to specify no extra
    * properties.
    *
    * If `closeio` is true, then `io` will be closed when this decoder is done
@@ -7741,7 +7711,7 @@
    *
    * @param io the i/o stream from which to load data.
    * @param closeio if true, close the i/o stream when done with it.
-   * @param props decoder-specific properties. May be zero.
+   * @param props decoder-specific properties. May be nullptr.
    * @post an audio decoder, ready to decode.
    *
    * @threadsafety It is safe to call this function from any thread.
@@ -7851,10 +7821,9 @@
    * happened.
    *
    * @param buffer the memory buffer to store decoded audio.
-   * @param buflen the maximum number of bytes to store to `buffer`.
    * @param spec the format that audio data will be stored to `buffer`.
-   * @returns number of bytes decoded, or -1 on error; call GetError() for more
-   *          information.
+   * @returns number of bytes decoded on success.
+   * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
@@ -7944,7 +7913,7 @@
  *
  * This function allows properties to be specified. This is intended to supply
  * file-specific settings, such as where to find SoundFonts for a MIDI file,
- * etc. In most cases, the caller should pass a zero to specify no extra
+ * etc. In most cases, the caller should pass a nullptr to specify no extra
  * properties.
  *
  * Properties are discussed in [SDL's
@@ -7957,7 +7926,7 @@
  * call to MIX.Init(), but does not need an actual Mixer to have been created.
  *
  * @param path the path on the filesystem from which to load data.
- * @param props decoder-specific properties. May be zero.
+ * @param props decoder-specific properties. May be nullptr.
  * @returns an audio decoder, ready to decode.
  *
  * @threadsafety It is safe to call this function from any thread.
@@ -7995,7 +7964,7 @@
  *
  * This function allows properties to be specified. This is intended to supply
  * file-specific settings, such as where to find SoundFonts for a MIDI file,
- * etc. In most cases, the caller should pass a zero to specify no extra
+ * etc. In most cases, the caller should pass a nullptr to specify no extra
  * properties.
  *
  * If `closeio` is true, then `io` will be closed when this decoder is done with
@@ -8010,7 +7979,7 @@
  *
  * @param io the i/o stream from which to load data.
  * @param closeio if true, close the i/o stream when done with it.
- * @param props decoder-specific properties. May be zero.
+ * @param props decoder-specific properties. May be nullptr.
  * @returns an audio decoder, ready to decode.
  *
  * @threadsafety It is safe to call this function from any thread.
@@ -8117,10 +8086,9 @@
  *
  * @param audiodecoder the decoder from which to retrieve more data.
  * @param buffer the memory buffer to store decoded audio.
- * @param buflen the maximum number of bytes to store to `buffer`.
  * @param spec the format that audio data will be stored to `buffer`.
- * @returns number of bytes decoded, or -1 on error; call GetError() for more
- *          information.
+ * @returns number of bytes decoded on success.
+ * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -8130,8 +8098,10 @@
                        TargetBytes buffer,
                        const AudioSpec& spec)
 {
-  return MIX_DecodeAudio(
-    audiodecoder, buffer.data(), buffer.size_bytes(), &spec);
+  return CheckError(
+    MIX_DecodeAudio(
+      audiodecoder, buffer.data(), narrowS32(buffer.size_bytes()), &spec),
+    -1);
 }
 
 inline int AudioDecoder::DecodeAudio(TargetBytes buffer, const AudioSpec& spec)
@@ -8143,4 +8113,6 @@
 
 } // namespace SDL
 
+#endif /* defined(SDL3PP_ENABLE_MIXER) || defined(SDL3PP_DOC) */
+
 #endif /* SDL3PP_MIXER_H_ */
