--- src/generated/SDL3pp_video.h
+++ include/SDL3pp/SDL3pp_video.h
@@ -9,7 +9,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryVideo Category Video
+ * @defgroup CategoryVideo Display and Window Management
  *
  * SDL's video subsystem is largely interested in abstracting window management
  * from the underlying operating system. You can create windows, manage them in
@@ -80,7 +80,7 @@
 struct GLContext;
 
 /// Alias to raw representation for GLContext.
-using GLContextRaw = SDL_GLContext*;
+using GLContextRaw = SDL_GLContext;
 
 // Forward decl
 struct GLContextScoped;
@@ -295,7 +295,7 @@
    * The primary display is often located at (0,0), but may be placed at a
    * different location depending on monitor layout.
    *
-   * @param rect the Rect structure filled in with the display bounds.
+   * @returns the Rect structure filled in with the display bounds.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -319,7 +319,7 @@
    * so these are good guidelines for the maximum space available to a
    * non-fullscreen window.
    *
-   * @param rect the Rect structure filled in with the display bounds.
+   * @returns the Rect structure filled in with the display bounds.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -397,12 +397,8 @@
    * - refresh rate -> highest to lowest
    * - pixel density -> lowest to highest
    *
-   * @param count a pointer filled in with the number of display modes returned,
-   *              may be nullptr.
    * @returns a nullptr terminated array of display mode pointers or nullptr on
-   *          failure; call GetError() for more information. This is a single
-   *          allocation that should be freed with free() when it is no longer
-   *          needed.
+   *          failure; call GetError() for more information.
    *
    * @threadsafety This function should only be called on the main thread.
    *
@@ -422,14 +418,12 @@
    * and finally checking the refresh rate. If all the available modes are too
    * small, then false is returned.
    *
-   * @param w the width in pixels of the desired display mode.
-   * @param h the height in pixels of the desired display mode.
+   * @param size the width and height in pixels of the desired display mode.
    * @param refresh_rate the refresh rate of the desired display mode, or 0.0f
    *                     for the desktop refresh rate.
    * @param include_high_density_modes boolean to include high density modes in
    *                                   the search.
-   * @param closest a pointer filled in with the closest display mode equal to
-   *                or larger than the desired mode.
+   * @returns the closest display mode equal to or larger than the desired mode.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -675,6 +669,7 @@
  * @since This enum is available since SDL 3.2.0.
  *
  * @sa HitTest
+ * @sa HitTestCB
  */
 using HitTestResult = SDL_HitTestResult;
 
@@ -731,11 +726,11 @@
  *
  * @param win the Window where hit-testing was set on.
  * @param area an Point which should be hit-tested.
- * @param data what was passed as `callback_data` to Window.SetHitTest().
  * @returns an HitTestResult value.
  *
- * @sa Window.SetHitTest
+ * @cat listener-callback
  *
+ * @sa Window.SetHitTest
  * @sa HitTest
  */
 using HitTestCB =
@@ -757,46 +752,23 @@
  */
 using ProgressState = SDL_ProgressState;
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 /// An invalid progress state indicating an error; check GetError()
 constexpr ProgressState PROGRESS_STATE_INVALID = SDL_PROGRESS_STATE_INVALID;
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 constexpr ProgressState PROGRESS_STATE_NONE =
-  SDL_PROGRESS_STATE_NONE; ///< No progress bar is shown
-
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
+  SDL_PROGRESS_STATE_NONE; ///< No progress bar is shown.
 
+/// The progress bar is shown in a indeterminate state.
 constexpr ProgressState PROGRESS_STATE_INDETERMINATE =
   SDL_PROGRESS_STATE_INDETERMINATE; ///< The progress bar is shown in a
                                     ///< indeterminate state
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 constexpr ProgressState PROGRESS_STATE_NORMAL =
   SDL_PROGRESS_STATE_NORMAL; ///< The progress bar is shown in a normal state
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 constexpr ProgressState PROGRESS_STATE_PAUSED =
   SDL_PROGRESS_STATE_PAUSED; ///< The progress bar is shown in a paused state
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * The progress bar is shown in a state indicating the application had an error
  */
@@ -809,9 +781,9 @@
  *
  * @since This struct is available since SDL 3.2.0.
  *
- * @sa Window.Window
- *
  * @cat resource
+ *
+ * @sa Window.Window
  */
 class Window
 {
@@ -855,21 +827,20 @@
    * Create a window and default renderer.
    *
    * @param title the title of the window, in UTF-8 encoding.
-   * @param width the width of the window.
-   * @param height the height of the window.
+   * @param size the width and height of the window.
    * @param window_flags the flags used to create the window (see
-   *                     Window.Window()).
-   * @param window a pointer filled with the window, or nullptr on error.
-   * @param renderer a pointer filled with the renderer, or nullptr on error.
-   * @post true on success or false on failure; call GetError() for more
-   *       information.
+   *                     Window.Window(StringParam,const
+   *                     PointRaw&,WindowFlags)).
+   * @param renderer a pointer filled with the renderer.
+   * @returns the created window on success.
+   * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
    *
    * @since This function is available since SDL 3.2.0.
    *
    * @sa Renderer.Renderer
-   * @sa Window.Window
+   * @sa Window.Window(StringParam,const PointRaw&,WindowFlags))
    */
   Window(StringParam title,
          const PointRaw& size,
@@ -947,8 +918,7 @@
    * in a future version of SDL.
    *
    * @param title the title of the window, in UTF-8 encoding.
-   * @param w the width of the window.
-   * @param h the height of the window.
+   * @param size the width and height of the window.
    * @param flags 0, or one or more WindowFlags OR'd together.
    * @post the window that was created or nullptr on failure; call GetError()
    *       for more information.
@@ -963,7 +933,7 @@
    * @sa Window.Destroy
    */
   Window(StringParam title, const PointRaw& size, WindowFlags flags = 0)
-    : m_resource(SDL_CreateWindow(title, size, flags))
+    : m_resource(SDL_CreateWindow(title, size.x, size.y, flags))
   {
   }
 
@@ -1021,12 +991,9 @@
    * hidden will be restored when the parent is shown.
    *
    * @param parent the parent of the window, must not be nullptr.
-   * @param offset_x the x position of the popup window relative to the origin
-   *                 of the parent.
-   * @param offset_y the y position of the popup window relative to the origin
-   *                 of the parent window.
-   * @param w the width of the window.
-   * @param h the height of the window.
+   * @param offset the x, y position of the popup window relative to the origin
+   *               of the parent.
+   * @param size the width and height of the window.
    * @param flags WINDOW_TOOLTIP or WINDOW_POPUP_MENU, and zero or more
    *              additional WindowFlags OR'd together.
    * @post the window that was created or nullptr on failure; call GetError()
@@ -1045,7 +1012,12 @@
          const PointRaw& offset,
          const PointRaw& size,
          WindowFlags flags = 0)
-    : m_resource(SDL_CreatePopupWindow(parent, offset, size, flags))
+    : m_resource(SDL_CreatePopupWindow(parent,
+                                       offset.x,
+                                       offset.y,
+                                       size.x,
+                                       size.y,
+                                       flags))
   {
   }
 
@@ -1393,7 +1365,6 @@
   /**
    * Get the raw ICC profile data for the screen the window is currently on.
    *
-   * @param size the size of the ICC profile.
    * @returns the raw ICC profile data on success.
    * @throws Error on failure.
    *
@@ -1672,9 +1643,7 @@
    * Additionally, as this is just a request, it can be denied by the windowing
    * system.
    *
-   * @param x the x coordinate of the window, or `WINDOWPOS_CENTERED` or
-   *          `WINDOWPOS_UNDEFINED`.
-   * @param y the y coordinate of the window, or `WINDOWPOS_CENTERED` or
+   * @param p the coordinates of the window, or `WINDOWPOS_CENTERED` or
    *          `WINDOWPOS_UNDEFINED`.
    * @throws Error on failure.
    *
@@ -1719,10 +1688,7 @@
    * If you do not need the value for one of the positions a nullptr may be
    * passed in the `x` or `y` parameter.
    *
-   * @param x a pointer filled in with the x position of the window, may be
-   *          nullptr.
-   * @param y a pointer filled in with the y position of the window, may be
-   *          nullptr.
+   * @returns the position on success.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1730,6 +1696,7 @@
    * @since This function is available since SDL 3.2.0.
    *
    * @sa Window.SetPosition
+   * @sa SetPosition(int *, int *)
    */
   Point GetPosition() const;
 
@@ -1754,8 +1721,7 @@
    * to remain within the usable desktop bounds). Additionally, as this is just
    * a request, it can be denied by the windowing system.
    *
-   * @param w the width of the window, must be > 0.
-   * @param h the height of the window, must be > 0.
+   * @param size the width and height of the window, must be > 0.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1797,8 +1763,7 @@
    * window is on a high pixel density display. Use Window.GetSizeInPixels() or
    * Renderer.GetOutputSize() to get the real client area size in pixels.
    *
-   * @param w a pointer filled in with the width of the window, may be nullptr.
-   * @param h a pointer filled in with the height of the window, may be nullptr.
+   * @returns a point with width and height on success
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1809,11 +1774,36 @@
    * @sa Window.GetSizeInPixels
    * @sa Window.SetSize
    * @sa EVENT_WINDOW_RESIZED
+   * @sa GetSize(int *, int *)
    */
   Point GetSize() const;
 
+  /**
+   * @brief Request the window's position and size to be set.
+   *
+   * @param rect the rect containing the new coordinates
+   * @throws Error on failure.
+   *
+   * @threadsafety This function should only be called on the main thread.
+   *
+   * @sa WindowRef.SetPosition()
+   * @sa WindowRef.SetSize()
+   */
   void SetRect(Rect rect);
 
+  /**
+   * Get the position and client size of a window.
+   *
+   * This is the current position of the window as last reported by the
+   * windowing system.
+   *
+   * The window pixel size may differ from its window coordinate size if the
+   * window is on a high pixel density display. Use Window.GetSizeInPixels()
+   * or RendererRef.GetOutputSize() to get the real client area size in pixels.
+   *
+   * @return Rect with the position and size
+   * @throws Error on failure.
+   */
   Rect GetRect() const;
 
   /**
@@ -1826,8 +1816,6 @@
    * into the rest of the window, but it should not contain visually important
    * or interactable content.
    *
-   * @param rect a pointer filled in with the client area that is safe for
-   *             interactive content.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1940,16 +1928,15 @@
    *
    * @sa Window.Window
    * @sa Window.GetSize
+   * @sa GetSizeInPixels()
    */
   void GetSizeInPixels(int* w, int* h) const;
 
   /**
    * Get the size of a window's client area, in pixels.
    *
-   * @param w a pointer to variable for storing the width in pixels, may be
-   *          nullptr.
-   * @param h a pointer to variable for storing the height in pixels, may be
-   *          nullptr.
+   * @returns the size on success or std::nullopt on failure; call GetError()
+   * for more information.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1958,14 +1945,14 @@
    *
    * @sa Window.Window
    * @sa Window.GetSize
+   * @sa GetSizeInPixels(int*, int*)
    */
   Point GetSizeInPixels() const;
 
   /**
    * Set the minimum size of a window's client area.
    *
-   * @param min_w the minimum width of the window, or 0 for no limit.
-   * @param min_h the minimum height of the window, or 0 for no limit.
+   * @param p the minimum width and heigh of the window, or 0 for no limit.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1998,8 +1985,7 @@
   /**
    * Set the maximum size of a window's client area.
    *
-   * @param max_w the maximum width of the window, or 0 for no limit.
-   * @param max_h the maximum height of the window, or 0 for no limit.
+   * @param p the maximum width and height of the window, or 0 for no limit.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -2379,8 +2365,8 @@
   /**
    * Get VSync for the window surface.
    *
-   * @param vsync an int filled with the current vertical refresh sync interval.
-   *              See Window.SetSurfaceVSync() for the meaning of the value.
+   * @returns the current vertical refresh sync interval. See
+   *          Window.SetSurfaceVSync() for the meaning of the value.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -2425,7 +2411,6 @@
    *
    * @param rects an array of Rect structures representing areas of the surface
    *              to copy, in pixels.
-   * @param numrects the number of rectangles.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -2674,10 +2659,8 @@
    * On platforms or desktops where this is unsupported, this function does
    * nothing.
    *
-   * @param x the x coordinate of the menu, relative to the origin (top-left) of
-   *          the client area.
-   * @param y the y coordinate of the menu, relative to the origin (top-left) of
-   *          the client area.
+   * @param p the x, y coordinates of the menu, relative to the origin
+   *          (top-left) of the client area.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -2761,12 +2744,13 @@
    * devoid of allocations, etc.
    *
    * @param callback the function to call when doing a hit-test.
-   * @param callback_data an app-defined void pointer passed to **callback**.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
    *
    * @since This function is available since SDL 3.2.0.
+   *
+   * @cat listener-callback
    */
   void SetHitTest(HitTestCB callback);
 
@@ -2823,10 +2807,6 @@
    */
   void SetProgressState(ProgressState state);
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
   /**
    * Get the state of the progress bar for the given window’s taskbar icon.
    *
@@ -2839,10 +2819,6 @@
    */
   ProgressState GetProgressState();
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
   /**
    * Sets the value of the progress bar for the given window’s taskbar icon.
    *
@@ -2856,10 +2832,6 @@
    */
   void SetProgressValue(float value);
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
   /**
    * Get the value of the progress bar for the given window’s taskbar icon.
    *
@@ -3131,8 +3103,7 @@
    * Note that this function will appear to succeed, but not actually move the
    * mouse when used over Microsoft Remote Desktop.
    *
-   * @param x the x coordinate within the window.
-   * @param y the y coordinate within the window.
+   * @param p the x, y coordinates within the window.
    *
    * @threadsafety This function should only be called on the main thread.
    *
@@ -3449,9 +3420,6 @@
   /**
    * Delete an OpenGL context.
    *
-   * @returns true on success or false on failure; call GetError() for more
-   *          information.
-   *
    * @threadsafety This function should only be called on the main thread.
    *
    * @since This function is available since SDL 3.2.0.
@@ -3466,7 +3434,6 @@
    * The context must have been created with a compatible window.
    *
    * @param window the window to associate with the context.
-   * @param context the OpenGL context to associate with the window.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -3867,17 +3834,19 @@
 /**
  * Get a list of currently connected displays.
  *
- * @param count a pointer filled in with the number of displays returned, may be
- *              nullptr.
  * @returns a 0 terminated array of display instance IDs or nullptr on failure;
- *          call GetError() for more information. This should be freed with
- *          free() when it is no longer needed.
+ *          call GetError() for more information.
  *
  * @threadsafety This function should only be called on the main thread.
  *
  * @since This function is available since SDL 3.2.0.
  */
-inline OwnArray<DisplayID> GetDisplays() { return SDL_GetDisplays(); }
+inline OwnArray<DisplayID> GetDisplays()
+{
+  int count = 0;
+  auto data = reinterpret_cast<DisplayID*>(SDL_GetDisplays(&count));
+  return OwnArray<DisplayID>{data, size_t(count)};
+}
 
 /**
  * Return the primary display.
@@ -3937,7 +3906,7 @@
  */
 inline PropertiesRef GetDisplayProperties(DisplayID displayID)
 {
-  return CheckError(SDL_GetDisplayProperties(displayID));
+  return {CheckError(SDL_GetDisplayProperties(displayID))};
 }
 
 inline PropertiesRef Display::GetProperties() const
@@ -3957,10 +3926,6 @@
 constexpr auto WAYLAND_WL_OUTPUT_POINTER =
   SDL_PROP_DISPLAY_WAYLAND_WL_OUTPUT_POINTER;
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 constexpr auto WINDOWS_HMONITOR_POINTER =
   SDL_PROP_DISPLAY_WINDOWS_HMONITOR_POINTER;
 
@@ -3998,7 +3963,7 @@
  * different location depending on monitor layout.
  *
  * @param displayID the instance ID of the display to query.
- * @param rect the Rect structure filled in with the display bounds.
+ * @returns the Rect structure filled in with the display bounds.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -4010,7 +3975,9 @@
  */
 inline Rect GetDisplayBounds(Display displayID)
 {
-  return CheckError(SDL_GetDisplayBounds(displayID));
+  Rect bounds;
+  SDL_GetDisplayBounds(displayID, &bounds);
+  return bounds;
 }
 
 inline Rect Display::GetBounds() const
@@ -4030,7 +3997,7 @@
  * window.
  *
  * @param displayID the instance ID of the display to query.
- * @param rect the Rect structure filled in with the display bounds.
+ * @returns the Rect structure filled in with the display bounds.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -4042,7 +4009,9 @@
  */
 inline Rect GetDisplayUsableBounds(Display displayID)
 {
-  return CheckError(SDL_GetDisplayUsableBounds(displayID));
+  Rect bounds;
+  SDL_GetDisplayUsableBounds(displayID, &bounds);
+  return bounds;
 }
 
 inline Rect Display::GetUsableBounds() const
@@ -4144,12 +4113,8 @@
  * - pixel density -> lowest to highest
  *
  * @param displayID the instance ID of the display to query.
- * @param count a pointer filled in with the number of display modes returned,
- *              may be nullptr.
  * @returns a nullptr terminated array of display mode pointers or nullptr on
- *          failure; call GetError() for more information. This is a single
- *          allocation that should be freed with free() when it is no longer
- *          needed.
+ *          failure; call GetError() for more information.
  *
  * @threadsafety This function should only be called on the main thread.
  *
@@ -4159,7 +4124,9 @@
  */
 inline OwnArray<DisplayMode*> GetFullscreenDisplayModes(Display displayID)
 {
-  return SDL_GetFullscreenDisplayModes(displayID);
+  int count = 0;
+  auto data = CheckError(SDL_GetFullscreenDisplayModes(displayID, &count));
+  return OwnArray<DisplayMode*>{data, size_t(count)};
 }
 
 inline OwnArray<DisplayMode*> Display::GetFullscreenModes() const
@@ -4178,14 +4145,12 @@
  * then false is returned.
  *
  * @param displayID the instance ID of the display to query.
- * @param w the width in pixels of the desired display mode.
- * @param h the height in pixels of the desired display mode.
+ * @param size the width and height in pixels of the desired display mode.
  * @param refresh_rate the refresh rate of the desired display mode, or 0.0f for
  *                     the desktop refresh rate.
  * @param include_high_density_modes boolean to include high density modes in
  *                                   the search.
- * @param closest a pointer filled in with the closest display mode equal to or
- *                larger than the desired mode.
+ * @returns the closest display mode equal to or larger than the desired mode.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -4201,8 +4166,14 @@
   float refresh_rate,
   bool include_high_density_modes)
 {
-  return CheckError(SDL_GetClosestFullscreenDisplayMode(
-    displayID, size, refresh_rate, include_high_density_modes));
+  DisplayMode mode;
+  CheckError(SDL_GetClosestFullscreenDisplayMode(displayID,
+                                                 size.x,
+                                                 size.y,
+                                                 refresh_rate,
+                                                 include_high_density_modes,
+                                                 &mode));
+  return mode;
 }
 
 inline DisplayMode Display::GetClosestFullscreenMode(
@@ -4235,7 +4206,7 @@
  */
 inline const DisplayMode& GetDesktopDisplayMode(DisplayID displayID)
 {
-  return SDL_GetDesktopDisplayMode(displayID);
+  return *SDL_GetDesktopDisplayMode(displayID);
 }
 
 inline const DisplayMode& Display::GetDesktopMode() const
@@ -4264,7 +4235,7 @@
  */
 inline const DisplayMode& GetCurrentDisplayMode(DisplayID displayID)
 {
-  return SDL_GetCurrentDisplayMode(displayID);
+  return *SDL_GetCurrentDisplayMode(displayID);
 }
 
 inline const DisplayMode& Display::GetCurrentMode() const
@@ -4288,7 +4259,7 @@
  */
 inline Display GetDisplayForPoint(const PointRaw& point)
 {
-  return SDL_GetDisplayForPoint(point);
+  return SDL_GetDisplayForPoint(&point);
 }
 
 inline Display Display::GetForPoint(const PointRaw& point)
@@ -4313,7 +4284,7 @@
  */
 inline Display GetDisplayForRect(const RectRaw& rect)
 {
-  return CheckError(SDL_GetDisplayForRect(rect));
+  return CheckError(SDL_GetDisplayForRect(&rect));
 }
 
 inline Display Display::GetForRect(const RectRaw& rect)
@@ -4462,7 +4433,7 @@
  */
 inline const DisplayMode& GetWindowFullscreenMode(WindowParam window)
 {
-  return SDL_GetWindowFullscreenMode(window);
+  return *SDL_GetWindowFullscreenMode(window);
 }
 
 inline const DisplayMode& Window::GetFullscreenMode() const
@@ -4474,7 +4445,6 @@
  * Get the raw ICC profile data for the screen the window is currently on.
  *
  * @param window the window to query.
- * @param size the size of the ICC profile.
  * @returns the raw ICC profile data on success.
  * @throws Error on failure.
  *
@@ -4484,7 +4454,8 @@
  */
 inline OwnPtr<void> GetWindowICCProfile(WindowParam window)
 {
-  return CheckError(SDL_GetWindowICCProfile(window));
+  size_t size;
+  return OwnPtr<void>{CheckError(SDL_GetWindowICCProfile(window, &size))};
 }
 
 inline OwnPtr<void> Window::GetICCProfile() const
@@ -4516,17 +4487,19 @@
 /**
  * Get a list of valid windows.
  *
- * @param count a pointer filled in with the number of windows returned, may be
- *              nullptr.
  * @returns a nullptr terminated array of Window pointers or nullptr on failure;
- *          call GetError() for more information. This is a single allocation
- *          that should be freed with free() when it is no longer needed.
+ *          call GetError() for more information.
  *
  * @threadsafety This function should only be called on the main thread.
  *
  * @since This function is available since SDL 3.2.0.
  */
-inline OwnArray<WindowRef> GetWindows() { return SDL_GetWindows(); }
+inline OwnArray<WindowRef> GetWindows()
+{
+  int count = 0;
+  auto data = CheckError(SDL_GetWindows(&count));
+  return OwnArray<WindowRef>{reinterpret_cast<WindowRef*>(data), size_t(count)};
+}
 
 /**
  * Create a window with the specified dimensions and flags.
@@ -4566,8 +4539,8 @@
  * - `WINDOW_TRANSPARENT`: window with transparent buffer
  * - `WINDOW_NOT_FOCUSABLE`: window should not be focusable
  *
- * The Window will be shown if WINDOW_HIDDEN is not set. If hidden at creation
- * time, Window.Show() can be used to show it later.
+ * The Window will be shown if WINDOW_HIDDEN is not set. If hidden at
+ * creation time, Window.Show() can be used to show it later.
  *
  * On Apple's macOS, you **must** set the NSHighResolutionCapable Info.plist
  * property to YES, otherwise you will not receive a High-DPI OpenGL canvas.
@@ -4598,8 +4571,7 @@
  * in a future version of SDL.
  *
  * @param title the title of the window, in UTF-8 encoding.
- * @param w the width of the window.
- * @param h the height of the window.
+ * @param size the width and height of the window.
  * @param flags 0, or one or more WindowFlags OR'd together.
  * @returns the window that was created or nullptr on failure; call GetError()
  *          for more information.
@@ -4674,12 +4646,9 @@
  * hidden will be restored when the parent is shown.
  *
  * @param parent the parent of the window, must not be nullptr.
- * @param offset_x the x position of the popup window relative to the origin of
- *                 the parent.
- * @param offset_y the y position of the popup window relative to the origin of
- *                 the parent window.
- * @param w the width of the window.
- * @param h the height of the window.
+ * @param offset the x, y position of the popup window relative to the origin
+ *               of the parent.
+ * @param size the width and height of the window.
  * @param flags WINDOW_TOOLTIP or WINDOW_POPUP_MENU, and zero or more additional
  *              WindowFlags OR'd together.
  * @returns the window that was created or nullptr on failure; call GetError()
@@ -4959,10 +4928,6 @@
 constexpr auto CREATE_EMSCRIPTEN_CANVAS_ID_STRING =
   SDL_PROP_WINDOW_CREATE_EMSCRIPTEN_CANVAS_ID_STRING;
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 constexpr auto CREATE_EMSCRIPTEN_KEYBOARD_ELEMENT_STRING =
   SDL_PROP_WINDOW_CREATE_EMSCRIPTEN_KEYBOARD_ELEMENT_STRING;
 
@@ -5013,6 +4978,10 @@
 constexpr auto OPENVR_OVERLAY_ID_NUMBER =
   SDL_PROP_WINDOW_OPENVR_OVERLAY_ID_NUMBER;
 
+#else
+
+constexpr auto OPENVR_OVERLAY_ID_NUMBER = SDL_PROP_WINDOW_OPENVR_OVERLAY_ID;
+
 #endif // SDL_VERSION_ATLEAST(3, 4, 0)
 
 constexpr auto VIVANTE_DISPLAY_POINTER =
@@ -5067,10 +5036,6 @@
 constexpr auto EMSCRIPTEN_CANVAS_ID_STRING =
   SDL_PROP_WINDOW_EMSCRIPTEN_CANVAS_ID_STRING;
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 constexpr auto EMSCRIPTEN_KEYBOARD_ELEMENT_STRING =
   SDL_PROP_WINDOW_EMSCRIPTEN_KEYBOARD_ELEMENT_STRING;
 
@@ -5119,7 +5084,7 @@
  */
 inline WindowRef GetWindowFromID(WindowID id)
 {
-  return SDL_GetWindowFromID(id);
+  return {SDL_GetWindowFromID(id)};
 }
 
 inline WindowRef Window::FromID(WindowID id)
@@ -5142,7 +5107,7 @@
  */
 inline WindowRef GetWindowParent(WindowParam window)
 {
-  return CheckError(SDL_GetWindowParent(window));
+  return {CheckError(SDL_GetWindowParent(window))};
 }
 
 inline WindowRef Window::GetParent() const
@@ -5276,7 +5241,7 @@
  */
 inline PropertiesRef GetWindowProperties(WindowParam window)
 {
-  return CheckError(SDL_GetWindowProperties(window));
+  return {CheckError(SDL_GetWindowProperties(window))};
 }
 
 inline PropertiesRef Window::GetProperties() const
@@ -5417,9 +5382,7 @@
  * this is just a request, it can be denied by the windowing system.
  *
  * @param window the window to reposition.
- * @param x the x coordinate of the window, or `WINDOWPOS_CENTERED` or
- *          `WINDOWPOS_UNDEFINED`.
- * @param y the y coordinate of the window, or `WINDOWPOS_CENTERED` or
+ * @param p the coordinates of the window, or `WINDOWPOS_CENTERED` or
  *          `WINDOWPOS_UNDEFINED`.
  * @throws Error on failure.
  *
@@ -5432,7 +5395,7 @@
  */
 inline void SetWindowPosition(WindowParam window, const PointRaw& p)
 {
-  CheckError(SDL_SetWindowPosition(window, p));
+  CheckError(SDL_SetWindowPosition(window, p.x, p.y));
 }
 
 inline void Window::SetPosition(const PointRaw& p)
@@ -5477,10 +5440,7 @@
  * in the `x` or `y` parameter.
  *
  * @param window the window to query.
- * @param x a pointer filled in with the x position of the window, may be
- *          nullptr.
- * @param y a pointer filled in with the y position of the window, may be
- *          nullptr.
+ * @returns the position on success.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5491,7 +5451,9 @@
  */
 inline Point GetWindowPosition(WindowParam window)
 {
-  static_assert(false, "Not implemented");
+  Point p;
+  GetWindowPosition(window, &p.x, &p.y);
+  return p;
 }
 
 inline void Window::GetPosition(int* x, int* y) const
@@ -5526,8 +5488,7 @@
  * request, it can be denied by the windowing system.
  *
  * @param window the window to change.
- * @param w the width of the window, must be > 0.
- * @param h the height of the window, must be > 0.
+ * @param size the width and height of the window, must be > 0.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5540,7 +5501,7 @@
  */
 inline void SetWindowSize(WindowParam window, const PointRaw& size)
 {
-  CheckError(SDL_SetWindowSize(window, size));
+  CheckError(SDL_SetWindowSize(window, size.x, size.y));
 }
 
 inline void Window::SetSize(const PointRaw& size)
@@ -5582,8 +5543,7 @@
  * Renderer.GetOutputSize() to get the real client area size in pixels.
  *
  * @param window the window to query the width and height from.
- * @param w a pointer filled in with the width of the window, may be nullptr.
- * @param h a pointer filled in with the height of the window, may be nullptr.
+ * @returns a point with width and height on success
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5597,7 +5557,9 @@
  */
 inline Point GetWindowSize(WindowParam window)
 {
-  static_assert(false, "Not implemented");
+  Point p;
+  GetWindowSize(window, &p.x, &p.y);
+  return p;
 }
 
 inline void Window::GetSize(int* w, int* h) const
@@ -5607,16 +5569,43 @@
 
 inline Point Window::GetSize() const { return SDL::GetWindowSize(m_resource); }
 
+/**
+ * @brief Request the window's position and size to be set.
+ *
+ * @param window the window to query the width and height from.
+ * @param rect the rect containing the new coordinates
+ * @throws Error on failure.
+ *
+ * @threadsafety This function should only be called on the main thread.
+ *
+ * @sa WindowRef.SetPosition()
+ * @sa WindowRef.SetSize()
+ */
 inline void SetWindowRect(WindowParam window, Rect rect)
 {
-  static_assert(false, "Not implemented");
+  SetWindowPosition(window, rect.GetTopLeft());
+  SetWindowSize(window, rect.GetSize());
 }
 
 inline void Window::SetRect(Rect rect) { SDL::SetWindowRect(m_resource, rect); }
 
+/**
+ * Get the position and client size of a window.
+ *
+ * This is the current position of the window as last reported by the
+ * windowing system.
+ *
+ * The window pixel size may differ from its window coordinate size if the
+ * window is on a high pixel density display. Use Window.GetSizeInPixels()
+ * or RendererRef.GetOutputSize() to get the real client area size in pixels.
+ *
+ * @param window the window to query the width and height from.
+ * @return Rect with the position and size
+ * @throws Error on failure.
+ */
 inline Rect GetWindowRect(WindowParam window)
 {
-  static_assert(false, "Not implemented");
+  return Rect{GetWindowPosition(window), GetWindowSize(window)};
 }
 
 inline Rect Window::GetRect() const { return SDL::GetWindowRect(m_resource); }
@@ -5632,8 +5621,6 @@
  * interactable content.
  *
  * @param window the window to query.
- * @param rect a pointer filled in with the client area that is safe for
- *             interactive content.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5642,7 +5629,9 @@
  */
 inline Rect GetWindowSafeArea(WindowParam window)
 {
-  return CheckError(SDL_GetWindowSafeArea(window));
+  Rect rect;
+  CheckError(SDL_GetWindowSafeArea(window, &rect));
+  return rect;
 }
 
 inline Rect Window::GetSafeArea() const
@@ -5803,10 +5792,8 @@
  * Get the size of a window's client area, in pixels.
  *
  * @param window the window from which the drawable size should be queried.
- * @param w a pointer to variable for storing the width in pixels, may be
- *          nullptr.
- * @param h a pointer to variable for storing the height in pixels, may be
- *          nullptr.
+ * @returns the size on success or std::nullopt on failure; call GetError()
+ * for more information.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5818,7 +5805,9 @@
  */
 inline Point GetWindowSizeInPixels(WindowParam window)
 {
-  static_assert(false, "Not implemented");
+  Point p;
+  GetWindowSizeInPixels(window, &p.x, &p.y);
+  return p;
 }
 
 inline void Window::GetSizeInPixels(int* w, int* h) const
@@ -5835,8 +5824,7 @@
  * Set the minimum size of a window's client area.
  *
  * @param window the window to change.
- * @param min_w the minimum width of the window, or 0 for no limit.
- * @param min_h the minimum height of the window, or 0 for no limit.
+ * @param p the minimum width and height of the window, or 0 for no limit.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5848,7 +5836,7 @@
  */
 inline void SetWindowMinimumSize(WindowParam window, const PointRaw& p)
 {
-  CheckError(SDL_SetWindowMinimumSize(window, p));
+  CheckError(SDL_SetWindowMinimumSize(window, p.x, p.y));
 }
 
 inline void Window::SetMinimumSize(const PointRaw& p)
@@ -5887,8 +5875,7 @@
  * Set the maximum size of a window's client area.
  *
  * @param window the window to change.
- * @param max_w the maximum width of the window, or 0 for no limit.
- * @param max_h the maximum height of the window, or 0 for no limit.
+ * @param p the maximum width and height of the window, or 0 for no limit.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5900,7 +5887,7 @@
  */
 inline void SetWindowMaximumSize(WindowParam window, const PointRaw& p)
 {
-  CheckError(SDL_SetWindowMaximumSize(window, p));
+  CheckError(SDL_SetWindowMaximumSize(window, p.x, p.y));
 }
 
 inline void Window::SetMaximumSize(const PointRaw& p)
@@ -6054,10 +6041,6 @@
   CheckError(SDL_SetWindowFillDocument(window, fill));
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 inline void Window::SetFillDocument(bool fill)
 {
   SDL::SetWindowFillDocument(m_resource, fill);
@@ -6359,7 +6342,7 @@
  */
 inline Surface GetWindowSurface(WindowParam window)
 {
-  return SDL_GetWindowSurface(window);
+  return Surface::Borrow(SDL_GetWindowSurface(window));
 }
 
 inline Surface Window::GetSurface()
@@ -6400,16 +6383,16 @@
   SDL::SetWindowSurfaceVSync(m_resource, vsync);
 }
 
+/// Constant to disable vsync
 constexpr int WINDOW_SURFACE_VSYNC_DISABLED = SDL_WINDOW_SURFACE_VSYNC_DISABLED;
 
+/// Constant to enable adaptive vsync
 constexpr int WINDOW_SURFACE_VSYNC_ADAPTIVE = SDL_WINDOW_SURFACE_VSYNC_ADAPTIVE;
 
 /**
  * Get VSync for the window surface.
  *
  * @param window the window to query.
- * @param vsync an int filled with the current vertical refresh sync interval.
- *              See Window.SetSurfaceVSync() for the meaning of the value.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -6420,7 +6403,9 @@
  */
 inline int GetWindowSurfaceVSync(WindowParam window)
 {
-  return CheckError(SDL_GetWindowSurfaceVSync(window));
+  int vsync;
+  CheckError(SDL_GetWindowSurfaceVSync(window, &vsync));
+  return vsync;
 }
 
 inline int Window::GetSurfaceVSync() const
@@ -6469,7 +6454,6 @@
  * @param window the window to update.
  * @param rects an array of Rect structures representing areas of the surface to
  *              copy, in pixels.
- * @param numrects the number of rectangles.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -6482,7 +6466,7 @@
 inline void UpdateWindowSurfaceRects(WindowParam window,
                                      SpanRef<const RectRaw> rects)
 {
-  CheckError(SDL_UpdateWindowSurfaceRects(window, rects));
+  CheckError(SDL_UpdateWindowSurfaceRects(window, rects.data(), rects.size()));
 }
 
 inline void Window::UpdateSurfaceRects(SpanRef<const RectRaw> rects)
@@ -6635,7 +6619,7 @@
  * @sa Window.SetMouseGrab
  * @sa Window.SetKeyboardGrab
  */
-inline WindowRef GetGrabbedWindow() { return SDL_GetGrabbedWindow(); }
+inline WindowRef GetGrabbedWindow() { return {SDL_GetGrabbedWindow()}; }
 
 inline WindowRef Window::GetGrabbed() { return SDL::GetGrabbedWindow(); }
 
@@ -6660,7 +6644,7 @@
  */
 inline void SetWindowMouseRect(WindowParam window, const RectRaw& rect)
 {
-  CheckError(SDL_SetWindowMouseRect(window, rect));
+  CheckError(SDL_SetWindowMouseRect(window, &rect));
 }
 
 inline void Window::SetMouseRect(const RectRaw& rect)
@@ -6848,9 +6832,7 @@
  * nothing.
  *
  * @param window the window for which the menu will be displayed.
- * @param x the x coordinate of the menu, relative to the origin (top-left) of
- *          the client area.
- * @param y the y coordinate of the menu, relative to the origin (top-left) of
+ * @param p the coordinates of the menu, relative to the origin (top-left) of
  *          the client area.
  * @throws Error on failure.
  *
@@ -6860,7 +6842,7 @@
  */
 inline void ShowWindowSystemMenu(WindowParam window, const PointRaw& p)
 {
-  CheckError(SDL_ShowWindowSystemMenu(window, p));
+  CheckError(SDL_ShowWindowSystemMenu(window, p.x, p.y));
 }
 
 inline void Window::ShowSystemMenu(const PointRaw& p)
@@ -6950,7 +6932,6 @@
  *
  * @param window the window to set hit-testing on.
  * @param callback the function to call when doing a hit-test.
- * @param callback_data an app-defined void pointer passed to **callback**.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -6959,7 +6940,7 @@
  */
 inline void SetWindowHitTest(WindowParam window, HitTestCB callback)
 {
-  static_assert(false, "Not implemented");
+  SetWindowHitTest(window, callback.wrapper, callback.data);
 }
 
 inline void Window::SetHitTest(HitTest callback, void* callback_data)
@@ -7033,8 +7014,8 @@
  * Sets the state of the progress bar for the given window’s taskbar icon.
  *
  * @param window the window whose progress state is to be modified.
- * @param state the progress state. `PROGRESS_STATE_NONE` stops displaying the
- *              progress bar.
+ * @param state the progress state. `PROGRESS_STATE_NONE` stops displaying
+ *              the progress bar.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -7046,19 +7027,11 @@
   CheckError(SDL_SetWindowProgressState(window, state));
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 inline void Window::SetProgressState(ProgressState state)
 {
   SDL::SetWindowProgressState(m_resource, state);
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Get the state of the progress bar for the given window’s taskbar icon.
  *
@@ -7075,19 +7048,11 @@
   return SDL_GetWindowProgressState(window);
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 inline ProgressState Window::GetProgressState()
 {
   return SDL::GetWindowProgressState(m_resource);
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Sets the value of the progress bar for the given window’s taskbar icon.
  *
@@ -7105,19 +7070,11 @@
   CheckError(SDL_SetWindowProgressValue(window, value));
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 inline void Window::SetProgressValue(float value)
 {
   SDL::SetWindowProgressValue(m_resource, value);
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Get the value of the progress bar for the given window’s taskbar icon.
  *
@@ -7134,10 +7091,6 @@
   return SDL_GetWindowProgressValue(window);
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 inline float Window::GetProgressValue()
 {
   return SDL::GetWindowProgressValue(m_resource);
@@ -7470,7 +7423,7 @@
  */
 inline void GL_MakeCurrent(WindowParam window, GLContext context)
 {
-  CheckError(SDL_GL_MakeCurrent(window, context));
+  CheckError(SDL_GL_MakeCurrent(window, context.get()));
 }
 
 inline void Window::MakeCurrent(GLContext context)
@@ -7480,7 +7433,7 @@
 
 inline void GLContext::MakeCurrent(WindowParam window)
 {
-  SDL::Window::MakeCurrent(m_resource, window);
+  SDL::GL_MakeCurrent(window, m_resource);
 }
 
 /**
@@ -7496,7 +7449,7 @@
  */
 inline WindowRef GL_GetCurrentWindow()
 {
-  return CheckError(SDL_GL_GetCurrentWindow());
+  return {CheckError(SDL_GL_GetCurrentWindow())};
 }
 
 /**
