--- build/generated/SDL3pp_hidapi.h
+++ include/SDL3pp/SDL3pp_hidapi.h
@@ -9,7 +9,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryHIDAPI Category HIDAPI
+ * @defgroup CategoryHIDAPI HIDAPI
  *
  * Header file for SDL HIDAPI functions.
  *
@@ -57,26 +57,33 @@
 constexpr hid_bus_type HID_API_BUS_UNKNOWN =
   SDL_HID_API_BUS_UNKNOWN; ///< Unknown bus type
 
-constexpr hid_bus_type HID_API_BUS_USB =
-  SDL_HID_API_BUS_USB; ///< USB bus Specifications: https://usb.org/hid
+/**
+ * USB bus Specifications:
+ *
+ * - https://usb.org/hid
+ */
+constexpr hid_bus_type HID_API_BUS_USB = SDL_HID_API_BUS_USB;
 
 /**
  * Bluetooth or Bluetooth LE bus Specifications:
- * https://www.bluetooth.com/specifications/specs/human-interface-device-profile-1-1-1/
- * https://www.bluetooth.com/specifications/specs/hid-service-1-0/
- * https://www.bluetooth.com/specifications/specs/hid-over-gatt-profile-1-0/
+ *
+ * * https://www.bluetooth.com/specifications/specs/human-interface-device-profile-1-1-1/
+ * * https://www.bluetooth.com/specifications/specs/hid-service-1-0/
+ * * https://www.bluetooth.com/specifications/specs/hid-over-gatt-profile-1-0/
  */
 constexpr hid_bus_type HID_API_BUS_BLUETOOTH = SDL_HID_API_BUS_BLUETOOTH;
 
 /**
  * I2C bus Specifications:
- * https://docs.microsoft.com/previous-versions/windows/hardware/design/dn642101(v=vs.85)
+ *
+ * * https://docs.microsoft.com/previous-versions/windows/hardware/design/dn642101(v=vs.85)
  */
 constexpr hid_bus_type HID_API_BUS_I2C = SDL_HID_API_BUS_I2C;
 
 /**
  * SPI bus Specifications:
- * https://www.microsoft.com/download/details.aspx?id=103325
+ *
+ * - https://www.microsoft.com/download/details.aspx?id=103325
  */
 constexpr hid_bus_type HID_API_BUS_SPI = SDL_HID_API_BUS_SPI;
 
@@ -134,7 +141,6 @@
    * @param product_id the Product ID (PID) of the device to open.
    * @param serial_number the Serial Number of the device to open (Optionally
    *                      nullptr).
-   * @post a pointer to a HidDevice object on success.
    * @throws Error on failure.
    *
    * @since This function is available since SDL 3.2.0.
@@ -150,7 +156,6 @@
    * platform-specific path name can be used (eg: /dev/hidraw0 on Linux).
    *
    * @param path the path name of the device to open.
-   * @post a pointer to a HidDevice object on success.
    * @throws Error on failure.
    *
    * @since This function is available since SDL 3.2.0.
@@ -173,8 +178,7 @@
   /**
    * Close a HID device.
    *
-   * @returns 0 on success or a negative error code on failure; call GetError()
-   *          for more information.
+   * @throws Error on failure.
    *
    * @since This function is available since SDL 3.2.0.
    */
@@ -217,7 +221,6 @@
    *
    * @param data the data to send, including the report number as the first
    *             byte.
-   * @param length the length in bytes of the data to send.
    * @returns the actual number of bytes written and -1 on on failure; call
    *          GetError() for more information.
    *
@@ -233,10 +236,7 @@
    * reports.
    *
    * @param data a buffer to put the read data into.
-   * @param length the number of bytes to read. For devices with multiple
-   *               reports, make sure to read an extra byte for the report
-   *               number.
-   * @param milliseconds timeout in milliseconds or -1 for blocking wait.
+   * @param timeout timeout in milliseconds
    * @returns the actual number of bytes read and -1 on on failure; call
    *          GetError() for more information. If no packet was available to be
    *          read within the timeout period, this function returns 0.
@@ -253,9 +253,6 @@
    * reports.
    *
    * @param data a buffer to put the read data into.
-   * @param length the number of bytes to read. For devices with multiple
-   *               reports, make sure to read an extra byte for the report
-   *               number.
    * @returns the actual number of bytes read and -1 on failure; call GetError()
    *          for more information. If no packet was available to be read and
    *          the handle is in non-blocking mode, this function returns 0.
@@ -274,9 +271,8 @@
    *
    * Nonblocking can be turned on and off at any time.
    *
-   * @param nonblock enable or not the nonblocking reads - 1 to enable
-   *                 nonblocking - 0 to disable nonblocking.
-   * @returns 0 on success.
+   * @param nonblock enable or not the nonblocking reads - true to enable
+   *                 nonblocking - false to disable nonblocking.
    * @throws Error on failure.
    *
    * @since This function is available since SDL 3.2.0.
@@ -298,8 +294,6 @@
    *
    * @param data the data to send, including the report number as the first
    *             byte.
-   * @param length the length in bytes of the data to send, including the report
-   *               number.
    * @returns the actual number of bytes written and -1 on failure; call
    *          GetError() for more information.
    *
@@ -319,8 +313,6 @@
    *             Set the first byte of `data` to the Report ID of the report to
    *             be read, or set it to zero if your device does not use numbered
    *             reports.
-   * @param length the number of bytes to read, including an extra byte for the
-   *               report ID. The buffer can be longer than the actual report.
    * @returns the number of bytes read plus one for the report ID (which is
    *          still in the first byte), or -1 on on failure; call GetError() for
    *          more information.
@@ -341,8 +333,6 @@
    *             Set the first byte of `data` to the Report ID of the report to
    *             be read, or set it to zero if your device does not use numbered
    *             reports.
-   * @param length the number of bytes to read, including an extra byte for the
-   *               report ID. The buffer can be longer than the actual report.
    * @returns the number of bytes read plus one for the report ID (which is
    *          still in the first byte), or -1 on on failure; call GetError() for
    *          more information.
@@ -356,7 +346,6 @@
    *
    * @param string a wide string buffer to put the data into.
    * @param maxlen the length of the buffer in multiples of wchar_t.
-   * @returns 0 on success.
    * @throws Error on failure.
    *
    * @since This function is available since SDL 3.2.0.
@@ -368,7 +357,6 @@
    *
    * @param string a wide string buffer to put the data into.
    * @param maxlen the length of the buffer in multiples of wchar_t.
-   * @returns 0 on success.
    * @throws Error on failure.
    *
    * @since This function is available since SDL 3.2.0.
@@ -380,7 +368,6 @@
    *
    * @param string a wide string buffer to put the data into.
    * @param maxlen the length of the buffer in multiples of wchar_t.
-   * @returns 0 on success.
    * @throws Error on failure.
    *
    * @since This function is available since SDL 3.2.0.
@@ -393,7 +380,6 @@
    * @param string_index the index of the string to get.
    * @param string a wide string buffer to put the data into.
    * @param maxlen the length of the buffer in multiples of wchar_t.
-   * @returns 0 on success.
    * @throws Error on failure.
    *
    * @since This function is available since SDL 3.2.0.
@@ -403,9 +389,10 @@
   /**
    * Get the device info from a HID device.
    *
-   * @returns a pointer to the hid_device_info for this hid_device or nullptr on
-   *          failure; call GetError() for more information. This struct is
-   *          valid until the device is closed with HidDevice.close().
+   * @returns a pointer to the hid_device_info for this hid_device on success.
+   *          This struct is valid until the device is closed with
+   *          HidDevice.close().
+   * @throws Error on failure.
    *
    * @since This function is available since SDL 3.2.0.
    */
@@ -418,7 +405,6 @@
    * to. The recommended size for a preallocated buffer is 4096 bytes.
    *
    * @param buf the buffer to copy descriptor into.
-   * @param buf_size the size of the buffer in bytes.
    * @returns the number of bytes actually copied or -1 on failure; call
    *          GetError() for more information.
    *
@@ -510,8 +496,6 @@
  *
  * Each call to this function should have a matching call to hid_exit()
  *
- * @returns 0 on success.
- *
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
@@ -526,8 +510,6 @@
  * This function frees all of the static data associated with HIDAPI. It should
  * be called at the end of execution to avoid memory leaks.
  *
- * @returns 0 on success.
- *
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
@@ -578,9 +560,9 @@
  * @param product_id the Product ID (PID) of the types of device to open, or 0
  *                   to match any product.
  * @returns a pointer to a linked list of type hid_device_info, containing
- *          information about the HID devices attached to the system, or nullptr
- *          in the case of failure. Free this linked list by calling
- *          hid_free_enumeration().
+ *          information about the HID devices attached to the system on success.
+ *          Free this linked list by calling hid_free_enumeration().
+ * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -617,7 +599,7 @@
  * @param product_id the Product ID (PID) of the device to open.
  * @param serial_number the Serial Number of the device to open (Optionally
  *                      nullptr).
- * @returns a pointer to a HidDevice object on success.
+ * @returns a HidDevice object on success.
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
@@ -648,7 +630,7 @@
  * platform-specific path name can be used (eg: /dev/hidraw0 on Linux).
  *
  * @param path the path name of the device to open.
- * @returns a pointer to a HidDevice object on success.
+ * @returns a HidDevice object on success.
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
@@ -684,6 +666,13 @@
   return SDL::hid_get_properties(get());
 }
 
+/**
+ * HIDAPI properties.
+ *
+ * @since This namespace is available since SDL 3.4.0.
+ *
+ * @sa Hidapi.GetProperties
+ */
 namespace prop::Hidapi {
 
 constexpr auto LIBUSB_DEVICE_HANDLE_POINTER =
@@ -710,7 +699,6 @@
  *
  * @param dev a device handle returned from hid_open().
  * @param data the data to send, including the report number as the first byte.
- * @param length the length in bytes of the data to send.
  * @returns the actual number of bytes written and -1 on on failure; call
  *          GetError() for more information.
  *
@@ -718,7 +706,7 @@
  */
 inline int hid_write(HidDeviceRef dev, SourceBytes data)
 {
-  return SDL_hid_write(dev, data.data(), data.size_bytes());
+  return SDL_hid_write(dev, data.data_as<Uint8>(), data.size_bytes());
 }
 
 inline int HidDevice::write(SourceBytes data)
@@ -735,9 +723,7 @@
  *
  * @param dev a device handle returned from hid_open().
  * @param data a buffer to put the read data into.
- * @param length the number of bytes to read. For devices with multiple reports,
- *               make sure to read an extra byte for the report number.
- * @param milliseconds timeout in milliseconds or -1 for blocking wait.
+ * @param timeout timeout in milliseconds
  * @returns the actual number of bytes read and -1 on on failure; call
  *          GetError() for more information. If no packet was available to be
  *          read within the timeout period, this function returns 0.
@@ -748,7 +734,8 @@
                             TargetBytes data,
                             Milliseconds timeout)
 {
-  return SDL_hid_read_timeout(dev, data.data(), data.size_bytes(), timeout);
+  return SDL_hid_read_timeout(
+    dev, data.data_as<Uint8>(), data.size_bytes(), narrowS32(timeout.count()));
 }
 
 inline int HidDevice::read_timeout(TargetBytes data, Milliseconds timeout)
@@ -765,8 +752,6 @@
  *
  * @param dev a device handle returned from hid_open().
  * @param data a buffer to put the read data into.
- * @param length the number of bytes to read. For devices with multiple reports,
- *               make sure to read an extra byte for the report number.
  * @returns the actual number of bytes read and -1 on failure; call GetError()
  *          for more information. If no packet was available to be read and the
  *          handle is in non-blocking mode, this function returns 0.
@@ -775,7 +760,7 @@
  */
 inline int hid_read(HidDeviceRef dev, TargetBytes data)
 {
-  return SDL_hid_read(dev, data.data(), data.size_bytes());
+  return SDL_hid_read(dev, data.data_as<Uint8>(), data.size_bytes());
 }
 
 inline int HidDevice::read(TargetBytes data)
@@ -793,9 +778,8 @@
  * Nonblocking can be turned on and off at any time.
  *
  * @param dev a device handle returned from hid_open().
- * @param nonblock enable or not the nonblocking reads - 1 to enable nonblocking
- *                 - 0 to disable nonblocking.
- * @returns 0 on success.
+ * @param nonblock enable or not the nonblocking reads - true to enable
+ *                 nonblocking - false to disable nonblocking.
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
@@ -825,8 +809,6 @@
  *
  * @param dev a device handle returned from hid_open().
  * @param data the data to send, including the report number as the first byte.
- * @param length the length in bytes of the data to send, including the report
- *               number.
  * @returns the actual number of bytes written and -1 on failure; call
  *          GetError() for more information.
  *
@@ -834,7 +816,8 @@
  */
 inline int hid_send_feature_report(HidDeviceRef dev, SourceBytes data)
 {
-  return SDL_hid_send_feature_report(dev, data.data(), data.size_bytes());
+  return SDL_hid_send_feature_report(
+    dev, data.data_as<Uint8>(), data.size_bytes());
 }
 
 inline int HidDevice::send_feature_report(SourceBytes data)
@@ -855,8 +838,6 @@
  *             the first byte of `data` to the Report ID of the report to be
  *             read, or set it to zero if your device does not use numbered
  *             reports.
- * @param length the number of bytes to read, including an extra byte for the
- *               report ID. The buffer can be longer than the actual report.
  * @returns the number of bytes read plus one for the report ID (which is still
  *          in the first byte), or -1 on on failure; call GetError() for more
  *          information.
@@ -865,7 +846,8 @@
  */
 inline int hid_get_feature_report(HidDeviceRef dev, TargetBytes data)
 {
-  return SDL_hid_get_feature_report(dev, data.data(), data.size_bytes());
+  return SDL_hid_get_feature_report(
+    dev, data.data_as<Uint8>(), data.size_bytes());
 }
 
 inline int HidDevice::get_feature_report(TargetBytes data)
@@ -886,8 +868,6 @@
  *             the first byte of `data` to the Report ID of the report to be
  *             read, or set it to zero if your device does not use numbered
  *             reports.
- * @param length the number of bytes to read, including an extra byte for the
- *               report ID. The buffer can be longer than the actual report.
  * @returns the number of bytes read plus one for the report ID (which is still
  *          in the first byte), or -1 on on failure; call GetError() for more
  *          information.
@@ -896,7 +876,8 @@
  */
 inline int hid_get_input_report(HidDeviceRef dev, TargetBytes data)
 {
-  return SDL_hid_get_input_report(dev, data.data(), data.size_bytes());
+  return SDL_hid_get_input_report(
+    dev, data.data_as<Uint8>(), data.size_bytes());
 }
 
 inline int HidDevice::get_input_report(TargetBytes data)
@@ -908,14 +889,13 @@
  * Close a HID device.
  *
  * @param dev a device handle returned from hid_open().
- * @returns 0 on success.
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
  */
-inline int hid_close(HidDeviceRaw dev)
+inline void hid_close(HidDeviceRaw dev)
 {
-  return CheckError(SDL_hid_close(dev));
+  CheckErrorIfNot(SDL_hid_close(dev), 0);
 }
 
 inline void HidDevice::close() { hid_close(release()); }
@@ -926,7 +906,6 @@
  * @param dev a device handle returned from hid_open().
  * @param string a wide string buffer to put the data into.
  * @param maxlen the length of the buffer in multiples of wchar_t.
- * @returns 0 on success.
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
@@ -949,7 +928,6 @@
  * @param dev a device handle returned from hid_open().
  * @param string a wide string buffer to put the data into.
  * @param maxlen the length of the buffer in multiples of wchar_t.
- * @returns 0 on success.
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
@@ -972,7 +950,6 @@
  * @param dev a device handle returned from hid_open().
  * @param string a wide string buffer to put the data into.
  * @param maxlen the length of the buffer in multiples of wchar_t.
- * @returns 0 on success.
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
@@ -996,7 +973,6 @@
  * @param string_index the index of the string to get.
  * @param string a wide string buffer to put the data into.
  * @param maxlen the length of the buffer in multiples of wchar_t.
- * @returns 0 on success.
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
@@ -1021,9 +997,10 @@
  * Get the device info from a HID device.
  *
  * @param dev a device handle returned from hid_open().
- * @returns a pointer to the hid_device_info for this hid_device or nullptr on
- *          failure; call GetError() for more information. This struct is valid
- *          until the device is closed with HidDevice.close().
+ * @returns a pointer to the hid_device_info for this hid_device on success.
+ *          This struct is valid until the device is closed with
+ *          HidDevice.close().
+ * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
  */
@@ -1045,7 +1022,6 @@
  *
  * @param dev a device handle returned from hid_open().
  * @param buf the buffer to copy descriptor into.
- * @param buf_size the size of the buffer in bytes.
  * @returns the number of bytes actually copied or -1 on failure; call
  *          GetError() for more information.
  *
@@ -1053,7 +1029,8 @@
  */
 inline int hid_get_report_descriptor(HidDeviceRef dev, TargetBytes buf)
 {
-  return SDL_hid_get_report_descriptor(dev, buf.data(), buf.size_bytes());
+  return SDL_hid_get_report_descriptor(
+    dev, buf.data_as<Uint8>(), buf.size_bytes());
 }
 
 inline int HidDevice::get_report_descriptor(TargetBytes buf)
