--- build/generated/SDL3pp_tray.h
+++ include/SDL3pp/SDL3pp_tray.h
@@ -8,7 +8,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryTray Category Tray
+ * @defgroup CategoryTray System Tray
  *
  * SDL offers a way to add items to the "system tray" (more correctly called the
  * "notification area" on Windows). On platforms that offer this concept, an SDL
@@ -388,8 +388,6 @@
   /**
    * Returns a list of entries in the menu, in order.
    *
-   * @param count An optional pointer to obtain the number of entries in the
-   *              menu.
    * @returns a nullptr-terminated list of entries within the given menu. The
    *          pointer becomes invalid when any function that inserts or deletes
    *          entries in the menu is called.
@@ -431,6 +429,30 @@
    */
   TrayEntry InsertEntry(int pos, StringParam label, TrayEntryFlags flags);
 
+  /**
+   * Appends a tray entry.
+   *
+   * If label is nullptr, the entry will be a separator. Many functions won't
+   * work for an entry that is a separator.
+   *
+   * An entry does not need to be destroyed; it will be destroyed with the tray.
+   *
+   * @param label the text to be displayed on the entry, in UTF-8 encoding, or
+   *              nullptr for a separator.
+   * @param flags a combination of flags, some of which are mandatory.
+   * @returns the newly created entry, or nullptr if pos is out of bounds.
+   *
+   * @threadsafety This function should be called on the thread that created the
+   *               tray.
+   *
+   * @since This function is available since SDL 3.2.0.
+   *
+   * @sa TrayMenu.InsertEntry
+   * @sa TrayEntryFlags
+   * @sa TrayMenu.GetEntries
+   * @sa TrayEntry.Remove
+   * @sa TrayEntry.GetParent
+   */
   TrayEntry AppendEntry(StringParam label, TrayEntryFlags flags);
 
   /**
@@ -512,7 +534,7 @@
   TrayEntry(TrayMenu menu, int pos, StringParam label, TrayEntryFlags flags);
 
   /**
-   * Insert a tray entry at a given position.
+   * Appends a tray entry.
    *
    * If label is nullptr, the entry will be a separator. Many functions won't
    * work for an entry that is a separator.
@@ -520,8 +542,6 @@
    * An entry does not need to be destroyed; it will be destroyed with the tray.
    *
    * @param menu the menu to append the entry to.
-   * @param pos the desired position for the new entry. Entries at or following
-   *            this place will be moved. If pos is -1, the entry is appended.
    * @param label the text to be displayed on the entry, in UTF-8 encoding, or
    *              nullptr for a separator.
    * @param flags a combination of flags, some of which are mandatory.
@@ -532,6 +552,7 @@
    *
    * @since This function is available since SDL 3.2.0.
    *
+   * @sa TrayMenu.InsertEntry
    * @sa TrayEntryFlags
    * @sa TrayMenu.GetEntries
    * @sa TrayEntry.Remove
@@ -729,8 +750,6 @@
    * Sets a callback to be invoked when the entry is selected.
    *
    * @param callback a callback to be invoked when the entry is selected.
-   * @param userdata an optional pointer to pass extra data to the callback when
-   *                 it will be invoked.
    *
    * @threadsafety This function should be called on the thread that created the
    *               tray.
@@ -986,7 +1005,6 @@
  * Returns a list of entries in the menu, in order.
  *
  * @param menu The menu to get entries from.
- * @param count An optional pointer to obtain the number of entries in the menu.
  * @returns a nullptr-terminated list of entries within the given menu. The
  *          pointer becomes invalid when any function that inserts or deletes
  *          entries in the menu is called.
@@ -1001,7 +1019,10 @@
  */
 inline std::span<TrayEntry> GetTrayEntries(TrayMenu menu)
 {
-  return SDL_GetTrayEntries(menu);
+  int count;
+  auto entries = SDL_GetTrayEntries(menu, &count);
+  return std::span<TrayEntry>{reinterpret_cast<TrayEntry*>(entries),
+                              size_t(count)};
 }
 
 inline std::span<TrayEntry> TrayMenu::GetEntries()
@@ -1078,20 +1099,19 @@
 inline TrayEntry::TrayEntry(TrayMenuRaw menu,
                             StringParam label,
                             TrayEntryFlags flags)
+  : TrayEntry(menu, -1, std::move(label), flags)
 {
 }
 
 /**
- * Insert a tray entry at a given position.
+ * Appends a tray entry.
  *
- * If label is nullptr, the entry will be a separator. Many functions won't work
- * for an entry that is a separator.
+ * If label is nullptr, the entry will be a separator. Many functions won't
+ * work for an entry that is a separator.
  *
  * An entry does not need to be destroyed; it will be destroyed with the tray.
  *
  * @param menu the menu to append the entry to.
- * @param pos the desired position for the new entry. Entries at or following
- *            this place will be moved. If pos is -1, the entry is appended.
  * @param label the text to be displayed on the entry, in UTF-8 encoding, or
  *              nullptr for a separator.
  * @param flags a combination of flags, some of which are mandatory.
@@ -1102,6 +1122,7 @@
  *
  * @since This function is available since SDL 3.2.0.
  *
+ * @sa TrayMenu.InsertEntry
  * @sa TrayEntryFlags
  * @sa TrayMenu.GetEntries
  * @sa TrayEntry.Remove
@@ -1308,8 +1329,6 @@
  *
  * @param entry the entry to be updated.
  * @param callback a callback to be invoked when the entry is selected.
- * @param userdata an optional pointer to pass extra data to the callback when
- *                 it will be invoked.
  *
  * @threadsafety This function should be called on the thread that created the
  *               tray.
@@ -1321,7 +1340,7 @@
  */
 inline void SetTrayEntryCallback(TrayEntry entry, TrayCB callback)
 {
-  static_assert(false, "Not implemented");
+  SetTrayEntryCallback(entry, callback.wrapper, callback.data);
 }
 
 inline void TrayEntry::SetCallback(TrayCallback callback, void* userdata)
