--- src/generated/SDL3pp_version.h
+++ include/SDL3pp/SDL3pp_version.h
@@ -6,7 +6,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryVersion Category Version
+ * @defgroup CategoryVersion Querying SDL Version
  *
  * Functionality to query the current SDL version, both as headers the app was
  * compiled against, and a library the app is linked to.
@@ -14,6 +14,8 @@
  * @{
  */
 
+#ifdef SDL3PP_DOC
+
 /**
  * The current major version of SDL headers.
  *
@@ -105,6 +107,8 @@
  */
 #define SDL_VERSION_ATLEAST(X, Y, Z) (SDL_VERSION >= SDL_VERSIONNUM(X, Y, Z))
 
+#endif // SDL3PP_DOC
+
 /**
  * Get the version of SDL that is linked against your program.
  *
@@ -165,6 +169,18 @@
 
 /// @}
 
+// Check if SDL_image is available
+#if !defined(SDL3PP_DISABLE_IMAGE) && !defined(SDL3PP_ENABLE_IMAGE) &&         \
+  __has_include(<SDL3_image/SDL_image.h>)
+#define SDL3PP_ENABLE_IMAGE
+#endif
+
+// Check if SDL_ttf is available
+#if !defined(SDL3PP_DISABLE_TTF) && !defined(SDL3PP_ENABLE_TTF) &&             \
+  __has_include(<SDL3_ttf/SDL_ttf.h>)
+#define SDL3PP_ENABLE_TTF
+#endif
+
 } // namespace SDL
 
 #endif /* SDL3PP_VERSION_H_ */
