--- src/generated/SDL3pp_iostream.h
+++ include/SDL3pp/SDL3pp_iostream.h
@@ -8,7 +8,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryIOStream Category IOStream
+ * @defgroup CategoryIOStream I/O Streams
  *
  * SDL provides an abstract interface for reading and writing data streams. It
  * offers implementations for files, memory, etc, and the app can provide their
@@ -178,7 +178,7 @@
    *   already exists its content is erased and the file is treated as a new
    *   empty file.
    * - "wx": Create an empty file for writing. If a file with the same name
-   *   already exists, the call fails.
+   *   already exists, the call fails. (Supported since SDL 3.4.0)
    * - "a": Append to a file. Writing operations append data at the end of the
    *   file. The file is created if it does not exist.
    * - "r+": Open a file for update both reading and writing. The file must
@@ -187,7 +187,7 @@
    *   the same name already exists its content is erased and the file is
    *   treated as a new empty file.
    * - "w+x": Create an empty file for both reading and writing. If a file with
-   *   the same name already exists, the call fails.
+   *   the same name already exists, the call fails. (Supported since SDL 3.4.0)
    * - "a+": Open a file for reading and appending. All writing operations are
    *   performed at the end of the file, protecting the previous content to be
    *   overwritten. You can reposition (fseek, rewind) the internal pointer to
@@ -279,10 +279,9 @@
    *   and called with the passed `mem` pointer when closing the stream. By
    *   default it is unset, i.e., the memory will not be freed.
    *
-   * @param mem a pointer to a buffer to feed an IOStream stream.
-   * @param size the buffer size, in bytes.
-   * @returns a pointer to a new IOStream structure or nullptr on failure; call
-   *          GetError() for more information.
+   * @param mem a buffer to feed an IOStream stream.
+   * @returns a valid IOStream on success.
+   * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
@@ -328,10 +327,9 @@
    *   and called with the passed `mem` pointer when closing the stream. By
    *   default it is unset, i.e., the memory will not be freed.
    *
-   * @param mem a pointer to a read-only buffer to feed an IOStream stream.
-   * @param size the buffer size, in bytes.
-   * @returns a pointer to a new IOStream structure or nullptr on failure; call
-   *          GetError() for more information.
+   * @param mem a read-only buffer to feed an IOStreamRef stream.
+   * @returns a valid IOStream on success.
+   * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
@@ -359,8 +357,8 @@
    * - `prop::IOStream.DYNAMIC_CHUNKSIZE_NUMBER`: memory will be allocated in
    *   multiples of this size, defaulting to 1024.
    *
-   * @returns a pointer to a new IOStream structure or nullptr on failure; call
-   *          GetError() for more information.
+   * @returns a valid IOStream on success.
+   * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
@@ -389,7 +387,7 @@
    * @param iface the interface that implements this IOStream, initialized using
    *              InitInterface().
    * @param userdata the pointer that will be passed to the interface functions.
-   * @returns a pointer to the allocated memory on success.
+   * @returns a valid stream on success.
    * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread.
@@ -460,8 +458,7 @@
    * and makes the system and your app operate less efficiently, so do so
    * sparingly.
    *
-   * @returns true on success or false on failure; call GetError() for more
-   *          information.
+   * @throws Error on failure.
    *
    * @threadsafety Do not use the same IOStream from two threads at once.
    *
@@ -559,9 +556,43 @@
    */
   Sint64 Tell() const;
 
+  /**
+   * Read from a data source.
+   *
+   * This function reads up `size` bytes from the data source to the area
+   * pointed at by `ptr`. This function may read less bytes than requested.
+   *
+   * This function will return zero when the data stream is completely read, and
+   * IOStreamRef.GetStatus() will return IO_STATUS_EOF. If zero is returned and
+   * the stream is not at EOF, IOStreamRef.GetStatus() will return a different
+   * error value and GetError() will offer a human-readable message.
+   *
+   * @param size the number of bytes to read from the data source.
+   * @returns the bytes, or empty string on end of file or other failure;
+   *          call GetError() for more information.
+   *
+   * @threadsafety This function is not thread safe.
+   *
+   * @since This function is available since SDL 3.2.0.
+   *
+   * @sa IOStream.Write
+   * @sa IOStream.GetStatus
+   */
   std::string Read(size_t size = -1)
   {
-    static_assert(false, "Not implemented");
+    Sint64 pos = Tell();
+    auto curSize = SDL_GetIOSize(get());
+    if ((curSize < 0 || pos < 0)) {
+      if (size == size_t(-1)) return {};
+    } else if (curSize - pos <= 0) {
+      return {};
+    } else if (curSize - pos < size) {
+      size = curSize - pos;
+    }
+    std::string result(size, 0);
+    auto actualSize = Read(result);
+    if (actualSize < size) result.resize(actualSize);
+    return result;
   }
 
   /**
@@ -579,8 +610,7 @@
    * without accessing the stream, so the stream status (EOF, err, etc) will not
    * change.
    *
-   * @param ptr a pointer to a buffer to read data into.
-   * @param size the number of bytes to read from the data source.
+   * @param buf a pointer to a buffer to read data into.
    * @returns the number of bytes read, or 0 on end of file or other failure;
    *          call GetError() for more information.
    *
@@ -611,8 +641,7 @@
    * without accessing the stream, so the stream status (EOF, err, etc) will not
    * change.
    *
-   * @param ptr a pointer to a buffer containing data to write.
-   * @param size the number of bytes to write.
+   * @param buf the bytes to write to
    * @returns the number of bytes written, which will be less than `size` on
    *          failure; call GetError() for more information.
    *
@@ -628,21 +657,37 @@
    */
   size_t Write(SourceBytes buf);
 
-  /// @cat formatted-string
+  /**
+   * Prints formatted string.
+   *
+   * @param fmt a std::format like format string
+   * @param args the arguments to be formatted
+   * @cat formatted-string
+   */
   size_t print(std::string_view fmt, auto... args)
   {
-    static_assert(false, "Not implemented");
+    return Write(std::vformat(fmt, std::make_format_args(args...)));
   }
 
-  /// @cat formatted-string
+  /**
+   * Prints formatted string.
+   *
+   * @param fmt a std::format like format string
+   * @param args the arguments to be formatted
+   * @cat formatted-string
+   */
   size_t println(std::string_view fmt, auto... args)
   {
-    static_assert(false, "Not implemented");
+    std::string result =
+      std::vformat(fmt, std::make_format_args(args...)) + "\n";
+    return Write(result);
   }
 
   /**
    * Print to an IOStream data stream.
    *
+   * @warning this is not typesafe! Prefer using print() and println()
+   *
    * This function does formatted printing to the stream.
    *
    * @param fmt a printf() style format string.
@@ -660,12 +705,21 @@
    */
   size_t printf(SDL_PRINTF_FORMAT_STRING const char* fmt, ...)
   {
-    return SDL_IOprintf(m_resource, fmt, ...);
+    va_list ap;
+    size_t result;
+
+    va_start(ap, fmt);
+    result = vprintf(fmt, ap);
+    va_end(ap);
+
+    return result;
   }
 
   /**
    * Print to an IOStream data stream.
    *
+   * @warning this is not typesafe! Prefer using print() and println()
+   *
    * This function does formatted printing to the stream.
    *
    * @param fmt a printf() style format string.
@@ -704,17 +758,10 @@
    * Load all the data from an SDL data stream.
    *
    * The data is allocated with a zero byte at the end (null terminated) for
-   * convenience. This extra byte is not included in the value reported via
-   * `datasize`.
+   * convenience.
    *
-   * The data should be freed with free().
-   *
-   * @param datasize a pointer filled in with the number of bytes read, may be
-   *                 nullptr.
-   * @param closeio if true, calls IOStream.Close() on `src` before returning,
-   *                even in the case of an error.
-   * @returns the data or nullptr on failure; call GetError() for more
-   *          information.
+   * @returns the data in bytes
+   * @throws Error on failure.
    *
    * @threadsafety Do not use the same IOStream from two threads at once.
    *
@@ -725,10 +772,29 @@
    */
   StringResult LoadFile();
 
+  /**
+   * Load all the data from an SDL data stream.
+   *
+   * The data is allocated with a zero byte at the end (null terminated) for
+   * convenience. This extra byte is not included in the value reported on
+   * the returned string.
+   *
+   * @returns the data in bytes
+   * @throws Error on failure.
+   *
+   * @threadsafety This function is not thread safe.
+   *
+   * @since This function is available since SDL 3.2.0.
+   *
+   * @sa LoadFile
+   * @sa IOStream.SaveFile
+   */
   template<class T>
   OwnArray<T> LoadFileAs()
   {
-    static_assert(false, "Not implemented");
+    size_t datasize = 0;
+    auto data = static_cast<T*>(SDL_LoadFile_IO(get(), &datasize, false));
+    return OwnArray<T>{CheckError(data), datasize / sizeof(T)};
   }
 
   /**
@@ -736,9 +802,6 @@
    *
    * @param data the data to be written. If datasize is 0, may be nullptr or a
    *             invalid pointer.
-   * @param datasize the number of bytes to be written.
-   * @param closeio if true, calls IOStream.Close() on `src` before returning,
-   *                even in the case of an error.
    * @throws Error on failure.
    *
    * @threadsafety Do not use the same IOStream from two threads at once.
@@ -758,8 +821,7 @@
    * and the stream is not at EOF, IOStream.GetStatus() will return a different
    * error value and GetError() will offer a human-readable message.
    *
-   * @param src the IOStream to read from.
-   * @param value a pointer filled in with the data read.
+   * @returns the data read on success.
    * @throws Error on failure.
    *
    * @threadsafety Do not use the same IOStream from two threads at once.
@@ -776,8 +838,7 @@
    * and the stream is not at EOF, IOStream.GetStatus() will return a different
    * error value and GetError() will offer a human-readable message.
    *
-   * @param src the IOStream to read from.
-   * @param value a pointer filled in with the data read.
+   * @returns the data read on success.
    * @throws Error on failure.
    *
    * @threadsafety Do not use the same IOStream from two threads at once.
@@ -798,8 +859,7 @@
    * and the stream is not at EOF, IOStream.GetStatus() will return a different
    * error value and GetError() will offer a human-readable message.
    *
-   * @param src the stream from which to read data.
-   * @param value a pointer filled in with the data read.
+   * @returns the data read on success.
    * @throws Error on failure.
    *
    * @threadsafety Do not use the same IOStream from two threads at once.
@@ -820,8 +880,7 @@
    * and the stream is not at EOF, IOStream.GetStatus() will return a different
    * error value and GetError() will offer a human-readable message.
    *
-   * @param src the stream from which to read data.
-   * @param value a pointer filled in with the data read.
+   * @returns the data read on success.
    * @throws Error on failure.
    *
    * @threadsafety Do not use the same IOStream from two threads at once.
@@ -842,8 +901,7 @@
    * and the stream is not at EOF, IOStream.GetStatus() will return a different
    * error value and GetError() will offer a human-readable message.
    *
-   * @param src the stream from which to read data.
-   * @param value a pointer filled in with the data read.
+   * @returns the data read on success.
    * @throws Error on failure.
    *
    * @threadsafety Do not use the same IOStream from two threads at once.
@@ -864,8 +922,7 @@
    * and the stream is not at EOF, IOStream.GetStatus() will return a different
    * error value and GetError() will offer a human-readable message.
    *
-   * @param src the stream from which to read data.
-   * @param value a pointer filled in with the data read.
+   * @returns the data read on success.
    * @throws Error on failure.
    *
    * @threadsafety Do not use the same IOStream from two threads at once.
@@ -886,8 +943,7 @@
    * and the stream is not at EOF, IOStream.GetStatus() will return a different
    * error value and GetError() will offer a human-readable message.
    *
-   * @param src the stream from which to read data.
-   * @param value a pointer filled in with the data read.
+   * @returns the data read on success.
    * @throws Error on failure.
    *
    * @threadsafety Do not use the same IOStream from two threads at once.
@@ -908,8 +964,7 @@
    * and the stream is not at EOF, IOStream.GetStatus() will return a different
    * error value and GetError() will offer a human-readable message.
    *
-   * @param src the stream from which to read data.
-   * @param value a pointer filled in with the data read.
+   * @returns the data read on success.
    * @throws Error on failure.
    *
    * @threadsafety Do not use the same IOStream from two threads at once.
@@ -930,8 +985,7 @@
    * and the stream is not at EOF, IOStream.GetStatus() will return a different
    * error value and GetError() will offer a human-readable message.
    *
-   * @param src the stream from which to read data.
-   * @param value a pointer filled in with the data read.
+   * @returns the data read on success.
    * @throws Error on failure.
    *
    * @threadsafety Do not use the same IOStream from two threads at once.
@@ -952,8 +1006,7 @@
    * and the stream is not at EOF, IOStream.GetStatus() will return a different
    * error value and GetError() will offer a human-readable message.
    *
-   * @param src the stream from which to read data.
-   * @param value a pointer filled in with the data read.
+   * @returns the data read on success.
    * @throws Error on failure.
    *
    * @threadsafety Do not use the same IOStream from two threads at once.
@@ -974,8 +1027,7 @@
    * and the stream is not at EOF, IOStream.GetStatus() will return a different
    * error value and GetError() will offer a human-readable message.
    *
-   * @param src the stream from which to read data.
-   * @param value a pointer filled in with the data read.
+   * @returns the data read on success.
    * @throws Error on failure.
    *
    * @threadsafety Do not use the same IOStream from two threads at once.
@@ -996,8 +1048,7 @@
    * and the stream is not at EOF, IOStream.GetStatus() will return a different
    * error value and GetError() will offer a human-readable message.
    *
-   * @param src the stream from which to read data.
-   * @param value a pointer filled in with the data read.
+   * @returns the data read on success.
    * @throws Error on failure.
    *
    * @threadsafety Do not use the same IOStream from two threads at once.
@@ -1018,8 +1069,7 @@
    * and the stream is not at EOF, IOStream.GetStatus() will return a different
    * error value and GetError() will offer a human-readable message.
    *
-   * @param src the stream from which to read data.
-   * @param value a pointer filled in with the data read.
+   * @returns the data read on success.
    * @throws Error on failure.
    *
    * @threadsafety Do not use the same IOStream from two threads at once.
@@ -1040,8 +1090,7 @@
    * and the stream is not at EOF, IOStream.GetStatus() will return a different
    * error value and GetError() will offer a human-readable message.
    *
-   * @param src the stream from which to read data.
-   * @param value a pointer filled in with the data read.
+   * @returns the data read on success.
    * @throws Error on failure.
    *
    * @threadsafety Do not use the same IOStream from two threads at once.
@@ -1050,68 +1099,332 @@
    */
   Sint64 ReadS64BE();
 
-  std::optional<Uint8> TryReadU8() { static_assert(false, "Not implemented"); }
+  /**
+   * Use this function to read a byte from an IOStreamRef.
+   *
+   * This function will return false when the data stream is completely read,
+   * and IOStreamRef.GetStatus() will return IO_STATUS_EOF. If false is
+   * returned and the stream is not at EOF, IOStreamRef.GetStatus() will return
+   * a different error value and GetError() will offer a human-readable message.
+   *
+   * @returns the data read on success, std::nullopt on failure.
+   *
+   * @threadsafety This function is not thread safe.
+   *
+   * @since This function is available since SDL 3.2.0.
+   */
+  std::optional<Uint8> TryReadU8()
+  {
+    if (Uint8 value; SDL_ReadU8(get(), &value)) return value;
+    return {};
+  }
 
-  std::optional<Sint8> TryReadS8() { static_assert(false, "Not implemented"); }
+  /**
+   * Use this function to read a byte from an IOStreamRef.
+   *
+   * This function will return false when the data stream is completely read,
+   * and IOStreamRef.GetStatus() will return IO_STATUS_EOF. If false is
+   * returned and the stream is not at EOF, IOStreamRef.GetStatus() will return
+   * a different error value and GetError() will offer a human-readable message.
+   *
+   * @returns the data read on success, std::nullopt on failure.
+   *
+   * @threadsafety This function is not thread safe.
+   *
+   * @since This function is available since SDL 3.2.0.
+   */
+  std::optional<Sint8> TryReadS8()
+  {
+    if (Sint8 value; SDL_ReadS8(get(), &value)) return value;
+    return {};
+  }
 
+  /**
+   * Use this function to read 16 bits of little-endian data from an
+   * IOStreamRef and return in native format.
+   *
+   * SDL byteswaps the data only if necessary, so the data returned will be in
+   * the native byte order.
+   *
+   * This function will return false when the data stream is completely read,
+   * and IOStreamRef.GetStatus() will return IO_STATUS_EOF. If false is
+   * returned and the stream is not at EOF, IOStreamRef.GetStatus() will return
+   * a different error value and GetError() will offer a human-readable message.
+   *
+   * @returns the data read on success, std::nullopt on failure.
+   *
+   * @threadsafety This function is not thread safe.
+   *
+   * @since This function is available since SDL 3.2.0.
+   */
   std::optional<Uint16> TryReadU16LE()
   {
-    static_assert(false, "Not implemented");
+    if (Uint16 value; SDL_ReadU16LE(get(), &value)) return value;
+    return {};
   }
 
+  /**
+   * Use this function to read 16 bits of little-endian data from an
+   * IOStreamRef and return in native format.
+   *
+   * SDL byteswaps the data only if necessary, so the data returned will be in
+   * the native byte order.
+   *
+   * This function will return false when the data stream is completely read,
+   * and IOStreamRef.GetStatus() will return IO_STATUS_EOF. If false is
+   * returned and the stream is not at EOF, IOStreamRef.GetStatus() will return
+   * a different error value and GetError() will offer a human-readable message.
+   *
+   * @returns the data read on success, std::nullopt on failure.
+   *
+   * @threadsafety This function is not thread safe.
+   *
+   * @since This function is available since SDL 3.2.0.
+   */
   std::optional<Sint16> TryReadS16LE()
   {
-    static_assert(false, "Not implemented");
+    if (Sint16 value; SDL_ReadS16LE(get(), &value)) return value;
+    return {};
   }
 
+  /**
+   * Use this function to read 16 bits of big-endian data from an IOStreamRef
+   * and return in native format.
+   *
+   * SDL byteswaps the data only if necessary, so the data returned will be in
+   * the native byte order.
+   *
+   * This function will return false when the data stream is completely read,
+   * and IOStreamRef.GetStatus() will return IO_STATUS_EOF. If false is
+   * returned and the stream is not at EOF, IOStreamRef.GetStatus() will return
+   * a different error value and GetError() will offer a human-readable message.
+   *
+   * @returns the data read on success, std::nullopt on failure.
+   *
+   * @threadsafety This function is not thread safe.
+   *
+   * @since This function is available since SDL 3.2.0.
+   */
   std::optional<Uint16> TryReadU16BE()
   {
-    static_assert(false, "Not implemented");
+    if (Uint16 value; SDL_ReadU16BE(get(), &value)) return value;
+    return {};
   }
 
+  /**
+   * Use this function to read 16 bits of big-endian data from an IOStreamRef
+   * and return in native format.
+   *
+   * SDL byteswaps the data only if necessary, so the data returned will be in
+   * the native byte order.
+   *
+   * This function will return false when the data stream is completely read,
+   * and IOStreamRef.GetStatus() will return IO_STATUS_EOF. If false is
+   * returned and the stream is not at EOF, IOStreamRef.GetStatus() will return
+   * a different error value and GetError() will offer a human-readable message.
+   *
+   * @returns the data read on success, std::nullopt on failure.
+   *
+   * @threadsafety This function is not thread safe.
+   *
+   * @since This function is available since SDL 3.2.0.
+   */
   std::optional<Sint16> TryReadS16BE()
   {
-    static_assert(false, "Not implemented");
+    if (Sint16 value; SDL_ReadS16BE(get(), &value)) return value;
+    return {};
   }
 
+  /**
+   * Use this function to read 32 bits of little-endian data from an
+   * IOStreamRef and return in native format.
+   *
+   * SDL byteswaps the data only if necessary, so the data returned will be in
+   * the native byte order.
+   *
+   * This function will return false when the data stream is completely read,
+   * and IOStreamRef.GetStatus() will return IO_STATUS_EOF. If false is
+   * returned and the stream is not at EOF, IOStreamRef.GetStatus() will return
+   * a different error value and GetError() will offer a human-readable message.
+   *
+   * @returns the data read on success, std::nullopt on failure.
+   *
+   * @threadsafety This function is not thread safe.
+   *
+   * @since This function is available since SDL 3.2.0.
+   */
   std::optional<Uint32> TryReadU32LE()
   {
-    static_assert(false, "Not implemented");
+    if (Uint32 value; SDL_ReadU32LE(get(), &value)) return value;
+    return {};
   }
 
+  /**
+   * Use this function to read 32 bits of little-endian data from an
+   * IOStreamRef and return in native format.
+   *
+   * SDL byteswaps the data only if necessary, so the data returned will be in
+   * the native byte order.
+   *
+   * This function will return false when the data stream is completely read,
+   * and IOStreamRef.GetStatus() will return IO_STATUS_EOF. If false is
+   * returned and the stream is not at EOF, IOStreamRef.GetStatus() will return
+   * a different error value and GetError() will offer a human-readable message.
+   *
+   * @returns the data read on success, std::nullopt on failure.
+   *
+   * @threadsafety This function is not thread safe.
+   *
+   * @since This function is available since SDL 3.2.0.
+   */
   std::optional<Sint32> TryReadS32LE()
   {
-    static_assert(false, "Not implemented");
+    if (Sint32 value; SDL_ReadS32LE(get(), &value)) return value;
+    return {};
   }
 
+  /**
+   * Use this function to read 32 bits of big-endian data from an IOStreamRef
+   * and return in native format.
+   *
+   * SDL byteswaps the data only if necessary, so the data returned will be in
+   * the native byte order.
+   *
+   * This function will return false when the data stream is completely read,
+   * and IOStreamRef.GetStatus() will return IO_STATUS_EOF. If false is
+   * returned and the stream is not at EOF, IOStreamRef.GetStatus() will return
+   * a different error value and GetError() will offer a human-readable message.
+   *
+   * @returns the data read on success, std::nullopt on failure.
+   *
+   * @threadsafety This function is not thread safe.
+   *
+   * @since This function is available since SDL 3.2.0.
+   */
   std::optional<Uint32> TryReadU32BE()
   {
-    static_assert(false, "Not implemented");
+    if (Uint32 value; SDL_ReadU32BE(get(), &value)) return value;
+    return {};
   }
 
+  /**
+   * Use this function to read 32 bits of big-endian data from an IOStreamRef
+   * and return in native format.
+   *
+   * SDL byteswaps the data only if necessary, so the data returned will be in
+   * the native byte order.
+   *
+   * This function will return false when the data stream is completely read,
+   * and IOStreamRef.GetStatus() will return IO_STATUS_EOF. If false is
+   * returned and the stream is not at EOF, IOStreamRef.GetStatus() will return
+   * a different error value and GetError() will offer a human-readable message.
+   *
+   * @returns the data read on success, std::nullopt on failure.
+   *
+   * @threadsafety This function is not thread safe.
+   *
+   * @since This function is available since SDL 3.2.0.
+   */
   std::optional<Sint32> TryReadS32BE()
   {
-    static_assert(false, "Not implemented");
+    if (Sint32 value; SDL_ReadS32BE(get(), &value)) return value;
+    return {};
   }
 
+  /**
+   * Use this function to read 64 bits of little-endian data from an
+   * IOStreamRef and return in native format.
+   *
+   * SDL byteswaps the data only if necessary, so the data returned will be in
+   * the native byte order.
+   *
+   * This function will return false when the data stream is completely read,
+   * and IOStreamRef.GetStatus() will return IO_STATUS_EOF. If false is
+   * returned and the stream is not at EOF, IOStreamRef.GetStatus() will return
+   * a different error value and GetError() will offer a human-readable message.
+   *
+   * @returns the data read on success, std::nullopt on failure.
+   *
+   * @threadsafety This function is not thread safe.
+   *
+   * @since This function is available since SDL 3.2.0.
+   */
   std::optional<Uint64> TryReadU64LE()
   {
-    static_assert(false, "Not implemented");
+    if (Uint64 value; SDL_ReadU64LE(get(), &value)) return value;
+    return {};
   }
 
+  /**
+   * Use this function to read 64 bits of little-endian data from an
+   * IOStreamRef and return in native format.
+   *
+   * SDL byteswaps the data only if necessary, so the data returned will be in
+   * the native byte order.
+   *
+   * This function will return false when the data stream is completely read,
+   * and IOStreamRef.GetStatus() will return IO_STATUS_EOF. If false is
+   * returned and the stream is not at EOF, IOStreamRef.GetStatus() will return
+   * a different error value and GetError() will offer a human-readable message.
+   *
+   * @returns the data read on success, std::nullopt on failure.
+   *
+   * @threadsafety This function is not thread safe.
+   *
+   * @since This function is available since SDL 3.2.0.
+   */
   std::optional<Sint64> TryReadS64LE()
   {
-    static_assert(false, "Not implemented");
+    if (Sint64 value; SDL_ReadS64LE(get(), &value)) return value;
+    return {};
   }
 
+  /**
+   * Use this function to read 64 bits of big-endian data from an IOStreamRef
+   * and return in native format.
+   *
+   * SDL byteswaps the data only if necessary, so the data returned will be in
+   * the native byte order.
+   *
+   * This function will return false when the data stream is completely read,
+   * and IOStreamRef.GetStatus() will return IO_STATUS_EOF. If false is
+   * returned and the stream is not at EOF, IOStreamRef.GetStatus() will return
+   * a different error value and GetError() will offer a human-readable message.
+   *
+   * @returns the data read on success, std::nullopt on failure.
+   *
+   * @threadsafety This function is not thread safe.
+   *
+   * @since This function is available since SDL 3.2.0.
+   */
   std::optional<Uint64> TryReadU64BE()
   {
-    static_assert(false, "Not implemented");
+    if (Uint64 value; SDL_ReadU64BE(get(), &value)) return value;
+    return {};
   }
 
+  /**
+   * Use this function to read 64 bits of big-endian data from an IOStreamRef
+   * and return in native format.
+   *
+   * SDL byteswaps the data only if necessary, so the data returned will be in
+   * the native byte order.
+   *
+   * This function will return false when the data stream is completely read,
+   * and IOStreamRef.GetStatus() will return IO_STATUS_EOF. If false is
+   * returned and the stream is not at EOF, IOStreamRef.GetStatus() will return
+   * a different error value and GetError() will offer a human-readable message.
+   *
+   * @returns the data read on success, std::nullopt on failure.
+   *
+   * @threadsafety This function is not thread safe.
+   *
+   * @since This function is available since SDL 3.2.0.
+   */
   std::optional<Sint64> TryReadS64BE()
   {
-    static_assert(false, "Not implemented");
+    if (Sint64 value; SDL_ReadS64BE(get(), &value)) return value;
+    return {};
   }
 
   /**
@@ -1389,7 +1702,7 @@
  * - "w": Create an empty file for writing. If a file with the same name already
  *   exists its content is erased and the file is treated as a new empty file.
  * - "wx": Create an empty file for writing. If a file with the same name
- *   already exists, the call fails.
+ *   already exists, the call fails. (Supported since SDL 3.4.0)
  * - "a": Append to a file. Writing operations append data at the end of the
  *   file. The file is created if it does not exist.
  * - "r+": Open a file for update both reading and writing. The file must exist.
@@ -1397,7 +1710,7 @@
  *   same name already exists its content is erased and the file is treated as a
  *   new empty file.
  * - "w+x": Create an empty file for both reading and writing. If a file with
- *   the same name already exists, the call fails.
+ *   the same name already exists, the call fails. (Supported since SDL 3.4.0)
  * - "a+": Open a file for reading and appending. All writing operations are
  *   performed at the end of the file, protecting the previous content to be
  *   overwritten. You can reposition (fseek, rewind) the internal pointer to
@@ -1527,10 +1840,9 @@
  *   and called with the passed `mem` pointer when closing the stream. By
  *   default it is unset, i.e., the memory will not be freed.
  *
- * @param mem a pointer to a buffer to feed an IOStream stream.
- * @param size the buffer size, in bytes.
- * @returns a pointer to a new IOStream structure or nullptr on failure; call
- *          GetError() for more information.
+ * @param mem a buffer to feed an IOStream stream.
+ * @returns a valid IOStream on success.
+ * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -1544,9 +1856,9 @@
  * @sa IOStream.Tell
  * @sa IOStream.Write
  */
-inline IOStream IOFromMem(void* mem, size_t size)
+inline IOStream IOFromMem(TargetBytes mem)
 {
-  return IOStream(SDL_IOFromMem(mem, size));
+  return IOStream(CheckError(SDL_IOFromMem(mem.data(), mem.size_bytes())));
 }
 
 inline IOStream IOStream::FromMem(TargetBytes mem)
@@ -1583,10 +1895,9 @@
  *   and called with the passed `mem` pointer when closing the stream. By
  *   default it is unset, i.e., the memory will not be freed.
  *
- * @param mem a pointer to a read-only buffer to feed an IOStream stream.
- * @param size the buffer size, in bytes.
- * @returns a pointer to a new IOStream structure or nullptr on failure; call
- *          GetError() for more information.
+ * @param mem a read-only buffer to feed an IOStreamRef stream.
+ * @returns a valid IOStream on success.
+ * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -1598,9 +1909,9 @@
  * @sa IOStream.Seek
  * @sa IOStream.Tell
  */
-inline IOStream IOFromConstMem(const void* mem, size_t size)
+inline IOStream IOFromConstMem(SourceBytes mem)
 {
-  return IOStream(SDL_IOFromConstMem(mem, size));
+  return IOStream(CheckError(SDL_IOFromConstMem(mem.data(), mem.size_bytes())));
 }
 
 inline IOStream IOStream::FromConstMem(SourceBytes mem)
@@ -1722,7 +2033,7 @@
  */
 inline PropertiesRef GetIOProperties(IOStreamParam context)
 {
-  return CheckError(SDL_GetIOProperties(context));
+  return {CheckError(SDL_GetIOProperties(context))};
 }
 
 inline PropertiesRef IOStream::GetProperties() const
@@ -1850,8 +2161,7 @@
  * change.
  *
  * @param context a pointer to an IOStream structure.
- * @param ptr a pointer to a buffer to read data into.
- * @param size the number of bytes to read from the data source.
+ * @param buf a pointer to a buffer to read data into.
  * @returns the number of bytes read, or 0 on end of file or other failure; call
  *          GetError() for more information.
  *
@@ -1891,8 +2201,7 @@
  * change.
  *
  * @param context a pointer to an IOStream structure.
- * @param ptr a pointer to a buffer containing data to write.
- * @param size the number of bytes to write.
+ * @param buf a pointer to a buffer containing data to write.
  * @returns the number of bytes written, which will be less than `size` on
  *          failure; call GetError() for more information.
  *
@@ -1919,6 +2228,8 @@
 /**
  * Print to an IOStream data stream.
  *
+ * @warning this is not typesafe! Prefer using print() and println()
+ *
  * This function does formatted printing to the stream.
  *
  * @param context a pointer to an IOStream structure.
@@ -1936,9 +2247,17 @@
  * @sa IOStream.Write
  */
 inline size_t IOprintf(IOStreamParam context,
-                       SDL_PRINTF_FORMAT_STRING const char* fmt)
+                       SDL_PRINTF_FORMAT_STRING const char* fmt,
+                       ...)
 {
-  return SDL_IOprintf(context, fmt);
+  va_list ap;
+  size_t result;
+
+  va_start(ap, fmt);
+  result = SDL_IOvprintf(context, fmt, ap);
+  va_end(ap);
+
+  return result;
 }
 
 /**
@@ -1997,14 +2316,12 @@
  * Load all the data from an SDL data stream.
  *
  * The data is allocated with a zero byte at the end (null terminated) for
- * convenience. This extra byte is not included in the value reported via
- * `datasize`.
+ * convenience. This extra byte is not included in the value reported on
+ * the returned string.
  *
  * The data should be freed with free().
  *
  * @param src the IOStream to read all available data from.
- * @param datasize a pointer filled in with the number of bytes read, may be
- *                 nullptr.
  * @param closeio if true, calls IOStream.Close() on `src` before returning,
  *                even in the case of an error.
  * @returns the data or nullptr on failure; call GetError() for more
@@ -2019,22 +2336,21 @@
  */
 inline StringResult LoadFile(IOStreamParam src, bool closeio = true)
 {
-  return SDL_LoadFile_IO(src, closeio);
+  size_t datasize = 0;
+  auto data = static_cast<char*>(SDL_LoadFile_IO(src, &datasize, closeio));
+  return StringResult{CheckError(data), datasize};
 }
 
 /**
  * Load all the data from a file path.
  *
  * The data is allocated with a zero byte at the end (null terminated) for
- * convenience. This extra byte is not included in the value reported via
- * `datasize`.
- *
- * The data should be freed with free().
+ * convenience. This extra byte is not included in the value reported on
+ * the returned string.
  *
  * @param file the path to read all available data from.
- * @param datasize if not nullptr, will store the number of bytes read.
- * @returns the data or nullptr on failure; call GetError() for more
- *          information.
+ * @returns the data.
+ * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -2043,23 +2359,49 @@
  * @sa IOStream.LoadFile
  * @sa SaveFile
  */
-inline StringResult LoadFile(StringParam file) { return SDL_LoadFile(file); }
+inline StringResult LoadFile(StringParam file)
+{
+  size_t datasize = 0;
+  auto data = static_cast<char*>(SDL_LoadFile(file, &datasize));
+  return StringResult{CheckError(data), datasize};
+}
 
-inline StringResult IOStream::LoadFile() { return SDL::LoadFile(m_resource); }
+inline StringResult IOStream::LoadFile()
+{
+  return SDL::LoadFile(m_resource, false);
+}
 
+/**
+ * Load all the data from a file path.
+ *
+ * The data is allocated with a zero byte at the end (null terminated) for
+ * convenience. This extra byte is not included in the value reported on
+ * the returned string.
+ *
+ * @param file the path to read all available data from.
+ * @returns the data.
+ * @throws Error on failure.
+ *
+ * @threadsafety This function is not thread safe.
+ *
+ * @since This function is available since SDL 3.2.0.
+ *
+ * @sa IOStream.LoadFile
+ * @sa SaveFile
+ */
 template<class T>
 inline OwnArray<T> LoadFileAs(StringParam file)
 {
-  static_assert(false, "Not implemented");
+  size_t datasize = 0;
+  auto data = static_cast<T*>(SDL_LoadFile(file, &datasize));
+  return OwnArray<T>{CheckError(data), datasize / sizeof(T)};
 }
 
 /**
  * Save all the data into an SDL data stream.
  *
  * @param src the IOStream to write all data to.
- * @param data the data to be written. If datasize is 0, may be nullptr or a
- *             invalid pointer.
- * @param datasize the number of bytes to be written.
+ * @param data the data to be written.
  * @param closeio if true, calls IOStream.Close() on `src` before returning,
  *                even in the case of an error.
  * @throws Error on failure.
@@ -2071,7 +2413,7 @@
  * @sa SaveFile
  * @sa IOStream.LoadFile
  */
-inline void SaveFile(IOStreamParam src, SourceBytes data, bool closeio = true)
+inline void SaveFile(IOStreamParam src, SourceBytes data, bool closeio = false)
 {
   CheckError(SDL_SaveFile_IO(src, data.data(), data.size_bytes(), closeio));
 }
@@ -2080,9 +2422,7 @@
  * Save all the data into a file path.
  *
  * @param file the path to write all available data into.
- * @param data the data to be written. If datasize is 0, may be nullptr or a
- *             invalid pointer.
- * @param datasize the number of bytes to be written.
+ * @param data the data to be written.
  * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
@@ -2111,16 +2451,19 @@
  * value and GetError() will offer a human-readable message.
  *
  * @param src the IOStream to read from.
- * @param value a pointer filled in with the data read.
+ * @return the  data read.
  * @throws Error on failure.
  *
  * @threadsafety Do not use the same IOStream from two threads at once.
  *
  * @since This function is available since SDL 3.2.0.
  */
-inline Uint8 ReadU8(IOStreamParam src) { return CheckError(SDL_ReadU8(src)); }
-
-inline Uint8 IOStream::ReadU8() { return SDL::ReadU8(m_resource); }
+inline Uint8 ReadU8(IOStreamParam src)
+{
+  Uint8 value;
+  CheckError(SDL_ReadU8(src, &value));
+  return value;
+}
 
 /**
  * Use this function to read a signed byte from an IOStream.
@@ -2131,16 +2474,19 @@
  * value and GetError() will offer a human-readable message.
  *
  * @param src the IOStream to read from.
- * @param value a pointer filled in with the data read.
+ * @return the  data read.
  * @throws Error on failure.
  *
  * @threadsafety Do not use the same IOStream from two threads at once.
  *
  * @since This function is available since SDL 3.2.0.
  */
-inline Sint8 ReadS8(IOStreamParam src) { return CheckError(SDL_ReadS8(src)); }
-
-inline Sint8 IOStream::ReadS8() { return SDL::ReadS8(m_resource); }
+inline Sint8 ReadS8(IOStreamParam src)
+{
+  Sint8 value;
+  CheckError(SDL_ReadS8(src, &value));
+  return value;
+}
 
 /**
  * Use this function to read 16 bits of little-endian data from an IOStream and
@@ -2155,7 +2501,7 @@
  * value and GetError() will offer a human-readable message.
  *
  * @param src the stream from which to read data.
- * @param value a pointer filled in with the data read.
+ * @return the  data read.
  * @throws Error on failure.
  *
  * @threadsafety Do not use the same IOStream from two threads at once.
@@ -2164,7 +2510,9 @@
  */
 inline Uint16 ReadU16LE(IOStreamParam src)
 {
-  return CheckError(SDL_ReadU16LE(src));
+  Uint16 value;
+  CheckError(SDL_ReadU16LE(src, &value));
+  return value;
 }
 
 inline Uint16 IOStream::ReadU16LE() { return SDL::ReadU16LE(m_resource); }
@@ -2182,7 +2530,7 @@
  * value and GetError() will offer a human-readable message.
  *
  * @param src the stream from which to read data.
- * @param value a pointer filled in with the data read.
+ * @return the  data read.
  * @throws Error on failure.
  *
  * @threadsafety Do not use the same IOStream from two threads at once.
@@ -2191,7 +2539,9 @@
  */
 inline Sint16 ReadS16LE(IOStreamParam src)
 {
-  return CheckError(SDL_ReadS16LE(src));
+  Sint16 value;
+  CheckError(SDL_ReadS16LE(src, &value));
+  return value;
 }
 
 inline Sint16 IOStream::ReadS16LE() { return SDL::ReadS16LE(m_resource); }
@@ -2209,7 +2559,7 @@
  * value and GetError() will offer a human-readable message.
  *
  * @param src the stream from which to read data.
- * @param value a pointer filled in with the data read.
+ * @return the  data read.
  * @throws Error on failure.
  *
  * @threadsafety Do not use the same IOStream from two threads at once.
@@ -2218,7 +2568,9 @@
  */
 inline Uint16 ReadU16BE(IOStreamParam src)
 {
-  return CheckError(SDL_ReadU16BE(src));
+  Uint16 value;
+  CheckError(SDL_ReadU16BE(src, &value));
+  return value;
 }
 
 inline Uint16 IOStream::ReadU16BE() { return SDL::ReadU16BE(m_resource); }
@@ -2236,7 +2588,7 @@
  * value and GetError() will offer a human-readable message.
  *
  * @param src the stream from which to read data.
- * @param value a pointer filled in with the data read.
+ * @return the  data read.
  * @throws Error on failure.
  *
  * @threadsafety Do not use the same IOStream from two threads at once.
@@ -2245,7 +2597,9 @@
  */
 inline Sint16 ReadS16BE(IOStreamParam src)
 {
-  return CheckError(SDL_ReadS16BE(src));
+  Sint16 value;
+  CheckError(SDL_ReadS16BE(src, &value));
+  return value;
 }
 
 inline Sint16 IOStream::ReadS16BE() { return SDL::ReadS16BE(m_resource); }
@@ -2263,7 +2617,7 @@
  * value and GetError() will offer a human-readable message.
  *
  * @param src the stream from which to read data.
- * @param value a pointer filled in with the data read.
+ * @return the  data read.
  * @throws Error on failure.
  *
  * @threadsafety Do not use the same IOStream from two threads at once.
@@ -2272,7 +2626,9 @@
  */
 inline Uint32 ReadU32LE(IOStreamParam src)
 {
-  return CheckError(SDL_ReadU32LE(src));
+  Uint32 value;
+  CheckError(SDL_ReadU32LE(src, &value));
+  return value;
 }
 
 inline Uint32 IOStream::ReadU32LE() { return SDL::ReadU32LE(m_resource); }
@@ -2290,7 +2646,7 @@
  * value and GetError() will offer a human-readable message.
  *
  * @param src the stream from which to read data.
- * @param value a pointer filled in with the data read.
+ * @return the  data read.
  * @throws Error on failure.
  *
  * @threadsafety Do not use the same IOStream from two threads at once.
@@ -2299,7 +2655,9 @@
  */
 inline Sint32 ReadS32LE(IOStreamParam src)
 {
-  return CheckError(SDL_ReadS32LE(src));
+  Sint32 value;
+  CheckError(SDL_ReadS32LE(src, &value));
+  return value;
 }
 
 inline Sint32 IOStream::ReadS32LE() { return SDL::ReadS32LE(m_resource); }
@@ -2317,7 +2675,7 @@
  * value and GetError() will offer a human-readable message.
  *
  * @param src the stream from which to read data.
- * @param value a pointer filled in with the data read.
+ * @return the  data read.
  * @throws Error on failure.
  *
  * @threadsafety Do not use the same IOStream from two threads at once.
@@ -2326,7 +2684,9 @@
  */
 inline Uint32 ReadU32BE(IOStreamParam src)
 {
-  return CheckError(SDL_ReadU32BE(src));
+  Uint32 value;
+  CheckError(SDL_ReadU32BE(src, &value));
+  return value;
 }
 
 inline Uint32 IOStream::ReadU32BE() { return SDL::ReadU32BE(m_resource); }
@@ -2344,7 +2704,7 @@
  * value and GetError() will offer a human-readable message.
  *
  * @param src the stream from which to read data.
- * @param value a pointer filled in with the data read.
+ * @return the  data read.
  * @throws Error on failure.
  *
  * @threadsafety Do not use the same IOStream from two threads at once.
@@ -2353,7 +2713,9 @@
  */
 inline Sint32 ReadS32BE(IOStreamParam src)
 {
-  return CheckError(SDL_ReadS32BE(src));
+  Sint32 value;
+  CheckError(SDL_ReadS32BE(src, &value));
+  return value;
 }
 
 inline Sint32 IOStream::ReadS32BE() { return SDL::ReadS32BE(m_resource); }
@@ -2371,7 +2733,7 @@
  * value and GetError() will offer a human-readable message.
  *
  * @param src the stream from which to read data.
- * @param value a pointer filled in with the data read.
+ * @return the  data read.
  * @throws Error on failure.
  *
  * @threadsafety Do not use the same IOStream from two threads at once.
@@ -2380,7 +2742,9 @@
  */
 inline Uint64 ReadU64LE(IOStreamParam src)
 {
-  return CheckError(SDL_ReadU64LE(src));
+  Uint64 value;
+  CheckError(SDL_ReadU64LE(src, &value));
+  return value;
 }
 
 inline Uint64 IOStream::ReadU64LE() { return SDL::ReadU64LE(m_resource); }
@@ -2398,7 +2762,7 @@
  * value and GetError() will offer a human-readable message.
  *
  * @param src the stream from which to read data.
- * @param value a pointer filled in with the data read.
+ * @return the  data read.
  * @throws Error on failure.
  *
  * @threadsafety Do not use the same IOStream from two threads at once.
@@ -2407,7 +2771,9 @@
  */
 inline Sint64 ReadS64LE(IOStreamParam src)
 {
-  return CheckError(SDL_ReadS64LE(src));
+  Sint64 value;
+  CheckError(SDL_ReadS64LE(src, &value));
+  return value;
 }
 
 inline Sint64 IOStream::ReadS64LE() { return SDL::ReadS64LE(m_resource); }
@@ -2425,7 +2791,7 @@
  * value and GetError() will offer a human-readable message.
  *
  * @param src the stream from which to read data.
- * @param value a pointer filled in with the data read.
+ * @return the  data read.
  * @throws Error on failure.
  *
  * @threadsafety Do not use the same IOStream from two threads at once.
@@ -2434,7 +2800,9 @@
  */
 inline Uint64 ReadU64BE(IOStreamParam src)
 {
-  return CheckError(SDL_ReadU64BE(src));
+  Uint64 value;
+  CheckError(SDL_ReadU64BE(src, &value));
+  return value;
 }
 
 inline Uint64 IOStream::ReadU64BE() { return SDL::ReadU64BE(m_resource); }
@@ -2452,7 +2820,7 @@
  * value and GetError() will offer a human-readable message.
  *
  * @param src the stream from which to read data.
- * @param value a pointer filled in with the data read.
+ * @return the  data read.
  * @throws Error on failure.
  *
  * @threadsafety Do not use the same IOStream from two threads at once.
@@ -2461,7 +2829,9 @@
  */
 inline Sint64 ReadS64BE(IOStreamParam src)
 {
-  return CheckError(SDL_ReadS64BE(src));
+  Sint64 value;
+  CheckError(SDL_ReadS64BE(src, &value));
+  return value;
 }
 
 inline Sint64 IOStream::ReadS64BE() { return SDL::ReadS64BE(m_resource); }
