--- build/generated/SDL3pp_gpu.h
+++ include/SDL3pp/SDL3pp_gpu.h
@@ -12,7 +12,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryGPU Category GPU
+ * @defgroup CategoryGPU 3D Rendering and GPU Compute
  *
  * The GPU API offers a cross-platform way for apps to talk to modern graphics
  * hardware. It offers both 3D graphics and compute support, in the style of
@@ -138,7 +138,7 @@
  * limited hardware support, this API is probably not for you.
  *
  * Examples demonstrating proper usage of this API can be found
- * [here](https://github.com/TheSpydog/SDL_gpu_examples) .
+ * [here](https://github.com/TheSpydog/SDL_gpu_examples).
  *
  * ## Performance considerations
  *
@@ -1256,7 +1256,6 @@
    * @param first_slot the vertex buffer slot to begin binding from.
    * @param bindings an array of GPUBufferBinding structs containing vertex
    *                 buffers and offset values.
-   * @param num_bindings the number of bindings in the bindings array.
    *
    * @since This function is available since SDL 3.2.0.
    */
@@ -1287,8 +1286,6 @@
    * @param first_slot the vertex sampler slot to begin binding from.
    * @param texture_sampler_bindings an array of texture-sampler binding
    *                                 structs.
-   * @param num_bindings the number of texture-sampler pairs to bind from the
-   *                     array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1309,7 +1306,6 @@
    *
    * @param first_slot the vertex storage texture slot to begin binding from.
    * @param storage_textures an array of storage textures.
-   * @param num_bindings the number of storage texture to bind from the array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1329,7 +1325,6 @@
    *
    * @param first_slot the vertex storage buffer slot to begin binding from.
    * @param storage_buffers an array of buffers.
-   * @param num_bindings the number of buffers to bind from the array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1349,8 +1344,6 @@
    * @param first_slot the fragment sampler slot to begin binding from.
    * @param texture_sampler_bindings an array of texture-sampler binding
    *                                 structs.
-   * @param num_bindings the number of texture-sampler pairs to bind from the
-   *                     array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1371,7 +1364,6 @@
    *
    * @param first_slot the fragment storage texture slot to begin binding from.
    * @param storage_textures an array of storage textures.
-   * @param num_bindings the number of storage textures to bind from the array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1392,7 +1384,6 @@
    *
    * @param first_slot the fragment storage buffer slot to begin binding from.
    * @param storage_buffers an array of storage buffers.
-   * @param num_bindings the number of storage buffers to bind from the array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1558,8 +1549,6 @@
    * @param first_slot the compute sampler slot to begin binding from.
    * @param texture_sampler_bindings an array of texture-sampler binding
    *                                 structs.
-   * @param num_bindings the number of texture-sampler bindings to bind from the
-   *                     array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1580,7 +1569,6 @@
    *
    * @param first_slot the compute storage texture slot to begin binding from.
    * @param storage_textures an array of storage textures.
-   * @param num_bindings the number of storage textures to bind from the array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1600,7 +1588,6 @@
    *
    * @param first_slot the compute storage buffer slot to begin binding from.
    * @param storage_buffers an array of storage buffer binding structs.
-   * @param num_bindings the number of storage buffers to bind from the array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -2144,7 +2131,6 @@
    *
    * @param slot_index the vertex uniform slot to push data to.
    * @param data client data to write.
-   * @param length the length of the data to write.
    *
    * @since This function is available since SDL 3.2.0.
    */
@@ -2161,7 +2147,6 @@
    *
    * @param slot_index the fragment uniform slot to push data to.
    * @param data client data to write.
-   * @param length the length of the data to write.
    *
    * @since This function is available since SDL 3.2.0.
    */
@@ -2178,7 +2163,6 @@
    *
    * @param slot_index the uniform slot to push data to.
    * @param data client data to write.
-   * @param length the length of the data to write.
    *
    * @since This function is available since SDL 3.2.0.
    */
@@ -2205,8 +2189,6 @@
    *
    * @param color_target_infos an array of texture subresources with
    *                           corresponding clear values and load/store ops.
-   * @param num_color_targets the number of color targets in the
-   *                          color_target_infos array.
    * @param depth_stencil_target_info a texture subresource with corresponding
    *                                  clear value and load/store ops, may be
    *                                  nullptr.
@@ -2244,12 +2226,8 @@
    *
    * @param storage_texture_bindings an array of writeable storage texture
    *                                 binding structs.
-   * @param num_storage_texture_bindings the number of storage textures to bind
-   *                                     from the array.
    * @param storage_buffer_bindings an array of writeable storage buffer binding
    *                                structs.
-   * @param num_storage_buffer_bindings the number of storage buffers to bind
-   *                                    from the array.
    * @returns a compute pass handle.
    *
    * @since This function is available since SDL 3.2.0.
@@ -2321,12 +2299,11 @@
    * than the one that created the window.
    *
    * @param window a window that has been claimed.
-   * @param swapchain_texture a pointer filled in with a swapchain texture
-   *                          handle.
    * @param swapchain_texture_width a pointer filled in with the swapchain
    *                                texture width, may be nullptr.
    * @param swapchain_texture_height a pointer filled in with the swapchain
    *                                 texture height, may be nullptr.
+   * @returns a swapchain texture handle.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called from the thread that
@@ -2371,12 +2348,11 @@
    * another reading operation.
    *
    * @param window a window that has been claimed.
-   * @param swapchain_texture a pointer filled in with a swapchain texture
-   *                          handle.
    * @param swapchain_texture_width a pointer filled in with the swapchain
    *                                texture width, may be nullptr.
    * @param swapchain_texture_height a pointer filled in with the swapchain
    *                                 texture height, may be nullptr.
+   * @returns a swapchain texture handle.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called from the thread that
@@ -2577,9 +2553,6 @@
  * format is supported before using it. However, there are a few guaranteed
  * formats.
  *
- * FIXME: Check universal support for 32-bit component formats FIXME: Check
- * universal support for SIMULTANEOUS_READ_WRITE
- *
  * For SAMPLER usage, the following formats are universally supported:
  *
  * - R8G8B8A8_UNORM
@@ -4025,7 +3998,6 @@
    * @param wait_all if 0, wait for any fence to be signaled, if 1, wait for all
    *                 fences to be signaled.
    * @param fences an array of fences to wait on.
-   * @param num_fences the number of fences in the fences array.
    * @throws Error on failure.
    *
    * @since This function is available since SDL 3.2.0.
@@ -4087,6 +4059,8 @@
   bool TextureSupportsSampleCount(GPUTextureFormat format,
                                   GPUSampleCount sample_count);
 
+#ifdef SDL_PLATFORM_GDK
+
   /**
    * Call this to suspend GPU operation on Xbox when you receive the
    * EVENT_DID_ENTER_BACKGROUND event.
@@ -4112,6 +4086,8 @@
    * @sa AddEventWatch
    */
   void GDKResumeGPU();
+
+#endif /* SDL_PLATFORM_GDK */
 };
 
 /**
@@ -5087,6 +5063,13 @@
   return GPUDevice(props);
 }
 
+/**
+ * Properties for GPUDevice creation.
+ *
+ * @sa CreateGPUDeviceWithProperties
+ * @sa GPUDevice.GetProperties
+ * @sa prop::GPUDevice
+ */
 namespace prop::GPUDevice::Create {
 
 constexpr auto DEBUGMODE_BOOLEAN = SDL_PROP_GPU_DEVICE_CREATE_DEBUGMODE_BOOLEAN;
@@ -5393,6 +5376,18 @@
   return SDL::GetGPUDeviceProperties(get());
 }
 
+/**
+ * Properties for GPUDevice creation and information.
+ *
+ * The properties in namespace `Create::` can be used in the properties object
+ * passed to CreateGPUDeviceWithProperties to specify the desired GPU device and
+ * driver features, while the remainder can be queried from an existing GPU
+ * device via GPUDevice.GetProperties.
+ *
+ * @sa CreateGPUDeviceWithProperties
+ * @sa GPUDevice.GetProperties
+ * @sa prop::GPUDevice::Create
+ */
 namespace prop::GPUDevice {
 
 constexpr auto NAME_STRING = SDL_PROP_GPU_DEVICE_NAME_STRING;
@@ -5474,6 +5469,11 @@
 {
 }
 
+/**
+ * Properties for GPUComputePipeline creation.
+ *
+ * @sa GPUComputePipeline.CreateComputePipeline
+ */
 namespace prop::GPUComputePipeline::Create {
 
 constexpr auto NAME_STRING = SDL_PROP_GPU_COMPUTEPIPELINE_CREATE_NAME_STRING;
@@ -5522,6 +5522,11 @@
 {
 }
 
+/**
+ * Properties for GPUGraphicsPipeline creation.
+ *
+ * @sa GPUGraphicsPipeline.CreateGraphicsPipeline
+ */
 namespace prop::GPUGraphicsPipeline::Create {
 
 constexpr auto NAME_STRING = SDL_PROP_GPU_GRAPHICSPIPELINE_CREATE_NAME_STRING;
@@ -5567,6 +5572,11 @@
 {
 }
 
+/**
+ * Properties for GPUSampler creation.
+ *
+ * @sa GPUDevice.CreateSampler
+ */
 namespace prop::GPUSampler::Create {
 
 constexpr auto NAME_STRING = SDL_PROP_GPU_SAMPLER_CREATE_NAME_STRING;
@@ -5661,6 +5671,11 @@
 {
 }
 
+/**
+ * Properties for GPUShader creation.
+ *
+ * @sa GPUDevice.CreateShader
+ */
 namespace prop::GPUShader::Create {
 
 constexpr auto NAME_STRING = SDL_PROP_GPU_SHADER_CREATE_NAME_STRING;
@@ -5743,6 +5758,11 @@
 {
 }
 
+/**
+ * Properties for GPUTexture creation.
+ *
+ * @sa GPUDevice.CreateTexture
+ */
 namespace prop::GPUTexture::Create {
 
 constexpr auto D3D12_CLEAR_R_FLOAT =
@@ -5831,6 +5851,11 @@
 {
 }
 
+/**
+ * Properties for GPUBuffer creation.
+ *
+ * @sa GPUDevice.CreateBuffer
+ */
 namespace prop::GPUBuffer::Create {
 
 constexpr auto NAME_STRING = SDL_PROP_GPU_BUFFER_CREATE_NAME_STRING;
@@ -5885,6 +5910,11 @@
 {
 }
 
+/**
+ * Properties for GPUTransferBuffer creation.
+ *
+ * @sa GPUDevice.CreateTransferBuffer
+ */
 namespace prop::GPUTransferBuffer::Create {
 
 constexpr auto NAME_STRING = SDL_PROP_GPU_TRANSFERBUFFER_CREATE_NAME_STRING;
@@ -6233,7 +6263,6 @@
  * @param command_buffer a command buffer.
  * @param slot_index the vertex uniform slot to push data to.
  * @param data client data to write.
- * @param length the length of the data to write.
  *
  * @since This function is available since SDL 3.2.0.
  */
@@ -6242,7 +6271,7 @@
                                      SourceBytes data)
 {
   SDL_PushGPUVertexUniformData(
-    command_buffer, slot_index, data.data(), data.size_bytes());
+    command_buffer, slot_index, data.data(), narrowU32(data.size_bytes()));
 }
 
 inline void GPUCommandBuffer::PushVertexUniformData(Uint32 slot_index,
@@ -6264,7 +6293,6 @@
  * @param command_buffer a command buffer.
  * @param slot_index the fragment uniform slot to push data to.
  * @param data client data to write.
- * @param length the length of the data to write.
  *
  * @since This function is available since SDL 3.2.0.
  */
@@ -6273,7 +6301,7 @@
                                        SourceBytes data)
 {
   SDL_PushGPUFragmentUniformData(
-    command_buffer, slot_index, data.data(), data.size_bytes());
+    command_buffer, slot_index, data.data(), narrowU32(data.size_bytes()));
 }
 
 inline void GPUCommandBuffer::PushFragmentUniformData(Uint32 slot_index,
@@ -6295,7 +6323,6 @@
  * @param command_buffer a command buffer.
  * @param slot_index the uniform slot to push data to.
  * @param data client data to write.
- * @param length the length of the data to write.
  *
  * @since This function is available since SDL 3.2.0.
  */
@@ -6304,7 +6331,7 @@
                                       SourceBytes data)
 {
   SDL_PushGPUComputeUniformData(
-    command_buffer, slot_index, data.data(), data.size_bytes());
+    command_buffer, slot_index, data.data(), narrowU32(data.size_bytes()));
 }
 
 inline void GPUCommandBuffer::PushComputeUniformData(Uint32 slot_index,
@@ -6336,8 +6363,6 @@
  * @param command_buffer a command buffer.
  * @param color_target_infos an array of texture subresources with corresponding
  *                           clear values and load/store ops.
- * @param num_color_targets the number of color targets in the
- *                          color_target_infos array.
  * @param depth_stencil_target_info a texture subresource with corresponding
  *                                  clear value and load/store ops, may be
  *                                  nullptr.
@@ -6352,8 +6377,10 @@
   std::span<const GPUColorTargetInfo> color_target_infos,
   OptionalRef<const GPUDepthStencilTargetInfo> depth_stencil_target_info)
 {
-  return SDL_BeginGPURenderPass(
-    command_buffer, color_target_infos, depth_stencil_target_info);
+  return SDL_BeginGPURenderPass(command_buffer,
+                                color_target_infos.data(),
+                                narrowU32(color_target_infos.size()),
+                                depth_stencil_target_info);
 }
 
 inline GPURenderPass GPUCommandBuffer::BeginRenderPass(
@@ -6414,7 +6441,7 @@
  */
 inline void SetGPUScissor(GPURenderPass render_pass, const RectRaw& scissor)
 {
-  SDL_SetGPUScissor(render_pass, scissor);
+  SDL_SetGPUScissor(render_pass, &scissor);
 }
 
 inline void GPURenderPass::SetScissor(const RectRaw& scissor)
@@ -6469,7 +6496,6 @@
  * @param first_slot the vertex buffer slot to begin binding from.
  * @param bindings an array of GPUBufferBinding structs containing vertex
  *                 buffers and offset values.
- * @param num_bindings the number of bindings in the bindings array.
  *
  * @since This function is available since SDL 3.2.0.
  */
@@ -6477,7 +6503,8 @@
                                  Uint32 first_slot,
                                  std::span<const GPUBufferBinding> bindings)
 {
-  SDL_BindGPUVertexBuffers(render_pass, first_slot, bindings);
+  SDL_BindGPUVertexBuffers(
+    render_pass, first_slot, bindings.data(), narrowU32(bindings.size()));
 }
 
 inline void GPURenderPass::BindVertexBuffers(
@@ -6522,8 +6549,6 @@
  * @param render_pass a render pass handle.
  * @param first_slot the vertex sampler slot to begin binding from.
  * @param texture_sampler_bindings an array of texture-sampler binding structs.
- * @param num_bindings the number of texture-sampler pairs to bind from the
- *                     array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -6534,7 +6559,10 @@
   Uint32 first_slot,
   std::span<const GPUTextureSamplerBinding> texture_sampler_bindings)
 {
-  SDL_BindGPUVertexSamplers(render_pass, first_slot, texture_sampler_bindings);
+  SDL_BindGPUVertexSamplers(render_pass,
+                            first_slot,
+                            texture_sampler_bindings.data(),
+                            narrowU32(texture_sampler_bindings.size()));
 }
 
 inline void GPURenderPass::BindVertexSamplers(
@@ -6557,7 +6585,6 @@
  * @param render_pass a render pass handle.
  * @param first_slot the vertex storage texture slot to begin binding from.
  * @param storage_textures an array of storage textures.
- * @param num_bindings the number of storage texture to bind from the array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -6568,7 +6595,10 @@
   Uint32 first_slot,
   SpanRef<const GPUTextureRaw> storage_textures)
 {
-  SDL_BindGPUVertexStorageTextures(render_pass, first_slot, storage_textures);
+  SDL_BindGPUVertexStorageTextures(render_pass,
+                                   first_slot,
+                                   storage_textures.data(),
+                                   narrowU32(storage_textures.size()));
 }
 
 inline void GPURenderPass::BindVertexStorageTextures(
@@ -6591,7 +6621,6 @@
  * @param render_pass a render pass handle.
  * @param first_slot the vertex storage buffer slot to begin binding from.
  * @param storage_buffers an array of buffers.
- * @param num_bindings the number of buffers to bind from the array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -6602,7 +6631,10 @@
   Uint32 first_slot,
   SpanRef<const GPUBufferRaw> storage_buffers)
 {
-  SDL_BindGPUVertexStorageBuffers(render_pass, first_slot, storage_buffers);
+  SDL_BindGPUVertexStorageBuffers(render_pass,
+                                  first_slot,
+                                  storage_buffers.data(),
+                                  narrowU32(storage_buffers.size()));
 }
 
 inline void GPURenderPass::BindVertexStorageBuffers(
@@ -6624,8 +6656,6 @@
  * @param render_pass a render pass handle.
  * @param first_slot the fragment sampler slot to begin binding from.
  * @param texture_sampler_bindings an array of texture-sampler binding structs.
- * @param num_bindings the number of texture-sampler pairs to bind from the
- *                     array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -6636,8 +6666,10 @@
   Uint32 first_slot,
   std::span<const GPUTextureSamplerBinding> texture_sampler_bindings)
 {
-  SDL_BindGPUFragmentSamplers(
-    render_pass, first_slot, texture_sampler_bindings);
+  SDL_BindGPUFragmentSamplers(render_pass,
+                              first_slot,
+                              texture_sampler_bindings.data(),
+                              narrowU32(texture_sampler_bindings.size()));
 }
 
 inline void GPURenderPass::BindFragmentSamplers(
@@ -6660,7 +6692,6 @@
  * @param render_pass a render pass handle.
  * @param first_slot the fragment storage texture slot to begin binding from.
  * @param storage_textures an array of storage textures.
- * @param num_bindings the number of storage textures to bind from the array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -6671,7 +6702,10 @@
   Uint32 first_slot,
   SpanRef<const GPUTextureRaw> storage_textures)
 {
-  SDL_BindGPUFragmentStorageTextures(render_pass, first_slot, storage_textures);
+  SDL_BindGPUFragmentStorageTextures(render_pass,
+                                     first_slot,
+                                     storage_textures.data(),
+                                     narrowU32(storage_textures.size()));
 }
 
 inline void GPURenderPass::BindFragmentStorageTextures(
@@ -6694,7 +6728,6 @@
  * @param render_pass a render pass handle.
  * @param first_slot the fragment storage buffer slot to begin binding from.
  * @param storage_buffers an array of storage buffers.
- * @param num_bindings the number of storage buffers to bind from the array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -6705,7 +6738,10 @@
   Uint32 first_slot,
   SpanRef<const GPUBufferRaw> storage_buffers)
 {
-  SDL_BindGPUFragmentStorageBuffers(render_pass, first_slot, storage_buffers);
+  SDL_BindGPUFragmentStorageBuffers(render_pass,
+                                    first_slot,
+                                    storage_buffers.data(),
+                                    narrowU32(storage_buffers.size()));
 }
 
 inline void GPURenderPass::BindFragmentStorageBuffers(
@@ -6912,12 +6948,8 @@
  * @param command_buffer a command buffer.
  * @param storage_texture_bindings an array of writeable storage texture binding
  *                                 structs.
- * @param num_storage_texture_bindings the number of storage textures to bind
- *                                     from the array.
  * @param storage_buffer_bindings an array of writeable storage buffer binding
  *                                structs.
- * @param num_storage_buffer_bindings the number of storage buffers to bind from
- *                                    the array.
  * @returns a compute pass handle.
  *
  * @since This function is available since SDL 3.2.0.
@@ -6929,8 +6961,11 @@
   std::span<const GPUStorageTextureReadWriteBinding> storage_texture_bindings,
   std::span<const GPUStorageBufferReadWriteBinding> storage_buffer_bindings)
 {
-  return SDL_BeginGPUComputePass(
-    command_buffer, storage_texture_bindings, storage_buffer_bindings);
+  return SDL_BeginGPUComputePass(command_buffer,
+                                 storage_texture_bindings.data(),
+                                 narrowU32(storage_texture_bindings.size()),
+                                 storage_buffer_bindings.data(),
+                                 narrowU32(storage_buffer_bindings.size()));
 }
 
 inline GPUComputePass GPUCommandBuffer::BeginComputePass(
@@ -6971,8 +7006,6 @@
  * @param compute_pass a compute pass handle.
  * @param first_slot the compute sampler slot to begin binding from.
  * @param texture_sampler_bindings an array of texture-sampler binding structs.
- * @param num_bindings the number of texture-sampler bindings to bind from the
- *                     array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -6983,8 +7016,10 @@
   Uint32 first_slot,
   std::span<const GPUTextureSamplerBinding> texture_sampler_bindings)
 {
-  SDL_BindGPUComputeSamplers(
-    compute_pass, first_slot, texture_sampler_bindings);
+  SDL_BindGPUComputeSamplers(compute_pass,
+                             first_slot,
+                             texture_sampler_bindings.data(),
+                             narrowU32(texture_sampler_bindings.size()));
 }
 
 inline void GPUComputePass::BindSamplers(
@@ -7007,7 +7042,6 @@
  * @param compute_pass a compute pass handle.
  * @param first_slot the compute storage texture slot to begin binding from.
  * @param storage_textures an array of storage textures.
- * @param num_bindings the number of storage textures to bind from the array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -7018,7 +7052,10 @@
   Uint32 first_slot,
   SpanRef<const GPUTextureRaw> storage_textures)
 {
-  SDL_BindGPUComputeStorageTextures(compute_pass, first_slot, storage_textures);
+  SDL_BindGPUComputeStorageTextures(compute_pass,
+                                    first_slot,
+                                    storage_textures.data(),
+                                    narrowU32(storage_textures.size()));
 }
 
 inline void GPUComputePass::BindStorageTextures(
@@ -7041,7 +7078,6 @@
  * @param compute_pass a compute pass handle.
  * @param first_slot the compute storage buffer slot to begin binding from.
  * @param storage_buffers an array of storage buffer binding structs.
- * @param num_bindings the number of storage buffers to bind from the array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -7052,7 +7088,10 @@
   Uint32 first_slot,
   SpanRef<const GPUBufferRaw> storage_buffers)
 {
-  SDL_BindGPUComputeStorageBuffers(compute_pass, first_slot, storage_buffers);
+  SDL_BindGPUComputeStorageBuffers(compute_pass,
+                                   first_slot,
+                                   storage_buffers.data(),
+                                   narrowU32(storage_buffers.size()));
 }
 
 inline void GPUComputePass::BindStorageBuffers(
@@ -7584,7 +7623,7 @@
  * supported via GPUDevice.WindowSupportsPresentMode /
  * GPUDevice.WindowSupportsSwapchainComposition prior to calling this function.
  *
- * GPU_PRESENTMODE_VSYNC with GPU_SWAPCHAINCOMPOSITION_SDR is always supported.
+ * GPU_PRESENTMODE_VSYNC with GPU_SWAPCHAINCOMPOSITION_SDR are always supported.
  *
  * @param device a GPU context.
  * @param window an Window that has been claimed.
@@ -7700,11 +7739,11 @@
  *
  * @param command_buffer a command buffer.
  * @param window a window that has been claimed.
- * @param swapchain_texture a pointer filled in with a swapchain texture handle.
  * @param swapchain_texture_width a pointer filled in with the swapchain texture
  *                                width, may be nullptr.
  * @param swapchain_texture_height a pointer filled in with the swapchain
  *                                 texture height, may be nullptr.
+ * @returns a swapchain texture handle.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called from the thread that
@@ -7727,8 +7766,13 @@
   Uint32* swapchain_texture_width = nullptr,
   Uint32* swapchain_texture_height = nullptr)
 {
-  return CheckError(SDL_AcquireGPUSwapchainTexture(
-    command_buffer, window, swapchain_texture_width, swapchain_texture_height));
+  GPUTextureRaw texture;
+  CheckError(SDL_AcquireGPUSwapchainTexture(command_buffer,
+                                            window,
+                                            &texture,
+                                            swapchain_texture_width,
+                                            swapchain_texture_height));
+  return texture;
 }
 
 inline GPUTexture GPUCommandBuffer::AcquireSwapchainTexture(
@@ -7792,11 +7836,11 @@
  *
  * @param command_buffer a command buffer.
  * @param window a window that has been claimed.
- * @param swapchain_texture a pointer filled in with a swapchain texture handle.
  * @param swapchain_texture_width a pointer filled in with the swapchain texture
  *                                width, may be nullptr.
  * @param swapchain_texture_height a pointer filled in with the swapchain
  *                                 texture height, may be nullptr.
+ * @returns a swapchain texture handle.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called from the thread that
@@ -7814,8 +7858,13 @@
   Uint32* swapchain_texture_width = nullptr,
   Uint32* swapchain_texture_height = nullptr)
 {
-  return CheckError(SDL_WaitAndAcquireGPUSwapchainTexture(
-    command_buffer, window, swapchain_texture_width, swapchain_texture_height));
+  GPUTextureRaw texture;
+  CheckError(SDL_WaitAndAcquireGPUSwapchainTexture(command_buffer,
+                                                   window,
+                                                   &texture,
+                                                   swapchain_texture_width,
+                                                   swapchain_texture_height));
+  return texture;
 }
 
 inline GPUTexture GPUCommandBuffer::WaitAndAcquireSwapchainTexture(
@@ -7949,7 +7998,6 @@
  * @param wait_all if 0, wait for any fence to be signaled, if 1, wait for all
  *                 fences to be signaled.
  * @param fences an array of fences to wait on.
- * @param num_fences the number of fences in the fences array.
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
@@ -7961,7 +8009,8 @@
                              bool wait_all,
                              std::span<GPUFence* const> fences)
 {
-  CheckError(SDL_WaitForGPUFences(device, wait_all, fences));
+  CheckError(SDL_WaitForGPUFences(
+    device, wait_all, fences.data(), narrowU32(fences.size())));
 }
 
 inline void GPUDevice::WaitForFences(bool wait_all,
@@ -8129,6 +8178,8 @@
 
 #endif // SDL_VERSION_ATLEAST(3, 4, 0)
 
+#ifdef SDL_PLATFORM_GDK
+
 /**
  * Call this to suspend GPU operation on Xbox when you receive the
  * EVENT_DID_ENTER_BACKGROUND event.
@@ -8163,6 +8214,8 @@
 
 inline void GPUDevice::GDKResumeGPU() { SDL::GDKResumeGPU(get()); }
 
+#endif /* SDL_PLATFORM_GDK */
+
 /// @}
 
 } // namespace SDL
