--- src/generated/SDL3pp_sensor.h
+++ include/SDL3pp/SDL3pp_sensor.h
@@ -10,7 +10,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategorySensor Category Sensor
+ * @defgroup CategorySensor Sensors
  *
  * SDL sensor management.
  *
@@ -367,15 +367,17 @@
 /**
  * Get a list of currently connected sensors.
  *
- * @param count a pointer filled in with the number of sensors returned, may be
- *              nullptr.
  * @returns a 0 terminated array of sensor instance IDs or nullptr on failure;
- *          call GetError() for more information. This should be freed with
- *          free() when it is no longer needed.
+ *          call GetError() for more information.
  *
  * @since This function is available since SDL 3.2.0.
  */
-inline OwnArray<SensorID> GetSensors() { return SDL_GetSensors(); }
+inline OwnArray<SensorID> GetSensors()
+{
+  int count = 0;
+  auto data = SDL_GetSensors(&count);
+  return OwnArray<SensorID>(data, size_t(count));
+}
 
 /**
  * Get the implementation dependent name of a sensor.
@@ -445,7 +447,7 @@
  */
 inline SensorRef GetSensorFromID(SensorID instance_id)
 {
-  return SDL_GetSensorFromID(instance_id);
+  return {SDL_GetSensorFromID(instance_id)};
 }
 
 /**
@@ -459,7 +461,7 @@
  */
 inline PropertiesRef GetSensorProperties(SensorParam sensor)
 {
-  return CheckError(SDL_GetSensorProperties(sensor));
+  return {CheckError(SDL_GetSensorProperties(sensor))};
 }
 
 inline PropertiesRef Sensor::GetProperties()
