--- src/generated/SDL3pp_storage.h
+++ include/SDL3pp/SDL3pp_storage.h
@@ -9,7 +9,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryStorage Category Storage
+ * @defgroup CategoryStorage Storage Abstraction
  *
  * The storage API is a high-level API designed to abstract away the portability
  * issues that come up when using something lower-level (in SDL's case, this
@@ -103,27 +103,27 @@
  * When using Storage, these types of problems are virtually impossible to trip
  * over:
  *
- * ```c
+ * ```cpp
  * void ReadGameData(void)
  * {
  *     extern char** fileNames;
  *     extern size_t numFiles;
  *
- *     Storage *title = Storage.Storage(nullptr, 0);
+ *     Storage title(nullptr, 0);
  *     if (title == nullptr) {
  *         // Something bad happened!
  *     }
- *     while (!Storage.Ready(title)) {
+ *     while (!title.Ready()) {
  *         Delay(1);
  *     }
  *
  *     for (size_t i = 0; i < numFiles; i += 1) {
  *         void* dst;
- *         Uint64 dstLen = 0;
+ *         Uint64 dstLen = title.GetSize(fileNames[i]);
  *
- *         if (Storage.GetFileSize(title, fileNames[i], &dstLen) && dstLen > 0)
- * { dst = malloc(dstLen); if (Storage.ReadFile(title, fileNames[i], dst,
- * dstLen)) {
+ *         if (dstLen > 0) {
+ *           dst = malloc(dstLen);
+ *           if (title.ReadFile(fileNames[i], dst, dstLen)) {
  *                 // A bunch of stuff happens here
  *             } else {
  *                 // Something bad happened!
@@ -133,24 +133,22 @@
  *             // Something bad happened!
  *         }
  *     }
- *
- *     Storage.Close(title);
  * }
  *
  * void ReadSave(void)
  * {
- *     Storage *user = Storage.Storage("libsdl", "Storage Example", 0);
+ *     Storage user("libsdl", "Storage Example", 0);
  *     if (user == nullptr) {
  *         // Something bad happened!
  *     }
- *     while (!Storage.Ready(user)) {
+ *     while (!user.Ready()) {
  *         Delay(1);
  *     }
  *
- *     Uint64 saveLen = 0;
- *     if (Storage.GetFileSize(user, "save0.sav", &saveLen) && saveLen > 0) {
+ *     Uint64 saveLen = user.GetFileSize();
+ *     if (saveLen > 0) {
  *         void* dst = malloc(saveLen);
- *         if (Storage.ReadFile(user, "save0.sav", dst, saveLen)) {
+ *         if (user.ReadFile("save0.sav", dst, saveLen)) {
  *             // A bunch of stuff happens here
  *         } else {
  *             // Something bad happened!
@@ -159,27 +157,23 @@
  *     } else {
  *         // Something bad happened!
  *     }
- *
- *     Storage.Close(user);
  * }
  *
  * void WriteSave(void)
  * {
- *     Storage *user = Storage.Storage("libsdl", "Storage Example", 0);
+ *     Storage user("libsdl", "Storage Example", 0);
  *     if (user == nullptr) {
  *         // Something bad happened!
  *     }
- *     while (!Storage.Ready(user)) {
+ *     while (!user.Ready()) {
  *         Delay(1);
  *     }
  *
  *     extern void *saveData; // A bunch of stuff happened here...
  *     extern Uint64 saveLen;
- *     if (!Storage.WriteFile(user, "save0.sav", saveData, saveLen)) {
+ *     if (!user.WriteFile("save0.sav", saveData, saveLen)) {
  *         // Something bad happened!
  *     }
- *
- *     Storage.Close(user);
  * }
  * ```
  *
@@ -287,8 +281,8 @@
  * An abstract interface for filesystem access.
  *
  * This is an opaque datatype. One can create this object using standard SDL
- * functions like Storage.Storage or Storage.Storage, etc, or create an object
- * with a custom implementation using Storage.Storage.
+ * functions like Storage.Storage, etc, or create an object with a custom
+ * implementation using Storage.Storage.
  *
  * @since This struct is available since SDL 3.2.0.
  *
@@ -436,7 +430,7 @@
   }
 
   /// Destructor
-  ~Storage() { SDL_CloseStorage(m_resource); }
+  ~Storage() { CheckError(SDL_CloseStorage(m_resource)); }
 
   /// Assignment operator.
   constexpr Storage& operator=(Storage&& other) noexcept
@@ -484,9 +478,6 @@
    * @since This function is available since SDL 3.2.0.
    *
    * @sa Storage.Storage
-   * @sa Storage.Storage
-   * @sa Storage.Storage
-   * @sa Storage.Storage
    */
   bool Close();
 
@@ -508,8 +499,7 @@
    * Query the size of a file within a storage container.
    *
    * @param path the relative path of the file to query.
-   * @param length a pointer to be filled with the file's length.
-   * @returns true if the file could be queried or false on failure; call
+   * @returns the file's length on success or 0 on failure; call
    *          GetError() for more information.
    *
    * @since This function is available since SDL 3.2.0.
@@ -529,7 +519,6 @@
    *
    * @param path the relative path of the file to read.
    * @param destination a client-provided buffer to read the file into.
-   * @param length the length of the destination buffer.
    * @returns true if the file was read or false on failure; call GetError() for
    *          more information.
    *
@@ -550,8 +539,6 @@
    * future release.
    *
    * @param path the relative path of the file to read.
-   * @param destination a client-provided buffer to read the file into.
-   * @param length the length of the destination buffer.
    * @returns true if the file was read or false on failure; call GetError() for
    *          more information.
    *
@@ -563,6 +550,24 @@
    */
   std::string ReadFile(StringParam path);
 
+  /**
+   * Synchronously read a file from a storage container into a client-provided
+   * buffer.
+   *
+   * The value of `length` must match the length of the file exactly; call
+   * Storage.GetFileSize() to get this value. This behavior may be relaxed
+   * in a future release.
+   *
+   * @param path the relative path of the file to read.
+   * @returns the content if the file was read or empty string on failure; call
+   *          GetError() for more information.
+   *
+   * @since This function is available since SDL 3.2.0.
+   *
+   * @sa Storage.GetFileSize
+   * @sa Storage.Ready
+   * @sa Storage.WriteFile
+   */
   template<class T>
   std::vector<T> ReadFileAs(StringParam path);
 
@@ -571,9 +576,7 @@
    *
    * @param path the relative path of the file to write.
    * @param source a client-provided buffer to write from.
-   * @param length the length of the source buffer.
-   * @returns true if the file was written or false on failure; call GetError()
-   *          for more information.
+   * @throws Error on failure.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -641,8 +644,7 @@
    *
    * @param path the path of the directory to enumerate, or nullptr for the
    *             root.
-   * @param callback a function that is called for each entry in the directory.
-   * @param userdata a pointer that is passed to `callback`.
+   * @returns all the directory contents.
    * @throws Error on failure.
    *
    * @since This function is available since SDL 3.2.0.
@@ -669,7 +671,6 @@
    * @param path the path of the directory to enumerate, or nullptr for the
    *             root.
    * @param callback a function that is called for each entry in the directory.
-   * @param userdata a pointer that is passed to `callback`.
    * @throws Error on failure.
    *
    * @since This function is available since SDL 3.2.0.
@@ -720,9 +721,8 @@
    * Get information about a filesystem path in a storage container.
    *
    * @param path the path to query.
-   * @param info a pointer filled in with information about the path, or nullptr
-   *             to check for the existence of a file.
-   * @throws Error on failure.
+   * @returns the info on success or false if the file doesn't exist, or another
+   *          failure; call GetError() for more information.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -767,8 +767,6 @@
    * @param pattern the pattern that files in the directory must match. Can be
    *                nullptr.
    * @param flags `SDL_GLOB_*` bitflags that affect this search.
-   * @param count on return, will be set to the number of items in the returned
-   *              array. Can be nullptr.
    * @returns an array of strings on success.
    * @throws Error on failure.
    *
@@ -978,8 +976,7 @@
  *
  * @param storage a storage container to query.
  * @param path the relative path of the file to query.
- * @param length a pointer to be filled with the file's length.
- * @returns true if the file could be queried or false on failure; call
+ * @returns the file's length on success or std::nullopt on failure; call
  *          GetError() for more information.
  *
  * @since This function is available since SDL 3.2.0.
@@ -990,7 +987,10 @@
 inline std::optional<Uint64> GetStorageFileSize(StorageParam storage,
                                                 StringParam path)
 {
-  return SDL_GetStorageFileSize(storage, path);
+  if (Uint64 length; SDL_GetStorageFileSize(storage, path, &length)) {
+    return length;
+  }
+  return {};
 }
 
 inline std::optional<Uint64> Storage::GetFileSize(StringParam path)
@@ -1002,14 +1002,13 @@
  * Synchronously read a file from a storage container into a client-provided
  * buffer.
  *
- * The value of `length` must match the length of the file exactly; call
- * Storage.GetFileSize() to get this value. This behavior may be relaxed in a
- * future release.
+ * The value of `destination.size()` must match the length of the file exactly;
+ * call Storage.GetFileSize() to get this value. This behavior may be relaxed in
+ * a future release.
  *
  * @param storage a storage container to read from.
  * @param path the relative path of the file to read.
  * @param destination a client-provided buffer to read the file into.
- * @param length the length of the destination buffer.
  * @returns true if the file was read or false on failure; call GetError() for
  *          more information.
  *
@@ -1031,16 +1030,10 @@
  * Synchronously read a file from a storage container into a client-provided
  * buffer.
  *
- * The value of `length` must match the length of the file exactly; call
- * Storage.GetFileSize() to get this value. This behavior may be relaxed in a
- * future release.
- *
  * @param storage a storage container to read from.
  * @param path the relative path of the file to read.
- * @param destination a client-provided buffer to read the file into.
- * @param length the length of the destination buffer.
- * @returns true if the file was read or false on failure; call GetError() for
- *          more information.
+ * @returns the content if the file was read.
+ * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -1050,7 +1043,11 @@
  */
 inline std::string ReadStorageFile(StorageParam storage, StringParam path)
 {
-  static_assert(false, "Not implemented");
+  auto sz = GetStorageFileSize(storage, path.c_str());
+  if (!sz || *sz == 0) return {};
+  std::string buffer(*sz, 0);
+  CheckError(ReadStorageFile(storage, std::move(path), buffer));
+  return buffer;
 }
 
 inline bool Storage::ReadFile(StringParam path, TargetBytes destination)
@@ -1064,16 +1061,35 @@
   return SDL::ReadStorageFile(m_resource, std::move(path));
 }
 
+/**
+ * Synchronously read a file from a storage container into a client-provided
+ * buffer.
+ *
+ * @param storage a storage container to read from.
+ * @param path the relative path of the file to read.
+ * @returns the content if the file was read.
+ * @throws Error on failure.
+ *
+ * @since This function is available since SDL 3.2.0.
+ *
+ * @sa Storage.GetFileSize
+ * @sa Storage.Ready
+ * @sa Storage.WriteFile
+ */
 template<class T>
 inline std::vector<T> ReadStorageFileAs(StorageParam storage, StringParam path)
 {
-  static_assert(false, "Not implemented");
+  auto sz = GetStorageFileSize(storage, path.c_str());
+  if (!sz || *sz == 0) return {};
+  std::vector<T> buffer(*sz / sizeof(T) + (*sz % sizeof(T) ? 1 : 0), 0);
+  CheckError(ReadFile(std::move(path), {buffer.data(), *sz}));
+  return buffer;
 }
 
 template<class T>
 inline std::vector<T> Storage::ReadFileAs(StringParam path)
 {
-  return SDL::ReadStorageFileAs(m_resource, std::move(path));
+  return SDL::ReadStorageFileAs<T>(m_resource, std::move(path));
 }
 
 /**
@@ -1082,9 +1098,7 @@
  * @param storage a storage container to write to.
  * @param path the relative path of the file to write.
  * @param source a client-provided buffer to write from.
- * @param length the length of the source buffer.
- * @returns true if the file was written or false on failure; call GetError()
- *          for more information.
+ * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -1096,7 +1110,8 @@
                              StringParam path,
                              SourceBytes source)
 {
-  SDL_WriteStorageFile(storage, path, source.data(), source.size_bytes());
+  CheckError(
+    SDL_WriteStorageFile(storage, path, source.data(), source.size_bytes()));
 }
 
 inline void Storage::WriteFile(StringParam path, SourceBytes source)
@@ -1174,7 +1189,6 @@
  * @param storage a storage container.
  * @param path the path of the directory to enumerate, or nullptr for the root.
  * @param callback a function that is called for each entry in the directory.
- * @param userdata a pointer that is passed to `callback`.
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
@@ -1185,7 +1199,14 @@
                                       StringParam path,
                                       EnumerateDirectoryCB callback)
 {
-  static_assert(false, "Not implemented");
+  EnumerateStorageDirectory(
+    storage,
+    std::move(path),
+    [](void* userdata, const char* dirname, const char* fname) {
+      auto& cb = *static_cast<EnumerateDirectoryCB*>(userdata);
+      return cb(dirname, fname);
+    },
+    &callback);
 }
 
 /**
@@ -1204,8 +1225,7 @@
  *
  * @param storage a storage container.
  * @param path the path of the directory to enumerate, or nullptr for the root.
- * @param callback a function that is called for each entry in the directory.
- * @param userdata a pointer that is passed to `callback`.
+ * @returns all the directory contents.
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
@@ -1215,7 +1235,12 @@
 inline std::vector<Path> EnumerateStorageDirectory(StorageParam storage,
                                                    StringParam path)
 {
-  static_assert(false, "Not implemented");
+  std::vector<Path> r;
+  EnumerateDirectory(std::move(path), [&](const char*, const char* fname) {
+    r.emplace_back(fname);
+    return ENUM_CONTINUE;
+  });
+  return r;
 }
 
 inline void Storage::EnumerateDirectory(StringParam path,
@@ -1311,9 +1336,8 @@
  *
  * @param storage a storage container.
  * @param path the path to query.
- * @param info a pointer filled in with information about the path, or nullptr
- *             to check for the existence of a file.
- * @throws Error on failure.
+ * @returns the info on success or false if the file doesn't exist, or another
+ *          failure; call GetError() for more information.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -1321,7 +1345,10 @@
  */
 inline PathInfo GetStoragePathInfo(StorageParam storage, StringParam path)
 {
-  return CheckError(SDL_GetStoragePathInfo(storage, path));
+  if (PathInfo info; SDL_GetStoragePathInfo(storage, path, &info)) {
+    return info;
+  }
+  return {};
 }
 
 inline PathInfo Storage::GetPathInfo(StringParam path)
@@ -1375,8 +1402,6 @@
  * @param pattern the pattern that files in the directory must match. Can be
  *                nullptr.
  * @param flags `SDL_GLOB_*` bitflags that affect this search.
- * @param count on return, will be set to the number of items in the returned
- *              array. Can be nullptr.
  * @returns an array of strings on success.
  * @throws Error on failure.
  *
@@ -1390,7 +1415,10 @@
                                             StringParam pattern,
                                             GlobFlags flags)
 {
-  return CheckError(SDL_GlobStorageDirectory(storage, path, pattern, flags));
+  int count;
+  auto data =
+    CheckError(SDL_GlobStorageDirectory(storage, path, pattern, flags, &count));
+  return OwnArray<char*>{data, size_t(count)};
 }
 
 inline OwnArray<char*> Storage::GlobDirectory(StringParam path,
