--- src/generated/SDL3pp_gamepad.h
+++ include/SDL3pp/SDL3pp_gamepad.h
@@ -14,7 +14,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryGamepad Category Gamepad
+ * @defgroup CategoryGamepad Gamepad Support
  *
  * SDL provides a low-level joystick API, which just treats joysticks as an
  * arbitrary pile of buttons, axes, and hat switches. If you're planning to
@@ -523,8 +523,7 @@
    * Details about mappings are discussed with AddGamepadMapping().
    *
    * @returns a string that has the gamepad's mapping or nullptr if no mapping
-   *          is available; call GetError() for more information. This should be
-   *          freed with free() when it is no longer needed.
+   *          is available; call GetError() for more information.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
@@ -1348,26 +1347,26 @@
 /**
  * Get the current gamepad mappings.
  *
- * @param count a pointer filled in with the number of mappings returned, can be
- *              nullptr.
  * @returns an array of the mapping strings, nullptr-terminated, or nullptr on
- *          failure; call GetError() for more information. This is a single
- *          allocation that should be freed with free() when it is no longer
- *          needed.
+ *          failure; call GetError() for more information.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
  * @since This function is available since SDL 3.2.0.
  */
-inline OwnArray<char*> GetGamepadMappings() { return SDL_GetGamepadMappings(); }
+inline OwnArray<char*> GetGamepadMappings()
+{
+  int count;
+  auto data = SDL_GetGamepadMappings(&count);
+  return OwnArray<char*>(data);
+}
 
 /**
  * Get the gamepad mapping string for a given GUID.
  *
  * @param guid a structure containing the GUID for which a mapping is desired.
  * @returns a mapping string or nullptr on failure; call GetError() for more
- *          information. This should be freed with free() when it is no longer
- *          needed.
+ *          information.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -1378,7 +1377,7 @@
  */
 inline StringResult GetGamepadMappingForGUID(GUID guid)
 {
-  return SDL_GetGamepadMappingForGUID(guid);
+  return StringResult(SDL_GetGamepadMappingForGUID(guid));
 }
 
 /**
@@ -1388,8 +1387,7 @@
  *
  * @param gamepad the gamepad you want to get the current mapping for.
  * @returns a string that has the gamepad's mapping or nullptr if no mapping is
- *          available; call GetError() for more information. This should be
- *          freed with free() when it is no longer needed.
+ *          available; call GetError() for more information.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -1402,7 +1400,7 @@
  */
 inline StringResult GetGamepadMapping(GamepadParam gamepad)
 {
-  return SDL_GetGamepadMapping(gamepad);
+  return StringResult(SDL_GetGamepadMapping(gamepad));
 }
 
 inline StringResult Gamepad::GetMapping()
@@ -1448,11 +1446,8 @@
 /**
  * Get a list of currently connected gamepads.
  *
- * @param count a pointer filled in with the number of gamepads returned, may be
- *              nullptr.
  * @returns a 0 terminated array of joystick instance IDs or nullptr on failure;
- *          call GetError() for more information. This should be freed with
- *          free() when it is no longer needed.
+ *          call GetError() for more information.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -1461,7 +1456,12 @@
  * @sa HasGamepad
  * @sa Gamepad.Gamepad
  */
-inline OwnArray<JoystickID> GetGamepads() { return SDL_GetGamepads(); }
+inline OwnArray<JoystickID> GetGamepads()
+{
+  int count;
+  auto r = reinterpret_cast<JoystickID*>(SDL_GetGamepads(&count));
+  return OwnArray<JoystickID>(r, count);
+}
 
 /**
  * Check if the given joystick is supported by the gamepad interface.
@@ -1727,7 +1727,7 @@
  */
 inline GamepadRef GetGamepadFromID(JoystickID instance_id)
 {
-  return CheckError(SDL_GetGamepadFromID(instance_id));
+  return {CheckError(SDL_GetGamepadFromID(instance_id))};
 }
 
 /**
@@ -1745,7 +1745,7 @@
  */
 inline GamepadRef GetGamepadFromPlayerIndex(int player_index)
 {
-  return SDL_GetGamepadFromPlayerIndex(player_index);
+  return {SDL_GetGamepadFromPlayerIndex(player_index)};
 }
 
 /**
@@ -1776,7 +1776,7 @@
  */
 inline PropertiesRef GetGamepadProperties(GamepadParam gamepad)
 {
-  return CheckError(SDL_GetGamepadProperties(gamepad));
+  return {CheckError(SDL_GetGamepadProperties(gamepad))};
 }
 
 inline PropertiesRef Gamepad::GetProperties()
@@ -2181,7 +2181,7 @@
  */
 inline JoystickRef GetGamepadJoystick(GamepadParam gamepad)
 {
-  return SDL_GetGamepadJoystick(gamepad);
+  return {SDL_GetGamepadJoystick(gamepad)};
 }
 
 inline JoystickRef Gamepad::GetJoystick()
