--- src/generated/SDL3pp_mixer.h
+++ include/SDL3pp/SDL3pp_mixer.h
@@ -1,10 +1,13 @@
 #ifndef SDL3PP_MIXER_H_
 #define SDL3PP_MIXER_H_
 
-#include <SDL3/SDL_mixer.h>
 #include "SDL3pp_audio.h"
 #include "SDL3pp_version.h"
 
+#if defined(SDL3PP_ENABLE_MIXER) || defined(SDL3PP_DOC)
+
+#include <SDL3_mixer/SDL_mixer.h>
+
 namespace SDL {
 
 /**
@@ -3413,6 +3416,8 @@
   constexpr operator GroupRaw() const noexcept { return get(); }
 };
 
+#ifdef SDL3PP_DOC
+
 /**
  * The current major version of SDL_mixer headers.
  *
@@ -3462,6 +3467,8 @@
    (SDL_MIXER_MAJOR_VERSION > X || SDL_MIXER_MINOR_VERSION > Y ||              \
     SDL_MIXER_MICRO_VERSION >= Z))
 
+#endif // SDL3PP_DOC
+
 namespace MIX {
 
 /**
@@ -4251,8 +4258,8 @@
                                 const AudioSpec& spec,
                                 bool freeWhenDone = false)
 {
-  return MIX_LoadRawAudioNoCopy(
-    mixer, data.data(), data.size_bytes(), &spec, freeWhenDone);
+  return Audio(MIX_LoadRawAudioNoCopy(
+    mixer, data.data(), data.size_bytes(), &spec, freeWhenDone));
 }
 
 inline Audio Mixer::LoadRawAudioNoCopy(SourceBytes data,
@@ -4303,7 +4310,7 @@
                                  float amplitude,
                                  Sint64 ms)
 {
-  return MIX_CreateSineWaveAudio(mixer, hz, amplitude, ms);
+  return Audio(MIX_CreateSineWaveAudio(mixer, hz, amplitude, ms));
 }
 
 inline Audio Mixer::CreateSineWaveAudio(int hz, float amplitude, Sint64 ms)
@@ -7295,4 +7302,6 @@
 
 } // namespace SDL
 
+#endif /* defined(SDL3PP_ENABLE_MIXER) || defined(SDL3PP_DOC) */
+
 #endif /* SDL3PP_MIXER_H_ */
