--- build/generated/SDL3pp_camera.h
+++ include/SDL3pp/SDL3pp_camera.h
@@ -11,7 +11,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryCamera Category Camera
+ * @defgroup CategoryCamera Camera Support
  *
  * Video capture for the SDL library.
  *
@@ -133,6 +133,16 @@
 
 constexpr CameraPermissionState CAMERA_PERMISSION_STATE_APPROVED =
   SDL_CAMERA_PERMISSION_STATE_APPROVED; ///< CAMERA_PERMISSION_STATE_APPROVED
+#else
+
+/**
+ * The current state of a request for camera access.
+ *
+ * @since This enum is available since SDL 3.4.0.
+ *
+ * @sa Camera.GetPermissionState
+ */
+using CameraPermissionState = int;
 
 #endif // SDL_VERSION_ATLEAST(3, 4, 0)
 
@@ -217,9 +227,9 @@
    *
    * @param instance_id the camera device instance ID.
    * @param spec the desired format for data the device will provide. Can be
-   *             nullptr.
-   * @post an Camera object or nullptr on failure; call GetError() for more
-   *       information.
+   *             std::nullopt.
+   * @post an Camera object on success..
+   * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
@@ -347,8 +357,7 @@
    * (or EVENT_CAMERA_DEVICE_DENIED) event, or poll Camera.GetPermissionState()
    * occasionally until it returns non-zero.
    *
-   * @param spec the CameraSpec to be initialized by this function.
-   * @throws Error on failure.
+   * @returns the CameraSpec or std::nullopt if waiting for user approval.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
@@ -417,7 +426,7 @@
    * The app should not use the surface again after calling this function;
    * assume the surface is freed and the pointer is invalid.
    *
-   * @param frame the video frame surface to release.
+   * @param lock the video frame surface to release.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
@@ -500,8 +509,12 @@
   constexpr operator CameraRaw() const noexcept { return get(); }
 };
 
-struct CameraFrame : Surface
+/// Camera Frame.
+class CameraFrame : public Surface
 {
+  CameraRef m_lock;
+
+public:
   /**
    * Acquire a frame.
    *
@@ -531,7 +544,7 @@
    * EVENT_CAMERA_DEVICE_DENIED) event, or poll Camera.GetPermissionState()
    * occasionally until it returns non-zero.
    *
-   * @param camera opened camera device.
+   * @param resource opened camera device.
    * @param timestampNS a pointer filled in with the frame's timestamp, or 0 on
    *                    error. Can be nullptr.
    * @post a new frame of video on success, nullptr if none is currently
@@ -543,14 +556,15 @@
    *
    * @sa Camera.ReleaseFrame
    */
-  CameraFrame(CameraRef resource);
+  CameraFrame(CameraRef resource, Uint64* timestampNS = nullptr);
 
   /// Copy constructor
   CameraFrame(const CameraFrame& other) = delete;
 
   /// Move constructor
   CameraFrame(CameraFrame&& other) noexcept
-    : m_lock(std::move(other.m_lock))
+    : Surface(std::move(other))
+    , m_lock(std::move(other.m_lock))
   {
   }
 
@@ -571,9 +585,6 @@
    * The app should not use the surface again after calling this function;
    * assume the surface is freed and the pointer is invalid.
    *
-   * @param camera opened camera device.
-   * @param frame the video frame surface to release.
-   *
    * @threadsafety It is safe to call this function from any thread.
    *
    * @since This function is available since SDL 3.2.0.
@@ -592,7 +603,10 @@
   }
 
   /// True if not locked.
-  constexpr operator bool() const { return bool(m_lock); }
+  constexpr operator bool() const
+  {
+    return bool(m_lock) && Surface::operator bool();
+  }
 
   /**
    * Release a frame of video acquired from a camera.
@@ -611,9 +625,6 @@
    * The app should not use the surface again after calling this function;
    * assume the surface is freed and the pointer is invalid.
    *
-   * @param camera opened camera device.
-   * @param frame the video frame surface to release.
-   *
    * @threadsafety It is safe to call this function from any thread.
    *
    * @since This function is available since SDL 3.2.0.
@@ -626,7 +637,11 @@
   CameraRef get() const { return m_lock; }
 
   /// Releases the lock without unlocking.
-  void release() { m_lock.release(); }
+  void release()
+  {
+    Surface::release();
+    m_lock.release();
+  }
 };
 
 /**
@@ -701,11 +716,8 @@
 /**
  * Get a list of currently connected camera devices.
  *
- * @param count a pointer filled in with the number of cameras returned, may be
- *              nullptr.
  * @returns a 0 terminated array of camera instance IDs or nullptr on failure;
- *          call GetError() for more information. This should be freed with
- *          free() when it is no longer needed.
+ *          call GetError() for more information.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -713,7 +725,12 @@
  *
  * @sa Camera.Camera
  */
-inline OwnArray<CameraID> GetCameras() { return SDL_GetCameras(); }
+inline OwnArray<CameraID> GetCameras()
+{
+  int count;
+  auto data = SDL_GetCameras(&count);
+  return OwnArray<CameraID>(data, count);
+}
 
 /**
  * Get the list of native formats/sizes a camera supports.
@@ -737,12 +754,8 @@
  * warning popup.
  *
  * @param instance_id the camera device instance ID.
- * @param count a pointer filled in with the number of elements in the list, may
- *              be nullptr.
  * @returns a nullptr terminated array of pointers to CameraSpec or nullptr on
- *          failure; call GetError() for more information. This is a single
- *          allocation that should be freed with free() when it is no longer
- *          needed.
+ *          failure; call GetError() for more information.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -753,7 +766,9 @@
  */
 inline OwnArray<CameraSpec*> GetCameraSupportedFormats(CameraID instance_id)
 {
-  return SDL_GetCameraSupportedFormats(instance_id);
+  int count;
+  auto data = SDL_GetCameraSupportedFormats(instance_id, &count);
+  return OwnArray<CameraSpec*>(data, count);
 }
 
 /**
@@ -829,9 +844,9 @@
  *
  * @param instance_id the camera device instance ID.
  * @param spec the desired format for data the device will provide. Can be
- *             nullptr.
- * @returns an Camera object or nullptr on failure; call GetError() for more
- *          information.
+ *             std::nullopt.
+ * @returns an Camera object on success.
+ * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -926,7 +941,7 @@
  */
 inline PropertiesRef GetCameraProperties(CameraRef camera)
 {
-  return CheckError(SDL_GetCameraProperties(camera));
+  return {CheckError(SDL_GetCameraProperties(camera))};
 }
 
 inline PropertiesRef Camera::GetProperties()
@@ -947,8 +962,7 @@
  * occasionally until it returns non-zero.
  *
  * @param camera opened camera device.
- * @param spec the CameraSpec to be initialized by this function.
- * @throws Error on failure.
+ * @returns the CameraSpec or std::nullopt if waiting for user approval.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -958,7 +972,8 @@
  */
 inline std::optional<CameraSpec> GetCameraFormat(CameraRef camera)
 {
-  return CheckError(SDL_GetCameraFormat(camera));
+  if (CameraSpec spec; SDL_GetCameraFormat(camera, &spec)) return spec;
+  return std::nullopt;
 }
 
 inline std::optional<CameraSpec> Camera::GetFormat()
@@ -1009,7 +1024,7 @@
 inline Surface AcquireCameraFrame(CameraRef camera,
                                   Uint64* timestampNS = nullptr)
 {
-  return SDL_AcquireCameraFrame(camera, timestampNS);
+  return Surface::Borrow(SDL_AcquireCameraFrame(camera, timestampNS));
 }
 
 inline CameraFrame Camera::AcquireFrame(Uint64* timestampNS)
@@ -1017,10 +1032,11 @@
   return {CameraRef(*this)};
 }
 
-inline CameraFrame::CameraFrame(CameraRef resource)
-  : m_lock(std::move(resource))
+inline CameraFrame::CameraFrame(CameraRef resource, Uint64* timestampNS)
+  : Surface(AcquireCameraFrame(resource, timestampNS))
+  , m_lock(std::move(resource))
 {
-  AcquireCameraFrame(m_lock);
+  if (!*this) m_lock.release();
 }
 
 /**
@@ -1063,7 +1079,7 @@
 inline void CameraFrame::reset()
 {
   if (!m_lock) return;
-  ReleaseCameraFrame(m_lock);
+  ReleaseCameraFrame(m_lock, Surface::release());
   m_lock = {};
 }
 
