--- build/generated/SDL3pp_hidapi.h
+++ include/SDL3pp/SDL3pp_hidapi.h
@@ -9,7 +9,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryHIDAPI Category HIDAPI
+ * @defgroup CategoryHIDAPI HIDAPI
  *
  * Header file for SDL HIDAPI functions.
  *
@@ -64,26 +64,33 @@
 constexpr hid_bus_type HID_API_BUS_UNKNOWN =
   SDL_HID_API_BUS_UNKNOWN; ///< Unknown bus type
 
-constexpr hid_bus_type HID_API_BUS_USB =
-  SDL_HID_API_BUS_USB; ///< USB bus Specifications: https://usb.org/hid
+/**
+ * USB bus Specifications:
+ *
+ * - https://usb.org/hid
+ */
+constexpr hid_bus_type HID_API_BUS_USB = SDL_HID_API_BUS_USB;
 
 /**
  * Bluetooth or Bluetooth LE bus Specifications:
- * https://www.bluetooth.com/specifications/specs/human-interface-device-profile-1-1-1/
- * https://www.bluetooth.com/specifications/specs/hid-service-1-0/
- * https://www.bluetooth.com/specifications/specs/hid-over-gatt-profile-1-0/
+ *
+ * * https://www.bluetooth.com/specifications/specs/human-interface-device-profile-1-1-1/
+ * * https://www.bluetooth.com/specifications/specs/hid-service-1-0/
+ * * https://www.bluetooth.com/specifications/specs/hid-over-gatt-profile-1-0/
  */
 constexpr hid_bus_type HID_API_BUS_BLUETOOTH = SDL_HID_API_BUS_BLUETOOTH;
 
 /**
  * I2C bus Specifications:
- * https://docs.microsoft.com/previous-versions/windows/hardware/design/dn642101(v=vs.85)
+ *
+ * * https://docs.microsoft.com/previous-versions/windows/hardware/design/dn642101(v=vs.85)
  */
 constexpr hid_bus_type HID_API_BUS_I2C = SDL_HID_API_BUS_I2C;
 
 /**
  * SPI bus Specifications:
- * https://www.microsoft.com/download/details.aspx?id=103325
+ *
+ * - https://www.microsoft.com/download/details.aspx?id=103325
  */
 constexpr hid_bus_type HID_API_BUS_SPI = SDL_HID_API_BUS_SPI;
 
@@ -106,8 +113,7 @@
   /**
    * Close a HID device.
    *
-   * @returns 0 on success or a negative error code on failure; call GetError()
-   *          for more information.
+   * @throws Error on failure.
    *
    * @since This function is available since SDL 3.2.0.
    */
@@ -150,7 +156,6 @@
    *
    * @param data the data to send, including the report number as the first
    *             byte.
-   * @param length the length in bytes of the data to send.
    * @returns the actual number of bytes written and -1 on on failure; call
    *          GetError() for more information.
    *
@@ -166,10 +171,7 @@
    * reports.
    *
    * @param data a buffer to put the read data into.
-   * @param length the number of bytes to read. For devices with multiple
-   *               reports, make sure to read an extra byte for the report
-   *               number.
-   * @param milliseconds timeout in milliseconds or -1 for blocking wait.
+   * @param timeout timeout in milliseconds
    * @returns the actual number of bytes read and -1 on on failure; call
    *          GetError() for more information. If no packet was available to be
    *          read within the timeout period, this function returns 0.
@@ -186,9 +188,6 @@
    * reports.
    *
    * @param data a buffer to put the read data into.
-   * @param length the number of bytes to read. For devices with multiple
-   *               reports, make sure to read an extra byte for the report
-   *               number.
    * @returns the actual number of bytes read and -1 on failure; call GetError()
    *          for more information. If no packet was available to be read and
    *          the handle is in non-blocking mode, this function returns 0.
@@ -206,9 +205,8 @@
    *
    * Nonblocking can be turned on and off at any time.
    *
-   * @param nonblock enable or not the nonblocking reads - 1 to enable
-   *                 nonblocking - 0 to disable nonblocking.
-   * @returns 0 on success.
+   * @param nonblock enable or not the nonblocking reads - true to enable
+   *                 nonblocking - false to disable nonblocking.
    * @throws Error on failure.
    *
    * @since This function is available since SDL 3.2.0.
@@ -230,8 +228,6 @@
    *
    * @param data the data to send, including the report number as the first
    *             byte.
-   * @param length the length in bytes of the data to send, including the report
-   *               number.
    * @returns the actual number of bytes written and -1 on failure; call
    *          GetError() for more information.
    *
@@ -251,8 +247,6 @@
    *             Set the first byte of `data` to the Report ID of the report to
    *             be read, or set it to zero if your device does not use numbered
    *             reports.
-   * @param length the number of bytes to read, including an extra byte for the
-   *               report ID. The buffer can be longer than the actual report.
    * @returns the number of bytes read plus one for the report ID (which is
    *          still in the first byte), or -1 on on failure; call GetError() for
    *          more information.
@@ -273,8 +267,6 @@
    *             Set the first byte of `data` to the Report ID of the report to
    *             be read, or set it to zero if your device does not use numbered
    *             reports.
-   * @param length the number of bytes to read, including an extra byte for the
-   *               report ID. The buffer can be longer than the actual report.
    * @returns the number of bytes read plus one for the report ID (which is
    *          still in the first byte), or -1 on on failure; call GetError() for
    *          more information.
@@ -288,7 +280,6 @@
    *
    * @param string a wide string buffer to put the data into.
    * @param maxlen the length of the buffer in multiples of wchar_t.
-   * @returns 0 on success.
    * @throws Error on failure.
    *
    * @since This function is available since SDL 3.2.0.
@@ -300,7 +291,6 @@
    *
    * @param string a wide string buffer to put the data into.
    * @param maxlen the length of the buffer in multiples of wchar_t.
-   * @returns 0 on success.
    * @throws Error on failure.
    *
    * @since This function is available since SDL 3.2.0.
@@ -312,7 +302,6 @@
    *
    * @param string a wide string buffer to put the data into.
    * @param maxlen the length of the buffer in multiples of wchar_t.
-   * @returns 0 on success.
    * @throws Error on failure.
    *
    * @since This function is available since SDL 3.2.0.
@@ -325,7 +314,6 @@
    * @param string_index the index of the string to get.
    * @param string a wide string buffer to put the data into.
    * @param maxlen the length of the buffer in multiples of wchar_t.
-   * @returns 0 on success.
    * @throws Error on failure.
    *
    * @since This function is available since SDL 3.2.0.
@@ -335,9 +323,10 @@
   /**
    * Get the device info from a HID device.
    *
-   * @returns a pointer to the hid_device_info for this hid_device or nullptr on
-   *          failure; call GetError() for more information. This struct is
-   *          valid until the device is closed with hid_close().
+   * @returns a pointer to the hid_device_info for this hid_device on success.
+   *          This struct is valid until the device is closed with
+   *          HidDevice.close().
+   * @throws Error on failure.
    *
    * @since This function is available since SDL 3.2.0.
    */
@@ -350,7 +339,6 @@
    * to. The recommended size for a preallocated buffer is 4096 bytes.
    *
    * @param buf the buffer to copy descriptor into.
-   * @param buf_size the size of the buffer in bytes.
    * @returns the number of bytes actually copied or -1 on failure; call
    *          GetError() for more information.
    *
@@ -402,7 +390,6 @@
    * @param product_id the Product ID (PID) of the device to open.
    * @param serial_number the Serial Number of the device to open (Optionally
    *                      nullptr).
-   * @post a pointer to a HidDevice object on success.
    * @throws Error on failure.
    *
    * @since This function is available since SDL 3.2.0.
@@ -418,7 +405,6 @@
    * platform-specific path name can be used (eg: /dev/hidraw0 on Linux).
    *
    * @param path the path name of the device to open.
-   * @post a pointer to a HidDevice object on success.
    * @throws Error on failure.
    *
    * @since This function is available since SDL 3.2.0.
@@ -450,8 +436,6 @@
  *
  * Each call to this function should have a matching call to hid_exit()
  *
- * @returns 0 on success.
- *
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
@@ -466,8 +450,6 @@
  * This function frees all of the static data associated with HIDAPI. It should
  * be called at the end of execution to avoid memory leaks.
  *
- * @returns 0 on success.
- *
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
@@ -518,9 +500,9 @@
  * @param product_id the Product ID (PID) of the types of device to open, or 0
  *                   to match any product.
  * @returns a pointer to a linked list of type hid_device_info, containing
- *          information about the HID devices attached to the system, or nullptr
- *          in the case of failure. Free this linked list by calling
- *          hid_free_enumeration().
+ *          information about the HID devices attached to the system on success.
+ *          Free this linked list by calling hid_free_enumeration().
+ * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -557,7 +539,7 @@
  * @param product_id the Product ID (PID) of the device to open.
  * @param serial_number the Serial Number of the device to open (Optionally
  *                      nullptr).
- * @returns a pointer to a HidDevice object on success.
+ * @returns a HidDevice object on success.
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
@@ -588,7 +570,7 @@
  * platform-specific path name can be used (eg: /dev/hidraw0 on Linux).
  *
  * @param path the path name of the device to open.
- * @returns a pointer to a HidDevice object on success.
+ * @returns a HidDevice object on success.
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
@@ -624,6 +606,13 @@
   return SDL::hid_get_properties(get());
 }
 
+/**
+ * HIDAPI properties.
+ *
+ * @since This namespace is available since SDL 3.4.0.
+ *
+ * @sa Hidapi.GetProperties
+ */
 namespace prop::Hidapi {
 
 constexpr auto LIBUSB_DEVICE_HANDLE_POINTER =
@@ -650,7 +639,6 @@
  *
  * @param dev a device handle returned from hid_open().
  * @param data the data to send, including the report number as the first byte.
- * @param length the length in bytes of the data to send.
  * @returns the actual number of bytes written and -1 on on failure; call
  *          GetError() for more information.
  *
@@ -658,7 +646,7 @@
  */
 inline int hid_write(HidDeviceRef dev, SourceBytes data)
 {
-  return SDL_hid_write(dev, data.data(), data.size_bytes());
+  return SDL_hid_write(dev, data.data_as<Uint8>(), data.size_bytes());
 }
 
 inline int HidDeviceBase::write(SourceBytes data)
@@ -675,9 +663,7 @@
  *
  * @param dev a device handle returned from hid_open().
  * @param data a buffer to put the read data into.
- * @param length the number of bytes to read. For devices with multiple reports,
- *               make sure to read an extra byte for the report number.
- * @param milliseconds timeout in milliseconds or -1 for blocking wait.
+ * @param timeout timeout in milliseconds
  * @returns the actual number of bytes read and -1 on on failure; call
  *          GetError() for more information. If no packet was available to be
  *          read within the timeout period, this function returns 0.
@@ -688,7 +674,8 @@
                             TargetBytes data,
                             Milliseconds timeout)
 {
-  return SDL_hid_read_timeout(dev, data.data(), data.size_bytes(), timeout);
+  return SDL_hid_read_timeout(
+    dev, data.data_as<Uint8>(), data.size_bytes(), narrowS32(timeout.count()));
 }
 
 inline int HidDeviceBase::read_timeout(TargetBytes data, Milliseconds timeout)
@@ -705,8 +692,6 @@
  *
  * @param dev a device handle returned from hid_open().
  * @param data a buffer to put the read data into.
- * @param length the number of bytes to read. For devices with multiple reports,
- *               make sure to read an extra byte for the report number.
  * @returns the actual number of bytes read and -1 on failure; call GetError()
  *          for more information. If no packet was available to be read and the
  *          handle is in non-blocking mode, this function returns 0.
@@ -715,7 +700,7 @@
  */
 inline int hid_read(HidDeviceRef dev, TargetBytes data)
 {
-  return SDL_hid_read(dev, data.data(), data.size_bytes());
+  return SDL_hid_read(dev, data.data_as<Uint8>(), data.size_bytes());
 }
 
 inline int HidDeviceBase::read(TargetBytes data)
@@ -733,9 +718,8 @@
  * Nonblocking can be turned on and off at any time.
  *
  * @param dev a device handle returned from hid_open().
- * @param nonblock enable or not the nonblocking reads - 1 to enable nonblocking
- *                 - 0 to disable nonblocking.
- * @returns 0 on success.
+ * @param nonblock enable or not the nonblocking reads - true to enable
+ *                 nonblocking - false to disable nonblocking.
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
@@ -765,8 +749,6 @@
  *
  * @param dev a device handle returned from hid_open().
  * @param data the data to send, including the report number as the first byte.
- * @param length the length in bytes of the data to send, including the report
- *               number.
  * @returns the actual number of bytes written and -1 on failure; call
  *          GetError() for more information.
  *
@@ -774,7 +756,8 @@
  */
 inline int hid_send_feature_report(HidDeviceRef dev, SourceBytes data)
 {
-  return SDL_hid_send_feature_report(dev, data.data(), data.size_bytes());
+  return SDL_hid_send_feature_report(
+    dev, data.data_as<Uint8>(), data.size_bytes());
 }
 
 inline int HidDeviceBase::send_feature_report(SourceBytes data)
@@ -795,8 +778,6 @@
  *             the first byte of `data` to the Report ID of the report to be
  *             read, or set it to zero if your device does not use numbered
  *             reports.
- * @param length the number of bytes to read, including an extra byte for the
- *               report ID. The buffer can be longer than the actual report.
  * @returns the number of bytes read plus one for the report ID (which is still
  *          in the first byte), or -1 on on failure; call GetError() for more
  *          information.
@@ -805,7 +786,8 @@
  */
 inline int hid_get_feature_report(HidDeviceRef dev, TargetBytes data)
 {
-  return SDL_hid_get_feature_report(dev, data.data(), data.size_bytes());
+  return SDL_hid_get_feature_report(
+    dev, data.data_as<Uint8>(), data.size_bytes());
 }
 
 inline int HidDeviceBase::get_feature_report(TargetBytes data)
@@ -826,8 +808,6 @@
  *             the first byte of `data` to the Report ID of the report to be
  *             read, or set it to zero if your device does not use numbered
  *             reports.
- * @param length the number of bytes to read, including an extra byte for the
- *               report ID. The buffer can be longer than the actual report.
  * @returns the number of bytes read plus one for the report ID (which is still
  *          in the first byte), or -1 on on failure; call GetError() for more
  *          information.
@@ -836,7 +816,8 @@
  */
 inline int hid_get_input_report(HidDeviceRef dev, TargetBytes data)
 {
-  return SDL_hid_get_input_report(dev, data.data(), data.size_bytes());
+  return SDL_hid_get_input_report(
+    dev, data.data_as<Uint8>(), data.size_bytes());
 }
 
 inline int HidDeviceBase::get_input_report(TargetBytes data)
@@ -848,14 +829,13 @@
  * Close a HID device.
  *
  * @param dev a device handle returned from hid_open().
- * @returns 0 on success.
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
  */
-inline int hid_close(HidDeviceRaw dev)
+inline void hid_close(HidDeviceRaw dev)
 {
-  return CheckError(SDL_hid_close(dev));
+  CheckErrorIfNot(SDL_hid_close(dev), 0);
 }
 
 inline void HidDeviceBase::close() { hid_close(release()); }
@@ -866,7 +846,6 @@
  * @param dev a device handle returned from hid_open().
  * @param string a wide string buffer to put the data into.
  * @param maxlen the length of the buffer in multiples of wchar_t.
- * @returns 0 on success.
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
@@ -890,7 +869,6 @@
  * @param dev a device handle returned from hid_open().
  * @param string a wide string buffer to put the data into.
  * @param maxlen the length of the buffer in multiples of wchar_t.
- * @returns 0 on success.
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
@@ -913,7 +891,6 @@
  * @param dev a device handle returned from hid_open().
  * @param string a wide string buffer to put the data into.
  * @param maxlen the length of the buffer in multiples of wchar_t.
- * @returns 0 on success.
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
@@ -938,7 +915,6 @@
  * @param string_index the index of the string to get.
  * @param string a wide string buffer to put the data into.
  * @param maxlen the length of the buffer in multiples of wchar_t.
- * @returns 0 on success.
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
@@ -963,9 +939,10 @@
  * Get the device info from a HID device.
  *
  * @param dev a device handle returned from hid_open().
- * @returns a pointer to the hid_device_info for this hid_device or nullptr on
- *          failure; call GetError() for more information. This struct is valid
- *          until the device is closed with hid_close().
+ * @returns a pointer to the hid_device_info for this hid_device on success.
+ *          This struct is valid until the device is closed with
+ *          HidDevice.close().
+ * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
  */
@@ -987,7 +964,6 @@
  *
  * @param dev a device handle returned from hid_open().
  * @param buf the buffer to copy descriptor into.
- * @param buf_size the size of the buffer in bytes.
  * @returns the number of bytes actually copied or -1 on failure; call
  *          GetError() for more information.
  *
@@ -995,7 +971,8 @@
  */
 inline int hid_get_report_descriptor(HidDeviceRef dev, TargetBytes buf)
 {
-  return SDL_hid_get_report_descriptor(dev, buf.data(), buf.size_bytes());
+  return SDL_hid_get_report_descriptor(
+    dev, buf.data_as<Uint8>(), buf.size_bytes());
 }
 
 inline int HidDeviceBase::get_report_descriptor(TargetBytes buf)
