--- src/generated/SDL3pp_hints.h
+++ include/SDL3pp/SDL3pp_hints.h
@@ -9,7 +9,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryHints Category Hints
+ * @defgroup CategoryHints Configuration Variables
  *
  * This file contains functions to set and get configuration hints, as well as
  * listing each of them alphabetically.
@@ -25,6 +25,8 @@
  * @{
  */
 
+#ifdef SDL3PP_DOC
+
 /**
  * Specify the behavior of Alt+Tab while the keyboard is grabbed.
  *
@@ -1338,7 +1340,7 @@
  *
  * `0xAAAA/0xBBBB,0xCCCC/0xDDDD`
  *
- * The variable can also take the form of @file, in which case the named file
+ * The variable can also take the form of file, in which case the named file
  * will be loaded and interpreted as the value of the variable.
  *
  * This hint can be set anytime.
@@ -2084,10 +2086,6 @@
  */
 #define SDL_HINT_JOYSTICK_HIDAPI_GIP "SDL_JOYSTICK_HIDAPI_GIP"
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * A variable controlling whether the new HIDAPI driver for wired Xbox One (GIP)
  * controllers should reset the controller if it can't get the metadata from the
@@ -3741,10 +3739,6 @@
 #define SDL_HINT_VIDEO_METAL_AUTO_RESIZE_DRAWABLE                              \
   "SDL_VIDEO_METAL_AUTO_RESIZE_DRAWABLE"
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * A variable controlling whether SDL will attempt to automatically set the
  * destination display to a mode most closely matching that of the previous
@@ -4679,6 +4673,8 @@
  */
 #define SDL_HINT_PEN_TOUCH_EVENTS "SDL_PEN_TOUCH_EVENTS"
 
+#endif // SDL3PP_DOC
+
 /**
  * An enumeration of hint priorities.
  *
@@ -4893,7 +4889,6 @@
  * @param name the hint to watch.
  * @param callback An HintCallback function that will be called when the hint
  *                 value changes.
- * @param userdata a pointer to pass to the callback function.
  * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
@@ -4904,7 +4899,7 @@
  */
 inline void AddHintCallback(StringParam name, HintCB callback)
 {
-  static_assert(false, "Not implemented");
+  AddHintCallback(std::move(name), callback.wrapper, callback.data);
 }
 
 /**
