--- build/generated/SDL3pp_main.h
+++ include/SDL3pp/SDL3pp_main.h
@@ -1,22 +1,34 @@
 #ifndef SDL3PP_MAIN_H_
 #define SDL3PP_MAIN_H_
 
+#ifdef SDL3PP_MAIN_HANDLED
+#define SDL_MAIN_HANDLED
+#endif // SDL3PP_MAIN_HANDLED
+
+#ifdef SDL3PP_MAIN_USE_CALLBACKS
+#define SDL_MAIN_USE_CALLBACKS
+#endif // SDL3PP_MAIN_USE_CALLBACKS
+
+#ifdef SDL3PP_MAIN_USE_CLASS_CALLBACKS
+#define SDL_MAIN_USE_CALLBACKS
+#endif // SDL3PP_MAIN_USE_CLASS_CALLBACKS
+
 #include <SDL3/SDL_main.h>
 #include "SDL3pp.h"
 
 namespace SDL {
 
 /**
- * @defgroup CategoryMain Category Main
+ * @defgroup CategoryMain Application entry points
  *
  * Redefine main() if necessary so that it is called by SDL.
  *
  * In order to make this consistent on all platforms, the application's main()
  * should look like this:
  *
- * ```c
- * #include <SDL3/SDL.h>
- * #include <SDL3/SDL_main.h>
+ * ```cpp
+ * #include <SDL3pp/SDL3pp.h>
+ * #include <SDL3pp/SDL3pp_main.h>
  *
  * int main(int argc, char *argv[])
  * {
@@ -39,6 +51,8 @@
  * @{
  */
 
+#ifdef SDL3PP_DOC
+
 /**
  * Inform SDL that the app is providing an entry point instead of SDL.
  *
@@ -78,6 +92,25 @@
 #define SDL3PP_MAIN_USE_CALLBACKS SDL_MAIN_USE_CALLBACKS
 
 /**
+ * Inform SDL to use the main class callbacks instead of main.
+ *
+ * SDL does not define this macro, but will check if it is defined when
+ * including `SDL_main.h`. If defined, SDL will expect the app to provide
+ * a function: SDL_AppCreate. The app should not provide a `main` function in
+ * this case, and doing so will likely cause the build to fail.
+ *
+ * @todo
+ *
+ * @since This macro is used by the headers since SDLpp 0.11.0.
+ *
+ * @sa SDL_AppCreate
+ * @sa SDL3PP_MAIN_USE_CALLBACKS
+ */
+#define SDL3PP_MAIN_USE_CLASS_CALLBACKS 1
+
+#endif // SDL3PP_DOC
+
+/**
  * The prototype for the application's main() function
  *
  * @param argc an ANSI-C style main function's argc.
@@ -140,7 +173,7 @@
                   main_func mainFunction,
                   void* reserved)
 {
-  return SDL_RunApp(argc, argv[], mainFunction, reserved);
+  return SDL_RunApp(argc, argv, mainFunction, reserved);
 }
 
 /**
@@ -176,9 +209,11 @@
                                  AppQuit_func appquit)
 {
   return SDL_EnterAppMainCallbacks(
-    argc, argv[], appinit, appiter, appevent, appquit);
+    argc, argv, appinit, appiter, appevent, appquit);
 }
 
+#if defined(SDL_PLATFORM_WINDOWS)
+
 /**
  * Register a win32 window class for SDL's use.
  *
@@ -191,12 +226,11 @@
  * Most applications do not need to, and should not, call this directly; SDL
  * will call it when initializing the video subsystem.
  *
- * If `name` is nullptr, SDL currently uses `(CS_BYTEALIGNCLIENT | CS_OWNDC)`
- * for the style, regardless of what is specified here.
- *
  * @param name the window class name, in UTF-8 encoding. If nullptr, SDL
  *             currently uses "SDL_app" but this isn't guaranteed.
- * @param style the value to use in WNDCLASSEX::style.
+ * @param style the value to use in WNDCLASSEX::style. If `name` is nullptr, SDL
+ *              currently uses `(CS_BYTEALIGNCLIENT | CS_OWNDC)` regardless of
+ *              what is specified here.
  * @param hInst the HINSTANCE to use in WNDCLASSEX::hInstance. If zero, SDL will
  *              use `GetModuleHandle(nullptr)` instead.
  * @returns true on success or false on failure; call GetError() for more
@@ -230,6 +264,8 @@
  */
 inline void UnregisterApp() { SDL_UnregisterApp(); }
 
+#endif /* defined(SDL_PLATFORM_WINDOWS) */
+
 /**
  * Callback from the application to let the suspend continue.
  *
@@ -257,6 +293,222 @@
  */
 inline void GDKSuspendComplete() { SDL_GDKSuspendComplete(); }
 
+/**
+ * Base class for SDL3PP_MAIN_USE_CLASS_CALLBACKS main callback classes.
+ *
+ * This is an alternative that already includes exception handling and logging,
+ * so you have sensible behavior for C++.
+ *
+ * Also, you can just override the functions you need, and rely on the default
+ * behavior of the others, which is to continue iterating until a quit event is
+ * received, at which point it will exit with success.
+ */
+struct AppInterface
+{
+  virtual ~AppInterface() = default;
+
+  /**
+   * Called to initialize the app. This is called once at startup.
+   *
+   * @return SDL::APP_CONTINUE to continue running the app, otherwise
+   * SDL::APP_SUCCESS to exit with success and SDL::APP_FAILURE to exit with
+   * failure.
+   */
+  virtual SDL::AppResult Init() { return SDL::APP_CONTINUE; }
+
+  /**
+   * Called to iterate the app.
+   *
+   * This is called repeatedly until it returns something other than
+   * SDL::APP_CONTINUE.
+   *
+   * @return SDL::APP_CONTINUE to continue running the app, otherwise
+   * SDL::APP_SUCCESS to exit with success and SDL::APP_FAILURE to exit with
+   * failure.
+   */
+  virtual SDL::AppResult Iterate() { return SDL::APP_CONTINUE; }
+
+  /**
+   * Called to handle an event.
+   *
+   * This is called when an event is received.
+   *
+   * @param ev The event to handle.
+   * @return SDL::APP_CONTINUE to continue running the app, otherwise
+   * SDL::APP_SUCCESS to exit with success and SDL::APP_FAILURE to exit with
+   * failure.
+   */
+  virtual SDL::AppResult Event(const SDL::Event& ev)
+  {
+    return ev.type == SDL_EVENT_QUIT ? SDL::APP_SUCCESS : SDL::APP_CONTINUE;
+  }
+
+  /**
+   * Called to quit the app.
+   *
+   * This is called when the app is quitting.
+   *
+   * @param result The result of the app.
+   */
+  virtual void quit(SDL::AppResult result) {}
+};
+
+/**
+ * Prototype for the application's main class creation function.
+ *
+ * You need to implement this function if you want to use
+ * SDL3PP_MAIN_USE_CLASS_CALLBACKS. It should return a pointer to an instance of
+ * your main class, which should be a subclass of AppInterface. This function is
+ * called by SDL to create the main class instance, and then SDL will call the
+ * appropriate functions on that instance for initialization, iteration, event
+ * handling, and quitting.
+ *
+ * You should not call this function directly; it is called by SDL3pp
+ * internally.
+ *
+ * You may also use the macros SDL3PP_DEFINE_CALLBACKS or
+ * SDL3PP_DEFINE_CLASS_CALLBACKS to define this function for you, if your main
+ * class is compatible with the default implementations of the main callbacks.
+ *
+ * @param argc an ANSI-C style main function's argc.
+ * @param argv an ANSI-C style main function's argv.
+ * @returns a pointer to the application's main class instance, or nullptr to
+ *          indicate early exit with APP_SUCCESS.
+ * @throws any exception, which will be caught and logged by SDL, and cause
+ *         the app to exit with APP_FAILURE.
+ */
+extern "C" SDLMAIN_DECLSPEC AppInterface* SDLCALL SDL_AppCreate(int argc,
+                                                                char* argv[]);
+
+#ifdef SDL3PP_MAIN_USE_CLASS_CALLBACKS
+
+/// @private
+extern "C" SDLMAIN_DECLSPEC SDL::AppResult SDL_AppInit(void** appstate,
+                                                       int argc,
+                                                       char* argv[])
+{
+  try {
+    auto app = SDL_AppCreate(argc, argv);
+    if (!app) return APP_SUCCESS;
+    *appstate = app;
+    app->Init();
+    return APP_CONTINUE;
+  } catch (std::exception& e) {
+    LOG_CATEGORY_APPLICATION.LogUnformatted(SDL3PP_APPCLASS_LOG_PRIORITY,
+                                            e.what());
+  } catch (...) {
+    LOG_CATEGORY_APPLICATION.LogUnformatted(
+      SDL3PP_APPCLASS_LOG_PRIORITY, "Critical error during app initialization");
+  }
+  return APP_FAILURE;
+}
+
+/// @private
+extern "C" SDLMAIN_DECLSPEC SDL::AppResult SDL_AppIterate(void* appstate)
+{
+  try {
+    return static_cast<AppInterface*>(appstate)->Iterate();
+  } catch (std::exception& e) {
+    LOG_CATEGORY_APPLICATION.LogUnformatted(SDL3PP_APPCLASS_LOG_PRIORITY,
+                                            e.what());
+  } catch (...) {
+    LOG_CATEGORY_APPLICATION.LogUnformatted(
+      SDL3PP_APPCLASS_LOG_PRIORITY, "Critical error during app iteration");
+  }
+  return APP_FAILURE;
+}
+
+/// @private
+extern "C" SDLMAIN_DECLSPEC SDL::AppResult SDL_AppEvent(void* appstate,
+                                                        SDL::Event* event)
+{
+  try {
+    return static_cast<AppInterface*>(appstate)->Event(*event);
+  } catch (std::exception& e) {
+    LOG_CATEGORY_APPLICATION.LogUnformatted(SDL3PP_APPCLASS_LOG_PRIORITY,
+                                            e.what());
+  } catch (...) {
+    LOG_CATEGORY_APPLICATION.LogUnformatted(
+      SDL3PP_APPCLASS_LOG_PRIORITY, "Critical error during app event handling");
+  }
+  return APP_FAILURE;
+}
+
+/// @private
+extern "C" SDLMAIN_DECLSPEC void SDL_AppQuit(void* appstate,
+                                             SDL::AppResult result)
+{
+  try {
+    auto app = static_cast<AppInterface*>(appstate);
+    if (!app) return;
+    app->quit(result);
+    delete app;
+  } catch (std::exception& e) {
+    LOG_CATEGORY_APPLICATION.LogUnformatted(SDL3PP_APPCLASS_LOG_PRIORITY,
+                                            e.what());
+  } catch (...) {
+    LOG_CATEGORY_APPLICATION.LogUnformatted(SDL3PP_APPCLASS_LOG_PRIORITY,
+                                            "Critical error during app exit");
+  }
+}
+
+/**
+ * Use this to define the callbacks for given class
+ * @param CLASS The class to wrap in callbacks.
+ */
+#define SDL3PP_DEFINE_CALLBACKS(CLASS)                                         \
+  extern "C" SDLMAIN_DECLSPEC SDL::AppInterface* SDLCALL SDL_AppCreate(        \
+    int, char*[])                                                              \
+  {                                                                            \
+    return new CLASS();                                                        \
+  }
+
+/**
+ * Use this to define the callbacks for given class, with app metadata.
+ * @param CLASS The class to wrap in callbacks.
+ * @param INIT_FLAGS The flags to pass to SDL::Init when initializing the app.
+ * @param APPNAME The application name to set in SDL_SetAppMetadata.
+ * @param APPVERSION The application version to set in SDL_SetAppMetadata.
+ * @param APPIDENTIFIER The application identifier to set in SDL_SetAppMetadata.
+ */
+#define SDL3PP_DEFINE_CLASS_CALLBACKS(                                         \
+  CLASS, INIT_FLAGS, APPNAME, APPVERSION, APPIDENTIFIER)                       \
+  extern "C" SDLMAIN_DECLSPEC SDL::AppInterface* SDLCALL SDL_AppCreate(        \
+    int, char*[])                                                              \
+  {                                                                            \
+    SDL::SetAppMetadata(APPNAME, APPVERSION, APPIDENTIFIER);                   \
+    SDL::Init(INIT_FLAGS);                                                     \
+    return new CLASS();                                                        \
+  }
+
+#else // SDL3PP_MAIN_USE_CLASS_CALLBACKS
+
+/**
+ * Use this to define the callbacks for given class
+ * @param CLASS The class to wrap in callbacks.
+ */
+#define SDL3PP_DEFINE_CALLBACKS(CLASS)                                         \
+  static_assert(SDL::HasIterateFunction<CLASS>, "Main class not compatible");  \
+  inline SDL::AppResult SDL_AppInit(void** appstate, int argc, char* argv[])   \
+  {                                                                            \
+    return SDL::InitClass(reinterpret_cast<CLASS**>(appstate),                 \
+                          SDL::AppArgs{argv, size_t(argc)});                   \
+  }                                                                            \
+  inline SDL::AppResult SDL_AppIterate(void* appstate)                         \
+  {                                                                            \
+    return SDL::IterateClass(static_cast<CLASS*>(appstate));                   \
+  }                                                                            \
+  inline SDL::AppResult SDL_AppEvent(void* appstate, SDL::Event* event)        \
+  {                                                                            \
+    return SDL::EventClass(static_cast<CLASS*>(appstate), *event);             \
+  }                                                                            \
+  inline void SDL_AppQuit(void* appstate, SDL::AppResult result)               \
+  {                                                                            \
+    SDL::QuitClass(static_cast<CLASS*>(appstate), result);                     \
+  }
+
+#endif // SDL3PP_MAIN_USE_CLASS_CALLBACKS
+
 /// @}
 
 } // namespace SDL
