--- src/generated/SDL3pp_gpu.h
+++ include/SDL3pp/SDL3pp_gpu.h
@@ -12,7 +12,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryGPU Category GPU
+ * @defgroup CategoryGPU 3D Rendering and GPU Compute
  *
  * The GPU API offers a cross-platform way for apps to talk to modern graphics
  * hardware. It offers both 3D graphics and compute support, in the style of
@@ -138,7 +138,7 @@
  * limited hardware support, this API is probably not for you.
  *
  * Examples demonstrating proper usage of this API can be found
- * [here](https://github.com/TheSpydog/SDL_gpu_examples) .
+ * [here](https://github.com/TheSpydog/SDL_gpu_examples).
  *
  * ## Performance considerations
  *
@@ -1304,7 +1304,6 @@
    * @param first_slot the vertex buffer slot to begin binding from.
    * @param bindings an array of GPUBufferBinding structs containing vertex
    *                 buffers and offset values.
-   * @param num_bindings the number of bindings in the bindings array.
    *
    * @since This function is available since SDL 3.2.0.
    */
@@ -1335,8 +1334,6 @@
    * @param first_slot the vertex sampler slot to begin binding from.
    * @param texture_sampler_bindings an array of texture-sampler binding
    *                                 structs.
-   * @param num_bindings the number of texture-sampler pairs to bind from the
-   *                     array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1357,7 +1354,6 @@
    *
    * @param first_slot the vertex storage texture slot to begin binding from.
    * @param storage_textures an array of storage textures.
-   * @param num_bindings the number of storage texture to bind from the array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1377,7 +1373,6 @@
    *
    * @param first_slot the vertex storage buffer slot to begin binding from.
    * @param storage_buffers an array of buffers.
-   * @param num_bindings the number of buffers to bind from the array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1397,8 +1392,6 @@
    * @param first_slot the fragment sampler slot to begin binding from.
    * @param texture_sampler_bindings an array of texture-sampler binding
    *                                 structs.
-   * @param num_bindings the number of texture-sampler pairs to bind from the
-   *                     array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1419,7 +1412,6 @@
    *
    * @param first_slot the fragment storage texture slot to begin binding from.
    * @param storage_textures an array of storage textures.
-   * @param num_bindings the number of storage textures to bind from the array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1440,7 +1432,6 @@
    *
    * @param first_slot the fragment storage buffer slot to begin binding from.
    * @param storage_buffers an array of storage buffers.
-   * @param num_bindings the number of storage buffers to bind from the array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1606,8 +1597,6 @@
    * @param first_slot the compute sampler slot to begin binding from.
    * @param texture_sampler_bindings an array of texture-sampler binding
    *                                 structs.
-   * @param num_bindings the number of texture-sampler bindings to bind from the
-   *                     array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1628,7 +1617,6 @@
    *
    * @param first_slot the compute storage texture slot to begin binding from.
    * @param storage_textures an array of storage textures.
-   * @param num_bindings the number of storage textures to bind from the array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1648,7 +1636,6 @@
    *
    * @param first_slot the compute storage buffer slot to begin binding from.
    * @param storage_buffers an array of storage buffer binding structs.
-   * @param num_bindings the number of storage buffers to bind from the array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -2191,7 +2178,6 @@
    *
    * @param slot_index the vertex uniform slot to push data to.
    * @param data client data to write.
-   * @param length the length of the data to write.
    *
    * @since This function is available since SDL 3.2.0.
    */
@@ -2208,7 +2194,6 @@
    *
    * @param slot_index the fragment uniform slot to push data to.
    * @param data client data to write.
-   * @param length the length of the data to write.
    *
    * @since This function is available since SDL 3.2.0.
    */
@@ -2225,7 +2210,6 @@
    *
    * @param slot_index the uniform slot to push data to.
    * @param data client data to write.
-   * @param length the length of the data to write.
    *
    * @since This function is available since SDL 3.2.0.
    */
@@ -2252,8 +2236,6 @@
    *
    * @param color_target_infos an array of texture subresources with
    *                           corresponding clear values and load/store ops.
-   * @param num_color_targets the number of color targets in the
-   *                          color_target_infos array.
    * @param depth_stencil_target_info a texture subresource with corresponding
    *                                  clear value and load/store ops, may be
    *                                  nullptr.
@@ -2291,12 +2273,8 @@
    *
    * @param storage_texture_bindings an array of writeable storage texture
    *                                 binding structs.
-   * @param num_storage_texture_bindings the number of storage textures to bind
-   *                                     from the array.
    * @param storage_buffer_bindings an array of writeable storage buffer binding
    *                                structs.
-   * @param num_storage_buffer_bindings the number of storage buffers to bind
-   *                                    from the array.
    * @returns a compute pass handle.
    *
    * @since This function is available since SDL 3.2.0.
@@ -2368,12 +2346,11 @@
    * than the one that created the window.
    *
    * @param window a window that has been claimed.
-   * @param swapchain_texture a pointer filled in with a swapchain texture
-   *                          handle.
    * @param swapchain_texture_width a pointer filled in with the swapchain
    *                                texture width, may be nullptr.
    * @param swapchain_texture_height a pointer filled in with the swapchain
    *                                 texture height, may be nullptr.
+   * @returns a swapchain texture handle.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called from the thread that
@@ -2418,12 +2395,11 @@
    * another reading operation.
    *
    * @param window a window that has been claimed.
-   * @param swapchain_texture a pointer filled in with a swapchain texture
-   *                          handle.
    * @param swapchain_texture_width a pointer filled in with the swapchain
    *                                texture width, may be nullptr.
    * @param swapchain_texture_height a pointer filled in with the swapchain
    *                                 texture height, may be nullptr.
+   * @returns a swapchain texture handle.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called from the thread that
@@ -4087,7 +4063,6 @@
    * @param wait_all if 0, wait for any fence to be signaled, if 1, wait for all
    *                 fences to be signaled.
    * @param fences an array of fences to wait on.
-   * @param num_fences the number of fences in the fences array.
    * @throws Error on failure.
    *
    * @since This function is available since SDL 3.2.0.
@@ -4149,6 +4124,8 @@
   bool TextureSupportsSampleCount(GPUTextureFormat format,
                                   GPUSampleCount sample_count);
 
+#ifdef SDL_PLATFORM_GDK
+
   /**
    * Call this to suspend GPU operation on Xbox when you receive the
    * EVENT_DID_ENTER_BACKGROUND event.
@@ -4174,6 +4151,8 @@
    * @sa AddEventWatch
    */
   void GDKResumeGPU();
+
+#endif /* SDL_PLATFORM_GDK */
 };
 
 /// Semi-safe reference for GPUDevice.
@@ -5173,24 +5152,12 @@
 constexpr auto CREATE_FEATURE_CLIP_DISTANCE_BOOLEAN =
   SDL_PROP_GPU_DEVICE_CREATE_FEATURE_CLIP_DISTANCE_BOOLEAN;
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 constexpr auto CREATE_FEATURE_DEPTH_CLAMPING_BOOLEAN =
   SDL_PROP_GPU_DEVICE_CREATE_FEATURE_DEPTH_CLAMPING_BOOLEAN;
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 constexpr auto CREATE_FEATURE_INDIRECT_DRAW_FIRST_INSTANCE_BOOLEAN =
   SDL_PROP_GPU_DEVICE_CREATE_FEATURE_INDIRECT_DRAW_FIRST_INSTANCE_BOOLEAN;
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 constexpr auto CREATE_FEATURE_ANISOTROPY_BOOLEAN =
   SDL_PROP_GPU_DEVICE_CREATE_FEATURE_ANISOTROPY_BOOLEAN;
 
@@ -5229,36 +5196,16 @@
 constexpr auto CREATE_VULKAN_REQUIRE_HARDWARE_ACCELERATION_BOOLEAN =
   SDL_PROP_GPU_DEVICE_CREATE_VULKAN_REQUIRE_HARDWARE_ACCELERATION_BOOLEAN;
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 constexpr auto CREATE_VULKAN_OPTIONS_POINTER =
   SDL_PROP_GPU_DEVICE_CREATE_VULKAN_OPTIONS_POINTER;
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 constexpr auto NAME_STRING = SDL_PROP_GPU_DEVICE_NAME_STRING;
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 constexpr auto DRIVER_NAME_STRING = SDL_PROP_GPU_DEVICE_DRIVER_NAME_STRING;
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 constexpr auto DRIVER_VERSION_STRING =
   SDL_PROP_GPU_DEVICE_DRIVER_VERSION_STRING;
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 constexpr auto DRIVER_INFO_STRING = SDL_PROP_GPU_DEVICE_DRIVER_INFO_STRING;
 
 #endif // SDL_VERSION_ATLEAST(3, 4, 0)
@@ -5479,10 +5426,6 @@
   return CheckError(SDL_GetGPUDeviceProperties(device));
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 inline PropertiesRef GPUDevice::GetProperties()
 {
   return SDL::GetGPUDeviceProperties(m_resource);
@@ -6268,7 +6211,6 @@
  * @param command_buffer a command buffer.
  * @param slot_index the vertex uniform slot to push data to.
  * @param data client data to write.
- * @param length the length of the data to write.
  *
  * @since This function is available since SDL 3.2.0.
  */
@@ -6299,7 +6241,6 @@
  * @param command_buffer a command buffer.
  * @param slot_index the fragment uniform slot to push data to.
  * @param data client data to write.
- * @param length the length of the data to write.
  *
  * @since This function is available since SDL 3.2.0.
  */
@@ -6330,7 +6271,6 @@
  * @param command_buffer a command buffer.
  * @param slot_index the uniform slot to push data to.
  * @param data client data to write.
- * @param length the length of the data to write.
  *
  * @since This function is available since SDL 3.2.0.
  */
@@ -6371,8 +6311,6 @@
  * @param command_buffer a command buffer.
  * @param color_target_infos an array of texture subresources with corresponding
  *                           clear values and load/store ops.
- * @param num_color_targets the number of color targets in the
- *                          color_target_infos array.
  * @param depth_stencil_target_info a texture subresource with corresponding
  *                                  clear value and load/store ops, may be
  *                                  nullptr.
@@ -6387,8 +6325,10 @@
   std::span<const GPUColorTargetInfo> color_target_infos,
   OptionalRef<const GPUDepthStencilTargetInfo> depth_stencil_target_info)
 {
-  return SDL_BeginGPURenderPass(
-    command_buffer, color_target_infos, depth_stencil_target_info);
+  return SDL_BeginGPURenderPass(command_buffer,
+                                color_target_infos.data(),
+                                color_target_infos.size(),
+                                depth_stencil_target_info);
 }
 
 inline GPURenderPass GPUCommandBuffer::BeginRenderPass(
@@ -6449,7 +6389,7 @@
  */
 inline void SetGPUScissor(GPURenderPass render_pass, const RectRaw& scissor)
 {
-  SDL_SetGPUScissor(render_pass, scissor);
+  SDL_SetGPUScissor(render_pass, &scissor);
 }
 
 inline void GPURenderPass::SetScissor(const RectRaw& scissor)
@@ -6504,7 +6444,6 @@
  * @param first_slot the vertex buffer slot to begin binding from.
  * @param bindings an array of GPUBufferBinding structs containing vertex
  *                 buffers and offset values.
- * @param num_bindings the number of bindings in the bindings array.
  *
  * @since This function is available since SDL 3.2.0.
  */
@@ -6512,7 +6451,8 @@
                                  Uint32 first_slot,
                                  std::span<const GPUBufferBinding> bindings)
 {
-  SDL_BindGPUVertexBuffers(render_pass, first_slot, bindings);
+  SDL_BindGPUVertexBuffers(
+    render_pass, first_slot, bindings.data(), bindings.size());
 }
 
 inline void GPURenderPass::BindVertexBuffers(
@@ -6557,8 +6497,6 @@
  * @param render_pass a render pass handle.
  * @param first_slot the vertex sampler slot to begin binding from.
  * @param texture_sampler_bindings an array of texture-sampler binding structs.
- * @param num_bindings the number of texture-sampler pairs to bind from the
- *                     array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -6569,7 +6507,10 @@
   Uint32 first_slot,
   std::span<const GPUTextureSamplerBinding> texture_sampler_bindings)
 {
-  SDL_BindGPUVertexSamplers(render_pass, first_slot, texture_sampler_bindings);
+  SDL_BindGPUVertexSamplers(render_pass,
+                            first_slot,
+                            texture_sampler_bindings.data(),
+                            texture_sampler_bindings.size());
 }
 
 inline void GPURenderPass::BindVertexSamplers(
@@ -6592,7 +6533,6 @@
  * @param render_pass a render pass handle.
  * @param first_slot the vertex storage texture slot to begin binding from.
  * @param storage_textures an array of storage textures.
- * @param num_bindings the number of storage texture to bind from the array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -6603,7 +6543,8 @@
   Uint32 first_slot,
   SpanRef<const GPUTextureRaw> storage_textures)
 {
-  SDL_BindGPUVertexStorageTextures(render_pass, first_slot, storage_textures);
+  SDL_BindGPUVertexStorageTextures(
+    render_pass, first_slot, storage_textures.data(), storage_textures.size());
 }
 
 inline void GPURenderPass::BindVertexStorageTextures(
@@ -6626,7 +6567,6 @@
  * @param render_pass a render pass handle.
  * @param first_slot the vertex storage buffer slot to begin binding from.
  * @param storage_buffers an array of buffers.
- * @param num_bindings the number of buffers to bind from the array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -6637,7 +6577,8 @@
   Uint32 first_slot,
   SpanRef<const GPUBufferRaw> storage_buffers)
 {
-  SDL_BindGPUVertexStorageBuffers(render_pass, first_slot, storage_buffers);
+  SDL_BindGPUVertexStorageBuffers(
+    render_pass, first_slot, storage_buffers.data(), storage_buffers.size());
 }
 
 inline void GPURenderPass::BindVertexStorageBuffers(
@@ -6659,8 +6600,6 @@
  * @param render_pass a render pass handle.
  * @param first_slot the fragment sampler slot to begin binding from.
  * @param texture_sampler_bindings an array of texture-sampler binding structs.
- * @param num_bindings the number of texture-sampler pairs to bind from the
- *                     array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -6671,8 +6610,10 @@
   Uint32 first_slot,
   std::span<const GPUTextureSamplerBinding> texture_sampler_bindings)
 {
-  SDL_BindGPUFragmentSamplers(
-    render_pass, first_slot, texture_sampler_bindings);
+  SDL_BindGPUFragmentSamplers(render_pass,
+                              first_slot,
+                              texture_sampler_bindings.data(),
+                              texture_sampler_bindings.size());
 }
 
 inline void GPURenderPass::BindFragmentSamplers(
@@ -6695,7 +6636,6 @@
  * @param render_pass a render pass handle.
  * @param first_slot the fragment storage texture slot to begin binding from.
  * @param storage_textures an array of storage textures.
- * @param num_bindings the number of storage textures to bind from the array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -6706,7 +6646,8 @@
   Uint32 first_slot,
   SpanRef<const GPUTextureRaw> storage_textures)
 {
-  SDL_BindGPUFragmentStorageTextures(render_pass, first_slot, storage_textures);
+  SDL_BindGPUFragmentStorageTextures(
+    render_pass, first_slot, storage_textures.data(), storage_textures.size());
 }
 
 inline void GPURenderPass::BindFragmentStorageTextures(
@@ -6729,7 +6670,6 @@
  * @param render_pass a render pass handle.
  * @param first_slot the fragment storage buffer slot to begin binding from.
  * @param storage_buffers an array of storage buffers.
- * @param num_bindings the number of storage buffers to bind from the array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -6740,7 +6680,8 @@
   Uint32 first_slot,
   SpanRef<const GPUBufferRaw> storage_buffers)
 {
-  SDL_BindGPUFragmentStorageBuffers(render_pass, first_slot, storage_buffers);
+  SDL_BindGPUFragmentStorageBuffers(
+    render_pass, first_slot, storage_buffers.data(), storage_buffers.size());
 }
 
 inline void GPURenderPass::BindFragmentStorageBuffers(
@@ -6947,12 +6888,8 @@
  * @param command_buffer a command buffer.
  * @param storage_texture_bindings an array of writeable storage texture binding
  *                                 structs.
- * @param num_storage_texture_bindings the number of storage textures to bind
- *                                     from the array.
  * @param storage_buffer_bindings an array of writeable storage buffer binding
  *                                structs.
- * @param num_storage_buffer_bindings the number of storage buffers to bind from
- *                                    the array.
  * @returns a compute pass handle.
  *
  * @since This function is available since SDL 3.2.0.
@@ -6964,8 +6901,11 @@
   std::span<const GPUStorageTextureReadWriteBinding> storage_texture_bindings,
   std::span<const GPUStorageBufferReadWriteBinding> storage_buffer_bindings)
 {
-  return SDL_BeginGPUComputePass(
-    command_buffer, storage_texture_bindings, storage_buffer_bindings);
+  return SDL_BeginGPUComputePass(command_buffer,
+                                 storage_texture_bindings.data(),
+                                 storage_texture_bindings.size(),
+                                 storage_buffer_bindings.data(),
+                                 storage_buffer_bindings.size());
 }
 
 inline GPUComputePass GPUCommandBuffer::BeginComputePass(
@@ -7006,8 +6946,6 @@
  * @param compute_pass a compute pass handle.
  * @param first_slot the compute sampler slot to begin binding from.
  * @param texture_sampler_bindings an array of texture-sampler binding structs.
- * @param num_bindings the number of texture-sampler bindings to bind from the
- *                     array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -7018,8 +6956,10 @@
   Uint32 first_slot,
   std::span<const GPUTextureSamplerBinding> texture_sampler_bindings)
 {
-  SDL_BindGPUComputeSamplers(
-    compute_pass, first_slot, texture_sampler_bindings);
+  SDL_BindGPUComputeSamplers(compute_pass,
+                             first_slot,
+                             texture_sampler_bindings.data(),
+                             texture_sampler_bindings.size());
 }
 
 inline void GPUComputePass::BindSamplers(
@@ -7042,7 +6982,6 @@
  * @param compute_pass a compute pass handle.
  * @param first_slot the compute storage texture slot to begin binding from.
  * @param storage_textures an array of storage textures.
- * @param num_bindings the number of storage textures to bind from the array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -7053,7 +6992,8 @@
   Uint32 first_slot,
   SpanRef<const GPUTextureRaw> storage_textures)
 {
-  SDL_BindGPUComputeStorageTextures(compute_pass, first_slot, storage_textures);
+  SDL_BindGPUComputeStorageTextures(
+    compute_pass, first_slot, storage_textures.data(), storage_textures.size());
 }
 
 inline void GPUComputePass::BindStorageTextures(
@@ -7076,7 +7016,6 @@
  * @param compute_pass a compute pass handle.
  * @param first_slot the compute storage buffer slot to begin binding from.
  * @param storage_buffers an array of storage buffer binding structs.
- * @param num_bindings the number of storage buffers to bind from the array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -7087,7 +7026,8 @@
   Uint32 first_slot,
   SpanRef<const GPUBufferRaw> storage_buffers)
 {
-  SDL_BindGPUComputeStorageBuffers(compute_pass, first_slot, storage_buffers);
+  SDL_BindGPUComputeStorageBuffers(
+    compute_pass, first_slot, storage_buffers.data(), storage_buffers.size());
 }
 
 inline void GPUComputePass::BindStorageBuffers(
@@ -7620,7 +7560,7 @@
  * supported via GPUDevice.WindowSupportsPresentMode /
  * GPUDevice.WindowSupportsSwapchainComposition prior to calling this function.
  *
- * GPU_PRESENTMODE_VSYNC with GPU_SWAPCHAINCOMPOSITION_SDR is always supported.
+ * GPU_PRESENTMODE_VSYNC with GPU_SWAPCHAINCOMPOSITION_SDR are always supported.
  *
  * @param device a GPU context.
  * @param window an Window that has been claimed.
@@ -7736,11 +7676,11 @@
  *
  * @param command_buffer a command buffer.
  * @param window a window that has been claimed.
- * @param swapchain_texture a pointer filled in with a swapchain texture handle.
  * @param swapchain_texture_width a pointer filled in with the swapchain texture
  *                                width, may be nullptr.
  * @param swapchain_texture_height a pointer filled in with the swapchain
  *                                 texture height, may be nullptr.
+ * @returns a swapchain texture handle.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called from the thread that
@@ -7763,8 +7703,13 @@
   Uint32* swapchain_texture_width = nullptr,
   Uint32* swapchain_texture_height = nullptr)
 {
-  return CheckError(SDL_AcquireGPUSwapchainTexture(
-    command_buffer, window, swapchain_texture_width, swapchain_texture_height));
+  GPUTextureRaw texture;
+  CheckError(SDL_AcquireGPUSwapchainTexture(command_buffer,
+                                            window,
+                                            &texture,
+                                            swapchain_texture_width,
+                                            swapchain_texture_height));
+  return texture;
 }
 
 inline GPUTexture GPUCommandBuffer::AcquireSwapchainTexture(
@@ -7828,11 +7773,11 @@
  *
  * @param command_buffer a command buffer.
  * @param window a window that has been claimed.
- * @param swapchain_texture a pointer filled in with a swapchain texture handle.
  * @param swapchain_texture_width a pointer filled in with the swapchain texture
  *                                width, may be nullptr.
  * @param swapchain_texture_height a pointer filled in with the swapchain
  *                                 texture height, may be nullptr.
+ * @returns a swapchain texture handle.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called from the thread that
@@ -7850,8 +7795,13 @@
   Uint32* swapchain_texture_width = nullptr,
   Uint32* swapchain_texture_height = nullptr)
 {
-  return CheckError(SDL_WaitAndAcquireGPUSwapchainTexture(
-    command_buffer, window, swapchain_texture_width, swapchain_texture_height));
+  GPUTextureRaw texture;
+  CheckError(SDL_WaitAndAcquireGPUSwapchainTexture(command_buffer,
+                                                   window,
+                                                   &texture,
+                                                   swapchain_texture_width,
+                                                   swapchain_texture_height));
+  return texture;
 }
 
 inline GPUTexture GPUCommandBuffer::WaitAndAcquireSwapchainTexture(
@@ -7985,7 +7935,6 @@
  * @param wait_all if 0, wait for any fence to be signaled, if 1, wait for all
  *                 fences to be signaled.
  * @param fences an array of fences to wait on.
- * @param num_fences the number of fences in the fences array.
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
@@ -7997,7 +7946,8 @@
                              bool wait_all,
                              std::span<GPUFence* const> fences)
 {
-  CheckError(SDL_WaitForGPUFences(device, wait_all, fences));
+  CheckError(
+    SDL_WaitForGPUFences(device, wait_all, fences.data(), fences.size()));
 }
 
 inline void GPUDevice::WaitForFences(bool wait_all,
@@ -8149,10 +8099,6 @@
   return SDL_GetPixelFormatFromGPUTextureFormat(format);
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Get the GPU texture format corresponding to an SDL pixel format.
  *
@@ -8169,6 +8115,8 @@
 
 #endif // SDL_VERSION_ATLEAST(3, 4, 0)
 
+#ifdef SDL_PLATFORM_GDK
+
 /**
  * Call this to suspend GPU operation on Xbox when you receive the
  * EVENT_DID_ENTER_BACKGROUND event.
@@ -8203,6 +8151,8 @@
 
 inline void GPUDevice::GDKResumeGPU() { SDL::GDKResumeGPU(m_resource); }
 
+#endif /* SDL_PLATFORM_GDK */
+
 /// @}
 
 } // namespace SDL
