--- src/generated/SDL3pp_messagebox.h
+++ include/SDL3pp/SDL3pp_messagebox.h
@@ -8,7 +8,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryMessagebox Category Messagebox
+ * @defgroup CategoryMessagebox Message Boxes
  *
  * SDL offers a simple message box API, which is useful for simple alerts, such
  * as informing the user when something fatal happens at startup without the
@@ -32,6 +32,11 @@
 struct MessageBox;
 
 /**
+ * @name MessageBoxFlags
+ * @{
+ */
+
+/**
  * Message box flags.
  *
  * If supported will display warning icon, etc.
@@ -55,6 +60,13 @@
 constexpr MessageBoxFlags MESSAGEBOX_BUTTONS_RIGHT_TO_LEFT =
   SDL_MESSAGEBOX_BUTTONS_RIGHT_TO_LEFT; ///< buttons placed right to left
 
+/// @}
+
+/**
+ * @name MessageBoxButtonFlags
+ * @{
+ */
+
 /**
  * MessageBoxButtonData flags.
  *
@@ -70,6 +82,8 @@
   SDL_MESSAGEBOX_BUTTON_ESCAPEKEY_DEFAULT; ///< Marks the default button when
                                            ///< escape is hit
 
+/// @}
+
 /**
  * Individual button data.
  *
@@ -84,6 +98,11 @@
  */
 using MessageBoxColor = SDL_MessageBoxColor;
 
+/**
+ * @name MessageBoxColorTypes
+ * @{
+ */
+
 /// An enumeration of indices inside the colors array of MessageBoxColorScheme.
 using MessageBoxColorType = SDL_MessageBoxColorType;
 
@@ -105,6 +124,7 @@
 constexpr MessageBoxColorType MESSAGEBOX_COLOR_COUNT =
   SDL_MESSAGEBOX_COLOR_COUNT; ///< Size of the colors array of
                               ///< MessageBoxColorScheme.
+/// @}
 
 /**
  * A set of colors to use for message box dialogs
@@ -137,38 +157,27 @@
    * @param window the value for window.
    * @param title the value for title.
    * @param message the value for message.
-   * @param numbuttons the value for numbuttons.
    * @param buttons the value for buttons.
    * @param colorScheme the value for colorScheme.
    */
-  constexpr MessageBox(SDL_MessageBoxFlags flags,
-                       SDL_Window* window,
-                       const char* title,
-                       const char* message,
-                       int numbuttons,
-                       const SDL_MessageBoxButtonData* buttons,
-                       const SDL_MessageBoxColorScheme* colorScheme) noexcept
+  constexpr MessageBox(
+    MessageBoxFlags flags,
+    WindowRef window,
+    const char* title,
+    const char* message,
+    std::span<const MessageBoxButtonData> buttons,
+    OptionalRef<const MessageBoxColorScheme> colorScheme) noexcept
     : MessageBoxRaw{flags,
-                    window,
+                    window.get(),
                     title,
                     message,
-                    numbuttons,
-                    buttons,
+                    int(buttons.size()),
+                    buttons.data(),
                     colorScheme}
   {
   }
 
   /**
-   * Check if valid.
-   *
-   * @returns True if valid state, false otherwise.
-   */
-  constexpr explicit operator bool() const noexcept
-  {
-    return *this != MessageBoxRaw{};
-  }
-
-  /**
    * Get the flags.
    *
    * @returns current flags value.
@@ -266,11 +275,12 @@
   /**
    * Get the buttons.
    *
-   * @returns current buttons value.
+   * @returns current buttons.
    */
-  constexpr const SDL_MessageBoxButtonData* GetButtons() const noexcept
+  constexpr std::span<const MessageBoxButtonData> GetButtons() const noexcept
   {
-    return buttons;
+    if (numbuttons == 0) return {};
+    return std::span(buttons, size_t(numbuttons));
   }
 
   /**
@@ -280,9 +290,15 @@
    * @returns Reference to self.
    */
   constexpr MessageBox& SetButtons(
-    const SDL_MessageBoxButtonData* newButtons) noexcept
+    std::span<const MessageBoxButtonData> newButtons) noexcept
   {
-    buttons = newButtons;
+    if (newButtons.empty()) {
+      numbuttons = 0;
+      buttons = nullptr;
+    } else {
+      numbuttons = newButtons.size();
+      buttons = newButtons.data();
+    }
     return *this;
   }
 
@@ -291,7 +307,7 @@
    *
    * @returns current colorScheme value.
    */
-  constexpr const SDL_MessageBoxColorScheme* GetColorScheme() const noexcept
+  constexpr const MessageBoxColorScheme* GetColorScheme() const noexcept
   {
     return colorScheme;
   }
@@ -303,7 +319,7 @@
    * @returns Reference to self.
    */
   constexpr MessageBox& SetColorScheme(
-    const SDL_MessageBoxColorScheme* newColorScheme) noexcept
+    OptionalRef<const MessageBoxColorScheme> newColorScheme) noexcept
   {
     colorScheme = newColorScheme;
     return *this;
@@ -376,12 +392,12 @@
  */
 inline void ShowMessageBox(const MessageBoxRaw& messageboxdata, int* buttonid)
 {
-  CheckError(SDL_ShowMessageBox(messageboxdata, buttonid));
+  CheckError(SDL_ShowMessageBox(&messageboxdata, buttonid));
 }
 
 inline void MessageBox::Show(int* buttonid) const
 {
-  SDL::ShowMessageBox(this, buttonid);
+  SDL::ShowMessageBox(*this, buttonid);
 }
 
 /**
