--- build/generated/SDL3pp_gamepad.h
+++ include/SDL3pp/SDL3pp_gamepad.h
@@ -14,7 +14,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryGamepad Category Gamepad
+ * @defgroup CategoryGamepad Gamepad Support
  *
  * SDL provides a low-level joystick API, which just treats joysticks as an
  * arbitrary pile of buttons, axes, and hat switches. If you're planning to
@@ -423,8 +423,7 @@
    * Details about mappings are discussed with AddGamepadMapping().
    *
    * @returns a string that has the gamepad's mapping or nullptr if no mapping
-   *          is available; call GetError() for more information. This should be
-   *          freed with free() when it is no longer needed.
+   *          is available; call GetError() for more information.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
@@ -717,9 +716,7 @@
    *
    * @param count a pointer filled in with the number of bindings returned.
    * @returns a nullptr terminated array of pointers to bindings or nullptr on
-   *          failure; call GetError() for more information. This is a single
-   *          allocation that should be freed with free() when it is no longer
-   *          needed.
+   *          failure; call GetError() for more information.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
@@ -1265,9 +1262,7 @@
 /**
  * Get the current gamepad mappings.
  *
- * @param count a pointer filled in with the number of mappings returned, can be
- *              nullptr.
- * @returns  nullptr-terminated on success.
+ * @returns an array of the mapping strings, nullptr-terminated.
  * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
@@ -1276,7 +1271,9 @@
  */
 inline OwnArray<char*> GetGamepadMappings()
 {
-  return CheckError(SDL_GetGamepadMappings());
+  int count;
+  auto data = CheckError(SDL_GetGamepadMappings(&count));
+  return OwnArray<char*>(data);
 }
 
 /**
@@ -1284,8 +1281,7 @@
  *
  * @param guid a structure containing the GUID for which a mapping is desired.
  * @returns a mapping string or nullptr on failure; call GetError() for more
- *          information. This should be freed with free() when it is no longer
- *          needed.
+ *          information.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -1296,7 +1292,7 @@
  */
 inline StringResult GetGamepadMappingForGUID(GUID guid)
 {
-  return SDL_GetGamepadMappingForGUID(guid);
+  return StringResult(SDL_GetGamepadMappingForGUID(guid));
 }
 
 /**
@@ -1306,8 +1302,7 @@
  *
  * @param gamepad the gamepad you want to get the current mapping for.
  * @returns a string that has the gamepad's mapping or nullptr if no mapping is
- *          available; call GetError() for more information. This should be
- *          freed with free() when it is no longer needed.
+ *          available; call GetError() for more information.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -1320,7 +1315,7 @@
  */
 inline StringResult GetGamepadMapping(GamepadRef gamepad)
 {
-  return SDL_GetGamepadMapping(gamepad);
+  return StringResult(SDL_GetGamepadMapping(gamepad));
 }
 
 inline StringResult GamepadBase::GetMapping()
@@ -1366,11 +1361,8 @@
 /**
  * Get a list of currently connected gamepads.
  *
- * @param count a pointer filled in with the number of gamepads returned, may be
- *              nullptr.
  * @returns a 0 terminated array of joystick instance IDs or nullptr on failure;
- *          call GetError() for more information. This should be freed with
- *          free() when it is no longer needed.
+ *          call GetError() for more information.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -1379,7 +1371,12 @@
  * @sa HasGamepad
  * @sa OpenGamepad
  */
-inline OwnArray<JoystickID> GetGamepads() { return SDL_GetGamepads(); }
+inline OwnArray<JoystickID> GetGamepads()
+{
+  int count;
+  auto r = reinterpret_cast<JoystickID*>(SDL_GetGamepads(&count));
+  return OwnArray<JoystickID>(r, count);
+}
 
 /**
  * Check if the given joystick is supported by the gamepad interface.
@@ -1650,7 +1647,7 @@
  */
 inline GamepadRef GetGamepadFromID(JoystickID instance_id)
 {
-  return CheckError(SDL_GetGamepadFromID(instance_id));
+  return {CheckError(SDL_GetGamepadFromID(instance_id))};
 }
 
 /**
@@ -1668,7 +1665,7 @@
  */
 inline GamepadRef GetGamepadFromPlayerIndex(int player_index)
 {
-  return SDL_GetGamepadFromPlayerIndex(player_index);
+  return {SDL_GetGamepadFromPlayerIndex(player_index)};
 }
 
 /**
@@ -1698,7 +1695,7 @@
  */
 inline PropertiesRef GetGamepadProperties(GamepadRef gamepad)
 {
-  return CheckError(SDL_GetGamepadProperties(gamepad));
+  return {CheckError(SDL_GetGamepadProperties(gamepad))};
 }
 
 inline PropertiesRef GamepadBase::GetProperties()
@@ -1706,6 +1703,16 @@
   return SDL::GetGamepadProperties(get());
 }
 
+/**
+ * Gamepad capability properties.
+ *
+ * Use Gamepad.GetProperties() to query these properties for a gamepad.
+ *
+ * @sa Gamepad.GetProperties
+ * @sa GetGamepadProperties
+ * @sa Joystick.GetProperties
+ * @sa prop::JoystickCap
+ */
 namespace prop::GamepadCap {
 
 constexpr auto MONO_LED_BOOLEAN =
@@ -2146,11 +2153,8 @@
  * Get the SDL joystick layer bindings for a gamepad.
  *
  * @param gamepad a gamepad.
- * @param count a pointer filled in with the number of bindings returned.
  * @returns a nullptr terminated array of pointers to bindings or nullptr on
- *          failure; call GetError() for more information. This is a single
- *          allocation that should be freed with free() when it is no longer
- *          needed.
+ *          failure; call GetError() for more information.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -2158,7 +2162,9 @@
  */
 inline OwnArray<GamepadBinding*> GetGamepadBindings(GamepadRef gamepad)
 {
-  return SDL_GetGamepadBindings(gamepad);
+  int count;
+  auto r = (SDL_GetGamepadBindings(gamepad, &count));
+  return OwnArray<GamepadBinding*>(r, count);
 }
 
 inline OwnArray<GamepadBinding*> GamepadBase::GetBindings()
