--- build/generated/SDL3pp_properties.h
+++ include/SDL3pp/SDL3pp_properties.h
@@ -11,14 +11,15 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryProperties Category Properties
+ * @defgroup CategoryProperties Object Properties
  *
  * A property is a variable that can be created and retrieved by name at
  * runtime.
  *
  * All properties are part of a property group (Properties). A property group
  * can be created with the CreateProperties function and destroyed with the
- * DestroyProperties function.
+ * DestroyProperties function, but the Properties destructor probably will do
+ * what you want to, automatically.
  *
  * Properties can be added to and retrieved from a property group through the
  * following functions:
@@ -246,6 +247,14 @@
    */
   void Unlock(PropertiesLock&& lock);
 
+  /**
+   * Set a property in a group of properties.
+   *
+   * @param name the name of the property to modify.
+   * @param value the new value of the property, or nullptr to delete the
+   *              property.
+   * @throws Error on failure.
+   */
   template<class V>
   void Set(StringParam name, V&& value);
 
@@ -299,7 +308,6 @@
    *              property.
    * @param cleanup the function to call when this property is deleted, or
    *                nullptr if no cleanup is necessary.
-   * @param userdata a pointer that is passed to the cleanup function.
    * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread.
@@ -581,7 +589,6 @@
    * properties. The properties are locked during enumeration.
    *
    * @param callback the function to call for each property.
-   * @param userdata a pointer that is passed to `callback`.
    * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread.
@@ -590,6 +597,13 @@
    */
   void Enumerate(EnumeratePropertiesCB callback) const;
 
+  /**
+   * Returns the number of properties this has
+   *
+   * This uses EnumerateProperties() internally, so might not be so fast
+   *
+   * @return Uint64
+   */
   Uint64 GetCount() const;
 };
 
@@ -599,6 +613,9 @@
  * @since This datatype is available since SDL 3.2.0.
  *
  * @cat resource
+ *
+ * @sa Properties.Create
+ * @sa prop
  */
 struct Properties : PropertiesBase
 {
@@ -689,7 +706,7 @@
    * or want to guarantee that properties being queried aren't freed in another
    * thread.
    *
-   * @param props the properties to lock.
+   * @param resource the properties to lock.
    * @post true on success or false on failure; call GetError() for more
    *       information.
    *
@@ -713,8 +730,6 @@
   /**
    * Unlock a group of properties.
    *
-   * @param props the properties to unlock.
-   *
    * @threadsafety It is safe to call this function from any thread.
    *
    * @since This function is available since SDL 3.2.0.
@@ -738,8 +753,6 @@
   /**
    * Unlock a group of properties.
    *
-   * @param props the properties to unlock.
-   *
    * @threadsafety It is safe to call this function from any thread.
    *
    * @since This function is available since SDL 3.2.0.
@@ -779,7 +792,7 @@
  *
  * @since This constant is available since SDL 3.4.0.
  */
-inline auto PROP_NAME_STRING = SDL_PROP_NAME_STRING;
+constexpr const char* PROP_NAME_STRING = SDL_PROP_NAME_STRING;
 
 #endif // SDL_VERSION_ATLEAST(3, 4, 0)
 
@@ -796,7 +809,7 @@
  */
 inline PropertiesRef GetGlobalProperties()
 {
-  return CheckError(SDL_GetGlobalProperties());
+  return {CheckError(SDL_GetGlobalProperties())};
 }
 
 /**
@@ -912,16 +925,52 @@
   m_lock = {};
 }
 
+/**
+ * Set a property in a group of properties.
+ *
+ * @param props the properties to modify.
+ * @param name the name of the property to modify.
+ * @param value the new value of the property, or nullptr to delete the
+ *              property.
+ * @throws Error on failure.
+ */
 template<class V>
 inline void SetProperty(PropertiesRef props, StringParam name, V&& value)
 {
-  static_assert(false, "Not implemented");
+  if constexpr (std::is_same_v<V, bool>) {
+    props.SetBooleanProperty(std::move(name), std::forward<V>(value));
+  } else if constexpr (std::is_integral_v<V>) {
+    props.SetNumberProperty(std::move(name), std::forward<V>(value));
+  } else {
+    struct Visitor
+    {
+      PropertiesRef props;
+      StringParam name;
+
+      void operator()(std::monostate) { props.ClearProperty(std::move(name)); }
+      void operator()(void* v) { props.SetPointerProperty(std::move(name), v); }
+
+      void operator()(StringParam&& v)
+      {
+        props.SetStringProperty(std::move(name), std::move(v));
+      }
+      void operator()(const char* v)
+      {
+        props.SetStringProperty(std::move(name), std::move(v));
+      }
+
+      void operator()(Sint64 v) { props.SetNumberProperty(std::move(name), v); }
+      void operator()(float v) { props.SetFloatProperty(std::move(name), v); }
+      void operator()(bool v) { props.SetBooleanProperty(std::move(name), v); }
+    };
+    Visitor{props, std::move(name)}(std::forward<V>(value));
+  }
 }
 
 template<class V>
 inline void PropertiesBase::Set(StringParam name, V&& value)
 {
-  SDL::SetProperty(get(), std::move(name), value);
+  SDL::SetProperty(get(), std::move(name), std::forward<V>(value));
 }
 
 /**
@@ -981,7 +1030,6 @@
  *              property.
  * @param cleanup the function to call when this property is deleted, or nullptr
  *                if no cleanup is necessary.
- * @param userdata a pointer that is passed to the cleanup function.
  * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
@@ -997,7 +1045,12 @@
                                           void* value,
                                           CleanupPropertyCB cleanup)
 {
-  static_assert(false, "Not implemented");
+  using Wrapper = CallbackWrapper<CleanupPropertyCB>;
+  SDL_SetPointerPropertyWithCleanup(props,
+                                    std::move(name),
+                                    value,
+                                    &Wrapper::CallOnce,
+                                    Wrapper::Wrap(std::move(cleanup)));
 }
 
 inline void PropertiesBase::SetPointerPropertyWithCleanup(
@@ -1015,7 +1068,8 @@
   void* value,
   CleanupPropertyCB cleanup)
 {
-  SDL::SetPointerPropertyWithCleanup(get(), std::move(name), value, cleanup);
+  SDL::SetPointerPropertyWithCleanup(
+    get(), std::move(name), value, std::move(cleanup));
 }
 
 /**
@@ -1433,7 +1487,6 @@
  *
  * @param props the properties to query.
  * @param callback the function to call for each property.
- * @param userdata a pointer that is passed to `callback`.
  * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
@@ -1443,7 +1496,13 @@
 inline void EnumerateProperties(PropertiesRef props,
                                 EnumeratePropertiesCB callback)
 {
-  static_assert(false, "Not implemented");
+  return EnumerateProperties(
+    props,
+    [](void* userdata, PropertiesID props, const char* name) {
+      auto& f = *static_cast<const EnumeratePropertiesCB*>(userdata);
+      f(props, name);
+    },
+    &callback);
 }
 
 inline void PropertiesBase::Enumerate(EnumeratePropertiesCallback callback,
@@ -1454,12 +1513,22 @@
 
 inline void PropertiesBase::Enumerate(EnumeratePropertiesCB callback) const
 {
-  SDL::EnumerateProperties(get(), callback);
+  SDL::EnumerateProperties(get(), std::move(callback));
 }
 
+/**
+ * Returns the number of properties this has
+ *
+ * This uses EnumerateProperties() internally, so might not be so fast
+ *
+ * @param props
+ * @return Uint64
+ */
 inline Uint64 CountProperties(PropertiesRef props)
 {
-  static_assert(false, "Not implemented");
+  Uint64 count = 0;
+  EnumerateProperties(props, [&count](auto, const char*) { count++; });
+  return count;
 }
 
 inline Uint64 PropertiesBase::GetCount() const
@@ -1492,6 +1561,17 @@
 
 /// @}
 
+/**
+ * @namespace SDL::prop
+ * @brief Constants for Properties keys.
+ *
+ * All subnamespaces in `prop` are intended to be used as namespaces for
+ * properties related to a specific type of object. For example, `prop::Surface`
+ * contains properties related to surfaces.
+ *
+ * @sa Properties
+ */
+
 } // namespace SDL
 
 #endif /* SDL3PP_PROPERTIES_H_ */
