diff --git a/include/SDL3pp/SDL3pp_bits.h b/include/SDL3pp/SDL3pp_bits.h
index c294f262..c818e3eb 100644
--- a/include/SDL3pp/SDL3pp_bits.h
+++ b/include/SDL3pp/SDL3pp_bits.h
@@ -1,10 +1,12 @@
 #ifndef SDL3PP_BITS_H_
 #define SDL3PP_BITS_H_
 
+#include <SDL3/SDL_bits.h>
+
 namespace SDL {
 
 /**
- * @defgroup CategoryBits Category Bits
+ * @defgroup CategoryBits Bit Manipulation
  *
  * Functions for fiddling with bits and bitmasks.
  *
@@ -29,9 +31,9 @@ namespace SDL {
  *
  * @since This function is available since SDL 3.2.0.
  */
-inline int MostSignificantBitIndex32(Uint32 x)
+constexpr int MostSignificantBitIndex32(Uint32 x)
 {
-  return SDL_MostSignificantBitIndex32(, x);
+  return SDL_MostSignificantBitIndex32(x);
 }
 
 /**
@@ -52,9 +54,9 @@ inline int MostSignificantBitIndex32(Uint32 x)
  *
  * @since This function is available since SDL 3.2.0.
  */
-inline bool HasExactlyOneBitSet32(Uint32 x)
+constexpr bool HasExactlyOneBitSet32(Uint32 x)
 {
-  return SDL_HasExactlyOneBitSet32(, x);
+  return SDL_HasExactlyOneBitSet32(x);
 }
 
 /// @}
