--- build/generated/SDL3pp_video.h
+++ include/SDL3pp/SDL3pp_video.h
@@ -9,7 +9,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryVideo Category Video
+ * @defgroup CategoryVideo Display and Window Management
  *
  * SDL's video subsystem is largely interested in abstracting window management
  * from the underlying operating system. You can create windows, manage them in
@@ -60,7 +60,7 @@
 struct GLContext;
 
 /// Alias to raw representation for GLContext.
-using GLContextRaw = SDL_GLContext*;
+using GLContextRaw = SDL_GLContext;
 
 // Forward decl
 struct GLContextScoped;
@@ -262,7 +262,7 @@
    * The primary display is often located at (0,0), but may be placed at a
    * different location depending on monitor layout.
    *
-   * @param rect the Rect structure filled in with the display bounds.
+   * @returns the Rect structure filled in with the display bounds.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -286,7 +286,7 @@
    * so these are good guidelines for the maximum space available to a
    * non-fullscreen window.
    *
-   * @param rect the Rect structure filled in with the display bounds.
+   * @returns the Rect structure filled in with the display bounds.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -364,12 +364,8 @@
    * - refresh rate -> highest to lowest
    * - pixel density -> lowest to highest
    *
-   * @param count a pointer filled in with the number of display modes returned,
-   *              may be nullptr.
    * @returns a nullptr terminated array of display mode pointers or nullptr on
-   *          failure; call GetError() for more information. This is a single
-   *          allocation that should be freed with free() when it is no longer
-   *          needed.
+   *          failure; call GetError() for more information.
    *
    * @threadsafety This function should only be called on the main thread.
    *
@@ -389,14 +385,12 @@
    * and finally checking the refresh rate. If all the available modes are too
    * small, then false is returned.
    *
-   * @param w the width in pixels of the desired display mode.
-   * @param h the height in pixels of the desired display mode.
+   * @param size the width and height in pixels of the desired display mode.
    * @param refresh_rate the refresh rate of the desired display mode, or 0.0f
    *                     for the desktop refresh rate.
    * @param include_high_density_modes boolean to include high density modes in
    *                                   the search.
-   * @param closest a pointer filled in with the closest display mode equal to
-   *                or larger than the desired mode.
+   * @returns the closest display mode equal to or larger than the desired mode.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -472,9 +466,9 @@
  * backend. Setting this property after the video subsystem has been initialized
  * has no effect, and reading it when the video subsystem is uninitialized will
  * either return the user provided value, if one was set prior to
- * initialization, or NULL. See docs/README-wayland.md for more information.
+ * initialization, or nullptr. See docs/README-wayland.md for more information.
  *
- * \since This macro is available since SDL 3.2.0.
+ * @since This constant is available since SDL 3.2.0.
  */
 constexpr auto VIDEO_WAYLAND_WL_DISPLAY_POINTER =
   SDL_PROP_GLOBAL_VIDEO_WAYLAND_WL_DISPLAY_POINTER;
@@ -642,6 +636,7 @@
  * @since This enum is available since SDL 3.2.0.
  *
  * @sa HitTest
+ * @sa HitTestCB
  */
 using HitTestResult = SDL_HitTestResult;
 
@@ -698,11 +693,11 @@
  *
  * @param win the Window where hit-testing was set on.
  * @param area an Point which should be hit-tested.
- * @param data what was passed as `callback_data` to SetWindowHitTest().
  * @returns an HitTestResult value.
  *
- * @sa SetWindowHitTest
+ * @cat listener-callback
  *
+ * @sa SetWindowHitTest
  * @sa HitTest
  */
 using HitTestCB =
@@ -728,8 +723,9 @@
 constexpr ProgressState PROGRESS_STATE_INVALID = SDL_PROGRESS_STATE_INVALID;
 
 constexpr ProgressState PROGRESS_STATE_NONE =
-  SDL_PROGRESS_STATE_NONE; ///< No progress bar is shown
+  SDL_PROGRESS_STATE_NONE; ///< No progress bar is shown.
 
+/// The progress bar is shown in a indeterminate state.
 constexpr ProgressState PROGRESS_STATE_INDETERMINATE =
   SDL_PROGRESS_STATE_INDETERMINATE; ///< The progress bar is shown in a
                                     ///< indeterminate state
@@ -916,7 +912,6 @@
   /**
    * Get the raw ICC profile data for the screen the window is currently on.
    *
-   * @param size the size of the ICC profile.
    * @returns the raw ICC profile data on success.
    * @throws Error on failure.
    *
@@ -1194,9 +1189,7 @@
    * Additionally, as this is just a request, it can be denied by the windowing
    * system.
    *
-   * @param x the x coordinate of the window, or `WINDOWPOS_CENTERED` or
-   *          `WINDOWPOS_UNDEFINED`.
-   * @param y the y coordinate of the window, or `WINDOWPOS_CENTERED` or
+   * @param p the coordinates of the window, or `WINDOWPOS_CENTERED` or
    *          `WINDOWPOS_UNDEFINED`.
    * @throws Error on failure.
    *
@@ -1241,10 +1234,7 @@
    * If you do not need the value for one of the positions a nullptr may be
    * passed in the `x` or `y` parameter.
    *
-   * @param x a pointer filled in with the x position of the window, may be
-   *          nullptr.
-   * @param y a pointer filled in with the y position of the window, may be
-   *          nullptr.
+   * @returns the position on success.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1276,8 +1266,7 @@
    * to remain within the usable desktop bounds). Additionally, as this is just
    * a request, it can be denied by the windowing system.
    *
-   * @param w the width of the window, must be > 0.
-   * @param h the height of the window, must be > 0.
+   * @param size the width and height of the window, must be > 0.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1319,8 +1308,7 @@
    * window is on a high pixel density display. Use GetWindowSizeInPixels() or
    * GetRenderOutputSize() to get the real client area size in pixels.
    *
-   * @param w a pointer filled in with the width of the window, may be nullptr.
-   * @param h a pointer filled in with the height of the window, may be nullptr.
+   * @returns a point with width and height on success
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1331,11 +1319,36 @@
    * @sa GetWindowSizeInPixels
    * @sa SetWindowSize
    * @sa EVENT_WINDOW_RESIZED
+   * @sa GetSize(int *, int *)
    */
   Point GetSize() const;
 
+  /**
+   * @brief Request the window's position and size to be set.
+   *
+   * @param rect the rect containing the new coordinates
+   * @throws Error on failure.
+   *
+   * @threadsafety This function should only be called on the main thread.
+   *
+   * @sa WindowRef.SetPosition()
+   * @sa WindowRef.SetSize()
+   */
   void SetRect(Rect rect);
 
+  /**
+   * Get the position and client size of a window.
+   *
+   * This is the current position of the window as last reported by the
+   * windowing system.
+   *
+   * The window pixel size may differ from its window coordinate size if the
+   * window is on a high pixel density display. Use Window.GetSizeInPixels()
+   * or RendererRef.GetOutputSize() to get the real client area size in pixels.
+   *
+   * @return Rect with the position and size
+   * @throws Error on failure.
+   */
   Rect GetRect() const;
 
   /**
@@ -1348,8 +1361,6 @@
    * into the rest of the window, but it should not contain visually important
    * or interactable content.
    *
-   * @param rect a pointer filled in with the client area that is safe for
-   *             interactive content.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1468,10 +1479,8 @@
   /**
    * Get the size of a window's client area, in pixels.
    *
-   * @param w a pointer to variable for storing the width in pixels, may be
-   *          nullptr.
-   * @param h a pointer to variable for storing the height in pixels, may be
-   *          nullptr.
+   * @returns the size on success or std::nullopt on failure; call GetError()
+   * for more information.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1486,8 +1495,7 @@
   /**
    * Set the minimum size of a window's client area.
    *
-   * @param min_w the minimum width of the window, or 0 for no limit.
-   * @param min_h the minimum height of the window, or 0 for no limit.
+   * @param p the minimum width and heigh of the window, or 0 for no limit.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1520,8 +1528,7 @@
   /**
    * Set the maximum size of a window's client area.
    *
-   * @param max_w the maximum width of the window, or 0 for no limit.
-   * @param max_h the maximum height of the window, or 0 for no limit.
+   * @param p the maximum width and height of the window, or 0 for no limit.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1902,8 +1909,8 @@
   /**
    * Get VSync for the window surface.
    *
-   * @param vsync an int filled with the current vertical refresh sync interval.
-   *              See SetWindowSurfaceVSync() for the meaning of the value.
+   * @returns the current vertical refresh sync interval. See
+   *          SetWindowSurfaceVSync() for the meaning of the value.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1948,7 +1955,6 @@
    *
    * @param rects an array of Rect structures representing areas of the surface
    *              to copy, in pixels.
-   * @param numrects the number of rectangles.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -2197,10 +2203,8 @@
    * On platforms or desktops where this is unsupported, this function does
    * nothing.
    *
-   * @param x the x coordinate of the menu, relative to the origin (top-left) of
-   *          the client area.
-   * @param y the y coordinate of the menu, relative to the origin (top-left) of
-   *          the client area.
+   * @param p the x, y coordinates of the menu, relative to the origin
+   *          (top-left) of the client area.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -2284,12 +2288,13 @@
    * devoid of allocations, etc.
    *
    * @param callback the function to call when doing a hit-test.
-   * @param callback_data an app-defined void pointer passed to **callback**.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
    *
    * @since This function is available since SDL 3.2.0.
+   *
+   * @cat listener-callback
    */
   void SetHitTest(HitTestCB callback);
 
@@ -2473,9 +2478,7 @@
    * activates an IME, which can prevent some key press events from being passed
    * through.
    *
-   * @param window the window to enable text input.
-   * @returns true on success or false on failure; call GetError() for more
-   *          information.
+   * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
    *
@@ -2524,10 +2527,8 @@
    *   overriding other properties. This is documented at
    *   https://developer.android.com/reference/android/text/InputType
    *
-   * @param window the window to enable text input.
    * @param props the properties to use.
-   * @returns true on success or false on failure; call GetError() for more
-   *          information.
+   * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
    *
@@ -2543,7 +2544,6 @@
   /**
    * Check whether or not Unicode text input events are enabled for a window.
    *
-   * @param window the window to check.
    * @returns true if text input events are enabled else false.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -2559,9 +2559,7 @@
    *
    * If StartTextInput() showed the screen keyboard, this function will hide it.
    *
-   * @param window the window to disable text input.
-   * @returns true on success or false on failure; call GetError() for more
-   *          information.
+   * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
    *
@@ -2574,9 +2572,7 @@
   /**
    * Dismiss the composition window/IME without disabling the subsystem.
    *
-   * @param window the window to affect.
-   * @returns true on success or false on failure; call GetError() for more
-   *          information.
+   * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
    *
@@ -2593,13 +2589,11 @@
    * Native input methods may place a window with word suggestions near the
    * cursor, without covering the text being entered.
    *
-   * @param window the window for which to set the text input area.
    * @param rect the Rect representing the text input area, in window
    *             coordinates, or nullptr to clear it.
    * @param cursor the offset of the current cursor location relative to
    *               `rect->x`, in window coordinates.
-   * @returns true on success or false on failure; call GetError() for more
-   *          information.
+   * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
    *
@@ -2615,13 +2609,11 @@
    *
    * This returns the values previously set by SetTextInputArea().
    *
-   * @param window the window for which to query the text input area.
    * @param rect a pointer to an Rect filled in with the text input area, may be
    *             nullptr.
    * @param cursor a pointer to the offset of the current cursor location
    *               relative to `rect->x`, may be nullptr.
-   * @returns true on success or false on failure; call GetError() for more
-   *          information.
+   * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
    *
@@ -2634,7 +2626,6 @@
   /**
    * Check whether the screen keyboard is shown for given window.
    *
-   * @param window the window for which screen keyboard should be queried.
    * @returns true if screen keyboard is shown or false if not.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -2655,10 +2646,7 @@
    * Note that this function will appear to succeed, but not actually move the
    * mouse when used over Microsoft Remote Desktop.
    *
-   * @param window the window to move the mouse into, or nullptr for the current
-   *               mouse focus.
-   * @param x the x coordinate within the window.
-   * @param y the y coordinate within the window.
+   * @param p the x, y coordinates within the window.
    *
    * @threadsafety This function should only be called on the main thread.
    *
@@ -2683,10 +2671,8 @@
    *
    * This function will flush any pending mouse motion for this window.
    *
-   * @param window the window to change.
    * @param enabled true to enable relative mode, false to disable.
-   * @returns true on success or false on failure; call GetError() for more
-   *          information.
+   * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
    *
@@ -2699,7 +2685,6 @@
   /**
    * Query whether relative mouse mode is enabled for a window.
    *
-   * @param window the window to query.
    * @returns true if relative mode is enabled for a window or false otherwise.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -2713,9 +2698,8 @@
   /**
    * Get the renderer associated with a window.
    *
-   * @param window the window to query.
-   * @returns the rendering context on success or nullptr on failure; call
-   *          GetError() for more information.
+   * @returns the rendering context on success.
+   * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
@@ -2729,9 +2713,9 @@
  *
  * @since This struct is available since SDL 3.2.0.
  *
- * @sa CreateWindow
- *
  * @cat resource
+ *
+ * @sa CreateWindow
  */
 struct Window : WindowBase
 {
@@ -2762,14 +2746,11 @@
    * Create a window and default renderer.
    *
    * @param title the title of the window, in UTF-8 encoding.
-   * @param width the width of the window.
-   * @param height the height of the window.
+   * @param size the width and height of the window.
    * @param window_flags the flags used to create the window (see
    *                     CreateWindow()).
-   * @param window a pointer filled with the window, or nullptr on error.
-   * @param renderer a pointer filled with the renderer, or nullptr on error.
-   * @post true on success or false on failure; call GetError() for more
-   *       information.
+   * @param renderer a pointer filled with the renderer.
+   * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
    *
@@ -2854,8 +2835,7 @@
    * in a future version of SDL.
    *
    * @param title the title of the window, in UTF-8 encoding.
-   * @param w the width of the window.
-   * @param h the height of the window.
+   * @param size the width and height of the window.
    * @param flags 0, or one or more WindowFlags OR'd together.
    * @post the window that was created or nullptr on failure; call GetError()
    *       for more information.
@@ -2925,12 +2905,9 @@
    * hidden will be restored when the parent is shown.
    *
    * @param parent the parent of the window, must not be nullptr.
-   * @param offset_x the x position of the popup window relative to the origin
-   *                 of the parent.
-   * @param offset_y the y position of the popup window relative to the origin
-   *                 of the parent window.
-   * @param w the width of the window.
-   * @param h the height of the window.
+   * @param offset the x, y position of the popup window relative to the origin
+   *               of the parent.
+   * @param size the width and height of the window.
    * @param flags WINDOW_TOOLTIP or WINDOW_POPUP_MENU, and zero or more
    *              additional WindowFlags OR'd together.
    * @post the window that was created or nullptr on failure; call GetError()
@@ -3270,9 +3247,6 @@
   /**
    * Delete an OpenGL context.
    *
-   * @returns true on success or false on failure; call GetError() for more
-   *          information.
-   *
    * @threadsafety This function should only be called on the main thread.
    *
    * @since This function is available since SDL 3.2.0.
@@ -3287,7 +3261,6 @@
    * The context must have been created with a compatible window.
    *
    * @param window the window to associate with the context.
-   * @param context the OpenGL context to associate with the window.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -3319,7 +3292,7 @@
   }
 
   /// Destructor
-  ~GLContextScoped() { Destroy(); }
+  ~GLContextScoped() { SDL_GL_DestroyContext(release()); }
 };
 
 /**
@@ -3694,17 +3667,19 @@
 /**
  * Get a list of currently connected displays.
  *
- * @param count a pointer filled in with the number of displays returned, may be
- *              nullptr.
  * @returns a 0 terminated array of display instance IDs or nullptr on failure;
- *          call GetError() for more information. This should be freed with
- *          free() when it is no longer needed.
+ *          call GetError() for more information.
  *
  * @threadsafety This function should only be called on the main thread.
  *
  * @since This function is available since SDL 3.2.0.
  */
-inline OwnArray<DisplayID> GetDisplays() { return SDL_GetDisplays(); }
+inline OwnArray<DisplayID> GetDisplays()
+{
+  int count = 0;
+  auto data = SDL_GetDisplays(&count);
+  return OwnArray<DisplayID>{data, size_t(count)};
+}
 
 /**
  * Return the primary display.
@@ -3772,6 +3747,11 @@
   return SDL::GetDisplayProperties(m_displayID);
 }
 
+/**
+ * Properties for Display.
+ *
+ * @sa Display.GetProperties
+ */
 namespace prop::Display {
 
 constexpr auto HDR_ENABLED_BOOLEAN =
@@ -3823,7 +3803,7 @@
  * different location depending on monitor layout.
  *
  * @param displayID the instance ID of the display to query.
- * @param rect the Rect structure filled in with the display bounds.
+ * @returns the Rect structure filled in with the display bounds.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -3835,7 +3815,9 @@
  */
 inline Rect GetDisplayBounds(Display displayID)
 {
-  return CheckError(SDL_GetDisplayBounds(displayID));
+  Rect bounds;
+  SDL_GetDisplayBounds(displayID, &bounds);
+  return bounds;
 }
 
 inline Rect Display::GetBounds() const
@@ -3855,7 +3837,7 @@
  * window.
  *
  * @param displayID the instance ID of the display to query.
- * @param rect the Rect structure filled in with the display bounds.
+ * @returns the Rect structure filled in with the display bounds.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -3867,7 +3849,9 @@
  */
 inline Rect GetDisplayUsableBounds(Display displayID)
 {
-  return CheckError(SDL_GetDisplayUsableBounds(displayID));
+  Rect bounds;
+  SDL_GetDisplayUsableBounds(displayID, &bounds);
+  return bounds;
 }
 
 inline Rect Display::GetUsableBounds() const
@@ -3969,12 +3953,8 @@
  * - pixel density -> lowest to highest
  *
  * @param displayID the instance ID of the display to query.
- * @param count a pointer filled in with the number of display modes returned,
- *              may be nullptr.
  * @returns a nullptr terminated array of display mode pointers or nullptr on
- *          failure; call GetError() for more information. This is a single
- *          allocation that should be freed with free() when it is no longer
- *          needed.
+ *          failure; call GetError() for more information.
  *
  * @threadsafety This function should only be called on the main thread.
  *
@@ -3984,7 +3964,9 @@
  */
 inline OwnArray<DisplayMode*> GetFullscreenDisplayModes(Display displayID)
 {
-  return SDL_GetFullscreenDisplayModes(displayID);
+  int count = 0;
+  auto data = CheckError(SDL_GetFullscreenDisplayModes(displayID, &count));
+  return OwnArray<DisplayMode*>{data, size_t(count)};
 }
 
 inline OwnArray<DisplayMode*> Display::GetFullscreenModes() const
@@ -4003,14 +3985,12 @@
  * then false is returned.
  *
  * @param displayID the instance ID of the display to query.
- * @param w the width in pixels of the desired display mode.
- * @param h the height in pixels of the desired display mode.
+ * @param size the width and height in pixels of the desired display mode.
  * @param refresh_rate the refresh rate of the desired display mode, or 0.0f for
  *                     the desktop refresh rate.
  * @param include_high_density_modes boolean to include high density modes in
  *                                   the search.
- * @param closest a pointer filled in with the closest display mode equal to or
- *                larger than the desired mode.
+ * @returns the closest display mode equal to or larger than the desired mode.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -4026,8 +4006,14 @@
   float refresh_rate,
   bool include_high_density_modes)
 {
-  return CheckError(SDL_GetClosestFullscreenDisplayMode(
-    displayID, size, refresh_rate, include_high_density_modes));
+  DisplayMode mode;
+  CheckError(SDL_GetClosestFullscreenDisplayMode(displayID,
+                                                 size.x,
+                                                 size.y,
+                                                 refresh_rate,
+                                                 include_high_density_modes,
+                                                 &mode));
+  return mode;
 }
 
 inline DisplayMode Display::GetClosestFullscreenMode(
@@ -4060,7 +4046,7 @@
  */
 inline const DisplayMode& GetDesktopDisplayMode(DisplayID displayID)
 {
-  return SDL_GetDesktopDisplayMode(displayID);
+  return *SDL_GetDesktopDisplayMode(displayID);
 }
 
 inline const DisplayMode& Display::GetDesktopMode() const
@@ -4089,7 +4075,7 @@
  */
 inline const DisplayMode& GetCurrentDisplayMode(DisplayID displayID)
 {
-  return SDL_GetCurrentDisplayMode(displayID);
+  return *SDL_GetCurrentDisplayMode(displayID);
 }
 
 inline const DisplayMode& Display::GetCurrentMode() const
@@ -4113,7 +4099,7 @@
  */
 inline Display GetDisplayForPoint(const PointRaw& point)
 {
-  return SDL_GetDisplayForPoint(point);
+  return SDL_GetDisplayForPoint(&point);
 }
 
 inline Display Display::GetForPoint(const PointRaw& point)
@@ -4138,7 +4124,7 @@
  */
 inline Display GetDisplayForRect(const RectRaw& rect)
 {
-  return CheckError(SDL_GetDisplayForRect(rect));
+  return CheckError(SDL_GetDisplayForRect(&rect));
 }
 
 inline Display Display::GetForRect(const RectRaw& rect)
@@ -4287,7 +4273,7 @@
  */
 inline const DisplayMode& GetWindowFullscreenMode(WindowRef window)
 {
-  return SDL_GetWindowFullscreenMode(window);
+  return *SDL_GetWindowFullscreenMode(window);
 }
 
 inline const DisplayMode& WindowBase::GetFullscreenMode() const
@@ -4299,7 +4285,6 @@
  * Get the raw ICC profile data for the screen the window is currently on.
  *
  * @param window the window to query.
- * @param size the size of the ICC profile.
  * @returns the raw ICC profile data on success.
  * @throws Error on failure.
  *
@@ -4309,7 +4294,8 @@
  */
 inline OwnPtr<void> GetWindowICCProfile(WindowRef window)
 {
-  return CheckError(SDL_GetWindowICCProfile(window));
+  size_t size;
+  return OwnPtr<void>{CheckError(SDL_GetWindowICCProfile(window, &size))};
 }
 
 inline OwnPtr<void> WindowBase::GetICCProfile() const
@@ -4341,17 +4327,19 @@
 /**
  * Get a list of valid windows.
  *
- * @param count a pointer filled in with the number of windows returned, may be
- *              nullptr.
  * @returns a nullptr terminated array of Window pointers or nullptr on failure;
- *          call GetError() for more information. This is a single allocation
- *          that should be freed with free() when it is no longer needed.
+ *          call GetError() for more information.
  *
  * @threadsafety This function should only be called on the main thread.
  *
  * @since This function is available since SDL 3.2.0.
  */
-inline OwnArray<WindowRef> GetWindows() { return SDL_GetWindows(); }
+inline OwnArray<WindowRef> GetWindows()
+{
+  int count = 0;
+  auto data = CheckError(SDL_GetWindows(&count));
+  return OwnArray<WindowRef>{reinterpret_cast<WindowRef*>(data), size_t(count)};
+}
 
 /**
  * Create a window with the specified dimensions and flags.
@@ -4423,8 +4411,7 @@
  * in a future version of SDL.
  *
  * @param title the title of the window, in UTF-8 encoding.
- * @param w the width of the window.
- * @param h the height of the window.
+ * @param size the width and height of the window.
  * @param flags 0, or one or more WindowFlags OR'd together.
  * @returns the window that was created or nullptr on failure; call GetError()
  *          for more information.
@@ -4448,7 +4435,7 @@
 inline Window::Window(StringParam title,
                       const PointRaw& size,
                       WindowFlags flags)
-  : Window(SDL_CreateWindow(title, size, flags))
+  : Window(SDL_CreateWindow(title, size.x, size.y, flags))
 {
 }
 
@@ -4456,7 +4443,8 @@
                       const PointRaw& offset,
                       const PointRaw& size,
                       WindowFlags flags)
-  : Window(SDL_CreatePopupWindow(parent, offset, size, flags))
+  : Window(
+      SDL_CreatePopupWindow(parent, offset.x, offset.y, size.x, size.y, flags))
 {
 }
 
@@ -4519,12 +4507,9 @@
  * hidden will be restored when the parent is shown.
  *
  * @param parent the parent of the window, must not be nullptr.
- * @param offset_x the x position of the popup window relative to the origin of
- *                 the parent.
- * @param offset_y the y position of the popup window relative to the origin of
- *                 the parent window.
- * @param w the width of the window.
- * @param h the height of the window.
+ * @param offset the x, y position of the popup window relative to the origin
+ *               of the parent.
+ * @param size the width and height of the window.
  * @param flags WINDOW_TOOLTIP or WINDOW_POPUP_MENU, and zero or more additional
  *              WindowFlags OR'd together.
  * @returns the window that was created or nullptr on failure; call GetError()
@@ -4695,6 +4680,11 @@
   return Window(props);
 }
 
+/**
+ * Properties for Window creation.
+ *
+ * @sa CreateWindowWithProperties
+ */
 namespace prop::Window::Create {
 
 constexpr auto ALWAYS_ON_TOP_BOOLEAN =
@@ -5044,6 +5034,17 @@
   return SDL::GetWindowProperties(get());
 }
 
+/**
+ * Properties for Window
+ *
+ * Properties within `Create` sub=namespace can be used in
+ * CreateWindowWithProperties() to specify window creation parameters. can be
+ * used to query information about an existing Window with
+ * Window.GetProperties()
+ *
+ * @sa CreateWindowWithProperties
+ * @sa Window.GetProperties
+ */
 namespace prop::Window {
 
 constexpr auto SHAPE_POINTER =
@@ -5106,6 +5107,10 @@
 constexpr auto OPENVR_OVERLAY_ID_NUMBER =
   SDL_PROP_WINDOW_OPENVR_OVERLAY_ID_NUMBER; ///< Number for openvr overlay id.
 
+#else
+
+constexpr auto OPENVR_OVERLAY_ID_NUMBER = SDL_PROP_WINDOW_OPENVR_OVERLAY_ID;
+
 #endif // SDL_VERSION_ATLEAST(3, 4, 0)
 
 constexpr auto VIVANTE_DISPLAY_POINTER =
@@ -5316,9 +5321,7 @@
  * this is just a request, it can be denied by the windowing system.
  *
  * @param window the window to reposition.
- * @param x the x coordinate of the window, or `WINDOWPOS_CENTERED` or
- *          `WINDOWPOS_UNDEFINED`.
- * @param y the y coordinate of the window, or `WINDOWPOS_CENTERED` or
+ * @param p the coordinates of the window, or `WINDOWPOS_CENTERED` or
  *          `WINDOWPOS_UNDEFINED`.
  * @throws Error on failure.
  *
@@ -5331,7 +5334,7 @@
  */
 inline void SetWindowPosition(WindowRef window, const PointRaw& p)
 {
-  CheckError(SDL_SetWindowPosition(window, p));
+  CheckError(SDL_SetWindowPosition(window, p.x, p.y));
 }
 
 inline void WindowBase::SetPosition(const PointRaw& p)
@@ -5376,10 +5379,7 @@
  * in the `x` or `y` parameter.
  *
  * @param window the window to query.
- * @param x a pointer filled in with the x position of the window, may be
- *          nullptr.
- * @param y a pointer filled in with the y position of the window, may be
- *          nullptr.
+ * @returns the position on success.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5390,7 +5390,9 @@
  */
 inline Point GetWindowPosition(WindowRef window)
 {
-  static_assert(false, "Not implemented");
+  Point p;
+  GetWindowPosition(window, &p.x, &p.y);
+  return p;
 }
 
 inline void WindowBase::GetPosition(int* x, int* y) const
@@ -5425,8 +5427,7 @@
  * request, it can be denied by the windowing system.
  *
  * @param window the window to change.
- * @param w the width of the window, must be > 0.
- * @param h the height of the window, must be > 0.
+ * @param size the width and height of the window, must be > 0.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5439,7 +5440,7 @@
  */
 inline void SetWindowSize(WindowRef window, const PointRaw& size)
 {
-  CheckError(SDL_SetWindowSize(window, size));
+  CheckError(SDL_SetWindowSize(window, size.x, size.y));
 }
 
 inline void WindowBase::SetSize(const PointRaw& size)
@@ -5481,8 +5482,7 @@
  * GetRenderOutputSize() to get the real client area size in pixels.
  *
  * @param window the window to query the width and height from.
- * @param w a pointer filled in with the width of the window, may be nullptr.
- * @param h a pointer filled in with the height of the window, may be nullptr.
+ * @returns a point with width and height on success
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5496,7 +5496,9 @@
  */
 inline Point GetWindowSize(WindowRef window)
 {
-  static_assert(false, "Not implemented");
+  Point p;
+  GetWindowSize(window, &p.x, &p.y);
+  return p;
 }
 
 inline void WindowBase::GetSize(int* w, int* h) const
@@ -5506,16 +5508,43 @@
 
 inline Point WindowBase::GetSize() const { return SDL::GetWindowSize(get()); }
 
+/**
+ * @brief Request the window's position and size to be set.
+ *
+ * @param window the window to query the width and height from.
+ * @param rect the rect containing the new coordinates
+ * @throws Error on failure.
+ *
+ * @threadsafety This function should only be called on the main thread.
+ *
+ * @sa WindowRef.SetPosition()
+ * @sa WindowRef.SetSize()
+ */
 inline void SetWindowRect(WindowRef window, Rect rect)
 {
-  static_assert(false, "Not implemented");
+  SetWindowPosition(window, rect.GetTopLeft());
+  SetWindowSize(window, rect.GetSize());
 }
 
 inline void WindowBase::SetRect(Rect rect) { SDL::SetWindowRect(get(), rect); }
 
+/**
+ * Get the position and client size of a window.
+ *
+ * This is the current position of the window as last reported by the
+ * windowing system.
+ *
+ * The window pixel size may differ from its window coordinate size if the
+ * window is on a high pixel density display. Use Window.GetSizeInPixels()
+ * or RendererRef.GetOutputSize() to get the real client area size in pixels.
+ *
+ * @param window the window to query the width and height from.
+ * @return Rect with the position and size
+ * @throws Error on failure.
+ */
 inline Rect GetWindowRect(WindowRef window)
 {
-  static_assert(false, "Not implemented");
+  return Rect{GetWindowPosition(window), GetWindowSize(window)};
 }
 
 inline Rect WindowBase::GetRect() const { return SDL::GetWindowRect(get()); }
@@ -5531,8 +5560,6 @@
  * interactable content.
  *
  * @param window the window to query.
- * @param rect a pointer filled in with the client area that is safe for
- *             interactive content.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5541,7 +5568,9 @@
  */
 inline Rect GetWindowSafeArea(WindowRef window)
 {
-  return CheckError(SDL_GetWindowSafeArea(window));
+  Rect rect;
+  CheckError(SDL_GetWindowSafeArea(window, &rect));
+  return rect;
 }
 
 inline Rect WindowBase::GetSafeArea() const
@@ -5703,10 +5732,8 @@
  * Get the size of a window's client area, in pixels.
  *
  * @param window the window from which the drawable size should be queried.
- * @param w a pointer to variable for storing the width in pixels, may be
- *          nullptr.
- * @param h a pointer to variable for storing the height in pixels, may be
- *          nullptr.
+ * @returns the size on success or std::nullopt on failure; call GetError()
+ * for more information.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5718,7 +5745,9 @@
  */
 inline Point GetWindowSizeInPixels(WindowRef window)
 {
-  static_assert(false, "Not implemented");
+  Point p;
+  GetWindowSizeInPixels(window, &p.x, &p.y);
+  return p;
 }
 
 inline void WindowBase::GetSizeInPixels(int* w, int* h) const
@@ -5735,8 +5764,7 @@
  * Set the minimum size of a window's client area.
  *
  * @param window the window to change.
- * @param min_w the minimum width of the window, or 0 for no limit.
- * @param min_h the minimum height of the window, or 0 for no limit.
+ * @param p the minimum width and height of the window, or 0 for no limit.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5748,7 +5776,7 @@
  */
 inline void SetWindowMinimumSize(WindowRef window, const PointRaw& p)
 {
-  CheckError(SDL_SetWindowMinimumSize(window, p));
+  CheckError(SDL_SetWindowMinimumSize(window, p.x, p.y));
 }
 
 inline void WindowBase::SetMinimumSize(const PointRaw& p)
@@ -5787,8 +5815,7 @@
  * Set the maximum size of a window's client area.
  *
  * @param window the window to change.
- * @param max_w the maximum width of the window, or 0 for no limit.
- * @param max_h the maximum height of the window, or 0 for no limit.
+ * @param p the maximum width and height of the window, or 0 for no limit.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5800,7 +5827,7 @@
  */
 inline void SetWindowMaximumSize(WindowRef window, const PointRaw& p)
 {
-  CheckError(SDL_SetWindowMaximumSize(window, p));
+  CheckError(SDL_SetWindowMaximumSize(window, p.x, p.y));
 }
 
 inline void WindowBase::SetMaximumSize(const PointRaw& p)
@@ -6247,7 +6274,7 @@
  */
 inline Surface GetWindowSurface(WindowRef window)
 {
-  return CheckError(SDL_GetWindowSurface(window));
+  return Surface::Borrow(CheckError(SDL_GetWindowSurface(window)));
 }
 
 inline Surface WindowBase::GetSurface() { return SDL::GetWindowSurface(get()); }
@@ -6285,16 +6312,18 @@
   SDL::SetWindowSurfaceVSync(get(), vsync);
 }
 
+/// Constant to disable vsync
 constexpr int WINDOW_SURFACE_VSYNC_DISABLED = SDL_WINDOW_SURFACE_VSYNC_DISABLED;
 
+/// Constant to enable adaptive vsync
 constexpr int WINDOW_SURFACE_VSYNC_ADAPTIVE = SDL_WINDOW_SURFACE_VSYNC_ADAPTIVE;
 
 /**
  * Get VSync for the window surface.
  *
  * @param window the window to query.
- * @param vsync an int filled with the current vertical refresh sync interval.
- *              See SetWindowSurfaceVSync() for the meaning of the value.
+ * @returns an int with the current vertical refresh sync interval. See
+ *          SetWindowSurfaceVSync() for the meaning of the value.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -6305,7 +6334,9 @@
  */
 inline int GetWindowSurfaceVSync(WindowRef window)
 {
-  return CheckError(SDL_GetWindowSurfaceVSync(window));
+  int vsync;
+  CheckError(SDL_GetWindowSurfaceVSync(window, &vsync));
+  return vsync;
 }
 
 inline int WindowBase::GetSurfaceVSync() const
@@ -6354,7 +6385,6 @@
  * @param window the window to update.
  * @param rects an array of Rect structures representing areas of the surface to
  *              copy, in pixels.
- * @param numrects the number of rectangles.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -6367,7 +6397,8 @@
 inline void UpdateWindowSurfaceRects(WindowRef window,
                                      SpanRef<const RectRaw> rects)
 {
-  CheckError(SDL_UpdateWindowSurfaceRects(window, rects));
+  CheckError(SDL_UpdateWindowSurfaceRects(
+    window, rects.data(), narrowS32(rects.size())));
 }
 
 inline void WindowBase::UpdateSurfaceRects(SpanRef<const RectRaw> rects)
@@ -6545,7 +6576,7 @@
  */
 inline void SetWindowMouseRect(WindowRef window, const RectRaw& rect)
 {
-  CheckError(SDL_SetWindowMouseRect(window, rect));
+  CheckError(SDL_SetWindowMouseRect(window, &rect));
 }
 
 inline void WindowBase::SetMouseRect(const RectRaw& rect)
@@ -6733,9 +6764,7 @@
  * nothing.
  *
  * @param window the window for which the menu will be displayed.
- * @param x the x coordinate of the menu, relative to the origin (top-left) of
- *          the client area.
- * @param y the y coordinate of the menu, relative to the origin (top-left) of
+ * @param p the coordinates of the menu, relative to the origin (top-left) of
  *          the client area.
  * @throws Error on failure.
  *
@@ -6745,7 +6774,7 @@
  */
 inline void ShowWindowSystemMenu(WindowRef window, const PointRaw& p)
 {
-  CheckError(SDL_ShowWindowSystemMenu(window, p));
+  CheckError(SDL_ShowWindowSystemMenu(window, p.x, p.y));
 }
 
 inline void WindowBase::ShowSystemMenu(const PointRaw& p)
@@ -6835,7 +6864,6 @@
  *
  * @param window the window to set hit-testing on.
  * @param callback the function to call when doing a hit-test.
- * @param callback_data an app-defined void pointer passed to **callback**.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -6844,7 +6872,7 @@
  */
 inline void SetWindowHitTest(WindowRef window, HitTestCB callback)
 {
-  static_assert(false, "Not implemented");
+  SetWindowHitTest(window, callback.wrapper, callback.data);
 }
 
 inline void WindowBase::SetHitTest(HitTest callback, void* callback_data)
@@ -7342,7 +7370,7 @@
 
 inline void GLContext::MakeCurrent(WindowRef window)
 {
-  SDL::WindowBase::MakeCurrent(get(), window);
+  GL_MakeCurrent(window, get());
 }
 
 /**
