--- src/generated/SDL3pp_mutex.h
+++ include/SDL3pp/SDL3pp_mutex.h
@@ -8,7 +8,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryMutex Category Mutex
+ * @defgroup CategoryMutex Thread Synchronization Primitives
  *
  * SDL offers several thread synchronization primitives. This document can't
  * cover the complicated topic of thread safety, but reading up on what each of
@@ -197,7 +197,7 @@
 
 public:
   /// Default ctor
-  constexpr Mutex(std::nullptr_t = nullptr) noexcept
+  constexpr Mutex(std::nullptr_t) noexcept
     : m_resource(0)
   {
   }
@@ -541,7 +541,7 @@
 
 public:
   /// Default ctor
-  constexpr RWLock(std::nullptr_t = nullptr) noexcept
+  constexpr RWLock(std::nullptr_t) noexcept
     : m_resource(0)
   {
   }
@@ -1252,8 +1252,8 @@
    * pointed to by `sem` has a positive value or the specified time has elapsed.
    * If the call is successful it will atomically decrement the semaphore value.
    *
-   * @param timeoutMS the length of the timeout, in milliseconds, or -1 to wait
-   *                  indefinitely.
+   * @param timeout the length of the timeout, in milliseconds, or -1 to wait
+   *                indefinitely.
    * @returns true if the wait succeeds or false if the wait times out.
    *
    * @since This function is available since SDL 3.2.0.
@@ -1421,8 +1421,8 @@
  * call is successful it will atomically decrement the semaphore value.
  *
  * @param sem the semaphore to wait on.
- * @param timeoutMS the length of the timeout, in milliseconds, or -1 to wait
- *                  indefinitely.
+ * @param timeout the length of the timeout, in milliseconds, or -1 to wait
+ *                indefinitely.
  * @returns true if the wait succeeds or false if the wait times out.
  *
  * @since This function is available since SDL 3.2.0.
@@ -1434,7 +1434,7 @@
 inline bool WaitSemaphoreTimeout(SemaphoreParam sem,
                                  std::chrono::milliseconds timeout)
 {
-  return SDL_WaitSemaphoreTimeout(sem, timeout);
+  return SDL_WaitSemaphoreTimeout(sem, timeout.count());
 }
 
 inline bool Semaphore::WaitTimeout(std::chrono::milliseconds timeout)
@@ -1496,7 +1496,7 @@
 
 public:
   /// Default ctor
-  constexpr Condition(std::nullptr_t = nullptr) noexcept
+  constexpr Condition(std::nullptr_t) noexcept
     : m_resource(0)
   {
   }
@@ -1656,8 +1656,8 @@
    * behavior.
    *
    * @param mutex the mutex used to coordinate thread access.
-   * @param timeoutMS the maximum time to wait, in milliseconds, or -1 to wait
-   *                  indefinitely.
+   * @param timeout the maximum time to wait, in milliseconds, or -1 to wait
+   *                indefinitely.
    * @returns true if the condition variable is signaled, false if the condition
    *          is not signaled in the allotted time.
    *
@@ -1828,8 +1828,8 @@
  *
  * @param cond the condition variable to wait on.
  * @param mutex the mutex used to coordinate thread access.
- * @param timeoutMS the maximum time to wait, in milliseconds, or -1 to wait
- *                  indefinitely.
+ * @param timeout the maximum time to wait, in milliseconds, or -1 to wait
+ *                indefinitely.
  * @returns true if the condition variable is signaled, false if the condition
  *          is not signaled in the allotted time.
  *
@@ -1845,7 +1845,7 @@
                                  MutexParam mutex,
                                  std::chrono::milliseconds timeout)
 {
-  return SDL_WaitConditionTimeout(cond, mutex, timeout);
+  return SDL_WaitConditionTimeout(cond, mutex, timeout.count());
 }
 
 inline bool Condition::WaitTimeout(MutexParam mutex,
@@ -1878,8 +1878,8 @@
  *
  * Here is an example of using this:
  *
- * ```c
- *    static SDL_AtomicInitState init;
+ * ```cpp
+ *    static SDL::InitState init;
  *
  *    bool InitSystem(void)
  *    {
