--- src/generated/SDL3pp_properties.h
+++ include/SDL3pp/SDL3pp_properties.h
@@ -10,14 +10,15 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryProperties Category Properties
+ * @defgroup CategoryProperties Object Properties
  *
  * A property is a variable that can be created and retrieved by name at
  * runtime.
  *
  * All properties are part of a property group (Properties). A property group
  * can be created with the Properties.Create function and destroyed with the
- * Properties.Destroy function.
+ * Properties.Destroy function, but the Properties destructor probably will do
+ * what you want to, automatically.
  *
  * Properties can be added to and retrieved from a property group through the
  * following functions:
@@ -110,8 +111,6 @@
  * @since This datatype is available since SDL 3.2.0.
  *
  * @sa Properties.Enumerate
- *
- * @sa EnumeratePropertiesCallback
  */
 using EnumeratePropertiesCB =
   std::function<void(PropertiesID props, const char* name)>;
@@ -173,6 +172,8 @@
  * @since This datatype is available since SDL 3.2.0.
  *
  * @cat resource
+ *
+ * @sa Properties.Create
  */
 class Properties
 {
@@ -381,7 +382,6 @@
    *              property.
    * @param cleanup the function to call when this property is deleted, or
    *                nullptr if no cleanup is necessary.
-   * @param userdata a pointer that is passed to the cleanup function.
    * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread.
@@ -664,7 +664,6 @@
    * properties. The properties are locked during enumeration.
    *
    * @param callback the function to call for each property.
-   * @param userdata a pointer that is passed to `callback`.
    * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread.
@@ -673,6 +672,13 @@
    */
   void Enumerate(EnumeratePropertiesCB callback);
 
+  /**
+   * Returns the number of properties this has
+   *
+   * This uses EnumerateProperties() internally, so might not be so fast
+   *
+   * @return Uint64
+   */
   Uint64 GetCount();
 };
 
@@ -783,7 +789,7 @@
    * or want to guarantee that properties being queried aren't freed in another
    * thread.
    *
-   * @param props the properties to lock.
+   * @param resource the properties to lock.
    * @post true on success or false on failure; call GetError() for more
    *       information.
    *
@@ -799,7 +805,7 @@
   PropertiesLock(const PropertiesLock& other) = delete;
 
   /// Move constructor
-  constexpr PropertiesLock(PropertiesLock&& other) noexcept
+  PropertiesLock(PropertiesLock&& other) noexcept
     : m_lock(other.m_lock)
   {
     other.m_lock = {};
@@ -808,8 +814,6 @@
   /**
    * Unlock a group of properties.
    *
-   * @param props the properties to unlock.
-   *
    * @threadsafety It is safe to call this function from any thread.
    *
    * @since This function is available since SDL 3.2.0.
@@ -833,8 +837,6 @@
   /**
    * Unlock a group of properties.
    *
-   * @param props the properties to unlock.
-   *
    * @threadsafety It is safe to call this function from any thread.
    *
    * @since This function is available since SDL 3.2.0.
@@ -891,7 +893,7 @@
  */
 inline PropertiesRef GetGlobalProperties()
 {
-  return CheckError(SDL_GetGlobalProperties());
+  return {CheckError(SDL_GetGlobalProperties())};
 }
 
 /**
@@ -1064,7 +1066,6 @@
  *              property.
  * @param cleanup the function to call when this property is deleted, or nullptr
  *                if no cleanup is necessary.
- * @param userdata a pointer that is passed to the cleanup function.
  * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
@@ -1080,7 +1081,12 @@
                                           void* value,
                                           CleanupPropertyCB cleanup)
 {
-  static_assert(false, "Not implemented");
+  using Wrapper = CallbackWrapper<CleanupPropertyCB>;
+  SDL_SetPointerPropertyWithCleanup(props,
+                                    std::move(name),
+                                    value,
+                                    &Wrapper::CallOnce,
+                                    Wrapper::Wrap(std::move(cleanup)));
 }
 
 inline void Properties::SetPointerPropertyWithCleanup(
@@ -1514,7 +1520,6 @@
  *
  * @param props the properties to query.
  * @param callback the function to call for each property.
- * @param userdata a pointer that is passed to `callback`.
  * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
@@ -1524,7 +1529,13 @@
 inline void EnumerateProperties(PropertiesRef props,
                                 EnumeratePropertiesCB callback)
 {
-  static_assert(false, "Not implemented");
+  return EnumerateProperties(
+    props,
+    [](void* userdata, PropertiesID props, const char* name) {
+      auto& f = *static_cast<EnumeratePropertiesCB*>(userdata);
+      f(props, name);
+    },
+    &callback);
 }
 
 inline void Properties::Enumerate(EnumeratePropertiesCallback callback,
@@ -1538,9 +1549,19 @@
   SDL::EnumerateProperties(m_resource, callback);
 }
 
+/**
+ * Returns the number of properties this has
+ *
+ * This uses EnumerateProperties() internally, so might not be so fast
+ *
+ * @param props
+ * @return Uint64
+ */
 inline Uint64 CountProperties(PropertiesRef props)
 {
-  static_assert(false, "Not implemented");
+  Uint64 count = 0;
+  EnumerateProperties(props, [&](auto, const char*) { count++; });
+  return count;
 }
 
 inline Uint64 Properties::GetCount()
