--- src/generated/SDL3pp_render.h
+++ include/SDL3pp/SDL3pp_render.h
@@ -10,7 +10,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryRender Category Render
+ * @defgroup CategoryRender 2D Accelerated Rendering
  *
  * Header file for SDL 2D rendering functions.
  *
@@ -147,6 +147,8 @@
   constexpr auto operator->() { return value; }
 };
 
+#if SDL_VERSION_ATLEAST(3, 4, 0)
+
 // Forward decl
 struct GPURenderState;
 
@@ -183,6 +185,8 @@
   constexpr operator GPURenderStateRaw() const { return value; }
 };
 
+#endif // SDL_VERSION_ATLEAST(3, 4, 0)
+
 /**
  * The name of the software renderer.
  *
@@ -239,8 +243,6 @@
  */
 using TextureAddressMode = SDL_TextureAddressMode;
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 6)
-
 constexpr TextureAddressMode TEXTURE_ADDRESS_INVALID =
   SDL_TEXTURE_ADDRESS_INVALID; ///< TEXTURE_ADDRESS_INVALID
 
@@ -256,6 +258,8 @@
 constexpr TextureAddressMode TEXTURE_ADDRESS_WRAP =
   SDL_TEXTURE_ADDRESS_WRAP; ///< The texture is repeated (tiled)
 
+#endif // SDL_VERSION_ATLEAST(3, 4, 0)
+
 /**
  * How the logical size is mapped to the output.
  *
@@ -347,8 +351,41 @@
 
   constexpr Renderer(RendererRef&& other) = delete;
 
-  /// Default ctor
-  Renderer(WindowParam window) {}
+  /**
+   * Create a 2D rendering context for a window.
+   *
+   * If you want a specific renderer, you can specify its name here. A list of
+   * available renderers can be obtained by calling GetRenderDriver()
+   * multiple times, with indices from 0 to GetNumRenderDrivers()-1. If you
+   * don't need a specific renderer, specify nullptr and SDL will attempt to
+   * choose the best option for you, based on what is available on the user's
+   * system.
+   *
+   * If `name` is a comma-separated list, SDL will try each name, in the order
+   * listed, until one succeeds or all of them fail.
+   *
+   * By default the rendering size matches the window size in pixels, but you
+   * can call Renderer.SetLogicalPresentation() to change the content size and
+   * scaling options.
+   *
+   * @param window the window where rendering is displayed.
+   * @throws Error on failure.
+   *
+   * @threadsafety This function should only be called on the main thread.
+   *
+   * @since This function is available since SDL 3.2.0.
+   *
+   * @sa Renderer.Renderer
+   * @sa Renderer.Renderer
+   * @sa Renderer.Destroy
+   * @sa GetNumRenderDrivers
+   * @sa GetRenderDriver
+   * @sa Renderer.GetName
+   */
+  Renderer(WindowParam window)
+    : m_resource(CheckError(SDL_CreateRenderer(window, nullptr)))
+  {
+  }
 
   /**
    * Create a 2D rendering context for a window.
@@ -369,8 +406,7 @@
    * @param window the window where rendering is displayed.
    * @param name the name of the rendering driver to initialize, or nullptr to
    *             let SDL choose one.
-   * @post a valid rendering context or nullptr if there was an error; call
-   *       GetError() for more information.
+   * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
    *
@@ -384,7 +420,7 @@
    * @sa Renderer.GetName
    */
   Renderer(WindowParam window, StringParam name)
-    : m_resource(SDL_CreateRenderer(window, name))
+    : m_resource(CheckError(SDL_CreateRenderer(window, name)))
   {
   }
 
@@ -437,8 +473,7 @@
    *   queue family index used for presentation.
    *
    * @param props the properties to use.
-   * @post a valid rendering context or nullptr if there was an error; call
-   *       GetError() for more information.
+   * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
    *
@@ -451,7 +486,7 @@
    * @sa Renderer.GetName
    */
   Renderer(PropertiesParam props)
-    : m_resource(SDL_CreateRendererWithProperties(props))
+    : m_resource(CheckError(SDL_CreateRendererWithProperties(props)))
   {
   }
 
@@ -465,8 +500,7 @@
    *
    * @param surface the Surface structure representing the surface where
    *                rendering is done.
-   * @post a valid rendering context or nullptr if there was an error; call
-   *       GetError() for more information.
+   * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
@@ -475,7 +509,7 @@
    * @sa Renderer.Destroy
    */
   Renderer(SurfaceParam surface)
-    : m_resource(SDL_CreateSoftwareRenderer(surface))
+    : m_resource(CheckError(SDL_CreateSoftwareRenderer(surface)))
   {
   }
 
@@ -567,7 +601,6 @@
    * @since This function is available since SDL 3.2.0.
    *
    * @sa Renderer.Renderer
-   * @sa Renderer.Renderer
    */
   const char* GetName() const;
 
@@ -687,8 +720,7 @@
    * For the output size of the current rendering target, with logical size
    * adjustments, use Renderer.GetCurrentOutputSize() instead.
    *
-   * @param w a pointer filled in with the width in pixels.
-   * @param h a pointer filled in with the height in pixels.
+   * @returns Point on success.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -729,8 +761,7 @@
    * Rendering target or not, the output will be adjusted by the current logical
    * presentation state, dictated by Renderer.SetLogicalPresentation().
    *
-   * @param w a pointer filled in with the current width.
-   * @param h a pointer filled in with the current height.
+   * @returns the size on success.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -748,8 +779,7 @@
    *
    * @param format one of the enumerated values in PixelFormat.
    * @param access one of the enumerated values in TextureAccess.
-   * @param w the width of the texture in pixels.
-   * @param h the height of the texture in pixels.
+   * @param size the width and height of the texture in pixels.
    * @returns the created texture or nullptr on failure; call GetError() for
    *          more information.
    *
@@ -943,6 +973,20 @@
    */
   void SetTarget(TextureParam texture);
 
+  /**
+   * Set target texture back to window
+   *
+   * This is equivalent to SetTarget(nullptr)
+   *
+   * @throws Error on failure.
+   *
+   * @threadsafety This function should only be called on the main thread.
+   *
+   * @since This function is available since SDL 3.2.0.
+   *
+   * @sa SetTarget(nullptr)
+   * @sa GetTarget()
+   */
   void ResetTarget();
 
   /**
@@ -992,8 +1036,7 @@
    * You can convert coordinates in an event into rendering coordinates using
    * Renderer.ConvertEventToRenderCoordinates().
    *
-   * @param w the width of the logical resolution.
-   * @param h the height of the logical resolution.
+   * @param size the width and height of the logical resolution.
    * @param mode the presentation mode used.
    * @throws Error on failure.
    *
@@ -1042,8 +1085,7 @@
    * Each render target has its own logical presentation state. This function
    * gets the state for the current render target.
    *
-   * @param w an int filled with the logical presentation width.
-   * @param h an int filled with the logical presentation height.
+   * @param size a Point to be filled with the width and height.
    * @param mode a variable filled with the logical presentation mode being
    *             used.
    * @throws Error on failure.
@@ -1068,8 +1110,7 @@
    * Each render target has its own logical presentation state. This function
    * gets the rectangle for the current render target.
    *
-   * @param rect a pointer filled in with the final presentation rectangle, may
-   *             be nullptr.
+   * @returns the final presentation rectangle.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1090,10 +1131,8 @@
    * - The scale (Renderer.SetScale)
    * - The viewport (Renderer.SetViewport)
    *
-   * @param window_x the x coordinate in window coordinates.
-   * @param window_y the y coordinate in window coordinates.
-   * @param x a pointer filled with the x coordinate in render coordinates.
-   * @param y a pointer filled with the y coordinate in render coordinates.
+   * @param window_coord the x, y coordinate in window coordinates.
+   * @returns a FPoint containing ther render coordinates on success.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1115,12 +1154,8 @@
    * - The scale (Renderer.SetScale)
    * - The viewport (Renderer.SetViewport)
    *
-   * @param x the x coordinate in render coordinates.
-   * @param y the y coordinate in render coordinates.
-   * @param window_x a pointer filled with the x coordinate in window
-   *                 coordinates.
-   * @param window_y a pointer filled with the y coordinate in window
-   *                 coordinates.
+   * @param coord the x, y coordinate in render coordinates.
+   * @returns a FPoint filled with window coordinates on success.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1157,8 +1192,7 @@
    * Once converted, coordinates may be outside the rendering area.
    *
    * @param event the event to modify.
-   * @returns true if the event is converted or doesn't need conversion, or
-   *          false on failure; call GetError() for more information.
+   * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
    *
@@ -1193,6 +1227,21 @@
    */
   void SetViewport(OptionalRef<const RectRaw> rect);
 
+  /**
+   * Reset the drawing area for rendering to the entire target
+   *
+   * This is equivalent to `SetViewport(std::nullopt)`
+   *
+   * @throws Error on failure.
+   *
+   * @threadsafety This function should only be called on the main thread.
+   *
+   * @since This function is available since SDL 3.2.0.
+   *
+   * @sa GetViewport()
+   * @sa SetViewport()
+   * @sa ViewportSet()
+   */
   void ResetViewport();
 
   /**
@@ -1201,7 +1250,7 @@
    * Each render target has its own viewport. This function gets the viewport
    * for the current render target.
    *
-   * @param rect an Rect structure filled in with the current drawing area.
+   * @returns an Rect with the current drawing area.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1244,8 +1293,7 @@
    * rendering into the rest of the render target, but it should not contain
    * visually important or interactible content.
    *
-   * @param rect a pointer filled in with the area that is safe for interactive
-   *             content.
+   * @returns the rect filled the area that is safe for interactive content.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1269,10 +1317,26 @@
    * @since This function is available since SDL 3.2.0.
    *
    * @sa Renderer.GetClipRect
+   * @sa Renderer.ResetClipRect()
    * @sa Renderer.IsClipEnabled
    */
   void SetClipRect(OptionalRef<const RectRaw> rect);
 
+  /**
+   * Reset the clip rectangle for rendering to the entire render target
+   *
+   * This is equivalent to `SetClipRect(std::nullopt)`
+   *
+   * @throws Error on failure.
+   *
+   * @threadsafety This function should only be called on the main thread.
+   *
+   * @since This function is available since SDL 3.2.0.
+   *
+   * @sa GetClipRect()
+   * @sa SetClipRect()
+   * @sa ClipEnabled()
+   */
   void ResetClipRect();
 
   /**
@@ -1281,8 +1345,8 @@
    * Each render target has its own clip rectangle. This function gets the
    * cliprect for the current render target.
    *
-   * @param rect an Rect structure filled in with the current clipping area or
-   *             an empty rectangle if clipping is disabled.
+   * @returns a Rect structure filled in with the current clipping area or an
+   *          empty rectangle if clipping is disabled.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1326,8 +1390,7 @@
    * Each render target has its own scale. This function sets the scale for the
    * current render target.
    *
-   * @param scaleX the horizontal scaling factor.
-   * @param scaleY the vertical scaling factor.
+   * @param scale the x, y scaling factors.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1362,8 +1425,7 @@
    * Each render target has its own scale. This function gets the scale for the
    * current render target.
    *
-   * @param scaleX a pointer filled in with the horizontal scaling factor.
-   * @param scaleY a pointer filled in with the vertical scaling factor.
+   * @returns the scaling factors on success.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1380,12 +1442,7 @@
    * Set the color for drawing or filling rectangles, lines, and points, and for
    * Renderer.RenderClear().
    *
-   * @param r the red value used to draw on the rendering target.
-   * @param g the green value used to draw on the rendering target.
-   * @param b the blue value used to draw on the rendering target.
-   * @param a the alpha value used to draw on the rendering target; usually
-   *          `ALPHA_OPAQUE` (255). Use Renderer.SetDrawBlendMode to specify how
-   *          the alpha channel is used.
+   * @param c the color value used to draw on the rendering target.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1403,12 +1460,7 @@
    * Set the color for drawing or filling rectangles, lines, and points, and for
    * Renderer.RenderClear().
    *
-   * @param r the red value used to draw on the rendering target.
-   * @param g the green value used to draw on the rendering target.
-   * @param b the blue value used to draw on the rendering target.
-   * @param a the alpha value used to draw on the rendering target. Use
-   *          Renderer.SetDrawBlendMode to specify how the alpha channel is
-   *          used.
+   * @param c the RGBA values used to draw on the rendering target.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1445,22 +1497,15 @@
   /**
    * Get the color used for drawing operations (Rect, Line and Clear).
    *
-   * @param r a pointer filled in with the red value used to draw on the
-   *          rendering target.
-   * @param g a pointer filled in with the green value used to draw on the
-   *          rendering target.
-   * @param b a pointer filled in with the blue value used to draw on the
-   *          rendering target.
-   * @param a a pointer filled in with the alpha value used to draw on the
-   *          rendering target; usually `ALPHA_OPAQUE` (255).
+   * @returns the color channel values used to draw on the rendering target.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
    *
    * @since This function is available since SDL 3.2.0.
    *
-   * @sa Renderer.GetDrawColorFloat
-   * @sa Renderer.SetDrawColor
+   * @sa GetDrawColor(SDL_FColor*)
+   * @sa SetDrawColor()
    */
   Color GetDrawColor() const;
 
@@ -1489,22 +1534,15 @@
   /**
    * Get the color used for drawing operations (Rect, Line and Clear).
    *
-   * @param r a pointer filled in with the red value used to draw on the
-   *          rendering target.
-   * @param g a pointer filled in with the green value used to draw on the
-   *          rendering target.
-   * @param b a pointer filled in with the blue value used to draw on the
-   *          rendering target.
-   * @param a a pointer filled in with the alpha value used to draw on the
-   *          rendering target.
+   * @returns the color on success.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
    *
    * @since This function is available since SDL 3.2.0.
    *
-   * @sa Renderer.SetDrawColorFloat
-   * @sa Renderer.GetDrawColor
+   * @sa GetDrawColor(SDL_FColor*)
+   * @sa SetDrawColor()
    */
   FColor GetDrawColorFloat() const;
 
@@ -1533,7 +1571,7 @@
   /**
    * Get the color scale used for render operations.
    *
-   * @param scale a pointer filled in with the current color scale value.
+   * @returns a float representing the current color scale on success.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1563,7 +1601,7 @@
   /**
    * Get the blend mode used for drawing operations.
    *
-   * @param blendMode a pointer filled in with the current BlendMode.
+   * @returns the current BlendMode on success.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1595,8 +1633,7 @@
   /**
    * Draw a point on the current rendering target at subpixel precision.
    *
-   * @param x the x coordinate of the point.
-   * @param y the y coordinate of the point.
+   * @param p the x, y coordinates of the point.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1611,7 +1648,6 @@
    * Draw multiple points on the current rendering target at subpixel precision.
    *
    * @param points the points to draw.
-   * @param count the number of points to draw.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1625,10 +1661,8 @@
   /**
    * Draw a line on the current rendering target at subpixel precision.
    *
-   * @param x1 the x coordinate of the start point.
-   * @param y1 the y coordinate of the start point.
-   * @param x2 the x coordinate of the end point.
-   * @param y2 the y coordinate of the end point.
+   * @param p1 the x,y coordinate of the start point.
+   * @param p2 the x,y coordinate of the end point.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1644,7 +1678,6 @@
    * subpixel precision.
    *
    * @param points the points along the lines.
-   * @param count the number of points, drawing count-1 lines.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1658,8 +1691,8 @@
   /**
    * Draw a rectangle on the current rendering target at subpixel precision.
    *
-   * @param rect a pointer to the destination rectangle, or nullptr to outline
-   *             the entire rendering target.
+   * @param rect a pointer to the destination rectangle, or std::nullopt to
+   *             outline the entire rendering target.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1675,7 +1708,6 @@
    * precision.
    *
    * @param rects a pointer to an array of destination rectangles.
-   * @param count the number of rectangles.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1690,7 +1722,7 @@
    * Fill a rectangle on the current rendering target with the drawing color at
    * subpixel precision.
    *
-   * @param rect a pointer to the destination rectangle, or nullptr for the
+   * @param rect a pointer to the destination rectangle, or std::nullopt for the
    *             entire rendering target.
    * @throws Error on failure.
    *
@@ -1707,7 +1739,6 @@
    * drawing color at subpixel precision.
    *
    * @param rects a pointer to an array of destination rectangles.
-   * @param count the number of rectangles.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1769,7 +1800,7 @@
                             OptionalRef<const FRectRaw> dstrect,
                             double angle,
                             OptionalRef<const FPointRaw> center,
-                            FlipMode flip);
+                            FlipMode flip = FlipMode::SDL_FLIP_NONE);
 
   /**
    * Copy a portion of the source texture to the current rendering target, with
@@ -1924,11 +1955,9 @@
    *
    * @param texture (optional) The SDL texture to use.
    * @param vertices vertices.
-   * @param num_vertices number of vertices.
    * @param indices (optional) An array of integer indices into the 'vertices'
    *                array, if nullptr all vertices will be rendered in
    *                sequential order.
-   * @param num_indices number of indices.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1940,7 +1969,7 @@
    */
   void RenderGeometry(TextureParam texture,
                       std::span<const Vertex> vertices,
-                      std::span<const int> indices);
+                      std::span<const int> indices = {});
 
   /**
    * Render a list of triangles, optionally using a texture and indices into the
@@ -2001,10 +2030,6 @@
   void SetRenderTextureAddressMode(TextureAddressMode u_mode,
                                    TextureAddressMode v_mode);
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 6)
-
-#if SDL_VERSION_ATLEAST(3, 3, 6)
-
   /**
    * Get the texture addressing mode used in Renderer.RenderGeometry().
    *
@@ -2224,8 +2249,7 @@
   /**
    * Get VSync of the given renderer.
    *
-   * @param vsync an int filled with the current vertical refresh sync interval.
-   *              See Renderer.SetVSync() for the meaning of the value.
+   * @returns the current vertical refresh sync interval on success.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -2261,8 +2285,8 @@
    *
    * The text is drawn in the color specified by Renderer.SetDrawColor().
    *
-   * @param x the x coordinate where the top-left corner of the text will draw.
-   * @param y the y coordinate where the top-left corner of the text will draw.
+   * @param p the x, y coordinates where the top-left corner of the text will
+   *          draw.
    * @param str the string to render.
    * @throws Error on failure.
    *
@@ -2278,18 +2302,19 @@
   /**
    * Draw debug text to an Renderer.
    *
-   * This function will render a printf()-style format string to a renderer.
-   * Note that this is a convenience function for debugging, with severe
-   * limitations, and is not intended to be used for production apps and games.
+   * This function will render a std::format()-style format string to a
+   * renderer. Note that this is a convenience function for debugging, with
+   * severe limitations, and is not intended to be used for production apps and
+   * games.
    *
    * For the full list of limitations and other useful information, see
    * Renderer.RenderDebugText.
    *
-   * @param x the x coordinate where the top-left corner of the text will draw.
-   * @param y the y coordinate where the top-left corner of the text will draw.
+   * @param p the x, y coordinates where the top-left corner of the text will
+   *          draw.
    * @param fmt the format string to draw.
-   * @param ... additional parameters matching % tokens in the `fmt` string, if
-   *            any.
+   * @param args additional parameters matching {} tokens in the `fmt` string,
+   *             if any.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -2322,10 +2347,6 @@
    */
   void SetDefaultTextureScaleMode(ScaleMode scale_mode);
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 6)
-
-#if SDL_VERSION_ATLEAST(3, 3, 6)
-
   /**
    * Get default texture scale mode of the given renderer.
    *
@@ -2342,10 +2363,6 @@
    */
   void GetDefaultTextureScaleMode(ScaleMode* scale_mode);
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 6)
-
-#if SDL_VERSION_ATLEAST(3, 3, 6)
-
   /**
    * Create custom GPU render state.
    *
@@ -2364,10 +2381,6 @@
    */
   GPURenderStateRef CreateGPURenderState(GPURenderStateCreateInfo* createinfo);
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 6)
-
-#if SDL_VERSION_ATLEAST(3, 3, 6)
-
   /**
    * Set custom GPU render state.
    *
@@ -2479,10 +2492,8 @@
    * @param renderer the rendering context.
    * @param format one of the enumerated values in PixelFormat.
    * @param access one of the enumerated values in TextureAccess.
-   * @param w the width of the texture in pixels.
-   * @param h the height of the texture in pixels.
-   * @post the created texture or nullptr on failure; call GetError() for more
-   *       information.
+   * @param size the width and height of the texture in pixels.
+   * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
    *
@@ -2498,7 +2509,8 @@
           PixelFormat format,
           TextureAccess access,
           const PointRaw& size)
-    : m_resource(SDL_CreateTexture(renderer, format, access, size))
+    : m_resource(
+        CheckError(SDL_CreateTexture(renderer, format, access, size.x, size.y)))
   {
   }
 
@@ -2516,19 +2528,17 @@
    * @param renderer the rendering context.
    * @param surface the Surface structure containing pixel data used to fill the
    *                texture.
-   * @post the created texture or nullptr on failure; call GetError() for more
-   *       information.
+   * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
    *
    * @since This function is available since SDL 3.2.0.
    *
    * @sa Texture.Texture
-   * @sa Texture.Texture
    * @sa Texture.Destroy
    */
   Texture(RendererParam renderer, SurfaceParam surface)
-    : m_resource(SDL_CreateTextureFromSurface(renderer, surface))
+    : m_resource(CheckError(SDL_CreateTextureFromSurface(renderer, surface)))
   {
   }
 
@@ -2642,8 +2652,7 @@
    *
    * @param renderer the rendering context.
    * @param props the properties to use.
-   * @post the created texture or nullptr on failure; call GetError() for more
-   *       information.
+   * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
    *
@@ -2657,7 +2666,7 @@
    * @sa Texture.Update
    */
   Texture(RendererParam renderer, PropertiesParam props)
-    : m_resource(SDL_CreateTextureWithProperties(renderer, props))
+    : m_resource(CheckError(SDL_CreateTextureWithProperties(renderer, props)))
   {
   }
 
@@ -2681,9 +2690,6 @@
    * If you would rather decode an image to an Surface (a buffer of pixels in
    * CPU memory), call Surface.Surface() instead.
    *
-   * When done with the returned texture, the app should dispose of it with a
-   * call to Texture.Destroy().
-   *
    * @param renderer the Renderer to use to create the GPU texture.
    * @param file a path on the filesystem to load an image from.
    * @post a new texture, or nullptr on error.
@@ -2724,9 +2730,6 @@
    * If you would rather decode an image to an Surface (a buffer of pixels in
    * CPU memory), call Surface.Surface() instead.
    *
-   * When done with the returned texture, the app should dispose of it with a
-   * call to Texture.Destroy().
-   *
    * @param renderer the Renderer to use to create the GPU texture.
    * @param src an IOStream that data will be read from.
    * @param closeio true to close/free the IOStream before returning, false to
@@ -2739,7 +2742,7 @@
    * @sa LoadTextureTyped
    * @sa Texture.Destroy
    */
-  Texture(RendererParam renderer, IOStreamParam src, bool closeio);
+  Texture(RendererParam renderer, IOStreamParam src, bool closeio = false);
 
   /**
    * Safely borrows the from TextureParam.
@@ -2936,27 +2939,19 @@
    */
   void GetSize(float* w, float* h) const;
 
-  /**
-   * Get the size of a texture, as floating point values.
-   *
-   * @param w a pointer filled in with the width of the texture in pixels. This
-   *          argument can be nullptr if you don't need this information.
-   * @param h a pointer filled in with the height of the texture in pixels. This
-   *          argument can be nullptr if you don't need this information.
-   * @throws Error on failure.
-   *
-   * @threadsafety This function should only be called on the main thread.
-   *
-   * @since This function is available since SDL 3.2.0.
-   */
+  /// Get the size of a texture.
   Point GetSize() const;
 
+  /// Get the size of a texture, as floating point values.
   FPoint GetSizeFloat() const;
 
+  /// Get the width in pixels.
   int GetWidth() const;
 
+  /// Get the height in pixels.
   int GetHeight() const;
 
+  /// Get the pixel format.
   PixelFormat GetFormat() const;
 
 #if SDL_VERSION_ATLEAST(3, 3, 6)
@@ -2981,10 +2976,6 @@
    */
   void SetPalette(PaletteParam palette);
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 6)
-
-#if SDL_VERSION_ATLEAST(3, 3, 6)
-
   /**
    * Get the palette used by a texture.
    *
@@ -3142,7 +3133,7 @@
   /**
    * Get the additional alpha value multiplied into render copy operations.
    *
-   * @param alpha a pointer filled in with the current alpha value.
+   * @returns the current alpha value.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -3158,7 +3149,7 @@
   /**
    * Get the additional alpha value multiplied into render copy operations.
    *
-   * @param alpha a pointer filled in with the current alpha value.
+   * @returns the current alpha value.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -3171,12 +3162,82 @@
    */
   float GetAlphaModFloat() const;
 
+  /**
+   * Set an additional color and alpha values multiplied into render copy
+   * operations.
+   *
+   * When this texture is rendered, during the copy operation each source color
+   * and alpha channels are modulated by the appropriate color value according
+   * to the following formula:
+   *
+   *      srcC = srcC * (color / 255)
+   *      srcA = srcA * (alpha / 255)
+   *
+   * Color and alpha modulation is not always supported by the renderer; it will
+   * return false if either modulation is not supported.
+   *
+   * @param c the color and alpha channel values multiplied into copy
+   *          operations.
+   * @throws Error on failure.
+   *
+   * @threadsafety This function should only be called on the main thread.
+   *
+   * @since This function is available since SDL 3.2.0.
+   */
   void SetMod(Color c);
 
+  /**
+   * Set an additional color and alpha values multiplied into render copy
+   * operations.
+   *
+   * When this texture is rendered, during the copy operation each source color
+   * and alpha channels are modulated by the appropriate color value according
+   * to the following formula:
+   *
+   *      srcC = srcC * (color / 255)
+   *      srcA = srcA * (alpha / 255)
+   *
+   * Color and alpha modulation is not always supported by the renderer; it will
+   * return false if either modulation is not supported.
+   *
+   * @param c the color and alpha channel values multiplied into copy
+   *          operations.
+   * @throws Error on failure.
+   *
+   * @threadsafety This function should only be called on the main thread.
+   *
+   * @since This function is available since SDL 3.2.0.
+   */
   void SetModFloat(FColor c);
 
+  /**
+   * Get the additional color value multiplied into render copy operations.
+   *
+   * @returns the color channels (0-1) on success.
+   * @throws Error on failure.
+   *
+   * @threadsafety This function should only be called on the main thread.
+   *
+   * @since This function is available since SDL 3.2.0.
+   *
+   * @sa GetAlphaMod()
+   * @sa SetColorMod()
+   */
   Color GetMod() const;
 
+  /**
+   * Get the additional color value multiplied into render copy operations.
+   *
+   * @returns the color channels (0-1) on success.
+   * @throws Error on failure.
+   *
+   * @threadsafety This function should only be called on the main thread.
+   *
+   * @since This function is available since SDL 3.2.0.
+   *
+   * @sa GetAlphaMod()
+   * @sa SetColorMod()
+   */
   FColor GetModFloat() const;
 
   /**
@@ -3199,7 +3260,7 @@
   /**
    * Get the blend mode used for texture copy operations.
    *
-   * @param blendMode a pointer filled in with the current BlendMode.
+   * @returns the current SDL_BlendMode on success.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -3231,7 +3292,7 @@
   /**
    * Get the scale mode used for texture scale operations.
    *
-   * @param scaleMode a pointer filled in with the current scale mode.
+   * @returns the current scale mode on success.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -3288,11 +3349,10 @@
    * While this function will work with streaming textures, for optimization
    * reasons you may not get the pixels back if you lock the texture afterward.
    *
-   * @param rect an Rect structure representing the area to update, or nullptr
-   *             to update the entire texture.
-   * @param pixels the raw pixel data in the format of the texture.
-   * @param pitch the number of bytes in a row of pixel data, including padding
-   *              between lines.
+   * @param surface the surface containing pixel data in the format of the
+   *                texture.
+   * @param rect an Rect structure representing the area to update, or
+   *             std::nullopt to update the entire texture.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -3419,8 +3479,8 @@
    *
    * @param rect a pointer to the rectangle to lock for access. If the rect is
    *             nullptr, the entire texture will be locked.
-   * @param surface a pointer to an SDL surface of size **rect**. Don't assume
-   *                any specific pixel content.
+   * @returns a surface of size **rect**. Don't assume any specific pixel
+   *          content.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -3503,12 +3563,10 @@
  * Create a window and default renderer.
  *
  * @param title the title of the window, in UTF-8 encoding.
- * @param width the width of the window.
- * @param height the height of the window.
+ * @param size the width and height of the window.
  * @param window_flags the flags used to create the window (see
  *                     Window.Window()).
- * @param window a pointer filled with the window, or nullptr on error.
- * @param renderer a pointer filled with the renderer, or nullptr on error.
+ * @returns a pair with window and renderer.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -3523,7 +3581,11 @@
   const PointRaw& size,
   WindowFlags window_flags = 0)
 {
-  return CheckError(SDL_CreateWindowAndRenderer(title, size, window_flags));
+  SDL_Window* window = nullptr;
+  SDL_Renderer* renderer = nullptr;
+  CheckError(SDL_CreateWindowAndRenderer(
+    title, size.x, size.y, window_flags, &window, &renderer));
+  return {Window{window}, Renderer(renderer)};
 }
 
 /**
@@ -3651,24 +3713,12 @@
 constexpr auto CREATE_GPU_DEVICE_POINTER =
   SDL_PROP_RENDERER_CREATE_GPU_DEVICE_POINTER;
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 2)
-
-#if SDL_VERSION_ATLEAST(3, 3, 2)
-
 constexpr auto CREATE_GPU_SHADERS_SPIRV_BOOLEAN =
   SDL_PROP_RENDERER_CREATE_GPU_SHADERS_SPIRV_BOOLEAN;
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 2)
-
-#if SDL_VERSION_ATLEAST(3, 3, 2)
-
 constexpr auto CREATE_GPU_SHADERS_DXIL_BOOLEAN =
   SDL_PROP_RENDERER_CREATE_GPU_SHADERS_DXIL_BOOLEAN;
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 2)
-
-#if SDL_VERSION_ATLEAST(3, 3, 2)
-
 constexpr auto CREATE_GPU_SHADERS_MSL_BOOLEAN =
   SDL_PROP_RENDERER_CREATE_GPU_SHADERS_MSL_BOOLEAN;
 
@@ -3799,10 +3849,6 @@
   return SDL_CreateGPURenderer(device, window);
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 6)
-
-#if SDL_VERSION_ATLEAST(3, 3, 6)
-
 /**
  * Return the GPU device used by a renderer.
  *
@@ -3819,10 +3865,6 @@
   return SDL_GetGPURendererDevice(renderer);
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 6)
-
-#if SDL_VERSION_ATLEAST(3, 3, 6)
-
 inline GPUDeviceRef Renderer::GetGPUDevice()
 {
   return SDL::GetGPURendererDevice(m_resource);
@@ -3856,7 +3898,7 @@
 
 inline RendererRef Window::GetRenderer() const
 {
-  return CheckError(SDL_GetRenderer(m_resource));
+  return {CheckError(SDL_GetRenderer(m_resource))};
 }
 
 /**
@@ -3872,7 +3914,7 @@
  */
 inline WindowRef GetRenderWindow(RendererParam renderer)
 {
-  return CheckError(SDL_GetRenderWindow(renderer));
+  return {CheckError(SDL_GetRenderWindow(renderer))};
 }
 
 inline WindowRef Renderer::GetWindow()
@@ -3991,7 +4033,7 @@
  */
 inline PropertiesRef GetRendererProperties(RendererParam renderer)
 {
-  return CheckError(SDL_GetRendererProperties(renderer));
+  return {CheckError(SDL_GetRendererProperties(renderer))};
 }
 
 inline PropertiesRef Renderer::GetProperties() const
@@ -4034,8 +4076,7 @@
  * adjustments, use Renderer.GetCurrentOutputSize() instead.
  *
  * @param renderer the rendering context.
- * @param w a pointer filled in with the width in pixels.
- * @param h a pointer filled in with the height in pixels.
+ * @returns the size on success.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -4046,7 +4087,9 @@
  */
 inline Point GetRenderOutputSize(RendererParam renderer)
 {
-  static_assert(false, "Not implemented");
+  Point p;
+  GetRenderOutputSize(renderer, &p.x, &p.y);
+  return p;
 }
 
 inline void Renderer::GetOutputSize(int* w, int* h) const
@@ -4094,8 +4137,7 @@
  * presentation state, dictated by Renderer.SetLogicalPresentation().
  *
  * @param renderer the rendering context.
- * @param w a pointer filled in with the current width.
- * @param h a pointer filled in with the current height.
+ * @returns the size on success.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -4106,7 +4148,9 @@
  */
 inline Point GetCurrentRenderOutputSize(RendererParam renderer)
 {
-  static_assert(false, "Not implemented");
+  Point p;
+  GetCurrentRenderOutputSize(renderer, &p.x, &p.y);
+  return p;
 }
 
 inline void Renderer::GetCurrentOutputSize(int* w, int* h) const
@@ -4127,8 +4171,7 @@
  * @param renderer the rendering context.
  * @param format one of the enumerated values in PixelFormat.
  * @param access one of the enumerated values in TextureAccess.
- * @param w the width of the texture in pixels.
- * @param h the height of the texture in pixels.
+ * @param size the width and height of the texture in pixels.
  * @returns the created texture or nullptr on failure; call GetError() for more
  *          information.
  *
@@ -4404,24 +4447,12 @@
 constexpr auto CREATE_GPU_TEXTURE_POINTER =
   SDL_PROP_TEXTURE_CREATE_GPU_TEXTURE_POINTER;
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 2)
-
-#if SDL_VERSION_ATLEAST(3, 3, 2)
-
 constexpr auto CREATE_GPU_TEXTURE_UV_POINTER =
   SDL_PROP_TEXTURE_CREATE_GPU_TEXTURE_UV_POINTER;
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 2)
-
-#if SDL_VERSION_ATLEAST(3, 3, 2)
-
 constexpr auto CREATE_GPU_TEXTURE_U_POINTER =
   SDL_PROP_TEXTURE_CREATE_GPU_TEXTURE_U_POINTER;
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 2)
-
-#if SDL_VERSION_ATLEAST(3, 3, 2)
-
 constexpr auto CREATE_GPU_TEXTURE_V_POINTER =
   SDL_PROP_TEXTURE_CREATE_GPU_TEXTURE_V_POINTER;
 
@@ -4496,22 +4527,10 @@
 
 constexpr auto GPU_TEXTURE_POINTER = SDL_PROP_TEXTURE_GPU_TEXTURE_POINTER;
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 2)
-
-#if SDL_VERSION_ATLEAST(3, 3, 2)
-
 constexpr auto GPU_TEXTURE_UV_POINTER = SDL_PROP_TEXTURE_GPU_TEXTURE_UV_POINTER;
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 2)
-
-#if SDL_VERSION_ATLEAST(3, 3, 2)
-
 constexpr auto GPU_TEXTURE_U_POINTER = SDL_PROP_TEXTURE_GPU_TEXTURE_U_POINTER;
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 2)
-
-#if SDL_VERSION_ATLEAST(3, 3, 2)
-
 constexpr auto GPU_TEXTURE_V_POINTER = SDL_PROP_TEXTURE_GPU_TEXTURE_V_POINTER;
 
 #endif // SDL_VERSION_ATLEAST(3, 3, 2)
@@ -4615,7 +4634,7 @@
  */
 inline PropertiesRef GetTextureProperties(TextureConstParam texture)
 {
-  return CheckError(SDL_GetTextureProperties(texture));
+  return {CheckError(SDL_GetTextureProperties(texture))};
 }
 
 inline PropertiesRef Texture::GetProperties() const
@@ -4636,7 +4655,7 @@
  */
 inline RendererRef GetRendererFromTexture(TextureConstParam texture)
 {
-  return SDL_GetRendererFromTexture(texture);
+  return {SDL_GetRendererFromTexture(texture)};
 }
 
 inline RendererRef Texture::GetRenderer() const
@@ -4663,23 +4682,10 @@
   CheckError(SDL_GetTextureSize(texture, w, h));
 }
 
-/**
- * Get the size of a texture, as floating point values.
- *
- * @param texture the texture to query.
- * @param w a pointer filled in with the width of the texture in pixels. This
- *          argument can be nullptr if you don't need this information.
- * @param h a pointer filled in with the height of the texture in pixels. This
- *          argument can be nullptr if you don't need this information.
- * @throws Error on failure.
- *
- * @threadsafety This function should only be called on the main thread.
- *
- * @since This function is available since SDL 3.2.0.
- */
+/// Get the size of a texture.
 inline Point GetTextureSize(TextureConstParam texture)
 {
-  static_assert(false, "Not implemented");
+  return Point(texture->w, texture->h);
 }
 
 inline void Texture::GetSize(float* w, float* h) const
@@ -4692,9 +4698,12 @@
   return SDL::GetTextureSize(m_resource);
 }
 
+/// Get the size of a texture, as floating point values.
 inline FPoint GetTextureSizeFloat(TextureConstParam texture)
 {
-  static_assert(false, "Not implemented");
+  FPoint p;
+  GetTextureSize(texture, &p.x, &p.y);
+  return p;
 }
 
 inline FPoint Texture::GetSizeFloat() const
@@ -4702,29 +4711,26 @@
   return SDL::GetTextureSizeFloat(m_resource);
 }
 
-inline int GetTextureWidth(TextureConstParam texture)
-{
-  static_assert(false, "Not implemented");
-}
+/// Get the width in pixels.
+inline int GetTextureWidth(TextureConstParam texture) { return texture->w; }
 
 inline int Texture::GetWidth() const
 {
   return SDL::GetTextureWidth(m_resource);
 }
 
-inline int GetTextureHeight(TextureConstParam texture)
-{
-  static_assert(false, "Not implemented");
-}
+/// Get the height in pixels.
+inline int GetTextureHeight(TextureConstParam texture) { return texture->h; }
 
 inline int Texture::GetHeight() const
 {
   return SDL::GetTextureHeight(m_resource);
 }
 
+/// Get the pixel format.
 inline PixelFormat GetTextureFormat(TextureConstParam texture)
 {
-  static_assert(false, "Not implemented");
+  return texture->format;
 }
 
 inline PixelFormat Texture::GetFormat() const
@@ -4758,19 +4764,11 @@
   CheckError(SDL_SetTexturePalette(texture, palette));
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 6)
-
-#if SDL_VERSION_ATLEAST(3, 3, 6)
-
 inline void Texture::SetPalette(PaletteParam palette)
 {
   SDL::SetTexturePalette(m_resource, palette);
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 6)
-
-#if SDL_VERSION_ATLEAST(3, 3, 6)
-
 /**
  * Get the palette used by a texture.
  *
@@ -4789,10 +4787,6 @@
   return SDL_GetTexturePalette(texture);
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 6)
-
-#if SDL_VERSION_ATLEAST(3, 3, 6)
-
 inline Palette Texture::GetPalette()
 {
   return SDL::GetTexturePalette(m_resource);
@@ -5005,7 +4999,7 @@
  * Get the additional alpha value multiplied into render copy operations.
  *
  * @param texture the texture to query.
- * @param alpha a pointer filled in with the current alpha value.
+ * @returns the current alpha value.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5018,7 +5012,9 @@
  */
 inline Uint8 GetTextureAlphaMod(TextureConstParam texture)
 {
-  return CheckError(SDL_GetTextureAlphaMod(texture));
+  Uint8 alpha;
+  CheckError(SDL_GetTextureAlphaMod(texture, &alpha));
+  return alpha;
 }
 
 inline Uint8 Texture::GetAlphaMod() const
@@ -5030,7 +5026,7 @@
  * Get the additional alpha value multiplied into render copy operations.
  *
  * @param texture the texture to query.
- * @param alpha a pointer filled in with the current alpha value.
+ * @returns the current alpha value.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5043,7 +5039,9 @@
  */
 inline float GetTextureAlphaModFloat(TextureConstParam texture)
 {
-  return CheckError(SDL_GetTextureAlphaModFloat(texture));
+  float alpha;
+  CheckError(SDL_GetTextureAlphaModFloat(texture, &alpha));
+  return alpha;
 }
 
 inline float Texture::GetAlphaModFloat() const
@@ -5051,16 +5049,64 @@
   return SDL::GetTextureAlphaModFloat(m_resource);
 }
 
+/**
+ * Set an additional color and alpha values multiplied into render copy
+ * operations.
+ *
+ * When this texture is rendered, during the copy operation each source color
+ * and alpha channels are modulated by the appropriate color value according
+ * to the following formula:
+ *
+ *      srcC = srcC * (color / 255)
+ *      srcA = srcA * (alpha / 255)
+ *
+ * Color and alpha modulation is not always supported by the renderer; it will
+ * return false if either modulation is not supported.
+ *
+ * @param texture the texture to update.
+ * @param c the color and alpha channel values multiplied into copy
+ *          operations.
+ * @throws Error on failure.
+ *
+ * @threadsafety This function should only be called on the main thread.
+ *
+ * @since This function is available since SDL 3.2.0.
+ */
 inline void SetTextureMod(TextureParam texture, Color c)
 {
-  static_assert(false, "Not implemented");
+  SetTextureColorMod(texture, c.r, c.g, c.b);
+  SetTextureAlphaMod(texture, c.a);
 }
 
 inline void Texture::SetMod(Color c) { SDL::SetTextureMod(m_resource, c); }
 
+/**
+ * Set an additional color and alpha values multiplied into render copy
+ * operations.
+ *
+ * When this texture is rendered, during the copy operation each source color
+ * and alpha channels are modulated by the appropriate color value according
+ * to the following formula:
+ *
+ *      srcC = srcC * (color / 255)
+ *      srcA = srcA * (alpha / 255)
+ *
+ * Color and alpha modulation is not always supported by the renderer; it will
+ * return false if either modulation is not supported.
+ *
+ * @param texture the texture to update.
+ * @param c the color and alpha channel values multiplied into copy
+ *          operations.
+ * @throws Error on failure.
+ *
+ * @threadsafety This function should only be called on the main thread.
+ *
+ * @since This function is available since SDL 3.2.0.
+ */
 inline void SetTextureModFloat(TextureParam texture, FColor c)
 {
-  static_assert(false, "Not implemented");
+  SetTextureColorModFloat(texture, c.r, c.g, c.b);
+  SetTextureAlphaModFloat(texture, c.a);
 }
 
 inline void Texture::SetModFloat(FColor c)
@@ -5068,16 +5114,50 @@
   SDL::SetTextureModFloat(m_resource, c);
 }
 
+/**
+ * Get the additional color value multiplied into render copy operations.
+ *
+ * @param texture the texture to query.
+ * @returns the color channels (0-1) on success.
+ * @throws Error on failure.
+ *
+ * @threadsafety This function should only be called on the main thread.
+ *
+ * @since This function is available since SDL 3.2.0.
+ *
+ * @sa GetAlphaMod()
+ * @sa SetColorMod()
+ */
 inline Color GetTextureMod(TextureConstParam texture)
 {
-  static_assert(false, "Not implemented");
+  Color c;
+  GetTextureColorMod(texture, &c.r, &c.g, &c.b);
+  c.a = GetTextureAlphaMod(texture);
+  return c;
 }
 
 inline Color Texture::GetMod() const { return SDL::GetTextureMod(m_resource); }
 
+/**
+ * Get the additional color value multiplied into render copy operations.
+ *
+ * @param texture the texture to query.
+ * @returns the color channels (0-1) on success.
+ * @throws Error on failure.
+ *
+ * @threadsafety This function should only be called on the main thread.
+ *
+ * @since This function is available since SDL 3.2.0.
+ *
+ * @sa GetAlphaMod()
+ * @sa SetColorMod()
+ */
 inline FColor GetTextureModFloat(TextureConstParam texture)
 {
-  static_assert(false, "Not implemented");
+  FColor c;
+  GetTextureColorModFloat(texture, &c.r, &c.g, &c.b);
+  c.a = GetTextureAlphaModFloat(texture);
+  return c;
 }
 
 inline FColor Texture::GetModFloat() const
@@ -5115,7 +5195,7 @@
  * Get the blend mode used for texture copy operations.
  *
  * @param texture the texture to query.
- * @param blendMode a pointer filled in with the current BlendMode.
+ * @return the current BlendMode.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5126,7 +5206,9 @@
  */
 inline BlendMode GetTextureBlendMode(TextureConstParam texture)
 {
-  return CheckError(SDL_GetTextureBlendMode(texture));
+  BlendMode blendMode;
+  CheckError(SDL_GetTextureBlendMode(texture, &blendMode));
+  return blendMode;
 }
 
 inline BlendMode Texture::GetBlendMode() const
@@ -5165,7 +5247,7 @@
  * Get the scale mode used for texture scale operations.
  *
  * @param texture the texture to query.
- * @param scaleMode a pointer filled in with the current scale mode.
+ * @return the current scale mode.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5176,7 +5258,9 @@
  */
 inline ScaleMode GetTextureScaleMode(TextureConstParam texture)
 {
-  return CheckError(SDL_GetTextureScaleMode(texture));
+  ScaleMode scaleMode;
+  CheckError(SDL_GetTextureScaleMode(texture, &scaleMode));
+  return scaleMode;
 }
 
 inline ScaleMode Texture::GetScaleMode() const
@@ -5238,11 +5322,10 @@
  * may not get the pixels back if you lock the texture afterward.
  *
  * @param texture the texture to update.
- * @param rect an Rect structure representing the area to update, or nullptr to
- *             update the entire texture.
- * @param pixels the raw pixel data in the format of the texture.
- * @param pitch the number of bytes in a row of pixel data, including padding
- *              between lines.
+ * @param surface the surface containing pixel data in the format of the
+ *                texture.
+ * @param rect an Rect structure representing the area to update, or
+ *             std::nullopt to update the entire texture.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5258,7 +5341,7 @@
                           SurfaceConstParam surface,
                           OptionalRef<const RectRaw> rect = std::nullopt)
 {
-  static_assert(false, "Not implemented");
+  UpdateTexture(texture, rect, surface->pixels, surface->pitch);
 }
 
 inline void Texture::Update(OptionalRef<const RectRaw> rect,
@@ -5431,8 +5514,7 @@
  *                `TEXTUREACCESS_STREAMING`.
  * @param rect a pointer to the rectangle to lock for access. If the rect is
  *             nullptr, the entire texture will be locked.
- * @param surface a pointer to an SDL surface of size **rect**. Don't assume any
- *                specific pixel content.
+ * @returns a surface of size **rect**. Don't assume any specific pixel content.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5446,7 +5528,9 @@
   TextureParam texture,
   OptionalRef<const SDL_Rect> rect = std::nullopt)
 {
-  return CheckError(SDL_LockTextureToSurface(texture, rect));
+  SurfaceRaw surface = nullptr;
+  CheckError(SDL_LockTextureToSurface(texture, rect, &surface));
+  return Surface::Borrow(surface);
 }
 
 inline Surface Texture::LockToSurface(OptionalRef<const SDL_Rect> rect)
@@ -5511,9 +5595,24 @@
   SDL::SetRenderTarget(m_resource, texture);
 }
 
+/**
+ * Set target texture back to window
+ *
+ * This is equivalent to SetTarget(nullptr)
+ *
+ * @param renderer the rendering context.
+ * @throws Error on failure.
+ *
+ * @threadsafety This function should only be called on the main thread.
+ *
+ * @since This function is available since SDL 3.2.0.
+ *
+ * @sa SetTarget(nullptr)
+ * @sa GetTarget()
+ */
 inline void ResetRenderTarget(RendererParam renderer)
 {
-  static_assert(false, "Not implemented");
+  SetRenderTarget(renderer, nullptr);
 }
 
 inline void Renderer::ResetTarget() { SDL::ResetRenderTarget(m_resource); }
@@ -5535,7 +5634,9 @@
  */
 inline Texture GetRenderTarget(RendererParam renderer)
 {
-  return SDL_GetRenderTarget(renderer);
+  TextureRaw texture = SDL_GetRenderTarget(renderer);
+  if (texture) return Texture::Borrow(texture);
+  return {};
 }
 
 inline Texture Renderer::GetTarget() const
@@ -5574,8 +5675,7 @@
  * Renderer.ConvertEventToRenderCoordinates().
  *
  * @param renderer the rendering context.
- * @param w the width of the logical resolution.
- * @param h the height of the logical resolution.
+ * @param size the width and height of the logical resolution.
  * @param mode the presentation mode used.
  * @throws Error on failure.
  *
@@ -5591,7 +5691,7 @@
                                          const PointRaw& size,
                                          RendererLogicalPresentation mode)
 {
-  CheckError(SDL_SetRenderLogicalPresentation(renderer, size, mode));
+  CheckError(SDL_SetRenderLogicalPresentation(renderer, size.x, size.y, mode));
 }
 
 inline void Renderer::SetLogicalPresentation(const PointRaw& size,
@@ -5639,8 +5739,8 @@
  * the state for the current render target.
  *
  * @param renderer the rendering context.
- * @param w an int filled with the logical presentation width.
- * @param h an int filled with the logical presentation height.
+ * @param size a Point to be filled  with the logical presentation width and
+ *             height.
  * @param mode a variable filled with the logical presentation mode being used.
  * @throws Error on failure.
  *
@@ -5654,7 +5754,10 @@
                                          PointRaw* size,
                                          RendererLogicalPresentation* mode)
 {
-  static_assert(false, "Not implemented");
+  if (size) {
+    return GetRenderLogicalPresentation(renderer, &size->x, &size->y, mode);
+  }
+  return GetRenderLogicalPresentation(renderer, nullptr, nullptr, mode);
 }
 
 inline void Renderer::GetLogicalPresentation(
@@ -5682,8 +5785,7 @@
  * the rectangle for the current render target.
  *
  * @param renderer the rendering context.
- * @param rect a pointer filled in with the final presentation rectangle, may be
- *             nullptr.
+ * @returns with the final presentation rectangle.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5694,7 +5796,9 @@
  */
 inline FRect GetRenderLogicalPresentationRect(RendererParam renderer)
 {
-  return CheckError(SDL_GetRenderLogicalPresentationRect(renderer));
+  FRect rect;
+  CheckError(SDL_GetRenderLogicalPresentationRect(renderer, &rect));
+  return rect;
 }
 
 inline FRect Renderer::GetLogicalPresentationRect() const
@@ -5713,10 +5817,8 @@
  * - The viewport (Renderer.SetViewport)
  *
  * @param renderer the rendering context.
- * @param window_x the x coordinate in window coordinates.
- * @param window_y the y coordinate in window coordinates.
- * @param x a pointer filled with the x coordinate in render coordinates.
- * @param y a pointer filled with the y coordinate in render coordinates.
+ * @param window_coord the x, y coordinate in window coordinates.
+ * @returns the x, y coordinate in render coordinates.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5729,7 +5831,10 @@
 inline FPoint RenderCoordinatesFromWindow(RendererParam renderer,
                                           const FPointRaw& window_coord)
 {
-  return CheckError(SDL_RenderCoordinatesFromWindow(renderer, window_coord));
+  FPoint p;
+  CheckError(SDL_RenderCoordinatesFromWindow(
+    renderer, window_coord.x, window_coord.y, &p.x, &p.y));
+  return p;
 }
 
 inline FPoint Renderer::RenderCoordinatesFromWindow(
@@ -5749,10 +5854,8 @@
  * - The viewport (Renderer.SetViewport)
  *
  * @param renderer the rendering context.
- * @param x the x coordinate in render coordinates.
- * @param y the y coordinate in render coordinates.
- * @param window_x a pointer filled with the x coordinate in window coordinates.
- * @param window_y a pointer filled with the y coordinate in window coordinates.
+ * @param coord the x, y coordinates in render coordinates.
+ * @returns x, y in window coordinates.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5766,7 +5869,10 @@
 inline FPoint RenderCoordinatesToWindow(RendererParam renderer,
                                         const FPointRaw& coord)
 {
-  return CheckError(SDL_RenderCoordinatesToWindow(renderer, coord));
+  FPoint p;
+  CheckError(
+    SDL_RenderCoordinatesToWindow(renderer, coord.x, coord.y, &p.x, &p.y));
+  return p;
 }
 
 inline FPoint Renderer::RenderCoordinatesToWindow(const FPointRaw& coord) const
@@ -5798,8 +5904,7 @@
  *
  * @param renderer the rendering context.
  * @param event the event to modify.
- * @returns true if the event is converted or doesn't need conversion, or false
- *          on failure; call GetError() for more information.
+ * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
  *
@@ -5853,9 +5958,25 @@
   SDL::SetRenderViewport(m_resource, rect);
 }
 
+/**
+ * Reset the drawing area for rendering to the entire target
+ *
+ * This is equivalent to `SetViewport(std::nullopt)`
+ *
+ * @param renderer the rendering context.
+ * @throws Error on failure.
+ *
+ * @threadsafety This function should only be called on the main thread.
+ *
+ * @since This function is available since SDL 3.2.0.
+ *
+ * @sa GetViewport()
+ * @sa SetViewport()
+ * @sa ViewportSet()
+ */
 inline void ResetRenderViewport(RendererParam renderer)
 {
-  static_assert(false, "Not implemented");
+  SetRenderViewport(renderer, std::nullopt);
 }
 
 inline void Renderer::ResetViewport() { SDL::ResetRenderViewport(m_resource); }
@@ -5867,7 +5988,7 @@
  * the current render target.
  *
  * @param renderer the rendering context.
- * @param rect an Rect structure filled in with the current drawing area.
+ * @returns an Rect structure filled in with the current drawing area.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5879,7 +6000,9 @@
  */
 inline Rect GetRenderViewport(RendererParam renderer)
 {
-  return CheckError(SDL_GetRenderViewport(renderer));
+  Rect rect;
+  CheckError(SDL_GetRenderViewport(renderer, &rect));
+  return rect;
 }
 
 inline Rect Renderer::GetViewport() const
@@ -5928,8 +6051,7 @@
  * visually important or interactible content.
  *
  * @param renderer the rendering context.
- * @param rect a pointer filled in with the area that is safe for interactive
- *             content.
+ * @returns the area that is safe for interactive content.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5938,7 +6060,9 @@
  */
 inline Rect GetRenderSafeArea(RendererParam renderer)
 {
-  return CheckError(SDL_GetRenderSafeArea(renderer));
+  Rect rect;
+  CheckError(SDL_GetRenderSafeArea(renderer, &rect));
+  return rect;
 }
 
 inline Rect Renderer::GetSafeArea() const
@@ -5975,9 +6099,25 @@
   SDL::SetRenderClipRect(m_resource, rect);
 }
 
+/**
+ * Reset the clip rectangle for rendering to the entire render target
+ *
+ * This is equivalent to `SetClipRect(std::nullopt)`
+ *
+ * @param renderer the rendering context.
+ * @throws Error on failure.
+ *
+ * @threadsafety This function should only be called on the main thread.
+ *
+ * @since This function is available since SDL 3.2.0.
+ *
+ * @sa GetClipRect()
+ * @sa SetClipRect()
+ * @sa ClipEnabled()
+ */
 inline void ResetRenderClipRect(RendererParam renderer)
 {
-  static_assert(false, "Not implemented");
+  SetRenderClipRect(renderer, std::nullopt);
 }
 
 inline void Renderer::ResetClipRect() { SDL::ResetRenderClipRect(m_resource); }
@@ -5989,8 +6129,8 @@
  * cliprect for the current render target.
  *
  * @param renderer the rendering context.
- * @param rect an Rect structure filled in with the current clipping area or an
- *             empty rectangle if clipping is disabled.
+ * @returns an Rect structure filled in with the current clipping area
+ *          or an empty rectangle if clipping is disabled.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -6002,7 +6142,9 @@
  */
 inline Rect GetRenderClipRect(RendererParam renderer)
 {
-  return CheckError(SDL_GetRenderClipRect(renderer));
+  Rect rect;
+  CheckError(SDL_GetRenderClipRect(renderer, &rect));
+  return rect;
 }
 
 inline Rect Renderer::GetClipRect() const
@@ -6052,8 +6194,7 @@
  * current render target.
  *
  * @param renderer the rendering context.
- * @param scaleX the horizontal scaling factor.
- * @param scaleY the vertical scaling factor.
+ * @param scale the horizontal and vertical scaling factors.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -6064,7 +6205,7 @@
  */
 inline void SetRenderScale(RendererParam renderer, const FPointRaw& scale)
 {
-  CheckError(SDL_SetRenderScale(renderer, scale));
+  CheckError(SDL_SetRenderScale(renderer, scale.x, scale.y));
 }
 
 inline void Renderer::SetScale(const FPointRaw& scale)
@@ -6100,9 +6241,7 @@
  * Each render target has its own scale. This function gets the scale for the
  * current render target.
  *
- * @param renderer the rendering context.
- * @param scaleX a pointer filled in with the horizontal scaling factor.
- * @param scaleY a pointer filled in with the vertical scaling factor.
+ * @returns the scaling factors on success.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -6113,7 +6252,9 @@
  */
 inline FPoint GetRenderScale(RendererParam renderer)
 {
-  static_assert(false, "Not implemented");
+  FPoint p;
+  GetRenderScale(renderer, &p.x, &p.y);
+  return p;
 }
 
 inline void Renderer::GetScale(float* scaleX, float* scaleY) const
@@ -6133,12 +6274,7 @@
  * Renderer.RenderClear().
  *
  * @param renderer the rendering context.
- * @param r the red value used to draw on the rendering target.
- * @param g the green value used to draw on the rendering target.
- * @param b the blue value used to draw on the rendering target.
- * @param a the alpha value used to draw on the rendering target; usually
- *          `ALPHA_OPAQUE` (255). Use Renderer.SetDrawBlendMode to specify how
- *          the alpha channel is used.
+ * @param c the color used to draw on the rendering target.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -6150,7 +6286,7 @@
  */
 inline void SetRenderDrawColor(RendererParam renderer, ColorRaw c)
 {
-  CheckError(SDL_SetRenderDrawColor(renderer, c));
+  CheckError(SDL_SetRenderDrawColor(renderer, c.r, c.g, c.b, c.a));
 }
 
 inline void Renderer::SetDrawColor(ColorRaw c)
@@ -6165,11 +6301,7 @@
  * Renderer.RenderClear().
  *
  * @param renderer the rendering context.
- * @param r the red value used to draw on the rendering target.
- * @param g the green value used to draw on the rendering target.
- * @param b the blue value used to draw on the rendering target.
- * @param a the alpha value used to draw on the rendering target. Use
- *          Renderer.SetDrawBlendMode to specify how the alpha channel is used.
+ * @param c the color used to draw on the rendering target.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -6181,7 +6313,7 @@
  */
 inline void SetRenderDrawColorFloat(RendererParam renderer, const FColorRaw& c)
 {
-  CheckError(SDL_SetRenderDrawColorFloat(renderer, c));
+  CheckError(SDL_SetRenderDrawColorFloat(renderer, c.r, c.g, c.b, c.a));
 }
 
 inline void Renderer::SetDrawColorFloat(const FColorRaw& c)
@@ -6223,14 +6355,7 @@
  * Get the color used for drawing operations (Rect, Line and Clear).
  *
  * @param renderer the rendering context.
- * @param r a pointer filled in with the red value used to draw on the rendering
- *          target.
- * @param g a pointer filled in with the green value used to draw on the
- *          rendering target.
- * @param b a pointer filled in with the blue value used to draw on the
- *          rendering target.
- * @param a a pointer filled in with the alpha value used to draw on the
- *          rendering target; usually `ALPHA_OPAQUE` (255).
+ * @returns the color channel values used to draw on the rendering target.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -6242,7 +6367,9 @@
  */
 inline Color GetRenderDrawColor(RendererParam renderer)
 {
-  static_assert(false, "Not implemented");
+  Color c;
+  GetRenderDrawColor(renderer, &c.r, &c.g, &c.b, &c.a);
+  return c;
 }
 
 inline void Renderer::GetDrawColor(Uint8* r, Uint8* g, Uint8* b, Uint8* a) const
@@ -6289,14 +6416,7 @@
  * Get the color used for drawing operations (Rect, Line and Clear).
  *
  * @param renderer the rendering context.
- * @param r a pointer filled in with the red value used to draw on the rendering
- *          target.
- * @param g a pointer filled in with the green value used to draw on the
- *          rendering target.
- * @param b a pointer filled in with the blue value used to draw on the
- *          rendering target.
- * @param a a pointer filled in with the alpha value used to draw on the
- *          rendering target.
+ * @returns the color on success.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -6308,7 +6428,9 @@
  */
 inline FColor GetRenderDrawColorFloat(RendererParam renderer)
 {
-  static_assert(false, "Not implemented");
+  FColor c;
+  GetRenderDrawColorFloat(renderer, &c.r, &c.g, &c.b, &c.a);
+  return c;
 }
 
 inline void Renderer::GetDrawColorFloat(float* r,
@@ -6358,7 +6480,7 @@
  * Get the color scale used for render operations.
  *
  * @param renderer the rendering context.
- * @param scale a pointer filled in with the current color scale value.
+ * @return the current color scale value.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -6369,7 +6491,9 @@
  */
 inline float GetRenderColorScale(RendererParam renderer)
 {
-  return CheckError(SDL_GetRenderColorScale(renderer));
+  float scale;
+  CheckError(SDL_GetRenderColorScale(renderer, &scale));
+  return scale;
 }
 
 inline float Renderer::GetColorScale() const
@@ -6406,7 +6530,7 @@
  * Get the blend mode used for drawing operations.
  *
  * @param renderer the rendering context.
- * @param blendMode a pointer filled in with the current BlendMode.
+ * @return the current BlendMode.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -6417,7 +6541,9 @@
  */
 inline BlendMode GetRenderDrawBlendMode(RendererParam renderer)
 {
-  return CheckError(SDL_GetRenderDrawBlendMode(renderer));
+  BlendMode blendMode;
+  CheckError(SDL_GetRenderDrawBlendMode(renderer, &blendMode));
+  return blendMode;
 }
 
 inline BlendMode Renderer::GetDrawBlendMode() const
@@ -6453,8 +6579,7 @@
  * Draw a point on the current rendering target at subpixel precision.
  *
  * @param renderer the renderer which should draw a point.
- * @param x the x coordinate of the point.
- * @param y the y coordinate of the point.
+ * @param p the x, y coordinate of the point.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -6465,7 +6590,7 @@
  */
 inline void RenderPoint(RendererParam renderer, const FPointRaw& p)
 {
-  CheckError(SDL_RenderPoint(renderer, p));
+  CheckError(SDL_RenderPoint(renderer, p.x, p.y));
 }
 
 inline void Renderer::RenderPoint(const FPointRaw& p)
@@ -6478,7 +6603,6 @@
  *
  * @param renderer the renderer which should draw multiple points.
  * @param points the points to draw.
- * @param count the number of points to draw.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -6490,7 +6614,7 @@
 inline void RenderPoints(RendererParam renderer,
                          SpanRef<const FPointRaw> points)
 {
-  CheckError(SDL_RenderPoints(renderer, points));
+  CheckError(SDL_RenderPoints(renderer, points.data(), points.size()));
 }
 
 inline void Renderer::RenderPoints(SpanRef<const FPointRaw> points)
@@ -6502,10 +6626,8 @@
  * Draw a line on the current rendering target at subpixel precision.
  *
  * @param renderer the renderer which should draw a line.
- * @param x1 the x coordinate of the start point.
- * @param y1 the y coordinate of the start point.
- * @param x2 the x coordinate of the end point.
- * @param y2 the y coordinate of the end point.
+ * @param p1 the x,y coordinates of the start point.
+ * @param p2 the x,y coordinates of the end point.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -6518,7 +6640,7 @@
                        const FPointRaw& p1,
                        const FPointRaw& p2)
 {
-  CheckError(SDL_RenderLine(renderer, p1, p2));
+  CheckError(SDL_RenderLine(renderer, p1.x, p1.y, p2.x, p2.y));
 }
 
 inline void Renderer::RenderLine(const FPointRaw& p1, const FPointRaw& p2)
@@ -6532,7 +6654,6 @@
  *
  * @param renderer the renderer which should draw multiple lines.
  * @param points the points along the lines.
- * @param count the number of points, drawing count-1 lines.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -6543,7 +6664,7 @@
  */
 inline void RenderLines(RendererParam renderer, SpanRef<const FPointRaw> points)
 {
-  CheckError(SDL_RenderLines(renderer, points));
+  CheckError(SDL_RenderLines(renderer, points.data(), points.size()));
 }
 
 inline void Renderer::RenderLines(SpanRef<const FPointRaw> points)
@@ -6581,7 +6702,6 @@
  *
  * @param renderer the renderer which should draw multiple rectangles.
  * @param rects a pointer to an array of destination rectangles.
- * @param count the number of rectangles.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -6592,7 +6712,7 @@
  */
 inline void RenderRects(RendererParam renderer, SpanRef<const FRectRaw> rects)
 {
-  CheckError(SDL_RenderRects(renderer, rects));
+  CheckError(SDL_RenderRects(renderer, rects.data(), rects.size()));
 }
 
 inline void Renderer::RenderRects(SpanRef<const FRectRaw> rects)
@@ -6632,7 +6752,6 @@
  *
  * @param renderer the renderer which should fill multiple rectangles.
  * @param rects a pointer to an array of destination rectangles.
- * @param count the number of rectangles.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -6644,7 +6763,7 @@
 inline void RenderFillRects(RendererParam renderer,
                             SpanRef<const FRectRaw> rects)
 {
-  CheckError(SDL_RenderFillRects(renderer, rects));
+  CheckError(SDL_RenderFillRects(renderer, rects.data(), rects.size()));
 }
 
 inline void Renderer::RenderFillRects(SpanRef<const FRectRaw> rects)
@@ -6717,7 +6836,7 @@
                                  OptionalRef<const FRectRaw> dstrect,
                                  double angle,
                                  OptionalRef<const FPointRaw> center,
-                                 FlipMode flip)
+                                 FlipMode flip = FlipMode::SDL_FLIP_NONE)
 {
   CheckError(SDL_RenderTextureRotated(
     renderer, texture, srcrect, dstrect, angle, center, flip));
@@ -6954,10 +7073,6 @@
                                          tileScale));
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 6)
-
-#if SDL_VERSION_ATLEAST(3, 3, 6)
-
 inline void Renderer::RenderTexture9GridTiled(TextureParam texture,
                                               const FRectRaw& srcrect,
                                               float left_width,
@@ -6990,11 +7105,9 @@
  * @param renderer the rendering context.
  * @param texture (optional) The SDL texture to use.
  * @param vertices vertices.
- * @param num_vertices number of vertices.
  * @param indices (optional) An array of integer indices into the 'vertices'
  *                array, if nullptr all vertices will be rendered in sequential
  *                order.
- * @param num_indices number of indices.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -7007,9 +7120,14 @@
 inline void RenderGeometry(RendererParam renderer,
                            TextureParam texture,
                            std::span<const Vertex> vertices,
-                           std::span<const int> indices)
+                           std::span<const int> indices = {})
 {
-  CheckError(SDL_RenderGeometry(renderer, texture, vertices, indices));
+  CheckError(SDL_RenderGeometry(renderer,
+                                texture,
+                                vertices.data(),
+                                vertices.size(),
+                                indices.data(),
+                                indices.size()));
 }
 
 inline void Renderer::RenderGeometry(TextureParam texture,
@@ -7124,20 +7242,12 @@
   CheckError(SDL_SetRenderTextureAddressMode(renderer, u_mode, v_mode));
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 6)
-
-#if SDL_VERSION_ATLEAST(3, 3, 6)
-
 inline void Renderer::SetRenderTextureAddressMode(TextureAddressMode u_mode,
                                                   TextureAddressMode v_mode)
 {
   SDL::SetRenderTextureAddressMode(m_resource, u_mode, v_mode);
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 6)
-
-#if SDL_VERSION_ATLEAST(3, 3, 6)
-
 /**
  * Get the texture addressing mode used in Renderer.RenderGeometry().
  *
@@ -7161,10 +7271,6 @@
   CheckError(SDL_GetRenderTextureAddressMode(renderer, u_mode, v_mode));
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 6)
-
-#if SDL_VERSION_ATLEAST(3, 3, 6)
-
 inline void Renderer::GetRenderTextureAddressMode(TextureAddressMode* u_mode,
                                                   TextureAddressMode* v_mode)
 {
@@ -7201,7 +7307,7 @@
 inline Surface RenderReadPixels(RendererParam renderer,
                                 OptionalRef<const RectRaw> rect = {})
 {
-  return CheckError(SDL_RenderReadPixels(renderer, rect));
+  return Surface{CheckError(SDL_RenderReadPixels(renderer, rect))};
 }
 
 inline Surface Renderer::ReadPixels(OptionalRef<const RectRaw> rect) const
@@ -7469,16 +7575,18 @@
   SDL::SetRenderVSync(m_resource, vsync);
 }
 
+/// Constant for disabling renderer vsync
 constexpr int RENDERER_VSYNC_DISABLED = SDL_RENDERER_VSYNC_DISABLED;
 
+/// Constant for enabling asaptive renderer vsync
 constexpr int RENDERER_VSYNC_ADAPTIVE = SDL_RENDERER_VSYNC_ADAPTIVE;
 
 /**
  * Get VSync of the given renderer.
  *
  * @param renderer the renderer to toggle.
- * @param vsync an int filled with the current vertical refresh sync interval.
- *              See Renderer.SetVSync() for the meaning of the value.
+ * @returns the current vertical refresh sync interval. See Renderer.SetVSync()
+ *          for the meaning of the value.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -7489,7 +7597,9 @@
  */
 inline int GetRenderVSync(RendererParam renderer)
 {
-  return CheckError(SDL_GetRenderVSync(renderer));
+  int vsync;
+  CheckError(SDL_GetRenderVSync(renderer, &vsync));
+  return vsync;
 }
 
 inline int Renderer::GetVSync() const
@@ -7535,8 +7645,7 @@
  * The text is drawn in the color specified by Renderer.SetDrawColor().
  *
  * @param renderer the renderer which should draw a line of text.
- * @param x the x coordinate where the top-left corner of the text will draw.
- * @param y the y coordinate where the top-left corner of the text will draw.
+ * @param p the x,y coordinates where the top-left corner of the text will draw.
  * @param str the string to render.
  * @throws Error on failure.
  *
@@ -7551,7 +7660,7 @@
                             const FPointRaw& p,
                             StringParam str)
 {
-  CheckError(SDL_RenderDebugText(renderer, p, str));
+  CheckError(SDL_RenderDebugText(renderer, p.x, p.y, str));
 }
 
 inline void Renderer::RenderDebugText(const FPointRaw& p, StringParam str)
@@ -7570,10 +7679,9 @@
  * Renderer.RenderDebugText.
  *
  * @param renderer the renderer which should draw the text.
- * @param x the x coordinate where the top-left corner of the text will draw.
- * @param y the y coordinate where the top-left corner of the text will draw.
+ * @param p the x,y coordinate where the top-left corner of the text will draw.
  * @param fmt the format string to draw.
- * @param ... additional parameters matching % tokens in the `fmt` string, if
+ * @param args additional parameters matching % tokens in the `fmt` string, if
  *            any.
  * @throws Error on failure.
  *
@@ -7590,7 +7698,8 @@
                                   std::string_view fmt,
                                   ARGS... args)
 {
-  CheckError(SDL_RenderDebugTextFormat(renderer, p, fmt, args));
+  RenderDebugText(
+    renderer, p, std::vformat(fmt, std::make_format_args(args...)));
 }
 
 template<class... ARGS>
@@ -7598,7 +7707,7 @@
                                             std::string_view fmt,
                                             ARGS... args)
 {
-  SDL::RenderDebugTextFormat(m_resource, p, fmt, args);
+  SDL::RenderDebugTextFormat(m_resource, p, fmt, args...);
 }
 
 #if SDL_VERSION_ATLEAST(3, 3, 6)
@@ -7624,19 +7733,11 @@
   CheckError(SDL_SetDefaultTextureScaleMode(renderer, scale_mode));
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 6)
-
-#if SDL_VERSION_ATLEAST(3, 3, 6)
-
 inline void Renderer::SetDefaultTextureScaleMode(ScaleMode scale_mode)
 {
   SDL::SetDefaultTextureScaleMode(m_resource, scale_mode);
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 6)
-
-#if SDL_VERSION_ATLEAST(3, 3, 6)
-
 /**
  * Get default texture scale mode of the given renderer.
  *
@@ -7658,19 +7759,11 @@
   CheckError(SDL_GetDefaultTextureScaleMode(renderer, scale_mode));
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 6)
-
-#if SDL_VERSION_ATLEAST(3, 3, 6)
-
 inline void Renderer::GetDefaultTextureScaleMode(ScaleMode* scale_mode)
 {
   SDL::GetDefaultTextureScaleMode(m_resource, scale_mode);
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 6)
-
-#if SDL_VERSION_ATLEAST(3, 3, 6)
-
 /**
  * A custom GPU render state.
  *
@@ -7689,10 +7782,7 @@
 
 public:
   /// Default ctor
-  constexpr GPURenderState(std::nullptr_t = nullptr) noexcept
-    : m_resource(0)
-  {
-  }
+  constexpr GPURenderState() = default;
 
   /**
    * Constructs from GPURenderStateParam.
@@ -7701,7 +7791,7 @@
    *
    * This assumes the ownership, call release() if you need to take back.
    */
-  constexpr explicit GPURenderState(const GPURenderStateRaw resource) noexcept
+  constexpr explicit GPURenderState(const GPURenderStateRaw resource)
     : m_resource(resource)
   {
   }
@@ -7710,7 +7800,7 @@
   constexpr GPURenderState(const GPURenderState& other) = delete;
 
   /// Move constructor
-  constexpr GPURenderState(GPURenderState&& other) noexcept
+  constexpr GPURenderState(GPURenderState&& other)
     : GPURenderState(other.release())
   {
   }
@@ -7719,8 +7809,6 @@
 
   constexpr GPURenderState(GPURenderStateRef&& other) = delete;
 
-#if SDL_VERSION_ATLEAST(3, 3, 6)
-
   /**
    * Create custom GPU render state.
    *
@@ -7743,29 +7831,21 @@
   {
   }
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 6)
-
   /// Destructor
   ~GPURenderState() { SDL_DestroyGPURenderState(m_resource); }
 
   /// Assignment operator.
-  constexpr GPURenderState& operator=(GPURenderState&& other) noexcept
+  GPURenderState& operator=(GPURenderState other)
   {
     std::swap(m_resource, other.m_resource);
     return *this;
   }
 
-protected:
-  /// Assignment operator.
-  constexpr GPURenderState& operator=(const GPURenderState& other) noexcept =
-    default;
-
-public:
   /// Retrieves underlying GPURenderStateRaw.
-  constexpr GPURenderStateRaw get() const noexcept { return m_resource; }
+  constexpr GPURenderStateRaw get() const { return m_resource; }
 
   /// Retrieves underlying GPURenderStateRaw and clear this.
-  constexpr GPURenderStateRaw release() noexcept
+  constexpr GPURenderStateRaw release()
   {
     auto r = m_resource;
     m_resource = nullptr;
@@ -7773,11 +7853,13 @@
   }
 
   /// Comparison
-  constexpr auto operator<=>(const GPURenderState& other) const noexcept =
-    default;
+  constexpr auto operator<=>(const GPURenderState& other) const = default;
+
+  /// Comparison
+  constexpr bool operator==(std::nullptr_t _) const { return !m_resource; }
 
   /// Converts to bool
-  constexpr explicit operator bool() const noexcept { return !!m_resource; }
+  constexpr explicit operator bool() const { return !!m_resource; }
 
   /// Converts to GPURenderStateParam
   constexpr operator GPURenderStateParam() const noexcept
@@ -7785,11 +7867,10 @@
     return {m_resource};
   }
 
-#if SDL_VERSION_ATLEAST(3, 3, 6)
-
   /**
    * Destroy custom GPU render state.
    *
+   *
    * @threadsafety This function should be called on the thread that created the
    *               renderer.
    *
@@ -7799,10 +7880,6 @@
    */
   void Destroy();
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 6)
-
-#if SDL_VERSION_ATLEAST(3, 3, 6)
-
   /**
    * Set fragment shader uniform variables in a custom GPU render state.
    *
@@ -7820,17 +7897,11 @@
    * @since This function is available since SDL 3.4.0.
    */
   void SetFragmentUniforms(Uint32 slot_index, const void* data, Uint32 length);
-
-#endif // SDL_VERSION_ATLEAST(3, 3, 6)
 };
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 6)
-
 /// Semi-safe reference for GPURenderState.
 struct GPURenderStateRef : GPURenderState
 {
-  using GPURenderState::GPURenderState;
-
   /**
    * Constructs from GPURenderStateParam.
    *
@@ -7838,25 +7909,13 @@
    *
    * This does not takes ownership!
    */
-  GPURenderStateRef(GPURenderStateParam resource) noexcept
+  GPURenderStateRef(GPURenderStateParam resource)
     : GPURenderState(resource.value)
   {
   }
 
-  /**
-   * Constructs from GPURenderStateParam.
-   *
-   * @param resource a GPURenderStateRaw or GPURenderState.
-   *
-   * This does not takes ownership!
-   */
-  GPURenderStateRef(GPURenderStateRaw resource) noexcept
-    : GPURenderState(resource)
-  {
-  }
-
   /// Copy constructor.
-  GPURenderStateRef(const GPURenderStateRef& other) noexcept
+  GPURenderStateRef(const GPURenderStateRef& other)
     : GPURenderState(other.get())
   {
   }
@@ -7865,8 +7924,6 @@
   ~GPURenderStateRef() { release(); }
 };
 
-#if SDL_VERSION_ATLEAST(3, 3, 6)
-
 /**
  * Create custom GPU render state.
  *
@@ -7890,20 +7947,12 @@
   return GPURenderState(renderer, createinfo);
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 6)
-
-#if SDL_VERSION_ATLEAST(3, 3, 6)
-
 inline GPURenderStateRef Renderer::CreateGPURenderState(
   GPURenderStateCreateInfo* createinfo)
 {
   return GPURenderState(m_resource, createinfo);
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 6)
-
-#if SDL_VERSION_ATLEAST(3, 3, 6)
-
 /**
  * Set fragment shader uniform variables in a custom GPU render state.
  *
@@ -7930,10 +7979,6 @@
     SDL_SetGPURenderStateFragmentUniforms(state, slot_index, data, length));
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 6)
-
-#if SDL_VERSION_ATLEAST(3, 3, 6)
-
 inline void GPURenderState::SetFragmentUniforms(Uint32 slot_index,
                                                 const void* data,
                                                 Uint32 length)
@@ -7941,10 +7986,6 @@
   SDL::SetGPURenderStateFragmentUniforms(m_resource, slot_index, data, length);
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 6)
-
-#if SDL_VERSION_ATLEAST(3, 3, 6)
-
 /**
  * Set custom GPU render state.
  *
@@ -7966,19 +8007,11 @@
   CheckError(SDL_SetGPURenderState(renderer, state));
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 6)
-
-#if SDL_VERSION_ATLEAST(3, 3, 6)
-
 inline void Renderer::SetGPURenderState(GPURenderStateParam state)
 {
   SDL::SetGPURenderState(m_resource, state);
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 6)
-
-#if SDL_VERSION_ATLEAST(3, 3, 6)
-
 /**
  * Destroy custom GPU render state.
  *
@@ -7996,10 +8029,6 @@
   SDL_DestroyGPURenderState(state);
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 3, 6)
-
-#if SDL_VERSION_ATLEAST(3, 3, 6)
-
 inline void GPURenderState::Destroy() { DestroyGPURenderState(release()); }
 
 #endif // SDL_VERSION_ATLEAST(3, 3, 6)
