--- src/generated/SDL3pp_ttf.h
+++ include/SDL3pp/SDL3pp_ttf.h
@@ -1,17 +1,25 @@
 #ifndef SDL3PP_TTF_H_
 #define SDL3PP_TTF_H_
 
-#include <SDL3/SDL_ttf.h>
 #include "SDL3pp_gpu.h"
 #include "SDL3pp_render.h"
 #include "SDL3pp_surface.h"
 #include "SDL3pp_version.h"
 #include "SDL3pp_video.h"
 
+#if !defined(SDL3PP_DISABLE_TTF) && !defined(SDL3PP_ENABLE_TTF) &&             \
+  __has_include(<SDL3_ttf/SDL_ttf.h>)
+#define SDL3PP_ENABLE_TTF
+#endif
+
+#if defined(SDL3PP_ENABLE_TTF) || defined(SDL3PP_DOC)
+
+#include <SDL3_ttf/SDL_ttf.h>
+
 namespace SDL {
 
 /**
- * @defgroup CategorySDLTTF Category SDLTTF
+ * @defgroup CategorySDLTTF True type font support
  *
  * Header file for SDL_ttf library
  *
@@ -165,25 +173,39 @@
   constexpr auto operator->() { return value; }
 };
 
-/// Printable format: "%d.%d.%d", MAJOR, MINOR, MICRO
+#ifdef SDL3PP_DOC
+
+/**
+ * @name TTF version
+ * @{
+ * Printable format: "%d.%d.%d", MAJOR, MINOR, MICRO
+ */
 #define SDL_TTF_MAJOR_VERSION
 
 #define SDL_TTF_MINOR_VERSION
 
 #define SDL_TTF_MICRO_VERSION
 
-/// This is the version number macro for the current SDL_ttf version.
+///@}
+
+/**
+ * This is the version number macro for the current SDL_ttf version.
+ */
 #define SDL_TTF_VERSION                                                        \
   SDL_VERSIONNUM(                                                              \
     SDL_TTF_MAJOR_VERSION, SDL_TTF_MINOR_VERSION, SDL_TTF_MICRO_VERSION)
 
-/// This macro will evaluate to true if compiled with SDL_ttf at least X.Y.Z.
+/**
+ * This macro will evaluate to true if compiled with SDL_ttf at least X.Y.Z.
+ */
 #define SDL_TTF_VERSION_ATLEAST(X, Y, Z)                                       \
   ((SDL_TTF_MAJOR_VERSION >= X) &&                                             \
    (SDL_TTF_MAJOR_VERSION > X || SDL_TTF_MINOR_VERSION >= Y) &&                \
    (SDL_TTF_MAJOR_VERSION > X || SDL_TTF_MINOR_VERSION > Y ||                  \
     SDL_TTF_MICRO_VERSION >= Z))
 
+#endif // SDL3PP_DOC
+
 namespace TTF {
 
 /**
@@ -384,7 +406,7 @@
  *
  * The values here are chosen to match
  * [hb_direction_t](https://harfbuzz.github.io/harfbuzz-hb-common.html#hb-direction-t)
- * .
+ *
  *
  * @since This enum is available since SDL_ttf 3.0.0.
  *
@@ -470,8 +492,6 @@
    * size becomes the index of choosing which size. If the value is too high,
    * the last indexed size will be the default.
    *
-   * When done with the returned Font, use Font.Close() to dispose of it.
-   *
    * @param file path to font file.
    * @param ptsize point size to use for the newly-opened font.
    * @post a valid Font on success.
@@ -498,12 +518,10 @@
    * If `closeio` is true, `src` will be automatically closed once the font is
    * closed. Otherwise you should keep `src` open until the font is closed.
    *
-   * When done with the returned Font, use Font.Close() to dispose of it.
-   *
    * @param src an IOStream to provide a font file's data.
+   * @param ptsize point size to use for the newly-opened font.
    * @param closeio true to close `src` when the font is closed, false to leave
    *                it open.
-   * @param ptsize point size to use for the newly-opened font.
    * @post a valid Font on success.
    * @throws Error on failure.
    *
@@ -514,7 +532,7 @@
    * @sa Font.Close
    */
   Font(IOStreamParam src, float ptsize, bool closeio = false)
-    : m_resource(CheckError(TTF_OpenFontIO(src, ptsize, closeio)))
+    : m_resource(CheckError(TTF_OpenFontIO(src, closeio, ptsize)))
   {
   }
 
@@ -619,7 +637,6 @@
    * @since This function is available since SDL_ttf 3.0.0.
    *
    * @sa Font.Font
-   * @sa Font.Font
    */
   void Close();
 
@@ -629,8 +646,6 @@
    * The copy will be distinct from the original, but will share the font file
    * and have the same size and style as the original.
    *
-   * When done with the returned Font, use Font.Close() to dispose of it.
-   *
    * @returns a valid Font on success.
    * @throws Error on failure.
    *
@@ -1235,7 +1250,7 @@
    * This updates any Text objects using this font.
    *
    * @param script an [ISO 15924
-   *               code](https://unicode.org/iso15924/iso15924-codes.html) .
+   *               code](https://unicode.org/iso15924/iso15924-codes.html).
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -1387,8 +1402,7 @@
    *
    * @param previous_ch the previous codepoint.
    * @param ch the current codepoint.
-   * @param kerning a pointer filled in with the kerning size between the two
-   *                glyphs, in pixels, may be nullptr.
+   * @returns the kerning size between the two glyphs, in pixels.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -1398,9 +1412,26 @@
    */
   int GetGlyphKerning(Uint32 previous_ch, Uint32 ch) const;
 
+  /**
+   * Calculate the dimensions of a rendered string of UTF-8 text.
+   *
+   * This will report the width and height, in pixels, of the space that the
+   * specified string will take to fully render.
+   *
+   * @param text text to calculate, in UTF-8 encoding.
+   * @returns return a Point with the width, height.
+   * @throws Error on failure.
+   *
+   * @threadsafety This function should be called on the thread that created the
+   *               font.
+   *
+   * @since This function is available since SDL_ttf 3.0.0.
+   */
   Point GetStringSize(std::string_view text) const
   {
-    static_assert(false, "Not implemented");
+    Point p;
+    GetStringSize(text, &p.x, &p.y);
+    return p;
   }
 
   /**
@@ -1409,10 +1440,7 @@
    * This will report the width and height, in pixels, of the space that the
    * specified string will take to fully render.
    *
-   * @param font the font to query.
    * @param text text to calculate, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param w will be filled with width, in pixels, on return.
    * @param h will be filled with height, in pixels, on return.
    * @throws Error on failure.
@@ -1424,9 +1452,32 @@
    */
   void GetStringSize(std::string_view text, int* w, int* h) const;
 
+  /**
+   * Calculate the dimensions of a rendered string of UTF-8 text.
+   *
+   * This will report the width and height, in pixels, of the space that the
+   * specified string will take to fully render.
+   *
+   * Text is wrapped to multiple lines on line endings and on word boundaries if
+   * it extends beyond `wrap_width` in pixels.
+   *
+   * If wrap_width is 0, this function will only wrap on newline characters.
+   *
+   * @param text text to calculate, in UTF-8 encoding.
+   * @param wrap_width the maximum width or 0 to wrap on newline characters.
+   * @returns return a Point with the width, height on success.
+   * @throws Error on failure.
+   *
+   * @threadsafety This function should be called on the thread that created the
+   *               font.
+   *
+   * @since This function is available since SDL_ttf 3.0.0.
+   */
   Point GetStringSizeWrapped(std::string_view text, int wrap_width) const
   {
-    static_assert(false, "Not implemented");
+    Point p;
+    GetStringSizeWrapped(text, wrap_width, &p.x, &p.y);
+    return p;
   }
 
   /**
@@ -1440,10 +1491,7 @@
    *
    * If wrap_width is 0, this function will only wrap on newline characters.
    *
-   * @param font the font to query.
    * @param text text to calculate, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param wrap_width the maximum width or 0 to wrap on newline characters.
    * @param w will be filled with width, in pixels, on return.
    * @param h will be filled with height, in pixels, on return.
@@ -1467,10 +1515,7 @@
    *
    * This does not need to render the string to do this calculation.
    *
-   * @param font the font to query.
    * @param text text to calculate, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param max_width maximum width, in pixels, available for the string, or 0
    *                  for unbounded width.
    * @param measured_width a pointer filled in with the width, in pixels, of the
@@ -1506,10 +1551,7 @@
    * You can render at other quality levels with Font.RenderText_Shaded,
    * Font.RenderText_Blended, and Font.RenderText_LCD.
    *
-   * @param font the font to render with.
    * @param text text to render, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param fg the foreground color for the text.
    * @returns a new 8-bit, palettized surface, or nullptr if there was an error.
    *
@@ -1541,10 +1583,7 @@
    * You can render at other quality levels with Font.RenderText_Shaded_Wrapped,
    * Font.RenderText_Blended_Wrapped, and Font.RenderText_LCD_Wrapped.
    *
-   * @param font the font to render with.
    * @param text text to render, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param fg the foreground color for the text.
    * @param wrapLength the maximum width of the text surface or 0 to wrap on
    *                   newline characters.
@@ -1610,10 +1649,7 @@
    * You can render at other quality levels with Font.RenderText_Solid,
    * Font.RenderText_Blended, and Font.RenderText_LCD.
    *
-   * @param font the font to render with.
    * @param text text to render, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param fg the foreground color for the text.
    * @param bg the background color for the text.
    * @returns a new 8-bit, palettized surface, or nullptr if there was an error.
@@ -1646,10 +1682,7 @@
    * You can render at other quality levels with Font.RenderText_Solid_Wrapped,
    * Font.RenderText_Blended_Wrapped, and Font.RenderText_LCD_Wrapped.
    *
-   * @param font the font to render with.
    * @param text text to render, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param fg the foreground color for the text.
    * @param bg the background color for the text.
    * @param wrap_width the maximum width of the text surface or 0 to wrap on
@@ -1718,10 +1751,7 @@
    * You can render at other quality levels with Font.RenderText_Solid,
    * Font.RenderText_Shaded, and Font.RenderText_LCD.
    *
-   * @param font the font to render with.
    * @param text text to render, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param fg the foreground color for the text.
    * @returns a new 32-bit, ARGB surface, or nullptr if there was an error.
    *
@@ -1752,10 +1782,7 @@
    * You can render at other quality levels with Font.RenderText_Solid_Wrapped,
    * Font.RenderText_Shaded_Wrapped, and Font.RenderText_LCD_Wrapped.
    *
-   * @param font the font to render with.
    * @param text text to render, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param fg the foreground color for the text.
    * @param wrap_width the maximum width of the text surface or 0 to wrap on
    *                   newline characters.
@@ -1820,10 +1847,7 @@
    * You can render at other quality levels with Font.RenderText_Solid,
    * Font.RenderText_Shaded, and Font.RenderText_Blended.
    *
-   * @param font the font to render with.
    * @param text text to render, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param fg the foreground color for the text.
    * @param bg the background color for the text.
    * @returns a new 32-bit, ARGB surface, or nullptr if there was an error.
@@ -1856,10 +1880,7 @@
    * You can render at other quality levels with Font.RenderText_Solid_Wrapped,
    * Font.RenderText_Shaded_Wrapped, and Font.RenderText_Blended_Wrapped.
    *
-   * @param font the font to render with.
    * @param text text to render, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param fg the foreground color for the text.
    * @param bg the background color for the text.
    * @param wrap_width the maximum width of the text surface or 0 to wrap on
@@ -1989,9 +2010,9 @@
  * When done with the returned Font, use Font.Close() to dispose of it.
  *
  * @param src an IOStream to provide a font file's data.
+ * @param ptsize point size to use for the newly-opened font.
  * @param closeio true to close `src` when the font is closed, false to leave it
  *                open.
- * @param ptsize point size to use for the newly-opened font.
  * @returns a valid Font on success.
  * @throws Error on failure.
  *
@@ -2113,7 +2134,7 @@
  */
 inline Font CopyFont(FontParam existing_font)
 {
-  return CheckError(TTF_CopyFont(existing_font));
+  return Font(CheckError(TTF_CopyFont(existing_font)));
 }
 
 inline Font Font::Copy() const { return SDL::CopyFont(m_resource); }
@@ -2140,7 +2161,7 @@
  */
 inline PropertiesRef GetFontProperties(FontParam font)
 {
-  return CheckError(TTF_GetFontProperties(font));
+  return {CheckError(TTF_GetFontProperties(font))};
 }
 
 inline PropertiesRef Font::GetProperties()
@@ -2605,79 +2626,35 @@
  */
 inline int GetFontWeight(FontParam font) { return TTF_GetFontWeight(font); }
 
-#endif // SDL_TTF_VERSION_ATLEAST(3, 2, 2)
-
-#if SDL_TTF_VERSION_ATLEAST(3, 2, 2)
-
 inline int Font::GetWeight() const { return SDL::GetFontWeight(m_resource); }
 
-#endif // SDL_TTF_VERSION_ATLEAST(3, 2, 2)
-
-#if SDL_TTF_VERSION_ATLEAST(3, 2, 2)
-
 constexpr int FONT_WEIGHT_THIN =
   TTF_FONT_WEIGHT_THIN; ///< Thin (100) named font weight value
 
-#endif // SDL_TTF_VERSION_ATLEAST(3, 2, 2)
-
-#if SDL_TTF_VERSION_ATLEAST(3, 2, 2)
-
 constexpr int FONT_WEIGHT_EXTRA_LIGHT =
   TTF_FONT_WEIGHT_EXTRA_LIGHT; ///< ExtraLight (200) named font weight value
 
-#endif // SDL_TTF_VERSION_ATLEAST(3, 2, 2)
-
-#if SDL_TTF_VERSION_ATLEAST(3, 2, 2)
-
 constexpr int FONT_WEIGHT_LIGHT =
   TTF_FONT_WEIGHT_LIGHT; ///< Light (300) named font weight value
 
-#endif // SDL_TTF_VERSION_ATLEAST(3, 2, 2)
-
-#if SDL_TTF_VERSION_ATLEAST(3, 2, 2)
-
 constexpr int FONT_WEIGHT_NORMAL =
   TTF_FONT_WEIGHT_NORMAL; ///< Normal (400) named font weight value
 
-#endif // SDL_TTF_VERSION_ATLEAST(3, 2, 2)
-
-#if SDL_TTF_VERSION_ATLEAST(3, 2, 2)
-
 constexpr int FONT_WEIGHT_MEDIUM =
   TTF_FONT_WEIGHT_MEDIUM; ///< Medium (500) named font weight value
 
-#endif // SDL_TTF_VERSION_ATLEAST(3, 2, 2)
-
-#if SDL_TTF_VERSION_ATLEAST(3, 2, 2)
-
 constexpr int FONT_WEIGHT_SEMI_BOLD =
   TTF_FONT_WEIGHT_SEMI_BOLD; ///< SemiBold (600) named font weight value
 
-#endif // SDL_TTF_VERSION_ATLEAST(3, 2, 2)
-
-#if SDL_TTF_VERSION_ATLEAST(3, 2, 2)
-
 constexpr int FONT_WEIGHT_BOLD =
   TTF_FONT_WEIGHT_BOLD; ///< Bold (700) named font weight value
 
-#endif // SDL_TTF_VERSION_ATLEAST(3, 2, 2)
-
-#if SDL_TTF_VERSION_ATLEAST(3, 2, 2)
-
 constexpr int FONT_WEIGHT_EXTRA_BOLD =
   TTF_FONT_WEIGHT_EXTRA_BOLD; ///< ExtraBold (800) named font weight value
 
-#endif // SDL_TTF_VERSION_ATLEAST(3, 2, 2)
-
-#if SDL_TTF_VERSION_ATLEAST(3, 2, 2)
-
 constexpr int FONT_WEIGHT_BLACK =
   TTF_FONT_WEIGHT_BLACK; ///< Black (900) named font weight value
 
-#endif // SDL_TTF_VERSION_ATLEAST(3, 2, 2)
-
-#if SDL_TTF_VERSION_ATLEAST(3, 2, 2)
-
 constexpr int FONT_WEIGHT_EXTRA_BLACK =
   TTF_FONT_WEIGHT_EXTRA_BLACK; ///< ExtraBlack (950) named font weight value
 
@@ -3060,7 +3037,7 @@
  *
  * @param font the font to modify.
  * @param script an [ISO 15924
- *               code](https://unicode.org/iso15924/iso15924-codes.html) .
+ *               code](https://unicode.org/iso15924/iso15924-codes.html).
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -3189,7 +3166,7 @@
  */
 inline Surface GetGlyphImage(FontParam font, Uint32 ch, ImageType* image_type)
 {
-  return TTF_GetGlyphImage(font, ch, image_type);
+  return Surface{TTF_GetGlyphImage(font, ch, image_type)};
 }
 
 inline Surface Font::GetGlyphImage(Uint32 ch, ImageType* image_type) const
@@ -3219,7 +3196,7 @@
                                      Uint32 glyph_index,
                                      ImageType* image_type)
 {
-  return TTF_GetGlyphImageForIndex(font, glyph_index, image_type);
+  return Surface(TTF_GetGlyphImageForIndex(font, glyph_index, image_type));
 }
 
 inline Surface Font::GetGlyphImageForIndex(Uint32 glyph_index,
@@ -3283,8 +3260,7 @@
  * @param font the font to query.
  * @param previous_ch the previous codepoint.
  * @param ch the current codepoint.
- * @param kerning a pointer filled in with the kerning size between the two
- *                glyphs, in pixels, may be nullptr.
+ * @returns the kerning size between the two glyphs, in pixels.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -3294,7 +3270,8 @@
  */
 inline int GetGlyphKerning(FontParam font, Uint32 previous_ch, Uint32 ch)
 {
-  return CheckError(TTF_GetGlyphKerning(font, previous_ch, ch));
+  if (int r; TTF_GetGlyphKerning(font, previous_ch, ch, &r)) return r;
+  throw Error();
 }
 
 inline int Font::GetGlyphKerning(Uint32 previous_ch, Uint32 ch) const
@@ -3310,8 +3287,6 @@
  *
  * @param font the font to query.
  * @param text text to calculate, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param w will be filled with width, in pixels, on return.
  * @param h will be filled with height, in pixels, on return.
  * @throws Error on failure.
@@ -3323,7 +3298,7 @@
  */
 inline void GetStringSize(FontParam font, std::string_view text, int* w, int* h)
 {
-  CheckError(TTF_GetStringSize(font, text, w, h));
+  CheckError(TTF_GetStringSize(font, text.data(), text.size(), w, h));
 }
 
 inline void Font::GetStringSize(std::string_view text, int* w, int* h) const
@@ -3344,8 +3319,6 @@
  *
  * @param font the font to query.
  * @param text text to calculate, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param wrap_width the maximum width or 0 to wrap on newline characters.
  * @param w will be filled with width, in pixels, on return.
  * @param h will be filled with height, in pixels, on return.
@@ -3362,7 +3335,8 @@
                                  int* w,
                                  int* h)
 {
-  CheckError(TTF_GetStringSizeWrapped(font, text, wrap_width, w, h));
+  CheckError(
+    TTF_GetStringSizeWrapped(font, text.data(), text.size(), wrap_width, w, h));
 }
 
 inline void Font::GetStringSizeWrapped(std::string_view text,
@@ -3383,8 +3357,6 @@
  *
  * @param font the font to query.
  * @param text text to calculate, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param max_width maximum width, in pixels, available for the string, or 0 for
  *                  unbounded width.
  * @param measured_width a pointer filled in with the width, in pixels, of the
@@ -3404,8 +3376,12 @@
                           int* measured_width,
                           size_t* measured_length)
 {
-  CheckError(
-    TTF_MeasureString(font, text, max_width, measured_width, measured_length));
+  CheckError(TTF_MeasureString(font,
+                               text.data(),
+                               text.size(),
+                               max_width,
+                               measured_width,
+                               measured_length));
 }
 
 inline void Font::MeasureString(std::string_view text,
@@ -3436,8 +3412,6 @@
  *
  * @param font the font to render with.
  * @param text text to render, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param fg the foreground color for the text.
  * @returns a new 8-bit, palettized surface, or nullptr if there was an error.
  *
@@ -3454,7 +3428,7 @@
  */
 inline Surface RenderText_Solid(FontParam font, std::string_view text, Color fg)
 {
-  return TTF_RenderText_Solid(font, text, fg);
+  return Surface{TTF_RenderText_Solid(font, text.data(), text.size(), fg)};
 }
 
 inline Surface Font::RenderText_Solid(std::string_view text, Color fg) const
@@ -3479,8 +3453,6 @@
  *
  * @param font the font to render with.
  * @param text text to render, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param fg the foreground color for the text.
  * @param wrapLength the maximum width of the text surface or 0 to wrap on
  *                   newline characters.
@@ -3501,7 +3473,8 @@
                                         Color fg,
                                         int wrapLength)
 {
-  return TTF_RenderText_Solid_Wrapped(font, text, fg, wrapLength);
+  return Surface(TTF_RenderText_Solid_Wrapped(
+    font, text.data(), text.size(), fg, wrapLength));
 }
 
 inline Surface Font::RenderText_Solid_Wrapped(std::string_view text,
@@ -3540,7 +3513,7 @@
  */
 inline Surface RenderGlyph_Solid(FontParam font, Uint32 ch, ColorRaw fg)
 {
-  return TTF_RenderGlyph_Solid(font, ch, fg);
+  return Surface(TTF_RenderGlyph_Solid(font, ch, fg));
 }
 
 inline Surface Font::RenderGlyph_Solid(Uint32 ch, ColorRaw fg) const
@@ -3568,8 +3541,6 @@
  *
  * @param font the font to render with.
  * @param text text to render, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param fg the foreground color for the text.
  * @param bg the background color for the text.
  * @returns a new 8-bit, palettized surface, or nullptr if there was an error.
@@ -3589,7 +3560,7 @@
                                  Color fg,
                                  Color bg)
 {
-  return TTF_RenderText_Shaded(font, text, fg, bg);
+  return Surface(TTF_RenderText_Shaded(font, text.data(), text.size(), fg, bg));
 }
 
 inline Surface Font::RenderText_Shaded(std::string_view text,
@@ -3617,8 +3588,6 @@
  *
  * @param font the font to render with.
  * @param text text to render, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param fg the foreground color for the text.
  * @param bg the background color for the text.
  * @param wrap_width the maximum width of the text surface or 0 to wrap on
@@ -3641,7 +3610,8 @@
                                          Color bg,
                                          int wrap_width)
 {
-  return TTF_RenderText_Shaded_Wrapped(font, text, fg, bg, wrap_width);
+  return Surface(TTF_RenderText_Shaded_Wrapped(
+    font, text.data(), text.size(), fg, bg, wrap_width));
 }
 
 inline Surface Font::RenderText_Shaded_Wrapped(std::string_view text,
@@ -3686,7 +3656,7 @@
                                   ColorRaw fg,
                                   ColorRaw bg)
 {
-  return TTF_RenderGlyph_Shaded(font, ch, fg, bg);
+  return Surface(TTF_RenderGlyph_Shaded(font, ch, fg, bg));
 }
 
 inline Surface Font::RenderGlyph_Shaded(Uint32 ch,
@@ -3715,8 +3685,6 @@
  *
  * @param font the font to render with.
  * @param text text to render, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param fg the foreground color for the text.
  * @returns a new 32-bit, ARGB surface, or nullptr if there was an error.
  *
@@ -3734,7 +3702,7 @@
                                   std::string_view text,
                                   Color fg)
 {
-  return TTF_RenderText_Blended(font, text, fg);
+  return Surface(TTF_RenderText_Blended(font, text.data(), text.size(), fg));
 }
 
 inline Surface Font::RenderText_Blended(std::string_view text, Color fg) const
@@ -3759,8 +3727,6 @@
  *
  * @param font the font to render with.
  * @param text text to render, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param fg the foreground color for the text.
  * @param wrap_width the maximum width of the text surface or 0 to wrap on
  *                   newline characters.
@@ -3781,7 +3747,8 @@
                                           Color fg,
                                           int wrap_width)
 {
-  return TTF_RenderText_Blended_Wrapped(font, text, fg, wrap_width);
+  return Surface(TTF_RenderText_Blended_Wrapped(
+    font, text.data(), text.size(), fg, wrap_width));
 }
 
 inline Surface Font::RenderText_Blended_Wrapped(std::string_view text,
@@ -3820,7 +3787,7 @@
  */
 inline Surface RenderGlyph_Blended(FontParam font, Uint32 ch, ColorRaw fg)
 {
-  return TTF_RenderGlyph_Blended(font, ch, fg);
+  return Surface(TTF_RenderGlyph_Blended(font, ch, fg));
 }
 
 inline Surface Font::RenderGlyph_Blended(Uint32 ch, ColorRaw fg) const
@@ -3847,8 +3814,6 @@
  *
  * @param font the font to render with.
  * @param text text to render, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param fg the foreground color for the text.
  * @param bg the background color for the text.
  * @returns a new 32-bit, ARGB surface, or nullptr if there was an error.
@@ -3868,7 +3833,7 @@
                               Color fg,
                               Color bg)
 {
-  return TTF_RenderText_LCD(font, text, fg, bg);
+  return Surface(TTF_RenderText_LCD(font, text.data(), text.size(), fg, bg));
 }
 
 inline Surface Font::RenderText_LCD(std::string_view text,
@@ -3895,8 +3860,6 @@
  *
  * @param font the font to render with.
  * @param text text to render, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param fg the foreground color for the text.
  * @param bg the background color for the text.
  * @param wrap_width the maximum width of the text surface or 0 to wrap on
@@ -3919,7 +3882,8 @@
                                       Color bg,
                                       int wrap_width)
 {
-  return TTF_RenderText_LCD_Wrapped(font, text, fg, bg, wrap_width);
+  return Surface(TTF_RenderText_LCD_Wrapped(
+    font, text.data(), text.size(), fg, bg, wrap_width));
 }
 
 inline Surface Font::RenderText_LCD_Wrapped(std::string_view text,
@@ -3964,7 +3928,7 @@
                                ColorRaw fg,
                                ColorRaw bg)
 {
-  return TTF_RenderGlyph_LCD(font, ch, fg, bg);
+  return Surface(TTF_RenderGlyph_LCD(font, ch, fg, bg));
 }
 
 inline Surface Font::RenderGlyph_LCD(Uint32 ch, ColorRaw fg, ColorRaw bg) const
@@ -4065,7 +4029,7 @@
   }
 
   /// Destructor
-  ~TextEngine() { Destroy(m_resource); }
+  virtual ~TextEngine() = default;
 
   /// Assignment operator.
   constexpr TextEngine& operator=(TextEngine&& other) noexcept
@@ -4099,16 +4063,14 @@
   /// Converts to TextEngineParam
   constexpr operator TextEngineParam() const noexcept { return {m_resource}; }
 
-  /// frees up textEngine.
-  void Destroy() { static_assert(false, "Not implemented"); }
+  /// frees up textEngine. Pure virtual
+  virtual void Destroy() = 0;
 
   /**
    * Create a text object from UTF-8 text and a text engine.
    *
    * @param font the font to render with.
    * @param text the text to use, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @returns a Text object or nullptr on failure; call GetError() for more
    *          information.
    *
@@ -4122,6 +4084,7 @@
   Text CreateText(FontParam font, std::string_view text);
 };
 
+/// A surface based text engine
 struct SurfaceTextEngine : TextEngine
 {
   /**
@@ -4138,7 +4101,7 @@
    * @sa Text.DrawSurface
    */
   SurfaceTextEngine()
-    : T(TTF_CreateSurfaceTextEngine())
+    : TextEngine(TTF_CreateSurfaceTextEngine())
   {
   }
 
@@ -4150,9 +4113,6 @@
    * All text created by this engine should be destroyed before calling this
    * function.
    *
-   * @param engine a TextEngine object created with
-   *               SurfaceTextEngine.SurfaceTextEngine().
-   *
    * @threadsafety This function should be called on the thread that created the
    *               engine.
    *
@@ -4160,9 +4120,10 @@
    *
    * @sa SurfaceTextEngine.SurfaceTextEngine
    */
-  void Destroy();
+  void Destroy() final;
 };
 
+/// A renderer based text engine
 struct RendererTextEngine : TextEngine
 {
   /**
@@ -4182,7 +4143,7 @@
    * @sa RendererTextEngine.RendererTextEngine
    */
   RendererTextEngine(RendererParam renderer)
-    : T(TTF_CreateRendererTextEngine(renderer))
+    : TextEngine(TTF_CreateRendererTextEngine(renderer))
   {
   }
 
@@ -4211,7 +4172,7 @@
    * @sa Text.DrawRenderer
    */
   RendererTextEngine(PropertiesParam props)
-    : T(TTF_CreateRendererTextEngineWithProperties(props))
+    : TextEngine(TTF_CreateRendererTextEngineWithProperties(props))
   {
   }
 
@@ -4223,9 +4184,6 @@
    * All text created by this engine should be destroyed before calling this
    * function.
    *
-   * @param engine a TextEngine object created with
-   *               RendererTextEngine.RendererTextEngine().
-   *
    * @threadsafety This function should be called on the thread that created the
    *               engine.
    *
@@ -4233,9 +4191,10 @@
    *
    * @sa RendererTextEngine.RendererTextEngine
    */
-  void Destroy();
+  void Destroy() final;
 };
 
+/// A GPU based text engine
 struct GPUTextEngine : TextEngine
 {
   /**
@@ -4255,7 +4214,7 @@
    * @sa Text.GetGPUDrawData
    */
   GPUTextEngine(GPUDeviceParam device)
-    : T(TTF_CreateGPUTextEngine(device))
+    : TextEngine(TTF_CreateGPUTextEngine(device))
   {
   }
 
@@ -4284,7 +4243,7 @@
    * @sa Text.GetGPUDrawData
    */
   GPUTextEngine(PropertiesParam props)
-    : T(TTF_CreateGPUTextEngineWithProperties(props))
+    : TextEngine(TTF_CreateGPUTextEngineWithProperties(props))
   {
   }
 
@@ -4294,8 +4253,6 @@
    * Sets the winding order of the vertices returned by Text.GetGPUDrawData for
    * a particular GPU text engine.
    *
-   * @param engine a TextEngine object created with
-   *               GPUTextEngine.GPUTextEngine().
    * @param winding the new winding order option.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4305,15 +4262,12 @@
    *
    * @sa GPUTextEngine.GetGPUWinding
    */
-  static void SetGPUWinding(TextEngineParam engine,
-                            GPUTextEngineWinding winding);
+  void SetGPUWinding(GPUTextEngineWinding winding);
 
   /**
    * Get the winding order of the vertices returned by Text.GetGPUDrawData for a
    * particular GPU text engine
    *
-   * @param engine a TextEngine object created with
-   *               GPUTextEngine.GPUTextEngine().
    * @returns the winding order used by the GPU text engine or
    *          GPU_TEXTENGINE_WINDING_INVALID in case of error.
    *
@@ -4324,7 +4278,7 @@
    *
    * @sa GPUTextEngine.SetGPUWinding
    */
-  GPUTextEngineWinding GetGPUWinding(TextEngineParam engine) const;
+  GPUTextEngineWinding GetGPUWinding() const;
 
   /**
    * Destroy a text engine created for drawing text with the SDL GPU API.
@@ -4332,9 +4286,6 @@
    * All text created by this engine should be destroyed before calling this
    * function.
    *
-   * @param engine a TextEngine object created with
-   *               GPUTextEngine.GPUTextEngine().
-   *
    * @threadsafety This function should be called on the thread that created the
    *               engine.
    *
@@ -4342,7 +4293,7 @@
    *
    * @sa GPUTextEngine.GPUTextEngine
    */
-  void Destroy();
+  void Destroy() final;
 };
 
 /**
@@ -4432,8 +4383,6 @@
    *               nullptr.
    * @param font the font to render with.
    * @param text the text to use, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @post a Text object or nullptr on failure; call GetError() for more
    *       information.
    *
@@ -4445,7 +4394,7 @@
    * @sa Text.Destroy
    */
   Text(TextEngineParam engine, FontParam font, std::string_view text)
-    : m_resource(TTF_CreateText(engine, font, text))
+    : m_resource(TTF_CreateText(engine, font, text.data(), text.size()))
   {
   }
 
@@ -4508,11 +4457,8 @@
    * `text` must have been created using a TextEngine from
    * SurfaceTextEngine.SurfaceTextEngine().
    *
-   * @param text the text to draw.
-   * @param x the x coordinate in pixels, positive from the left edge towards
-   *          the right.
-   * @param y the y coordinate in pixels, positive from the top edge towards the
-   *          bottom.
+   * @param p the x,y coordinates in pixels, positive from the top-left edge
+   *          towards the bottom-right.
    * @param surface the surface to draw on.
    * @throws Error on failure.
    *
@@ -4533,11 +4479,8 @@
    * RendererTextEngine.RendererTextEngine(), and will draw using the renderer
    * passed to that function.
    *
-   * @param text the text to draw.
-   * @param x the x coordinate in pixels, positive from the left edge towards
-   *          the right.
-   * @param y the y coordinate in pixels, positive from the top edge towards the
-   *          bottom.
+   * @param p the x,y coordinates in pixels, positive from the top-left edge
+   *          towards the bottom-right.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4696,7 +4639,7 @@
    * This returns false if SDL_ttf isn't built with HarfBuzz support.
    *
    * @param script an [ISO 15924
-   *               code](https://unicode.org/iso15924/iso15924-codes.html) .
+   *               code](https://unicode.org/iso15924/iso15924-codes.html).
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4731,11 +4674,7 @@
    *
    * The default text color is white (255, 255, 255, 255).
    *
-   * @param text the Text to modify.
-   * @param r the red color value in the range of 0-255.
-   * @param g the green color value in the range of 0-255.
-   * @param b the blue color value in the range of 0-255.
-   * @param a the alpha value in the range of 0-255.
+   * @param c the color values in the range of 0-255.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4753,11 +4692,7 @@
    *
    * The default text color is white (1.0f, 1.0f, 1.0f, 1.0f).
    *
-   * @param text the Text to modify.
-   * @param r the red color value, normally in the range of 0-1.
-   * @param g the green color value, normally in the range of 0-1.
-   * @param b the blue color value, normally in the range of 0-1.
-   * @param a the alpha value in the range of 0-1.
+   * @param c the color value, normally in the range of 0-1.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4796,14 +4731,7 @@
   /**
    * Get the color of a text object.
    *
-   * @param r a pointer filled in with the red color value in the range of
-   *          0-255, may be nullptr.
-   * @param g a pointer filled in with the green color value in the range of
-   *          0-255, may be nullptr.
-   * @param b a pointer filled in with the blue color value in the range of
-   *          0-255, may be nullptr.
-   * @param a a pointer filled in with the alpha value in the range of 0-255,
-   *          may be nullptr.
+   * @return The color in the range of 0-255 on success.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4814,7 +4742,7 @@
    * @sa Text.GetColorFloat
    * @sa Text.SetColor
    */
-  Color GetColor();
+  Color GetColor() const;
 
   /**
    * Get the color of a text object.
@@ -4842,14 +4770,7 @@
   /**
    * Get the color of a text object.
    *
-   * @param r a pointer filled in with the red color value, normally in the
-   *          range of 0-1, may be nullptr.
-   * @param g a pointer filled in with the green color value, normally in the
-   *          range of 0-1, may be nullptr.
-   * @param b a pointer filled in with the blue color value, normally in the
-   *          range of 0-1, may be nullptr.
-   * @param a a pointer filled in with the alpha value in the range of 0-1, may
-   *          be nullptr.
+   * @return The color in the range of 0-1 on success.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4860,7 +4781,7 @@
    * @sa Text.GetColor
    * @sa Text.SetColorFloat
    */
-  FColor GetColorFloat();
+  FColor GetColorFloat() const;
 
   /**
    * Set the position of a text object.
@@ -4870,9 +4791,7 @@
    *
    * This function may cause the internal text representation to be rebuilt.
    *
-   * @param text the Text to modify.
-   * @param x the x offset of the upper left corner of this text in pixels.
-   * @param y the y offset of the upper left corner of this text in pixels.
+   * @param p the x, y offset of the upper left corner of this text in pixels.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4905,10 +4824,9 @@
   /**
    * Get the position of a text object.
    *
-   * @param x a pointer filled in with the x offset of the upper left corner of
-   *          this text in pixels, may be nullptr.
-   * @param y a pointer filled in with the y offset of the upper left corner of
-   *          this text in pixels, may be nullptr.
+   * @returns a Point with the offset of the upper left corner of this text in
+   *          pixels on success.
+   * @throws Error on failure.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4918,7 +4836,7 @@
    *
    * @sa Text.SetPosition
    */
-  Point GetPosition();
+  Point GetPosition() const;
 
   /**
    * Set whether wrapping is enabled on a text object.
@@ -4941,9 +4859,8 @@
   /**
    * Get whether wrapping is enabled on a text object.
    *
-   * @param text the Text to query.
-   * @param wrap_width a pointer filled in with the maximum width in pixels or 0
-   *                   if the text is being wrapped on newline characters.
+   * @returns an int with the maximum width in pixels or 0 if the text is being
+   *          wrapped on newline characters on success.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4998,10 +4915,7 @@
    *
    * This function may cause the internal text representation to be rebuilt.
    *
-   * @param text the Text to modify.
-   * @param string the UTF-8 text to use, may be nullptr.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
+   * @param string the UTF-8 text to use.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -5020,14 +4934,11 @@
    *
    * This function may cause the internal text representation to be rebuilt.
    *
-   * @param text the Text to modify.
    * @param offset the offset, in bytes, from the beginning of the string if >=
    *               0, the offset from the end of the string if < 0. Note that
    *               this does not do UTF-8 validation, so you should only insert
    *               at UTF-8 sequence boundaries.
    * @param string the UTF-8 text to insert.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -5046,10 +4957,7 @@
    *
    * This function may cause the internal text representation to be rebuilt.
    *
-   * @param text the Text to modify.
    * @param string the UTF-8 text to insert.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -5112,18 +5020,17 @@
    * The size of the text may change when the font or font style and size
    * change.
    *
-   * @param w a pointer filled in with the width of the text, in pixels, may be
-   *          nullptr.
-   * @param h a pointer filled in with the height of the text, in pixels, may be
-   *          nullptr.
+   * @returns Point containing the size on x and y axis on success.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
    *               text.
    *
    * @since This function is available since SDL_ttf 3.0.0.
+   *
+   * @sa GetSize(int*,int*) if might want only one of the coordinates
    */
-  Point GetSize();
+  Point GetSize() const;
 
   /**
    * Get the substring of a text object that surrounds a text offset.
@@ -5146,14 +5053,32 @@
    */
   void GetSubString(int offset, SubString* substring) const;
 
-  SubStringIterator begin() const { static_assert(false, "Not implemented"); }
+  /**
+   * Get iterator to first substring
+   */
+  SubStringIterator begin() const;
 
-  SubStringIterator end() const { static_assert(false, "Not implemented"); }
+  /**
+   * Get iterator to one past last substring
+   */
+  SubStringIterator end() const;
 
-  SubStringIterator GetSubStringForLine(int line) const
-  {
-    static_assert(false, "Not implemented");
-  }
+  /**
+   * Get iterator to substring of a text object that contains the given line.
+   *
+   * If `line` is greater than or equal to `text->num_lines` this will return an
+   * iterator equal to end().
+   *
+   * @param line a zero-based line index, in the range [0 .. text->num_lines-1].
+   * @returns iterator on success or false on failure; call GetError() for more
+   *          information.
+   *
+   * @threadsafety This function should be called on the thread that created the
+   *               text.
+   *
+   * @since This function is available since SDL_ttf 3.0.0.
+   */
+  SubStringIterator GetSubStringForLine(int line) const;
 
   /**
    * Get the substring of a text object that contains the given line.
@@ -5175,9 +5100,20 @@
    */
   void GetSubStringForLine(int line, SubString* substring) const;
 
+  /**
+   * Get all substrings of a text object.
+   *
+   * @returns a nullptr terminated array of substring pointers or nullptr on
+   *          failure; call GetError() for more information.
+   *
+   * @threadsafety This function should be called on the thread that created the
+   *               text.
+   *
+   * @since This function is available since SDL_ttf 3.0.0.
+   */
   OwnArray<SubString*> GetSubStrings() const
   {
-    static_assert(false, "Not implemented");
+    return GetSubStringsForRange(0);
   }
 
   /**
@@ -5186,8 +5122,6 @@
    * @param offset a byte offset into the text string.
    * @param length the length of the range being queried, in bytes, or -1 for
    *               the remainder of the string.
-   * @param count a pointer filled in with the number of substrings returned,
-   *              may be nullptr.
    * @returns a nullptr terminated array of substring pointers or nullptr on
    *          failure; call GetError() for more information. This is a single
    *          allocation that should be freed with free() when it is no longer
@@ -5198,22 +5132,31 @@
    *
    * @since This function is available since SDL_ttf 3.0.0.
    */
-  OwnArray<SubString*> GetSubStringsForRange(int offset, int length) const;
-
-  SubStringIterator GetSubStringForPoint(Point p) const
-  {
-    static_assert(false, "Not implemented");
-  }
+  OwnArray<SubString*> GetSubStringsForRange(int offset, int length = -1) const;
 
   /**
    * Get the portion of a text string that is closest to a point.
    *
    * This will return the closest substring of text to the given point.
    *
-   * @param text the Text to query.
-   * @param x the x coordinate relative to the left side of the text, may be
+   * @param p the coordinates relative to the top-left side of the text, may be
    *          outside the bounds of the text area.
-   * @param y the y coordinate relative to the top side of the text, may be
+   * @returns the iterator on success or false on failure; call GetError() for
+   *          more information.
+   *
+   * @threadsafety This function should be called on the thread that created the
+   *               text.
+   *
+   * @since This function is available since SDL_ttf 3.0.0.
+   */
+  SubStringIterator GetSubStringForPoint(Point p) const;
+
+  /**
+   * Get the portion of a text string that is closest to a point.
+   *
+   * This will return the closest substring of text to the given point.
+   *
+   * @param p the coordinates relative to the top-left side of the text, may be
    *          outside the bounds of the text area.
    * @param substring a pointer filled in with the closest substring of text to
    *                  the given point.
@@ -5282,10 +5225,10 @@
    * A copy of the UTF-8 string that this text object represents, useful for
    * layout, debugging and retrieving substring text
    */
-  const char* GetText() const { static_assert(false, "Not implemented"); }
+  const char* GetText() const { return m_resource->text; }
 
   /// The number of lines in the text, 0 if it's empty
-  int GetNumLines() const { static_assert(false, "Not implemented"); }
+  int GetNumLines() const { return m_resource->num_lines; }
 };
 
 /// Semi-safe reference for Text.
@@ -5300,7 +5243,7 @@
    *
    * This does not takes ownership!
    */
-  TextRef(TextParam resource) noexcept
+  TextRef(TextParam resource = nullptr) noexcept
     : Text(resource.value)
   {
   }
@@ -5327,53 +5270,75 @@
   ~TextRef() { release(); }
 };
 
+/**
+ * Iterator for substrings
+ *
+ */
 class SubStringIterator
 {
   TextRef m_text;
 
   SubString m_subString;
 
-  constexpr SubStringIterator(TextRef text) {}
-
-public:
-  constexpr SubStringIterator() {}
-
-  constexpr operator bool() const { static_assert(false, "Not implemented"); }
-
-  constexpr const SubString& operator*() const
+  constexpr SubStringIterator(TextRef text)
+    : m_text(text)
+    , m_subString(0)
   {
-    static_assert(false, "Not implemented");
   }
 
-  constexpr const SubString* operator->() const
+public:
+  /// Default constructor.
+  constexpr SubStringIterator()
+    : SubStringIterator(TextRef{})
   {
-    static_assert(false, "Not implemented");
   }
 
+  /// True if pointing to valid SubString.
+  constexpr operator bool() const { return bool(m_text); }
+
+  /// Retrieve SubString
+  constexpr const SubString& operator*() const { return m_subString; }
+
+  /// Retrieve SubString.
+  constexpr const SubString* operator->() const { return &m_subString; }
+
+  /// Comparison.
   constexpr bool operator==(const SubStringIterator& other) const
   {
-    static_assert(false, "Not implemented");
+    return m_subString.offset == other.m_subString.offset;
   }
 
+  /// Increment operator.
   constexpr SubStringIterator& operator++()
   {
-    static_assert(false, "Not implemented");
+    m_text.GetNextSubString(m_subString, &m_subString);
+    return *this;
   }
 
+  /// Increment operator.
   constexpr SubStringIterator operator++(int)
   {
-    static_assert(false, "Not implemented");
+    auto curr = *this;
+    m_text.GetNextSubString(m_subString, &m_subString);
+    return curr;
   }
 
+  /// Decrement operator.
   constexpr SubStringIterator& operator--()
   {
-    static_assert(false, "Not implemented");
+    m_text.GetPreviousSubString(m_subString, &m_subString);
+    return *this;
   }
 
+  /// Decrement operator.
   constexpr SubStringIterator operator--(int)
   {
-    static_assert(false, "Not implemented");
+    auto curr = *this;
+    m_text.GetPreviousSubString(m_subString, &m_subString);
+    return curr;
   }
+
+  friend class Text;
 };
 
 /**
@@ -5401,10 +5366,8 @@
  * SurfaceTextEngine.SurfaceTextEngine().
  *
  * @param text the text to draw.
- * @param x the x coordinate in pixels, positive from the left edge towards the
- *          right.
- * @param y the y coordinate in pixels, positive from the top edge towards the
- *          bottom.
+ * @param p the (x, y) coordinate in pixels, positive from the left edge
+ *          towards the right and from the top edge towards the bottom.
  * @param surface the surface to draw on.
  * @throws Error on failure.
  *
@@ -5418,7 +5381,7 @@
  */
 inline void DrawSurfaceText(TextConstParam text, Point p, SurfaceParam surface)
 {
-  CheckError(TTF_DrawSurfaceText(text, p, surface));
+  CheckError(TTF_DrawSurfaceText(text, p.x, p.y, surface));
 }
 
 inline void Text::DrawSurface(Point p, SurfaceParam surface) const
@@ -5510,10 +5473,6 @@
 constexpr auto RENDERER_POINTER =
   TTF_PROP_RENDERER_TEXT_ENGINE_RENDERER_POINTER;
 
-#endif // SDL_TTF_VERSION_ATLEAST(3, 2, 3)
-
-#if SDL_TTF_VERSION_ATLEAST(3, 2, 3)
-
 constexpr auto ATLAS_TEXTURE_SIZE_NUMBER =
   TTF_PROP_RENDERER_TEXT_ENGINE_ATLAS_TEXTURE_SIZE_NUMBER;
 
@@ -5529,10 +5488,8 @@
  * passed to that function.
  *
  * @param text the text to draw.
- * @param x the x coordinate in pixels, positive from the left edge towards the
- *          right.
- * @param y the y coordinate in pixels, positive from the top edge towards the
- *          bottom.
+ * @param p the (x, y) coordinate in pixels, positive from the left edge
+ *          towards the right and from the top edge towards the bottom.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -5541,11 +5498,11 @@
  * @since This function is available since SDL_ttf 3.0.0.
  *
  * @sa RendererTextEngine.RendererTextEngine
- * @sa Text.Text
+ * @sa GPUTextEngine.GPUTextEngine
  */
 inline void DrawRendererText(TextConstParam text, FPoint p)
 {
-  CheckError(TTF_DrawRendererText(text, p));
+  CheckError(TTF_DrawRendererText(text, p.x, p.y));
 }
 
 inline void Text::DrawRenderer(FPoint p) const
@@ -5635,10 +5592,6 @@
 
 constexpr auto DEVICE_POINTER = TTF_PROP_GPU_TEXT_ENGINE_DEVICE_POINTER;
 
-#endif // SDL_TTF_VERSION_ATLEAST(3, 2, 3)
-
-#if SDL_TTF_VERSION_ATLEAST(3, 2, 3)
-
 constexpr auto ATLAS_TEXTURE_SIZE_NUMBER =
   TTF_PROP_GPU_TEXT_ENGINE_ATLAS_TEXTURE_SIZE_NUMBER;
 
@@ -5670,7 +5623,7 @@
  * @since This function is available since SDL_ttf 3.0.0.
  *
  * @sa GPUTextEngine.GPUTextEngine
- * @sa Text.Text
+ * @sa GPUTextEngine.GPUTextEngine
  */
 inline GPUAtlasDrawSequence* GetGPUTextDrawData(TextConstParam text)
 {
@@ -5704,16 +5657,56 @@
 
 inline void GPUTextEngine::Destroy() { DestroyGPUTextEngine(release()); }
 
-inline void GPUTextEngine::SetGPUWinding(TextEngineParam engine,
-                                         GPUTextEngineWinding winding)
+/**
+ * Sets the winding order of the vertices returned by Text.GetGPUDrawData
+ * for a particular GPU text engine.
+ *
+ * @param engine a TextEngine object created with
+ *               GPUTextEngine.GPUTextEngine().
+ * @param winding the new winding order option.
+ *
+ * @threadsafety This function should be called on the thread that created the
+ *               engine.
+ *
+ * @since This function is available since SDL_ttf 3.0.0.
+ *
+ * @sa GPUTextEngine.GetGPUWinding
+ */
+inline void SetGPUTextEngineWinding(TextEngineParam engine,
+                                    GPUTextEngineWinding winding)
+{
+  TTF_SetGPUTextEngineWinding(engine, winding);
+}
+
+inline void GPUTextEngine::SetGPUWinding(GPUTextEngineWinding winding)
+{
+  SDL::SetGPUTextEngineWinding(get(), winding);
+}
+
+/**
+ * Get the winding order of the vertices returned by Text.GetGPUDrawData
+ * for a particular GPU text engine
+ *
+ * @param engine a TextEngine object created with
+ *               GPUTextEngine.GPUTextEngine().
+ * @returns the winding order used by the GPU text engine or
+ *          GPU_TEXTENGINE_WINDING_INVALID in case of error.
+ *
+ * @threadsafety This function should be called on the thread that created the
+ *               engine.
+ *
+ * @since This function is available since SDL_ttf 3.0.0.
+ *
+ * @sa GPUTextEngine.SetGPUWinding
+ */
+inline GPUTextEngineWinding GetGPUTextEngineWinding(TextEngineParam engine)
 {
-  SDL::SetGPUTextEngineWinding(engine, winding);
+  return TTF_GetGPUTextEngineWinding(engine);
 }
 
-inline GPUTextEngineWinding GPUTextEngine::GetGPUWinding(
-  TextEngineParam engine) const
+inline GPUTextEngineWinding GPUTextEngine::GetGPUWinding() const
 {
-  return SDL::GetGPUTextEngineWinding(engine);
+  return SDL::GetGPUTextEngineWinding(get());
 }
 
 /**
@@ -5723,8 +5716,6 @@
  *               nullptr.
  * @param font the font to render with.
  * @param text the text to use, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @returns a Text object or nullptr on failure; call GetError() for more
  *          information.
  *
@@ -5761,7 +5752,7 @@
  */
 inline PropertiesRef GetTextProperties(TextConstParam text)
 {
-  return CheckError(TTF_GetTextProperties(text));
+  return {CheckError(TTF_GetTextProperties(text))};
 }
 
 inline PropertiesRef Text::GetProperties() const
@@ -5866,7 +5857,7 @@
  */
 inline FontRef GetTextFont(TextConstParam text)
 {
-  return CheckError(TTF_GetTextFont(text));
+  return {CheckError(TTF_GetTextFont(text))};
 }
 
 inline FontRef Text::GetFont() const { return SDL::GetTextFont(m_resource); }
@@ -5926,7 +5917,7 @@
  *
  * @param text the text to modify.
  * @param script an [ISO 15924
- *               code](https://unicode.org/iso15924/iso15924-codes.html) .
+ *               code](https://unicode.org/iso15924/iso15924-codes.html).
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -5976,10 +5967,7 @@
  * The default text color is white (255, 255, 255, 255).
  *
  * @param text the Text to modify.
- * @param r the red color value in the range of 0-255.
- * @param g the green color value in the range of 0-255.
- * @param b the blue color value in the range of 0-255.
- * @param a the alpha value in the range of 0-255.
+ * @param c the color value in the range of 0-255.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -5992,7 +5980,7 @@
  */
 inline void SetTextColor(TextParam text, Color c)
 {
-  CheckError(TTF_SetTextColor(text, c));
+  CheckError(TTF_SetTextColor(text, c.r, c.g, c.b, c.a));
 }
 
 inline void Text::SetColor(Color c) { SDL::SetTextColor(m_resource, c); }
@@ -6003,10 +5991,7 @@
  * The default text color is white (1.0f, 1.0f, 1.0f, 1.0f).
  *
  * @param text the Text to modify.
- * @param r the red color value, normally in the range of 0-1.
- * @param g the green color value, normally in the range of 0-1.
- * @param b the blue color value, normally in the range of 0-1.
- * @param a the alpha value in the range of 0-1.
+ * @param c the color value, normally in the range of 0-1.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -6019,7 +6004,7 @@
  */
 inline void SetTextColorFloat(TextParam text, FColor c)
 {
-  CheckError(TTF_SetTextColorFloat(text, c));
+  CheckError(TTF_SetTextColorFloat(text, c.r, c.g, c.b, c.a));
 }
 
 inline void Text::SetColorFloat(FColor c)
@@ -6062,14 +6047,7 @@
  * Get the color of a text object.
  *
  * @param text the Text to query.
- * @param r a pointer filled in with the red color value in the range of 0-255,
- *          may be nullptr.
- * @param g a pointer filled in with the green color value in the range of
- *          0-255, may be nullptr.
- * @param b a pointer filled in with the blue color value in the range of 0-255,
- *          may be nullptr.
- * @param a a pointer filled in with the alpha value in the range of 0-255, may
- *          be nullptr.
+ * @return The color in the range of 0-255 on success.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -6082,7 +6060,9 @@
  */
 inline Color GetTextColor(TextParam text)
 {
-  static_assert(false, "Not implemented");
+  Color c;
+  GetTextColor(text, &c.r, &c.g, &c.b, &c.a);
+  return c;
 }
 
 inline void Text::GetColor(Uint8* r, Uint8* g, Uint8* b, Uint8* a) const
@@ -6090,7 +6070,7 @@
   SDL::GetTextColor(m_resource, r, g, b, a);
 }
 
-inline Color Text::GetColor() { return SDL::GetTextColor(m_resource); }
+inline Color Text::GetColor() const { return SDL::GetTextColor(m_resource); }
 
 /**
  * Get the color of a text object.
@@ -6127,14 +6107,7 @@
  * Get the color of a text object.
  *
  * @param text the Text to query.
- * @param r a pointer filled in with the red color value, normally in the range
- *          of 0-1, may be nullptr.
- * @param g a pointer filled in with the green color value, normally in the
- *          range of 0-1, may be nullptr.
- * @param b a pointer filled in with the blue color value, normally in the range
- *          of 0-1, may be nullptr.
- * @param a a pointer filled in with the alpha value in the range of 0-1, may be
- *          nullptr.
+ * @return The color in the range of 0-1 on success.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -6147,7 +6120,9 @@
  */
 inline FColor GetTextColorFloat(TextParam text)
 {
-  static_assert(false, "Not implemented");
+  FColor c;
+  GetTextColorFloat(text, &c.r, &c.g, &c.b, &c.a);
+  return c;
 }
 
 inline void Text::GetColorFloat(float* r, float* g, float* b, float* a) const
@@ -6155,7 +6130,7 @@
   SDL::GetTextColorFloat(m_resource, r, g, b, a);
 }
 
-inline FColor Text::GetColorFloat()
+inline FColor Text::GetColorFloat() const
 {
   return SDL::GetTextColorFloat(m_resource);
 }
@@ -6169,8 +6144,7 @@
  * This function may cause the internal text representation to be rebuilt.
  *
  * @param text the Text to modify.
- * @param x the x offset of the upper left corner of this text in pixels.
- * @param y the y offset of the upper left corner of this text in pixels.
+ * @param p the x, y offset of the upper left corner of this text in pixels.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -6182,7 +6156,7 @@
  */
 inline void SetTextPosition(TextParam text, Point p)
 {
-  CheckError(TTF_SetTextPosition(text, p));
+  CheckError(TTF_SetTextPosition(text, p.x, p.y));
 }
 
 inline void Text::SetPosition(Point p) { SDL::SetTextPosition(m_resource, p); }
@@ -6213,10 +6187,8 @@
  * Get the position of a text object.
  *
  * @param text the Text to query.
- * @param x a pointer filled in with the x offset of the upper left corner of
- *          this text in pixels, may be nullptr.
- * @param y a pointer filled in with the y offset of the upper left corner of
- *          this text in pixels, may be nullptr.
+ * @returns a Point with the offset of the upper left corner of this text in
+ *          pixels on success.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -6228,7 +6200,9 @@
  */
 inline Point GetTextPosition(TextParam text)
 {
-  static_assert(false, "Not implemented");
+  Point p;
+  GetTextPosition(text, &p.x, &p.y);
+  return p;
 }
 
 inline void Text::GetPosition(int* x, int* y) const
@@ -6236,7 +6210,10 @@
   SDL::GetTextPosition(m_resource, x, y);
 }
 
-inline Point Text::GetPosition() { return SDL::GetTextPosition(m_resource); }
+inline Point Text::GetPosition() const
+{
+  return SDL::GetTextPosition(m_resource);
+}
 
 /**
  * Set whether wrapping is enabled on a text object.
@@ -6269,8 +6246,8 @@
  * Get whether wrapping is enabled on a text object.
  *
  * @param text the Text to query.
- * @param wrap_width a pointer filled in with the maximum width in pixels or 0
- *                   if the text is being wrapped on newline characters.
+ * @returns maximum width in pixels or 0 if the text is being wrapped on newline
+ *          characters.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -6282,7 +6259,9 @@
  */
 inline int GetTextWrapWidth(TextConstParam text)
 {
-  return CheckError(TTF_GetTextWrapWidth(text));
+  int w;
+  CheckError(TTF_GetTextWrapWidth(text, &w));
+  return w;
 }
 
 inline int Text::GetWrapWidth() const
@@ -6350,8 +6329,6 @@
  *
  * @param text the Text to modify.
  * @param string the UTF-8 text to use, may be nullptr.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -6365,7 +6342,7 @@
  */
 inline void SetTextString(TextParam text, std::string_view string)
 {
-  CheckError(TTF_SetTextString(text, string));
+  CheckError(TTF_SetTextString(text, string.data(), string.size()));
 }
 
 inline void Text::SetString(std::string_view string)
@@ -6384,8 +6361,6 @@
  *               does not do UTF-8 validation, so you should only insert at
  *               UTF-8 sequence boundaries.
  * @param string the UTF-8 text to insert.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -6401,7 +6376,7 @@
                              int offset,
                              std::string_view string)
 {
-  CheckError(TTF_InsertTextString(text, offset, string));
+  CheckError(TTF_InsertTextString(text, offset, string.data(), string.size()));
 }
 
 inline void Text::InsertString(int offset, std::string_view string)
@@ -6416,8 +6391,6 @@
  *
  * @param text the Text to modify.
  * @param string the UTF-8 text to insert.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -6431,7 +6404,7 @@
  */
 inline void AppendTextString(TextParam text, std::string_view string)
 {
-  CheckError(TTF_AppendTextString(text, string));
+  CheckError(TTF_AppendTextString(text, string.data(), string.size()));
 }
 
 inline void Text::AppendString(std::string_view string)
@@ -6500,10 +6473,8 @@
  * The size of the text may change when the font or font style and size change.
  *
  * @param text the Text to query.
- * @param w a pointer filled in with the width of the text, in pixels, may be
- *          nullptr.
- * @param h a pointer filled in with the height of the text, in pixels, may be
- *          nullptr.
+ * @returns Point containing the size on x and y axis on success or
+ *          std::nullopt on failure; call GetError() for more information.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -6513,7 +6484,9 @@
  */
 inline Point GetTextSize(TextParam text)
 {
-  static_assert(false, "Not implemented");
+  Point p;
+  GetTextSize(text, &p.x, &p.y);
+  return p;
 }
 
 inline void Text::GetSize(int* w, int* h) const
@@ -6521,7 +6494,7 @@
   SDL::GetTextSize(m_resource, w, h);
 }
 
-inline Point Text::GetSize() { return SDL::GetTextSize(m_resource); }
+inline Point Text::GetSize() const { return SDL::GetTextSize(m_resource); }
 
 /**
  * Get the substring of a text object that surrounds a text offset.
@@ -6593,12 +6566,8 @@
  * @param offset a byte offset into the text string.
  * @param length the length of the range being queried, in bytes, or -1 for the
  *               remainder of the string.
- * @param count a pointer filled in with the number of substrings returned, may
- *              be nullptr.
  * @returns a nullptr terminated array of substring pointers or nullptr on
- *          failure; call GetError() for more information. This is a single
- *          allocation that should be freed with free() when it is no longer
- *          needed.
+ *          failure; call GetError() for more information.
  *
  * @threadsafety This function should be called on the thread that created the
  *               text.
@@ -6609,7 +6578,9 @@
                                                       int offset,
                                                       int length)
 {
-  return TTF_GetTextSubStringsForRange(text, offset, length);
+  int count = 0;
+  auto data = TTF_GetTextSubStringsForRange(text, offset, length, &count);
+  return OwnArray<SubString*>{data, size_t(count)};
 }
 
 inline OwnArray<SubString*> Text::GetSubStringsForRange(int offset,
@@ -6624,9 +6595,7 @@
  * This will return the closest substring of text to the given point.
  *
  * @param text the Text to query.
- * @param x the x coordinate relative to the left side of the text, may be
- *          outside the bounds of the text area.
- * @param y the y coordinate relative to the top side of the text, may be
+ * @param p the coordinates relative to the top-left side of the text, may be
  *          outside the bounds of the text area.
  * @param substring a pointer filled in with the closest substring of text to
  *                  the given point.
@@ -6641,7 +6610,7 @@
                                      Point p,
                                      SubString* substring)
 {
-  CheckError(TTF_GetTextSubStringForPoint(text, p, substring));
+  CheckError(TTF_GetTextSubStringForPoint(text, p.x, p.y, substring));
 }
 
 inline void Text::GetSubStringForPoint(Point p, SubString* substring) const
@@ -6737,7 +6706,7 @@
  *
  * @since This function is available since SDL_ttf 3.0.0.
  *
- * @sa Text.Text
+ * @sa RendererTextEngine.RendererTextEngine
  */
 inline void DestroyText(TextRaw text) { TTF_DestroyText(text); }
 
@@ -6773,4 +6742,6 @@
 
 } // namespace SDL
 
+#endif // defined(SDL3PP_ENABLE_TTF) || defined(SDL3PP_DOC)
+
 #endif /* SDL3PP_TTF_H_ */
