--- src/generated/SDL3pp_log.h
+++ include/SDL3pp/SDL3pp_log.h
@@ -10,7 +10,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryLog Category Log
+ * @defgroup CategoryLog Log Handling
  *
  * Simple log messages with priorities and categories. A message's LogPriority
  * signifies how important the message is. A message's LogCategory signifies
@@ -66,6 +66,14 @@
 struct LogCategory;
 
 /**
+ * @name LogPriorities
+ *
+ * The priorities assignable for a LogCategory.
+ *
+ * @{
+ */
+
+/**
  * The predefined log priorities
  *
  * @since This enum is available since SDL 3.2.0.
@@ -93,6 +101,16 @@
 
 constexpr LogPriority LOG_PRIORITY_COUNT = SDL_LOG_PRIORITY_COUNT; ///< COUNT
 
+/// @}
+
+/**
+ * @name LogCategories
+ *
+ * The logging categories. see LogCategory for more info
+ *
+ * @{
+ */
+
 /**
  * The predefined log categories
  *
@@ -101,6 +119,10 @@
  * VERBOSE level and all other categories are enabled at the ERROR level.
  *
  * @since This enum is available since SDL 3.2.0.
+ *
+ * @cat wrap-state
+ *
+ * @sa wrap-state
  */
 class LogCategory
 {
@@ -112,7 +134,7 @@
    *
    * @param category the value to be wrapped
    */
-  constexpr LogCategory(LogCategoryRaw category = {})
+  constexpr LogCategory(LogCategoryRaw category = SDL_LOG_CATEGORY_APPLICATION)
     : m_category(category)
   {
   }
@@ -122,7 +144,10 @@
    *
    * @param category the value to be wrapped
    */
-  constexpr explicit LogCategory(int category) {}
+  constexpr explicit LogCategory(int category)
+    : m_category(SDL_LogCategory(category))
+  {
+  }
 
   /**
    * Unwraps to the underlying LogCategory.
@@ -134,7 +159,6 @@
   /**
    * Set the priority of a particular log category.
    *
-   * @param category the category to assign a priority to.
    * @param priority the LogPriority to assign.
    *
    * @threadsafety It is safe to call this function from any thread.
@@ -150,7 +174,6 @@
   /**
    * Get the priority of a particular log category.
    *
-   * @param category the category to query.
    * @returns the LogPriority for the requested category.
    *
    * @threadsafety It is safe to call this function from any thread.
@@ -161,25 +184,47 @@
    */
   LogPriority GetLogPriority() const;
 
+  /**
+   * Log an unformatted message with the specified priority.
+   *
+   * @param priority the priority of the message.
+   * @param message string to output.
+   *
+   * @threadsafety It is safe to call this function from any thread.
+   *
+   * @since This function is available since SDL 3.2.0.
+   *
+   * @sa SDL::Log()
+   * @sa Log()
+   * @sa LogCritical()
+   * @sa LogDebug()
+   * @sa LogError()
+   * @sa LogInfo()
+   * @sa LogTrace()
+   * @sa LogVerbose()
+   * @sa LogWarn()
+   */
   void LogUnformatted(LogPriority priority, StringParam message) const
   {
-    static_assert(false, "Not implemented");
+    SDL_LogMessage(m_category, priority, "%s", (const char*)(message));
   }
 
   /**
-   * Log a message with the specified category and priority.
+   * Log a message with the specified priority.
    *
-   * @param category the category of the message.
    * @param priority the priority of the message.
-   * @param fmt a printf() style message format string.
-   * @param ... additional parameters matching % tokens in the **fmt** string,
-   *            if any.
+   * @param fmt a
+   * [std::format/fmt](https://en.cppreference.com/w/cpp/utility/format/spec)
+   * style message format string.
+   * @param args additional parameters matching the `{}` tokens in the format
+   * string, if any.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
    * @since This function is available since SDL 3.2.0.
    *
-   * @sa Log
+   * @cat formatted-string
+   *
    * @sa LogCategory.LogCritical
    * @sa LogCategory.LogDebug
    * @sa LogCategory.LogError
@@ -197,16 +242,17 @@
   /**
    * Log a message with LOG_PRIORITY_TRACE.
    *
-   * @param category the category of the message.
-   * @param fmt a printf() style message format string.
-   * @param ... additional parameters matching % tokens in the **fmt** string,
-   *            if any.
-   *
+   * @param fmt a
+   * [std::format/fmt](https://en.cppreference.com/w/cpp/utility/format/spec)
+   * style message format string.
+   * @param args additional parameters matching the `{}` tokens in the format
+   * string, if any.
    * @threadsafety It is safe to call this function from any thread.
    *
    * @since This function is available since SDL 3.2.0.
    *
-   * @sa Log
+   * @cat formatted-string
+   *
    * @sa LogCategory.LogCritical
    * @sa LogCategory.LogDebug
    * @sa LogCategory.LogError
@@ -223,16 +269,18 @@
   /**
    * Log a message with LOG_PRIORITY_VERBOSE.
    *
-   * @param category the category of the message.
-   * @param fmt a printf() style message format string.
-   * @param ... additional parameters matching % tokens in the **fmt** string,
-   *            if any.
+   * @param fmt a
+   * [std::format/fmt](https://en.cppreference.com/w/cpp/utility/format/spec)
+   * style message format string.
+   * @param args additional parameters matching the `{}` tokens in the format
+   * string, if any.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
    * @since This function is available since SDL 3.2.0.
    *
-   * @sa Log
+   * @cat formatted-string
+   *
    * @sa LogCategory.LogCritical
    * @sa LogCategory.LogDebug
    * @sa LogCategory.LogError
@@ -247,16 +295,18 @@
   /**
    * Log a message with LOG_PRIORITY_DEBUG.
    *
-   * @param category the category of the message.
-   * @param fmt a printf() style message format string.
-   * @param ... additional parameters matching % tokens in the **fmt** string,
-   *            if any.
+   * @param fmt a
+   * [std::format/fmt](https://en.cppreference.com/w/cpp/utility/format/spec)
+   * style message format string.
+   * @param args additional parameters matching the `{}` tokens in the format
+   * string, if any.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
    * @since This function is available since SDL 3.2.0.
    *
-   * @sa Log
+   * @cat formatted-string
+   *
    * @sa LogCategory.LogCritical
    * @sa LogCategory.LogError
    * @sa LogCategory.LogInfo
@@ -272,16 +322,18 @@
   /**
    * Log a message with LOG_PRIORITY_INFO.
    *
-   * @param category the category of the message.
-   * @param fmt a printf() style message format string.
-   * @param ... additional parameters matching % tokens in the **fmt** string,
-   *            if any.
+   * @param fmt a
+   * [std::format/fmt](https://en.cppreference.com/w/cpp/utility/format/spec)
+   * style message format string.
+   * @param args additional parameters matching the `{}` tokens in the format
+   * string, if any.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
    * @since This function is available since SDL 3.2.0.
    *
-   * @sa Log
+   * @cat formatted-string
+   *
    * @sa LogCategory.LogCritical
    * @sa LogCategory.LogDebug
    * @sa LogCategory.LogError
@@ -297,16 +349,18 @@
   /**
    * Log a message with LOG_PRIORITY_WARN.
    *
-   * @param category the category of the message.
-   * @param fmt a printf() style message format string.
-   * @param ... additional parameters matching % tokens in the **fmt** string,
-   *            if any.
+   * @param fmt a
+   * [std::format/fmt](https://en.cppreference.com/w/cpp/utility/format/spec)
+   * style message format string.
+   * @param args additional parameters matching the `{}` tokens in the format
+   * string, if any.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
    * @since This function is available since SDL 3.2.0.
    *
-   * @sa Log
+   * @cat formatted-string
+   *
    * @sa LogCategory.LogCritical
    * @sa LogCategory.LogDebug
    * @sa LogCategory.LogError
@@ -322,16 +376,18 @@
   /**
    * Log a message with LOG_PRIORITY_ERROR.
    *
-   * @param category the category of the message.
-   * @param fmt a printf() style message format string.
-   * @param ... additional parameters matching % tokens in the **fmt** string,
-   *            if any.
+   * @param fmt a
+   * [std::format/fmt](https://en.cppreference.com/w/cpp/utility/format/spec)
+   * style message format string.
+   * @param args additional parameters matching the `{}` tokens in the format
+   * string, if any.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
    * @since This function is available since SDL 3.2.0.
    *
-   * @sa Log
+   * @cat formatted-string
+   *
    * @sa LogCategory.LogCritical
    * @sa LogCategory.LogDebug
    * @sa LogCategory.LogInfo
@@ -347,23 +403,26 @@
   /**
    * Log a message with LOG_PRIORITY_CRITICAL.
    *
-   * @param category the category of the message.
-   * @param fmt a printf() style message format string.
-   * @param ... additional parameters matching % tokens in the **fmt** string,
-   *            if any.
+   * @param fmt a
+   * [std::format/fmt](https://en.cppreference.com/w/cpp/utility/format/spec)
+   * style message format string.
+   * @param args additional parameters matching the `{}` tokens in the format
+   * string, if any.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
    * @since This function is available since SDL 3.2.0.
    *
-   * @sa Log
-   * @sa LogCategory.LogDebug
-   * @sa LogCategory.LogError
-   * @sa LogCategory.LogInfo
-   * @sa LogCategory.LogMessage
-   * @sa LogUnformatted
-   * @sa LogCategory.LogTrace
-   * @sa LogCategory.LogVerbose
+   * @cat formatted-string
+   *
+   * @sa formatted-string
+   * @sa LogCategory.LogDebug()
+   * @sa LogCategory.LogError()
+   * @sa LogCategory.LogInfo()
+   * @sa LogUnformatted()
+   * @sa LogCategory.LogUnformatted()
+   * @sa LogCategory.LogTrace()
+   * @sa LogCategory.LogVerbose()
    * @sa LogCategory.LogWarn
    */
   template<class... ARGS>
@@ -521,39 +580,39 @@
 }
 
 /**
- * Log a message with LOG_CATEGORY_APPLICATION and LOG_PRIORITY_INFO.
+ * Log an unformatted message with LOG_CATEGORY_APPLICATION and
+ * LOG_PRIORITY_INFO.
  *
- * @param fmt a printf() style message format string.
- * @param ... additional parameters matching % tokens in the `fmt` string, if
- *            any.
+ * @param category the category of the message.
+ * @param priority the priority of the message.
+ * @param message string to output.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
  * @since This function is available since SDL 3.2.0.
  *
+ * @sa Log
  * @sa LogCategory.LogCritical
  * @sa LogCategory.LogDebug
  * @sa LogCategory.LogError
  * @sa LogCategory.LogInfo
  * @sa LogCategory.LogMessage
- * @sa LogUnformatted
  * @sa LogCategory.LogTrace
  * @sa LogCategory.LogVerbose
  * @sa LogCategory.LogWarn
  */
-template<class... ARGS>
-inline void Log(std::string_view fmt, ARGS&&... args)
+inline void LogUnformatted(LogCategory category,
+                           LogPriority priority,
+                           StringParam message)
 {
-  SDL_Log(fmt, args);
+  SDL_LogMessage(category, priority, "%s", static_cast<const char*>(message));
 }
 
 /**
- * Log a message with LOG_PRIORITY_TRACE.
+ * Log an unformatted message with LOG_CATEGORY_APPLICATION and
+ * LOG_PRIORITY_INFO.
  *
- * @param category the category of the message.
- * @param fmt a printf() style message format string.
- * @param ... additional parameters matching % tokens in the **fmt** string, if
- *            any.
+ * @param message string to output.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -565,30 +624,56 @@
  * @sa LogCategory.LogError
  * @sa LogCategory.LogInfo
  * @sa LogCategory.LogMessage
- * @sa LogUnformatted
  * @sa LogCategory.LogTrace
  * @sa LogCategory.LogVerbose
  * @sa LogCategory.LogWarn
  */
-template<class... ARGS>
-inline void LogTrace(LogCategory category, std::string_view fmt, ARGS&&... args)
+inline void LogUnformatted(StringParam message)
 {
-  SDL_LogTrace(category, fmt, args);
+  SDL_Log("%s", static_cast<const char*>(message));
 }
 
+/**
+ * Log a message with LOG_CATEGORY_APPLICATION and LOG_PRIORITY_INFO.
+ *
+ * @param fmt a
+ *            [std::format/fmt](https://en.cppreference.com/w/cpp/utility/format/spec)
+ *            style message format string.
+ * @param args additional parameters matching the `{}` tokens in the format
+ *             string, if any.
+ *
+ * @threadsafety It is safe to call this function from any thread.
+ *
+ * @since This function is available since SDL 3.2.0.
+ *
+ * @cat formatted-string
+ *
+ * @sa formatted-string
+ * @sa LogUnformatted()
+ * @sa LogCategory.LogCritical
+ * @sa LogCategory.LogDebug
+ * @sa LogCategory.LogError
+ * @sa LogCategory.LogInfo
+ * @sa LogCategory.LogMessage
+ * @sa LogUnformatted
+ * @sa LogCategory.LogTrace
+ * @sa LogCategory.LogVerbose
+ * @sa LogCategory.LogWarn
+ */
 template<class... ARGS>
-inline void LogCategory::LogTrace(std::string_view fmt, ARGS&&... args) const
+inline void Log(std::string_view fmt, ARGS&&... args)
 {
-  SDL::LogTrace(m_category, fmt, args);
+  LOG_CATEGORY_APPLICATION.LogInfo(fmt, std::forward<ARGS>(args)...);
 }
 
 /**
- * Log a message with LOG_PRIORITY_VERBOSE.
+ * Log a message with the specified category and priority.
  *
  * @param category the category of the message.
+ * @param priority the priority of the message.
  * @param fmt a printf() style message format string.
- * @param ... additional parameters matching % tokens in the **fmt** string, if
- *            any.
+ * @param args additional parameters matching % tokens in the **fmt** string,
+ *             if any.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -599,31 +684,36 @@
  * @sa LogCategory.LogDebug
  * @sa LogCategory.LogError
  * @sa LogCategory.LogInfo
- * @sa LogCategory.LogMessage
  * @sa LogUnformatted
+ * @sa LogCategory.LogTrace
+ * @sa LogCategory.LogVerbose
  * @sa LogCategory.LogWarn
  */
 template<class... ARGS>
-inline void LogVerbose(LogCategory category,
+inline void LogMessage(LogCategory category,
+                       LogPriority priority,
                        std::string_view fmt,
-                       ARGS&&... args)
+                       ARGS... args)
 {
-  SDL_LogVerbose(category, fmt, args);
+  LogUnformatted(
+    category, priority, std::vformat(fmt, std::make_format_args(args...)));
 }
 
 template<class... ARGS>
-inline void LogCategory::LogVerbose(std::string_view fmt, ARGS&&... args) const
+inline void LogCategory::LogMessage(LogPriority priority,
+                                    std::string_view fmt,
+                                    ARGS... args) const
 {
-  SDL::LogVerbose(m_category, fmt, args);
+  SDL::LogMessage(m_category, priority, fmt, args...);
 }
 
 /**
- * Log a message with LOG_PRIORITY_DEBUG.
+ * Log a message with LOG_PRIORITY_TRACE.
  *
  * @param category the category of the message.
  * @param fmt a printf() style message format string.
- * @param ... additional parameters matching % tokens in the **fmt** string, if
- *            any.
+ * @param args additional parameters matching % tokens in the **fmt** string,
+ *             if any.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -631,6 +721,7 @@
  *
  * @sa Log
  * @sa LogCategory.LogCritical
+ * @sa LogCategory.LogDebug
  * @sa LogCategory.LogError
  * @sa LogCategory.LogInfo
  * @sa LogCategory.LogMessage
@@ -640,24 +731,24 @@
  * @sa LogCategory.LogWarn
  */
 template<class... ARGS>
-inline void LogDebug(LogCategory category, std::string_view fmt, ARGS&&... args)
+inline void LogTrace(LogCategory category, std::string_view fmt, ARGS&&... args)
 {
-  SDL_LogDebug(category, fmt, args);
+  LogMessage(category, LOG_PRIORITY_TRACE, fmt, args...);
 }
 
 template<class... ARGS>
-inline void LogCategory::LogDebug(std::string_view fmt, ARGS&&... args) const
+inline void LogCategory::LogTrace(std::string_view fmt, ARGS&&... args) const
 {
-  SDL::LogDebug(m_category, fmt, args);
+  SDL::LogTrace(m_category, fmt, args...);
 }
 
 /**
- * Log a message with LOG_PRIORITY_INFO.
+ * Log a message with LOG_PRIORITY_VERBOSE.
  *
  * @param category the category of the message.
  * @param fmt a printf() style message format string.
- * @param ... additional parameters matching % tokens in the **fmt** string, if
- *            any.
+ * @param args additional parameters matching % tokens in the **fmt** string,
+ *             if any.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -667,31 +758,32 @@
  * @sa LogCategory.LogCritical
  * @sa LogCategory.LogDebug
  * @sa LogCategory.LogError
+ * @sa LogCategory.LogInfo
  * @sa LogCategory.LogMessage
  * @sa LogUnformatted
- * @sa LogCategory.LogTrace
- * @sa LogCategory.LogVerbose
  * @sa LogCategory.LogWarn
  */
 template<class... ARGS>
-inline void LogInfo(LogCategory category, std::string_view fmt, ARGS&&... args)
+inline void LogVerbose(LogCategory category,
+                       std::string_view fmt,
+                       ARGS&&... args)
 {
-  SDL_LogInfo(category, fmt, args);
+  LogMessage(category, LOG_PRIORITY_VERBOSE, fmt, args...);
 }
 
 template<class... ARGS>
-inline void LogCategory::LogInfo(std::string_view fmt, ARGS&&... args) const
+inline void LogCategory::LogVerbose(std::string_view fmt, ARGS&&... args) const
 {
-  SDL::LogInfo(m_category, fmt, args);
+  SDL::LogVerbose(m_category, fmt, args...);
 }
 
 /**
- * Log a message with LOG_PRIORITY_WARN.
+ * Log a message with LOG_PRIORITY_DEBUG.
  *
  * @param category the category of the message.
  * @param fmt a printf() style message format string.
- * @param ... additional parameters matching % tokens in the **fmt** string, if
- *            any.
+ * @param args additional parameters matching % tokens in the **fmt** string,
+ *             if any.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -699,33 +791,33 @@
  *
  * @sa Log
  * @sa LogCategory.LogCritical
- * @sa LogCategory.LogDebug
  * @sa LogCategory.LogError
  * @sa LogCategory.LogInfo
  * @sa LogCategory.LogMessage
  * @sa LogUnformatted
  * @sa LogCategory.LogTrace
  * @sa LogCategory.LogVerbose
+ * @sa LogCategory.LogWarn
  */
 template<class... ARGS>
-inline void LogWarn(LogCategory category, std::string_view fmt, ARGS&&... args)
+inline void LogDebug(LogCategory category, std::string_view fmt, ARGS&&... args)
 {
-  SDL_LogWarn(category, fmt, args);
+  LogMessage(category, LOG_PRIORITY_DEBUG, fmt, args...);
 }
 
 template<class... ARGS>
-inline void LogCategory::LogWarn(std::string_view fmt, ARGS&&... args) const
+inline void LogCategory::LogDebug(std::string_view fmt, ARGS&&... args) const
 {
-  SDL::LogWarn(m_category, fmt, args);
+  SDL::LogDebug(m_category, fmt, args...);
 }
 
 /**
- * Log a message with LOG_PRIORITY_ERROR.
+ * Log a message with LOG_PRIORITY_INFO.
  *
  * @param category the category of the message.
  * @param fmt a printf() style message format string.
- * @param ... additional parameters matching % tokens in the **fmt** string, if
- *            any.
+ * @param args additional parameters matching % tokens in the **fmt** string,
+ *             if any.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -734,7 +826,7 @@
  * @sa Log
  * @sa LogCategory.LogCritical
  * @sa LogCategory.LogDebug
- * @sa LogCategory.LogInfo
+ * @sa LogCategory.LogError
  * @sa LogCategory.LogMessage
  * @sa LogUnformatted
  * @sa LogCategory.LogTrace
@@ -742,30 +834,31 @@
  * @sa LogCategory.LogWarn
  */
 template<class... ARGS>
-inline void LogError(LogCategory category, std::string_view fmt, ARGS&&... args)
+inline void LogInfo(LogCategory category, std::string_view fmt, ARGS&&... args)
 {
-  SDL_LogError(category, fmt, args);
+  LogMessage(category, LOG_PRIORITY_INFO, fmt, args...);
 }
 
 template<class... ARGS>
-inline void LogCategory::LogError(std::string_view fmt, ARGS&&... args) const
+inline void LogCategory::LogInfo(std::string_view fmt, ARGS&&... args) const
 {
-  SDL::LogError(m_category, fmt, args);
+  SDL::LogInfo(m_category, fmt, args...);
 }
 
 /**
- * Log a message with LOG_PRIORITY_CRITICAL.
+ * Log a message with LOG_PRIORITY_WARN.
  *
  * @param category the category of the message.
  * @param fmt a printf() style message format string.
- * @param ... additional parameters matching % tokens in the **fmt** string, if
- *            any.
+ * @param args additional parameters matching % tokens in the **fmt** string,
+ *             if any.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
  * @since This function is available since SDL 3.2.0.
  *
  * @sa Log
+ * @sa LogCategory.LogCritical
  * @sa LogCategory.LogDebug
  * @sa LogCategory.LogError
  * @sa LogCategory.LogInfo
@@ -773,30 +866,26 @@
  * @sa LogUnformatted
  * @sa LogCategory.LogTrace
  * @sa LogCategory.LogVerbose
- * @sa LogCategory.LogWarn
  */
 template<class... ARGS>
-inline void LogCritical(LogCategory category,
-                        std::string_view fmt,
-                        ARGS&&... args)
+inline void LogWarn(LogCategory category, std::string_view fmt, ARGS&&... args)
 {
-  SDL_LogCritical(category, fmt, args);
+  LogMessage(category, SDL_LOG_PRIORITY_WARN, fmt, args...);
 }
 
 template<class... ARGS>
-inline void LogCategory::LogCritical(std::string_view fmt, ARGS&&... args) const
+inline void LogCategory::LogWarn(std::string_view fmt, ARGS&&... args) const
 {
-  SDL::LogCritical(m_category, fmt, args);
+  SDL::LogWarn(m_category, fmt, args...);
 }
 
 /**
- * Log a message with the specified category and priority.
+ * Log a message with LOG_PRIORITY_ERROR.
  *
  * @param category the category of the message.
- * @param priority the priority of the message.
  * @param fmt a printf() style message format string.
- * @param ... additional parameters matching % tokens in the **fmt** string, if
- *            any.
+ * @param args additional parameters matching % tokens in the **fmt** string,
+ *             if any.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -805,53 +894,60 @@
  * @sa Log
  * @sa LogCategory.LogCritical
  * @sa LogCategory.LogDebug
- * @sa LogCategory.LogError
  * @sa LogCategory.LogInfo
+ * @sa LogCategory.LogMessage
  * @sa LogUnformatted
  * @sa LogCategory.LogTrace
  * @sa LogCategory.LogVerbose
  * @sa LogCategory.LogWarn
  */
 template<class... ARGS>
-inline void LogMessage(LogCategory category,
-                       LogPriority priority,
-                       std::string_view fmt,
-                       ARGS... args)
+inline void LogError(LogCategory category, std::string_view fmt, ARGS&&... args)
 {
-  SDL_LogMessage(category, priority, fmt, args);
+  LogMessage(category, SDL_LOG_PRIORITY_ERROR, fmt, args...);
 }
 
 template<class... ARGS>
-inline void LogCategory::LogMessage(LogPriority priority,
-                                    std::string_view fmt,
-                                    ARGS... args) const
+inline void LogCategory::LogError(std::string_view fmt, ARGS&&... args) const
 {
-  SDL::LogMessage(m_category, priority, fmt, args);
+  SDL::LogError(m_category, fmt, args...);
 }
 
 /**
- * Log a message with the specified category and priority.
+ * Log a message with LOG_PRIORITY_CRITICAL.
  *
  * @param category the category of the message.
- * @param priority the priority of the message.
  * @param fmt a printf() style message format string.
- * @param ap a variable argument list.
+ * @param args additional parameters matching % tokens in the **fmt** string,
+ *            if any.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
  * @since This function is available since SDL 3.2.0.
  *
  * @sa Log
- * @sa LogCategory.LogCritical
  * @sa LogCategory.LogDebug
  * @sa LogCategory.LogError
  * @sa LogCategory.LogInfo
  * @sa LogCategory.LogMessage
+ * @sa LogUnformatted
  * @sa LogCategory.LogTrace
  * @sa LogCategory.LogVerbose
  * @sa LogCategory.LogWarn
  */
-inline void LogUnformatted(StringParam message) { SDL_LogMessageV(message); }
+template<class... ARGS>
+inline void LogCritical(LogCategory category,
+                        std::string_view fmt,
+                        ARGS&&... args)
+{
+  LogMessage(category, SDL_LOG_PRIORITY_CRITICAL, fmt, args...);
+}
+
+template<class... ARGS>
+inline void LogCategory::LogCritical(std::string_view fmt, ARGS&&... args) const
+{
+  SDL::LogCritical(m_category, fmt, args...);
+}
 
 /**
  * The prototype for the log output callback function.
@@ -938,6 +1034,7 @@
  *
  * @sa GetDefaultLogOutputFunction
  * @sa GetLogOutputFunction
+ * @sa ResetLogOutputFunction
  */
 inline void SetLogOutputFunction(LogOutputFunction callback, void* userdata)
 {
@@ -948,7 +1045,6 @@
  * Replace the default log output function with one of your own.
  *
  * @param callback an LogOutputFunction to call instead of the default.
- * @param userdata a pointer that is passed to `callback`.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -956,15 +1052,26 @@
  *
  * @sa GetDefaultLogOutputFunction
  * @sa GetLogOutputFunction
+ * @sa ResetLogOutputFunction
  */
 inline void SetLogOutputFunction(LogOutputCB callback)
 {
-  static_assert(false, "Not implemented");
+  SetLogOutputFunction(callback.wrapper, callback.data);
 }
 
+/**
+ * Replace the current log output function with the default one
+ *
+ * @threadsafety It is safe to call this function from any thread.
+ *
+ * @since This function is available since SDL 3.2.0.
+ *
+ * @sa GetDefaultLogOutputFunction
+ * @sa GetLogOutputFunction
+ */
 inline void ResetLogOutputFunction()
 {
-  static_assert(false, "Not implemented");
+  return SetLogOutputFunction(GetDefaultLogOutputFunction(), nullptr);
 }
 
 /// @}
