--- build/generated/SDL3pp_ttf.h
+++ include/SDL3pp/SDL3pp_ttf.h
@@ -1,17 +1,20 @@
 #ifndef SDL3PP_TTF_H_
 #define SDL3PP_TTF_H_
 
-#include <SDL3/SDL_ttf.h>
 #include "SDL3pp_gpu.h"
 #include "SDL3pp_render.h"
 #include "SDL3pp_surface.h"
 #include "SDL3pp_version.h"
 #include "SDL3pp_video.h"
 
+#if defined(SDL3PP_ENABLE_TTF) || defined(SDL3PP_DOC)
+
+#include <SDL3_ttf/SDL_ttf.h>
+
 namespace SDL {
 
 /**
- * @defgroup CategorySDLTTF Category SDLTTF
+ * @defgroup CategorySDLTTF True type font support
  *
  * Header file for SDL_ttf library
  *
@@ -54,25 +57,39 @@
 /// Safely wrap Text for non owning const parameters
 using TextConstRef = ResourceConstRef<TextRaw, TextRawConst>;
 
-/// Printable format: "%d.%d.%d", MAJOR, MINOR, MICRO
+#ifdef SDL3PP_DOC
+
+/**
+ * @name TTF version
+ * @{
+ * Printable format: "%d.%d.%d", MAJOR, MINOR, MICRO
+ */
 #define SDL_TTF_MAJOR_VERSION
 
 #define SDL_TTF_MINOR_VERSION
 
 #define SDL_TTF_MICRO_VERSION
 
-/// This is the version number macro for the current SDL_ttf version.
+///@}
+
+/**
+ * This is the version number macro for the current SDL_ttf version.
+ */
 #define SDL_TTF_VERSION                                                        \
   SDL_VERSIONNUM(                                                              \
     SDL_TTF_MAJOR_VERSION, SDL_TTF_MINOR_VERSION, SDL_TTF_MICRO_VERSION)
 
-/// This macro will evaluate to true if compiled with SDL_ttf at least X.Y.Z.
+/**
+ * This macro will evaluate to true if compiled with SDL_ttf at least X.Y.Z.
+ */
 #define SDL_TTF_VERSION_ATLEAST(X, Y, Z)                                       \
   ((SDL_TTF_MAJOR_VERSION >= X) &&                                             \
    (SDL_TTF_MAJOR_VERSION > X || SDL_TTF_MINOR_VERSION >= Y) &&                \
    (SDL_TTF_MAJOR_VERSION > X || SDL_TTF_MINOR_VERSION > Y ||                  \
     SDL_TTF_MICRO_VERSION >= Z))
 
+#endif // SDL3PP_DOC
+
 namespace TTF {
 
 /**
@@ -277,7 +294,7 @@
  *
  * The values here are chosen to match
  * [hb_direction_t](https://harfbuzz.github.io/harfbuzz-hb-common.html#hb-direction-t)
- * .
+ *
  *
  * @since This enum is available since SDL_ttf 3.0.0.
  *
@@ -363,8 +380,6 @@
    * size becomes the index of choosing which size. If the value is too high,
    * the last indexed size will be the default.
    *
-   * When done with the returned Font, use Font.Close() to dispose of it.
-   *
    * @param file path to font file.
    * @param ptsize point size to use for the newly-opened font.
    * @post a valid Font on success.
@@ -386,14 +401,12 @@
    * the last indexed size will be the default.
    *
    * If `closeio` is true, `src` will be automatically closed once the font is
-   * closed. Otherwise you should close `src` yourself after closing the font.
-   *
-   * When done with the returned Font, use Font.Close() to dispose of it.
+   * closed. Otherwise you should keep `src` open until the font is closed.
    *
    * @param src an IOStream to provide a font file's data.
+   * @param ptsize point size to use for the newly-opened font.
    * @param closeio true to close `src` when the font is closed, false to leave
    *                it open.
-   * @param ptsize point size to use for the newly-opened font.
    * @post a valid Font on success.
    * @throws Error on failure.
    *
@@ -508,8 +521,6 @@
    * The copy will be distinct from the original, but will share the font file
    * and have the same size and style as the original.
    *
-   * When done with the returned Font, use Font.Close() to dispose of it.
-   *
    * @returns a valid Font on success.
    * @throws Error on failure.
    *
@@ -871,7 +882,7 @@
    * @threadsafety This function should be called on the thread that created the
    *               font.
    *
-   * @since This function is available since SDL_ttf 3.4.0.
+   * @since This function is available since SDL_ttf 3.2.2.
    */
   int GetWeight() const;
 
@@ -1114,7 +1125,7 @@
    * This updates any Text objects using this font.
    *
    * @param script an [ISO 15924
-   *               code](https://unicode.org/iso15924/iso15924-codes.html) .
+   *               code](https://unicode.org/iso15924/iso15924-codes.html).
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -1266,8 +1277,7 @@
    *
    * @param previous_ch the previous codepoint.
    * @param ch the current codepoint.
-   * @param kerning a pointer filled in with the kerning size between the two
-   *                glyphs, in pixels, may be nullptr.
+   * @returns the kerning size between the two glyphs, in pixels.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -1277,9 +1287,26 @@
    */
   int GetGlyphKerning(Uint32 previous_ch, Uint32 ch) const;
 
+  /**
+   * Calculate the dimensions of a rendered string of UTF-8 text.
+   *
+   * This will report the width and height, in pixels, of the space that the
+   * specified string will take to fully render.
+   *
+   * @param text text to calculate, in UTF-8 encoding.
+   * @returns return a Point with the width, height.
+   * @throws Error on failure.
+   *
+   * @threadsafety This function should be called on the thread that created the
+   *               font.
+   *
+   * @since This function is available since SDL_ttf 3.0.0.
+   */
   Point GetStringSize(std::string_view text) const
   {
-    static_assert(false, "Not implemented");
+    Point p;
+    GetStringSize(text, &p.x, &p.y);
+    return p;
   }
 
   /**
@@ -1288,10 +1315,7 @@
    * This will report the width and height, in pixels, of the space that the
    * specified string will take to fully render.
    *
-   * @param font the font to query.
    * @param text text to calculate, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param w will be filled with width, in pixels, on return.
    * @param h will be filled with height, in pixels, on return.
    * @throws Error on failure.
@@ -1303,9 +1327,32 @@
    */
   void GetStringSize(std::string_view text, int* w, int* h) const;
 
+  /**
+   * Calculate the dimensions of a rendered string of UTF-8 text.
+   *
+   * This will report the width and height, in pixels, of the space that the
+   * specified string will take to fully render.
+   *
+   * Text is wrapped to multiple lines on line endings and on word boundaries if
+   * it extends beyond `wrap_width` in pixels.
+   *
+   * If wrap_width is 0, this function will only wrap on newline characters.
+   *
+   * @param text text to calculate, in UTF-8 encoding.
+   * @param wrap_width the maximum width or 0 to wrap on newline characters.
+   * @returns return a Point with the width, height on success.
+   * @throws Error on failure.
+   *
+   * @threadsafety This function should be called on the thread that created the
+   *               font.
+   *
+   * @since This function is available since SDL_ttf 3.0.0.
+   */
   Point GetStringSizeWrapped(std::string_view text, int wrap_width) const
   {
-    static_assert(false, "Not implemented");
+    Point p;
+    GetStringSizeWrapped(text, wrap_width, &p.x, &p.y);
+    return p;
   }
 
   /**
@@ -1319,10 +1366,7 @@
    *
    * If wrap_width is 0, this function will only wrap on newline characters.
    *
-   * @param font the font to query.
    * @param text text to calculate, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param wrap_width the maximum width or 0 to wrap on newline characters.
    * @param w will be filled with width, in pixels, on return.
    * @param h will be filled with height, in pixels, on return.
@@ -1346,10 +1390,7 @@
    *
    * This does not need to render the string to do this calculation.
    *
-   * @param font the font to query.
    * @param text text to calculate, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param max_width maximum width, in pixels, available for the string, or 0
    *                  for unbounded width.
    * @param measured_width a pointer filled in with the width, in pixels, of the
@@ -1385,10 +1426,7 @@
    * You can render at other quality levels with Font.RenderText_Shaded,
    * Font.RenderText_Blended, and Font.RenderText_LCD.
    *
-   * @param font the font to render with.
    * @param text text to render, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param fg the foreground color for the text.
    * @returns a new 8-bit, palettized surface, or nullptr if there was an error.
    *
@@ -1420,10 +1458,7 @@
    * You can render at other quality levels with Font.RenderText_Shaded_Wrapped,
    * Font.RenderText_Blended_Wrapped, and Font.RenderText_LCD_Wrapped.
    *
-   * @param font the font to render with.
    * @param text text to render, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param fg the foreground color for the text.
    * @param wrapLength the maximum width of the text surface or 0 to wrap on
    *                   newline characters.
@@ -1489,10 +1524,7 @@
    * You can render at other quality levels with Font.RenderText_Solid,
    * Font.RenderText_Blended, and Font.RenderText_LCD.
    *
-   * @param font the font to render with.
    * @param text text to render, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param fg the foreground color for the text.
    * @param bg the background color for the text.
    * @returns a new 8-bit, palettized surface, or nullptr if there was an error.
@@ -1525,10 +1557,7 @@
    * You can render at other quality levels with Font.RenderText_Solid_Wrapped,
    * Font.RenderText_Blended_Wrapped, and Font.RenderText_LCD_Wrapped.
    *
-   * @param font the font to render with.
    * @param text text to render, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param fg the foreground color for the text.
    * @param bg the background color for the text.
    * @param wrap_width the maximum width of the text surface or 0 to wrap on
@@ -1597,10 +1626,7 @@
    * You can render at other quality levels with Font.RenderText_Solid,
    * Font.RenderText_Shaded, and Font.RenderText_LCD.
    *
-   * @param font the font to render with.
    * @param text text to render, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param fg the foreground color for the text.
    * @returns a new 32-bit, ARGB surface, or nullptr if there was an error.
    *
@@ -1631,10 +1657,7 @@
    * You can render at other quality levels with Font.RenderText_Solid_Wrapped,
    * Font.RenderText_Shaded_Wrapped, and Font.RenderText_LCD_Wrapped.
    *
-   * @param font the font to render with.
    * @param text text to render, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param fg the foreground color for the text.
    * @param wrap_width the maximum width of the text surface or 0 to wrap on
    *                   newline characters.
@@ -1699,10 +1722,7 @@
    * You can render at other quality levels with Font.RenderText_Solid,
    * Font.RenderText_Shaded, and Font.RenderText_Blended.
    *
-   * @param font the font to render with.
    * @param text text to render, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param fg the foreground color for the text.
    * @param bg the background color for the text.
    * @returns a new 32-bit, ARGB surface, or nullptr if there was an error.
@@ -1735,10 +1755,7 @@
    * You can render at other quality levels with Font.RenderText_Solid_Wrapped,
    * Font.RenderText_Shaded_Wrapped, and Font.RenderText_Blended_Wrapped.
    *
-   * @param font the font to render with.
    * @param text text to render, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param fg the foreground color for the text.
    * @param bg the background color for the text.
    * @param wrap_width the maximum width of the text surface or 0 to wrap on
@@ -1894,7 +1911,7 @@
 }
 
 inline Font::Font(IOStreamRef src, float ptsize, bool closeio)
-  : m_resource(CheckError(TTF_OpenFontIO(src, ptsize, closeio)))
+  : m_resource(CheckError(TTF_OpenFontIO(src, closeio, ptsize)))
 {
 }
 
@@ -1911,14 +1928,14 @@
  * last indexed size will be the default.
  *
  * If `closeio` is true, `src` will be automatically closed once the font is
- * closed. Otherwise you should close `src` yourself after closing the font.
+ * closed. Otherwise you should keep `src` open until the font is closed.
  *
  * When done with the returned Font, use Font.Close() to dispose of it.
  *
  * @param src an IOStream to provide a font file's data.
+ * @param ptsize point size to use for the newly-opened font.
  * @param closeio true to close `src` when the font is closed, false to leave it
  *                open.
- * @param ptsize point size to use for the newly-opened font.
  * @returns a valid Font on success.
  * @throws Error on failure.
  *
@@ -1977,6 +1994,16 @@
  */
 inline Font OpenFontWithProperties(PropertiesRef props) { return Font(props); }
 
+/**
+ * Properties for Font.
+ *
+ * Properties prefixed with CREATE_ are used when creating a font with
+ * OpenFontWithProperties(). Poperties without the prefix can retrieved by using
+ * Font.GetProperties() on an existing font.
+ *
+ * @sa OpenFontWithProperties
+ * @sa Font.GetProperties
+ */
 namespace prop::Font {
 
 constexpr auto CREATE_FILENAME_STRING = TTF_PROP_FONT_CREATE_FILENAME_STRING;
@@ -2036,7 +2063,7 @@
  */
 inline Font CopyFont(FontRef existing_font)
 {
-  return CheckError(TTF_CopyFont(existing_font));
+  return Font(CheckError(TTF_CopyFont(existing_font)));
 }
 
 inline Font Font::Copy() const { return SDL::CopyFont(m_resource); }
@@ -2524,7 +2551,7 @@
  * @threadsafety This function should be called on the thread that created the
  *               font.
  *
- * @since This function is available since SDL_ttf 3.4.0.
+ * @since This function is available since SDL_ttf 3.2.2.
  */
 inline int GetFontWeight(FontRef font) { return TTF_GetFontWeight(font); }
 
@@ -2939,7 +2966,7 @@
  *
  * @param font the font to modify.
  * @param script an [ISO 15924
- *               code](https://unicode.org/iso15924/iso15924-codes.html) .
+ *               code](https://unicode.org/iso15924/iso15924-codes.html).
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -3068,7 +3095,7 @@
  */
 inline Surface GetGlyphImage(FontRef font, Uint32 ch, ImageType* image_type)
 {
-  return CheckError(TTF_GetGlyphImage(font, ch, image_type));
+  return Surface{CheckError(TTF_GetGlyphImage(font, ch, image_type))};
 }
 
 inline Surface Font::GetGlyphImage(Uint32 ch, ImageType* image_type) const
@@ -3098,7 +3125,8 @@
                                      Uint32 glyph_index,
                                      ImageType* image_type)
 {
-  return CheckError(TTF_GetGlyphImageForIndex(font, glyph_index, image_type));
+  return Surface(
+    CheckError(TTF_GetGlyphImageForIndex(font, glyph_index, image_type)));
 }
 
 inline Surface Font::GetGlyphImageForIndex(Uint32 glyph_index,
@@ -3162,8 +3190,7 @@
  * @param font the font to query.
  * @param previous_ch the previous codepoint.
  * @param ch the current codepoint.
- * @param kerning a pointer filled in with the kerning size between the two
- *                glyphs, in pixels, may be nullptr.
+ * @returns the kerning size between the two glyphs, in pixels.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -3173,7 +3200,8 @@
  */
 inline int GetGlyphKerning(FontRef font, Uint32 previous_ch, Uint32 ch)
 {
-  return CheckError(TTF_GetGlyphKerning(font, previous_ch, ch));
+  if (int r; TTF_GetGlyphKerning(font, previous_ch, ch, &r)) return r;
+  throw Error();
 }
 
 inline int Font::GetGlyphKerning(Uint32 previous_ch, Uint32 ch) const
@@ -3189,8 +3217,6 @@
  *
  * @param font the font to query.
  * @param text text to calculate, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param w will be filled with width, in pixels, on return.
  * @param h will be filled with height, in pixels, on return.
  * @throws Error on failure.
@@ -3202,7 +3228,7 @@
  */
 inline void GetStringSize(FontRef font, std::string_view text, int* w, int* h)
 {
-  CheckError(TTF_GetStringSize(font, text, w, h));
+  CheckError(TTF_GetStringSize(font, text.data(), text.size(), w, h));
 }
 
 inline void Font::GetStringSize(std::string_view text, int* w, int* h) const
@@ -3223,8 +3249,6 @@
  *
  * @param font the font to query.
  * @param text text to calculate, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param wrap_width the maximum width or 0 to wrap on newline characters.
  * @param w will be filled with width, in pixels, on return.
  * @param h will be filled with height, in pixels, on return.
@@ -3241,7 +3265,8 @@
                                  int* w,
                                  int* h)
 {
-  CheckError(TTF_GetStringSizeWrapped(font, text, wrap_width, w, h));
+  CheckError(
+    TTF_GetStringSizeWrapped(font, text.data(), text.size(), wrap_width, w, h));
 }
 
 inline void Font::GetStringSizeWrapped(std::string_view text,
@@ -3262,8 +3287,6 @@
  *
  * @param font the font to query.
  * @param text text to calculate, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param max_width maximum width, in pixels, available for the string, or 0 for
  *                  unbounded width.
  * @param measured_width a pointer filled in with the width, in pixels, of the
@@ -3283,8 +3306,12 @@
                           int* measured_width,
                           size_t* measured_length)
 {
-  CheckError(
-    TTF_MeasureString(font, text, max_width, measured_width, measured_length));
+  CheckError(TTF_MeasureString(font,
+                               text.data(),
+                               text.size(),
+                               max_width,
+                               measured_width,
+                               measured_length));
 }
 
 inline void Font::MeasureString(std::string_view text,
@@ -3315,8 +3342,6 @@
  *
  * @param font the font to render with.
  * @param text text to render, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param fg the foreground color for the text.
  * @returns a new 8-bit, palettized surface, or nullptr if there was an error.
  *
@@ -3333,7 +3358,7 @@
  */
 inline Surface RenderText_Solid(FontRef font, std::string_view text, Color fg)
 {
-  return TTF_RenderText_Solid(font, text, fg);
+  return Surface{TTF_RenderText_Solid(font, text.data(), text.size(), fg)};
 }
 
 inline Surface Font::RenderText_Solid(std::string_view text, Color fg) const
@@ -3358,8 +3383,6 @@
  *
  * @param font the font to render with.
  * @param text text to render, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param fg the foreground color for the text.
  * @param wrapLength the maximum width of the text surface or 0 to wrap on
  *                   newline characters.
@@ -3380,7 +3403,8 @@
                                         Color fg,
                                         int wrapLength)
 {
-  return TTF_RenderText_Solid_Wrapped(font, text, fg, wrapLength);
+  return Surface(TTF_RenderText_Solid_Wrapped(
+    font, text.data(), text.size(), fg, wrapLength));
 }
 
 inline Surface Font::RenderText_Solid_Wrapped(std::string_view text,
@@ -3419,7 +3443,7 @@
  */
 inline Surface RenderGlyph_Solid(FontRef font, Uint32 ch, ColorRaw fg)
 {
-  return TTF_RenderGlyph_Solid(font, ch, fg);
+  return Surface(TTF_RenderGlyph_Solid(font, ch, fg));
 }
 
 inline Surface Font::RenderGlyph_Solid(Uint32 ch, ColorRaw fg) const
@@ -3447,8 +3471,6 @@
  *
  * @param font the font to render with.
  * @param text text to render, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param fg the foreground color for the text.
  * @param bg the background color for the text.
  * @returns a new 8-bit, palettized surface, or nullptr if there was an error.
@@ -3468,7 +3490,7 @@
                                  Color fg,
                                  Color bg)
 {
-  return TTF_RenderText_Shaded(font, text, fg, bg);
+  return Surface(TTF_RenderText_Shaded(font, text.data(), text.size(), fg, bg));
 }
 
 inline Surface Font::RenderText_Shaded(std::string_view text,
@@ -3496,8 +3518,6 @@
  *
  * @param font the font to render with.
  * @param text text to render, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param fg the foreground color for the text.
  * @param bg the background color for the text.
  * @param wrap_width the maximum width of the text surface or 0 to wrap on
@@ -3520,7 +3540,8 @@
                                          Color bg,
                                          int wrap_width)
 {
-  return TTF_RenderText_Shaded_Wrapped(font, text, fg, bg, wrap_width);
+  return Surface(TTF_RenderText_Shaded_Wrapped(
+    font, text.data(), text.size(), fg, bg, wrap_width));
 }
 
 inline Surface Font::RenderText_Shaded_Wrapped(std::string_view text,
@@ -3565,7 +3586,7 @@
                                   ColorRaw fg,
                                   ColorRaw bg)
 {
-  return TTF_RenderGlyph_Shaded(font, ch, fg, bg);
+  return Surface(TTF_RenderGlyph_Shaded(font, ch, fg, bg));
 }
 
 inline Surface Font::RenderGlyph_Shaded(Uint32 ch,
@@ -3594,8 +3615,6 @@
  *
  * @param font the font to render with.
  * @param text text to render, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param fg the foreground color for the text.
  * @returns a new 32-bit, ARGB surface, or nullptr if there was an error.
  *
@@ -3611,7 +3630,7 @@
  */
 inline Surface RenderText_Blended(FontRef font, std::string_view text, Color fg)
 {
-  return TTF_RenderText_Blended(font, text, fg);
+  return Surface(TTF_RenderText_Blended(font, text.data(), text.size(), fg));
 }
 
 inline Surface Font::RenderText_Blended(std::string_view text, Color fg) const
@@ -3636,8 +3655,6 @@
  *
  * @param font the font to render with.
  * @param text text to render, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param fg the foreground color for the text.
  * @param wrap_width the maximum width of the text surface or 0 to wrap on
  *                   newline characters.
@@ -3658,7 +3675,8 @@
                                           Color fg,
                                           int wrap_width)
 {
-  return TTF_RenderText_Blended_Wrapped(font, text, fg, wrap_width);
+  return Surface(TTF_RenderText_Blended_Wrapped(
+    font, text.data(), text.size(), fg, wrap_width));
 }
 
 inline Surface Font::RenderText_Blended_Wrapped(std::string_view text,
@@ -3697,7 +3715,7 @@
  */
 inline Surface RenderGlyph_Blended(FontRef font, Uint32 ch, ColorRaw fg)
 {
-  return TTF_RenderGlyph_Blended(font, ch, fg);
+  return Surface(TTF_RenderGlyph_Blended(font, ch, fg));
 }
 
 inline Surface Font::RenderGlyph_Blended(Uint32 ch, ColorRaw fg) const
@@ -3724,8 +3742,6 @@
  *
  * @param font the font to render with.
  * @param text text to render, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param fg the foreground color for the text.
  * @param bg the background color for the text.
  * @returns a new 32-bit, ARGB surface, or nullptr if there was an error.
@@ -3745,7 +3761,7 @@
                               Color fg,
                               Color bg)
 {
-  return TTF_RenderText_LCD(font, text, fg, bg);
+  return Surface(TTF_RenderText_LCD(font, text.data(), text.size(), fg, bg));
 }
 
 inline Surface Font::RenderText_LCD(std::string_view text,
@@ -3772,8 +3788,6 @@
  *
  * @param font the font to render with.
  * @param text text to render, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param fg the foreground color for the text.
  * @param bg the background color for the text.
  * @param wrap_width the maximum width of the text surface or 0 to wrap on
@@ -3796,7 +3810,8 @@
                                       Color bg,
                                       int wrap_width)
 {
-  return TTF_RenderText_LCD_Wrapped(font, text, fg, bg, wrap_width);
+  return Surface(TTF_RenderText_LCD_Wrapped(
+    font, text.data(), text.size(), fg, bg, wrap_width));
 }
 
 inline Surface Font::RenderText_LCD_Wrapped(std::string_view text,
@@ -3841,7 +3856,7 @@
                                ColorRaw fg,
                                ColorRaw bg)
 {
-  return TTF_RenderGlyph_LCD(font, ch, fg, bg);
+  return Surface(TTF_RenderGlyph_LCD(font, ch, fg, bg));
 }
 
 inline Surface Font::RenderGlyph_LCD(Uint32 ch, ColorRaw fg, ColorRaw bg) const
@@ -3941,7 +3956,7 @@
   }
 
   /// Destructor
-  ~TextEngine() { Destroy(m_resource); }
+  virtual ~TextEngine() { SDL_assert_paranoid(!m_resource); }
 
   /// Assignment operator.
   constexpr TextEngine& operator=(TextEngine&& other) noexcept
@@ -3970,16 +3985,14 @@
   /// Converts to bool
   constexpr explicit operator bool() const noexcept { return !!m_resource; }
 
-  /// frees up textEngine.
-  void Destroy() { static_assert(false, "Not implemented"); }
+  /// frees up textEngine. Pure virtual
+  virtual void Destroy() = 0;
 
   /**
    * Create a text object from UTF-8 text and a text engine.
    *
    * @param font the font to render with.
    * @param text the text to use, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @returns a Text object or nullptr on failure; call GetError() for more
    *          information.
    *
@@ -3993,6 +4006,7 @@
   Text CreateText(FontRef font, std::string_view text);
 };
 
+/// A surface based text engine
 struct SurfaceTextEngine : TextEngine
 {
   /**
@@ -4010,7 +4024,11 @@
    */
   SurfaceTextEngine();
 
-  ~SurfaceTextEngine() { Destroy(); }
+  SurfaceTextEngine(const SurfaceTextEngine&) = delete;
+
+  SurfaceTextEngine& operator=(const SurfaceTextEngine&) = delete;
+
+  ~SurfaceTextEngine() final { Destroy(); }
 
   /**
    * Destroy a text engine created for drawing text on SDL surfaces.
@@ -4018,8 +4036,6 @@
    * All text created by this engine should be destroyed before calling this
    * function.
    *
-   * @param engine a TextEngine object created with CreateSurfaceTextEngine().
-   *
    * @threadsafety This function should be called on the thread that created the
    *               engine.
    *
@@ -4027,9 +4043,10 @@
    *
    * @sa CreateSurfaceTextEngine
    */
-  void Destroy();
+  void Destroy() final;
 };
 
+/// A renderer based text engine
 struct RendererTextEngine : TextEngine
 {
   /**
@@ -4056,10 +4073,10 @@
    *
    * These are the supported properties:
    *
-   * - `prop::RendererTextEngine.RENDERER`: the renderer to use for creating
-   *   textures and drawing text
-   * - `prop::RendererTextEngine.ATLAS_TEXTURE_SIZE`: the size of the texture
-   *   atlas
+   * - `prop::RendererTextEngine.RENDERER_POINTER`: the renderer to use for
+   *   creating textures and drawing text
+   * - `prop::RendererTextEngine.ATLAS_TEXTURE_SIZE_NUMBER`: the size of the
+   *   texture atlas
    *
    * @param props the properties to use.
    * @post a TextEngine object or nullptr on failure; call GetError() for more
@@ -4076,7 +4093,11 @@
    */
   RendererTextEngine(PropertiesRef props);
 
-  ~RendererTextEngine() { Destroy(); }
+  RendererTextEngine(const RendererTextEngine&) = delete;
+
+  RendererTextEngine& operator=(const RendererTextEngine&) = delete;
+
+  ~RendererTextEngine() final { Destroy(); }
 
   /**
    * Destroy a text engine created for drawing text on an SDL renderer.
@@ -4084,8 +4105,6 @@
    * All text created by this engine should be destroyed before calling this
    * function.
    *
-   * @param engine a TextEngine object created with CreateRendererTextEngine().
-   *
    * @threadsafety This function should be called on the thread that created the
    *               engine.
    *
@@ -4093,9 +4112,10 @@
    *
    * @sa CreateRendererTextEngine
    */
-  void Destroy();
+  void Destroy() final;
 };
 
+/// A GPU based text engine
 struct GPUTextEngine : TextEngine
 {
   /**
@@ -4122,9 +4142,10 @@
    *
    * These are the supported properties:
    *
-   * - `prop::GpuTextEngine.DEVICE`: the GPUDevice to use for creating textures
-   *   and drawing text.
-   * - `prop::GpuTextEngine.ATLAS_TEXTURE_SIZE`: the size of the texture atlas
+   * - `prop::GpuTextEngine.DEVICE_POINTER`: the GPUDevice to use for creating
+   *   textures and drawing text.
+   * - `prop::GpuTextEngine.ATLAS_TEXTURE_SIZE_NUMBER`: the size of the texture
+   *   atlas
    *
    * @param props the properties to use.
    * @post a TextEngine object or nullptr on failure; call GetError() for more
@@ -4141,13 +4162,16 @@
    */
   GPUTextEngine(PropertiesRef props);
 
-  ~GPUTextEngine() { Destroy(); }
+  GPUTextEngine(const GPUTextEngine&) = delete;
+
+  GPUTextEngine& operator=(const GPUTextEngine&) = delete;
+
+  ~GPUTextEngine() final { Destroy(); }
 
   /**
    * Sets the winding order of the vertices returned by Text.GetGPUDrawData for
    * a particular GPU text engine.
    *
-   * @param engine a TextEngine object created with CreateGPUTextEngine().
    * @param winding the new winding order option.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4157,13 +4181,12 @@
    *
    * @sa GPUTextEngine.GetGPUWinding
    */
-  static void SetGPUWinding(TextEngineRef engine, GPUTextEngineWinding winding);
+  void SetGPUWinding(GPUTextEngineWinding winding);
 
   /**
    * Get the winding order of the vertices returned by Text.GetGPUDrawData for a
    * particular GPU text engine
    *
-   * @param engine a TextEngine object created with CreateGPUTextEngine().
    * @returns the winding order used by the GPU text engine or
    *          GPU_TEXTENGINE_WINDING_INVALID in case of error.
    *
@@ -4174,7 +4197,7 @@
    *
    * @sa GPUTextEngine.SetGPUWinding
    */
-  GPUTextEngineWinding GetGPUWinding(TextEngineRef engine) const;
+  GPUTextEngineWinding GetGPUWinding() const;
 
   /**
    * Destroy a text engine created for drawing text with the SDL GPU API.
@@ -4182,8 +4205,6 @@
    * All text created by this engine should be destroyed before calling this
    * function.
    *
-   * @param engine a TextEngine object created with CreateGPUTextEngine().
-   *
    * @threadsafety This function should be called on the thread that created the
    *               engine.
    *
@@ -4191,7 +4212,7 @@
    *
    * @sa CreateGPUTextEngine
    */
-  void Destroy();
+  void Destroy() final;
 };
 
 /**
@@ -4281,8 +4302,6 @@
    *               nullptr.
    * @param font the font to render with.
    * @param text the text to use, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @post a Text object or nullptr on failure; call GetError() for more
    *       information.
    *
@@ -4352,11 +4371,8 @@
    * `text` must have been created using a TextEngine from
    * CreateSurfaceTextEngine().
    *
-   * @param text the text to draw.
-   * @param x the x coordinate in pixels, positive from the left edge towards
-   *          the right.
-   * @param y the y coordinate in pixels, positive from the top edge towards the
-   *          bottom.
+   * @param p the x,y coordinates in pixels, positive from the top-left edge
+   *          towards the bottom-right.
    * @param surface the surface to draw on.
    * @throws Error on failure.
    *
@@ -4377,11 +4393,8 @@
    * CreateRendererTextEngine(), and will draw using the renderer passed to that
    * function.
    *
-   * @param text the text to draw.
-   * @param x the x coordinate in pixels, positive from the left edge towards
-   *          the right.
-   * @param y the y coordinate in pixels, positive from the top edge towards the
-   *          bottom.
+   * @param p the x,y coordinates in pixels, positive from the top-left edge
+   *          towards the bottom-right.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4540,7 +4553,7 @@
    * This returns false if SDL_ttf isn't built with HarfBuzz support.
    *
    * @param script an [ISO 15924
-   *               code](https://unicode.org/iso15924/iso15924-codes.html) .
+   *               code](https://unicode.org/iso15924/iso15924-codes.html).
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4575,11 +4588,7 @@
    *
    * The default text color is white (255, 255, 255, 255).
    *
-   * @param text the Text to modify.
-   * @param r the red color value in the range of 0-255.
-   * @param g the green color value in the range of 0-255.
-   * @param b the blue color value in the range of 0-255.
-   * @param a the alpha value in the range of 0-255.
+   * @param c the color values in the range of 0-255.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4597,11 +4606,7 @@
    *
    * The default text color is white (1.0f, 1.0f, 1.0f, 1.0f).
    *
-   * @param text the Text to modify.
-   * @param r the red color value, normally in the range of 0-1.
-   * @param g the green color value, normally in the range of 0-1.
-   * @param b the blue color value, normally in the range of 0-1.
-   * @param a the alpha value in the range of 0-1.
+   * @param c the color value, normally in the range of 0-1.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4640,14 +4645,7 @@
   /**
    * Get the color of a text object.
    *
-   * @param r a pointer filled in with the red color value in the range of
-   *          0-255, may be nullptr.
-   * @param g a pointer filled in with the green color value in the range of
-   *          0-255, may be nullptr.
-   * @param b a pointer filled in with the blue color value in the range of
-   *          0-255, may be nullptr.
-   * @param a a pointer filled in with the alpha value in the range of 0-255,
-   *          may be nullptr.
+   * @return The color in the range of 0-255 on success.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4658,7 +4656,7 @@
    * @sa Text.GetColorFloat
    * @sa Text.SetColor
    */
-  Color GetColor();
+  Color GetColor() const;
 
   /**
    * Get the color of a text object.
@@ -4686,14 +4684,7 @@
   /**
    * Get the color of a text object.
    *
-   * @param r a pointer filled in with the red color value, normally in the
-   *          range of 0-1, may be nullptr.
-   * @param g a pointer filled in with the green color value, normally in the
-   *          range of 0-1, may be nullptr.
-   * @param b a pointer filled in with the blue color value, normally in the
-   *          range of 0-1, may be nullptr.
-   * @param a a pointer filled in with the alpha value in the range of 0-1, may
-   *          be nullptr.
+   * @return The color in the range of 0-1 on success.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4704,7 +4695,7 @@
    * @sa Text.GetColor
    * @sa Text.SetColorFloat
    */
-  FColor GetColorFloat();
+  FColor GetColorFloat() const;
 
   /**
    * Set the position of a text object.
@@ -4714,8 +4705,8 @@
    *
    * This function may cause the internal text representation to be rebuilt.
    *
-   * @param x the x offset of the upper left corner of this text in pixels.
-   * @param y the y offset of the upper left corner of this text in pixels.
+   * @param p the x, y offset of the upper left corner of this text in pixels.
+   * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
    *               text.
@@ -4733,6 +4724,7 @@
    *          this text in pixels, may be nullptr.
    * @param y a pointer filled in with the y offset of the upper left corner of
    *          this text in pixels, may be nullptr.
+   * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
    *               text.
@@ -4746,10 +4738,9 @@
   /**
    * Get the position of a text object.
    *
-   * @param x a pointer filled in with the x offset of the upper left corner of
-   *          this text in pixels, may be nullptr.
-   * @param y a pointer filled in with the y offset of the upper left corner of
-   *          this text in pixels, may be nullptr.
+   * @returns a Point with the offset of the upper left corner of this text in
+   *          pixels on success.
+   * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
    *               text.
@@ -4758,7 +4749,7 @@
    *
    * @sa Text.SetPosition
    */
-  Point GetPosition();
+  Point GetPosition() const;
 
   /**
    * Set whether wrapping is enabled on a text object.
@@ -4781,9 +4772,8 @@
   /**
    * Get whether wrapping is enabled on a text object.
    *
-   * @param text the Text to query.
-   * @param wrap_width a pointer filled in with the maximum width in pixels or 0
-   *                   if the text is being wrapped on newline characters.
+   * @returns an int with the maximum width in pixels or 0 if the text is being
+   *          wrapped on newline characters on success.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4838,10 +4828,7 @@
    *
    * This function may cause the internal text representation to be rebuilt.
    *
-   * @param text the Text to modify.
-   * @param string the UTF-8 text to use, may be nullptr.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
+   * @param string the UTF-8 text to use.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4860,14 +4847,11 @@
    *
    * This function may cause the internal text representation to be rebuilt.
    *
-   * @param text the Text to modify.
    * @param offset the offset, in bytes, from the beginning of the string if >=
    *               0, the offset from the end of the string if < 0. Note that
    *               this does not do UTF-8 validation, so you should only insert
    *               at UTF-8 sequence boundaries.
    * @param string the UTF-8 text to insert.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4886,10 +4870,7 @@
    *
    * This function may cause the internal text representation to be rebuilt.
    *
-   * @param text the Text to modify.
    * @param string the UTF-8 text to insert.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4952,18 +4933,17 @@
    * The size of the text may change when the font or font style and size
    * change.
    *
-   * @param w a pointer filled in with the width of the text, in pixels, may be
-   *          nullptr.
-   * @param h a pointer filled in with the height of the text, in pixels, may be
-   *          nullptr.
+   * @returns Point containing the size on x and y axis on success.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
    *               text.
    *
    * @since This function is available since SDL_ttf 3.0.0.
+   *
+   * @sa GetSize(int*,int*) if might want only one of the coordinates
    */
-  Point GetSize();
+  Point GetSize() const;
 
   /**
    * Get the substring of a text object that surrounds a text offset.
@@ -4986,14 +4966,32 @@
    */
   void GetSubString(int offset, SubString* substring) const;
 
-  SubStringIterator begin() const { static_assert(false, "Not implemented"); }
+  /**
+   * Get iterator to first substring
+   */
+  SubStringIterator begin() const;
 
-  SubStringIterator end() const { static_assert(false, "Not implemented"); }
+  /**
+   * Get iterator to one past last substring
+   */
+  SubStringIterator end() const;
 
-  SubStringIterator GetSubStringForLine(int line) const
-  {
-    static_assert(false, "Not implemented");
-  }
+  /**
+   * Get iterator to substring of a text object that contains the given line.
+   *
+   * If `line` is greater than or equal to `text->num_lines` this will return an
+   * iterator equal to end().
+   *
+   * @param line a zero-based line index, in the range [0 .. text->num_lines-1].
+   * @returns iterator on success or false on failure; call GetError() for more
+   *          information.
+   *
+   * @threadsafety This function should be called on the thread that created the
+   *               text.
+   *
+   * @since This function is available since SDL_ttf 3.0.0.
+   */
+  SubStringIterator GetSubStringForLine(int line) const;
 
   /**
    * Get the substring of a text object that contains the given line.
@@ -5015,9 +5013,20 @@
    */
   void GetSubStringForLine(int line, SubString* substring) const;
 
+  /**
+   * Get all substrings of a text object.
+   *
+   * @returns a nullptr terminated array of substring pointers or nullptr on
+   *          failure; call GetError() for more information.
+   *
+   * @threadsafety This function should be called on the thread that created the
+   *               text.
+   *
+   * @since This function is available since SDL_ttf 3.0.0.
+   */
   OwnArray<SubString*> GetSubStrings() const
   {
-    static_assert(false, "Not implemented");
+    return GetSubStringsForRange(0);
   }
 
   /**
@@ -5026,8 +5035,6 @@
    * @param offset a byte offset into the text string.
    * @param length the length of the range being queried, in bytes, or -1 for
    *               the remainder of the string.
-   * @param count a pointer filled in with the number of substrings returned,
-   *              may be nullptr.
    * @returns a nullptr terminated array of substring pointers or nullptr on
    *          failure; call GetError() for more information. This is a single
    *          allocation that should be freed with free() when it is no longer
@@ -5038,22 +5045,31 @@
    *
    * @since This function is available since SDL_ttf 3.0.0.
    */
-  OwnArray<SubString*> GetSubStringsForRange(int offset, int length) const;
-
-  SubStringIterator GetSubStringForPoint(Point p) const
-  {
-    static_assert(false, "Not implemented");
-  }
+  OwnArray<SubString*> GetSubStringsForRange(int offset, int length = -1) const;
 
   /**
    * Get the portion of a text string that is closest to a point.
    *
    * This will return the closest substring of text to the given point.
    *
-   * @param text the Text to query.
-   * @param x the x coordinate relative to the left side of the text, may be
+   * @param p the coordinates relative to the top-left side of the text, may be
    *          outside the bounds of the text area.
-   * @param y the y coordinate relative to the top side of the text, may be
+   * @returns the iterator on success or false on failure; call GetError() for
+   *          more information.
+   *
+   * @threadsafety This function should be called on the thread that created the
+   *               text.
+   *
+   * @since This function is available since SDL_ttf 3.0.0.
+   */
+  SubStringIterator GetSubStringForPoint(Point p) const;
+
+  /**
+   * Get the portion of a text string that is closest to a point.
+   *
+   * This will return the closest substring of text to the given point.
+   *
+   * @param p the coordinates relative to the top-left side of the text, may be
    *          outside the bounds of the text area.
    * @param substring a pointer filled in with the closest substring of text to
    *                  the given point.
@@ -5073,6 +5089,8 @@
    * substring with the SUBSTRING_TEXT_START flag set.
    *
    * @param substring the SubString to query.
+   * @param previous a pointer filled in with the previous substring in the text
+   *                 object.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -5120,10 +5138,10 @@
    * A copy of the UTF-8 string that this text object represents, useful for
    * layout, debugging and retrieving substring text
    */
-  const char* GetText() const { static_assert(false, "Not implemented"); }
+  const char* GetText() const { return m_resource->text; }
 
   /// The number of lines in the text, 0 if it's empty
-  int GetNumLines() const { static_assert(false, "Not implemented"); }
+  int GetNumLines() const { return m_resource->num_lines; }
 };
 
 /**
@@ -5198,53 +5216,75 @@
   constexpr operator TextRaw() const noexcept { return get(); }
 };
 
+/**
+ * Iterator for substrings
+ *
+ */
 class SubStringIterator
 {
   TextRef m_text;
 
   SubString m_subString;
 
-  constexpr SubStringIterator(TextRef text) {}
-
-public:
-  constexpr SubStringIterator() {}
-
-  constexpr operator bool() const { static_assert(false, "Not implemented"); }
-
-  constexpr const SubString& operator*() const
+  SubStringIterator(TextRef text)
+    : m_text(text)
+    , m_subString()
   {
-    static_assert(false, "Not implemented");
   }
 
-  constexpr const SubString* operator->() const
+public:
+  /// Default constructor.
+  SubStringIterator()
+    : SubStringIterator(TextRef{})
   {
-    static_assert(false, "Not implemented");
   }
 
+  /// True if pointing to valid SubString.
+  constexpr operator bool() const { return bool(m_text); }
+
+  /// Retrieve SubString
+  constexpr const SubString& operator*() const { return m_subString; }
+
+  /// Retrieve SubString.
+  constexpr const SubString* operator->() const { return &m_subString; }
+
+  /// Comparison.
   constexpr bool operator==(const SubStringIterator& other) const
   {
-    static_assert(false, "Not implemented");
+    return m_subString.offset == other.m_subString.offset;
   }
 
-  constexpr SubStringIterator& operator++()
+  /// Increment operator.
+  SubStringIterator& operator++()
   {
-    static_assert(false, "Not implemented");
+    m_text.GetNextSubString(m_subString, &m_subString);
+    return *this;
   }
 
-  constexpr SubStringIterator operator++(int)
+  /// Increment operator.
+  SubStringIterator operator++(int)
   {
-    static_assert(false, "Not implemented");
+    auto curr = *this;
+    m_text.GetNextSubString(m_subString, &m_subString);
+    return curr;
   }
 
-  constexpr SubStringIterator& operator--()
+  /// Decrement operator.
+  SubStringIterator& operator--()
   {
-    static_assert(false, "Not implemented");
+    m_text.GetPreviousSubString(m_subString, &m_subString);
+    return *this;
   }
 
-  constexpr SubStringIterator operator--(int)
+  /// Decrement operator.
+  SubStringIterator operator--(int)
   {
-    static_assert(false, "Not implemented");
+    auto curr = *this;
+    m_text.GetPreviousSubString(m_subString, &m_subString);
+    return curr;
   }
+
+  friend class Text;
 };
 
 /**
@@ -5266,7 +5306,7 @@
 }
 
 inline SurfaceTextEngine::SurfaceTextEngine()
-  : T(TTF_CreateSurfaceTextEngine())
+  : TextEngine(TTF_CreateSurfaceTextEngine())
 {
 }
 
@@ -5277,10 +5317,8 @@
  * CreateSurfaceTextEngine().
  *
  * @param text the text to draw.
- * @param x the x coordinate in pixels, positive from the left edge towards the
- *          right.
- * @param y the y coordinate in pixels, positive from the top edge towards the
- *          bottom.
+ * @param p the (x, y) coordinate in pixels, positive from the left edge
+ *          towards the right and from the top edge towards the bottom.
  * @param surface the surface to draw on.
  * @throws Error on failure.
  *
@@ -5294,7 +5332,7 @@
  */
 inline void DrawSurfaceText(TextConstRef text, Point p, SurfaceRef surface)
 {
-  CheckError(TTF_DrawSurfaceText(text, p, surface));
+  CheckError(TTF_DrawSurfaceText(text, p.x, p.y, surface));
 }
 
 inline void Text::DrawSurface(Point p, SurfaceRef surface) const
@@ -5349,12 +5387,12 @@
 }
 
 inline RendererTextEngine::RendererTextEngine(RendererRef renderer)
-  : T(TTF_CreateRendererTextEngine(renderer))
+  : TextEngine(TTF_CreateRendererTextEngine(renderer))
 {
 }
 
 inline RendererTextEngine::RendererTextEngine(PropertiesRef props)
-  : T(TTF_CreateRendererTextEngineWithProperties(props))
+  : TextEngine(TTF_CreateRendererTextEngineWithProperties(props))
 {
 }
 
@@ -5364,10 +5402,10 @@
  *
  * These are the supported properties:
  *
- * - `prop::RendererTextEngine.RENDERER`: the renderer to use for creating
- *   textures and drawing text
- * - `prop::RendererTextEngine.ATLAS_TEXTURE_SIZE`: the size of the texture
- *   atlas
+ * - `prop::RendererTextEngine.RENDERER_POINTER`: the renderer to use for
+ *   creating textures and drawing text
+ * - `prop::RendererTextEngine.ATLAS_TEXTURE_SIZE_NUMBER`: the size of the
+ *   texture atlas
  *
  * @param props the properties to use.
  * @returns a TextEngine object or nullptr on failure; call GetError() for more
@@ -5388,11 +5426,19 @@
   return RendererTextEngine(props);
 }
 
+/**
+ * Properties for RendererTextEngine.
+ *
+ * Use these with CreateRendererTextEngineWithProperties() to specify properties
+ * when creating a renderer text engine.
+ *
+ * @sa CreateRendererTextEngineWithProperties
+ */
 namespace prop::RendererTextEngine {
 
-constexpr auto RENDERER = TTF_PROP_RENDERER_TEXT_ENGINE_RENDERER;
+constexpr auto RENDERER_POINTER = TTF_PROP_RENDERER_TEXT_ENGINE_RENDERER;
 
-constexpr auto ATLAS_TEXTURE_SIZE =
+constexpr auto ATLAS_TEXTURE_SIZE_NUMBER =
   TTF_PROP_RENDERER_TEXT_ENGINE_ATLAS_TEXTURE_SIZE;
 
 } // namespace prop::RendererTextEngine
@@ -5405,10 +5451,8 @@
  * function.
  *
  * @param text the text to draw.
- * @param x the x coordinate in pixels, positive from the left edge towards the
- *          right.
- * @param y the y coordinate in pixels, positive from the top edge towards the
- *          bottom.
+ * @param p the (x, y) coordinate in pixels, positive from the left edge
+ *          towards the right and from the top edge towards the bottom.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -5421,7 +5465,7 @@
  */
 inline void DrawRendererText(TextConstRef text, FPoint p)
 {
-  CheckError(TTF_DrawRendererText(text, p));
+  CheckError(TTF_DrawRendererText(text, p.x, p.y));
 }
 
 inline void Text::DrawRenderer(FPoint p) const
@@ -5476,12 +5520,12 @@
 }
 
 inline GPUTextEngine::GPUTextEngine(GPUDeviceRef device)
-  : T(TTF_CreateGPUTextEngine(device))
+  : TextEngine(TTF_CreateGPUTextEngine(device))
 {
 }
 
 inline GPUTextEngine::GPUTextEngine(PropertiesRef props)
-  : T(TTF_CreateGPUTextEngineWithProperties(props))
+  : TextEngine(TTF_CreateGPUTextEngineWithProperties(props))
 {
 }
 
@@ -5491,9 +5535,10 @@
  *
  * These are the supported properties:
  *
- * - `prop::GpuTextEngine.DEVICE`: the GPUDevice to use for creating textures
- *   and drawing text.
- * - `prop::GpuTextEngine.ATLAS_TEXTURE_SIZE`: the size of the texture atlas
+ * - `prop::GpuTextEngine.DEVICE_POINTER`: the GPUDevice to use for creating
+ *   textures and drawing text.
+ * - `prop::GpuTextEngine.ATLAS_TEXTURE_SIZE_NUMBER`: the size of the texture
+ *   atlas
  *
  * @param props the properties to use.
  * @returns a TextEngine object or nullptr on failure; call GetError() for more
@@ -5513,11 +5558,20 @@
   return GPUTextEngine(props);
 }
 
+/**
+ * Properties for GPUTextEngine.
+ *
+ * Use these with CreateGPUTextEngineWithProperties() to specify properties when
+ * creating a GPU text engine.
+ *
+ * @sa CreateGPUTextEngineWithProperties
+ */
 namespace prop::GpuTextEngine {
 
-constexpr auto DEVICE = TTF_PROP_GPU_TEXT_ENGINE_DEVICE;
+constexpr auto DEVICE_POINTER = TTF_PROP_GPU_TEXT_ENGINE_DEVICE;
 
-constexpr auto ATLAS_TEXTURE_SIZE = TTF_PROP_GPU_TEXT_ENGINE_ATLAS_TEXTURE_SIZE;
+constexpr auto ATLAS_TEXTURE_SIZE_NUMBER =
+  TTF_PROP_GPU_TEXT_ENGINE_ATLAS_TEXTURE_SIZE;
 
 } // namespace prop::GpuTextEngine
 
@@ -5578,16 +5632,56 @@
 
 inline void GPUTextEngine::Destroy() { DestroyGPUTextEngine(release()); }
 
-inline void GPUTextEngine::SetGPUWinding(TextEngineRef engine,
-                                         GPUTextEngineWinding winding)
+/**
+ * Sets the winding order of the vertices returned by Text.GetGPUDrawData
+ * for a particular GPU text engine.
+ *
+ * @param engine a TextEngine object created with
+ *               GPUTextEngine.GPUTextEngine().
+ * @param winding the new winding order option.
+ *
+ * @threadsafety This function should be called on the thread that created the
+ *               engine.
+ *
+ * @since This function is available since SDL_ttf 3.0.0.
+ *
+ * @sa GPUTextEngine.GetGPUWinding
+ */
+inline void SetGPUTextEngineWinding(TextEngineRef engine,
+                                    GPUTextEngineWinding winding)
 {
-  SDL::SetGPUTextEngineWinding(engine, winding);
+  TTF_SetGPUTextEngineWinding(engine, winding);
 }
 
-inline GPUTextEngineWinding GPUTextEngine::GetGPUWinding(
-  TextEngineRef engine) const
+inline void GPUTextEngine::SetGPUWinding(GPUTextEngineWinding winding)
 {
-  return SDL::GetGPUTextEngineWinding(engine);
+  SDL::SetGPUTextEngineWinding(get(), winding);
+}
+
+/**
+ * Get the winding order of the vertices returned by Text.GetGPUDrawData
+ * for a particular GPU text engine
+ *
+ * @param engine a TextEngine object created with
+ *               GPUTextEngine.GPUTextEngine().
+ * @returns the winding order used by the GPU text engine or
+ *          GPU_TEXTENGINE_WINDING_INVALID in case of error.
+ *
+ * @threadsafety This function should be called on the thread that created the
+ *               engine.
+ *
+ * @since This function is available since SDL_ttf 3.0.0.
+ *
+ * @sa GPUTextEngine.SetGPUWinding
+ */
+inline GPUTextEngineWinding GetGPUTextEngineWinding(TextEngineRef engine)
+{
+  return TTF_GetGPUTextEngineWinding(engine);
+}
+
+inline GPUTextEngineWinding GPUTextEngine::GetGPUWinding() const
+{
+  return SDL::GetGPUTextEngineWinding(get());
 }
 
 /**
@@ -5597,8 +5691,6 @@
  *               nullptr.
  * @param font the font to render with.
  * @param text the text to use, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @returns a Text object or nullptr on failure; call GetError() for more
  *          information.
  *
@@ -5622,7 +5714,7 @@
 }
 
 inline Text::Text(TextEngineRef engine, FontRef font, std::string_view text)
-  : m_resource(TTF_CreateText(engine, font, text))
+  : m_resource(TTF_CreateText(engine, font, text.data(), text.size()))
 {
 }
 
@@ -5640,7 +5732,7 @@
  */
 inline PropertiesRef GetTextProperties(TextConstRef text)
 {
-  return CheckError(TTF_GetTextProperties(text));
+  return {CheckError(TTF_GetTextProperties(text))};
 }
 
 inline PropertiesRef Text::GetProperties() const
@@ -5745,7 +5837,7 @@
  */
 inline FontRef GetTextFont(TextConstRef text)
 {
-  return CheckError(TTF_GetTextFont(text));
+  return {CheckError(TTF_GetTextFont(text))};
 }
 
 inline FontRef Text::GetFont() const { return SDL::GetTextFont(m_resource); }
@@ -5805,7 +5897,7 @@
  *
  * @param text the text to modify.
  * @param script an [ISO 15924
- *               code](https://unicode.org/iso15924/iso15924-codes.html) .
+ *               code](https://unicode.org/iso15924/iso15924-codes.html).
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -5855,10 +5947,7 @@
  * The default text color is white (255, 255, 255, 255).
  *
  * @param text the Text to modify.
- * @param r the red color value in the range of 0-255.
- * @param g the green color value in the range of 0-255.
- * @param b the blue color value in the range of 0-255.
- * @param a the alpha value in the range of 0-255.
+ * @param c the color value in the range of 0-255.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -5871,7 +5960,7 @@
  */
 inline void SetTextColor(TextRef text, Color c)
 {
-  CheckError(TTF_SetTextColor(text, c));
+  CheckError(TTF_SetTextColor(text, c.r, c.g, c.b, c.a));
 }
 
 inline void Text::SetColor(Color c) { SDL::SetTextColor(m_resource, c); }
@@ -5882,10 +5971,7 @@
  * The default text color is white (1.0f, 1.0f, 1.0f, 1.0f).
  *
  * @param text the Text to modify.
- * @param r the red color value, normally in the range of 0-1.
- * @param g the green color value, normally in the range of 0-1.
- * @param b the blue color value, normally in the range of 0-1.
- * @param a the alpha value in the range of 0-1.
+ * @param c the color value, normally in the range of 0-1.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -5898,7 +5984,7 @@
  */
 inline void SetTextColorFloat(TextRef text, FColor c)
 {
-  CheckError(TTF_SetTextColorFloat(text, c));
+  CheckError(TTF_SetTextColorFloat(text, c.r, c.g, c.b, c.a));
 }
 
 inline void Text::SetColorFloat(FColor c)
@@ -5941,14 +6027,7 @@
  * Get the color of a text object.
  *
  * @param text the Text to query.
- * @param r a pointer filled in with the red color value in the range of 0-255,
- *          may be nullptr.
- * @param g a pointer filled in with the green color value in the range of
- *          0-255, may be nullptr.
- * @param b a pointer filled in with the blue color value in the range of 0-255,
- *          may be nullptr.
- * @param a a pointer filled in with the alpha value in the range of 0-255, may
- *          be nullptr.
+ * @return The color in the range of 0-255 on success.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -5961,7 +6040,9 @@
  */
 inline Color GetTextColor(TextRef text)
 {
-  static_assert(false, "Not implemented");
+  Color c;
+  GetTextColor(text, &c.r, &c.g, &c.b, &c.a);
+  return c;
 }
 
 inline void Text::GetColor(Uint8* r, Uint8* g, Uint8* b, Uint8* a) const
@@ -5969,7 +6050,7 @@
   SDL::GetTextColor(m_resource, r, g, b, a);
 }
 
-inline Color Text::GetColor() { return SDL::GetTextColor(m_resource); }
+inline Color Text::GetColor() const { return SDL::GetTextColor(m_resource); }
 
 /**
  * Get the color of a text object.
@@ -6006,14 +6087,7 @@
  * Get the color of a text object.
  *
  * @param text the Text to query.
- * @param r a pointer filled in with the red color value, normally in the range
- *          of 0-1, may be nullptr.
- * @param g a pointer filled in with the green color value, normally in the
- *          range of 0-1, may be nullptr.
- * @param b a pointer filled in with the blue color value, normally in the range
- *          of 0-1, may be nullptr.
- * @param a a pointer filled in with the alpha value in the range of 0-1, may be
- *          nullptr.
+ * @return The color in the range of 0-1 on success.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -6026,7 +6100,9 @@
  */
 inline FColor GetTextColorFloat(TextRef text)
 {
-  static_assert(false, "Not implemented");
+  FColor c;
+  GetTextColorFloat(text, &c.r, &c.g, &c.b, &c.a);
+  return c;
 }
 
 inline void Text::GetColorFloat(float* r, float* g, float* b, float* a) const
@@ -6034,7 +6110,7 @@
   SDL::GetTextColorFloat(m_resource, r, g, b, a);
 }
 
-inline FColor Text::GetColorFloat()
+inline FColor Text::GetColorFloat() const
 {
   return SDL::GetTextColorFloat(m_resource);
 }
@@ -6048,8 +6124,8 @@
  * This function may cause the internal text representation to be rebuilt.
  *
  * @param text the Text to modify.
- * @param x the x offset of the upper left corner of this text in pixels.
- * @param y the y offset of the upper left corner of this text in pixels.
+ * @param p the x, y offset of the upper left corner of this text in pixels.
+ * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
  *               text.
@@ -6060,7 +6136,7 @@
  */
 inline void SetTextPosition(TextRef text, const PointRaw& p)
 {
-  CheckError(TTF_SetTextPosition(text, p));
+  CheckError(TTF_SetTextPosition(text, p.x, p.y));
 }
 
 inline void Text::SetPosition(const PointRaw& p)
@@ -6076,6 +6152,7 @@
  *          this text in pixels, may be nullptr.
  * @param y a pointer filled in with the y offset of the upper left corner of
  *          this text in pixels, may be nullptr.
+ * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
  *               text.
@@ -6093,10 +6170,9 @@
  * Get the position of a text object.
  *
  * @param text the Text to query.
- * @param x a pointer filled in with the x offset of the upper left corner of
- *          this text in pixels, may be nullptr.
- * @param y a pointer filled in with the y offset of the upper left corner of
- *          this text in pixels, may be nullptr.
+ * @returns a Point with the offset of the upper left corner of this text in
+ *          pixels on success.
+ * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
  *               text.
@@ -6107,7 +6183,9 @@
  */
 inline Point GetTextPosition(TextRef text)
 {
-  static_assert(false, "Not implemented");
+  Point p;
+  GetTextPosition(text, &p.x, &p.y);
+  return p;
 }
 
 inline void Text::GetPosition(int* x, int* y) const
@@ -6115,7 +6193,10 @@
   SDL::GetTextPosition(m_resource, x, y);
 }
 
-inline Point Text::GetPosition() { return SDL::GetTextPosition(m_resource); }
+inline Point Text::GetPosition() const
+{
+  return SDL::GetTextPosition(m_resource);
+}
 
 /**
  * Set whether wrapping is enabled on a text object.
@@ -6148,8 +6229,8 @@
  * Get whether wrapping is enabled on a text object.
  *
  * @param text the Text to query.
- * @param wrap_width a pointer filled in with the maximum width in pixels or 0
- *                   if the text is being wrapped on newline characters.
+ * @returns maximum width in pixels or 0 if the text is being wrapped on newline
+ *          characters.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -6161,7 +6242,9 @@
  */
 inline int GetTextWrapWidth(TextConstRef text)
 {
-  return CheckError(TTF_GetTextWrapWidth(text));
+  int w;
+  CheckError(TTF_GetTextWrapWidth(text, &w));
+  return w;
 }
 
 inline int Text::GetWrapWidth() const
@@ -6229,8 +6312,6 @@
  *
  * @param text the Text to modify.
  * @param string the UTF-8 text to use, may be nullptr.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -6244,7 +6325,7 @@
  */
 inline void SetTextString(TextRef text, std::string_view string)
 {
-  CheckError(TTF_SetTextString(text, string));
+  CheckError(TTF_SetTextString(text, string.data(), string.size()));
 }
 
 inline void Text::SetString(std::string_view string)
@@ -6263,8 +6344,6 @@
  *               does not do UTF-8 validation, so you should only insert at
  *               UTF-8 sequence boundaries.
  * @param string the UTF-8 text to insert.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -6278,7 +6357,7 @@
  */
 inline void InsertTextString(TextRef text, int offset, std::string_view string)
 {
-  CheckError(TTF_InsertTextString(text, offset, string));
+  CheckError(TTF_InsertTextString(text, offset, string.data(), string.size()));
 }
 
 inline void Text::InsertString(int offset, std::string_view string)
@@ -6293,8 +6372,6 @@
  *
  * @param text the Text to modify.
  * @param string the UTF-8 text to insert.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -6308,7 +6385,7 @@
  */
 inline void AppendTextString(TextRef text, std::string_view string)
 {
-  CheckError(TTF_AppendTextString(text, string));
+  CheckError(TTF_AppendTextString(text, string.data(), string.size()));
 }
 
 inline void Text::AppendString(std::string_view string)
@@ -6377,10 +6454,8 @@
  * The size of the text may change when the font or font style and size change.
  *
  * @param text the Text to query.
- * @param w a pointer filled in with the width of the text, in pixels, may be
- *          nullptr.
- * @param h a pointer filled in with the height of the text, in pixels, may be
- *          nullptr.
+ * @returns Point containing the size on x and y axis on success or
+ *          std::nullopt on failure; call GetError() for more information.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -6390,7 +6465,9 @@
  */
 inline Point GetTextSize(TextRef text)
 {
-  static_assert(false, "Not implemented");
+  Point p;
+  GetTextSize(text, &p.x, &p.y);
+  return p;
 }
 
 inline void Text::GetSize(int* w, int* h) const
@@ -6398,7 +6475,7 @@
   SDL::GetTextSize(m_resource, w, h);
 }
 
-inline Point Text::GetSize() { return SDL::GetTextSize(m_resource); }
+inline Point Text::GetSize() const { return SDL::GetTextSize(m_resource); }
 
 /**
  * Get the substring of a text object that surrounds a text offset.
@@ -6470,12 +6547,8 @@
  * @param offset a byte offset into the text string.
  * @param length the length of the range being queried, in bytes, or -1 for the
  *               remainder of the string.
- * @param count a pointer filled in with the number of substrings returned, may
- *              be nullptr.
  * @returns a nullptr terminated array of substring pointers or nullptr on
- *          failure; call GetError() for more information. This is a single
- *          allocation that should be freed with free() when it is no longer
- *          needed.
+ *          failure; call GetError() for more information.
  *
  * @threadsafety This function should be called on the thread that created the
  *               text.
@@ -6486,7 +6559,9 @@
                                                       int offset,
                                                       int length)
 {
-  return TTF_GetTextSubStringsForRange(text, offset, length);
+  int count = 0;
+  auto data = TTF_GetTextSubStringsForRange(text, offset, length, &count);
+  return OwnArray<SubString*>{data, size_t(count)};
 }
 
 inline OwnArray<SubString*> Text::GetSubStringsForRange(int offset,
@@ -6501,9 +6576,7 @@
  * This will return the closest substring of text to the given point.
  *
  * @param text the Text to query.
- * @param x the x coordinate relative to the left side of the text, may be
- *          outside the bounds of the text area.
- * @param y the y coordinate relative to the top side of the text, may be
+ * @param p the coordinates relative to the top-left side of the text, may be
  *          outside the bounds of the text area.
  * @param substring a pointer filled in with the closest substring of text to
  *                  the given point.
@@ -6518,7 +6591,7 @@
                                      Point p,
                                      SubString* substring)
 {
-  CheckError(TTF_GetTextSubStringForPoint(text, p, substring));
+  CheckError(TTF_GetTextSubStringForPoint(text, p.x, p.y, substring));
 }
 
 inline void Text::GetSubStringForPoint(Point p, SubString* substring) const
@@ -6534,6 +6607,8 @@
  *
  * @param text the Text to query.
  * @param substring the SubString to query.
+ * @param previous a pointer filled in with the previous substring in the text
+ *                 object.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -6648,4 +6723,6 @@
 
 } // namespace SDL
 
+#endif // defined(SDL3PP_ENABLE_TTF) || defined(SDL3PP_DOC)
+
 #endif /* SDL3PP_TTF_H_ */
