--- src/generated/SDL3pp_image.h
+++ include/SDL3pp/SDL3pp_image.h
@@ -1,17 +1,25 @@
 #ifndef SDL3PP_IMAGE_H_
 #define SDL3PP_IMAGE_H_
 
-#include <SDL3/SDL_image.h>
 #include "SDL3pp_error.h"
 #include "SDL3pp_render.h"
 #include "SDL3pp_surface.h"
 #include "SDL3pp_version.h"
 #include "SDL3pp_video.h"
 
+#if !defined(SDL3PP_DISABLE_IMAGE) && !defined(SDL3PP_ENABLE_IMAGE) &&         \
+  __has_include(<SDL3_image/SDL_image.h>)
+#define SDL3PP_ENABLE_IMAGE
+#endif
+
+#if defined(SDL3PP_ENABLE_IMAGE) || defined(SDL3PP_DOC)
+
+#include <SDL3_image/SDL_image.h>
+
 namespace SDL {
 
 /**
- * @defgroup CategorySDLImage Category SDLImage
+ * @defgroup CategorySDLImage Load images from several formats
  *
  * Header file for SDL_image library
  *
@@ -95,14 +103,24 @@
   constexpr auto operator->() { return value; }
 };
 
-/// Printable format: "%d.%d.%d", MAJOR, MINOR, MICRO
-#define SDL_IMAGE_MAJOR_VERSION 3
+#ifdef SDL3PP_DOC
+
+/**
+ * @name Image version
+ * @{
+ * Printable format: "%d.%d.%d", MAJOR, MINOR, MICRO
+ */
+#define SDL_IMAGE_MAJOR_VERSION
+
+#define SDL_IMAGE_MINOR_VERSION
 
-#define SDL_IMAGE_MINOR_VERSION 2
+#define SDL_IMAGE_MICRO_VERSION
 
-#define SDL_IMAGE_MICRO_VERSION 5
+/// @}
 
-/// This is the version number macro for the current SDL_image version.
+/**
+ * This is the version number macro for the current SDL_image version.
+ */
 #define SDL_IMAGE_VERSION                                                      \
   SDL_VERSIONNUM(                                                              \
     SDL_IMAGE_MAJOR_VERSION, SDL_IMAGE_MINOR_VERSION, SDL_IMAGE_MICRO_VERSION)
@@ -114,6 +132,8 @@
    (SDL_IMAGE_MAJOR_VERSION > X || SDL_IMAGE_MINOR_VERSION > Y ||              \
     SDL_IMAGE_MICRO_VERSION >= Z))
 
+#endif // SDL3PP_DOC
+
 namespace IMG {
 
 /**
@@ -123,12 +143,10 @@
  *
  * @since This function is available since SDL_image 3.0.0.
  */
-inline int Version();
+inline int Version() { return IMG_Version(); }
 
 } // namespace IMG
 
-inline int IMG::Version() { return SDL::Version(); }
-
 /**
  * Load an image from an SDL data source into a software surface.
  *
@@ -170,14 +188,11 @@
  * images directly into an Texture for use by the GPU without using a software
  * surface: call LoadTextureTyped() instead.
  *
- * When done with the returned surface, the app should dispose of it with a call
- * to Surface.Destroy().
- *
  * @param src an IOStream that data will be read from.
- * @param closeio true to close/free the IOStream before returning, false to
- *                leave it open.
  * @param type a filename extension that represent this data ("BMP", "GIF",
  *             "PNG", etc).
+ * @param closeio true to close/free the IOStream before returning, false
+ *                to leave it open.
  * @returns a new SDL surface, or nullptr on error.
  *
  * @since This function is available since SDL_image 3.0.0.
@@ -190,7 +205,7 @@
                                 StringParam type,
                                 bool closeio = false)
 {
-  return IMG_LoadTyped_IO(src, type, closeio);
+  return Surface(IMG_LoadTyped_IO(src, closeio, type));
 }
 
 /**
@@ -219,9 +234,6 @@
  * images directly into an Texture for use by the GPU without using a software
  * surface: call Texture.Texture() instead.
  *
- * When done with the returned surface, the app should dispose of it with a call
- * to [Surface.Destroy](https://wiki.libsdl.org/SDL3/Surface.Destroy) ().
- *
  * @param file a path on the filesystem to load an image from.
  * @returns a new SDL surface, or nullptr on error.
  *
@@ -231,7 +243,7 @@
  * @sa Surface.Surface
  * @sa Surface.Destroy
  */
-inline Surface LoadSurface(StringParam file) { return IMG_Load(file); }
+inline Surface LoadSurface(StringParam file) { return Surface(IMG_Load(file)); }
 
 /**
  * Load an image from an SDL data source into a software surface.
@@ -268,9 +280,6 @@
  * images directly into an Texture for use by the GPU without using a software
  * surface: call Texture.Texture() instead.
  *
- * When done with the returned surface, the app should dispose of it with a call
- * to Surface.Destroy().
- *
  * @param src an IOStream that data will be read from.
  * @param closeio true to close/free the IOStream before returning, false to
  *                leave it open.
@@ -282,9 +291,9 @@
  * @sa LoadSurfaceTyped
  * @sa Surface.Destroy
  */
-inline Surface LoadSurface(IOStreamParam src, bool closeio)
+inline Surface LoadSurface(IOStreamParam src, bool closeio = false)
 {
-  return IMG_Load_IO(src, closeio);
+  return Surface(IMG_Load_IO(src, closeio));
 }
 
 inline Surface::Surface(StringParam file)
@@ -316,9 +325,6 @@
  * If you would rather decode an image to an Surface (a buffer of pixels in CPU
  * memory), call Surface.Surface() instead.
  *
- * When done with the returned texture, the app should dispose of it with a call
- * to Texture.Destroy().
- *
  * @param renderer the Renderer to use to create the GPU texture.
  * @param file a path on the filesystem to load an image from.
  * @returns a new texture, or nullptr on error.
@@ -330,7 +336,7 @@
  */
 inline Texture LoadTexture(RendererParam renderer, StringParam file)
 {
-  return IMG_LoadTexture(renderer, file);
+  return Texture(IMG_LoadTexture(renderer, file));
 }
 
 /**
@@ -361,9 +367,6 @@
  * If you would rather decode an image to an Surface (a buffer of pixels in CPU
  * memory), call Surface.Surface() instead.
  *
- * When done with the returned texture, the app should dispose of it with a call
- * to Texture.Destroy().
- *
  * @param renderer the Renderer to use to create the GPU texture.
  * @param src an IOStream that data will be read from.
  * @param closeio true to close/free the IOStream before returning, false to
@@ -378,9 +381,9 @@
  */
 inline Texture LoadTexture(RendererParam renderer,
                            IOStreamParam src,
-                           bool closeio)
+                           bool closeio = false)
 {
-  return IMG_LoadTexture_IO(renderer, src, closeio);
+  return Texture(IMG_LoadTexture_IO(renderer, src, closeio));
 }
 
 inline Texture::Texture(RendererParam renderer, StringParam file)
@@ -427,21 +430,17 @@
  * If you would rather decode an image to an Surface (a buffer of pixels in CPU
  * memory), call LoadSurfaceTyped() instead.
  *
- * When done with the returned texture, the app should dispose of it with a call
- * to Texture.Destroy().
- *
  * @param renderer the Renderer to use to create the GPU texture.
  * @param src an IOStream that data will be read from.
- * @param closeio true to close/free the IOStream before returning, false to
- *                leave it open.
  * @param type a filename extension that represent this data ("BMP", "GIF",
  *             "PNG", etc).
+ * @param closeio true to close/free the IOStream before returning, false to
+ *                leave it open.
  * @returns a new texture, or nullptr on error.
  *
  * @since This function is available since SDL_image 3.0.0.
  *
  * @sa Texture.Texture
- * @sa Texture.Texture
  * @sa Texture.Destroy
  */
 inline Texture LoadTextureTyped(RendererParam renderer,
@@ -449,7 +448,7 @@
                                 StringParam type,
                                 bool closeio = false)
 {
-  return IMG_LoadTextureTyped_IO(renderer, src, type, closeio);
+  return Texture(IMG_LoadTextureTyped_IO(renderer, src, closeio, type));
 }
 
 /**
@@ -1241,7 +1240,10 @@
  * @sa LoadXV
  * @sa LoadWEBP
  */
-inline Surface LoadAVIF(IOStreamParam src) { return IMG_LoadAVIF_IO(src); }
+inline Surface LoadAVIF(IOStreamParam src)
+{
+  return Surface(IMG_LoadAVIF_IO(src));
+}
 
 /**
  * Load a ICO image directly.
@@ -1275,7 +1277,10 @@
  * @sa LoadXV
  * @sa LoadWEBP
  */
-inline Surface LoadICO(IOStreamParam src) { return IMG_LoadICO_IO(src); }
+inline Surface LoadICO(IOStreamParam src)
+{
+  return Surface{IMG_LoadICO_IO(src)};
+}
 
 /**
  * Load a CUR image directly.
@@ -1309,7 +1314,10 @@
  * @sa LoadXV
  * @sa LoadWEBP
  */
-inline Surface LoadCUR(IOStreamParam src) { return IMG_LoadCUR_IO(src); }
+inline Surface LoadCUR(IOStreamParam src)
+{
+  return Surface{IMG_LoadCUR_IO(src)};
+}
 
 /**
  * Load a BMP image directly.
@@ -1343,7 +1351,10 @@
  * @sa LoadXV
  * @sa LoadWEBP
  */
-inline Surface LoadBMP(IOStreamParam src) { return IMG_LoadBMP_IO(src); }
+inline Surface LoadBMP(IOStreamParam src)
+{
+  return Surface{IMG_LoadBMP_IO(src)};
+}
 
 /**
  * Load a GIF image directly.
@@ -1377,7 +1388,10 @@
  * @sa LoadXV
  * @sa LoadWEBP
  */
-inline Surface LoadGIF(IOStreamParam src) { return IMG_LoadGIF_IO(src); }
+inline Surface LoadGIF(IOStreamParam src)
+{
+  return Surface{IMG_LoadGIF_IO(src)};
+}
 
 /**
  * Load a JPG image directly.
@@ -1411,7 +1425,10 @@
  * @sa LoadXV
  * @sa LoadWEBP
  */
-inline Surface LoadJPG(IOStreamParam src) { return IMG_LoadJPG_IO(src); }
+inline Surface LoadJPG(IOStreamParam src)
+{
+  return Surface{IMG_LoadJPG_IO(src)};
+}
 
 /**
  * Load a JXL image directly.
@@ -1445,7 +1462,10 @@
  * @sa LoadXV
  * @sa LoadWEBP
  */
-inline Surface LoadJXL(IOStreamParam src) { return IMG_LoadJXL_IO(src); }
+inline Surface LoadJXL(IOStreamParam src)
+{
+  return Surface{IMG_LoadJXL_IO(src)};
+}
 
 /**
  * Load a LBM image directly.
@@ -1479,7 +1499,10 @@
  * @sa LoadXV
  * @sa LoadWEBP
  */
-inline Surface LoadLBM(IOStreamParam src) { return IMG_LoadLBM_IO(src); }
+inline Surface LoadLBM(IOStreamParam src)
+{
+  return Surface{IMG_LoadLBM_IO(src)};
+}
 
 /**
  * Load a PCX image directly.
@@ -1513,7 +1536,10 @@
  * @sa LoadXV
  * @sa LoadWEBP
  */
-inline Surface LoadPCX(IOStreamParam src) { return IMG_LoadPCX_IO(src); }
+inline Surface LoadPCX(IOStreamParam src)
+{
+  return Surface{IMG_LoadPCX_IO(src)};
+}
 
 /**
  * Load a PNG image directly.
@@ -1547,7 +1573,10 @@
  * @sa LoadXV
  * @sa LoadWEBP
  */
-inline Surface LoadPNG(IOStreamParam src) { return IMG_LoadPNG_IO(src); }
+inline Surface LoadPNG(IOStreamParam src)
+{
+  return Surface{IMG_LoadPNG_IO(src)};
+}
 
 /**
  * Load a PNM image directly.
@@ -1581,7 +1610,10 @@
  * @sa LoadXV
  * @sa LoadWEBP
  */
-inline Surface LoadPNM(IOStreamParam src) { return IMG_LoadPNM_IO(src); }
+inline Surface LoadPNM(IOStreamParam src)
+{
+  return Surface{IMG_LoadPNM_IO(src)};
+}
 
 /**
  * Load a SVG image directly.
@@ -1615,7 +1647,10 @@
  * @sa LoadXV
  * @sa LoadWEBP
  */
-inline Surface LoadSVG(IOStreamParam src) { return IMG_LoadSVG_IO(src); }
+inline Surface LoadSVG(IOStreamParam src)
+{
+  return Surface{IMG_LoadSVG_IO(src)};
+}
 
 /**
  * Load a QOI image directly.
@@ -1649,7 +1684,10 @@
  * @sa LoadXV
  * @sa LoadWEBP
  */
-inline Surface LoadQOI(IOStreamParam src) { return IMG_LoadQOI_IO(src); }
+inline Surface LoadQOI(IOStreamParam src)
+{
+  return Surface{IMG_LoadQOI_IO(src)};
+}
 
 /**
  * Load a TGA image directly.
@@ -1683,7 +1721,10 @@
  * @sa LoadXV
  * @sa LoadWEBP
  */
-inline Surface LoadTGA(IOStreamParam src) { return IMG_LoadTGA_IO(src); }
+inline Surface LoadTGA(IOStreamParam src)
+{
+  return Surface{IMG_LoadTGA_IO(src)};
+}
 
 /**
  * Load a TIFF image directly.
@@ -1717,7 +1758,10 @@
  * @sa LoadXV
  * @sa LoadWEBP
  */
-inline Surface LoadTIF(IOStreamParam src) { return IMG_LoadTIF_IO(src); }
+inline Surface LoadTIF(IOStreamParam src)
+{
+  return Surface{IMG_LoadTIF_IO(src)};
+}
 
 /**
  * Load a XCF image directly.
@@ -1751,7 +1795,10 @@
  * @sa LoadXV
  * @sa LoadWEBP
  */
-inline Surface LoadXCF(IOStreamParam src) { return IMG_LoadXCF_IO(src); }
+inline Surface LoadXCF(IOStreamParam src)
+{
+  return Surface{IMG_LoadXCF_IO(src)};
+}
 
 /**
  * Load a XPM image directly.
@@ -1785,7 +1832,10 @@
  * @sa LoadXV
  * @sa LoadWEBP
  */
-inline Surface LoadXPM(IOStreamParam src) { return IMG_LoadXPM_IO(src); }
+inline Surface LoadXPM(IOStreamParam src)
+{
+  return Surface{IMG_LoadXPM_IO(src)};
+}
 
 /**
  * Load a XV image directly.
@@ -1819,7 +1869,7 @@
  * @sa LoadXPM
  * @sa LoadWEBP
  */
-inline Surface LoadXV(IOStreamParam src) { return IMG_LoadXV_IO(src); }
+inline Surface LoadXV(IOStreamParam src) { return Surface{IMG_LoadXV_IO(src)}; }
 
 /**
  * Load a WEBP image directly.
@@ -1853,7 +1903,10 @@
  * @sa LoadXPM
  * @sa LoadXV
  */
-inline Surface LoadWEBP(IOStreamParam src) { return IMG_LoadWEBP_IO(src); }
+inline Surface LoadWEBP(IOStreamParam src)
+{
+  return Surface{IMG_LoadWEBP_IO(src)};
+}
 
 /**
  * Load an SVG image, scaled to a specific size.
@@ -1864,19 +1917,15 @@
  * Either width or height may be 0 and the image will be auto-sized to preserve
  * aspect ratio.
  *
- * When done with the returned surface, the app should dispose of it with a call
- * to Surface.Destroy().
- *
  * @param src an IOStream to load SVG data from.
- * @param width desired width of the generated surface, in pixels.
- * @param height desired height of the generated surface, in pixels.
+ * @param size desired width and height of the generated surface, in pixels.
  * @returns a new SDL surface, or nullptr on error.
  *
  * @since This function is available since SDL_image 3.0.0.
  */
 inline Surface LoadSizedSVG(IOStreamParam src, const PointRaw& size)
 {
-  return IMG_LoadSizedSVG_IO(src, size);
+  return Surface(IMG_LoadSizedSVG_IO(src, size.x, size.y));
 }
 
 /**
@@ -1886,9 +1935,6 @@
  * it will be 32bpp. If you always want 32-bit data, use
  * ReadXPMFromArrayToRGB888() instead.
  *
- * When done with the returned surface, the app should dispose of it with a call
- * to Surface.Destroy().
- *
  * @param xpm a null-terminated array of strings that comprise XPM data.
  * @returns a new SDL surface, or nullptr on error.
  *
@@ -1898,7 +1944,7 @@
  */
 inline Surface ReadXPMFromArray(char** xpm)
 {
-  return IMG_ReadXPMFromArray(xpm);
+  return Surface{IMG_ReadXPMFromArray(xpm)};
 }
 
 /**
@@ -1907,9 +1953,6 @@
  * The returned surface will always be a 32-bit RGB surface. If you want 8-bit
  * indexed colors (and the XPM data allows it), use ReadXPMFromArray() instead.
  *
- * When done with the returned surface, the app should dispose of it with a call
- * to Surface.Destroy().
- *
  * @param xpm a null-terminated array of strings that comprise XPM data.
  * @returns a new SDL surface, or nullptr on error.
  *
@@ -1919,7 +1962,7 @@
  */
 inline Surface ReadXPMFromArrayToRGB888(char** xpm)
 {
-  return IMG_ReadXPMFromArrayToRGB888(xpm);
+  return Surface{IMG_ReadXPMFromArrayToRGB888(xpm)};
 }
 
 /**
@@ -1934,8 +1977,6 @@
  * @throws Error on failure.
  *
  * @since This function is available since SDL_image 3.0.0.
- *
- * @sa SaveAVIF
  */
 inline void SaveAVIF(SurfaceParam surface, StringParam file, int quality)
 {
@@ -1952,22 +1993,20 @@
  *
  * @param surface the SDL surface to save.
  * @param dst the IOStream to save the image data to.
- * @param closeio true to close/free the IOStream before returning, false to
- *                leave it open.
  * @param quality the desired quality, ranging between 0 (lowest) and 100
  *                (highest).
+ * @param closeio true to close/free the IOStream before returning, false to
+ *                leave it open.
  * @throws Error on failure.
  *
  * @since This function is available since SDL_image 3.0.0.
- *
- * @sa SaveAVIF
  */
 inline void SaveAVIF(SurfaceParam surface,
                      IOStreamParam dst,
                      int quality,
                      bool closeio = false)
 {
-  CheckError(IMG_SaveAVIF_IO(surface, dst, quality, closeio));
+  CheckError(IMG_SaveAVIF_IO(surface, dst, closeio, quality));
 }
 
 /**
@@ -2003,8 +2042,6 @@
  * @throws Error on failure.
  *
  * @since This function is available since SDL_image 3.0.0.
- *
- * @sa SavePNG
  */
 inline void SavePNG(SurfaceParam surface,
                     IOStreamParam dst,
@@ -2025,8 +2062,6 @@
  * @throws Error on failure.
  *
  * @since This function is available since SDL_image 3.0.0.
- *
- * @sa SaveJPG
  */
 inline void SaveJPG(SurfaceParam surface, StringParam file, int quality)
 {
@@ -2043,22 +2078,20 @@
  *
  * @param surface the SDL surface to save.
  * @param dst the IOStream to save the image data to.
- * @param closeio true to close/free the IOStream before returning, false to
- *                leave it open.
  * @param quality [0; 33] is Lowest quality, [34; 66] is Middle quality, [67;
  *                100] is Highest quality.
+ * @param closeio true to close/free the IOStream before returning, false to
+ *                leave it open.
  * @throws Error on failure.
  *
  * @since This function is available since SDL_image 3.0.0.
- *
- * @sa SaveJPG
  */
 inline void SaveJPG(SurfaceParam surface,
                     IOStreamParam dst,
                     int quality,
                     bool closeio = false)
 {
-  CheckError(IMG_SaveJPG_IO(surface, dst, quality, closeio));
+  CheckError(IMG_SaveJPG_IO(surface, dst, closeio, quality));
 }
 
 /**
@@ -2138,7 +2171,7 @@
    *
    * @sa Animation.Free
    */
-  Animation(IOStreamParam src, bool closeio)
+  Animation(IOStreamParam src, bool closeio = false)
     : m_resource(IMG_LoadAnimation_IO(src, closeio))
   {
   }
@@ -2202,21 +2235,34 @@
    * @since This function is available since SDL_image 3.0.0.
    *
    * @sa Animation.Animation
-   * @sa Animation.Animation
    * @sa LoadAnimationTyped
    */
   void Free();
 
+  /// Get the width in pixels.
   int GetWidth() const;
 
+  /// Get the height in pixels.
   int GetHeight() const;
 
+  /// Get the size in pixels.
   Point GetSize() const;
 
+  /// Return the number of frames.
   int GetCount() const;
 
+  /**
+   * Return the frame image under given index.
+   *
+   * @param index the index to get frame, within [0, GetCount() - 1]
+   */
   Surface GetFrame(int index) const;
 
+  /**
+   * Return the frame delay under given index.
+   *
+   * @param index the index to get frame, within [0, GetCount() - 1]
+   */
   int GetDelay(int index) const;
 };
 
@@ -2259,29 +2305,26 @@
   ~AnimationRef() { release(); }
 };
 
-inline int GetAnimationWidth(AnimationConstParam anim)
-{
-  static_assert(false, "Not implemented");
-}
+/// Get the width in pixels.
+inline int GetAnimationWidth(AnimationConstParam anim) { return anim->w; }
 
 inline int Animation::GetWidth() const
 {
   return SDL::GetAnimationWidth(m_resource);
 }
 
-inline int GetAnimationHeight(AnimationConstParam anim)
-{
-  static_assert(false, "Not implemented");
-}
+/// Get the height in pixels.
+inline int GetAnimationHeight(AnimationConstParam anim) { return anim->h; }
 
 inline int Animation::GetHeight() const
 {
   return SDL::GetAnimationHeight(m_resource);
 }
 
+/// Get the size in pixels.
 inline Point GetAnimationSize(AnimationConstParam anim)
 {
-  static_assert(false, "Not implemented");
+  return {anim->w, anim->h};
 }
 
 inline Point Animation::GetSize() const
@@ -2289,19 +2332,23 @@
   return SDL::GetAnimationSize(m_resource);
 }
 
-inline int GetAnimationCount(AnimationConstParam anim)
-{
-  static_assert(false, "Not implemented");
-}
+/// Return the number of frames.
+inline int GetAnimationCount(AnimationConstParam anim) { return anim->count; }
 
 inline int Animation::GetCount() const
 {
   return SDL::GetAnimationCount(m_resource);
 }
 
+/**
+ * Return the frame image under given index.
+ *
+ * @param anim Animation to dispose of.
+ * @param index the index to get frame, within [0, GetCount() - 1]
+ */
 inline Surface GetAnimationFrame(AnimationConstParam anim, int index)
 {
-  static_assert(false, "Not implemented");
+  return Surface::Borrow(anim->frames[index]);
 }
 
 inline Surface Animation::GetFrame(int index) const
@@ -2309,9 +2356,15 @@
   return SDL::GetAnimationFrame(m_resource, index);
 }
 
+/**
+ * Return the frame delay under given index.
+ *
+ * @param anim Animation to dispose of.
+ * @param index the index to get frame, within [0, GetCount() - 1]
+ */
 inline int GetAnimationDelay(AnimationConstParam anim, int index)
 {
-  static_assert(false, "Not implemented");
+  return anim->delays[index];
 }
 
 inline int Animation::GetDelay(int index) const
@@ -2356,7 +2409,7 @@
  *
  * @sa Animation.Free
  */
-inline Animation LoadAnimation(IOStreamParam src, bool closeio)
+inline Animation LoadAnimation(IOStreamParam src, bool closeio = false)
 {
   return Animation(src, closeio);
 }
@@ -2378,22 +2431,21 @@
  * call to Animation.Free().
  *
  * @param src an IOStream that data will be read from.
+ * @param type a filename extension that represent this data ("GIF", etc).
  * @param closeio true to close/free the IOStream before returning, false to
  *                leave it open.
- * @param type a filename extension that represent this data ("GIF", etc).
  * @returns a new Animation, or nullptr on error.
  *
  * @since This function is available since SDL_image 3.0.0.
  *
  * @sa Animation.Animation
- * @sa Animation.Animation
  * @sa Animation.Free
  */
 inline Animation LoadAnimationTyped(IOStreamParam src,
                                     StringParam type,
                                     bool closeio = false)
 {
-  return IMG_LoadAnimationTyped_IO(src, type, closeio);
+  return Animation(IMG_LoadAnimationTyped_IO(src, closeio, type));
 }
 
 /**
@@ -2406,7 +2458,6 @@
  * @since This function is available since SDL_image 3.0.0.
  *
  * @sa Animation.Animation
- * @sa Animation.Animation
  * @sa LoadAnimationTyped
  */
 inline void FreeAnimation(AnimationRaw anim) { IMG_FreeAnimation(anim); }
@@ -2427,13 +2478,12 @@
  * @since This function is available since SDL_image 3.0.0.
  *
  * @sa Animation.Animation
- * @sa Animation.Animation
  * @sa LoadAnimationTyped
  * @sa Animation.Free
  */
-inline AnimationRef LoadGIFAnimation(IOStreamParam src)
+inline Animation LoadGIFAnimation(IOStreamParam src)
 {
-  return IMG_LoadGIFAnimation_IO(src);
+  return Animation(IMG_LoadGIFAnimation_IO(src));
 }
 
 /**
@@ -2450,17 +2500,18 @@
  * @since This function is available since SDL_image 3.0.0.
  *
  * @sa Animation.Animation
- * @sa Animation.Animation
  * @sa LoadAnimationTyped
  * @sa Animation.Free
  */
-inline AnimationRef LoadWEBPAnimation(IOStreamParam src)
+inline Animation LoadWEBPAnimation(IOStreamParam src)
 {
-  return IMG_LoadWEBPAnimation_IO(src);
+  return Animation{IMG_LoadWEBPAnimation_IO(src)};
 }
 
 /// @}
 
 } // namespace SDL
 
+#endif // defined(SDL3PP_ENABLE_IMAGE) || defined(SDL3PP_DOC)
+
 #endif /* SDL3PP_IMAGE_H_ */
