--- src/generated/SDL3pp_metal.h
+++ include/SDL3pp/SDL3pp_metal.h
@@ -7,7 +7,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryMetal Category Metal
+ * @defgroup CategoryMetal Metal Support
  *
  * Functions to creating Metal layers and views on SDL windows.
  *
@@ -109,7 +109,7 @@
    *
    * @since This function is available since SDL 3.2.0.
    *
-   * @sa MetalView.Metal_DestroyView
+   * @sa MetalView.Destroy
    * @sa MetalView.GetLayer
    */
   MetalView(WindowParam window)
@@ -165,7 +165,7 @@
    *
    * @sa MetalView.MetalView
    */
-  void Metal_DestroyView();
+  void Destroy();
 
   /**
    * Get a pointer to the backing CAMetalLayer for the given view.
@@ -231,7 +231,7 @@
  *
  * @since This function is available since SDL 3.2.0.
  *
- * @sa MetalView.Metal_DestroyView
+ * @sa MetalView.Destroy
  * @sa MetalView.GetLayer
  */
 inline MetalView Metal_CreateView(WindowParam window)
@@ -256,7 +256,7 @@
   SDL_Metal_DestroyView(view);
 }
 
-inline void MetalView::Metal_DestroyView() { Metal_DestroyView(release()); }
+inline void MetalView::Destroy() { SDL::Metal_DestroyView(release()); }
 
 /**
  * Get a pointer to the backing CAMetalLayer for the given view.
