--- build/generated/SDL3pp_gpu.h
+++ include/SDL3pp/SDL3pp_gpu.h
@@ -12,7 +12,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryGPU Category GPU
+ * @defgroup CategoryGPU 3D Rendering and GPU Compute
  *
  * The GPU API offers a cross-platform way for apps to talk to modern graphics
  * hardware. It offers both 3D graphics and compute support, in the style of
@@ -138,7 +138,7 @@
  * limited hardware support, this API is probably not for you.
  *
  * Examples demonstrating proper usage of this API can be found
- * [here](https://github.com/TheSpydog/SDL_gpu_examples) .
+ * [here](https://github.com/TheSpydog/SDL_gpu_examples).
  *
  * ## Performance considerations
  *
@@ -1253,7 +1253,6 @@
    * @param first_slot the vertex buffer slot to begin binding from.
    * @param bindings an array of GPUBufferBinding structs containing vertex
    *                 buffers and offset values.
-   * @param num_bindings the number of bindings in the bindings array.
    *
    * @since This function is available since SDL 3.2.0.
    */
@@ -1284,8 +1283,6 @@
    * @param first_slot the vertex sampler slot to begin binding from.
    * @param texture_sampler_bindings an array of texture-sampler binding
    *                                 structs.
-   * @param num_bindings the number of texture-sampler pairs to bind from the
-   *                     array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1306,7 +1303,6 @@
    *
    * @param first_slot the vertex storage texture slot to begin binding from.
    * @param storage_textures an array of storage textures.
-   * @param num_bindings the number of storage texture to bind from the array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1326,7 +1322,6 @@
    *
    * @param first_slot the vertex storage buffer slot to begin binding from.
    * @param storage_buffers an array of buffers.
-   * @param num_bindings the number of buffers to bind from the array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1346,8 +1341,6 @@
    * @param first_slot the fragment sampler slot to begin binding from.
    * @param texture_sampler_bindings an array of texture-sampler binding
    *                                 structs.
-   * @param num_bindings the number of texture-sampler pairs to bind from the
-   *                     array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1368,7 +1361,6 @@
    *
    * @param first_slot the fragment storage texture slot to begin binding from.
    * @param storage_textures an array of storage textures.
-   * @param num_bindings the number of storage textures to bind from the array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1389,7 +1381,6 @@
    *
    * @param first_slot the fragment storage buffer slot to begin binding from.
    * @param storage_buffers an array of storage buffers.
-   * @param num_bindings the number of storage buffers to bind from the array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1555,8 +1546,6 @@
    * @param first_slot the compute sampler slot to begin binding from.
    * @param texture_sampler_bindings an array of texture-sampler binding
    *                                 structs.
-   * @param num_bindings the number of texture-sampler bindings to bind from the
-   *                     array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1577,7 +1566,6 @@
    *
    * @param first_slot the compute storage texture slot to begin binding from.
    * @param storage_textures an array of storage textures.
-   * @param num_bindings the number of storage textures to bind from the array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1597,7 +1585,6 @@
    *
    * @param first_slot the compute storage buffer slot to begin binding from.
    * @param storage_buffers an array of storage buffer binding structs.
-   * @param num_bindings the number of storage buffers to bind from the array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -2141,7 +2128,6 @@
    *
    * @param slot_index the vertex uniform slot to push data to.
    * @param data client data to write.
-   * @param length the length of the data to write.
    *
    * @since This function is available since SDL 3.2.0.
    */
@@ -2158,7 +2144,6 @@
    *
    * @param slot_index the fragment uniform slot to push data to.
    * @param data client data to write.
-   * @param length the length of the data to write.
    *
    * @since This function is available since SDL 3.2.0.
    */
@@ -2175,7 +2160,6 @@
    *
    * @param slot_index the uniform slot to push data to.
    * @param data client data to write.
-   * @param length the length of the data to write.
    *
    * @since This function is available since SDL 3.2.0.
    */
@@ -2202,8 +2186,6 @@
    *
    * @param color_target_infos an array of texture subresources with
    *                           corresponding clear values and load/store ops.
-   * @param num_color_targets the number of color targets in the
-   *                          color_target_infos array.
    * @param depth_stencil_target_info a texture subresource with corresponding
    *                                  clear value and load/store ops, may be
    *                                  nullptr.
@@ -2241,12 +2223,8 @@
    *
    * @param storage_texture_bindings an array of writeable storage texture
    *                                 binding structs.
-   * @param num_storage_texture_bindings the number of storage textures to bind
-   *                                     from the array.
    * @param storage_buffer_bindings an array of writeable storage buffer binding
    *                                structs.
-   * @param num_storage_buffer_bindings the number of storage buffers to bind
-   *                                    from the array.
    * @returns a compute pass handle.
    *
    * @since This function is available since SDL 3.2.0.
@@ -2318,12 +2296,11 @@
    * than the one that created the window.
    *
    * @param window a window that has been claimed.
-   * @param swapchain_texture a pointer filled in with a swapchain texture
-   *                          handle.
    * @param swapchain_texture_width a pointer filled in with the swapchain
    *                                texture width, may be nullptr.
    * @param swapchain_texture_height a pointer filled in with the swapchain
    *                                 texture height, may be nullptr.
+   * @returns a swapchain texture handle.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called from the thread that
@@ -2368,12 +2345,11 @@
    * another reading operation.
    *
    * @param window a window that has been claimed.
-   * @param swapchain_texture a pointer filled in with a swapchain texture
-   *                          handle.
    * @param swapchain_texture_width a pointer filled in with the swapchain
    *                                texture width, may be nullptr.
    * @param swapchain_texture_height a pointer filled in with the swapchain
    *                                 texture height, may be nullptr.
+   * @returns a swapchain texture handle.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called from the thread that
@@ -4052,7 +4028,6 @@
    * @param wait_all if 0, wait for any fence to be signaled, if 1, wait for all
    *                 fences to be signaled.
    * @param fences an array of fences to wait on.
-   * @param num_fences the number of fences in the fences array.
    * @throws Error on failure.
    *
    * @since This function is available since SDL 3.2.0.
@@ -4114,6 +4089,8 @@
   bool TextureSupportsSampleCount(GPUTextureFormat format,
                                   GPUSampleCount sample_count);
 
+#ifdef SDL_PLATFORM_GDK
+
   /**
    * Call this to suspend GPU operation on Xbox when you receive the
    * EVENT_DID_ENTER_BACKGROUND event.
@@ -4139,6 +4116,8 @@
    * @sa AddEventWatch
    */
   void GDKResumeGPU();
+
+#endif /* SDL_PLATFORM_GDK */
 };
 
 /**
@@ -6336,7 +6315,6 @@
  * @param command_buffer a command buffer.
  * @param slot_index the vertex uniform slot to push data to.
  * @param data client data to write.
- * @param length the length of the data to write.
  *
  * @since This function is available since SDL 3.2.0.
  */
@@ -6367,7 +6345,6 @@
  * @param command_buffer a command buffer.
  * @param slot_index the fragment uniform slot to push data to.
  * @param data client data to write.
- * @param length the length of the data to write.
  *
  * @since This function is available since SDL 3.2.0.
  */
@@ -6398,7 +6375,6 @@
  * @param command_buffer a command buffer.
  * @param slot_index the uniform slot to push data to.
  * @param data client data to write.
- * @param length the length of the data to write.
  *
  * @since This function is available since SDL 3.2.0.
  */
@@ -6439,8 +6415,6 @@
  * @param command_buffer a command buffer.
  * @param color_target_infos an array of texture subresources with corresponding
  *                           clear values and load/store ops.
- * @param num_color_targets the number of color targets in the
- *                          color_target_infos array.
  * @param depth_stencil_target_info a texture subresource with corresponding
  *                                  clear value and load/store ops, may be
  *                                  nullptr.
@@ -6455,8 +6429,10 @@
   std::span<const GPUColorTargetInfo> color_target_infos,
   OptionalRef<const GPUDepthStencilTargetInfo> depth_stencil_target_info)
 {
-  return SDL_BeginGPURenderPass(
-    command_buffer, color_target_infos, depth_stencil_target_info);
+  return SDL_BeginGPURenderPass(command_buffer,
+                                color_target_infos.data(),
+                                color_target_infos.size(),
+                                depth_stencil_target_info);
 }
 
 inline GPURenderPass GPUCommandBuffer::BeginRenderPass(
@@ -6517,7 +6493,7 @@
  */
 inline void SetGPUScissor(GPURenderPass render_pass, const RectRaw& scissor)
 {
-  SDL_SetGPUScissor(render_pass, scissor);
+  SDL_SetGPUScissor(render_pass, &scissor);
 }
 
 inline void GPURenderPass::SetScissor(const RectRaw& scissor)
@@ -6572,7 +6548,6 @@
  * @param first_slot the vertex buffer slot to begin binding from.
  * @param bindings an array of GPUBufferBinding structs containing vertex
  *                 buffers and offset values.
- * @param num_bindings the number of bindings in the bindings array.
  *
  * @since This function is available since SDL 3.2.0.
  */
@@ -6580,7 +6555,8 @@
                                  Uint32 first_slot,
                                  std::span<const GPUBufferBinding> bindings)
 {
-  SDL_BindGPUVertexBuffers(render_pass, first_slot, bindings);
+  SDL_BindGPUVertexBuffers(
+    render_pass, first_slot, bindings.data(), bindings.size());
 }
 
 inline void GPURenderPass::BindVertexBuffers(
@@ -6625,8 +6601,6 @@
  * @param render_pass a render pass handle.
  * @param first_slot the vertex sampler slot to begin binding from.
  * @param texture_sampler_bindings an array of texture-sampler binding structs.
- * @param num_bindings the number of texture-sampler pairs to bind from the
- *                     array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -6637,7 +6611,10 @@
   Uint32 first_slot,
   std::span<const GPUTextureSamplerBinding> texture_sampler_bindings)
 {
-  SDL_BindGPUVertexSamplers(render_pass, first_slot, texture_sampler_bindings);
+  SDL_BindGPUVertexSamplers(render_pass,
+                            first_slot,
+                            texture_sampler_bindings.data(),
+                            texture_sampler_bindings.size());
 }
 
 inline void GPURenderPass::BindVertexSamplers(
@@ -6660,7 +6637,6 @@
  * @param render_pass a render pass handle.
  * @param first_slot the vertex storage texture slot to begin binding from.
  * @param storage_textures an array of storage textures.
- * @param num_bindings the number of storage texture to bind from the array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -6671,7 +6647,8 @@
   Uint32 first_slot,
   SpanRef<const GPUTextureRaw> storage_textures)
 {
-  SDL_BindGPUVertexStorageTextures(render_pass, first_slot, storage_textures);
+  SDL_BindGPUVertexStorageTextures(
+    render_pass, first_slot, storage_textures.data(), storage_textures.size());
 }
 
 inline void GPURenderPass::BindVertexStorageTextures(
@@ -6694,7 +6671,6 @@
  * @param render_pass a render pass handle.
  * @param first_slot the vertex storage buffer slot to begin binding from.
  * @param storage_buffers an array of buffers.
- * @param num_bindings the number of buffers to bind from the array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -6705,7 +6681,8 @@
   Uint32 first_slot,
   SpanRef<const GPUBufferRaw> storage_buffers)
 {
-  SDL_BindGPUVertexStorageBuffers(render_pass, first_slot, storage_buffers);
+  SDL_BindGPUVertexStorageBuffers(
+    render_pass, first_slot, storage_buffers.data(), storage_buffers.size());
 }
 
 inline void GPURenderPass::BindVertexStorageBuffers(
@@ -6727,8 +6704,6 @@
  * @param render_pass a render pass handle.
  * @param first_slot the fragment sampler slot to begin binding from.
  * @param texture_sampler_bindings an array of texture-sampler binding structs.
- * @param num_bindings the number of texture-sampler pairs to bind from the
- *                     array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -6739,8 +6714,10 @@
   Uint32 first_slot,
   std::span<const GPUTextureSamplerBinding> texture_sampler_bindings)
 {
-  SDL_BindGPUFragmentSamplers(
-    render_pass, first_slot, texture_sampler_bindings);
+  SDL_BindGPUFragmentSamplers(render_pass,
+                              first_slot,
+                              texture_sampler_bindings.data(),
+                              texture_sampler_bindings.size());
 }
 
 inline void GPURenderPass::BindFragmentSamplers(
@@ -6763,7 +6740,6 @@
  * @param render_pass a render pass handle.
  * @param first_slot the fragment storage texture slot to begin binding from.
  * @param storage_textures an array of storage textures.
- * @param num_bindings the number of storage textures to bind from the array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -6774,7 +6750,8 @@
   Uint32 first_slot,
   SpanRef<const GPUTextureRaw> storage_textures)
 {
-  SDL_BindGPUFragmentStorageTextures(render_pass, first_slot, storage_textures);
+  SDL_BindGPUFragmentStorageTextures(
+    render_pass, first_slot, storage_textures.data(), storage_textures.size());
 }
 
 inline void GPURenderPass::BindFragmentStorageTextures(
@@ -6797,7 +6774,6 @@
  * @param render_pass a render pass handle.
  * @param first_slot the fragment storage buffer slot to begin binding from.
  * @param storage_buffers an array of storage buffers.
- * @param num_bindings the number of storage buffers to bind from the array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -6808,7 +6784,8 @@
   Uint32 first_slot,
   SpanRef<const GPUBufferRaw> storage_buffers)
 {
-  SDL_BindGPUFragmentStorageBuffers(render_pass, first_slot, storage_buffers);
+  SDL_BindGPUFragmentStorageBuffers(
+    render_pass, first_slot, storage_buffers.data(), storage_buffers.size());
 }
 
 inline void GPURenderPass::BindFragmentStorageBuffers(
@@ -7015,12 +6992,8 @@
  * @param command_buffer a command buffer.
  * @param storage_texture_bindings an array of writeable storage texture binding
  *                                 structs.
- * @param num_storage_texture_bindings the number of storage textures to bind
- *                                     from the array.
  * @param storage_buffer_bindings an array of writeable storage buffer binding
  *                                structs.
- * @param num_storage_buffer_bindings the number of storage buffers to bind from
- *                                    the array.
  * @returns a compute pass handle.
  *
  * @since This function is available since SDL 3.2.0.
@@ -7032,8 +7005,11 @@
   std::span<const GPUStorageTextureReadWriteBinding> storage_texture_bindings,
   std::span<const GPUStorageBufferReadWriteBinding> storage_buffer_bindings)
 {
-  return SDL_BeginGPUComputePass(
-    command_buffer, storage_texture_bindings, storage_buffer_bindings);
+  return SDL_BeginGPUComputePass(command_buffer,
+                                 storage_texture_bindings.data(),
+                                 storage_texture_bindings.size(),
+                                 storage_buffer_bindings.data(),
+                                 storage_buffer_bindings.size());
 }
 
 inline GPUComputePass GPUCommandBuffer::BeginComputePass(
@@ -7074,8 +7050,6 @@
  * @param compute_pass a compute pass handle.
  * @param first_slot the compute sampler slot to begin binding from.
  * @param texture_sampler_bindings an array of texture-sampler binding structs.
- * @param num_bindings the number of texture-sampler bindings to bind from the
- *                     array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -7086,8 +7060,10 @@
   Uint32 first_slot,
   std::span<const GPUTextureSamplerBinding> texture_sampler_bindings)
 {
-  SDL_BindGPUComputeSamplers(
-    compute_pass, first_slot, texture_sampler_bindings);
+  SDL_BindGPUComputeSamplers(compute_pass,
+                             first_slot,
+                             texture_sampler_bindings.data(),
+                             texture_sampler_bindings.size());
 }
 
 inline void GPUComputePass::BindSamplers(
@@ -7110,7 +7086,6 @@
  * @param compute_pass a compute pass handle.
  * @param first_slot the compute storage texture slot to begin binding from.
  * @param storage_textures an array of storage textures.
- * @param num_bindings the number of storage textures to bind from the array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -7121,7 +7096,8 @@
   Uint32 first_slot,
   SpanRef<const GPUTextureRaw> storage_textures)
 {
-  SDL_BindGPUComputeStorageTextures(compute_pass, first_slot, storage_textures);
+  SDL_BindGPUComputeStorageTextures(
+    compute_pass, first_slot, storage_textures.data(), storage_textures.size());
 }
 
 inline void GPUComputePass::BindStorageTextures(
@@ -7144,7 +7120,6 @@
  * @param compute_pass a compute pass handle.
  * @param first_slot the compute storage buffer slot to begin binding from.
  * @param storage_buffers an array of storage buffer binding structs.
- * @param num_bindings the number of storage buffers to bind from the array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -7155,7 +7130,8 @@
   Uint32 first_slot,
   SpanRef<const GPUBufferRaw> storage_buffers)
 {
-  SDL_BindGPUComputeStorageBuffers(compute_pass, first_slot, storage_buffers);
+  SDL_BindGPUComputeStorageBuffers(
+    compute_pass, first_slot, storage_buffers.data(), storage_buffers.size());
 }
 
 inline void GPUComputePass::BindStorageBuffers(
@@ -7687,7 +7663,7 @@
  * supported via GPUDevice.WindowSupportsPresentMode /
  * GPUDevice.WindowSupportsSwapchainComposition prior to calling this function.
  *
- * GPU_PRESENTMODE_VSYNC with GPU_SWAPCHAINCOMPOSITION_SDR is always supported.
+ * GPU_PRESENTMODE_VSYNC with GPU_SWAPCHAINCOMPOSITION_SDR are always supported.
  *
  * @param device a GPU context.
  * @param window an Window that has been claimed.
@@ -7803,11 +7779,11 @@
  *
  * @param command_buffer a command buffer.
  * @param window a window that has been claimed.
- * @param swapchain_texture a pointer filled in with a swapchain texture handle.
  * @param swapchain_texture_width a pointer filled in with the swapchain texture
  *                                width, may be nullptr.
  * @param swapchain_texture_height a pointer filled in with the swapchain
  *                                 texture height, may be nullptr.
+ * @returns a swapchain texture handle.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called from the thread that
@@ -7830,8 +7806,13 @@
   Uint32* swapchain_texture_width = nullptr,
   Uint32* swapchain_texture_height = nullptr)
 {
-  return CheckError(SDL_AcquireGPUSwapchainTexture(
-    command_buffer, window, swapchain_texture_width, swapchain_texture_height));
+  GPUTextureRaw texture;
+  CheckError(SDL_AcquireGPUSwapchainTexture(command_buffer,
+                                            window,
+                                            &texture,
+                                            swapchain_texture_width,
+                                            swapchain_texture_height));
+  return texture;
 }
 
 inline GPUTexture GPUCommandBuffer::AcquireSwapchainTexture(
@@ -7895,11 +7876,11 @@
  *
  * @param command_buffer a command buffer.
  * @param window a window that has been claimed.
- * @param swapchain_texture a pointer filled in with a swapchain texture handle.
  * @param swapchain_texture_width a pointer filled in with the swapchain texture
  *                                width, may be nullptr.
  * @param swapchain_texture_height a pointer filled in with the swapchain
  *                                 texture height, may be nullptr.
+ * @returns a swapchain texture handle.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called from the thread that
@@ -7917,8 +7898,13 @@
   Uint32* swapchain_texture_width = nullptr,
   Uint32* swapchain_texture_height = nullptr)
 {
-  return CheckError(SDL_WaitAndAcquireGPUSwapchainTexture(
-    command_buffer, window, swapchain_texture_width, swapchain_texture_height));
+  GPUTextureRaw texture;
+  CheckError(SDL_WaitAndAcquireGPUSwapchainTexture(command_buffer,
+                                                   window,
+                                                   &texture,
+                                                   swapchain_texture_width,
+                                                   swapchain_texture_height));
+  return texture;
 }
 
 inline GPUTexture GPUCommandBuffer::WaitAndAcquireSwapchainTexture(
@@ -8052,7 +8038,6 @@
  * @param wait_all if 0, wait for any fence to be signaled, if 1, wait for all
  *                 fences to be signaled.
  * @param fences an array of fences to wait on.
- * @param num_fences the number of fences in the fences array.
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
@@ -8064,7 +8049,8 @@
                              bool wait_all,
                              std::span<GPUFence* const> fences)
 {
-  CheckError(SDL_WaitForGPUFences(device, wait_all, fences));
+  CheckError(
+    SDL_WaitForGPUFences(device, wait_all, fences.data(), fences.size()));
 }
 
 inline void GPUDevice::WaitForFences(bool wait_all,
@@ -8232,6 +8218,8 @@
 
 #endif // SDL_VERSION_ATLEAST(3, 4, 0)
 
+#ifdef SDL_PLATFORM_GDK
+
 /**
  * Call this to suspend GPU operation on Xbox when you receive the
  * EVENT_DID_ENTER_BACKGROUND event.
@@ -8266,6 +8254,8 @@
 
 inline void GPUDevice::GDKResumeGPU() { SDL::GDKResumeGPU(m_resource); }
 
+#endif /* SDL_PLATFORM_GDK */
+
 /// @}
 
 } // namespace SDL
