--- src/generated/SDL3pp_pen.h
+++ include/SDL3pp/SDL3pp_pen.h
@@ -8,7 +8,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryPen Category Pen
+ * @defgroup CategoryPen Pen Support
  *
  * SDL pen event handling.
  *
@@ -184,8 +184,6 @@
  */
 using PenDeviceType = SDL_PenDeviceType;
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
 constexpr PenDeviceType PEN_DEVICE_TYPE_INVALID =
   SDL_PEN_DEVICE_TYPE_INVALID; ///< Not a valid pen device.
 
@@ -199,8 +197,6 @@
   SDL_PEN_DEVICE_TYPE_INDIRECT; ///< Pen touches something that isn't the
                                 ///< display.
 
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Get the device type of the given pen.
  *
