--- build/generated/SDL3pp_camera.h
+++ include/SDL3pp/SDL3pp_camera.h
@@ -11,7 +11,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryCamera Category Camera
+ * @defgroup CategoryCamera Camera Support
  *
  * Video capture for the SDL library.
  *
@@ -140,6 +140,16 @@
 
 constexpr CameraPermissionState CAMERA_PERMISSION_STATE_APPROVED =
   SDL_CAMERA_PERMISSION_STATE_APPROVED; ///< CAMERA_PERMISSION_STATE_APPROVED
+#else
+
+/**
+ * The current state of a request for camera access.
+ *
+ * @since This enum is available since SDL 3.4.0.
+ *
+ * @sa Camera.GetPermissionState
+ */
+using CameraPermissionState = int;
 
 #endif // SDL_VERSION_ATLEAST(3, 4, 0)
 
@@ -237,8 +247,7 @@
    * (or EVENT_CAMERA_DEVICE_DENIED) event, or poll GetCameraPermissionState()
    * occasionally until it returns non-zero.
    *
-   * @param spec the CameraSpec to be initialized by this function.
-   * @throws Error on failure.
+   * @returns the CameraSpec or std::nullopt if waiting for user approval.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
@@ -265,11 +274,11 @@
    * is called, so real failure here is almost always an out of memory
    * condition.
    *
-   * After use, the frame should be released with CameraBase.ReleaseFrame(). If
-   * you don't do this, the system may stop providing more video!
+   * After use, the frame should be released with ReleaseCameraFrame(). If you
+   * don't do this, the system may stop providing more video!
    *
    * Do not call DestroySurface() on the returned surface! It must be given back
-   * to the camera subsystem with CameraBase.ReleaseFrame!
+   * to the camera subsystem with ReleaseCameraFrame!
    *
    * If the system is waiting for the user to approve access to the camera, as
    * some platforms require, this will return nullptr (no frames available); you
@@ -286,9 +295,9 @@
    *
    * @since This function is available since SDL 3.2.0.
    *
-   * @sa CameraBase.ReleaseFrame
+   * @sa ReleaseCameraFrame
    */
-  Surface AcquireFrame(Uint64* timestampNS = nullptr);
+  CameraFrame AcquireFrame(Uint64* timestampNS = nullptr);
 
   /**
    * Release a frame of video acquired from a camera.
@@ -296,9 +305,9 @@
    * Let the back-end re-use the internal buffer for camera.
    *
    * This function _must_ be called only on surface objects returned by
-   * CameraBase.AcquireFrame(). This function should be called as quickly as
-   * possible after acquisition, as SDL keeps a small FIFO queue of surfaces for
-   * video frames; if surfaces aren't released in a timely manner, SDL may drop
+   * AcquireCameraFrame(). This function should be called as quickly as possible
+   * after acquisition, as SDL keeps a small FIFO queue of surfaces for video
+   * frames; if surfaces aren't released in a timely manner, SDL may drop
    * upcoming video frames from the camera.
    *
    * If the app needs to keep the surface for a significant time, they should
@@ -307,15 +316,15 @@
    * The app should not use the surface again after calling this function;
    * assume the surface is freed and the pointer is invalid.
    *
-   * @param frame the video frame surface to release.
+   * @param lock the video frame surface to release.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
    * @since This function is available since SDL 3.2.0.
    *
-   * @sa CameraBase.AcquireFrame
+   * @sa AcquireCameraFrame
    */
-  void ReleaseFrame(SurfaceRef frame);
+  void ReleaseFrame(CameraFrame&& lock);
 };
 
 /**
@@ -383,9 +392,9 @@
    *
    * @param instance_id the camera device instance ID.
    * @param spec the desired format for data the device will provide. Can be
-   *             nullptr.
-   * @post an Camera object or nullptr on failure; call GetError() for more
-   *       information.
+   *             std::nullopt.
+   * @post an Camera object on success..
+   * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
@@ -408,79 +417,14 @@
 
   /// Assignment operator.
   Camera& operator=(const Camera& other) = delete;
-
-  /**
-   * Acquire a frame.
-   *
-   * The frame is a memory pointer to the image data, whose size and format are
-   * given by the spec requested when opening the device.
-   *
-   * This is a non blocking API. If there is a frame available, a non-nullptr
-   * surface is returned, and timestampNS will be filled with a non-zero value.
-   *
-   * Note that an error case can also return nullptr, but a nullptr by itself is
-   * normal and just signifies that a new frame is not yet available. Note that
-   * even if a camera device fails outright (a USB camera is unplugged while in
-   * use, etc), SDL will send an event separately to notify the app, but
-   * continue to provide blank frames at ongoing intervals until CloseCamera()
-   * is called, so real failure here is almost always an out of memory
-   * condition.
-   *
-   * After use, the frame should be released with CameraBase.ReleaseFrame(). If
-   * you don't do this, the system may stop providing more video!
-   *
-   * Do not call DestroySurface() on the returned surface! It must be given back
-   * to the camera subsystem with CameraBase.ReleaseFrame!
-   *
-   * If the system is waiting for the user to approve access to the camera, as
-   * some platforms require, this will return nullptr (no frames available); you
-   * should either wait for an EVENT_CAMERA_DEVICE_APPROVED (or
-   * EVENT_CAMERA_DEVICE_DENIED) event, or poll GetCameraPermissionState()
-   * occasionally until it returns non-zero.
-   *
-   * @param timestampNS a pointer filled in with the frame's timestamp, or 0 on
-   *                    error. Can be nullptr.
-   * @returns a new frame of video on success, nullptr if none is currently
-   *          available.
-   *
-   * @threadsafety It is safe to call this function from any thread.
-   *
-   * @since This function is available since SDL 3.2.0.
-   *
-   * @sa CameraBase.ReleaseFrame
-   */
-  CameraFrame AcquireFrame(Uint64* timestampNS);
-
-  /**
-   * Release a frame of video acquired from a camera.
-   *
-   * Let the back-end re-use the internal buffer for camera.
-   *
-   * This function _must_ be called only on surface objects returned by
-   * CameraBase.AcquireFrame(). This function should be called as quickly as
-   * possible after acquisition, as SDL keeps a small FIFO queue of surfaces for
-   * video frames; if surfaces aren't released in a timely manner, SDL may drop
-   * upcoming video frames from the camera.
-   *
-   * If the app needs to keep the surface for a significant time, they should
-   * make a copy of it and release the original.
-   *
-   * The app should not use the surface again after calling this function;
-   * assume the surface is freed and the pointer is invalid.
-   *
-   * @param frame the video frame surface to release.
-   *
-   * @threadsafety It is safe to call this function from any thread.
-   *
-   * @since This function is available since SDL 3.2.0.
-   *
-   * @sa CameraBase.AcquireFrame
-   */
-  void ReleaseFrame(CameraFrame&& lock);
 };
 
-struct CameraFrame : Surface
+/// Camera Frame.
+class CameraFrame : public Surface
 {
+  CameraRef m_lock;
+
+public:
   /**
    * Acquire a frame.
    *
@@ -498,11 +442,11 @@
    * is called, so real failure here is almost always an out of memory
    * condition.
    *
-   * After use, the frame should be released with CameraBase.ReleaseFrame(). If
-   * you don't do this, the system may stop providing more video!
+   * After use, the frame should be released with ReleaseCameraFrame(). If you
+   * don't do this, the system may stop providing more video!
    *
    * Do not call DestroySurface() on the returned surface! It must be given back
-   * to the camera subsystem with CameraBase.ReleaseFrame!
+   * to the camera subsystem with ReleaseCameraFrame!
    *
    * If the system is waiting for the user to approve access to the camera, as
    * some platforms require, this will return nullptr (no frames available); you
@@ -510,7 +454,7 @@
    * EVENT_CAMERA_DEVICE_DENIED) event, or poll GetCameraPermissionState()
    * occasionally until it returns non-zero.
    *
-   * @param camera opened camera device.
+   * @param resource opened camera device.
    * @param timestampNS a pointer filled in with the frame's timestamp, or 0 on
    *                    error. Can be nullptr.
    * @post a new frame of video on success, nullptr if none is currently
@@ -520,16 +464,17 @@
    *
    * @since This function is available since SDL 3.2.0.
    *
-   * @sa CameraBase.ReleaseFrame
+   * @sa ReleaseCameraFrame
    */
-  CameraFrame(CameraRef resource);
+  CameraFrame(CameraRef resource, Uint64* timestampNS = nullptr);
 
   /// Copy constructor
   CameraFrame(const CameraFrame& other) = delete;
 
   /// Move constructor
   CameraFrame(CameraFrame&& other) noexcept
-    : m_lock(std::move(other.m_lock))
+    : Surface(std::move(other))
+    , m_lock(std::move(other.m_lock))
   {
   }
 
@@ -539,9 +484,9 @@
    * Let the back-end re-use the internal buffer for camera.
    *
    * This function _must_ be called only on surface objects returned by
-   * CameraBase.AcquireFrame(). This function should be called as quickly as
-   * possible after acquisition, as SDL keeps a small FIFO queue of surfaces for
-   * video frames; if surfaces aren't released in a timely manner, SDL may drop
+   * AcquireCameraFrame(). This function should be called as quickly as possible
+   * after acquisition, as SDL keeps a small FIFO queue of surfaces for video
+   * frames; if surfaces aren't released in a timely manner, SDL may drop
    * upcoming video frames from the camera.
    *
    * If the app needs to keep the surface for a significant time, they should
@@ -550,14 +495,11 @@
    * The app should not use the surface again after calling this function;
    * assume the surface is freed and the pointer is invalid.
    *
-   * @param camera opened camera device.
-   * @param frame the video frame surface to release.
-   *
    * @threadsafety It is safe to call this function from any thread.
    *
    * @since This function is available since SDL 3.2.0.
    *
-   * @sa CameraBase.AcquireFrame
+   * @sa AcquireCameraFrame
    */
   ~CameraFrame() { reset(); }
 
@@ -570,18 +512,15 @@
     return *this;
   }
 
-  /// True if not locked.
-  constexpr operator bool() const { return bool(m_lock); }
-
   /**
    * Release a frame of video acquired from a camera.
    *
    * Let the back-end re-use the internal buffer for camera.
    *
    * This function _must_ be called only on surface objects returned by
-   * CameraBase.AcquireFrame(). This function should be called as quickly as
-   * possible after acquisition, as SDL keeps a small FIFO queue of surfaces for
-   * video frames; if surfaces aren't released in a timely manner, SDL may drop
+   * AcquireCameraFrame(). This function should be called as quickly as possible
+   * after acquisition, as SDL keeps a small FIFO queue of surfaces for video
+   * frames; if surfaces aren't released in a timely manner, SDL may drop
    * upcoming video frames from the camera.
    *
    * If the app needs to keep the surface for a significant time, they should
@@ -590,22 +529,16 @@
    * The app should not use the surface again after calling this function;
    * assume the surface is freed and the pointer is invalid.
    *
-   * @param camera opened camera device.
-   * @param frame the video frame surface to release.
-   *
    * @threadsafety It is safe to call this function from any thread.
    *
    * @since This function is available since SDL 3.2.0.
    *
-   * @sa CameraBase.AcquireFrame
+   * @sa AcquireCameraFrame
    */
   void reset();
 
   /// Get the reference to locked resource.
   CameraRef resource() const { return m_lock; }
-
-  /// Releases the lock without unlocking.
-  void release() { m_lock.release(); }
 };
 
 /**
@@ -680,11 +613,8 @@
 /**
  * Get a list of currently connected camera devices.
  *
- * @param count a pointer filled in with the number of cameras returned, may be
- *              nullptr.
  * @returns a 0 terminated array of camera instance IDs or nullptr on failure;
- *          call GetError() for more information. This should be freed with
- *          free() when it is no longer needed.
+ *          call GetError() for more information.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -692,7 +622,12 @@
  *
  * @sa OpenCamera
  */
-inline OwnArray<CameraID> GetCameras() { return SDL_GetCameras(); }
+inline OwnArray<CameraID> GetCameras()
+{
+  int count;
+  auto data = SDL_GetCameras(&count);
+  return OwnArray<CameraID>(data, count);
+}
 
 /**
  * Get the list of native formats/sizes a camera supports.
@@ -716,12 +651,8 @@
  * warning popup.
  *
  * @param instance_id the camera device instance ID.
- * @param count a pointer filled in with the number of elements in the list, may
- *              be nullptr.
  * @returns a nullptr terminated array of pointers to CameraSpec or nullptr on
- *          failure; call GetError() for more information. This is a single
- *          allocation that should be freed with free() when it is no longer
- *          needed.
+ *          failure; call GetError() for more information.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -732,7 +663,9 @@
  */
 inline OwnArray<CameraSpec*> GetCameraSupportedFormats(CameraID instance_id)
 {
-  return SDL_GetCameraSupportedFormats(instance_id);
+  int count;
+  auto data = SDL_GetCameraSupportedFormats(instance_id, &count);
+  return OwnArray<CameraSpec*>(data, count);
 }
 
 /**
@@ -808,9 +741,9 @@
  *
  * @param instance_id the camera device instance ID.
  * @param spec the desired format for data the device will provide. Can be
- *             nullptr.
- * @returns an Camera object or nullptr on failure; call GetError() for more
- *          information.
+ *             std::nullopt.
+ * @returns an Camera object on success.
+ * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -905,7 +838,7 @@
  */
 inline PropertiesRef GetCameraProperties(CameraRef camera)
 {
-  return CheckError(SDL_GetCameraProperties(camera));
+  return {CheckError(SDL_GetCameraProperties(camera))};
 }
 
 inline PropertiesRef CameraBase::GetProperties()
@@ -926,8 +859,7 @@
  * occasionally until it returns non-zero.
  *
  * @param camera opened camera device.
- * @param spec the CameraSpec to be initialized by this function.
- * @throws Error on failure.
+ * @returns the CameraSpec or std::nullopt if waiting for user approval.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -937,7 +869,8 @@
  */
 inline std::optional<CameraSpec> GetCameraFormat(CameraRef camera)
 {
-  return CheckError(SDL_GetCameraFormat(camera));
+  if (CameraSpec spec; SDL_GetCameraFormat(camera, &spec)) return spec;
+  return std::nullopt;
 }
 
 inline std::optional<CameraSpec> CameraBase::GetFormat()
@@ -961,11 +894,11 @@
  * to provide blank frames at ongoing intervals until CloseCamera() is called,
  * so real failure here is almost always an out of memory condition.
  *
- * After use, the frame should be released with CameraBase.ReleaseFrame(). If
- * you don't do this, the system may stop providing more video!
+ * After use, the frame should be released with ReleaseCameraFrame(). If you
+ * don't do this, the system may stop providing more video!
  *
  * Do not call DestroySurface() on the returned surface! It must be given back
- * to the camera subsystem with CameraBase.ReleaseFrame!
+ * to the camera subsystem with ReleaseCameraFrame!
  *
  * If the system is waiting for the user to approve access to the camera, as
  * some platforms require, this will return nullptr (no frames available); you
@@ -983,28 +916,24 @@
  *
  * @since This function is available since SDL 3.2.0.
  *
- * @sa CameraBase.ReleaseFrame
+ * @sa ReleaseCameraFrame
  */
 inline Surface AcquireCameraFrame(CameraRef camera,
                                   Uint64* timestampNS = nullptr)
 {
-  return SDL_AcquireCameraFrame(camera, timestampNS);
-}
-
-inline Surface CameraBase::AcquireFrame(Uint64* timestampNS)
-{
-  return SDL::AcquireCameraFrame(get(), timestampNS);
+  return Surface::Borrow(SDL_AcquireCameraFrame(camera, timestampNS));
 }
 
-inline CameraFrame Camera::AcquireFrame(Uint64* timestampNS)
+inline CameraFrame CameraBase::AcquireFrame(Uint64* timestampNS)
 {
-  return {CameraRef(*this)};
+  return {CameraRef(*this), timestampNS};
 }
 
-inline CameraFrame::CameraFrame(CameraRef resource)
-  : m_lock(std::move(resource))
+inline CameraFrame::CameraFrame(CameraRef resource, Uint64* timestampNS)
+  : Surface(AcquireCameraFrame(resource, timestampNS))
+  , m_lock(std::move(resource))
 {
-  AcquireCameraFrame(m_lock);
+  if (!*this) m_lock.release();
 }
 
 /**
@@ -1013,10 +942,10 @@
  * Let the back-end re-use the internal buffer for camera.
  *
  * This function _must_ be called only on surface objects returned by
- * CameraBase.AcquireFrame(). This function should be called as quickly as
- * possible after acquisition, as SDL keeps a small FIFO queue of surfaces for
- * video frames; if surfaces aren't released in a timely manner, SDL may drop
- * upcoming video frames from the camera.
+ * AcquireCameraFrame(). This function should be called as quickly as possible
+ * after acquisition, as SDL keeps a small FIFO queue of surfaces for video
+ * frames; if surfaces aren't released in a timely manner, SDL may drop upcoming
+ * video frames from the camera.
  *
  * If the app needs to keep the surface for a significant time, they should make
  * a copy of it and release the original.
@@ -1031,19 +960,14 @@
  *
  * @since This function is available since SDL 3.2.0.
  *
- * @sa CameraBase.AcquireFrame
+ * @sa AcquireCameraFrame
  */
 inline void ReleaseCameraFrame(CameraRef camera, SurfaceRef frame)
 {
   SDL_ReleaseCameraFrame(camera, frame);
 }
 
-inline void CameraBase::ReleaseFrame(SurfaceRef frame)
-{
-  SDL::ReleaseCameraFrame(get(), frame);
-}
-
-inline void Camera::ReleaseFrame(CameraFrame&& lock)
+inline void CameraBase::ReleaseFrame(CameraFrame&& lock)
 {
   SDL_assert_paranoid(lock.resource() == *this);
   std::move(lock).reset();
@@ -1052,7 +976,7 @@
 inline void CameraFrame::reset()
 {
   if (!m_lock) return;
-  ReleaseCameraFrame(m_lock);
+  ReleaseCameraFrame(m_lock, Surface::release());
   m_lock = {};
 }
 
