--- build/generated/SDL3pp_mutex.h
+++ include/SDL3pp/SDL3pp_mutex.h
@@ -8,7 +8,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryMutex Category Mutex
+ * @defgroup CategoryMutex Thread Synchronization Primitives
  *
  * SDL offers several thread synchronization primitives. This document can't
  * cover the complicated topic of thread safety, but reading up on what each of
@@ -89,7 +89,7 @@
 
 public:
   /// Default ctor
-  constexpr Mutex(std::nullptr_t = nullptr) noexcept
+  constexpr Mutex(std::nullptr_t) noexcept
     : m_resource(nullptr)
   {
   }
@@ -487,7 +487,7 @@
 
 public:
   /// Default ctor
-  constexpr RWLock(std::nullptr_t = nullptr) noexcept
+  constexpr RWLock(std::nullptr_t) noexcept
     : m_resource(nullptr)
   {
   }
@@ -1268,8 +1268,8 @@
    * pointed to by `sem` has a positive value or the specified time has elapsed.
    * If the call is successful it will atomically decrement the semaphore value.
    *
-   * @param timeoutMS the length of the timeout, in milliseconds, or -1 to wait
-   *                  indefinitely.
+   * @param timeout the length of the timeout, in milliseconds, or -1 to wait
+   *                indefinitely.
    * @returns true if the wait succeeds or false if the wait times out.
    *
    * @threadsafety It is safe to call this function from any thread.
@@ -1489,8 +1489,8 @@
  * call is successful it will atomically decrement the semaphore value.
  *
  * @param sem the semaphore to wait on.
- * @param timeoutMS the length of the timeout, in milliseconds, or -1 to wait
- *                  indefinitely.
+ * @param timeout the length of the timeout, in milliseconds, or -1 to wait
+ *                indefinitely.
  * @returns true if the wait succeeds or false if the wait times out.
  *
  * @threadsafety It is safe to call this function from any thread.
@@ -1504,7 +1504,7 @@
 inline bool WaitSemaphoreTimeout(SemaphoreRef sem,
                                  std::chrono::milliseconds timeout)
 {
-  return SDL_WaitSemaphoreTimeout(sem, timeout);
+  return SDL_WaitSemaphoreTimeout(sem, narrowS32(timeout.count()));
 }
 
 inline bool Semaphore::WaitTimeout(std::chrono::milliseconds timeout)
@@ -1570,7 +1570,7 @@
 
 public:
   /// Default ctor
-  constexpr Condition(std::nullptr_t = nullptr) noexcept
+  constexpr Condition(std::nullptr_t) noexcept
     : m_resource(nullptr)
   {
   }
@@ -1731,8 +1731,8 @@
    * behavior.
    *
    * @param mutex the mutex used to coordinate thread access.
-   * @param timeoutMS the maximum time to wait, in milliseconds, or -1 to wait
-   *                  indefinitely.
+   * @param timeout the maximum time to wait, in milliseconds, or -1 to wait
+   *                indefinitely.
    * @returns true if the condition variable is signaled, false if the condition
    *          is not signaled in the allotted time.
    *
@@ -1945,8 +1945,8 @@
  *
  * @param cond the condition variable to wait on.
  * @param mutex the mutex used to coordinate thread access.
- * @param timeoutMS the maximum time to wait, in milliseconds, or -1 to wait
- *                  indefinitely.
+ * @param timeout the maximum time to wait, in milliseconds, or -1 to wait
+ *                indefinitely.
  * @returns true if the condition variable is signaled, false if the condition
  *          is not signaled in the allotted time.
  *
@@ -1962,7 +1962,7 @@
                                  MutexRef mutex,
                                  std::chrono::milliseconds timeout)
 {
-  return SDL_WaitConditionTimeout(cond, mutex, timeout);
+  return SDL_WaitConditionTimeout(cond, mutex, narrowS32(timeout.count()));
 }
 
 inline bool Condition::WaitTimeout(MutexRef mutex,
@@ -1995,8 +1995,8 @@
  *
  * Here is an example of using this:
  *
- * ```c
- *    static InitState init;
+ * ```cpp
+ *    static SDL::InitState init;
  *
  *    bool InitSystem(void)
  *    {
@@ -2050,7 +2050,7 @@
 struct InitState : InitStateRaw
 {
   constexpr InitState()
-    : SDL_InitState{0}
+    : SDL_InitState{}
   {
   }
 
