--- build/generated/SDL3pp_video.h
+++ include/SDL3pp/SDL3pp_video.h
@@ -9,7 +9,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryVideo Category Video
+ * @defgroup CategoryVideo Display and Window Management
  *
  * SDL's video subsystem is largely interested in abstracting window management
  * from the underlying operating system. You can create windows, manage them in
@@ -53,7 +53,7 @@
 struct GLContext;
 
 /// Alias to raw representation for GLContext.
-using GLContextRaw = SDL_GLContext*;
+using GLContextRaw = SDL_GLContext;
 
 // Forward decl
 struct GLContextScoped;
@@ -244,7 +244,7 @@
    * The primary display is often located at (0,0), but may be placed at a
    * different location depending on monitor layout.
    *
-   * @param rect the Rect structure filled in with the display bounds.
+   * @returns the Rect structure filled in with the display bounds.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -268,7 +268,7 @@
    * so these are good guidelines for the maximum space available to a
    * non-fullscreen window.
    *
-   * @param rect the Rect structure filled in with the display bounds.
+   * @returns the Rect structure filled in with the display bounds.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -346,12 +346,8 @@
    * - refresh rate -> highest to lowest
    * - pixel density -> lowest to highest
    *
-   * @param count a pointer filled in with the number of display modes returned,
-   *              may be nullptr.
    * @returns a nullptr terminated array of display mode pointers or nullptr on
-   *          failure; call GetError() for more information. This is a single
-   *          allocation that should be freed with free() when it is no longer
-   *          needed.
+   *          failure; call GetError() for more information.
    *
    * @threadsafety This function should only be called on the main thread.
    *
@@ -371,14 +367,12 @@
    * and finally checking the refresh rate. If all the available modes are too
    * small, then false is returned.
    *
-   * @param w the width in pixels of the desired display mode.
-   * @param h the height in pixels of the desired display mode.
+   * @param size the width and height in pixels of the desired display mode.
    * @param refresh_rate the refresh rate of the desired display mode, or 0.0f
    *                     for the desktop refresh rate.
    * @param include_high_density_modes boolean to include high density modes in
    *                                   the search.
-   * @param closest a pointer filled in with the closest display mode equal to
-   *                or larger than the desired mode.
+   * @returns the closest display mode equal to or larger than the desired mode.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -624,6 +618,7 @@
  * @since This enum is available since SDL 3.2.0.
  *
  * @sa HitTest
+ * @sa HitTestCB
  */
 using HitTestResult = SDL_HitTestResult;
 
@@ -680,11 +675,11 @@
  *
  * @param win the Window where hit-testing was set on.
  * @param area an Point which should be hit-tested.
- * @param data what was passed as `callback_data` to Window.SetHitTest().
  * @returns an HitTestResult value.
  *
- * @sa Window.SetHitTest
+ * @cat listener-callback
  *
+ * @sa Window.SetHitTest
  * @sa HitTest
  */
 using HitTestCB =
@@ -710,8 +705,9 @@
 constexpr ProgressState PROGRESS_STATE_INVALID = SDL_PROGRESS_STATE_INVALID;
 
 constexpr ProgressState PROGRESS_STATE_NONE =
-  SDL_PROGRESS_STATE_NONE; ///< No progress bar is shown
+  SDL_PROGRESS_STATE_NONE; ///< No progress bar is shown.
 
+/// The progress bar is shown in a indeterminate state.
 constexpr ProgressState PROGRESS_STATE_INDETERMINATE =
   SDL_PROGRESS_STATE_INDETERMINATE; ///< The progress bar is shown in a
                                     ///< indeterminate state
@@ -734,9 +730,9 @@
  *
  * @since This struct is available since SDL 3.2.0.
  *
- * @sa Window.Window
- *
  * @cat resource
+ *
+ * @sa Window.Window
  */
 class Window
 {
@@ -783,21 +779,19 @@
    * Create a window and default renderer.
    *
    * @param title the title of the window, in UTF-8 encoding.
-   * @param width the width of the window.
-   * @param height the height of the window.
+   * @param size the width and height of the window.
    * @param window_flags the flags used to create the window (see
-   *                     Window.Window()).
-   * @param window a pointer filled with the window, or nullptr on error.
-   * @param renderer a pointer filled with the renderer, or nullptr on error.
-   * @post true on success or false on failure; call GetError() for more
-   *       information.
+   *                     Window.Window(StringParam,const
+   *                     PointRaw&,WindowFlags)).
+   * @param renderer a pointer filled with the renderer.
+   * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
    *
    * @since This function is available since SDL 3.2.0.
    *
    * @sa Renderer.Renderer
-   * @sa Window.Window
+   * @sa Window.Window(StringParam,const PointRaw&,WindowFlags))
    */
   Window(StringParam title,
          const PointRaw& size,
@@ -875,8 +869,7 @@
    * in a future version of SDL.
    *
    * @param title the title of the window, in UTF-8 encoding.
-   * @param w the width of the window.
-   * @param h the height of the window.
+   * @param size the width and height of the window.
    * @param flags 0, or one or more WindowFlags OR'd together.
    * @post the window that was created or nullptr on failure; call GetError()
    *       for more information.
@@ -946,12 +939,9 @@
    * hidden will be restored when the parent is shown.
    *
    * @param parent the parent of the window, must not be nullptr.
-   * @param offset_x the x position of the popup window relative to the origin
-   *                 of the parent.
-   * @param offset_y the y position of the popup window relative to the origin
-   *                 of the parent window.
-   * @param w the width of the window.
-   * @param h the height of the window.
+   * @param offset the x, y position of the popup window relative to the origin
+   *               of the parent.
+   * @param size the width and height of the window.
    * @param flags WINDOW_TOOLTIP or WINDOW_POPUP_MENU, and zero or more
    *              additional WindowFlags OR'd together.
    * @post the window that was created or nullptr on failure; call GetError()
@@ -1309,7 +1299,6 @@
   /**
    * Get the raw ICC profile data for the screen the window is currently on.
    *
-   * @param size the size of the ICC profile.
    * @returns the raw ICC profile data on success.
    * @throws Error on failure.
    *
@@ -1588,9 +1577,7 @@
    * Additionally, as this is just a request, it can be denied by the windowing
    * system.
    *
-   * @param x the x coordinate of the window, or `WINDOWPOS_CENTERED` or
-   *          `WINDOWPOS_UNDEFINED`.
-   * @param y the y coordinate of the window, or `WINDOWPOS_CENTERED` or
+   * @param p the coordinates of the window, or `WINDOWPOS_CENTERED` or
    *          `WINDOWPOS_UNDEFINED`.
    * @throws Error on failure.
    *
@@ -1635,10 +1622,7 @@
    * If you do not need the value for one of the positions a nullptr may be
    * passed in the `x` or `y` parameter.
    *
-   * @param x a pointer filled in with the x position of the window, may be
-   *          nullptr.
-   * @param y a pointer filled in with the y position of the window, may be
-   *          nullptr.
+   * @returns the position on success.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1646,6 +1630,7 @@
    * @since This function is available since SDL 3.2.0.
    *
    * @sa Window.SetPosition
+   * @sa SetPosition(int *, int *)
    */
   Point GetPosition() const;
 
@@ -1670,8 +1655,7 @@
    * to remain within the usable desktop bounds). Additionally, as this is just
    * a request, it can be denied by the windowing system.
    *
-   * @param w the width of the window, must be > 0.
-   * @param h the height of the window, must be > 0.
+   * @param size the width and height of the window, must be > 0.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1713,8 +1697,7 @@
    * window is on a high pixel density display. Use Window.GetSizeInPixels() or
    * Renderer.GetOutputSize() to get the real client area size in pixels.
    *
-   * @param w a pointer filled in with the width of the window, may be nullptr.
-   * @param h a pointer filled in with the height of the window, may be nullptr.
+   * @returns a point with width and height on success
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1725,11 +1708,36 @@
    * @sa Window.GetSizeInPixels
    * @sa Window.SetSize
    * @sa EVENT_WINDOW_RESIZED
+   * @sa GetSize(int *, int *)
    */
   Point GetSize() const;
 
+  /**
+   * @brief Request the window's position and size to be set.
+   *
+   * @param rect the rect containing the new coordinates
+   * @throws Error on failure.
+   *
+   * @threadsafety This function should only be called on the main thread.
+   *
+   * @sa WindowRef.SetPosition()
+   * @sa WindowRef.SetSize()
+   */
   void SetRect(Rect rect);
 
+  /**
+   * Get the position and client size of a window.
+   *
+   * This is the current position of the window as last reported by the
+   * windowing system.
+   *
+   * The window pixel size may differ from its window coordinate size if the
+   * window is on a high pixel density display. Use Window.GetSizeInPixels()
+   * or RendererRef.GetOutputSize() to get the real client area size in pixels.
+   *
+   * @return Rect with the position and size
+   * @throws Error on failure.
+   */
   Rect GetRect() const;
 
   /**
@@ -1742,8 +1750,6 @@
    * into the rest of the window, but it should not contain visually important
    * or interactable content.
    *
-   * @param rect a pointer filled in with the client area that is safe for
-   *             interactive content.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1856,16 +1862,15 @@
    *
    * @sa Window.Window
    * @sa Window.GetSize
+   * @sa GetSizeInPixels()
    */
   void GetSizeInPixels(int* w, int* h) const;
 
   /**
    * Get the size of a window's client area, in pixels.
    *
-   * @param w a pointer to variable for storing the width in pixels, may be
-   *          nullptr.
-   * @param h a pointer to variable for storing the height in pixels, may be
-   *          nullptr.
+   * @returns the size on success or std::nullopt on failure; call GetError()
+   * for more information.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1874,14 +1879,14 @@
    *
    * @sa Window.Window
    * @sa Window.GetSize
+   * @sa GetSizeInPixels(int*, int*)
    */
   Point GetSizeInPixels() const;
 
   /**
    * Set the minimum size of a window's client area.
    *
-   * @param min_w the minimum width of the window, or 0 for no limit.
-   * @param min_h the minimum height of the window, or 0 for no limit.
+   * @param p the minimum width and heigh of the window, or 0 for no limit.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1914,8 +1919,7 @@
   /**
    * Set the maximum size of a window's client area.
    *
-   * @param max_w the maximum width of the window, or 0 for no limit.
-   * @param max_h the maximum height of the window, or 0 for no limit.
+   * @param p the maximum width and height of the window, or 0 for no limit.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -2296,8 +2300,8 @@
   /**
    * Get VSync for the window surface.
    *
-   * @param vsync an int filled with the current vertical refresh sync interval.
-   *              See Window.SetSurfaceVSync() for the meaning of the value.
+   * @returns the current vertical refresh sync interval. See
+   *          Window.SetSurfaceVSync() for the meaning of the value.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -2342,7 +2346,6 @@
    *
    * @param rects an array of Rect structures representing areas of the surface
    *              to copy, in pixels.
-   * @param numrects the number of rectangles.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -2591,10 +2594,8 @@
    * On platforms or desktops where this is unsupported, this function does
    * nothing.
    *
-   * @param x the x coordinate of the menu, relative to the origin (top-left) of
-   *          the client area.
-   * @param y the y coordinate of the menu, relative to the origin (top-left) of
-   *          the client area.
+   * @param p the x, y coordinates of the menu, relative to the origin
+   *          (top-left) of the client area.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -2678,12 +2679,13 @@
    * devoid of allocations, etc.
    *
    * @param callback the function to call when doing a hit-test.
-   * @param callback_data an app-defined void pointer passed to **callback**.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
    *
    * @since This function is available since SDL 3.2.0.
+   *
+   * @cat listener-callback
    */
   void SetHitTest(HitTestCB callback);
 
@@ -3036,8 +3038,7 @@
    * Note that this function will appear to succeed, but not actually move the
    * mouse when used over Microsoft Remote Desktop.
    *
-   * @param x the x coordinate within the window.
-   * @param y the y coordinate within the window.
+   * @param p the x, y coordinates within the window.
    *
    * @threadsafety This function should only be called on the main thread.
    *
@@ -3390,9 +3391,6 @@
   /**
    * Delete an OpenGL context.
    *
-   * @returns true on success or false on failure; call GetError() for more
-   *          information.
-   *
    * @threadsafety This function should only be called on the main thread.
    *
    * @since This function is available since SDL 3.2.0.
@@ -3407,7 +3405,6 @@
    * The context must have been created with a compatible window.
    *
    * @param window the window to associate with the context.
-   * @param context the OpenGL context to associate with the window.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -3433,7 +3430,7 @@
   }
 
   /// Destructor
-  ~GLContextScoped() { Destroy(); }
+  ~GLContextScoped() { SDL_GL_DestroyContext(release()); }
 };
 
 /**
@@ -3808,17 +3805,19 @@
 /**
  * Get a list of currently connected displays.
  *
- * @param count a pointer filled in with the number of displays returned, may be
- *              nullptr.
  * @returns a 0 terminated array of display instance IDs or nullptr on failure;
- *          call GetError() for more information. This should be freed with
- *          free() when it is no longer needed.
+ *          call GetError() for more information.
  *
  * @threadsafety This function should only be called on the main thread.
  *
  * @since This function is available since SDL 3.2.0.
  */
-inline OwnArray<DisplayID> GetDisplays() { return SDL_GetDisplays(); }
+inline OwnArray<DisplayID> GetDisplays()
+{
+  int count = 0;
+  auto data = SDL_GetDisplays(&count);
+  return OwnArray<DisplayID>{data, size_t(count)};
+}
 
 /**
  * Return the primary display.
@@ -3878,7 +3877,7 @@
  */
 inline PropertiesRef GetDisplayProperties(DisplayID displayID)
 {
-  return CheckError(SDL_GetDisplayProperties(displayID));
+  return {CheckError(SDL_GetDisplayProperties(displayID))};
 }
 
 inline PropertiesRef Display::GetProperties() const
@@ -3935,7 +3934,7 @@
  * different location depending on monitor layout.
  *
  * @param displayID the instance ID of the display to query.
- * @param rect the Rect structure filled in with the display bounds.
+ * @returns the Rect structure filled in with the display bounds.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -3947,7 +3946,9 @@
  */
 inline Rect GetDisplayBounds(Display displayID)
 {
-  return CheckError(SDL_GetDisplayBounds(displayID));
+  Rect bounds;
+  SDL_GetDisplayBounds(displayID, &bounds);
+  return bounds;
 }
 
 inline Rect Display::GetBounds() const
@@ -3967,7 +3968,7 @@
  * window.
  *
  * @param displayID the instance ID of the display to query.
- * @param rect the Rect structure filled in with the display bounds.
+ * @returns the Rect structure filled in with the display bounds.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -3979,7 +3980,9 @@
  */
 inline Rect GetDisplayUsableBounds(Display displayID)
 {
-  return CheckError(SDL_GetDisplayUsableBounds(displayID));
+  Rect bounds;
+  SDL_GetDisplayUsableBounds(displayID, &bounds);
+  return bounds;
 }
 
 inline Rect Display::GetUsableBounds() const
@@ -4081,12 +4084,8 @@
  * - pixel density -> lowest to highest
  *
  * @param displayID the instance ID of the display to query.
- * @param count a pointer filled in with the number of display modes returned,
- *              may be nullptr.
  * @returns a nullptr terminated array of display mode pointers or nullptr on
- *          failure; call GetError() for more information. This is a single
- *          allocation that should be freed with free() when it is no longer
- *          needed.
+ *          failure; call GetError() for more information.
  *
  * @threadsafety This function should only be called on the main thread.
  *
@@ -4096,7 +4095,9 @@
  */
 inline OwnArray<DisplayMode*> GetFullscreenDisplayModes(Display displayID)
 {
-  return SDL_GetFullscreenDisplayModes(displayID);
+  int count = 0;
+  auto data = CheckError(SDL_GetFullscreenDisplayModes(displayID, &count));
+  return OwnArray<DisplayMode*>{data, size_t(count)};
 }
 
 inline OwnArray<DisplayMode*> Display::GetFullscreenModes() const
@@ -4115,14 +4116,12 @@
  * then false is returned.
  *
  * @param displayID the instance ID of the display to query.
- * @param w the width in pixels of the desired display mode.
- * @param h the height in pixels of the desired display mode.
+ * @param size the width and height in pixels of the desired display mode.
  * @param refresh_rate the refresh rate of the desired display mode, or 0.0f for
  *                     the desktop refresh rate.
  * @param include_high_density_modes boolean to include high density modes in
  *                                   the search.
- * @param closest a pointer filled in with the closest display mode equal to or
- *                larger than the desired mode.
+ * @returns the closest display mode equal to or larger than the desired mode.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -4138,8 +4137,14 @@
   float refresh_rate,
   bool include_high_density_modes)
 {
-  return CheckError(SDL_GetClosestFullscreenDisplayMode(
-    displayID, size, refresh_rate, include_high_density_modes));
+  DisplayMode mode;
+  CheckError(SDL_GetClosestFullscreenDisplayMode(displayID,
+                                                 size.x,
+                                                 size.y,
+                                                 refresh_rate,
+                                                 include_high_density_modes,
+                                                 &mode));
+  return mode;
 }
 
 inline DisplayMode Display::GetClosestFullscreenMode(
@@ -4172,7 +4177,7 @@
  */
 inline const DisplayMode& GetDesktopDisplayMode(DisplayID displayID)
 {
-  return SDL_GetDesktopDisplayMode(displayID);
+  return *SDL_GetDesktopDisplayMode(displayID);
 }
 
 inline const DisplayMode& Display::GetDesktopMode() const
@@ -4201,7 +4206,7 @@
  */
 inline const DisplayMode& GetCurrentDisplayMode(DisplayID displayID)
 {
-  return SDL_GetCurrentDisplayMode(displayID);
+  return *SDL_GetCurrentDisplayMode(displayID);
 }
 
 inline const DisplayMode& Display::GetCurrentMode() const
@@ -4225,7 +4230,7 @@
  */
 inline Display GetDisplayForPoint(const PointRaw& point)
 {
-  return SDL_GetDisplayForPoint(point);
+  return SDL_GetDisplayForPoint(&point);
 }
 
 inline Display Display::GetForPoint(const PointRaw& point)
@@ -4250,7 +4255,7 @@
  */
 inline Display GetDisplayForRect(const RectRaw& rect)
 {
-  return CheckError(SDL_GetDisplayForRect(rect));
+  return CheckError(SDL_GetDisplayForRect(&rect));
 }
 
 inline Display Display::GetForRect(const RectRaw& rect)
@@ -4399,7 +4404,7 @@
  */
 inline const DisplayMode& GetWindowFullscreenMode(WindowRef window)
 {
-  return SDL_GetWindowFullscreenMode(window);
+  return *SDL_GetWindowFullscreenMode(window);
 }
 
 inline const DisplayMode& Window::GetFullscreenMode() const
@@ -4411,7 +4416,6 @@
  * Get the raw ICC profile data for the screen the window is currently on.
  *
  * @param window the window to query.
- * @param size the size of the ICC profile.
  * @returns the raw ICC profile data on success.
  * @throws Error on failure.
  *
@@ -4421,7 +4425,8 @@
  */
 inline OwnPtr<void> GetWindowICCProfile(WindowRef window)
 {
-  return CheckError(SDL_GetWindowICCProfile(window));
+  size_t size;
+  return OwnPtr<void>{CheckError(SDL_GetWindowICCProfile(window, &size))};
 }
 
 inline OwnPtr<void> Window::GetICCProfile() const
@@ -4453,17 +4458,19 @@
 /**
  * Get a list of valid windows.
  *
- * @param count a pointer filled in with the number of windows returned, may be
- *              nullptr.
  * @returns a nullptr terminated array of Window pointers or nullptr on failure;
- *          call GetError() for more information. This is a single allocation
- *          that should be freed with free() when it is no longer needed.
+ *          call GetError() for more information.
  *
  * @threadsafety This function should only be called on the main thread.
  *
  * @since This function is available since SDL 3.2.0.
  */
-inline OwnArray<WindowRef> GetWindows() { return SDL_GetWindows(); }
+inline OwnArray<WindowRef> GetWindows()
+{
+  int count = 0;
+  auto data = CheckError(SDL_GetWindows(&count));
+  return OwnArray<WindowRef>{reinterpret_cast<WindowRef*>(data), size_t(count)};
+}
 
 /**
  * Create a window with the specified dimensions and flags.
@@ -4503,8 +4510,8 @@
  * - `WINDOW_TRANSPARENT`: window with transparent buffer
  * - `WINDOW_NOT_FOCUSABLE`: window should not be focusable
  *
- * The Window will be shown if WINDOW_HIDDEN is not set. If hidden at creation
- * time, Window.Show() can be used to show it later.
+ * The Window will be shown if WINDOW_HIDDEN is not set. If hidden at
+ * creation time, Window.Show() can be used to show it later.
  *
  * On Apple's macOS, you **must** set the NSHighResolutionCapable Info.plist
  * property to YES, otherwise you will not receive a High-DPI OpenGL canvas.
@@ -4535,8 +4542,7 @@
  * in a future version of SDL.
  *
  * @param title the title of the window, in UTF-8 encoding.
- * @param w the width of the window.
- * @param h the height of the window.
+ * @param size the width and height of the window.
  * @param flags 0, or one or more WindowFlags OR'd together.
  * @returns the window that was created or nullptr on failure; call GetError()
  *          for more information.
@@ -4560,7 +4566,7 @@
 inline Window::Window(StringParam title,
                       const PointRaw& size,
                       WindowFlags flags)
-  : m_resource(SDL_CreateWindow(title, size, flags))
+  : m_resource(SDL_CreateWindow(title, size.x, size.y, flags))
 {
 }
 
@@ -4568,7 +4574,8 @@
                       const PointRaw& offset,
                       const PointRaw& size,
                       WindowFlags flags)
-  : m_resource(SDL_CreatePopupWindow(parent, offset, size, flags))
+  : m_resource(
+      SDL_CreatePopupWindow(parent, offset.x, offset.y, size.x, size.y, flags))
 {
 }
 
@@ -4631,12 +4638,9 @@
  * hidden will be restored when the parent is shown.
  *
  * @param parent the parent of the window, must not be nullptr.
- * @param offset_x the x position of the popup window relative to the origin of
- *                 the parent.
- * @param offset_y the y position of the popup window relative to the origin of
- *                 the parent window.
- * @param w the width of the window.
- * @param h the height of the window.
+ * @param offset the x, y position of the popup window relative to the origin
+ *               of the parent.
+ * @param size the width and height of the window.
  * @param flags WINDOW_TOOLTIP or WINDOW_POPUP_MENU, and zero or more additional
  *              WindowFlags OR'd together.
  * @returns the window that was created or nullptr on failure; call GetError()
@@ -4966,6 +4970,10 @@
 constexpr auto OPENVR_OVERLAY_ID_NUMBER =
   SDL_PROP_WINDOW_OPENVR_OVERLAY_ID_NUMBER;
 
+#else
+
+constexpr auto OPENVR_OVERLAY_ID_NUMBER = SDL_PROP_WINDOW_OPENVR_OVERLAY_ID;
+
 #endif // SDL_VERSION_ATLEAST(3, 4, 0)
 
 constexpr auto VIVANTE_DISPLAY_POINTER =
@@ -5068,7 +5076,7 @@
  */
 inline WindowRef GetWindowFromID(WindowID id)
 {
-  return SDL_GetWindowFromID(id);
+  return {SDL_GetWindowFromID(id)};
 }
 
 inline WindowRef Window::FromID(WindowID id)
@@ -5225,7 +5233,7 @@
  */
 inline PropertiesRef GetWindowProperties(WindowRef window)
 {
-  return CheckError(SDL_GetWindowProperties(window));
+  return {CheckError(SDL_GetWindowProperties(window))};
 }
 
 inline PropertiesRef Window::GetProperties() const
@@ -5366,9 +5374,7 @@
  * this is just a request, it can be denied by the windowing system.
  *
  * @param window the window to reposition.
- * @param x the x coordinate of the window, or `WINDOWPOS_CENTERED` or
- *          `WINDOWPOS_UNDEFINED`.
- * @param y the y coordinate of the window, or `WINDOWPOS_CENTERED` or
+ * @param p the coordinates of the window, or `WINDOWPOS_CENTERED` or
  *          `WINDOWPOS_UNDEFINED`.
  * @throws Error on failure.
  *
@@ -5381,7 +5387,7 @@
  */
 inline void SetWindowPosition(WindowRef window, const PointRaw& p)
 {
-  CheckError(SDL_SetWindowPosition(window, p));
+  CheckError(SDL_SetWindowPosition(window, p.x, p.y));
 }
 
 inline void Window::SetPosition(const PointRaw& p)
@@ -5426,10 +5432,7 @@
  * in the `x` or `y` parameter.
  *
  * @param window the window to query.
- * @param x a pointer filled in with the x position of the window, may be
- *          nullptr.
- * @param y a pointer filled in with the y position of the window, may be
- *          nullptr.
+ * @returns the position on success.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5440,7 +5443,9 @@
  */
 inline Point GetWindowPosition(WindowRef window)
 {
-  static_assert(false, "Not implemented");
+  Point p;
+  GetWindowPosition(window, &p.x, &p.y);
+  return p;
 }
 
 inline void Window::GetPosition(int* x, int* y) const
@@ -5475,8 +5480,7 @@
  * request, it can be denied by the windowing system.
  *
  * @param window the window to change.
- * @param w the width of the window, must be > 0.
- * @param h the height of the window, must be > 0.
+ * @param size the width and height of the window, must be > 0.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5489,7 +5493,7 @@
  */
 inline void SetWindowSize(WindowRef window, const PointRaw& size)
 {
-  CheckError(SDL_SetWindowSize(window, size));
+  CheckError(SDL_SetWindowSize(window, size.x, size.y));
 }
 
 inline void Window::SetSize(const PointRaw& size)
@@ -5531,8 +5535,7 @@
  * Renderer.GetOutputSize() to get the real client area size in pixels.
  *
  * @param window the window to query the width and height from.
- * @param w a pointer filled in with the width of the window, may be nullptr.
- * @param h a pointer filled in with the height of the window, may be nullptr.
+ * @returns a point with width and height on success
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5546,7 +5549,9 @@
  */
 inline Point GetWindowSize(WindowRef window)
 {
-  static_assert(false, "Not implemented");
+  Point p;
+  GetWindowSize(window, &p.x, &p.y);
+  return p;
 }
 
 inline void Window::GetSize(int* w, int* h) const
@@ -5556,16 +5561,43 @@
 
 inline Point Window::GetSize() const { return SDL::GetWindowSize(m_resource); }
 
+/**
+ * @brief Request the window's position and size to be set.
+ *
+ * @param window the window to query the width and height from.
+ * @param rect the rect containing the new coordinates
+ * @throws Error on failure.
+ *
+ * @threadsafety This function should only be called on the main thread.
+ *
+ * @sa WindowRef.SetPosition()
+ * @sa WindowRef.SetSize()
+ */
 inline void SetWindowRect(WindowRef window, Rect rect)
 {
-  static_assert(false, "Not implemented");
+  SetWindowPosition(window, rect.GetTopLeft());
+  SetWindowSize(window, rect.GetSize());
 }
 
 inline void Window::SetRect(Rect rect) { SDL::SetWindowRect(m_resource, rect); }
 
+/**
+ * Get the position and client size of a window.
+ *
+ * This is the current position of the window as last reported by the
+ * windowing system.
+ *
+ * The window pixel size may differ from its window coordinate size if the
+ * window is on a high pixel density display. Use Window.GetSizeInPixels()
+ * or RendererRef.GetOutputSize() to get the real client area size in pixels.
+ *
+ * @param window the window to query the width and height from.
+ * @return Rect with the position and size
+ * @throws Error on failure.
+ */
 inline Rect GetWindowRect(WindowRef window)
 {
-  static_assert(false, "Not implemented");
+  return Rect{GetWindowPosition(window), GetWindowSize(window)};
 }
 
 inline Rect Window::GetRect() const { return SDL::GetWindowRect(m_resource); }
@@ -5581,8 +5613,6 @@
  * interactable content.
  *
  * @param window the window to query.
- * @param rect a pointer filled in with the client area that is safe for
- *             interactive content.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5591,7 +5621,9 @@
  */
 inline Rect GetWindowSafeArea(WindowRef window)
 {
-  return CheckError(SDL_GetWindowSafeArea(window));
+  Rect rect;
+  CheckError(SDL_GetWindowSafeArea(window, &rect));
+  return rect;
 }
 
 inline Rect Window::GetSafeArea() const
@@ -5752,10 +5784,8 @@
  * Get the size of a window's client area, in pixels.
  *
  * @param window the window from which the drawable size should be queried.
- * @param w a pointer to variable for storing the width in pixels, may be
- *          nullptr.
- * @param h a pointer to variable for storing the height in pixels, may be
- *          nullptr.
+ * @returns the size on success or std::nullopt on failure; call GetError()
+ * for more information.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5767,7 +5797,9 @@
  */
 inline Point GetWindowSizeInPixels(WindowRef window)
 {
-  static_assert(false, "Not implemented");
+  Point p;
+  GetWindowSizeInPixels(window, &p.x, &p.y);
+  return p;
 }
 
 inline void Window::GetSizeInPixels(int* w, int* h) const
@@ -5784,8 +5816,7 @@
  * Set the minimum size of a window's client area.
  *
  * @param window the window to change.
- * @param min_w the minimum width of the window, or 0 for no limit.
- * @param min_h the minimum height of the window, or 0 for no limit.
+ * @param p the minimum width and height of the window, or 0 for no limit.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5797,7 +5828,7 @@
  */
 inline void SetWindowMinimumSize(WindowRef window, const PointRaw& p)
 {
-  CheckError(SDL_SetWindowMinimumSize(window, p));
+  CheckError(SDL_SetWindowMinimumSize(window, p.x, p.y));
 }
 
 inline void Window::SetMinimumSize(const PointRaw& p)
@@ -5836,8 +5867,7 @@
  * Set the maximum size of a window's client area.
  *
  * @param window the window to change.
- * @param max_w the maximum width of the window, or 0 for no limit.
- * @param max_h the maximum height of the window, or 0 for no limit.
+ * @param p the maximum width and height of the window, or 0 for no limit.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5849,7 +5879,7 @@
  */
 inline void SetWindowMaximumSize(WindowRef window, const PointRaw& p)
 {
-  CheckError(SDL_SetWindowMaximumSize(window, p));
+  CheckError(SDL_SetWindowMaximumSize(window, p.x, p.y));
 }
 
 inline void Window::SetMaximumSize(const PointRaw& p)
@@ -6296,7 +6326,7 @@
  */
 inline Surface GetWindowSurface(WindowRef window)
 {
-  return CheckError(SDL_GetWindowSurface(window));
+  return Surface::Borrow(CheckError(SDL_GetWindowSurface(window)));
 }
 
 inline Surface Window::GetSurface()
@@ -6337,16 +6367,16 @@
   SDL::SetWindowSurfaceVSync(m_resource, vsync);
 }
 
+/// Constant to disable vsync
 constexpr int WINDOW_SURFACE_VSYNC_DISABLED = SDL_WINDOW_SURFACE_VSYNC_DISABLED;
 
+/// Constant to enable adaptive vsync
 constexpr int WINDOW_SURFACE_VSYNC_ADAPTIVE = SDL_WINDOW_SURFACE_VSYNC_ADAPTIVE;
 
 /**
  * Get VSync for the window surface.
  *
  * @param window the window to query.
- * @param vsync an int filled with the current vertical refresh sync interval.
- *              See Window.SetSurfaceVSync() for the meaning of the value.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -6357,7 +6387,9 @@
  */
 inline int GetWindowSurfaceVSync(WindowRef window)
 {
-  return CheckError(SDL_GetWindowSurfaceVSync(window));
+  int vsync;
+  CheckError(SDL_GetWindowSurfaceVSync(window, &vsync));
+  return vsync;
 }
 
 inline int Window::GetSurfaceVSync() const
@@ -6406,7 +6438,6 @@
  * @param window the window to update.
  * @param rects an array of Rect structures representing areas of the surface to
  *              copy, in pixels.
- * @param numrects the number of rectangles.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -6419,7 +6450,7 @@
 inline void UpdateWindowSurfaceRects(WindowRef window,
                                      SpanRef<const RectRaw> rects)
 {
-  CheckError(SDL_UpdateWindowSurfaceRects(window, rects));
+  CheckError(SDL_UpdateWindowSurfaceRects(window, rects.data(), rects.size()));
 }
 
 inline void Window::UpdateSurfaceRects(SpanRef<const RectRaw> rects)
@@ -6572,7 +6603,7 @@
  * @sa Window.SetMouseGrab
  * @sa Window.SetKeyboardGrab
  */
-inline WindowRef GetGrabbedWindow() { return SDL_GetGrabbedWindow(); }
+inline WindowRef GetGrabbedWindow() { return {SDL_GetGrabbedWindow()}; }
 
 inline WindowRef Window::GetGrabbed() { return SDL::GetGrabbedWindow(); }
 
@@ -6597,7 +6628,7 @@
  */
 inline void SetWindowMouseRect(WindowRef window, const RectRaw& rect)
 {
-  CheckError(SDL_SetWindowMouseRect(window, rect));
+  CheckError(SDL_SetWindowMouseRect(window, &rect));
 }
 
 inline void Window::SetMouseRect(const RectRaw& rect)
@@ -6785,9 +6816,7 @@
  * nothing.
  *
  * @param window the window for which the menu will be displayed.
- * @param x the x coordinate of the menu, relative to the origin (top-left) of
- *          the client area.
- * @param y the y coordinate of the menu, relative to the origin (top-left) of
+ * @param p the coordinates of the menu, relative to the origin (top-left) of
  *          the client area.
  * @throws Error on failure.
  *
@@ -6797,7 +6826,7 @@
  */
 inline void ShowWindowSystemMenu(WindowRef window, const PointRaw& p)
 {
-  CheckError(SDL_ShowWindowSystemMenu(window, p));
+  CheckError(SDL_ShowWindowSystemMenu(window, p.x, p.y));
 }
 
 inline void Window::ShowSystemMenu(const PointRaw& p)
@@ -6887,7 +6916,6 @@
  *
  * @param window the window to set hit-testing on.
  * @param callback the function to call when doing a hit-test.
- * @param callback_data an app-defined void pointer passed to **callback**.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -6896,7 +6924,7 @@
  */
 inline void SetWindowHitTest(WindowRef window, HitTestCB callback)
 {
-  static_assert(false, "Not implemented");
+  SetWindowHitTest(window, callback.wrapper, callback.data);
 }
 
 inline void Window::SetHitTest(HitTest callback, void* callback_data)
@@ -6970,8 +6998,8 @@
  * Sets the state of the progress bar for the given window’s taskbar icon.
  *
  * @param window the window whose progress state is to be modified.
- * @param state the progress state. `PROGRESS_STATE_NONE` stops displaying the
- *              progress bar.
+ * @param state the progress state. `PROGRESS_STATE_NONE` stops displaying
+ *              the progress bar.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -7384,7 +7412,7 @@
  */
 inline void GL_MakeCurrent(WindowRef window, GLContext context)
 {
-  CheckError(SDL_GL_MakeCurrent(window, context));
+  CheckError(SDL_GL_MakeCurrent(window, context.get()));
 }
 
 inline void Window::MakeCurrent(GLContext context)
@@ -7394,7 +7422,7 @@
 
 inline void GLContext::MakeCurrent(WindowRef window)
 {
-  SDL::Window::MakeCurrent(m_resource, window);
+  SDL::GL_MakeCurrent(window, m_resource);
 }
 
 /**
@@ -7410,7 +7438,7 @@
  */
 inline WindowRef GL_GetCurrentWindow()
 {
-  return CheckError(SDL_GL_GetCurrentWindow());
+  return {CheckError(SDL_GL_GetCurrentWindow())};
 }
 
 /**
