--- build/generated/SDL3pp_ttf.h
+++ include/SDL3pp/SDL3pp_ttf.h
@@ -1,17 +1,20 @@
 #ifndef SDL3PP_TTF_H_
 #define SDL3PP_TTF_H_
 
-#include <SDL3/SDL_ttf.h>
 #include "SDL3pp_gpu.h"
 #include "SDL3pp_render.h"
 #include "SDL3pp_surface.h"
 #include "SDL3pp_version.h"
 #include "SDL3pp_video.h"
 
+#if defined(SDL3PP_ENABLE_TTF) || defined(SDL3PP_DOC)
+
+#include <SDL3_ttf/SDL_ttf.h>
+
 namespace SDL {
 
 /**
- * @defgroup CategorySDLTTF Category SDLTTF
+ * @defgroup CategorySDLTTF True type font support
  *
  * Header file for SDL_ttf library
  *
@@ -48,7 +51,7 @@
  *
  * This does not take ownership!
  */
-using TextEngineRef = TextEngine;
+using TextEngineRef = ResourceRefT<TextEngine>;
 
 // Forward decl
 struct TextBase;
@@ -72,25 +75,39 @@
 /// Safely wrap Text for non owning const parameters
 using TextConstRef = ResourceConstRef<TextRaw, TextRawConst>;
 
-/// Printable format: "%d.%d.%d", MAJOR, MINOR, MICRO
+#ifdef SDL3PP_DOC
+
+/**
+ * @name TTF version
+ * @{
+ * Printable format: "%d.%d.%d", MAJOR, MINOR, MICRO
+ */
 #define SDL_TTF_MAJOR_VERSION
 
 #define SDL_TTF_MINOR_VERSION
 
 #define SDL_TTF_MICRO_VERSION
 
-/// This is the version number macro for the current SDL_ttf version.
+///@}
+
+/**
+ * This is the version number macro for the current SDL_ttf version.
+ */
 #define SDL_TTF_VERSION                                                        \
   SDL_VERSIONNUM(                                                              \
     SDL_TTF_MAJOR_VERSION, SDL_TTF_MINOR_VERSION, SDL_TTF_MICRO_VERSION)
 
-/// This macro will evaluate to true if compiled with SDL_ttf at least X.Y.Z.
+/**
+ * This macro will evaluate to true if compiled with SDL_ttf at least X.Y.Z.
+ */
 #define SDL_TTF_VERSION_ATLEAST(X, Y, Z)                                       \
   ((SDL_TTF_MAJOR_VERSION >= X) &&                                             \
    (SDL_TTF_MAJOR_VERSION > X || SDL_TTF_MINOR_VERSION >= Y) &&                \
    (SDL_TTF_MAJOR_VERSION > X || SDL_TTF_MINOR_VERSION > Y ||                  \
     SDL_TTF_MICRO_VERSION >= Z))
 
+#endif // SDL3PP_DOC
+
 namespace TTF {
 
 /**
@@ -295,7 +312,7 @@
  *
  * The values here are chosen to match
  * [hb_direction_t](https://harfbuzz.github.io/harfbuzz-hb-common.html#hb-direction-t)
- * .
+ *
  *
  * @since This enum is available since SDL_ttf 3.0.0.
  *
@@ -368,8 +385,6 @@
    * The copy will be distinct from the original, but will share the font file
    * and have the same size and style as the original.
    *
-   * When done with the returned Font, use CloseFont() to dispose of it.
-   *
    * @returns a valid Font on success.
    * @throws Error on failure.
    *
@@ -731,7 +746,7 @@
    * @threadsafety This function should be called on the thread that created the
    *               font.
    *
-   * @since This function is available since SDL_ttf 3.4.0.
+   * @since This function is available since SDL_ttf 3.2.2.
    */
   int GetWeight() const;
 
@@ -974,7 +989,7 @@
    * This updates any Text objects using this font.
    *
    * @param script an [ISO 15924
-   *               code](https://unicode.org/iso15924/iso15924-codes.html) .
+   *               code](https://unicode.org/iso15924/iso15924-codes.html).
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -1126,8 +1141,7 @@
    *
    * @param previous_ch the previous codepoint.
    * @param ch the current codepoint.
-   * @param kerning a pointer filled in with the kerning size between the two
-   *                glyphs, in pixels, may be nullptr.
+   * @returns the kerning size between the two glyphs, in pixels.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -1137,9 +1151,26 @@
    */
   int GetGlyphKerning(Uint32 previous_ch, Uint32 ch) const;
 
+  /**
+   * Calculate the dimensions of a rendered string of UTF-8 text.
+   *
+   * This will report the width and height, in pixels, of the space that the
+   * specified string will take to fully render.
+   *
+   * @param text text to calculate, in UTF-8 encoding.
+   * @returns return a Point with the width, height.
+   * @throws Error on failure.
+   *
+   * @threadsafety This function should be called on the thread that created the
+   *               font.
+   *
+   * @since This function is available since SDL_ttf 3.0.0.
+   */
   Point GetStringSize(std::string_view text) const
   {
-    static_assert(false, "Not implemented");
+    Point p;
+    GetStringSize(text, &p.x, &p.y);
+    return p;
   }
 
   /**
@@ -1148,10 +1179,7 @@
    * This will report the width and height, in pixels, of the space that the
    * specified string will take to fully render.
    *
-   * @param font the font to query.
    * @param text text to calculate, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param w will be filled with width, in pixels, on return.
    * @param h will be filled with height, in pixels, on return.
    * @throws Error on failure.
@@ -1163,9 +1191,32 @@
    */
   void GetStringSize(std::string_view text, int* w, int* h) const;
 
+  /**
+   * Calculate the dimensions of a rendered string of UTF-8 text.
+   *
+   * This will report the width and height, in pixels, of the space that the
+   * specified string will take to fully render.
+   *
+   * Text is wrapped to multiple lines on line endings and on word boundaries if
+   * it extends beyond `wrap_width` in pixels.
+   *
+   * If wrap_width is 0, this function will only wrap on newline characters.
+   *
+   * @param text text to calculate, in UTF-8 encoding.
+   * @param wrap_width the maximum width or 0 to wrap on newline characters.
+   * @returns return a Point with the width, height on success.
+   * @throws Error on failure.
+   *
+   * @threadsafety This function should be called on the thread that created the
+   *               font.
+   *
+   * @since This function is available since SDL_ttf 3.0.0.
+   */
   Point GetStringSizeWrapped(std::string_view text, int wrap_width) const
   {
-    static_assert(false, "Not implemented");
+    Point p;
+    GetStringSizeWrapped(text, wrap_width, &p.x, &p.y);
+    return p;
   }
 
   /**
@@ -1179,10 +1230,7 @@
    *
    * If wrap_width is 0, this function will only wrap on newline characters.
    *
-   * @param font the font to query.
    * @param text text to calculate, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param wrap_width the maximum width or 0 to wrap on newline characters.
    * @param w will be filled with width, in pixels, on return.
    * @param h will be filled with height, in pixels, on return.
@@ -1206,10 +1254,7 @@
    *
    * This does not need to render the string to do this calculation.
    *
-   * @param font the font to query.
    * @param text text to calculate, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param max_width maximum width, in pixels, available for the string, or 0
    *                  for unbounded width.
    * @param measured_width a pointer filled in with the width, in pixels, of the
@@ -1245,10 +1290,7 @@
    * You can render at other quality levels with RenderText_Shaded,
    * RenderText_Blended, and RenderText_LCD.
    *
-   * @param font the font to render with.
    * @param text text to render, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param fg the foreground color for the text.
    * @returns a new 8-bit, palettized surface, or nullptr if there was an error.
    *
@@ -1280,10 +1322,7 @@
    * You can render at other quality levels with RenderText_Shaded_Wrapped,
    * RenderText_Blended_Wrapped, and RenderText_LCD_Wrapped.
    *
-   * @param font the font to render with.
    * @param text text to render, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param fg the foreground color for the text.
    * @param wrapLength the maximum width of the text surface or 0 to wrap on
    *                   newline characters.
@@ -1349,10 +1388,7 @@
    * You can render at other quality levels with RenderText_Solid,
    * RenderText_Blended, and RenderText_LCD.
    *
-   * @param font the font to render with.
    * @param text text to render, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param fg the foreground color for the text.
    * @param bg the background color for the text.
    * @returns a new 8-bit, palettized surface, or nullptr if there was an error.
@@ -1385,10 +1421,7 @@
    * You can render at other quality levels with RenderText_Solid_Wrapped,
    * RenderText_Blended_Wrapped, and RenderText_LCD_Wrapped.
    *
-   * @param font the font to render with.
    * @param text text to render, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param fg the foreground color for the text.
    * @param bg the background color for the text.
    * @param wrap_width the maximum width of the text surface or 0 to wrap on
@@ -1457,10 +1490,7 @@
    * You can render at other quality levels with RenderText_Solid,
    * RenderText_Shaded, and RenderText_LCD.
    *
-   * @param font the font to render with.
    * @param text text to render, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param fg the foreground color for the text.
    * @returns a new 32-bit, ARGB surface, or nullptr if there was an error.
    *
@@ -1491,10 +1521,7 @@
    * You can render at other quality levels with RenderText_Solid_Wrapped,
    * RenderText_Shaded_Wrapped, and RenderText_LCD_Wrapped.
    *
-   * @param font the font to render with.
    * @param text text to render, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param fg the foreground color for the text.
    * @param wrap_width the maximum width of the text surface or 0 to wrap on
    *                   newline characters.
@@ -1559,10 +1586,7 @@
    * You can render at other quality levels with RenderText_Solid,
    * RenderText_Shaded, and RenderText_Blended.
    *
-   * @param font the font to render with.
    * @param text text to render, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param fg the foreground color for the text.
    * @param bg the background color for the text.
    * @returns a new 32-bit, ARGB surface, or nullptr if there was an error.
@@ -1595,10 +1619,7 @@
    * You can render at other quality levels with RenderText_Solid_Wrapped,
    * RenderText_Shaded_Wrapped, and RenderText_Blended_Wrapped.
    *
-   * @param font the font to render with.
    * @param text text to render, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param fg the foreground color for the text.
    * @param bg the background color for the text.
    * @param wrap_width the maximum width of the text surface or 0 to wrap on
@@ -1687,8 +1708,6 @@
    * size becomes the index of choosing which size. If the value is too high,
    * the last indexed size will be the default.
    *
-   * When done with the returned Font, use CloseFont() to dispose of it.
-   *
    * @param file path to font file.
    * @param ptsize point size to use for the newly-opened font.
    * @post a valid Font on success.
@@ -1710,14 +1729,12 @@
    * the last indexed size will be the default.
    *
    * If `closeio` is true, `src` will be automatically closed once the font is
-   * closed. Otherwise you should close `src` yourself after closing the font.
-   *
-   * When done with the returned Font, use CloseFont() to dispose of it.
+   * closed. Otherwise you should keep `src` open until the font is closed.
    *
    * @param src an IOStream to provide a font file's data.
+   * @param ptsize point size to use for the newly-opened font.
    * @param closeio true to close `src` when the font is closed, false to leave
    *                it open.
-   * @param ptsize point size to use for the newly-opened font.
    * @post a valid Font on success.
    * @throws Error on failure.
    *
@@ -1815,7 +1832,7 @@
 }
 
 inline Font::Font(IOStreamRef src, float ptsize, bool closeio)
-  : Font(CheckError(TTF_OpenFontIO(src, ptsize, closeio)))
+  : Font(CheckError(TTF_OpenFontIO(src, closeio, ptsize)))
 {
 }
 
@@ -1832,14 +1849,14 @@
  * last indexed size will be the default.
  *
  * If `closeio` is true, `src` will be automatically closed once the font is
- * closed. Otherwise you should close `src` yourself after closing the font.
+ * closed. Otherwise you should keep `src` open until the font is closed.
  *
  * When done with the returned Font, use CloseFont() to dispose of it.
  *
  * @param src an IOStream to provide a font file's data.
+ * @param ptsize point size to use for the newly-opened font.
  * @param closeio true to close `src` when the font is closed, false to leave it
  *                open.
- * @param ptsize point size to use for the newly-opened font.
  * @returns a valid Font on success.
  * @throws Error on failure.
  *
@@ -1898,6 +1915,11 @@
  */
 inline Font OpenFontWithProperties(PropertiesRef props) { return Font(props); }
 
+/**
+ * Properties for Font creation.
+ *
+ * @sa OpenFontWithProperties
+ */
 namespace prop::Font::Create {
 
 constexpr auto FILENAME_STRING =
@@ -1955,7 +1977,7 @@
  */
 inline Font CopyFont(FontRef existing_font)
 {
-  return CheckError(TTF_CopyFont(existing_font));
+  return Font(CheckError(TTF_CopyFont(existing_font)));
 }
 
 inline Font FontBase::Copy() const { return SDL::CopyFont(get()); }
@@ -1990,6 +2012,16 @@
   return SDL::GetFontProperties(get());
 }
 
+/**
+ * Properties for Font.
+ *
+ * Properties within subnamespace Create are used when creating a font with
+ * OpenFontWithProperties(). Poperties without the prefix can retrieved by using
+ * Font.GetProperties() on an existing font.
+ *
+ * @sa OpenFontWithProperties
+ * @sa Font.GetProperties
+ */
 namespace prop::Font {
 
 constexpr auto OUTLINE_LINE_CAP_NUMBER =
@@ -2450,7 +2482,7 @@
  * @threadsafety This function should be called on the thread that created the
  *               font.
  *
- * @since This function is available since SDL_ttf 3.4.0.
+ * @since This function is available since SDL_ttf 3.2.2.
  */
 inline int GetFontWeight(FontRef font) { return TTF_GetFontWeight(font); }
 
@@ -2862,7 +2894,7 @@
  *
  * @param font the font to modify.
  * @param script an [ISO 15924
- *               code](https://unicode.org/iso15924/iso15924-codes.html) .
+ *               code](https://unicode.org/iso15924/iso15924-codes.html).
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -2991,7 +3023,7 @@
  */
 inline Surface GetGlyphImage(FontRef font, Uint32 ch, ImageType* image_type)
 {
-  return CheckError(TTF_GetGlyphImage(font, ch, image_type));
+  return Surface{CheckError(TTF_GetGlyphImage(font, ch, image_type))};
 }
 
 inline Surface FontBase::GetGlyphImage(Uint32 ch, ImageType* image_type) const
@@ -3021,7 +3053,8 @@
                                      Uint32 glyph_index,
                                      ImageType* image_type)
 {
-  return CheckError(TTF_GetGlyphImageForIndex(font, glyph_index, image_type));
+  return Surface(
+    CheckError(TTF_GetGlyphImageForIndex(font, glyph_index, image_type)));
 }
 
 inline Surface FontBase::GetGlyphImageForIndex(Uint32 glyph_index,
@@ -3085,8 +3118,7 @@
  * @param font the font to query.
  * @param previous_ch the previous codepoint.
  * @param ch the current codepoint.
- * @param kerning a pointer filled in with the kerning size between the two
- *                glyphs, in pixels, may be nullptr.
+ * @returns the kerning size between the two glyphs, in pixels.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -3096,7 +3128,8 @@
  */
 inline int GetGlyphKerning(FontRef font, Uint32 previous_ch, Uint32 ch)
 {
-  return CheckError(TTF_GetGlyphKerning(font, previous_ch, ch));
+  if (int r; TTF_GetGlyphKerning(font, previous_ch, ch, &r)) return r;
+  throw Error();
 }
 
 inline int FontBase::GetGlyphKerning(Uint32 previous_ch, Uint32 ch) const
@@ -3112,8 +3145,6 @@
  *
  * @param font the font to query.
  * @param text text to calculate, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param w will be filled with width, in pixels, on return.
  * @param h will be filled with height, in pixels, on return.
  * @throws Error on failure.
@@ -3125,7 +3156,7 @@
  */
 inline void GetStringSize(FontRef font, std::string_view text, int* w, int* h)
 {
-  CheckError(TTF_GetStringSize(font, text, w, h));
+  CheckError(TTF_GetStringSize(font, text.data(), text.size(), w, h));
 }
 
 inline void FontBase::GetStringSize(std::string_view text, int* w, int* h) const
@@ -3146,8 +3177,6 @@
  *
  * @param font the font to query.
  * @param text text to calculate, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param wrap_width the maximum width or 0 to wrap on newline characters.
  * @param w will be filled with width, in pixels, on return.
  * @param h will be filled with height, in pixels, on return.
@@ -3164,7 +3193,8 @@
                                  int* w,
                                  int* h)
 {
-  CheckError(TTF_GetStringSizeWrapped(font, text, wrap_width, w, h));
+  CheckError(
+    TTF_GetStringSizeWrapped(font, text.data(), text.size(), wrap_width, w, h));
 }
 
 inline void FontBase::GetStringSizeWrapped(std::string_view text,
@@ -3185,8 +3215,6 @@
  *
  * @param font the font to query.
  * @param text text to calculate, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param max_width maximum width, in pixels, available for the string, or 0 for
  *                  unbounded width.
  * @param measured_width a pointer filled in with the width, in pixels, of the
@@ -3206,8 +3234,12 @@
                           int* measured_width,
                           size_t* measured_length)
 {
-  CheckError(
-    TTF_MeasureString(font, text, max_width, measured_width, measured_length));
+  CheckError(TTF_MeasureString(font,
+                               text.data(),
+                               text.size(),
+                               max_width,
+                               measured_width,
+                               measured_length));
 }
 
 inline void FontBase::MeasureString(std::string_view text,
@@ -3237,8 +3269,6 @@
  *
  * @param font the font to render with.
  * @param text text to render, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param fg the foreground color for the text.
  * @returns a new 8-bit, palettized surface, or nullptr if there was an error.
  *
@@ -3255,7 +3285,7 @@
  */
 inline Surface RenderText_Solid(FontRef font, std::string_view text, Color fg)
 {
-  return TTF_RenderText_Solid(font, text, fg);
+  return Surface{TTF_RenderText_Solid(font, text.data(), text.size(), fg)};
 }
 
 inline Surface FontBase::RenderText_Solid(std::string_view text, Color fg) const
@@ -3280,8 +3310,6 @@
  *
  * @param font the font to render with.
  * @param text text to render, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param fg the foreground color for the text.
  * @param wrapLength the maximum width of the text surface or 0 to wrap on
  *                   newline characters.
@@ -3302,7 +3330,8 @@
                                         Color fg,
                                         int wrapLength)
 {
-  return TTF_RenderText_Solid_Wrapped(font, text, fg, wrapLength);
+  return Surface(TTF_RenderText_Solid_Wrapped(
+    font, text.data(), text.size(), fg, wrapLength));
 }
 
 inline Surface FontBase::RenderText_Solid_Wrapped(std::string_view text,
@@ -3341,7 +3370,7 @@
  */
 inline Surface RenderGlyph_Solid(FontRef font, Uint32 ch, ColorRaw fg)
 {
-  return TTF_RenderGlyph_Solid(font, ch, fg);
+  return Surface(TTF_RenderGlyph_Solid(font, ch, fg));
 }
 
 inline Surface FontBase::RenderGlyph_Solid(Uint32 ch, ColorRaw fg) const
@@ -3369,8 +3398,6 @@
  *
  * @param font the font to render with.
  * @param text text to render, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param fg the foreground color for the text.
  * @param bg the background color for the text.
  * @returns a new 8-bit, palettized surface, or nullptr if there was an error.
@@ -3390,7 +3417,7 @@
                                  Color fg,
                                  Color bg)
 {
-  return TTF_RenderText_Shaded(font, text, fg, bg);
+  return Surface(TTF_RenderText_Shaded(font, text.data(), text.size(), fg, bg));
 }
 
 inline Surface FontBase::RenderText_Shaded(std::string_view text,
@@ -3418,8 +3445,6 @@
  *
  * @param font the font to render with.
  * @param text text to render, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param fg the foreground color for the text.
  * @param bg the background color for the text.
  * @param wrap_width the maximum width of the text surface or 0 to wrap on
@@ -3442,7 +3467,8 @@
                                          Color bg,
                                          int wrap_width)
 {
-  return TTF_RenderText_Shaded_Wrapped(font, text, fg, bg, wrap_width);
+  return Surface(TTF_RenderText_Shaded_Wrapped(
+    font, text.data(), text.size(), fg, bg, wrap_width));
 }
 
 inline Surface FontBase::RenderText_Shaded_Wrapped(std::string_view text,
@@ -3487,7 +3513,7 @@
                                   ColorRaw fg,
                                   ColorRaw bg)
 {
-  return TTF_RenderGlyph_Shaded(font, ch, fg, bg);
+  return Surface(TTF_RenderGlyph_Shaded(font, ch, fg, bg));
 }
 
 inline Surface FontBase::RenderGlyph_Shaded(Uint32 ch,
@@ -3516,8 +3542,6 @@
  *
  * @param font the font to render with.
  * @param text text to render, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param fg the foreground color for the text.
  * @returns a new 32-bit, ARGB surface, or nullptr if there was an error.
  *
@@ -3533,7 +3557,7 @@
  */
 inline Surface RenderText_Blended(FontRef font, std::string_view text, Color fg)
 {
-  return TTF_RenderText_Blended(font, text, fg);
+  return Surface(TTF_RenderText_Blended(font, text.data(), text.size(), fg));
 }
 
 inline Surface FontBase::RenderText_Blended(std::string_view text,
@@ -3559,8 +3583,6 @@
  *
  * @param font the font to render with.
  * @param text text to render, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param fg the foreground color for the text.
  * @param wrap_width the maximum width of the text surface or 0 to wrap on
  *                   newline characters.
@@ -3581,7 +3603,8 @@
                                           Color fg,
                                           int wrap_width)
 {
-  return TTF_RenderText_Blended_Wrapped(font, text, fg, wrap_width);
+  return Surface(TTF_RenderText_Blended_Wrapped(
+    font, text.data(), text.size(), fg, wrap_width));
 }
 
 inline Surface FontBase::RenderText_Blended_Wrapped(std::string_view text,
@@ -3620,7 +3643,7 @@
  */
 inline Surface RenderGlyph_Blended(FontRef font, Uint32 ch, ColorRaw fg)
 {
-  return TTF_RenderGlyph_Blended(font, ch, fg);
+  return Surface(TTF_RenderGlyph_Blended(font, ch, fg));
 }
 
 inline Surface FontBase::RenderGlyph_Blended(Uint32 ch, ColorRaw fg) const
@@ -3646,8 +3669,6 @@
  *
  * @param font the font to render with.
  * @param text text to render, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param fg the foreground color for the text.
  * @param bg the background color for the text.
  * @returns a new 32-bit, ARGB surface, or nullptr if there was an error.
@@ -3667,7 +3688,7 @@
                               Color fg,
                               Color bg)
 {
-  return TTF_RenderText_LCD(font, text, fg, bg);
+  return Surface(TTF_RenderText_LCD(font, text.data(), text.size(), fg, bg));
 }
 
 inline Surface FontBase::RenderText_LCD(std::string_view text,
@@ -3694,8 +3715,6 @@
  *
  * @param font the font to render with.
  * @param text text to render, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param fg the foreground color for the text.
  * @param bg the background color for the text.
  * @param wrap_width the maximum width of the text surface or 0 to wrap on
@@ -3718,7 +3737,8 @@
                                       Color bg,
                                       int wrap_width)
 {
-  return TTF_RenderText_LCD_Wrapped(font, text, fg, bg, wrap_width);
+  return Surface(TTF_RenderText_LCD_Wrapped(
+    font, text.data(), text.size(), fg, bg, wrap_width));
 }
 
 inline Surface FontBase::RenderText_LCD_Wrapped(std::string_view text,
@@ -3763,7 +3783,7 @@
                                ColorRaw fg,
                                ColorRaw bg)
 {
-  return TTF_RenderGlyph_LCD(font, ch, fg, bg);
+  return Surface(TTF_RenderGlyph_LCD(font, ch, fg, bg));
 }
 
 inline Surface FontBase::RenderGlyph_LCD(Uint32 ch,
@@ -3837,40 +3857,10 @@
   using ResourceBaseT::ResourceBaseT;
 
   /**
-   * Constructs from raw TextEngine.
-   *
-   * @param resource a TextEngineRaw to be wrapped.
-   *
-   * This assumes the ownership, call release() if you need to take back.
-   */
-  constexpr explicit TextEngine(TextEngineRaw resource) noexcept
-    : TextEngineBase(resource)
-  {
-  }
-
-  /// Move constructor
-  constexpr TextEngine(TextEngine&& other) noexcept
-    : TextEngine(other.release())
-  {
-  }
-
-  /// Destructor
-  ~TextEngine() { Destroy(get()); }
-
-  /// Assignment operator.
-  constexpr TextEngine& operator=(TextEngine&& other) noexcept
-  {
-    swap(*this, other);
-    return *this;
-  }
-
-  /**
    * Create a text object from UTF-8 text and a text engine.
    *
    * @param font the font to render with.
    * @param text the text to use, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @returns a Text object or nullptr on failure; call GetError() for more
    *          information.
    *
@@ -3884,8 +3874,36 @@
   Text CreateText(FontRef font, std::string_view text);
 };
 
+/// A surface based text engine
 struct SurfaceTextEngine : TextEngine
 {
+  using TextEngine::TextEngine;
+
+  /**
+   * Constructs from raw TextEngine.
+   *
+   * @param resource a TextEngineRaw to be wrapped.
+   *
+   * This assumes the ownership, call release() if you need to take back.
+   */
+  constexpr explicit SurfaceTextEngine(TextEngineRaw resource) noexcept
+    : TextEngine(resource)
+  {
+  }
+
+  /// Move constructor
+  constexpr SurfaceTextEngine(SurfaceTextEngine&& other) noexcept
+    : TextEngine(other.release())
+  {
+  }
+
+  /// Move assignment operator
+  constexpr SurfaceTextEngine& operator=(SurfaceTextEngine&& other) noexcept
+  {
+    swap(*this, other);
+    return *this;
+  }
+
   /**
    * Create a text engine for drawing text on SDL surfaces.
    *
@@ -3901,6 +3919,7 @@
    */
   SurfaceTextEngine();
 
+  /// Destructor
   ~SurfaceTextEngine() { Destroy(); }
 
   /**
@@ -3909,8 +3928,6 @@
    * All text created by this engine should be destroyed before calling this
    * function.
    *
-   * @param engine a TextEngine object created with CreateSurfaceTextEngine().
-   *
    * @threadsafety This function should be called on the thread that created the
    *               engine.
    *
@@ -3921,8 +3938,36 @@
   void Destroy();
 };
 
+/// A renderer based text engine
 struct RendererTextEngine : TextEngine
 {
+  using TextEngine::TextEngine;
+
+  /**
+   * Constructs from raw TextEngine.
+   *
+   * @param resource a TextEngineRaw to be wrapped.
+   *
+   * This assumes the ownership, call release() if you need to take back.
+   */
+  constexpr explicit RendererTextEngine(TextEngineRaw resource) noexcept
+    : TextEngine(resource)
+  {
+  }
+
+  /// Move constructor
+  constexpr RendererTextEngine(RendererTextEngine&& other) noexcept
+    : TextEngine(other.release())
+  {
+  }
+
+  /// Move assignment operator
+  constexpr RendererTextEngine& operator=(RendererTextEngine&& other) noexcept
+  {
+    swap(*this, other);
+    return *this;
+  }
+
   /**
    * Create a text engine for drawing text on an SDL renderer.
    *
@@ -3967,6 +4012,7 @@
    */
   RendererTextEngine(PropertiesRef props);
 
+  /// Destructor
   ~RendererTextEngine() { Destroy(); }
 
   /**
@@ -3975,8 +4021,6 @@
    * All text created by this engine should be destroyed before calling this
    * function.
    *
-   * @param engine a TextEngine object created with CreateRendererTextEngine().
-   *
    * @threadsafety This function should be called on the thread that created the
    *               engine.
    *
@@ -3987,8 +4031,36 @@
   void Destroy();
 };
 
+/// A GPU based text engine
 struct GPUTextEngine : TextEngine
 {
+  using TextEngine::TextEngine;
+
+  /**
+   * Constructs from raw TextEngine.
+   *
+   * @param resource a TextEngineRaw to be wrapped.
+   *
+   * This assumes the ownership, call release() if you need to take back.
+   */
+  constexpr explicit GPUTextEngine(TextEngineRaw resource) noexcept
+    : TextEngine(resource)
+  {
+  }
+
+  /// Move constructor
+  constexpr GPUTextEngine(GPUTextEngine&& other) noexcept
+    : TextEngine(other.release())
+  {
+  }
+
+  /// Move assignment operator
+  constexpr GPUTextEngine& operator=(GPUTextEngine&& other) noexcept
+  {
+    swap(*this, other);
+    return *this;
+  }
+
   /**
    * Create a text engine for drawing text with the SDL GPU API.
    *
@@ -4033,13 +4105,13 @@
    */
   GPUTextEngine(PropertiesRef props);
 
+  /// Destructor
   ~GPUTextEngine() { Destroy(); }
 
   /**
    * Sets the winding order of the vertices returned by GetGPUTextDrawData for a
    * particular GPU text engine.
    *
-   * @param engine a TextEngine object created with CreateGPUTextEngine().
    * @param winding the new winding order option.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4049,13 +4121,12 @@
    *
    * @sa GPUTextEngine.GetGPUWinding
    */
-  static void SetGPUWinding(TextEngineRef engine, GPUTextEngineWinding winding);
+  void SetGPUWinding(GPUTextEngineWinding winding);
 
   /**
    * Get the winding order of the vertices returned by GetGPUTextDrawData for a
    * particular GPU text engine
    *
-   * @param engine a TextEngine object created with CreateGPUTextEngine().
    * @returns the winding order used by the GPU text engine or
    *          GPU_TEXTENGINE_WINDING_INVALID in case of error.
    *
@@ -4066,7 +4137,7 @@
    *
    * @sa GPUTextEngine.SetGPUWinding
    */
-  GPUTextEngineWinding GetGPUWinding(TextEngineRef engine) const;
+  GPUTextEngineWinding GetGPUWinding() const;
 
   /**
    * Destroy a text engine created for drawing text with the SDL GPU API.
@@ -4074,8 +4145,6 @@
    * All text created by this engine should be destroyed before calling this
    * function.
    *
-   * @param engine a TextEngine object created with CreateGPUTextEngine().
-   *
    * @threadsafety This function should be called on the thread that created the
    *               engine.
    *
@@ -4149,11 +4218,8 @@
    * `text` must have been created using a TextEngine from
    * CreateSurfaceTextEngine().
    *
-   * @param text the text to draw.
-   * @param x the x coordinate in pixels, positive from the left edge towards
-   *          the right.
-   * @param y the y coordinate in pixels, positive from the top edge towards the
-   *          bottom.
+   * @param p the x,y coordinates in pixels, positive from the top-left edge
+   *          towards the bottom-right.
    * @param surface the surface to draw on.
    * @throws Error on failure.
    *
@@ -4174,11 +4240,8 @@
    * CreateRendererTextEngine(), and will draw using the renderer passed to that
    * function.
    *
-   * @param text the text to draw.
-   * @param x the x coordinate in pixels, positive from the left edge towards
-   *          the right.
-   * @param y the y coordinate in pixels, positive from the top edge towards the
-   *          bottom.
+   * @param p the x,y coordinates in pixels, positive from the top-left edge
+   *          towards the bottom-right.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4337,7 +4400,7 @@
    * This returns false if SDL_ttf isn't built with HarfBuzz support.
    *
    * @param script an [ISO 15924
-   *               code](https://unicode.org/iso15924/iso15924-codes.html) .
+   *               code](https://unicode.org/iso15924/iso15924-codes.html).
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4372,11 +4435,7 @@
    *
    * The default text color is white (255, 255, 255, 255).
    *
-   * @param text the Text to modify.
-   * @param r the red color value in the range of 0-255.
-   * @param g the green color value in the range of 0-255.
-   * @param b the blue color value in the range of 0-255.
-   * @param a the alpha value in the range of 0-255.
+   * @param c the color values in the range of 0-255.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4394,11 +4453,7 @@
    *
    * The default text color is white (1.0f, 1.0f, 1.0f, 1.0f).
    *
-   * @param text the Text to modify.
-   * @param r the red color value, normally in the range of 0-1.
-   * @param g the green color value, normally in the range of 0-1.
-   * @param b the blue color value, normally in the range of 0-1.
-   * @param a the alpha value in the range of 0-1.
+   * @param c the color value, normally in the range of 0-1.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4437,14 +4492,7 @@
   /**
    * Get the color of a text object.
    *
-   * @param r a pointer filled in with the red color value in the range of
-   *          0-255, may be nullptr.
-   * @param g a pointer filled in with the green color value in the range of
-   *          0-255, may be nullptr.
-   * @param b a pointer filled in with the blue color value in the range of
-   *          0-255, may be nullptr.
-   * @param a a pointer filled in with the alpha value in the range of 0-255,
-   *          may be nullptr.
+   * @return The color in the range of 0-255 on success.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4455,7 +4503,7 @@
    * @sa GetTextColorFloat
    * @sa SetTextColor
    */
-  Color GetColor();
+  Color GetColor() const;
 
   /**
    * Get the color of a text object.
@@ -4483,14 +4531,7 @@
   /**
    * Get the color of a text object.
    *
-   * @param r a pointer filled in with the red color value, normally in the
-   *          range of 0-1, may be nullptr.
-   * @param g a pointer filled in with the green color value, normally in the
-   *          range of 0-1, may be nullptr.
-   * @param b a pointer filled in with the blue color value, normally in the
-   *          range of 0-1, may be nullptr.
-   * @param a a pointer filled in with the alpha value in the range of 0-1, may
-   *          be nullptr.
+   * @return The color in the range of 0-1 on success.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4501,7 +4542,7 @@
    * @sa GetTextColor
    * @sa SetTextColorFloat
    */
-  FColor GetColorFloat();
+  FColor GetColorFloat() const;
 
   /**
    * Set the position of a text object.
@@ -4511,8 +4552,8 @@
    *
    * This function may cause the internal text representation to be rebuilt.
    *
-   * @param x the x offset of the upper left corner of this text in pixels.
-   * @param y the y offset of the upper left corner of this text in pixels.
+   * @param p the x, y offset of the upper left corner of this text in pixels.
+   * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
    *               text.
@@ -4530,6 +4571,7 @@
    *          this text in pixels, may be nullptr.
    * @param y a pointer filled in with the y offset of the upper left corner of
    *          this text in pixels, may be nullptr.
+   * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
    *               text.
@@ -4543,10 +4585,9 @@
   /**
    * Get the position of a text object.
    *
-   * @param x a pointer filled in with the x offset of the upper left corner of
-   *          this text in pixels, may be nullptr.
-   * @param y a pointer filled in with the y offset of the upper left corner of
-   *          this text in pixels, may be nullptr.
+   * @returns a Point with the offset of the upper left corner of this text in
+   *          pixels on success.
+   * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
    *               text.
@@ -4555,7 +4596,7 @@
    *
    * @sa SetTextPosition
    */
-  Point GetPosition();
+  Point GetPosition() const;
 
   /**
    * Set whether wrapping is enabled on a text object.
@@ -4578,9 +4619,8 @@
   /**
    * Get whether wrapping is enabled on a text object.
    *
-   * @param text the Text to query.
-   * @param wrap_width a pointer filled in with the maximum width in pixels or 0
-   *                   if the text is being wrapped on newline characters.
+   * @returns an int with the maximum width in pixels or 0 if the text is being
+   *          wrapped on newline characters on success.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4635,10 +4675,7 @@
    *
    * This function may cause the internal text representation to be rebuilt.
    *
-   * @param text the Text to modify.
-   * @param string the UTF-8 text to use, may be nullptr.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
+   * @param string the UTF-8 text to use.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4657,14 +4694,11 @@
    *
    * This function may cause the internal text representation to be rebuilt.
    *
-   * @param text the Text to modify.
    * @param offset the offset, in bytes, from the beginning of the string if >=
    *               0, the offset from the end of the string if < 0. Note that
    *               this does not do UTF-8 validation, so you should only insert
    *               at UTF-8 sequence boundaries.
    * @param string the UTF-8 text to insert.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4683,10 +4717,7 @@
    *
    * This function may cause the internal text representation to be rebuilt.
    *
-   * @param text the Text to modify.
    * @param string the UTF-8 text to insert.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4749,18 +4780,17 @@
    * The size of the text may change when the font or font style and size
    * change.
    *
-   * @param w a pointer filled in with the width of the text, in pixels, may be
-   *          nullptr.
-   * @param h a pointer filled in with the height of the text, in pixels, may be
-   *          nullptr.
+   * @returns Point containing the size on x and y axis on success.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
    *               text.
    *
    * @since This function is available since SDL_ttf 3.0.0.
+   *
+   * @sa GetSize(int*,int*) if might want only one of the coordinates
    */
-  Point GetSize();
+  Point GetSize() const;
 
   /**
    * Get the substring of a text object that surrounds a text offset.
@@ -4783,14 +4813,32 @@
    */
   void GetSubString(int offset, SubString* substring) const;
 
-  SubStringIterator begin() const { static_assert(false, "Not implemented"); }
+  /**
+   * Get iterator to first substring
+   */
+  SubStringIterator begin() const;
 
-  SubStringIterator end() const { static_assert(false, "Not implemented"); }
+  /**
+   * Get iterator to one past last substring
+   */
+  SubStringIterator end() const;
 
-  SubStringIterator GetSubStringForLine(int line) const
-  {
-    static_assert(false, "Not implemented");
-  }
+  /**
+   * Get iterator to substring of a text object that contains the given line.
+   *
+   * If `line` is greater than or equal to `text->num_lines` this will return an
+   * iterator equal to end().
+   *
+   * @param line a zero-based line index, in the range [0 .. text->num_lines-1].
+   * @returns iterator on success or false on failure; call GetError() for more
+   *          information.
+   *
+   * @threadsafety This function should be called on the thread that created the
+   *               text.
+   *
+   * @since This function is available since SDL_ttf 3.0.0.
+   */
+  SubStringIterator GetSubStringForLine(int line) const;
 
   /**
    * Get the substring of a text object that contains the given line.
@@ -4812,9 +4860,20 @@
    */
   void GetSubStringForLine(int line, SubString* substring) const;
 
+  /**
+   * Get all substrings of a text object.
+   *
+   * @returns a nullptr terminated array of substring pointers or nullptr on
+   *          failure; call GetError() for more information.
+   *
+   * @threadsafety This function should be called on the thread that created the
+   *               text.
+   *
+   * @since This function is available since SDL_ttf 3.0.0.
+   */
   OwnArray<SubString*> GetSubStrings() const
   {
-    static_assert(false, "Not implemented");
+    return GetSubStringsForRange(0);
   }
 
   /**
@@ -4823,8 +4882,6 @@
    * @param offset a byte offset into the text string.
    * @param length the length of the range being queried, in bytes, or -1 for
    *               the remainder of the string.
-   * @param count a pointer filled in with the number of substrings returned,
-   *              may be nullptr.
    * @returns a nullptr terminated array of substring pointers or nullptr on
    *          failure; call GetError() for more information. This is a single
    *          allocation that should be freed with free() when it is no longer
@@ -4835,22 +4892,31 @@
    *
    * @since This function is available since SDL_ttf 3.0.0.
    */
-  OwnArray<SubString*> GetSubStringsForRange(int offset, int length) const;
-
-  SubStringIterator GetSubStringForPoint(Point p) const
-  {
-    static_assert(false, "Not implemented");
-  }
+  OwnArray<SubString*> GetSubStringsForRange(int offset, int length = -1) const;
 
   /**
    * Get the portion of a text string that is closest to a point.
    *
    * This will return the closest substring of text to the given point.
    *
-   * @param text the Text to query.
-   * @param x the x coordinate relative to the left side of the text, may be
+   * @param p the coordinates relative to the top-left side of the text, may be
    *          outside the bounds of the text area.
-   * @param y the y coordinate relative to the top side of the text, may be
+   * @returns the iterator on success or false on failure; call GetError() for
+   *          more information.
+   *
+   * @threadsafety This function should be called on the thread that created the
+   *               text.
+   *
+   * @since This function is available since SDL_ttf 3.0.0.
+   */
+  SubStringIterator GetSubStringForPoint(Point p) const;
+
+  /**
+   * Get the portion of a text string that is closest to a point.
+   *
+   * This will return the closest substring of text to the given point.
+   *
+   * @param p the coordinates relative to the top-left side of the text, may be
    *          outside the bounds of the text area.
    * @param substring a pointer filled in with the closest substring of text to
    *                  the given point.
@@ -4870,6 +4936,8 @@
    * substring with the SUBSTRING_TEXT_START flag set.
    *
    * @param substring the SubString to query.
+   * @param previous a pointer filled in with the previous substring in the text
+   *                 object.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4917,10 +4985,10 @@
    * A copy of the UTF-8 string that this text object represents, useful for
    * layout, debugging and retrieving substring text
    */
-  const char* GetText() const { static_assert(false, "Not implemented"); }
+  const char* GetText() const { return get()->text; }
 
   /// The number of lines in the text, 0 if it's empty
-  int GetNumLines() const { static_assert(false, "Not implemented"); }
+  int GetNumLines() const { return get()->num_lines; }
 };
 
 /**
@@ -4963,8 +5031,6 @@
    *               nullptr.
    * @param font the font to render with.
    * @param text the text to use, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @post a Text object or nullptr on failure; call GetError() for more
    *       information.
    *
@@ -4988,53 +5054,75 @@
   }
 };
 
+/**
+ * Iterator for substrings
+ *
+ */
 class SubStringIterator
 {
   TextRef m_text;
 
   SubString m_subString;
 
-  constexpr SubStringIterator(TextRef text) {}
-
-public:
-  constexpr SubStringIterator() {}
-
-  constexpr operator bool() const { static_assert(false, "Not implemented"); }
-
-  constexpr const SubString& operator*() const
+  SubStringIterator(TextRef text)
+    : m_text(text)
+    , m_subString()
   {
-    static_assert(false, "Not implemented");
   }
 
-  constexpr const SubString* operator->() const
+public:
+  /// Default constructor.
+  SubStringIterator()
+    : SubStringIterator(TextRef{})
   {
-    static_assert(false, "Not implemented");
   }
 
+  /// True if pointing to valid SubString.
+  constexpr operator bool() const { return bool(m_text); }
+
+  /// Retrieve SubString
+  constexpr const SubString& operator*() const { return m_subString; }
+
+  /// Retrieve SubString.
+  constexpr const SubString* operator->() const { return &m_subString; }
+
+  /// Comparison.
   constexpr bool operator==(const SubStringIterator& other) const
   {
-    static_assert(false, "Not implemented");
+    return m_subString.offset == other.m_subString.offset;
   }
 
-  constexpr SubStringIterator& operator++()
+  /// Increment operator.
+  SubStringIterator& operator++()
   {
-    static_assert(false, "Not implemented");
+    m_text.GetNextSubString(m_subString, &m_subString);
+    return *this;
   }
 
-  constexpr SubStringIterator operator++(int)
+  /// Increment operator.
+  SubStringIterator operator++(int)
   {
-    static_assert(false, "Not implemented");
+    auto curr = *this;
+    m_text.GetNextSubString(m_subString, &m_subString);
+    return curr;
   }
 
-  constexpr SubStringIterator& operator--()
+  /// Decrement operator.
+  SubStringIterator& operator--()
   {
-    static_assert(false, "Not implemented");
+    m_text.GetPreviousSubString(m_subString, &m_subString);
+    return *this;
   }
 
-  constexpr SubStringIterator operator--(int)
+  /// Decrement operator.
+  SubStringIterator operator--(int)
   {
-    static_assert(false, "Not implemented");
+    auto curr = *this;
+    m_text.GetPreviousSubString(m_subString, &m_subString);
+    return curr;
   }
+
+  friend class Text;
 };
 
 /**
@@ -5056,7 +5144,7 @@
 }
 
 inline SurfaceTextEngine::SurfaceTextEngine()
-  : T(TTF_CreateSurfaceTextEngine())
+  : TextEngine(TTF_CreateSurfaceTextEngine())
 {
 }
 
@@ -5067,10 +5155,8 @@
  * CreateSurfaceTextEngine().
  *
  * @param text the text to draw.
- * @param x the x coordinate in pixels, positive from the left edge towards the
- *          right.
- * @param y the y coordinate in pixels, positive from the top edge towards the
- *          bottom.
+ * @param p the (x, y) coordinate in pixels, positive from the left edge
+ *          towards the right and from the top edge towards the bottom.
  * @param surface the surface to draw on.
  * @throws Error on failure.
  *
@@ -5084,7 +5170,7 @@
  */
 inline void DrawSurfaceText(TextConstRef text, Point p, SurfaceRef surface)
 {
-  CheckError(TTF_DrawSurfaceText(text, p, surface));
+  CheckError(TTF_DrawSurfaceText(text, p.x, p.y, surface));
 }
 
 inline void TextBase::DrawSurface(Point p, SurfaceRef surface) const
@@ -5139,12 +5225,12 @@
 }
 
 inline RendererTextEngine::RendererTextEngine(RendererRef renderer)
-  : T(TTF_CreateRendererTextEngine(renderer))
+  : TextEngine(TTF_CreateRendererTextEngine(renderer))
 {
 }
 
 inline RendererTextEngine::RendererTextEngine(PropertiesRef props)
-  : T(TTF_CreateRendererTextEngineWithProperties(props))
+  : TextEngine(TTF_CreateRendererTextEngineWithProperties(props))
 {
 }
 
@@ -5178,6 +5264,14 @@
   return RendererTextEngine(props);
 }
 
+/**
+ * Properties for RendererTextEngine.
+ *
+ * Use these with CreateRendererTextEngineWithProperties() to specify properties
+ * when creating a renderer text engine.
+ *
+ * @sa CreateRendererTextEngineWithProperties
+ */
 namespace prop::RendererTextEngine {
 
 constexpr auto RENDERER_POINTER =
@@ -5197,10 +5291,8 @@
  * function.
  *
  * @param text the text to draw.
- * @param x the x coordinate in pixels, positive from the left edge towards the
- *          right.
- * @param y the y coordinate in pixels, positive from the top edge towards the
- *          bottom.
+ * @param p the (x, y) coordinate in pixels, positive from the left edge
+ *          towards the right and from the top edge towards the bottom.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -5213,7 +5305,7 @@
  */
 inline void DrawRendererText(TextConstRef text, FPoint p)
 {
-  CheckError(TTF_DrawRendererText(text, p));
+  CheckError(TTF_DrawRendererText(text, p.x, p.y));
 }
 
 inline void TextBase::DrawRenderer(FPoint p) const
@@ -5268,12 +5360,12 @@
 }
 
 inline GPUTextEngine::GPUTextEngine(GPUDeviceRef device)
-  : T(TTF_CreateGPUTextEngine(device))
+  : TextEngine(TTF_CreateGPUTextEngine(device))
 {
 }
 
 inline GPUTextEngine::GPUTextEngine(PropertiesRef props)
-  : T(TTF_CreateGPUTextEngineWithProperties(props))
+  : TextEngine(TTF_CreateGPUTextEngineWithProperties(props))
 {
 }
 
@@ -5306,6 +5398,14 @@
   return GPUTextEngine(props);
 }
 
+/**
+ * Properties for GPUTextEngine.
+ *
+ * Use these with CreateGPUTextEngineWithProperties() to specify properties when
+ * creating a GPU text engine.
+ *
+ * @sa CreateGPUTextEngineWithProperties
+ */
 namespace prop::GpuTextEngine {
 
 constexpr auto DEVICE_POINTER =
@@ -5374,16 +5474,56 @@
 
 inline void GPUTextEngine::Destroy() { DestroyGPUTextEngine(release()); }
 
-inline void GPUTextEngine::SetGPUWinding(TextEngineRef engine,
-                                         GPUTextEngineWinding winding)
+/**
+ * Sets the winding order of the vertices returned by Text.GetGPUDrawData
+ * for a particular GPU text engine.
+ *
+ * @param engine a TextEngine object created with
+ *               GPUTextEngine.GPUTextEngine().
+ * @param winding the new winding order option.
+ *
+ * @threadsafety This function should be called on the thread that created the
+ *               engine.
+ *
+ * @since This function is available since SDL_ttf 3.0.0.
+ *
+ * @sa GPUTextEngine.GetGPUWinding
+ */
+inline void SetGPUTextEngineWinding(TextEngineRef engine,
+                                    GPUTextEngineWinding winding)
+{
+  TTF_SetGPUTextEngineWinding(engine, winding);
+}
+
+inline void GPUTextEngine::SetGPUWinding(GPUTextEngineWinding winding)
+{
+  SDL::SetGPUTextEngineWinding(get(), winding);
+}
+
+/**
+ * Get the winding order of the vertices returned by Text.GetGPUDrawData
+ * for a particular GPU text engine
+ *
+ * @param engine a TextEngine object created with
+ *               GPUTextEngine.GPUTextEngine().
+ * @returns the winding order used by the GPU text engine or
+ *          GPU_TEXTENGINE_WINDING_INVALID in case of error.
+ *
+ * @threadsafety This function should be called on the thread that created the
+ *               engine.
+ *
+ * @since This function is available since SDL_ttf 3.0.0.
+ *
+ * @sa GPUTextEngine.SetGPUWinding
+ */
+inline GPUTextEngineWinding GetGPUTextEngineWinding(TextEngineRef engine)
 {
-  SDL::SetGPUTextEngineWinding(engine, winding);
+  return TTF_GetGPUTextEngineWinding(engine);
 }
 
-inline GPUTextEngineWinding GPUTextEngine::GetGPUWinding(
-  TextEngineRef engine) const
+inline GPUTextEngineWinding GPUTextEngine::GetGPUWinding() const
 {
-  return SDL::GetGPUTextEngineWinding(engine);
+  return SDL::GetGPUTextEngineWinding(get());
 }
 
 /**
@@ -5393,8 +5533,6 @@
  *               nullptr.
  * @param font the font to render with.
  * @param text the text to use, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @returns a Text object or nullptr on failure; call GetError() for more
  *          information.
  *
@@ -5418,7 +5556,7 @@
 }
 
 inline Text::Text(TextEngineRef engine, FontRef font, std::string_view text)
-  : Text(TTF_CreateText(engine, font, text))
+  : Text(TTF_CreateText(engine, font, text.data(), text.size()))
 {
 }
 
@@ -5436,7 +5574,7 @@
  */
 inline PropertiesRef GetTextProperties(TextConstRef text)
 {
-  return CheckError(TTF_GetTextProperties(text));
+  return {CheckError(TTF_GetTextProperties(text))};
 }
 
 inline PropertiesRef TextBase::GetProperties() const
@@ -5601,7 +5739,7 @@
  *
  * @param text the text to modify.
  * @param script an [ISO 15924
- *               code](https://unicode.org/iso15924/iso15924-codes.html) .
+ *               code](https://unicode.org/iso15924/iso15924-codes.html).
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -5651,10 +5789,7 @@
  * The default text color is white (255, 255, 255, 255).
  *
  * @param text the Text to modify.
- * @param r the red color value in the range of 0-255.
- * @param g the green color value in the range of 0-255.
- * @param b the blue color value in the range of 0-255.
- * @param a the alpha value in the range of 0-255.
+ * @param c the color value in the range of 0-255.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -5667,7 +5802,7 @@
  */
 inline void SetTextColor(TextRef text, Color c)
 {
-  CheckError(TTF_SetTextColor(text, c));
+  CheckError(TTF_SetTextColor(text, c.r, c.g, c.b, c.a));
 }
 
 inline void TextBase::SetColor(Color c) { SDL::SetTextColor(get(), c); }
@@ -5678,10 +5813,7 @@
  * The default text color is white (1.0f, 1.0f, 1.0f, 1.0f).
  *
  * @param text the Text to modify.
- * @param r the red color value, normally in the range of 0-1.
- * @param g the green color value, normally in the range of 0-1.
- * @param b the blue color value, normally in the range of 0-1.
- * @param a the alpha value in the range of 0-1.
+ * @param c the color value, normally in the range of 0-1.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -5694,13 +5826,10 @@
  */
 inline void SetTextColorFloat(TextRef text, FColor c)
 {
-  CheckError(TTF_SetTextColorFloat(text, c));
+  CheckError(TTF_SetTextColorFloat(text, c.r, c.g, c.b, c.a));
 }
 
-inline void TextBase::SetColorFloat(FColor c)
-{
-  SDL::SetTextColorFloat(get(), c);
-}
+inline void TextBase::SetColorFloat(FColor c) { SetTextColorFloat(get(), c); }
 
 /**
  * Get the color of a text object.
@@ -5737,14 +5866,7 @@
  * Get the color of a text object.
  *
  * @param text the Text to query.
- * @param r a pointer filled in with the red color value in the range of 0-255,
- *          may be nullptr.
- * @param g a pointer filled in with the green color value in the range of
- *          0-255, may be nullptr.
- * @param b a pointer filled in with the blue color value in the range of 0-255,
- *          may be nullptr.
- * @param a a pointer filled in with the alpha value in the range of 0-255, may
- *          be nullptr.
+ * @return The color in the range of 0-255 on success.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -5757,7 +5879,9 @@
  */
 inline Color GetTextColor(TextRef text)
 {
-  static_assert(false, "Not implemented");
+  Color c;
+  GetTextColor(text, &c.r, &c.g, &c.b, &c.a);
+  return c;
 }
 
 inline void TextBase::GetColor(Uint8* r, Uint8* g, Uint8* b, Uint8* a) const
@@ -5765,7 +5889,7 @@
   SDL::GetTextColor(get(), r, g, b, a);
 }
 
-inline Color TextBase::GetColor() { return SDL::GetTextColor(get()); }
+inline Color TextBase::GetColor() const { return SDL::GetTextColor(get()); }
 
 /**
  * Get the color of a text object.
@@ -5802,14 +5926,7 @@
  * Get the color of a text object.
  *
  * @param text the Text to query.
- * @param r a pointer filled in with the red color value, normally in the range
- *          of 0-1, may be nullptr.
- * @param g a pointer filled in with the green color value, normally in the
- *          range of 0-1, may be nullptr.
- * @param b a pointer filled in with the blue color value, normally in the range
- *          of 0-1, may be nullptr.
- * @param a a pointer filled in with the alpha value in the range of 0-1, may be
- *          nullptr.
+ * @return The color in the range of 0-1 on success.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -5822,7 +5939,9 @@
  */
 inline FColor GetTextColorFloat(TextRef text)
 {
-  static_assert(false, "Not implemented");
+  FColor c;
+  GetTextColorFloat(text, &c.r, &c.g, &c.b, &c.a);
+  return c;
 }
 
 inline void TextBase::GetColorFloat(float* r,
@@ -5833,7 +5952,7 @@
   SDL::GetTextColorFloat(get(), r, g, b, a);
 }
 
-inline FColor TextBase::GetColorFloat()
+inline FColor TextBase::GetColorFloat() const
 {
   return SDL::GetTextColorFloat(get());
 }
@@ -5847,8 +5966,8 @@
  * This function may cause the internal text representation to be rebuilt.
  *
  * @param text the Text to modify.
- * @param x the x offset of the upper left corner of this text in pixels.
- * @param y the y offset of the upper left corner of this text in pixels.
+ * @param p the x, y offset of the upper left corner of this text in pixels.
+ * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
  *               text.
@@ -5859,7 +5978,7 @@
  */
 inline void SetTextPosition(TextRef text, const PointRaw& p)
 {
-  CheckError(TTF_SetTextPosition(text, p));
+  CheckError(TTF_SetTextPosition(text, p.x, p.y));
 }
 
 inline void TextBase::SetPosition(const PointRaw& p)
@@ -5875,6 +5994,7 @@
  *          this text in pixels, may be nullptr.
  * @param y a pointer filled in with the y offset of the upper left corner of
  *          this text in pixels, may be nullptr.
+ * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
  *               text.
@@ -5892,10 +6012,9 @@
  * Get the position of a text object.
  *
  * @param text the Text to query.
- * @param x a pointer filled in with the x offset of the upper left corner of
- *          this text in pixels, may be nullptr.
- * @param y a pointer filled in with the y offset of the upper left corner of
- *          this text in pixels, may be nullptr.
+ * @returns a Point with the offset of the upper left corner of this text in
+ *          pixels on success.
+ * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
  *               text.
@@ -5906,7 +6025,9 @@
  */
 inline Point GetTextPosition(TextRef text)
 {
-  static_assert(false, "Not implemented");
+  Point p;
+  GetTextPosition(text, &p.x, &p.y);
+  return p;
 }
 
 inline void TextBase::GetPosition(int* x, int* y) const
@@ -5914,7 +6035,7 @@
   SDL::GetTextPosition(get(), x, y);
 }
 
-inline Point TextBase::GetPosition() { return SDL::GetTextPosition(get()); }
+inline Point TextBase::GetPosition() const { return GetTextPosition(get()); }
 
 /**
  * Set whether wrapping is enabled on a text object.
@@ -5947,8 +6068,8 @@
  * Get whether wrapping is enabled on a text object.
  *
  * @param text the Text to query.
- * @param wrap_width a pointer filled in with the maximum width in pixels or 0
- *                   if the text is being wrapped on newline characters.
+ * @returns maximum width in pixels or 0 if the text is being wrapped on newline
+ *          characters.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -5960,13 +6081,12 @@
  */
 inline int GetTextWrapWidth(TextConstRef text)
 {
-  return CheckError(TTF_GetTextWrapWidth(text));
+  int w;
+  CheckError(TTF_GetTextWrapWidth(text, &w));
+  return w;
 }
 
-inline int TextBase::GetWrapWidth() const
-{
-  return SDL::GetTextWrapWidth(get());
-}
+inline int TextBase::GetWrapWidth() const { return GetTextWrapWidth(get()); }
 
 /**
  * Set whether whitespace should be visible when wrapping a text object.
@@ -6028,8 +6148,6 @@
  *
  * @param text the Text to modify.
  * @param string the UTF-8 text to use, may be nullptr.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -6043,7 +6161,7 @@
  */
 inline void SetTextString(TextRef text, std::string_view string)
 {
-  CheckError(TTF_SetTextString(text, string));
+  CheckError(TTF_SetTextString(text, string.data(), string.size()));
 }
 
 inline void TextBase::SetString(std::string_view string)
@@ -6062,8 +6180,6 @@
  *               does not do UTF-8 validation, so you should only insert at
  *               UTF-8 sequence boundaries.
  * @param string the UTF-8 text to insert.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -6077,7 +6193,7 @@
  */
 inline void InsertTextString(TextRef text, int offset, std::string_view string)
 {
-  CheckError(TTF_InsertTextString(text, offset, string));
+  CheckError(TTF_InsertTextString(text, offset, string.data(), string.size()));
 }
 
 inline void TextBase::InsertString(int offset, std::string_view string)
@@ -6092,8 +6208,6 @@
  *
  * @param text the Text to modify.
  * @param string the UTF-8 text to insert.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -6107,7 +6221,7 @@
  */
 inline void AppendTextString(TextRef text, std::string_view string)
 {
-  CheckError(TTF_AppendTextString(text, string));
+  CheckError(TTF_AppendTextString(text, string.data(), string.size()));
 }
 
 inline void TextBase::AppendString(std::string_view string)
@@ -6176,10 +6290,8 @@
  * The size of the text may change when the font or font style and size change.
  *
  * @param text the Text to query.
- * @param w a pointer filled in with the width of the text, in pixels, may be
- *          nullptr.
- * @param h a pointer filled in with the height of the text, in pixels, may be
- *          nullptr.
+ * @returns Point containing the size on x and y axis on success or
+ *          std::nullopt on failure; call GetError() for more information.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -6189,7 +6301,9 @@
  */
 inline Point GetTextSize(TextRef text)
 {
-  static_assert(false, "Not implemented");
+  Point p;
+  GetTextSize(text, &p.x, &p.y);
+  return p;
 }
 
 inline void TextBase::GetSize(int* w, int* h) const
@@ -6197,7 +6311,7 @@
   SDL::GetTextSize(get(), w, h);
 }
 
-inline Point TextBase::GetSize() { return SDL::GetTextSize(get()); }
+inline Point TextBase::GetSize() const { return GetTextSize(get()); }
 
 /**
  * Get the substring of a text object that surrounds a text offset.
@@ -6269,12 +6383,8 @@
  * @param offset a byte offset into the text string.
  * @param length the length of the range being queried, in bytes, or -1 for the
  *               remainder of the string.
- * @param count a pointer filled in with the number of substrings returned, may
- *              be nullptr.
  * @returns a nullptr terminated array of substring pointers or nullptr on
- *          failure; call GetError() for more information. This is a single
- *          allocation that should be freed with free() when it is no longer
- *          needed.
+ *          failure; call GetError() for more information.
  *
  * @threadsafety This function should be called on the thread that created the
  *               text.
@@ -6285,7 +6395,9 @@
                                                       int offset,
                                                       int length)
 {
-  return TTF_GetTextSubStringsForRange(text, offset, length);
+  int count = 0;
+  auto data = TTF_GetTextSubStringsForRange(text, offset, length, &count);
+  return OwnArray<SubString*>{data, size_t(count)};
 }
 
 inline OwnArray<SubString*> TextBase::GetSubStringsForRange(int offset,
@@ -6300,9 +6412,7 @@
  * This will return the closest substring of text to the given point.
  *
  * @param text the Text to query.
- * @param x the x coordinate relative to the left side of the text, may be
- *          outside the bounds of the text area.
- * @param y the y coordinate relative to the top side of the text, may be
+ * @param p the coordinates relative to the top-left side of the text, may be
  *          outside the bounds of the text area.
  * @param substring a pointer filled in with the closest substring of text to
  *                  the given point.
@@ -6317,7 +6427,7 @@
                                      Point p,
                                      SubString* substring)
 {
-  CheckError(TTF_GetTextSubStringForPoint(text, p, substring));
+  CheckError(TTF_GetTextSubStringForPoint(text, p.x, p.y, substring));
 }
 
 inline void TextBase::GetSubStringForPoint(Point p, SubString* substring) const
@@ -6333,6 +6443,8 @@
  *
  * @param text the Text to query.
  * @param substring the SubString to query.
+ * @param previous a pointer filled in with the previous substring in the text
+ *                 object.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -6447,4 +6559,6 @@
 
 } // namespace SDL
 
+#endif // defined(SDL3PP_ENABLE_TTF) || defined(SDL3PP_DOC)
+
 #endif /* SDL3PP_TTF_H_ */
