--- build/generated/SDL3pp_time.h
+++ include/SDL3pp/SDL3pp_time.h
@@ -7,7 +7,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryTime Category Time
+ * @defgroup CategoryTime Date and Time
  *
  * SDL realtime clock and date/time routines.
  *
@@ -84,7 +84,6 @@
    * DateTime format.
    *
    * @param ticks the Time to be converted.
-   * @param dt the resulting DateTime.
    * @param localTime the resulting DateTime will be expressed in local time if
    *                  true, otherwise it will be in Universal Coordinated Time
    *                  (UTC).
@@ -103,7 +102,8 @@
    */
   constexpr explicit operator bool() const noexcept
   {
-    return *this != DateTimeRaw{};
+    return year != 0 || month != 0 || day != 0 || hour != 0 || minute != 0 ||
+           second != 0 || nanosecond != 0;
   }
 
   /**
@@ -283,8 +283,7 @@
    * This function ignores the day_of_week member of the DateTime struct, so it
    * may remain unset.
    *
-   * @param dt the source DateTime.
-   * @param ticks the resulting Time.
+   * @returns time on success.
    * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread.
@@ -360,19 +359,24 @@
  *
  * @since This function is available since SDL 3.2.0.
  */
-inline Time GetCurrentTime() { return CheckError(SDL_GetCurrentTime()); }
+inline Time GetCurrentTime()
+{
+  SDL_Time t;
+  CheckError(SDL_GetCurrentTime(&t));
+  return Time::FromNS(t);
+}
 
-inline Time Time::Current() { return CheckError(SDL_GetCurrentTime(m_time)); }
+inline Time Time::Current() { return GetCurrentTime(); }
 
 /**
  * Converts an Time in nanoseconds since the epoch to a calendar time in the
  * DateTime format.
  *
  * @param ticks the Time to be converted.
- * @param dt the resulting DateTime.
  * @param localTime the resulting DateTime will be expressed in local time if
  *                  true, otherwise it will be in Universal Coordinated Time
  *                  (UTC).
+ * @returns the resulting DateTime.
  * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
@@ -385,8 +389,8 @@
 }
 
 inline DateTime::DateTime(Time ticks, bool localTime)
-  : DateTimeRaw(CheckError(SDL_TimeToDateTime(ticks, localTime)))
 {
+  CheckError(SDL_TimeToDateTime(ticks.ToNS(), this, localTime));
 }
 
 /**
@@ -396,7 +400,7 @@
  * may remain unset.
  *
  * @param dt the source DateTime.
- * @param ticks the resulting Time.
+ * @returns the resulting Time.
  * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
@@ -405,10 +409,12 @@
  */
 inline Time DateTimeToTime(const DateTimeRaw& dt)
 {
-  return CheckError(SDL_DateTimeToTime(dt));
+  SDL_Time t;
+  CheckError(SDL_DateTimeToTime(&dt, &t));
+  return Time::FromNS(t);
 }
 
-inline DateTime::operator Time() const { return SDL::DateTimeToTime(); }
+inline DateTime::operator Time() const { return SDL::DateTimeToTime(*this); }
 
 /**
  * Converts an SDL time into a Windows FILETIME (100-nanosecond intervals since
@@ -430,12 +436,12 @@
                           Uint32* dwLowDateTime,
                           Uint32* dwHighDateTime)
 {
-  SDL_TimeToWindows(ticks, dwLowDateTime, dwHighDateTime);
+  SDL_TimeToWindows(ticks.ToNS(), dwLowDateTime, dwHighDateTime);
 }
 
 inline void Time::ToWindows(Uint32* dwLowDateTime, Uint32* dwHighDateTime) const
 {
-  SDL_TimeToWindows(m_time, dwLowDateTime, dwHighDateTime);
+  return TimeToWindows(*this, dwLowDateTime, dwHighDateTime);
 }
 
 /**
@@ -455,12 +461,12 @@
  */
 inline Time TimeFromWindows(Uint32 dwLowDateTime, Uint32 dwHighDateTime)
 {
-  return SDL_TimeFromWindows(dwLowDateTime, dwHighDateTime);
+  return Time::FromNS(SDL_TimeFromWindows(dwLowDateTime, dwHighDateTime));
 }
 
 inline Time Time::FromWindows(Uint32 dwLowDateTime, Uint32 dwHighDateTime)
 {
-  return SDL_TimeFromWindows(dwLowDateTime, dwHighDateTime);
+  return TimeFromWindows(dwLowDateTime, dwHighDateTime);
 }
 
 /**
@@ -468,8 +474,8 @@
  *
  * @param year the year.
  * @param month the month [1-12].
- * @returns the number of days in the requested month or -1 on failure; call
- *          GetError() for more information.
+ * @returns the number of days in the requested month.
+ * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -477,7 +483,7 @@
  */
 inline int GetDaysInMonth(int year, int month)
 {
-  return SDL_GetDaysInMonth(year, month);
+  return CheckError(SDL_GetDaysInMonth(year, month), -1);
 }
 
 /**
@@ -486,8 +492,8 @@
  * @param year the year component of the date.
  * @param month the month component of the date.
  * @param day the day component of the date.
- * @returns the day of year [0-365] if the date is valid or -1 on failure; call
- *          GetError() for more information.
+ * @returns the day of year [0-365] if the date is valid.
+ * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -495,7 +501,7 @@
  */
 inline int GetDayOfYear(int year, int month, int day)
 {
-  return SDL_GetDayOfYear(year, month, day);
+  return CheckError(SDL_GetDayOfYear(year, month, day), -1);
 }
 
 /**
@@ -504,8 +510,8 @@
  * @param year the year component of the date.
  * @param month the month component of the date.
  * @param day the day component of the date.
- * @returns a value between 0 and 6 (0 being Sunday) if the date is valid or -1
- *          on failure; call GetError() for more information.
+ * @returns a value between 0 and 6 (0 being Sunday) if the date is valid.
+ * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -513,7 +519,7 @@
  */
 inline int GetDayOfWeek(int year, int month, int day)
 {
-  return SDL_GetDayOfWeek(year, month, day);
+  return CheckError(SDL_GetDayOfWeek(year, month, day), -1);
 }
 
 /// @}
