--- src/generated/SDL3pp_hints.h
+++ include/SDL3pp/SDL3pp_hints.h
@@ -9,7 +9,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryHints Category Hints
+ * @defgroup CategoryHints Configuration Variables
  *
  * This file contains functions to set and get configuration hints, as well as
  * listing each of them alphabetically.
@@ -25,6 +25,8 @@
  * @{
  */
 
+#ifdef SDL3PP_DOC
+
 /**
  * Specify the behavior of Alt+Tab while the keyboard is grabbed.
  *
@@ -1324,7 +1326,7 @@
  *
  * `0xAAAA/0xBBBB,0xCCCC/0xDDDD`
  *
- * The variable can also take the form of @file, in which case the named file
+ * The variable can also take the form of file, in which case the named file
  * will be loaded and interpreted as the value of the variable.
  *
  * This hint can be set anytime.
@@ -4661,6 +4663,8 @@
  */
 #define SDL_HINT_PEN_TOUCH_EVENTS "SDL_PEN_TOUCH_EVENTS"
 
+#endif // SDL3PP_DOC
+
 /**
  * An enumeration of hint priorities.
  *
@@ -4875,7 +4879,6 @@
  * @param name the hint to watch.
  * @param callback An HintCallback function that will be called when the hint
  *                 value changes.
- * @param userdata a pointer to pass to the callback function.
  * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
@@ -4886,7 +4889,7 @@
  */
 inline void AddHintCallback(StringParam name, HintCB callback)
 {
-  static_assert(false, "Not implemented");
+  AddHintCallback(std::move(name), callback.wrapper, callback.data);
 }
 
 /**
