--- src/generated/SDL3pp_endian.h
+++ include/SDL3pp/SDL3pp_endian.h
@@ -7,7 +7,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryEndian Category Endian
+ * @defgroup CategoryEndian Byte Order and Byte Swapping
  *
  * Functions converting endian-specific values to different byte orders.
  *
@@ -25,13 +25,15 @@
  * @{
  */
 
+#ifdef SDL3PP_DOC
+
 /**
  * A value to represent littleendian byteorder.
  *
  * This is used with the preprocessor macro SDL_BYTEORDER, to determine a
  * platform's byte ordering:
  *
- * ```c
+ * ```cpp
  * #if SDL_BYTEORDER == SDL_LIL_ENDIAN
  * Log("This system is littleendian.");
  * #endif
@@ -50,7 +52,7 @@
  * This is used with the preprocessor macro SDL_BYTEORDER, to determine a
  * platform's byte ordering:
  *
- * ```c
+ * ```cpp
  * #if SDL_BYTEORDER == SDL_BIG_ENDIAN
  * Log("This system is bigendian.");
  * #endif
@@ -103,6 +105,8 @@
  */
 #define SDL_FLOATWORDORDER SDL_LIL_ENDIAN___or_maybe___SDL_BIG_ENDIAN
 
+#endif // SDL3PP_DOC
+
 /**
  * Byte-swap an unsigned 16-bit number.
  *
@@ -122,7 +126,7 @@
  *
  * @since This function is available since SDL 3.2.0.
  */
-constexpr Uint16 Swap16(Uint16 x) { return SDL_Swap16(x); }
+inline Uint16 Swap16(Uint16 x) { return SDL_Swap16(x); }
 
 /**
  * Byte-swap an unsigned 32-bit number.
@@ -143,7 +147,7 @@
  *
  * @since This function is available since SDL 3.2.0.
  */
-constexpr Uint32 Swap32(Uint32 x) { return SDL_Swap32(x); }
+inline Uint32 Swap32(Uint32 x) { return SDL_Swap32(x); }
 
 /**
  * Byte-swap an unsigned 64-bit number.
@@ -164,7 +168,7 @@
  *
  * @since This function is available since SDL 3.2.0.
  */
-constexpr Uint64 Swap64(Uint64 x) { return SDL_Swap64(x); }
+inline Uint64 Swap64(Uint64 x) { return SDL_Swap64(x); }
 
 /**
  * Byte-swap a floating point number.
@@ -185,7 +189,7 @@
  *
  * @since This function is available since SDL 3.2.0.
  */
-constexpr float SwapFloat(float x) { return SDL_SwapFloat(x); }
+inline float SwapFloat(float x) { return SDL_SwapFloat(x); }
 
 /**
  * Swap a 16-bit value from littleendian to native byte order.
@@ -265,7 +269,7 @@
  *
  * @since This function is available since SDL 3.2.0.
  */
-constexpr Uint16 Swap16BE(Uint16 x) { return SDL_Swap16BE(x); }
+inline Uint16 Swap16BE(Uint16 x) { return SDL_Swap16BE(x); }
 
 /**
  * Swap a 32-bit value from bigendian to native byte order.
@@ -281,7 +285,7 @@
  *
  * @since This function is available since SDL 3.2.0.
  */
-constexpr Uint32 Swap32BE(Uint32 x) { return SDL_Swap32BE(x); }
+inline Uint32 Swap32BE(Uint32 x) { return SDL_Swap32BE(x); }
 
 /**
  * Swap a 64-bit value from bigendian to native byte order.
@@ -297,7 +301,7 @@
  *
  * @since This function is available since SDL 3.2.0.
  */
-constexpr Uint64 Swap64BE(Uint64 x) { return SDL_Swap64BE(x); }
+inline Uint64 Swap64BE(Uint64 x) { return SDL_Swap64BE(x); }
 
 /**
  * Swap a floating point value from bigendian to native byte order.
@@ -313,7 +317,7 @@
  *
  * @since This function is available since SDL 3.2.0.
  */
-constexpr float SwapFloatBE(float x) { return SDL_SwapFloatBE(x); }
+inline float SwapFloatBE(float x) { return SDL_SwapFloatBE(x); }
 
 /// @}
 
