--- src/generated/SDL3pp_system.h
+++ include/SDL3pp/SDL3pp_system.h
@@ -2,12 +2,14 @@
 #define SDL3PP_SYSTEM_H_
 
 #include <SDL3/SDL_system.h>
+#include "SDL3pp_keyboard.h"
 #include "SDL3pp_stdinc.h"
+#include "SDL3pp_video.h"
 
 namespace SDL {
 
 /**
- * @defgroup CategorySystem Category System
+ * @defgroup CategorySystem Platform-specific Functionality
  *
  * Platform-specific SDL API functions. These are functions that deal with needs
  * of specific operating systems, that didn't make sense to offer as
@@ -20,6 +22,12 @@
  * @{
  */
 
+/*
+ * Platform specific functions for Windows
+ */
+#if defined(SDL_PLATFORM_WINDOWS) || defined(SDL3PP_DOC)
+
+/// MSG
 using MSG = ::MSG;
 
 /**
@@ -94,7 +102,6 @@
  * should continue to be processed, or false to prevent further processing.
  *
  * @param callback the WindowsMessageHook function to call.
- * @param userdata a pointer to pass to every iteration of `callback`.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -103,9 +110,14 @@
  */
 inline void SetWindowsMessageHook(WindowsMessageHookCB callback)
 {
-  static_assert(false, "Not implemented");
+  using Wrapper = UniqueCallbackWrapper<WindowsMessageHookCB>;
+  SetWindowsMessageHook(&Wrapper::CallSuffixed,
+                        Wrapper::Wrap(std::move(callback)));
 }
+#endif // SDL_PLATFORM_WINDOWS
 
+#if defined(SDL_PLATFORM_WIN32) || defined(SDL_PLATFORM_WINGDK) ||             \
+  defined(SDL3PP_DOC)
 /**
  * Get the D3D9 adapter index that matches the specified display.
  *
@@ -144,6 +156,9 @@
   CheckError(SDL_GetDXGIOutputInfo(displayID, adapterIndex, outputIndex));
 }
 
+#endif // defined(SDL_PLATFORM_WIN32) || defined(SDL_PLATFORM_WINGDK)
+
+/// this is defined in Xlib's headers, just need a simple declaration here
 using XEvent = ::XEvent;
 
 /**
@@ -214,15 +229,18 @@
  * continue to be processed, or false to prevent further processing.
  *
  * @param callback the X11EventHook function to call.
- * @param userdata a pointer to pass to every iteration of `callback`.
  *
  * @since This function is available since SDL 3.2.0.
  */
 inline void SetX11EventHook(X11EventHookCB callback)
 {
-  static_assert(false, "Not implemented");
+  using Wrapper = UniqueCallbackWrapper<X11EventHookCB>;
+  SDL_SetX11EventHook(&Wrapper::Call, Wrapper::Wrap(std::move(callback)));
 }
 
+/* Platform specific functions for Linux*/
+#if defined(SDL_PLATFORM_LINUX) || defined(SDL3PP_DOC)
+
 /**
  * Sets the UNIX nice value for a thread.
  *
@@ -260,6 +278,13 @@
     SDL_SetLinuxThreadPriorityAndPolicy(threadID, sdlPriority, schedPolicy));
 }
 
+#endif // SDL_PLATFORM_LINUX
+
+/*
+ * Platform specific functions for iOS
+ */
+#if defined(SDL_PLATFORM_IOS) || defined(SDL3PP_DOC)
+
 /**
  * The prototype for an Apple iOS animation callback.
  *
@@ -365,7 +390,6 @@
  * @param window the window for which the animation callback should be set.
  * @param interval the number of frames after which **callback** will be called.
  * @param callback the function to call for every frame.
- * @param callbackParam a pointer that is passed to `callback`.
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
@@ -376,7 +400,8 @@
                                     int interval,
                                     iOSAnimationCB callback)
 {
-  static_assert(false, "Not implemented");
+  using Wrapper = UniqueCallbackWrapper<iOSAnimationCB>;
+  SetiOSAnimationCallback(&Wrapper::Call, Wrapper::Wrap(std::move(callback)));
 }
 
 /**
@@ -392,6 +417,13 @@
  */
 inline void SetiOSEventPump(bool enabled) { SDL_SetiOSEventPump(enabled); }
 
+#endif // defined(SDL_PLATFORM_IOS)
+
+/*
+ * Platform specific functions for Android
+ */
+#if defined(SDL_PLATFORM_ANDROID) || defined(SDL3PP_DOC)
+
 /**
  * Get the Android Java Native Interface Environment of the current thread.
  *
@@ -520,8 +552,6 @@
 constexpr Uint32 ANDROID_EXTERNAL_STORAGE_WRITE =
   SDL_ANDROID_EXTERNAL_STORAGE_WRITE;
 
-#error "RequestAndroidPermissionCB (plc)"
-
 /**
  * Get the path used for internal storage for this Android application.
  *
@@ -635,6 +665,21 @@
 using RequestAndroidPermissionCallback = SDL_RequestAndroidPermissionCallback;
 
 /**
+ * Callback that presents a response from a RequestAndroidPermission call.
+ *
+ * @param userdata an app-controlled pointer that is passed to the callback.
+ * @param permission the Android-specific permission name that was requested.
+ * @param granted true if permission is granted, false if denied.
+ *
+ * @since This datatype is available since SDL 3.2.0.
+ *
+ * @sa RequestAndroidPermission
+ * @sa RequestAndroidPermissionCallback
+ */
+using RequestAndroidPermissionCB =
+  std::function<void(const char* permission, bool granted)>;
+
+/**
  * Request permissions at runtime, asynchronously.
  *
  * You do not need to call this for built-in functionality of SDL; recording
@@ -700,7 +745,6 @@
  *
  * @param permission the permission to request.
  * @param cb the callback to trigger when the request has a response.
- * @param userdata an app-controlled pointer that is passed to the callback.
  * @returns true if the request was submitted, false if there was an error
  *          submitting. The result of the request is only ever reported through
  *          the callback, not this return value.
@@ -712,7 +756,12 @@
 inline bool RequestAndroidPermission(StringParam permission,
                                      RequestAndroidPermissionCB cb)
 {
-  static_assert(false, "Not implemented");
+  using Wrapper = CallbackWrapper<RequestAndroidPermissionCB>;
+  auto callback = Wrapper::Wrap(std::move(cb));
+  if (!RequestAndroidPermission(
+        std::move(permission), &Wrapper::CallOnce, callback)) {
+    Wrapper::release(callback);
+  }
 }
 
 /**
@@ -768,6 +817,8 @@
   CheckError(SDL_SendAndroidMessage(command, param));
 }
 
+#endif // defined(SDL_PLATFORM_ANDROID)
+
 /**
  * Query if the current device is a tablet.
  *
@@ -934,6 +985,8 @@
   SDL_OnApplicationDidEnterForeground();
 }
 
+#if defined(SDL_PLATFORM_IOS) || defined(SDL3PP_DOC)
+
 /**
  * Let iOS apps with external event handling report
  * onApplicationDidChangeStatusBarOrientation.
@@ -954,8 +1007,14 @@
   SDL_OnApplicationDidChangeStatusBarOrientation();
 }
 
+#endif // defined(SDL_PLATFORM_IOS)
+
+#if defined(SDL_PLATFORM_GDK) || defined(SDL3PP_DOC)
+
+/// XTaskQueueHandle
 using XTaskQueueHandle = ::XTaskQueueHandle;
 
+/// XUserHandle
 using XUserHandle = ::XUserHandle;
 
 /**
@@ -991,6 +1050,7 @@
 {
   return SDL_GetGDKDefaultUser(outUserHandle);
 }
+#endif // defined(SDL_PLATFORM_GDK)
 
 /// @}
 
