--- src/generated/SDL3pp_haptic.h
+++ include/SDL3pp/SDL3pp_haptic.h
@@ -9,7 +9,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryHaptic Category Haptic
+ * @defgroup CategoryHaptic Force Feedback Support
  *
  * The SDL haptic subsystem manages haptic (force feedback) devices.
  *
@@ -140,6 +140,15 @@
   constexpr operator HapticRaw() const { return value; }
 };
 
+/**
+ * @name Haptic effects
+ * @{
+ */
+
+/**
+ * Haptic effects flag constants
+ *
+ */
 using HapticEffectType = Uint32;
 
 /**
@@ -353,6 +362,14 @@
  */
 constexpr HapticEffectType HAPTIC_PAUSE = SDL_HAPTIC_PAUSE;
 
+/// @}
+
+/**
+ * @name Direction encodings
+ * @{
+ */
+
+/// Direction encoding
 using HapticDirectionType = Uint8;
 
 /**
@@ -394,6 +411,8 @@
  */
 constexpr HapticDirectionType HAPTIC_STEERING_AXIS = SDL_HAPTIC_STEERING_AXIS;
 
+/// @}
+
 /**
  * Used to play a device an infinite number of times.
  *
@@ -1268,17 +1287,19 @@
 /**
  * Get a list of currently connected haptic devices.
  *
- * @param count a pointer filled in with the number of haptic devices returned,
- *              may be nullptr.
  * @returns a 0 terminated array of haptic device instance IDs or nullptr on
- *          failure; call GetError() for more information. This should be freed
- *          with free() when it is no longer needed.
+ *          failure; call GetError() for more information.
  *
  * @since This function is available since SDL 3.2.0.
  *
  * @sa Haptic.Haptic
  */
-inline OwnArray<HapticID> GetHaptics() { return SDL_GetHaptics(); }
+inline OwnArray<HapticID> GetHaptics()
+{
+  int count;
+  auto data = SDL_GetHaptics(&count);
+  return OwnArray<HapticID>{data};
+}
 
 /**
  * Get the implementation dependent name of a haptic device.
@@ -1335,7 +1356,7 @@
  */
 inline HapticRef GetHapticFromID(HapticID instance_id)
 {
-  return CheckError(SDL_GetHapticFromID(instance_id));
+  return {CheckError(SDL_GetHapticFromID(instance_id))};
 }
 
 /**
