--- src/generated/SDL3pp_mouse.h
+++ include/SDL3pp/SDL3pp_mouse.h
@@ -8,7 +8,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryMouse Category Mouse
+ * @defgroup CategoryMouse Mouse Support
  *
  * Any GUI application has to deal with the mouse, and SDL provides functions to
  * manage mouse input and the displayed cursor.
@@ -240,12 +240,9 @@
    *
    * @param data the color value for each pixel of the cursor.
    * @param mask the mask value for each pixel of the cursor.
-   * @param w the width of the cursor.
-   * @param h the height of the cursor.
-   * @param hot_x the x-axis offset from the left of the cursor image to the
-   *              mouse x position, in the range of 0 to `w` - 1.
-   * @param hot_y the y-axis offset from the top of the cursor image to the
-   *              mouse y position, in the range of 0 to `h` - 1.
+   * @param size the width and height of the cursor.
+   * @param hot the x position of the cursor hot spot, from the top-left, in the
+   *            range of 0 to `size.x` - 1 and 0 to `size.y` - 1.
    * @post a new cursor with the specified parameters on success.
    * @throws Error on failure.
    *
@@ -263,7 +260,8 @@
          const Uint8* mask,
          const PointRaw& size,
          const PointRaw& hot)
-    : m_resource(CheckError(SDL_CreateCursor(data, mask, size, hot)))
+    : m_resource(
+        CheckError(SDL_CreateCursor(data, mask, size.x, size.y, hot.x, hot.y)))
   {
   }
 
@@ -282,8 +280,7 @@
    * used instead.
    *
    * @param surface an Surface structure representing the cursor image.
-   * @param hot_x the x position of the cursor hot spot.
-   * @param hot_y the y position of the cursor hot spot.
+   * @param hot the x, y position of the cursor hot spot.
    * @post the new cursor on success.
    * @throws Error on failure.
    *
@@ -299,7 +296,7 @@
    * @sa Cursor.Set
    */
   Cursor(SurfaceParam surface, const PointRaw& hot)
-    : m_resource(CheckError(SDL_CreateColorCursor(surface, hot)))
+    : m_resource(CheckError(SDL_CreateColorCursor(surface, hot.x, hot.y)))
   {
   }
 
@@ -455,17 +452,21 @@
 
 #endif // SDL_VERSION_ATLEAST(3, 4, 0)
 
+/**
+ * Represents a button index.
+ *
+ */
 using MouseButton = Uint8;
 
-constexpr MouseButton BUTTON_LEFT = SDL_BUTTON_LEFT; ///< LEFT
+constexpr MouseButton BUTTON_LEFT = SDL_BUTTON_LEFT; ///< Left button
 
-constexpr MouseButton BUTTON_MIDDLE = SDL_BUTTON_MIDDLE; ///< MIDDLE
+constexpr MouseButton BUTTON_MIDDLE = SDL_BUTTON_MIDDLE; ///< Middle button
 
-constexpr MouseButton BUTTON_RIGHT = SDL_BUTTON_RIGHT; ///< RIGHT
+constexpr MouseButton BUTTON_RIGHT = SDL_BUTTON_RIGHT; ///< Right button
 
-constexpr MouseButton BUTTON_X1 = SDL_BUTTON_X1; ///< X1
+constexpr MouseButton BUTTON_X1 = SDL_BUTTON_X1; ///< X1 button
 
-constexpr MouseButton BUTTON_X2 = SDL_BUTTON_X2; ///< X2
+constexpr MouseButton BUTTON_X2 = SDL_BUTTON_X2; ///< X2 button
 
 /**
  * A bitmask of pressed mouse buttons, as reported by GetMouseState, etc.
@@ -494,6 +495,7 @@
 
 constexpr MouseButtonFlags BUTTON_X2MASK = SDL_BUTTON_X2MASK; ///< X2MASK
 
+/** Returns mask for button */
 constexpr MouseButtonFlags ButtonMask(MouseButton button)
 {
   return SDL_BUTTON_MASK(button);
@@ -531,16 +533,7 @@
  *
  * @sa SetRelativeMouseTransform
  */
-using MouseMotionTransformCallback = void(SDLCALL*)(void* userdata,
-                                                    Uint64 timestamp,
-                                                    SDL_Window* window,
-                                                    MouseID mouseID,
-                                                    float* x,
-                                                    float* y);
-
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
+using MouseMotionTransformCallback = SDL_MouseMotionTransformCallback;
 
 /**
  * A callback used to transform mouse motion delta from raw values.
@@ -601,11 +594,8 @@
  * functionality, including some game controllers, KVM switches, etc. You should
  * wait for input from a device before you consider it actively in use.
  *
- * @param count a pointer filled in with the number of mice returned, may be
- *              nullptr.
- * @returns a 0 terminated array of mouse instance IDs or nullptr on failure;
- *          call GetError() for more information. This should be freed with
- *          free() when it is no longer needed.
+ * @returns a 0 terminated array of mouse instance IDs.
+ * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
  *
@@ -614,7 +604,12 @@
  * @sa GetMouseNameForID
  * @sa HasMouse
  */
-inline OwnArray<MouseID> GetMice() { return SDL_GetMice(); }
+inline OwnArray<MouseID> GetMice()
+{
+  int count;
+  auto data = CheckError(SDL_GetMice(&count));
+  return OwnArray<MouseID>{data, size_t(count)};
+}
 
 /**
  * Get the name of a mouse.
@@ -645,7 +640,7 @@
  *
  * @since This function is available since SDL 3.2.0.
  */
-inline WindowRef GetMouseFocus() { return SDL_GetMouseFocus(); }
+inline WindowRef GetMouseFocus() { return {SDL_GetMouseFocus()}; }
 
 /**
  * Query SDL's cache for the synchronous mouse button state and the
@@ -763,7 +758,7 @@
 
 inline void Window::WarpMouse(const FPointRaw& p)
 {
-  SDL_WarpMouseInWindow(m_resource, p);
+  SDL_WarpMouseInWindow(m_resource, p.x, p.y);
 }
 
 /**
@@ -777,8 +772,7 @@
  * Note that this function will appear to succeed, but not actually move the
  * mouse when used over Microsoft Remote Desktop.
  *
- * @param x the x coordinate.
- * @param y the y coordinate.
+ * @param p the x, y coordinates;
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -789,7 +783,7 @@
  */
 inline void WarpMouse(const FPointRaw& p)
 {
-  CheckError(SDL_WarpMouseGlobal(p));
+  CheckError(SDL_WarpMouseGlobal(p.x, p.y));
 }
 
 #if SDL_VERSION_ATLEAST(3, 4, 0)
@@ -815,10 +809,6 @@
   CheckError(SDL_SetRelativeMouseTransform(callback, userdata));
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Set a user-defined function by which to transform relative mouse inputs.
  *
@@ -827,7 +817,6 @@
  *
  * @param callback a callback used to transform relative mouse motion, or
  *                 nullptr for default behavior.
- * @param userdata a pointer that will be passed to `callback`.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -836,7 +825,7 @@
  */
 inline void SetRelativeMouseTransform(MouseMotionTransformCB callback)
 {
-  static_assert(false, "Not implemented");
+  SetRelativeMouseTransform(callback.wrapper, callback.data);
 }
 
 #endif // SDL_VERSION_ATLEAST(3, 4, 0)
@@ -926,12 +915,9 @@
  *
  * @param data the color value for each pixel of the cursor.
  * @param mask the mask value for each pixel of the cursor.
- * @param w the width of the cursor.
- * @param h the height of the cursor.
- * @param hot_x the x-axis offset from the left of the cursor image to the mouse
- *              x position, in the range of 0 to `w` - 1.
- * @param hot_y the y-axis offset from the top of the cursor image to the mouse
- *              y position, in the range of 0 to `h` - 1.
+ * @param size the width and height of the cursor.
+ * @param hot the x position of the cursor hot spot, from the top-left, in the
+ *            range of 0 to `size.x` - 1 and 0 to `size.y` - 1.
  * @returns a new cursor with the specified parameters on success.
  * @throws Error on failure.
  *
@@ -967,8 +953,7 @@
  * closest smaller image will be upscaled and be used instead.
  *
  * @param surface an Surface structure representing the cursor image.
- * @param hot_x the x position of the cursor hot spot.
- * @param hot_y the y position of the cursor hot spot.
+ * @param hot the position of the cursor hot spot.
  * @returns the new cursor on success.
  * @throws Error on failure.
  *
@@ -1095,7 +1080,7 @@
  *
  * @sa Cursor.Set
  */
-inline CursorRef GetCursor() { return SDL_GetCursor(); }
+inline CursorRef GetCursor() { return {SDL_GetCursor()}; }
 
 /**
  * Get the default cursor.
@@ -1113,7 +1098,7 @@
  */
 inline CursorRef GetDefaultCursor()
 {
-  return CheckError(SDL_GetDefaultCursor());
+  return {CheckError(SDL_GetDefaultCursor())};
 }
 
 /**
