--- src/generated/SDL3pp_locale.h
+++ include/SDL3pp/SDL3pp_locale.h
@@ -7,7 +7,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryLocale Category Locale
+ * @defgroup CategoryLocale Locale Info
  *
  * SDL locale services.
  *
@@ -65,8 +65,6 @@
  * possible, and you can call this function again to get an updated copy of
  * preferred locales.
  *
- * @param count a pointer filled in with the number of locales returned, may be
- *              nullptr.
  * @returns a nullptr terminated array of locale pointers on success.
  * @throws Error on failure.
  *
@@ -76,7 +74,9 @@
  */
 inline OwnArray<Locale*> GetPreferredLocales()
 {
-  return CheckError(SDL_GetPreferredLocales());
+  int count = 0;
+  auto data = SDL_GetPreferredLocales(&count);
+  return OwnArray<Locale*>{CheckError(data), size_t(count)};
 }
 
 /// @}
