--- src/generated/SDL3pp_touch.h
+++ include/SDL3pp/SDL3pp_touch.h
@@ -9,7 +9,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryTouch Category Touch
+ * @defgroup CategoryTouch Touch Support
  *
  * SDL offers touch input, on platforms that support it. It can manage multiple
  * touch devices and track multiple fingers on those devices.
@@ -120,10 +120,7 @@
    *
    * @returns True if valid state, false otherwise.
    */
-  constexpr explicit operator bool() const noexcept
-  {
-    return *this != FingerRaw{};
-  }
+  constexpr explicit operator bool() const noexcept { return id != 0; }
 
   /**
    * Get the id.
@@ -223,15 +220,17 @@
  * Therefore the returned list might be empty, although devices are available.
  * After using all devices at least once the number will be correct.
  *
- * @param count a pointer filled in with the number of devices returned, may be
- *              nullptr.
  * @returns a 0 terminated array of touch device IDs or nullptr on failure; call
- *          GetError() for more information. This should be freed with free()
- *          when it is no longer needed.
+ *          GetError() for more information.
  *
  * @since This function is available since SDL 3.2.0.
  */
-inline OwnArray<TouchID> GetTouchDevices() { return SDL_GetTouchDevices(); }
+inline OwnArray<TouchID> GetTouchDevices()
+{
+  int count = 0;
+  auto data = SDL_GetTouchDevices(&count);
+  return OwnArray<TouchID>(data, size_t(count));
+}
 
 /**
  * Get the touch device name as reported from the driver.
@@ -264,17 +263,16 @@
  * Get a list of active fingers for a given touch device.
  *
  * @param touchID the ID of a touch device.
- * @param count a pointer filled in with the number of fingers returned, can be
- *              nullptr.
  * @returns a nullptr terminated array of Finger pointers or nullptr on failure;
- *          call GetError() for more information. This is a single allocation
- *          that should be freed with free() when it is no longer needed.
+ *          call GetError() for more information.
  *
  * @since This function is available since SDL 3.2.0.
  */
 inline OwnArray<Finger*> GetTouchFingers(TouchID touchID)
 {
-  return SDL_GetTouchFingers(touchID);
+  int count = 0;
+  auto data = reinterpret_cast<Finger**>(SDL_GetTouchFingers(touchID, &count));
+  return OwnArray<Finger*>(data, size_t(count));
 }
 
 /// @}
