--- src/generated/SDL3pp_properties.h
+++ include/SDL3pp/SDL3pp_properties.h
@@ -10,14 +10,15 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryProperties Category Properties
+ * @defgroup CategoryProperties Object Properties
  *
  * A property is a variable that can be created and retrieved by name at
  * runtime.
  *
  * All properties are part of a property group (Properties). A property group
  * can be created with the Properties.Create function and destroyed with the
- * Properties.Destroy function.
+ * Properties.Destroy function, but the Properties destructor probably will do
+ * what you want to, automatically.
  *
  * Properties can be added to and retrieved from a property group through the
  * following functions:
@@ -134,8 +135,6 @@
  * @since This datatype is available since SDL 3.2.0.
  *
  * @sa Properties.Enumerate
- *
- * @sa EnumeratePropertiesCallback
  */
 using EnumeratePropertiesCB =
   std::function<void(PropertiesID props, const char* name)>;
@@ -197,6 +196,8 @@
  * @since This datatype is available since SDL 3.2.0.
  *
  * @cat resource
+ *
+ * @sa Properties.Create
  */
 class Properties
 {
@@ -241,8 +242,8 @@
    *
    * All properties are automatically destroyed when Quit() is called.
    *
-   * @returns an ID for a new group of properties, or 0 on failure; call
-   *          GetError() for more information.
+   * @returns a valid ID for a new group of properties on success;
+   * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
@@ -407,7 +408,6 @@
    *              property.
    * @param cleanup the function to call when this property is deleted, or
    *                nullptr if no cleanup is necessary.
-   * @param userdata a pointer that is passed to the cleanup function.
    * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread.
@@ -690,7 +690,6 @@
    * properties. The properties are locked during enumeration.
    *
    * @param callback the function to call for each property.
-   * @param userdata a pointer that is passed to `callback`.
    * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread.
@@ -699,6 +698,13 @@
    */
   void Enumerate(EnumeratePropertiesCB callback);
 
+  /**
+   * Returns the number of properties this has
+   *
+   * This uses EnumerateProperties() internally, so might not be so fast
+   *
+   * @return Uint64
+   */
   Uint64 GetCount();
 };
 
@@ -780,7 +786,7 @@
  */
 inline PropertiesRef GetGlobalProperties()
 {
-  return CheckError(SDL_GetGlobalProperties());
+  return {CheckError(SDL_GetGlobalProperties())};
 }
 
 /**
@@ -788,8 +794,8 @@
  *
  * All properties are automatically destroyed when Quit() is called.
  *
- * @returns an ID for a new group of properties, or 0 on failure; call
- *          GetError() for more information.
+ * @returns a valid ID for a new group of properties on success;
+ * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -799,7 +805,7 @@
  */
 inline Properties CreateProperties()
 {
-  return Properties(SDL_CreateProperties());
+  return Properties(CheckError(SDL_CreateProperties()));
 }
 
 inline Properties Properties::Create() { return SDL::CreateProperties(); }
@@ -935,7 +941,6 @@
  *              property.
  * @param cleanup the function to call when this property is deleted, or nullptr
  *                if no cleanup is necessary.
- * @param userdata a pointer that is passed to the cleanup function.
  * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
@@ -951,7 +956,12 @@
                                           void* value,
                                           CleanupPropertyCB cleanup)
 {
-  static_assert(false, "Not implemented");
+  using Wrapper = CallbackWrapper<CleanupPropertyCB>;
+  SDL_SetPointerPropertyWithCleanup(props,
+                                    std::move(name),
+                                    value,
+                                    &Wrapper::CallOnce,
+                                    Wrapper::Wrap(std::move(cleanup)));
 }
 
 inline void Properties::SetPointerPropertyWithCleanup(
@@ -1387,7 +1397,6 @@
  *
  * @param props the properties to query.
  * @param callback the function to call for each property.
- * @param userdata a pointer that is passed to `callback`.
  * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
@@ -1397,7 +1406,13 @@
 inline void EnumerateProperties(PropertiesParam props,
                                 EnumeratePropertiesCB callback)
 {
-  static_assert(false, "Not implemented");
+  return EnumerateProperties(
+    props,
+    [](void* userdata, PropertiesID props, const char* name) {
+      auto& f = *static_cast<EnumeratePropertiesCB*>(userdata);
+      f(props, name);
+    },
+    &callback);
 }
 
 inline void Properties::Enumerate(EnumeratePropertiesCallback callback,
@@ -1411,9 +1426,19 @@
   SDL::EnumerateProperties(m_resource, callback);
 }
 
+/**
+ * Returns the number of properties this has
+ *
+ * This uses EnumerateProperties() internally, so might not be so fast
+ *
+ * @param props
+ * @return Uint64
+ */
 inline Uint64 CountProperties(PropertiesParam props)
 {
-  static_assert(false, "Not implemented");
+  Uint64 count = 0;
+  EnumerateProperties(props, [&](auto, const char*) { count++; });
+  return count;
 }
 
 inline Uint64 Properties::GetCount()
