--- src/generated/SDL3pp_mixer.h
+++ include/SDL3pp/SDL3pp_mixer.h
@@ -1,10 +1,13 @@
 #ifndef SDL3PP_MIXER_H_
 #define SDL3PP_MIXER_H_
 
-#include <SDL3/SDL_mixer.h>
 #include "SDL3pp_audio.h"
 #include "SDL3pp_version.h"
 
+#if defined(SDL3PP_ENABLE_MIXER) || defined(SDL3PP_DOC)
+
+#include <SDL3_mixer/SDL_mixer.h>
+
 namespace SDL {
 
 /**
@@ -22,9 +25,8 @@
  * To use the library, first call MIX.Init(). Then create a mixer with
  * Mixer.Mixer() (or Mixer.Mixer() to render to memory).
  *
- * Once you have a mixer, you can load sound data with Audio.Audio(),
- * Audio.Audio(), or Audio.Audio(). Data gets loaded once and can be played over
- * and over.
+ * Once you have a mixer, you can load sound data with Audio.Audio(). Data gets
+ * loaded once and can be played over and over.
  *
  * When loading audio, SDL_mixer can parse out several metadata tag formats,
  * such as ID3 and APE tags, and exposes this information through the
@@ -515,7 +517,7 @@
    *
    * This function loads data from an IOStream. There is also a version that
    * loads from a path on the filesystem (Audio.Audio()), and one that accepts
-   * properties for ultimate control (Audio.Audio()).
+   * properties for ultimate control (LoadAudioWithProperties()).
    *
    * The IOStream provided must be able to seek, or loading will fail. If the
    * stream can't seek (data is coming from an HTTP connection, etc), consider
@@ -526,9 +528,8 @@
    * @param predecode if true, data will be fully uncompressed before returning.
    * @param closeio true if SDL_mixer should close `io` before returning
    *                (success or failure).
-   * @returns an audio object that can be used to make sound on a mixer on
-   *          success.
-   * @throws Error on failure.
+   * @returns an audio object that can be used to make sound on a mixer, or
+   *          nullptr on failure; call GetError() for more information.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
@@ -537,7 +538,6 @@
    * @sa Audio.Destroy
    * @sa Track.SetAudio
    * @sa Audio.Audio
-   * @sa Audio.Audio
    */
   Audio LoadAudio_IO(IOStreamRef io, bool predecode, bool closeio = false);
 
@@ -546,19 +546,18 @@
    *
    * This is equivalent to calling:
    *
-   * ```c
-   * Audio.Audio(mixer, IOStream.FromFile(path, "rb"), predecode, true);
+   * ```cpp
+   * Audio audio(mixer, IOStream.FromFile(path, "rb"), predecode, true);
    * ```
    *
    * This function loads data from a path on the filesystem. There is also a
    * version that loads from an IOStream (Audio.Audio()), and one that accepts
-   * properties for ultimate control (Audio.Audio()).
+   * properties for ultimate control (LoadAudioWithProperties()).
    *
    * @param path the path on the filesystem to load data from.
    * @param predecode if true, data will be fully uncompressed before returning.
-   * @returns an audio object that can be used to make sound on a mixer on
-   *          success.
-   * @throws Error on failure.
+   * @returns an audio object that can be used to make sound on a mixer, or
+   *          nullptr on failure; call GetError() for more information.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
@@ -567,7 +566,6 @@
    * @sa Audio.Destroy
    * @sa Track.SetAudio
    * @sa Audio.Audio
-   * @sa Audio.Audio
    */
   Audio LoadAudio(StringParam path, bool predecode);
 
@@ -586,11 +584,11 @@
    * This function is meant to maximize efficiency: if the data is already in
    * memory and can remain there, don't copy it. This data can be in any
    * supported audio file format (WAV, MP3, etc); it will be decoded on the fly
-   * while mixing. Unlike Audio.Audio(), there is no `predecode` option offered
-   * here, as this is meant to optimize for data that's already in memory and
-   * intends to exist there for significant time; since predecoding would only
-   * need the file format data once, upfront, one could simply wrap it in
-   * SDL_CreateIOFromConstMem() and pass that to Audio.Audio().
+   * while mixing. Unlike Mixer.LoadAudio(), there is no `predecode` option
+   * offered here, as this is meant to optimize for data that's already in
+   * memory and intends to exist there for significant time; since predecoding
+   * would only need the file format data once, upfront, one could simply wrap
+   * it in SDL_CreateIOFromConstMem() and pass that to Mixer.LoadAudio_IO().
    *
    * Audio objects can be shared between multiple mixers. The `mixer` parameter
    * just suggests the most likely mixer to use this audio, in case some
@@ -661,7 +659,7 @@
    * @sa Track.SetAudio
    * @sa Audio.Audio
    * @sa Mixer.LoadRawAudioNoCopy
-   * @sa Audio.Audio
+   * @sa Mixer.LoadAudio_IO
    */
   Audio LoadRawAudio_IO(IOStreamRef io,
                         const AudioSpec& spec,
@@ -699,7 +697,7 @@
    * @sa Track.SetAudio
    * @sa Audio.Audio
    * @sa Mixer.LoadRawAudioNoCopy
-   * @sa Audio.Audio
+   * @sa Mixer.LoadAudio_IO
    */
   Audio LoadRawAudio(SourceBytes data, const AudioSpec& spec);
 
@@ -741,7 +739,7 @@
    * @sa Track.SetAudio
    * @sa Audio.Audio
    * @sa Audio.Audio
-   * @sa Audio.Audio
+   * @sa Mixer.LoadAudio_IO
    */
   Audio LoadRawAudioNoCopy(SourceBytes data,
                            const AudioSpec& spec,
@@ -780,7 +778,7 @@
    *
    * @sa Audio.Destroy
    * @sa Track.SetAudio
-   * @sa Audio.Audio
+   * @sa Mixer.LoadAudio_IO
    */
   Audio CreateSineWaveAudio(int hz, float amplitude, Sint64 ms);
 
@@ -817,7 +815,7 @@
    * @param tag the tag to search.
    * @param count a pointer filled in with the number of tracks returned, can be
    *              nullptr.
-   * @returns  nullptr-terminated on success.
+   * @returns an array of the tracks, nullptr-terminated on success.
    * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread.
@@ -889,7 +887,7 @@
    * @since This function is available since SDL_mixer 3.0.0.
    *
    * @sa Track.Play
-   * @sa Audio.Audio
+   * @sa Mixer.LoadAudio
    */
   bool PlayAudio(AudioRef audio);
 
@@ -1267,7 +1265,7 @@
    * initialized, if available tracks to mix run out, the end of the buffer will
    * be initialized with silence; this silence will not be counted in the return
    * value, so the caller has the option to identify how much of the buffer has
-   * legimitate contents vs appended silence. As such, any value >= 0 signifies
+   * legitimate contents vs appended silence. As such, any value >= 0 signifies
    * success. A return value of -1 means failure (out of memory, invalid
    * parameters, etc).
    *
@@ -1444,8 +1442,8 @@
    * @param predecode if true, data will be fully uncompressed before returning.
    * @param closeio true if SDL_mixer should close `io` before returning
    *                (success or failure).
-   * @post an audio object that can be used to make sound on a mixer on success.
-   * @throws Error on failure.
+   * @post an audio object that can be used to make sound on a mixer, or nullptr
+   *       on failure; call GetError() for more information.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
@@ -1475,8 +1473,8 @@
    *              nullptr.
    * @param path the path on the filesystem to load data from.
    * @param predecode if true, data will be fully uncompressed before returning.
-   * @post an audio object that can be used to make sound on a mixer on success.
-   * @throws Error on failure.
+   * @post an audio object that can be used to make sound on a mixer, or nullptr
+   *       on failure; call GetError() for more information.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
@@ -1569,7 +1567,7 @@
    * @sa Track.SetAudio
    * @sa Audio.Audio
    * @sa Mixer.LoadRawAudioNoCopy
-   * @sa Audio.Audio
+   * @sa Mixer.LoadAudio_IO
    */
   Audio(MixerRef mixer,
         IOStreamRef io,
@@ -1609,7 +1607,7 @@
    * @sa Track.SetAudio
    * @sa Audio.Audio
    * @sa Mixer.LoadRawAudioNoCopy
-   * @sa Audio.Audio
+   * @sa Mixer.LoadAudio_IO
    */
   Audio(MixerRef mixer, SourceBytes data, const AudioSpec& spec);
 
@@ -1920,7 +1918,7 @@
  * A callback that fires when a Track is mixing at various stages.
  *
  * This callback is fired for different parts of the mixing pipeline, and gives
- * the app visbility into the audio data that is being generated at various
+ * the app visibility into the audio data that is being generated at various
  * stages.
  *
  * The audio data passed through here is _not_ const data; the app is permitted
@@ -2348,7 +2346,7 @@
    *
    * @param count a pointer filled in with the number of tags returned, can be
    *              nullptr.
-   * @returns  nullptr-terminated on success.
+   * @returns an array of the tags, nullptr-terminated on success.
    * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread.
@@ -3553,6 +3551,8 @@
   constexpr operator GroupRaw() const noexcept { return get(); }
 };
 
+#ifdef SDL3PP_DOC
+
 /**
  * The current major version of SDL_mixer headers.
  *
@@ -3602,6 +3602,8 @@
    (SDL_MIXER_MAJOR_VERSION > X || SDL_MIXER_MINOR_VERSION > Y ||              \
     SDL_MIXER_MICRO_VERSION >= Z))
 
+#endif // SDL3PP_DOC
+
 namespace MIX {
 
 /**
@@ -4028,7 +4030,7 @@
  *
  * This function loads data from an IOStream. There is also a version that loads
  * from a path on the filesystem (Audio.Audio()), and one that accepts
- * properties for ultimate control (Audio.Audio()).
+ * properties for ultimate control (LoadAudioWithProperties()).
  *
  * The IOStream provided must be able to seek, or loading will fail. If the
  * stream can't seek (data is coming from an HTTP connection, etc), consider
@@ -4040,9 +4042,8 @@
  * @param predecode if true, data will be fully uncompressed before returning.
  * @param closeio true if SDL_mixer should close `io` before returning (success
  *                or failure).
- * @returns an audio object that can be used to make sound on a mixer on
- *          success.
- * @throws Error on failure.
+ * @returns an audio object that can be used to make sound on a mixer, or
+ *          nullptr on failure; call GetError() for more information.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -4051,7 +4052,7 @@
  * @sa Audio.Destroy
  * @sa Track.SetAudio
  * @sa Audio.Audio
- * @sa Audio.Audio
+ * @sa LoadAudioWithProperties
  */
 inline Audio LoadAudio_IO(MixerRef mixer,
                           IOStreamRef io,
@@ -4070,12 +4071,12 @@
                     IOStreamRef io,
                     bool predecode,
                     bool closeio)
-  : m_resource(CheckError(MIX_LoadAudio_IO(mixer, io, predecode, closeio)))
+  : m_resource(MIX_LoadAudio_IO(mixer, io, predecode, closeio))
 {
 }
 
 inline Audio::Audio(MixerRef mixer, StringParam path, bool predecode)
-  : m_resource(CheckError(MIX_LoadAudio(mixer, path, predecode)))
+  : m_resource(MIX_LoadAudio(mixer, path, predecode))
 {
 }
 
@@ -4103,20 +4104,19 @@
  *
  * This is equivalent to calling:
  *
- * ```c
- * Audio.Audio(mixer, IOStream.FromFile(path, "rb"), predecode, true);
+ * ```cpp
+ * Audio audio(mixer, IOStream.FromFile(path, "rb"), predecode, true);
  * ```
  *
  * This function loads data from a path on the filesystem. There is also a
  * version that loads from an IOStream (Audio.Audio()), and one that accepts
- * properties for ultimate control (Audio.Audio()).
+ * properties for ultimate control (LoadAudioWithProperties()).
  *
  * @param mixer a mixer this audio is intended to be used with. May be nullptr.
  * @param path the path on the filesystem to load data from.
  * @param predecode if true, data will be fully uncompressed before returning.
- * @returns an audio object that can be used to make sound on a mixer on
- *          success.
- * @throws Error on failure.
+ * @returns an audio object that can be used to make sound on a mixer, or
+ *          nullptr on failure; call GetError() for more information.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -4124,8 +4124,8 @@
  *
  * @sa Audio.Destroy
  * @sa Track.SetAudio
- * @sa Audio.Audio
- * @sa Audio.Audio
+ * @sa Mixer.LoadAudio_IO
+ * @sa LoadAudioWithProperties
  */
 inline Audio LoadAudio(MixerRef mixer, StringParam path, bool predecode)
 {
@@ -4155,7 +4155,7 @@
  * this is meant to optimize for data that's already in memory and intends to
  * exist there for significant time; since predecoding would only need the file
  * format data once, upfront, one could simply wrap it in
- * SDL_CreateIOFromConstMem() and pass that to Audio.Audio().
+ * SDL_CreateIOFromConstMem() and pass that to Mixer.LoadAudio_IO().
  *
  * Audio objects can be shared between multiple mixers. The `mixer` parameter
  * just suggests the most likely mixer to use this audio, in case some
@@ -4192,14 +4192,14 @@
  * @sa Audio.Destroy
  * @sa Track.SetAudio
  * @sa Mixer.LoadRawAudioNoCopy
- * @sa Audio.Audio
+ * @sa Mixer.LoadAudio_IO
  */
 inline Audio LoadAudioNoCopy(MixerRef mixer,
                              SourceBytes data,
                              bool free_when_done)
 {
-  return CheckError(
-    MIX_LoadAudioNoCopy(mixer, data.data(), data.size_bytes(), free_when_done));
+  return Audio(CheckError(MIX_LoadAudioNoCopy(
+    mixer, data.data(), data.size_bytes(), free_when_done)));
 }
 
 inline Audio Mixer::LoadAudioNoCopy(SourceBytes data, bool free_when_done)
@@ -4251,7 +4251,6 @@
  * @sa Audio.Destroy
  * @sa Track.SetAudio
  * @sa Audio.Audio
- * @sa Audio.Audio
  */
 inline Audio LoadAudioWithProperties(PropertiesRef props)
 {
@@ -4308,7 +4307,6 @@
  * @sa Track.SetAudio
  * @sa Audio.Audio
  * @sa Mixer.LoadRawAudioNoCopy
- * @sa Audio.Audio
  */
 inline Audio LoadRawAudio_IO(MixerRef mixer,
                              IOStreamRef io,
@@ -4358,7 +4356,6 @@
  * @sa Track.SetAudio
  * @sa Audio.Audio
  * @sa Mixer.LoadRawAudioNoCopy
- * @sa Audio.Audio
  */
 inline Audio LoadRawAudio(MixerRef mixer,
                           SourceBytes data,
@@ -4411,15 +4408,14 @@
  * @sa Track.SetAudio
  * @sa Audio.Audio
  * @sa Audio.Audio
- * @sa Audio.Audio
  */
 inline Audio LoadRawAudioNoCopy(MixerRef mixer,
                                 SourceBytes data,
                                 const AudioSpec& spec,
                                 bool free_when_done)
 {
-  return CheckError(MIX_LoadRawAudioNoCopy(
-    mixer, data.data(), data.size_bytes(), &spec, free_when_done));
+  return Audio(CheckError(MIX_LoadRawAudioNoCopy(
+    mixer, data.data(), data.size_bytes(), &spec, free_when_done)));
 }
 
 inline Audio Mixer::LoadRawAudioNoCopy(SourceBytes data,
@@ -4471,7 +4467,7 @@
                                  float amplitude,
                                  Sint64 ms)
 {
-  return CheckError(MIX_CreateSineWaveAudio(mixer, hz, amplitude, ms));
+  return Audio(CheckError(MIX_CreateSineWaveAudio(mixer, hz, amplitude, ms)));
 }
 
 inline Audio Mixer::CreateSineWaveAudio(int hz, float amplitude, Sint64 ms)
@@ -4506,7 +4502,8 @@
  *   out of sound to generate. This isn't necessarily always known to SDL_mixer,
  *   though.
  *
- * Other properties, documented with Audio.Audio(), may also be present.
+ * Other properties, documented with LoadAudioWithProperties(), may also be
+ * present.
  *
  * Note that the metadata properties are whatever SDL_mixer finds in things like
  * ID3 tags, and they often have very little standardized formatting, may be
@@ -4594,8 +4591,10 @@
 
 inline Sint64 Audio::GetDuration() { return SDL::GetAudioDuration(m_resource); }
 
+/// Unknown duration, when the length of the audio can't be determined.
 constexpr Sint64 DURATION_UNKNOWN = MIX_DURATION_UNKNOWN;
 
+/// Infinite duration, when the audio never runs out of sound to generate.
 constexpr Sint64 DURATION_INFINITE = MIX_DURATION_INFINITE;
 
 /**
@@ -5037,7 +5036,7 @@
  * @param track the track to query.
  * @param count a pointer filled in with the number of tags returned, can be
  *              nullptr.
- * @returns  nullptr-terminated on success.
+ * @returns an array of the tags, nullptr-terminated on success.
  * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
@@ -5063,7 +5062,7 @@
  * @param tag the tag to search.
  * @param count a pointer filled in with the number of tracks returned, can be
  *              nullptr.
- * @returns  nullptr-terminated on success.
+ * @returns an array of the tracks, nullptr-terminated on success.
  * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
@@ -6934,7 +6933,7 @@
  * initialized, if available tracks to mix run out, the end of the buffer will
  * be initialized with silence; this silence will not be counted in the return
  * value, so the caller has the option to identify how much of the buffer has
- * legimitate contents vs appended silence. As such, any value >= 0 signifies
+ * legitimate contents vs appended silence. As such, any value >= 0 signifies
  * success. A return value of -1 means failure (out of memory, invalid
  * parameters, etc).
  *
@@ -7023,7 +7022,7 @@
    *
    * This function allows properties to be specified. This is intended to supply
    * file-specific settings, such as where to find SoundFonts for a MIDI file,
-   * etc. In most cases, the caller should pass a zero to specify no extra
+   * etc. In most cases, the caller should pass a nullptr to specify no extra
    * properties.
    *
    * Properties are discussed in [SDL's
@@ -7036,7 +7035,7 @@
    * call to MIX.Init(), but does not need an actual Mixer to have been created.
    *
    * @param path the path on the filesystem from which to load data.
-   * @param props decoder-specific properties. May be zero.
+   * @param props decoder-specific properties. May be nullptr.
    * @post an audio decoder, ready to decode.
    *
    * @threadsafety It is safe to call this function from any thread.
@@ -7058,7 +7057,7 @@
    *
    * This function allows properties to be specified. This is intended to supply
    * file-specific settings, such as where to find SoundFonts for a MIDI file,
-   * etc. In most cases, the caller should pass a zero to specify no extra
+   * etc. In most cases, the caller should pass a nullptr to specify no extra
    * properties.
    *
    * If `closeio` is true, then `io` will be closed when this decoder is done
@@ -7073,7 +7072,7 @@
    *
    * @param io the i/o stream from which to load data.
    * @param closeio if true, close the i/o stream when done with it.
-   * @param props decoder-specific properties. May be zero.
+   * @param props decoder-specific properties. May be nullptr.
    * @post an audio decoder, ready to decode.
    *
    * @threadsafety It is safe to call this function from any thread.
@@ -7266,7 +7265,7 @@
  *
  * This function allows properties to be specified. This is intended to supply
  * file-specific settings, such as where to find SoundFonts for a MIDI file,
- * etc. In most cases, the caller should pass a zero to specify no extra
+ * etc. In most cases, the caller should pass a nullptr to specify no extra
  * properties.
  *
  * Properties are discussed in [SDL's
@@ -7279,7 +7278,7 @@
  * call to MIX.Init(), but does not need an actual Mixer to have been created.
  *
  * @param path the path on the filesystem from which to load data.
- * @param props decoder-specific properties. May be zero.
+ * @param props decoder-specific properties. May be nullptr.
  * @returns an audio decoder, ready to decode.
  *
  * @threadsafety It is safe to call this function from any thread.
@@ -7317,7 +7316,7 @@
  *
  * This function allows properties to be specified. This is intended to supply
  * file-specific settings, such as where to find SoundFonts for a MIDI file,
- * etc. In most cases, the caller should pass a zero to specify no extra
+ * etc. In most cases, the caller should pass a nullptr to specify no extra
  * properties.
  *
  * If `closeio` is true, then `io` will be closed when this decoder is done with
@@ -7332,7 +7331,7 @@
  *
  * @param io the i/o stream from which to load data.
  * @param closeio if true, close the i/o stream when done with it.
- * @param props decoder-specific properties. May be zero.
+ * @param props decoder-specific properties. May be nullptr.
  * @returns an audio decoder, ready to decode.
  *
  * @threadsafety It is safe to call this function from any thread.
@@ -7467,4 +7466,6 @@
 
 } // namespace SDL
 
+#endif /* defined(SDL3PP_ENABLE_MIXER) || defined(SDL3PP_DOC) */
+
 #endif /* SDL3PP_MIXER_H_ */
