--- build/generated/SDL3pp_net.h
+++ include/SDL3pp/SDL3pp_net.h
@@ -1,14 +1,17 @@
 #ifndef SDL3PP_NET_H_
 #define SDL3PP_NET_H_
 
-#include <SDL3/SDL_net.h>
 #include "SDL3pp_properties.h"
 #include "SDL3pp_version.h"
 
+#ifdef SDL3PP_ENABLE_NET
+
+#include <SDL3_net/SDL_net.h>
+
 namespace SDL {
 
 /**
- * @defgroup CategorySDLNet Category SDLNet
+ * @defgroup CategorySDLNet A small cross-platform networking library
  *
  * SDL_net is a simple library to help with networking.
  *
@@ -181,6 +184,8 @@
 /// Safely wrap Datagram for non owning const parameters
 using DatagramConstRef = ResourceConstRef<DatagramRaw, DatagramRawConst>;
 
+#ifdef SDL3PP_DOC
+
 /**
  * The current major version of the SDL_net headers.
  *
@@ -230,6 +235,8 @@
    (SDL_NET_MAJOR_VERSION > X || SDL_NET_MINOR_VERSION > Y ||                  \
     SDL_NET_MICRO_VERSION >= Z))
 
+#endif /* SDL3PP_DOC */
+
 namespace NET {
 
 /**
@@ -761,12 +768,6 @@
   {
   }
 
-  /// Copy constructor
-  constexpr Address(const Address& other)
-    : Address(borrow(other.get()))
-  {
-  }
-
   /// Move constructor
   constexpr Address(Address&& other) noexcept
     : Address(other.release())
@@ -847,7 +848,7 @@
    *
    * @since This function is available since SDL_net 3.0.0.
    */
-  Address(AddressRef address);
+  Address(const Address& address);
 
   /**
    * Safely borrows the from AddressRaw.
@@ -859,7 +860,7 @@
   static Address borrow(AddressRaw resource)
   {
     if (resource) {
-      ++resource->NET_RefAddress;
+      NET_RefAddress(resource);
       return Address(resource);
     }
     return {};
@@ -929,8 +930,8 @@
 {
 }
 
-inline Address::Address(AddressRef address)
-  : Address(NET_RefAddress(address))
+inline Address::Address(const Address& address)
+  : Address(address ? NET_RefAddress(address.get()) : nullptr)
 {
 }
 
@@ -1231,17 +1232,18 @@
 
 inline bool AddressBase::operator==(const AddressBase& other) const
 {
-  static_assert(false, "Not implemented");
+  return Compare(other) == 0;
 }
 
 inline auto AddressBase::operator<=>(const AddressBase& other) const
 {
-  static_assert(false, "Not implemented");
+  return Compare(other) <=> 0;
 }
 
 /// @private
 struct LocalAddressesArrayDeleter
 {
+  /// @private
   static void operator()(AddressRef* addresses);
 };
 
@@ -1270,13 +1272,8 @@
  * nullptr-terminated. You can also pass a pointer to an int, which will return
  * the final count, not counting the nullptr at the end of the array.
  *
- * Pass the returned array to FreeLocalAddresses when you are done with it. It
- * is safe to keep any addresses you want from this array even after calling
- * that function, as long as you called RefAddress() on them.
- *
- * @param num_addresses on exit, will be set to the number of addresses
- *                      returned. Can be nullptr.
- * @returns  one for each bindable address on the system on success.
+ * @returns A nullptr-terminated array of AddressRef, one for each bindable
+ *          address on the system on success.
  * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
@@ -1301,15 +1298,19 @@
  *
  * It is safe to pass a nullptr in here, it will be ignored.
  *
+ * You most likely don't need to call this function directly, as the
+ * LocalAddressesArray type will call it for you when the array goes out of
+ * scope.
+ *
  * @param addresses A pointer returned by GetLocalAddresses().
  *
  * @threadsafety It is safe to call this function from any thread.
  *
  * @since This function is available since SDL_net 3.0.0.
  */
-inline void FreeLocalAddresses(NET_Address** addresses)
+inline void FreeLocalAddresses(AddressRef* addresses)
 {
-  NET_FreeLocalAddresses(addresses);
+  NET_FreeLocalAddresses(reinterpret_cast<AddressRaw*>(addresses));
 }
 
 inline void LocalAddressesArrayDeleter::operator()(AddressRef* addresses)
@@ -2247,7 +2248,7 @@
  */
 inline Address GetStreamSocketAddress(StreamSocketRef sock)
 {
-  return CheckError(NET_GetStreamSocketAddress(sock));
+  return Address(CheckError(NET_GetStreamSocketAddress(sock)));
 }
 
 inline Address StreamSocketBase::GetAddress()
@@ -2733,7 +2734,7 @@
    * should assume it is no longer usable and should destroy it with
    * SDL_DestroyDatagramSocket().
    *
-   * @param dgram a pointer to the datagram packet pointer.
+   * @param dgram a reference to the datagram packet object.
    * @returns true if data sent or queued for transmission, false on failure;
    *          call GetError() for details.
    *
@@ -2780,9 +2781,8 @@
    * should assume it is no longer usable and should destroy it with
    * SDL_DestroyDatagramSocket().
    *
-   * @param dgram a pointer to the datagram packet pointer.
-   * @returns true if data sent or queued for transmission, false on failure;
-   *          call GetError() for details.
+   * @return a valid Datagram object if data sent or queued for transmission,
+   *         nullptr on failure; call GetError() for details.
    *
    * @threadsafety You should not operate on the same socket from multiple
    *               threads at the same time without supplying a serialization
@@ -3037,7 +3037,6 @@
    * SDL_DestroyDatagramSocket().
    *
    * @param sock the datagram socket to send data through.
-   * @param dgram a pointer to the datagram packet pointer.
    * @returns true if data sent or queued for transmission, false on failure;
    *          call GetError() for details.
    *
@@ -3115,12 +3114,11 @@
    * "connection" to fail at this level, but may report failure for
    * unrecoverable system-level conditions; once a datagram socket fails, you
    * should assume it is no longer usable and should destroy it with
-   * SDL_DestroyDatagramSocket().
+   * DestroyDatagramSocket().
    *
    * @param sock the datagram socket to send data through.
-   * @param dgram a pointer to the datagram packet pointer.
-   * @post true if data sent or queued for transmission, false on failure; call
-   *       GetError() for details.
+   * @post a valid Datagram object if data sent or queued for transmission,
+   *       nullptr on failure; call GetError() for details.
    *
    * @threadsafety You should not operate on the same socket from multiple
    *               threads at the same time without supplying a serialization
@@ -3437,9 +3435,8 @@
  * is no longer usable and should destroy it with SDL_DestroyDatagramSocket().
  *
  * @param sock the datagram socket to send data through.
- * @param dgram a pointer to the datagram packet pointer.
- * @returns true if data sent or queued for transmission, false on failure; call
- *          GetError() for details.
+ * @return a valid Datagram object if data sent or queued for transmission,
+ *         nullptr on failure; call GetError() for details.
  *
  * @threadsafety You should not operate on the same socket from multiple threads
  *               at the same time without supplying a serialization mechanism.
@@ -3460,7 +3457,7 @@
 
 inline bool DatagramSocketBase::ReceiveDatagram(Datagram& dgram)
 {
-  return dgram.Receive(sock);
+  return dgram.Receive(*this);
 }
 
 inline Datagram DatagramSocketBase::ReceiveDatagram()
@@ -3468,11 +3465,17 @@
   return SDL::ReceiveDatagram(get());
 }
 
-inline Datagram::Datagram(DatagramSocketRef sock) {}
+inline Datagram::Datagram(DatagramSocketRef sock)
+  : Datagram(sock.ReceiveDatagram())
+{
+}
 
 inline bool DatagramBase::Receive(DatagramSocketRef sock)
 {
-  static_assert(false, "Not implemented");
+  DatagramRaw dgram;
+  if (!NET_ReceiveDatagram(sock, &dgram)) return false;
+  *this = Datagram(dgram);
+  return true;
 }
 
 /**
@@ -3632,4 +3635,6 @@
 
 } // namespace SDL
 
+#endif /* SDL3PP_ENABLE_NET */
+
 #endif /* SDL3PP_NET_H_ */
