--- src/generated/SDL3pp_audio.h
+++ include/SDL3pp/SDL3pp_audio.h
@@ -9,7 +9,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryAudio Category Audio
+ * @defgroup CategoryAudio Audio Playback, Recording, and Mixing
  *
  * Audio functionality for the SDL library.
  *
@@ -196,7 +196,7 @@
 /**
  * Mask of bits in an AudioFormat that contains the format bit size.
  *
- * Generally one should use AudioFormat.GetBitSize instead of this macro
+ * Generally one should use AudioFormat.GetBitSize instead of this constant
  * directly.
  *
  * @since This constant is available since SDL 3.2.0.
@@ -206,7 +206,8 @@
 /**
  * Mask of bits in an AudioFormat that contain the floating point flag.
  *
- * Generally one should use AudioFormat.IsFloat instead of this macro directly.
+ * Generally one should use AudioFormat.IsFloat instead of this constant
+ * directly.
  *
  * @since This constant is available since SDL 3.2.0.
  */
@@ -216,7 +217,7 @@
  * Mask of bits in an AudioFormat that contain the bigendian flag.
  *
  * Generally one should use AudioFormat.IsBigEndian or
- * AudioFormat.IsLittleEndian instead of this macro directly.
+ * AudioFormat.IsLittleEndian instead of this constant directly.
  *
  * @since This constant is available since SDL 3.2.0.
  */
@@ -225,7 +226,8 @@
 /**
  * Mask of bits in an AudioFormat that contain the signed data flag.
  *
- * Generally one should use AudioFormat.IsSigned instead of this macro directly.
+ * Generally one should use AudioFormat.IsSigned instead of this constant
+ * directly.
  *
  * @since This constant is available since SDL 3.2.0.
  */
@@ -282,7 +284,7 @@
    * AudioFormat.AudioFormat(1, 0, 0, 32)
    * ```
    *
-   * @param signed 1 for signed data, 0 for unsigned data.
+   * @param sign 1 for signed data, 0 for unsigned data.
    * @param bigendian 1 for bigendian data, 0 for littleendian data.
    * @param flt 1 for floating point data, 0 for integer data.
    * @param size number of bits per sample.
@@ -293,7 +295,8 @@
    * @since This function is available since SDL 3.2.0.
    */
   constexpr AudioFormat(bool sign, bool bigendian, bool flt, Uint16 size)
-    : m_audioFormat(SDL_DEFINE_AUDIO_FORMAT(sign, bigendian, flt, size))
+    : m_audioFormat(
+        AudioFormatRaw(SDL_DEFINE_AUDIO_FORMAT(sign, bigendian, flt, size)))
   {
   }
 
@@ -435,6 +438,13 @@
   int GetSilenceValue() const;
 };
 
+// Unfortunate name clash with SDL_oldnames.h
+#undef AUDIO_U8
+#undef AUDIO_S8
+#undef AUDIO_S16
+#undef AUDIO_S32
+#undef AUDIO_F32
+
 constexpr AudioFormat AUDIO_UNKNOWN =
   SDL_AUDIO_UNKNOWN; ///< Unspecified audio format
 
@@ -477,7 +487,7 @@
  * AudioFormat.AudioFormat(1, 0, 0, 32)
  * ```
  *
- * @param signed 1 for signed data, 0 for unsigned data.
+ * @param sign 1 for signed data, 0 for unsigned data.
  * @param bigendian 1 for bigendian data, 0 for littleendian data.
  * @param flt 1 for floating point data, 0 for integer data.
  * @param size number of bits per sample.
@@ -722,7 +732,6 @@
  *             device.
  * @param buffer the buffer of audio samples to be submitted. The callback can
  *               inspect and/or modify this data.
- * @param buflen the size of `buffer` in bytes.
  *
  * @threadsafety This will run from a background thread owned by SDL. The
  *               application is responsible for locking resources the callback
@@ -999,8 +1008,6 @@
    */
   void Close();
 
-  constexpr auto operator<=>(AudioDeviceParam other) const = default;
-
   /**
    * Get the human-readable name of a specific audio device.
    *
@@ -1013,8 +1020,8 @@
    * default might change in general, and the string will be the name of a
    * specific device and not the abstract system default.
    *
-   * @returns the name of the audio device, or nullptr on failure; call
-   *          GetError() for more information.
+   * @returns the name of the audio device on success.
+   * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
@@ -1047,9 +1054,9 @@
    * Buffer size is only important if you need low-level control over the audio
    * playback timing. Most apps do not need this.
    *
-   * @param spec on return, will be filled with device details.
    * @param sample_frames pointer to store device buffer size, in sample frames.
    *                      Can be nullptr.
+   * @returns the device format details on success.
    * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread.
@@ -1067,11 +1074,8 @@
    * Audio devices usually have no remapping applied. This is represented by
    * returning nullptr, and does not signify an error.
    *
-   * @param count On output, set to number of channels in the map. Can be
-   *              nullptr.
    * @returns an array of the current channel mapping, with as many elements as
-   *          the current output spec's channels, or nullptr if default. This
-   *          should be freed with free() when it is no longer needed.
+   *          the current output spec's channels, or nullptr if default
    *
    * @threadsafety It is safe to call this function from any thread.
    *
@@ -1273,7 +1277,6 @@
    * end's format.
    *
    * @param streams an array of audio streams to bind.
-   * @param num_streams number streams listed in the `streams` array.
    * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread.
@@ -1400,7 +1403,6 @@
    * Setting a nullptr callback function disables any previously-set callback.
    *
    * @param callback a callback function to be called. Can be nullptr.
-   * @param userdata app-controlled pointer passed to callback. Can be nullptr.
    * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread.
@@ -1464,9 +1466,9 @@
    * @sa AudioStream.GetDevice
    * @sa AudioStream.ResumeDevice
    */
-  AudioStreamRef OpenStream(const AudioSpec& spec,
-                            AudioStreamCallback callback,
-                            void* userdata);
+  AudioStream OpenStream(OptionalRef<const AudioSpec> spec,
+                         AudioStreamCallback callback,
+                         void* userdata);
 
   /**
    * Convenience function for straightforward audio init for the common case.
@@ -1508,11 +1510,7 @@
    *
    * @param spec the audio stream's data format. Can be nullptr.
    * @param callback a callback where the app will provide new data for
-   *                 playback, or receive new data for recording. Can be
-   *                 nullptr, in which case the app will need to call
-   *                 AudioStream.PutData or AudioStream.GetData as necessary.
-   * @param userdata app-controlled pointer passed to callback. Can be nullptr.
-   *                 Ignored if callback is nullptr.
+   *                 playback, or receive new data for recording.
    * @returns an audio stream on success.
    * @throws Error on failure.
    *
@@ -1572,7 +1570,7 @@
  *
  * @since This constant is available since SDL 3.2.0.
  */
-constexpr AudioDeviceRaw AUDIO_DEVICE_DEFAULT_PLAYBACK =
+constexpr AudioDeviceID AUDIO_DEVICE_DEFAULT_PLAYBACK =
   SDL_AUDIO_DEVICE_DEFAULT_PLAYBACK;
 
 /**
@@ -1584,7 +1582,7 @@
  *
  * @since This constant is available since SDL 3.2.0.
  */
-constexpr AudioDeviceRaw AUDIO_DEVICE_DEFAULT_RECORDING =
+constexpr AudioDeviceID AUDIO_DEVICE_DEFAULT_RECORDING =
   SDL_AUDIO_DEVICE_DEFAULT_RECORDING;
 
 /**
@@ -1602,7 +1600,7 @@
  */
 constexpr int AudioFrameSize(const AudioSpec& x)
 {
-  return SDL_AUDIO_FRAMESIZE(&x);
+  return SDL_AUDIO_FRAMESIZE(x);
 }
 
 #if SDL_VERSION_ATLEAST(3, 4, 0)
@@ -1637,10 +1635,6 @@
                                                        const void* buf,
                                                        int buflen);
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * A callback that fires for completed AudioStream.PutDataNoCopy() data.
  *
@@ -1828,7 +1822,59 @@
   {
   }
 
-  /// Default ctor
+  /**
+   * Convenience function for straightforward audio init for the common case.
+   *
+   * If all your app intends to do is provide a single source of PCM audio, this
+   * function allows you to do all your audio setup in a single call.
+   *
+   * This is also intended to be a clean means to migrate apps from SDL2.
+   *
+   * This function will open an audio device, create a stream and bind it.
+   * Unlike other methods of setup, the audio device will be closed when this
+   * stream is destroyed, so the app can treat the returned AudioStream as
+   * the only object needed to manage audio playback.
+   *
+   * Also unlike other functions, the audio device begins paused. This is to map
+   * more closely to SDL2-style behavior, since there is no extra step here to
+   * bind a stream to begin audio flowing. The audio device should be resumed
+   * with `AudioStream.ResumeDevice();`
+   *
+   * This function works with both playback and recording devices.
+   *
+   * The `spec` parameter represents the app's side of the audio stream. That
+   * is, for recording audio, this will be the output format, and for playing
+   * audio, this will be the input format. If spec is nullptr, the system will
+   * choose the format, and the app can use AudioStream.GetFormat() to obtain
+   * this information later.
+   *
+   * If you don't care about opening a specific audio device, you can (and
+   * probably _should_), use AUDIO_DEVICE_DEFAULT_PLAYBACK for playback and
+   * AUDIO_DEVICE_DEFAULT_RECORDING for recording.
+   *
+   * One can optionally provide a callback function; if nullptr, the app is
+   * expected to queue audio data for playback (or unqueue audio data if
+   * capturing). Otherwise, the callback will begin to fire once the device is
+   * unpaused.
+   *
+   * Destroying the returned stream with AudioStream.Destroy will also close
+   * the audio device associated with this stream.
+   *
+   * @param devid an audio device to open, or AUDIO_DEVICE_DEFAULT_PLAYBACK
+   *              or AUDIO_DEVICE_DEFAULT_RECORDING.
+   * @param spec the audio stream's data format. Can be nullptr.
+   * @param callback a callback where the app will provide new data for
+   *                 playback, or receive new data for recording.
+   * @post an audio stream on success.
+   * @throws Error on failure.
+   *
+   * @threadsafety It is safe to call this function from any thread.
+   *
+   * @since This function is available since SDL 3.2.0.
+   *
+   * @sa AudioStream.GetDevice
+   * @sa AudioStream.ResumeDevice
+   */
   AudioStream(AudioDeviceParam devid,
               OptionalRef<const AudioSpec> spec,
               AudioStreamCB callback);
@@ -1910,9 +1956,45 @@
    */
   PropertiesRef GetProperties() const;
 
-  AudioSpec GetInputFormat() const { static_assert(false, "Not implemented"); }
+  /**
+   * Query the current input format of an audio stream.
+   *
+   * @returns the input audio format.
+   * @throws Error on failure.
+   *
+   * @threadsafety It is safe to call this function from any thread, as it holds
+   *               a stream-specific mutex while running.
+   *
+   * @since This function is available since SDL 3.2.0.
+   *
+   * @sa AudioStream.SetFormat
+   */
+  AudioSpec GetInputFormat() const
+  {
+    AudioSpec spec;
+    GetFormat(&spec, nullptr);
+    return spec;
+  }
 
-  AudioSpec GetOutputFormat() const { static_assert(false, "Not implemented"); }
+  /**
+   * Query the current output format of an audio stream.
+   *
+   * @returns the output audio format.
+   * @throws Error on failure.
+   *
+   * @threadsafety It is safe to call this function from any thread, as it holds
+   *               a stream-specific mutex while running.
+   *
+   * @since This function is available since SDL 3.2.0.
+   *
+   * @sa AudioStream.SetFormat
+   */
+  AudioSpec GetOutputFormat() const
+  {
+    AudioSpec spec;
+    GetFormat(nullptr, &spec);
+    return spec;
+  }
 
   /**
    * Query the current format of an audio stream.
@@ -1930,15 +2012,69 @@
    */
   void GetFormat(AudioSpec* src_spec, AudioSpec* dst_spec) const;
 
-  void SetInputFormat(const AudioSpec& spec)
-  {
-    static_assert(false, "Not implemented");
-  }
+  /**
+   * Change the input format of an audio stream.
+   *
+   * Future calls to and AudioStreamRef.GetAvailable and
+   * AudioStreamRef.GetData will reflect the new format, and future calls to
+   * AudioStreamRef.PutData must provide data in the new input formats.
+   *
+   * Data that was previously queued in the stream will still be operated on in
+   * the format that was current when it was added, which is to say you can put
+   * the end of a sound file in one format to a stream, change formats for the
+   * next sound file, and start putting that new data while the previous sound
+   * file is still queued, and everything will still play back correctly.
+   *
+   * If a stream is bound to a device, then the format of the side of the stream
+   * bound to a device cannot be changed (src_spec for recording devices,
+   * dst_spec for playback devices). Attempts to make a change to this side will
+   * be ignored, but this will not report an error. The other side's format can
+   * be changed.
+   *
+   * @param spec the new format of the audio input.
+   * @throws Error on failure.
+   *
+   * @threadsafety It is safe to call this function from any thread, as it holds
+   *               a stream-specific mutex while running.
+   *
+   * @since This function is available since SDL 3.2.0.
+   *
+   * @sa AudioStream.GetFormat
+   * @sa AudioStream.SetFrequencyRatio
+   */
+  void SetInputFormat(const AudioSpec& spec) { SetFormat(spec, std::nullopt); }
 
-  void SetOutputFormat(const AudioSpec& spec)
-  {
-    static_assert(false, "Not implemented");
-  }
+  /**
+   * Change the output format of an audio stream.
+   *
+   * Future calls to and AudioStreamRef.GetAvailable and
+   * AudioStreamRef.GetData will reflect the new format, and future calls to
+   * AudioStreamRef.PutData must provide data in the new input formats.
+   *
+   * Data that was previously queued in the stream will still be operated on in
+   * the format that was current when it was added, which is to say you can put
+   * the end of a sound file in one format to a stream, change formats for the
+   * next sound file, and start putting that new data while the previous sound
+   * file is still queued, and everything will still play back correctly.
+   *
+   * If a stream is bound to a device, then the format of the side of the stream
+   * bound to a device cannot be changed (src_spec for recording devices,
+   * dst_spec for playback devices). Attempts to make a change to this side will
+   * be ignored, but this will not report an error. The other side's format can
+   * be changed.
+   *
+   * @param spec the new format of the audio output.
+   * @throws Error on failure.
+   *
+   * @threadsafety It is safe to call this function from any thread, as it holds
+   *               a stream-specific mutex while running.
+   *
+   * @since This function is available since SDL 3.2.0.
+   *
+   * @sa AudioStream.GetFormat
+   * @sa AudioStream.SetFrequencyRatio
+   */
+  void SetOutputFormat(const AudioSpec& spec) { SetFormat(std::nullopt, spec); }
 
   /**
    * Change the input and output formats of an audio stream.
@@ -2069,11 +2205,8 @@
    * Audio streams default to no remapping applied. This is represented by
    * returning nullptr, and does not signify an error.
    *
-   * @param count On output, set to number of channels in the map. Can be
-   *              nullptr.
    * @returns an array of the current channel mapping, with as many elements as
-   *          the current output spec's channels, or nullptr if default. This
-   *          should be freed with free() when it is no longer needed.
+   *          the current output spec's channels, or nullptr if default.
    *
    * @threadsafety It is safe to call this function from any thread, as it holds
    *               a stream-specific mutex while running.
@@ -2093,11 +2226,8 @@
    * Audio streams default to no remapping applied. This is represented by
    * returning nullptr, and does not signify an error.
    *
-   * @param count On output, set to number of channels in the map. Can be
-   *              nullptr.
    * @returns an array of the current channel mapping, with as many elements as
-   *          the current output spec's channels, or nullptr if default. This
-   *          should be freed with free() when it is no longer needed.
+   *          the current output spec's channels, or nullptr if default.
    *
    * @threadsafety It is safe to call this function from any thread, as it holds
    *               a stream-specific mutex while running.
@@ -2152,7 +2282,6 @@
    * mapping, but previously-added data will still have the prior mapping.
    *
    * @param chmap the new channel map, nullptr to reset to default.
-   * @param count The number of channels in the map.
    * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread, as it holds
@@ -2208,7 +2337,6 @@
    * before feeding it to the device for playback.
    *
    * @param chmap the new channel map, nullptr to reset to default.
-   * @param count The number of channels in the map.
    * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread, as it holds
@@ -2234,7 +2362,6 @@
    * Get call would just dequeue the previously-converted data.
    *
    * @param buf a pointer to the audio data to add.
-   * @param len the number of bytes to write to the stream.
    * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread, but if the
@@ -2277,7 +2404,6 @@
    * AudioStream.PutData(), which allocates and buffers in blocks.
    *
    * @param buf a pointer to the audio data to add.
-   * @param len the number of bytes to add to the stream.
    * @param callback the callback function to call when the data is no longer
    *                 needed by the stream. May be nullptr.
    * @param userdata an opaque pointer provided to the callback for its own
@@ -2299,10 +2425,6 @@
                      AudioStreamDataCompleteCallback callback,
                      void* userdata);
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
   /**
    * Add external data to an audio stream without copying it.
    *
@@ -2328,11 +2450,8 @@
    * AudioStream.PutData(), which allocates and buffers in blocks.
    *
    * @param buf a pointer to the audio data to add.
-   * @param len the number of bytes to add to the stream.
    * @param callback the callback function to call when the data is no longer
    *                 needed by the stream. May be nullptr.
-   * @param userdata an opaque pointer provided to the callback for its own
-   *                 personal use.
    * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread, but if the
@@ -2363,7 +2482,6 @@
    * the stream and requesting the output just copied the converted data.
    *
    * @param buf a buffer to fill with audio data.
-   * @param len the maximum number of bytes to fill.
    * @returns the number of bytes read from the stream or -1 on failure; call
    *          GetError() for more information.
    *
@@ -2657,8 +2775,6 @@
    *
    * @param callback the new callback function to call when data is requested
    *                 from the stream.
-   * @param userdata an opaque pointer provided to the callback for its own
-   *                 personal use.
    * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread.
@@ -2751,8 +2867,6 @@
    *
    * @param callback the new callback function to call when data is added to the
    *                 stream.
-   * @param userdata an opaque pointer provided to the callback for its own
-   *                 personal use.
    * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread.
@@ -2906,15 +3020,7 @@
  * As this is just a wrapper over Mutex.Lock for an internal lock; it has all
  * the same attributes (recursive locks are allowed, etc).
  *
- * @param stream the audio stream to lock.
- * @returns true on success or false on failure; call GetError() for more
- *          information.
- *
- * @threadsafety It is safe to call this function from any thread.
- *
- * @since This function is available since SDL 3.2.0.
- *
- * @sa AudioStream.Unlock
+ * @sa AudioStream.Lock
  */
 class AudioStreamLock
 {
@@ -3020,8 +3126,7 @@
    * This unlocks an audio stream after a call to AudioStream.Lock.
    *
    * @param stream the audio stream to unlock.
-   * @returns true on success or false on failure; call GetError() for more
-   *          information.
+   * @throws Error on failure.
    *
    * @threadsafety You should only call this from the same thread that
    *               previously called AudioStream.Lock.
@@ -3122,11 +3227,8 @@
  * If this function returns nullptr, to signify an error, `*count` will be set
  * to zero.
  *
- * @param count a pointer filled in with the number of devices returned, may be
- *              nullptr.
- * @returns a 0 terminated array of device instance IDs or nullptr on error;
- *          call GetError() for more information. This should be freed with
- *          free() when it is no longer needed.
+ * @returns a 0 terminated array of device instance IDs on success.
+ * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -3137,7 +3239,10 @@
  */
 inline OwnArray<AudioDeviceRef> GetAudioPlaybackDevices()
 {
-  return SDL_GetAudioPlaybackDevices();
+  int count;
+  auto data = CheckError(SDL_GetAudioPlaybackDevices(&count));
+  return OwnArray<AudioDeviceRef>{reinterpret_cast<AudioDeviceRef*>(data),
+                                  size_t(count)};
 }
 
 /**
@@ -3154,11 +3259,8 @@
  * If this function returns nullptr, to signify an error, `*count` will be set
  * to zero.
  *
- * @param count a pointer filled in with the number of devices returned, may be
- *              nullptr.
- * @returns a 0 terminated array of device instance IDs, or nullptr on failure;
- *          call GetError() for more information. This should be freed with
- *          free() when it is no longer needed.
+ * @returns a 0 terminated array of device instance IDs on success.
+ * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -3169,7 +3271,10 @@
  */
 inline OwnArray<AudioDeviceRef> GetAudioRecordingDevices()
 {
-  return SDL_GetAudioRecordingDevices();
+  int count;
+  auto data = CheckError(SDL_GetAudioRecordingDevices(&count));
+  return OwnArray<AudioDeviceRef>{reinterpret_cast<AudioDeviceRef*>(data),
+                                  size_t(count)};
 }
 
 /**
@@ -3185,8 +3290,8 @@
  * not the abstract system default.
  *
  * @param devid the instance ID of the device to query.
- * @returns the name of the audio device, or nullptr on failure; call GetError()
- *          for more information.
+ * @returns the name of the audio device on success.
+ * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -3197,7 +3302,7 @@
  */
 inline const char* GetAudioDeviceName(AudioDeviceParam devid)
 {
-  return SDL_GetAudioDeviceName(devid);
+  return CheckError(SDL_GetAudioDeviceName(devid));
 }
 
 inline const char* AudioDevice::GetName() const
@@ -3227,7 +3332,6 @@
  * playback timing. Most apps do not need this.
  *
  * @param devid the instance ID of the device to query.
- * @param spec on return, will be filled with device details.
  * @param sample_frames pointer to store device buffer size, in sample frames.
  *                      Can be nullptr.
  * @throws Error on failure.
@@ -3239,7 +3343,9 @@
 inline AudioSpec GetAudioDeviceFormat(AudioDeviceParam devid,
                                       int* sample_frames = nullptr)
 {
-  return CheckError(SDL_GetAudioDeviceFormat(devid, sample_frames));
+  AudioSpec spec;
+  CheckError(SDL_GetAudioDeviceFormat(devid, &spec, sample_frames));
+  return spec;
 }
 
 inline AudioSpec AudioDevice::GetFormat(int* sample_frames) const
@@ -3257,10 +3363,8 @@
  * returning nullptr, and does not signify an error.
  *
  * @param devid the instance ID of the device to query.
- * @param count On output, set to number of channels in the map. Can be nullptr.
  * @returns an array of the current channel mapping, with as many elements as
- *          the current output spec's channels, or nullptr if default. This
- *          should be freed with free() when it is no longer needed.
+ *          the current output spec's channels, or nullptr if default
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -3270,7 +3374,9 @@
  */
 inline OwnArray<int> GetAudioDeviceChannelMap(AudioDeviceParam devid)
 {
-  return SDL_GetAudioDeviceChannelMap(devid);
+  int count;
+  auto data = SDL_GetAudioDeviceChannelMap(devid, &count);
+  return OwnArray<int>{data, size_t(count)};
 }
 
 inline OwnArray<int> AudioDevice::GetChannelMap() const
@@ -3349,7 +3455,7 @@
  * @sa AudioDevice.GetFormat
  */
 inline AudioDevice OpenAudioDevice(AudioDeviceParam devid,
-                                   const AudioSpec& spec)
+                                   OptionalRef<const AudioSpec> spec)
 {
   return AudioDevice(devid, spec);
 }
@@ -3630,7 +3736,6 @@
  *
  * @param devid an audio device to bind a stream to.
  * @param streams an array of audio streams to bind.
- * @param num_streams number streams listed in the `streams` array.
  * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
@@ -3644,7 +3749,10 @@
 inline void BindAudioStreams(AudioDeviceParam devid,
                              std::span<AudioStreamRef> streams)
 {
-  CheckError(SDL_BindAudioStreams(devid, streams));
+  CheckError(SDL_BindAudioStreams(
+    devid,
+    reinterpret_cast<SDL_AudioStream* const*>(streams.data()),
+    streams.size()));
 }
 
 inline void AudioDevice::BindAudioStreams(std::span<AudioStreamRef> streams)
@@ -3691,7 +3799,6 @@
  *
  * @param streams an array of audio streams to unbind. Can be nullptr or contain
  *                nullptr.
- * @param num_streams number streams listed in the `streams` array.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -3701,7 +3808,8 @@
  */
 inline void UnbindAudioStreams(std::span<AudioStreamRef> streams)
 {
-  SDL_UnbindAudioStreams(streams);
+  SDL_UnbindAudioStreams(
+    reinterpret_cast<SDL_AudioStream* const*>(streams.data()), streams.size());
 }
 
 /**
@@ -3745,7 +3853,7 @@
  */
 inline AudioDeviceRef GetAudioStreamDevice(AudioStreamParam stream)
 {
-  return SDL_GetAudioStreamDevice(stream);
+  return {SDL_GetAudioStreamDevice(stream)};
 }
 
 inline AudioDeviceRef AudioStream::GetDevice() const
@@ -3773,8 +3881,8 @@
  * @sa AudioStream.SetFormat
  * @sa AudioStream.Destroy
  */
-inline AudioStream CreateAudioStream(const AudioSpec& src_spec,
-                                     const AudioSpec& dst_spec)
+inline AudioStream CreateAudioStream(OptionalRef<const AudioSpec> src_spec,
+                                     OptionalRef<const AudioSpec> dst_spec)
 {
   return AudioStream(src_spec, dst_spec);
 }
@@ -3803,7 +3911,7 @@
  */
 inline PropertiesRef GetAudioStreamProperties(AudioStreamParam stream)
 {
-  return CheckError(SDL_GetAudioStreamProperties(stream));
+  return {CheckError(SDL_GetAudioStreamProperties(stream))};
 }
 
 inline PropertiesRef AudioStream::GetProperties() const
@@ -3811,17 +3919,17 @@
   return SDL::GetAudioStreamProperties(m_resource);
 }
 
-namespace prop::AudioStream {
-
 #if SDL_VERSION_ATLEAST(3, 4, 0)
 
+namespace prop::AudioStream {
+
 constexpr auto _AUTO_CLEANUP_BOOLEAN =
   SDL_PROP_AUDIOSTREAM_AUTO_CLEANUP_BOOLEAN;
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
 } // namespace prop::AudioStream
 
+#endif // SDL_VERSION_ATLEAST(3, 4, 0)
+
 /**
  * Query the current format of an audio stream.
  *
@@ -4027,10 +4135,8 @@
  * returning nullptr, and does not signify an error.
  *
  * @param stream the AudioStream to query.
- * @param count On output, set to number of channels in the map. Can be nullptr.
  * @returns an array of the current channel mapping, with as many elements as
- *          the current output spec's channels, or nullptr if default. This
- *          should be freed with free() when it is no longer needed.
+ *          the current output spec's channels, or nullptr if default.
  *
  * @threadsafety It is safe to call this function from any thread, as it holds a
  *               stream-specific mutex while running.
@@ -4041,7 +4147,10 @@
  */
 inline OwnArray<int> GetAudioStreamInputChannelMap(AudioStreamParam stream)
 {
-  return SDL_GetAudioStreamInputChannelMap(stream);
+  int count;
+  auto data = SDL_GetAudioStreamInputChannelMap(stream, &count);
+  if (!data) return {};
+  return OwnArray<int>{data, size_t(count)};
 }
 
 inline OwnArray<int> AudioStream::GetInputChannelMap() const
@@ -4059,10 +4168,8 @@
  * returning nullptr, and does not signify an error.
  *
  * @param stream the AudioStream to query.
- * @param count On output, set to number of channels in the map. Can be nullptr.
  * @returns an array of the current channel mapping, with as many elements as
- *          the current output spec's channels, or nullptr if default. This
- *          should be freed with free() when it is no longer needed.
+ *          the current output spec's channels, or nullptr if default.
  *
  * @threadsafety It is safe to call this function from any thread, as it holds a
  *               stream-specific mutex while running.
@@ -4073,7 +4180,10 @@
  */
 inline OwnArray<int> GetAudioStreamOutputChannelMap(AudioStreamParam stream)
 {
-  return SDL_GetAudioStreamOutputChannelMap(stream);
+  int count;
+  auto data = SDL_GetAudioStreamOutputChannelMap(stream, &count);
+  if (!data) return {};
+  return OwnArray<int>{data, size_t(count)};
 }
 
 inline OwnArray<int> AudioStream::GetOutputChannelMap() const
@@ -4126,7 +4236,6 @@
  *
  * @param stream the AudioStream to change.
  * @param chmap the new channel map, nullptr to reset to default.
- * @param count The number of channels in the map.
  * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread, as it holds a
@@ -4141,7 +4250,8 @@
 inline void SetAudioStreamInputChannelMap(AudioStreamParam stream,
                                           std::span<int> chmap)
 {
-  CheckError(SDL_SetAudioStreamInputChannelMap(stream, chmap));
+  CheckError(
+    SDL_SetAudioStreamInputChannelMap(stream, chmap.data(), chmap.size()));
 }
 
 inline void AudioStream::SetInputChannelMap(std::span<int> chmap)
@@ -4192,7 +4302,6 @@
  *
  * @param stream the AudioStream to change.
  * @param chmap the new channel map, nullptr to reset to default.
- * @param count The number of channels in the map.
  * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread, as it holds a
@@ -4207,7 +4316,8 @@
 inline void SetAudioStreamOutputChannelMap(AudioStreamParam stream,
                                            std::span<int> chmap)
 {
-  CheckError(SDL_SetAudioStreamOutputChannelMap(stream, chmap));
+  CheckError(
+    SDL_SetAudioStreamOutputChannelMap(stream, chmap.data(), chmap.size()));
 }
 
 inline void AudioStream::SetOutputChannelMap(std::span<int> chmap)
@@ -4228,7 +4338,6 @@
  *
  * @param stream the stream the audio data is being added to.
  * @param buf a pointer to the audio data to add.
- * @param len the number of bytes to write to the stream.
  * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread, but if the
@@ -4280,7 +4389,6 @@
  *
  * @param stream the stream the audio data is being added to.
  * @param buf a pointer to the audio data to add.
- * @param len the number of bytes to add to the stream.
  * @param callback the callback function to call when the data is no longer
  *                 needed by the stream. May be nullptr.
  * @param userdata an opaque pointer provided to the callback for its own
@@ -4307,10 +4415,6 @@
     stream, buf.data(), buf.size_bytes(), callback, userdata));
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Add external data to an audio stream without copying it.
  *
@@ -4337,11 +4441,8 @@
  *
  * @param stream the stream the audio data is being added to.
  * @param buf a pointer to the audio data to add.
- * @param len the number of bytes to add to the stream.
  * @param callback the callback function to call when the data is no longer
  *                 needed by the stream. May be nullptr.
- * @param userdata an opaque pointer provided to the callback for its own
- *                 personal use.
  * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread, but if the
@@ -4359,13 +4460,13 @@
                                      SourceBytes buf,
                                      AudioStreamDataCompleteCB callback)
 {
-  static_assert(false, "Not implemented");
+  using Wrapper = CallbackWrapper<AudioStreamDataCompleteCB>;
+  PutAudioStreamDataNoCopy(stream,
+                           std::move(buf),
+                           &Wrapper::CallOnce,
+                           Wrapper::Wrap(std::move(callback)));
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 inline void AudioStream::PutDataNoCopy(SourceBytes buf,
                                        AudioStreamDataCompleteCallback callback,
                                        void* userdata)
@@ -4373,20 +4474,12 @@
   SDL::PutAudioStreamDataNoCopy(m_resource, std::move(buf), callback, userdata);
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 inline void AudioStream::PutDataNoCopy(SourceBytes buf,
                                        AudioStreamDataCompleteCB callback)
 {
   SDL::PutAudioStreamDataNoCopy(m_resource, std::move(buf), callback);
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Add data to the stream with each channel in a separate array.
  *
@@ -4444,10 +4537,6 @@
     stream, channel_buffers, num_channels, num_samples));
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 inline void AudioStream::PutPlanarData(const void* const* channel_buffers,
                                        int num_channels,
                                        int num_samples)
@@ -4472,7 +4561,6 @@
  *
  * @param stream the stream the audio is being requested from.
  * @param buf a buffer to fill with audio data.
- * @param len the maximum number of bytes to fill.
  * @returns the number of bytes read from the stream or -1 on failure; call
  *          GetError() for more information.
  *
@@ -4861,8 +4949,6 @@
  * @param stream the audio stream to set the new callback on.
  * @param callback the new callback function to call when data is requested from
  *                 the stream.
- * @param userdata an opaque pointer provided to the callback for its own
- *                 personal use.
  * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
@@ -4874,7 +4960,7 @@
 inline void SetAudioStreamGetCallback(AudioStreamParam stream,
                                       AudioStreamCB callback)
 {
-  static_assert(false, "Not implemented");
+  SetAudioStreamGetCallback(stream, callback.wrapper, callback.data);
 }
 
 inline void AudioStream::SetGetCallback(AudioStreamCallback callback,
@@ -4975,8 +5061,6 @@
  * @param stream the audio stream to set the new callback on.
  * @param callback the new callback function to call when data is added to the
  *                 stream.
- * @param userdata an opaque pointer provided to the callback for its own
- *                 personal use.
  * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
@@ -4988,7 +5072,7 @@
 inline void SetAudioStreamPutCallback(AudioStreamParam stream,
                                       AudioStreamCB callback)
 {
-  static_assert(false, "Not implemented");
+  SetAudioStreamPutCallback(stream, callback.wrapper, callback.data);
 }
 
 inline void AudioStream::SetPutCallback(AudioStreamCallback callback,
@@ -5085,9 +5169,9 @@
  * @sa AudioStream.ResumeDevice
  */
 inline AudioStream OpenAudioDeviceStream(AudioDeviceParam devid,
-                                         const AudioSpec& spec,
-                                         AudioStreamCallback callback,
-                                         void* userdata)
+                                         OptionalRef<const AudioSpec> spec,
+                                         AudioStreamCallback callback = nullptr,
+                                         void* userdata = nullptr)
 {
   return AudioStream(devid, spec, callback, userdata);
 }
@@ -5134,11 +5218,7 @@
  *              AUDIO_DEVICE_DEFAULT_RECORDING.
  * @param spec the audio stream's data format. Can be nullptr.
  * @param callback a callback where the app will provide new data for playback,
- *                 or receive new data for recording. Can be nullptr, in which
- *                 case the app will need to call AudioStream.PutData or
- *                 AudioStream.GetData as necessary.
- * @param userdata app-controlled pointer passed to callback. Can be nullptr.
- *                 Ignored if callback is nullptr.
+ *                 or receive new data for recording.
  * @returns an audio stream on success.
  * @throws Error on failure.
  *
@@ -5156,9 +5236,9 @@
   return AudioStream(devid, spec, callback);
 }
 
-inline AudioStreamRef AudioDevice::OpenStream(const AudioSpec& spec,
-                                              AudioStreamCallback callback,
-                                              void* userdata)
+inline AudioStream AudioDevice::OpenStream(OptionalRef<const AudioSpec> spec,
+                                           AudioStreamCallback callback,
+                                           void* userdata)
 {
   return AudioStream(m_resource, spec, callback, userdata);
 }
@@ -5166,13 +5246,19 @@
 inline AudioStream AudioDevice::OpenStream(OptionalRef<const AudioSpec> spec,
                                            AudioStreamCB callback)
 {
-  return SDL::OpenAudioDeviceStream(m_resource, spec, callback);
+  return AudioStream(m_resource, spec, callback);
 }
 
 inline AudioStream::AudioStream(AudioDeviceParam devid,
                                 OptionalRef<const AudioSpec> spec,
                                 AudioStreamCB callback)
+  : AudioStream(devid, spec)
 {
+  if (IsAudioDevicePlayback(devid)) {
+    SetGetCallback(std::move(callback));
+  } else {
+    SetPutCallback(std::move(callback));
+  }
 }
 
 /**
@@ -5277,7 +5363,6 @@
  *
  * @param devid the ID of an opened audio device.
  * @param callback a callback function to be called. Can be nullptr.
- * @param userdata app-controlled pointer passed to callback. Can be nullptr.
  * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
@@ -5287,7 +5372,7 @@
 inline void SetAudioPostmixCallback(AudioDeviceParam devid,
                                     AudioPostmixCB callback)
 {
-  static_assert(false, "Not implemented");
+  SetAudioPostmixCallback(devid, callback.wrapper, callback.data);
 }
 
 inline void AudioDevice::SetPostmixCallback(AudioPostmixCallback callback,
@@ -5340,46 +5425,41 @@
  *
  * Example:
  *
- * ```c
- * LoadWAV(IOStream.FromFile("sample.wav", "rb"), true, &spec, &buf, &len);
+ * ```cpp
+ * LoadWAV(IOStream.FromFile("sample.wav", "rb"), spec);
  * ```
  *
  * Note that the LoadWAV function does this same thing for you, but in a less
  * messy way:
  *
- * ```c
- * LoadWAV("sample.wav", &spec, &buf, &len);
+ * ```cpp
+ * LoadWAV("sample.wav", &spec);
  * ```
  *
  * @param src the data source for the WAVE data.
- * @param closeio if true, calls IOStream.Close() on `src` before returning,
- *                even in the case of an error.
  * @param spec a pointer to an AudioSpec that will be set to the WAVE data's
  *             format details on successful return.
- * @param audio_buf a pointer filled with the audio data, allocated by the
- *                  function.
- * @param audio_len a pointer filled with the length of the audio data buffer in
- *                  bytes.
+ * @param closeio if true, calls IOStream.Close() on `src` before returning,
+ *                even in the case of an error.
  * @throws Error on failure.
  *
- * This function returns false if the .WAV file cannot be opened, uses an
- * unknown data format, or is corrupt; call GetError() for more information.
- *
- * When the application is done with the data returned in `audio_buf`, it should
- * call free() to dispose of it.
+ * This function throws if the .WAV file cannot be opened, uses an unknown data
+ * format, or is corrupt; call GetError() for more information.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
  * @since This function is available since SDL 3.2.0.
  *
- * @sa free
  * @sa LoadWAV
  */
 inline OwnArray<Uint8> LoadWAV(IOStreamParam src,
                                AudioSpec* spec,
                                bool closeio = false)
 {
-  return CheckError(SDL_LoadWAV_IO(src, spec, closeio));
+  Uint8* buf;
+  Uint32 len;
+  if (!SDL_LoadWAV_IO(src, closeio, spec, &buf, &len)) return {};
+  return OwnArray<Uint8>{buf, size_t(len)};
 }
 
 /**
@@ -5394,28 +5474,23 @@
  * @param path the file path of the WAV file to open.
  * @param spec a pointer to an AudioSpec that will be set to the WAVE data's
  *             format details on successful return.
- * @param audio_buf a pointer filled with the audio data, allocated by the
- *                  function.
- * @param audio_len a pointer filled with the length of the audio data buffer in
- *                  bytes.
  * @throws Error on failure.
  *
- * This function returns false if the .WAV file cannot be opened, uses an
- * unknown data format, or is corrupt; call GetError() for more information.
- *
- * When the application is done with the data returned in `audio_buf`, it should
- * call free() to dispose of it.
+ * This function throws if the .WAV file cannot be opened, uses an unknown data
+ * format, or is corrupt,
  *
  * @threadsafety It is safe to call this function from any thread.
  *
  * @since This function is available since SDL 3.2.0.
  *
- * @sa free
  * @sa LoadWAV
  */
 inline OwnArray<Uint8> LoadWAV(StringParam path, AudioSpec* spec)
 {
-  return CheckError(SDL_LoadWAV(path, spec));
+  Uint8* buf;
+  Uint32 len;
+  if (!SDL_LoadWAV(path, spec, &buf, &len)) return {};
+  return OwnArray<Uint8>{buf, size_t(len)};
 }
 
 /**
@@ -5441,7 +5516,6 @@
  * @param src the source audio buffer to be mixed.
  * @param format the AudioFormat structure representing the desired audio
  *               format.
- * @param len the length of the audio buffer in bytes.
  * @param volume ranges from 0.0 - 1.0, and should be set to 1.0 for full audio
  *               volume.
  * @throws Error on failure.
@@ -5455,7 +5529,8 @@
                      AudioFormat format,
                      float volume)
 {
-  CheckError(SDL_MixAudio(dst, src.data(), src.size_bytes(), format, volume));
+  CheckError(
+    SDL_MixAudio(dst, src.data_as<Uint8>(), format, src.size_bytes(), volume));
 }
 
 /**
@@ -5481,7 +5556,6 @@
  * @param src the source audio buffer to be mixed.
  * @param format the AudioFormat structure representing the desired audio
  *               format.
- * @param len the length of the audio buffer in bytes.
  * @param volume ranges from 0.0 - 1.0, and should be set to 1.0 for full audio
  *               volume.
  * @throws Error on failure.
@@ -5495,7 +5569,13 @@
                      AudioFormat format,
                      float volume)
 {
-  static_assert(false, "Not implemented");
+  if (dst.size_bytes() < src.size_bytes()) {
+    MixAudio(dst.data_as<Uint8>(),
+             SourceBytes{src.data(), dst.size_bytes()},
+             format,
+             volume);
+  } else
+    MixAudio(dst.data_as<Uint8>(), src, format, volume);
 }
 
 /**
@@ -5513,11 +5593,7 @@
  *
  * @param src_spec the format details of the input audio.
  * @param src_data the audio data to be converted.
- * @param src_len the len of src_data.
  * @param dst_spec the format details of the output audio.
- * @param dst_data will be filled with a pointer to converted audio data, which
- *                 should be freed with free(). On error, it will be nullptr.
- * @param dst_len will be filled with the len of dst_data.
  * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
@@ -5528,8 +5604,15 @@
                                            SourceBytes src_data,
                                            const AudioSpec& dst_spec)
 {
-  return CheckError(SDL_ConvertAudioSamples(
-    &src_spec, src_data.data(), src_data.size_bytes(), &dst_spec));
+  Uint8* buf;
+  int len;
+  CheckError(SDL_ConvertAudioSamples(&src_spec,
+                                     src_data.data_as<Uint8>(),
+                                     src_data.size_bytes(),
+                                     &dst_spec,
+                                     &buf,
+                                     &len));
+  return OwnArray<Uint8>{buf, size_t(len)};
 }
 
 /**
