--- src/generated/SDL3pp_tray.h
+++ include/SDL3pp/SDL3pp_tray.h
@@ -8,7 +8,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryTray Category Tray
+ * @defgroup CategoryTray System Tray
  *
  * SDL offers a way to add items to the "system tray" (more correctly called the
  * "notification area" on Windows). On platforms that offer this concept, an SDL
@@ -55,6 +55,9 @@
 };
 
 /// Alias to raw representation for TrayMenu.
+struct TrayMenu;
+
+/// Alias to raw representation for TrayMenu.
 using TrayMenuRaw = SDL_TrayMenu*;
 
 // Forward decl
@@ -418,8 +421,6 @@
   /**
    * Returns a list of entries in the menu, in order.
    *
-   * @param count An optional pointer to obtain the number of entries in the
-   *              menu.
    * @returns a nullptr-terminated list of entries within the given menu. The
    *          pointer becomes invalid when any function that inserts or deletes
    *          entries in the menu is called.
@@ -461,6 +462,30 @@
    */
   TrayEntry InsertEntry(int pos, StringParam label, TrayEntryFlags flags);
 
+  /**
+   * Appends a tray entry.
+   *
+   * If label is nullptr, the entry will be a separator. Many functions won't
+   * work for an entry that is a separator.
+   *
+   * An entry does not need to be destroyed; it will be destroyed with the tray.
+   *
+   * @param label the text to be displayed on the entry, in UTF-8 encoding, or
+   *              nullptr for a separator.
+   * @param flags a combination of flags, some of which are mandatory.
+   * @returns the newly created entry, or nullptr if pos is out of bounds.
+   *
+   * @threadsafety This function should be called on the thread that created the
+   *               tray.
+   *
+   * @since This function is available since SDL 3.2.0.
+   *
+   * @sa TrayMenu.InsertEntry
+   * @sa TrayEntryFlags
+   * @sa TrayMenu.GetEntries
+   * @sa TrayEntry.Remove
+   * @sa TrayEntry.GetParent
+   */
   TrayEntry AppendEntry(StringParam label, TrayEntryFlags flags);
 
   /**
@@ -744,6 +769,19 @@
    */
   bool GetEnabled() const;
 
+  /**
+   * Sets a callback to be invoked when the entry is selected.
+   *
+   * @param callback a callback to be invoked when the entry is selected.
+   *
+   * @threadsafety This function should be called on the thread that created the
+   *               tray.
+   *
+   * @since This function is available since SDL 3.2.0.
+   *
+   * @sa TrayMenu.GetEntries
+   * @sa TrayMenu.InsertEntry
+   */
   void SetCallback(TrayCB callback);
 
   /**
@@ -1002,7 +1040,6 @@
  * Returns a list of entries in the menu, in order.
  *
  * @param menu The menu to get entries from.
- * @param count An optional pointer to obtain the number of entries in the menu.
  * @returns a nullptr-terminated list of entries within the given menu. The
  *          pointer becomes invalid when any function that inserts or deletes
  *          entries in the menu is called.
@@ -1017,7 +1054,10 @@
  */
 inline std::span<TrayEntry> GetTrayEntries(TrayMenu menu)
 {
-  return SDL_GetTrayEntries(menu);
+  int count;
+  auto entries = SDL_GetTrayEntries(menu, &count);
+  return std::span<TrayEntry>{reinterpret_cast<TrayEntry*>(entries),
+                              size_t(count)};
 }
 
 inline std::span<TrayEntry> TrayMenu::GetEntries()
@@ -1397,6 +1437,16 @@
 
 /// @}
 
+inline TrayEntry TrayMenu::AppendEntry(StringParam label, TrayEntryFlags flags)
+{
+  return InsertEntry(-1, std::move(label), flags);
+}
+
+inline void TrayEntry::SetCallback(TrayCB callback)
+{
+  SetCallback(callback.wrapper, callback.data);
+}
+
 } // namespace SDL
 
 #endif /* SDL3PP_TRAY_H_ */
