--- src/generated/SDL3pp_video.h
+++ include/SDL3pp/SDL3pp_video.h
@@ -9,7 +9,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryVideo Category Video
+ * @defgroup CategoryVideo Display and Window Management
  *
  * SDL's video subsystem is largely interested in abstracting window management
  * from the underlying operating system. You can create windows, manage them in
@@ -80,7 +80,7 @@
 struct GLContext;
 
 /// Alias to raw representation for GLContext.
-using GLContextRaw = SDL_GLContext*;
+using GLContextRaw = SDL_GLContext;
 
 // Forward decl
 struct GLContextScoped;
@@ -285,7 +285,7 @@
    * The primary display is often located at (0,0), but may be placed at a
    * different location depending on monitor layout.
    *
-   * @param rect the Rect structure filled in with the display bounds.
+   * @returns the Rect structure filled in with the display bounds.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -309,7 +309,7 @@
    * so these are good guidelines for the maximum space available to a
    * non-fullscreen window.
    *
-   * @param rect the Rect structure filled in with the display bounds.
+   * @returns the Rect structure filled in with the display bounds.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -387,12 +387,8 @@
    * - refresh rate -> highest to lowest
    * - pixel density -> lowest to highest
    *
-   * @param count a pointer filled in with the number of display modes returned,
-   *              may be nullptr.
    * @returns a nullptr terminated array of display mode pointers or nullptr on
-   *          failure; call GetError() for more information. This is a single
-   *          allocation that should be freed with free() when it is no longer
-   *          needed.
+   *          failure; call GetError() for more information.
    *
    * @threadsafety This function should only be called on the main thread.
    *
@@ -412,14 +408,12 @@
    * and finally checking the refresh rate. If all the available modes are too
    * small, then false is returned.
    *
-   * @param w the width in pixels of the desired display mode.
-   * @param h the height in pixels of the desired display mode.
+   * @param size the width and height in pixels of the desired display mode.
    * @param refresh_rate the refresh rate of the desired display mode, or 0.0f
    *                     for the desktop refresh rate.
    * @param include_high_density_modes boolean to include high density modes in
    *                                   the search.
-   * @param closest a pointer filled in with the closest display mode equal to
-   *                or larger than the desired mode.
+   * @returns the closest display mode equal to or larger than the desired mode.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -650,6 +644,7 @@
  * @since This enum is available since SDL 3.2.0.
  *
  * @sa HitTest
+ * @sa HitTestCB
  */
 using HitTestResult = SDL_HitTestResult;
 
@@ -701,7 +696,18 @@
                                             const PointRaw* area,
                                             void* data);
 
-/// @sa HitTest
+/**
+ * Callback used for hit-testing.
+ *
+ * @param win the WindowRef where hit-testing was set on.
+ * @param area a Point const reference which should be hit-tested.
+ * @returns an SDL::HitTestResult value.
+ *
+ * @cat listener-callback
+ *
+ * @sa HitTest
+ * @sa Window.SetHitTest
+ */
 using HitTestCB =
   std::function<HitTestResult(WindowRaw window, const Point& area)>;
 
@@ -717,9 +723,9 @@
  *
  * @since This struct is available since SDL 3.2.0.
  *
- * @sa Window.Window
- *
  * @cat resource
+ *
+ * @sa Window.Window
  */
 class Window
 {
@@ -826,8 +832,7 @@
    * in a future version of SDL.
    *
    * @param title the title of the window, in UTF-8 encoding.
-   * @param w the width of the window.
-   * @param h the height of the window.
+   * @param size the width and height of the window.
    * @param flags 0, or one or more WindowFlags OR'd together.
    * @post the window that was created or nullptr on failure; call GetError()
    *       for more information.
@@ -842,7 +847,7 @@
    * @sa Window.Destroy
    */
   Window(StringParam title, const PointRaw& size, WindowFlags flags = 0)
-    : m_resource(SDL_CreateWindow(title, size, flags))
+    : m_resource(SDL_CreateWindow(title, size.x, size.y, flags))
   {
   }
 
@@ -900,12 +905,9 @@
    * hidden will be restored when the parent is shown.
    *
    * @param parent the parent of the window, must not be nullptr.
-   * @param offset_x the x position of the popup window relative to the origin
-   *                 of the parent.
-   * @param offset_y the y position of the popup window relative to the origin
-   *                 of the parent window.
-   * @param w the width of the window.
-   * @param h the height of the window.
+   * @param offset the x, y position of the popup window relative to the origin
+   *               of the parent.
+   * @param size the width and height of the window.
    * @param flags WINDOW_TOOLTIP or WINDOW_POPUP_MENU, and zero or more
    *              additional WindowFlags OR'd together.
    * @post the window that was created or nullptr on failure; call GetError()
@@ -924,7 +926,12 @@
          const PointRaw& offset,
          const PointRaw& size,
          WindowFlags flags = 0)
-    : m_resource(SDL_CreatePopupWindow(parent, offset, size, flags))
+    : m_resource(SDL_CreatePopupWindow(parent,
+                                       offset.x,
+                                       offset.y,
+                                       size.x,
+                                       size.y,
+                                       flags))
   {
   }
 
@@ -1250,7 +1257,6 @@
   /**
    * Get the raw ICC profile data for the screen the window is currently on.
    *
-   * @param size the size of the ICC profile.
    * @returns the raw ICC profile data on success.
    * @throws Error on failure.
    *
@@ -1518,9 +1524,7 @@
    * Additionally, as this is just a request, it can be denied by the windowing
    * system.
    *
-   * @param x the x coordinate of the window, or `WINDOWPOS_CENTERED` or
-   *          `WINDOWPOS_UNDEFINED`.
-   * @param y the y coordinate of the window, or `WINDOWPOS_CENTERED` or
+   * @param p the coordinates of the window, or `WINDOWPOS_CENTERED` or
    *          `WINDOWPOS_UNDEFINED`.
    * @throws Error on failure.
    *
@@ -1565,10 +1569,7 @@
    * If you do not need the value for one of the positions a nullptr may be
    * passed in the `x` or `y` parameter.
    *
-   * @param x a pointer filled in with the x position of the window, may be
-   *          nullptr.
-   * @param y a pointer filled in with the y position of the window, may be
-   *          nullptr.
+   * @returns the position on success.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1576,6 +1577,7 @@
    * @since This function is available since SDL 3.2.0.
    *
    * @sa Window.SetPosition
+   * @sa SetPosition(int *, int *)
    */
   Point GetPosition() const;
 
@@ -1600,8 +1602,7 @@
    * to remain within the usable desktop bounds). Additionally, as this is just
    * a request, it can be denied by the windowing system.
    *
-   * @param w the width of the window, must be > 0.
-   * @param h the height of the window, must be > 0.
+   * @param size the width and height of the window, must be > 0.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1642,8 +1643,7 @@
    * window is on a high pixel density display. Use Window.GetSizeInPixels() or
    * Renderer.GetOutputSize() to get the real client area size in pixels.
    *
-   * @param w a pointer filled in with the width of the window, may be nullptr.
-   * @param h a pointer filled in with the height of the window, may be nullptr.
+   * @returns a point with width and height on success
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1653,11 +1653,36 @@
    * @sa Renderer.GetOutputSize
    * @sa Window.GetSizeInPixels
    * @sa Window.SetSize
+   * @sa GetSize(int *, int *)
    */
   Point GetSize() const;
 
+  /**
+   * @brief Request the window's position and size to be set.
+   *
+   * @param rect the rect containing the new coordinates
+   * @throws Error on failure.
+   *
+   * @threadsafety This function should only be called on the main thread.
+   *
+   * @sa WindowRef.SetPosition()
+   * @sa WindowRef.SetSize()
+   */
   void SetRect(Rect rect);
 
+  /**
+   * Get the position and client size of a window.
+   *
+   * This is the current position of the window as last reported by the
+   * windowing system.
+   *
+   * The window pixel size may differ from its window coordinate size if the
+   * window is on a high pixel density display. Use Window.GetSizeInPixels()
+   * or RendererRef.GetOutputSize() to get the real client area size in pixels.
+   *
+   * @return Rect with the position and size
+   * @throws Error on failure.
+   */
   Rect GetRect() const;
 
   /**
@@ -1670,8 +1695,6 @@
    * into the rest of the window, but it should not contain visually important
    * or interactible content.
    *
-   * @param rect a pointer filled in with the client area that is safe for
-   *             interactive content.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1784,16 +1807,15 @@
    *
    * @sa Window.Window
    * @sa Window.GetSize
+   * @sa GetSizeInPixels()
    */
   void GetSizeInPixels(int* w, int* h) const;
 
   /**
    * Get the size of a window's client area, in pixels.
    *
-   * @param w a pointer to variable for storing the width in pixels, may be
-   *          nullptr.
-   * @param h a pointer to variable for storing the height in pixels, may be
-   *          nullptr.
+   * @returns the size on success or std::nullopt on failure; call GetError()
+   * for more information.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1802,14 +1824,14 @@
    *
    * @sa Window.Window
    * @sa Window.GetSize
+   * @sa GetSizeInPixels(int*, int*)
    */
   Point GetSizeInPixels() const;
 
   /**
    * Set the minimum size of a window's client area.
    *
-   * @param min_w the minimum width of the window, or 0 for no limit.
-   * @param min_h the minimum height of the window, or 0 for no limit.
+   * @param p the minimum width and heigh of the window, or 0 for no limit.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -1842,8 +1864,7 @@
   /**
    * Set the maximum size of a window's client area.
    *
-   * @param max_w the maximum width of the window, or 0 for no limit.
-   * @param max_h the maximum height of the window, or 0 for no limit.
+   * @param p the maximum width and height of the window, or 0 for no limit.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -2190,8 +2211,8 @@
   /**
    * Get VSync for the window surface.
    *
-   * @param vsync an int filled with the current vertical refresh sync interval.
-   *              See Window.SetSurfaceVSync() for the meaning of the value.
+   * @returns the current vertical refresh sync interval. See
+   *          Window.SetSurfaceVSync() for the meaning of the value.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -2236,7 +2257,6 @@
    *
    * @param rects an array of Rect structures representing areas of the surface
    *              to copy, in pixels.
-   * @param numrects the number of rectangles.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -2486,10 +2506,8 @@
    * On platforms or desktops where this is unsupported, this function does
    * nothing.
    *
-   * @param x the x coordinate of the menu, relative to the origin (top-left) of
-   *          the client area.
-   * @param y the y coordinate of the menu, relative to the origin (top-left) of
-   *          the client area.
+   * @param p the x, y coordinates of the menu, relative to the origin
+   *          (top-left) of the client area.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -2573,12 +2591,13 @@
    * devoid of allocations, etc.
    *
    * @param callback the function to call when doing a hit-test.
-   * @param callback_data an app-defined void pointer passed to **callback**.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
    *
    * @since This function is available since SDL 3.2.0.
+   *
+   * @cat listener-callback
    */
   void SetHitTest(HitTestCB callback);
 
@@ -2871,8 +2890,7 @@
    * Note that this function will appear to succeed, but not actually move the
    * mouse when used over Microsoft Remote Desktop.
    *
-   * @param x the x coordinate within the window.
-   * @param y the y coordinate within the window.
+   * @param p the x, y coordinates within the window.
    *
    * @threadsafety This function should only be called on the main thread.
    *
@@ -3162,9 +3180,6 @@
   /**
    * Delete an OpenGL context.
    *
-   * @returns true on success or false on failure; call GetError() for more
-   *          information.
-   *
    * @threadsafety This function should only be called on the main thread.
    *
    * @since This function is available since SDL 3.2.0.
@@ -3179,7 +3194,6 @@
    * The context must have been created with a compatible window.
    *
    * @param window the window to associate with the context.
-   * @param context the OpenGL context to associate with the window.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called on the main thread.
@@ -3577,17 +3591,19 @@
 /**
  * Get a list of currently connected displays.
  *
- * @param count a pointer filled in with the number of displays returned, may be
- *              nullptr.
  * @returns a 0 terminated array of display instance IDs or nullptr on failure;
- *          call GetError() for more information. This should be freed with
- *          free() when it is no longer needed.
+ *          call GetError() for more information.
  *
  * @threadsafety This function should only be called on the main thread.
  *
  * @since This function is available since SDL 3.2.0.
  */
-inline OwnArray<DisplayID> GetDisplays() { return SDL_GetDisplays(); }
+inline OwnArray<DisplayID> GetDisplays()
+{
+  int count = 0;
+  auto data = reinterpret_cast<DisplayID*>(SDL_GetDisplays(&count));
+  return OwnArray<DisplayID>{data, size_t(count)};
+}
 
 /**
  * Return the primary display.
@@ -3637,7 +3653,7 @@
  */
 inline PropertiesRef GetDisplayProperties(DisplayID displayID)
 {
-  return CheckError(SDL_GetDisplayProperties(displayID));
+  return {CheckError(SDL_GetDisplayProperties(displayID))};
 }
 
 inline PropertiesRef Display::GetProperties() const
@@ -3684,7 +3700,7 @@
  * different location depending on monitor layout.
  *
  * @param displayID the instance ID of the display to query.
- * @param rect the Rect structure filled in with the display bounds.
+ * @returns the Rect structure filled in with the display bounds.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -3696,7 +3712,9 @@
  */
 inline Rect GetDisplayBounds(Display displayID)
 {
-  return CheckError(SDL_GetDisplayBounds(displayID));
+  Rect bounds;
+  SDL_GetDisplayBounds(displayID, &bounds);
+  return bounds;
 }
 
 inline Rect Display::GetBounds() const
@@ -3716,7 +3734,7 @@
  * window.
  *
  * @param displayID the instance ID of the display to query.
- * @param rect the Rect structure filled in with the display bounds.
+ * @returns the Rect structure filled in with the display bounds.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -3728,7 +3746,9 @@
  */
 inline Rect GetDisplayUsableBounds(Display displayID)
 {
-  return CheckError(SDL_GetDisplayUsableBounds(displayID));
+  Rect bounds;
+  SDL_GetDisplayUsableBounds(displayID, &bounds);
+  return bounds;
 }
 
 inline Rect Display::GetUsableBounds() const
@@ -3830,12 +3850,8 @@
  * - pixel density -> lowest to highest
  *
  * @param displayID the instance ID of the display to query.
- * @param count a pointer filled in with the number of display modes returned,
- *              may be nullptr.
  * @returns a nullptr terminated array of display mode pointers or nullptr on
- *          failure; call GetError() for more information. This is a single
- *          allocation that should be freed with free() when it is no longer
- *          needed.
+ *          failure; call GetError() for more information.
  *
  * @threadsafety This function should only be called on the main thread.
  *
@@ -3845,7 +3861,9 @@
  */
 inline OwnArray<DisplayMode*> GetFullscreenDisplayModes(Display displayID)
 {
-  return SDL_GetFullscreenDisplayModes(displayID);
+  int count = 0;
+  auto data = CheckError(SDL_GetFullscreenDisplayModes(displayID, &count));
+  return OwnArray<DisplayMode*>{data, size_t(count)};
 }
 
 inline OwnArray<DisplayMode*> Display::GetFullscreenModes() const
@@ -3864,14 +3882,12 @@
  * then false is returned.
  *
  * @param displayID the instance ID of the display to query.
- * @param w the width in pixels of the desired display mode.
- * @param h the height in pixels of the desired display mode.
+ * @param size the width and height in pixels of the desired display mode.
  * @param refresh_rate the refresh rate of the desired display mode, or 0.0f for
  *                     the desktop refresh rate.
  * @param include_high_density_modes boolean to include high density modes in
  *                                   the search.
- * @param closest a pointer filled in with the closest display mode equal to or
- *                larger than the desired mode.
+ * @returns the closest display mode equal to or larger than the desired mode.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -3887,8 +3903,14 @@
   float refresh_rate,
   bool include_high_density_modes)
 {
-  return CheckError(SDL_GetClosestFullscreenDisplayMode(
-    displayID, size, refresh_rate, include_high_density_modes));
+  DisplayMode mode;
+  CheckError(SDL_GetClosestFullscreenDisplayMode(displayID,
+                                                 size.x,
+                                                 size.y,
+                                                 refresh_rate,
+                                                 include_high_density_modes,
+                                                 &mode));
+  return mode;
 }
 
 inline DisplayMode Display::GetClosestFullscreenMode(
@@ -3921,7 +3943,7 @@
  */
 inline const DisplayMode& GetDesktopDisplayMode(DisplayID displayID)
 {
-  return SDL_GetDesktopDisplayMode(displayID);
+  return *SDL_GetDesktopDisplayMode(displayID);
 }
 
 inline const DisplayMode& Display::GetDesktopMode() const
@@ -3950,7 +3972,7 @@
  */
 inline const DisplayMode& GetCurrentDisplayMode(DisplayID displayID)
 {
-  return SDL_GetCurrentDisplayMode(displayID);
+  return *SDL_GetCurrentDisplayMode(displayID);
 }
 
 inline const DisplayMode& Display::GetCurrentMode() const
@@ -3974,7 +3996,7 @@
  */
 inline Display GetDisplayForPoint(const PointRaw& point)
 {
-  return SDL_GetDisplayForPoint(point);
+  return SDL_GetDisplayForPoint(&point);
 }
 
 inline Display Display::GetForPoint(const PointRaw& point)
@@ -3999,7 +4021,7 @@
  */
 inline Display GetDisplayForRect(const RectRaw& rect)
 {
-  return CheckError(SDL_GetDisplayForRect(rect));
+  return CheckError(SDL_GetDisplayForRect(&rect));
 }
 
 inline Display Display::GetForRect(const RectRaw& rect)
@@ -4148,7 +4170,7 @@
  */
 inline const DisplayMode& GetWindowFullscreenMode(WindowParam window)
 {
-  return SDL_GetWindowFullscreenMode(window);
+  return *SDL_GetWindowFullscreenMode(window);
 }
 
 inline const DisplayMode& Window::GetFullscreenMode() const
@@ -4160,7 +4182,6 @@
  * Get the raw ICC profile data for the screen the window is currently on.
  *
  * @param window the window to query.
- * @param size the size of the ICC profile.
  * @returns the raw ICC profile data on success.
  * @throws Error on failure.
  *
@@ -4170,7 +4191,8 @@
  */
 inline OwnPtr<void> GetWindowICCProfile(WindowParam window)
 {
-  return CheckError(SDL_GetWindowICCProfile(window));
+  size_t size;
+  return OwnPtr<void>{CheckError(SDL_GetWindowICCProfile(window, &size))};
 }
 
 inline OwnPtr<void> Window::GetICCProfile() const
@@ -4202,17 +4224,19 @@
 /**
  * Get a list of valid windows.
  *
- * @param count a pointer filled in with the number of windows returned, may be
- *              nullptr.
  * @returns a nullptr terminated array of Window pointers or nullptr on failure;
- *          call GetError() for more information. This is a single allocation
- *          that should be freed with free() when it is no longer needed.
+ *          call GetError() for more information.
  *
  * @threadsafety This function should only be called on the main thread.
  *
  * @since This function is available since SDL 3.2.0.
  */
-inline OwnArray<WindowRef> GetWindows() { return SDL_GetWindows(); }
+inline OwnArray<WindowRef> GetWindows()
+{
+  int count = 0;
+  auto data = CheckError(SDL_GetWindows(&count));
+  return OwnArray<WindowRef>{reinterpret_cast<WindowRef*>(data), size_t(count)};
+}
 
 /**
  * Create a window with the specified dimensions and flags.
@@ -4285,8 +4309,7 @@
  * in a future version of SDL.
  *
  * @param title the title of the window, in UTF-8 encoding.
- * @param w the width of the window.
- * @param h the height of the window.
+ * @param size the width and height of the window.
  * @param flags 0, or one or more WindowFlags OR'd together.
  * @returns the window that was created or nullptr on failure; call GetError()
  *          for more information.
@@ -4361,12 +4384,9 @@
  * hidden will be restored when the parent is shown.
  *
  * @param parent the parent of the window, must not be nullptr.
- * @param offset_x the x position of the popup window relative to the origin of
- *                 the parent.
- * @param offset_y the y position of the popup window relative to the origin of
- *                 the parent window.
- * @param w the width of the window.
- * @param h the height of the window.
+ * @param offset the x, y position of the popup window relative to the origin
+ *               of the parent.
+ * @param size the width and height of the window.
  * @param flags WINDOW_TOOLTIP or WINDOW_POPUP_MENU, and zero or more additional
  *              WindowFlags OR'd together.
  * @returns the window that was created or nullptr on failure; call GetError()
@@ -4746,7 +4766,7 @@
  */
 inline WindowRef GetWindowFromID(WindowID id)
 {
-  return SDL_GetWindowFromID(id);
+  return {SDL_GetWindowFromID(id)};
 }
 
 inline WindowRef Window::FromID(WindowID id)
@@ -4769,7 +4789,7 @@
  */
 inline WindowRef GetWindowParent(WindowParam window)
 {
-  return CheckError(SDL_GetWindowParent(window));
+  return {CheckError(SDL_GetWindowParent(window))};
 }
 
 inline WindowRef Window::GetParent() const
@@ -4896,7 +4916,7 @@
  */
 inline PropertiesRef GetWindowProperties(WindowParam window)
 {
-  return CheckError(SDL_GetWindowProperties(window));
+  return {CheckError(SDL_GetWindowProperties(window))};
 }
 
 inline PropertiesRef Window::GetProperties() const
@@ -5034,9 +5054,7 @@
  * this is just a request, it can be denied by the windowing system.
  *
  * @param window the window to reposition.
- * @param x the x coordinate of the window, or `WINDOWPOS_CENTERED` or
- *          `WINDOWPOS_UNDEFINED`.
- * @param y the y coordinate of the window, or `WINDOWPOS_CENTERED` or
+ * @param p the coordinates of the window, or `WINDOWPOS_CENTERED` or
  *          `WINDOWPOS_UNDEFINED`.
  * @throws Error on failure.
  *
@@ -5049,7 +5067,7 @@
  */
 inline void SetWindowPosition(WindowParam window, const PointRaw& p)
 {
-  CheckError(SDL_SetWindowPosition(window, p));
+  CheckError(SDL_SetWindowPosition(window, p.x, p.y));
 }
 
 inline void Window::SetPosition(const PointRaw& p)
@@ -5094,10 +5112,7 @@
  * in the `x` or `y` parameter.
  *
  * @param window the window to query.
- * @param x a pointer filled in with the x position of the window, may be
- *          nullptr.
- * @param y a pointer filled in with the y position of the window, may be
- *          nullptr.
+ * @returns the position on success.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5108,7 +5123,9 @@
  */
 inline Point GetWindowPosition(WindowParam window)
 {
-  static_assert(false, "Not implemented");
+  Point p;
+  GetWindowPosition(window, &p.x, &p.y);
+  return p;
 }
 
 inline void Window::GetPosition(int* x, int* y) const
@@ -5143,8 +5160,7 @@
  * request, it can be denied by the windowing system.
  *
  * @param window the window to change.
- * @param w the width of the window, must be > 0.
- * @param h the height of the window, must be > 0.
+ * @param size the width and height of the window, must be > 0.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5157,7 +5173,7 @@
  */
 inline void SetWindowSize(WindowParam window, const PointRaw& size)
 {
-  CheckError(SDL_SetWindowSize(window, size));
+  CheckError(SDL_SetWindowSize(window, size.x, size.y));
 }
 
 inline void Window::SetSize(const PointRaw& size)
@@ -5198,8 +5214,7 @@
  * Renderer.GetOutputSize() to get the real client area size in pixels.
  *
  * @param window the window to query the width and height from.
- * @param w a pointer filled in with the width of the window, may be nullptr.
- * @param h a pointer filled in with the height of the window, may be nullptr.
+ * @returns a point with width and height on success
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5212,7 +5227,9 @@
  */
 inline Point GetWindowSize(WindowParam window)
 {
-  static_assert(false, "Not implemented");
+  Point p;
+  GetWindowSize(window, &p.x, &p.y);
+  return p;
 }
 
 inline void Window::GetSize(int* w, int* h) const
@@ -5222,16 +5239,43 @@
 
 inline Point Window::GetSize() const { return SDL::GetWindowSize(m_resource); }
 
+/**
+ * @brief Request the window's position and size to be set.
+ *
+ * @param window the window to query the width and height from.
+ * @param rect the rect containing the new coordinates
+ * @throws Error on failure.
+ *
+ * @threadsafety This function should only be called on the main thread.
+ *
+ * @sa WindowRef.SetPosition()
+ * @sa WindowRef.SetSize()
+ */
 inline void SetWindowRect(WindowParam window, Rect rect)
 {
-  static_assert(false, "Not implemented");
+  SetWindowPosition(window, rect.GetTopLeft());
+  SetWindowSize(window, rect.GetSize());
 }
 
 inline void Window::SetRect(Rect rect) { SDL::SetWindowRect(m_resource, rect); }
 
+/**
+ * Get the position and client size of a window.
+ *
+ * This is the current position of the window as last reported by the
+ * windowing system.
+ *
+ * The window pixel size may differ from its window coordinate size if the
+ * window is on a high pixel density display. Use Window.GetSizeInPixels()
+ * or RendererRef.GetOutputSize() to get the real client area size in pixels.
+ *
+ * @param window the window to query the width and height from.
+ * @return Rect with the position and size
+ * @throws Error on failure.
+ */
 inline Rect GetWindowRect(WindowParam window)
 {
-  static_assert(false, "Not implemented");
+  return Rect{GetWindowPosition(window), GetWindowSize(window)};
 }
 
 inline Rect Window::GetRect() const { return SDL::GetWindowRect(m_resource); }
@@ -5247,8 +5291,6 @@
  * interactible content.
  *
  * @param window the window to query.
- * @param rect a pointer filled in with the client area that is safe for
- *             interactive content.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5257,7 +5299,9 @@
  */
 inline Rect GetWindowSafeArea(WindowParam window)
 {
-  return CheckError(SDL_GetWindowSafeArea(window));
+  Rect rect;
+  CheckError(SDL_GetWindowSafeArea(window, &rect));
+  return rect;
 }
 
 inline Rect Window::GetSafeArea() const
@@ -5418,10 +5462,8 @@
  * Get the size of a window's client area, in pixels.
  *
  * @param window the window from which the drawable size should be queried.
- * @param w a pointer to variable for storing the width in pixels, may be
- *          nullptr.
- * @param h a pointer to variable for storing the height in pixels, may be
- *          nullptr.
+ * @returns the size on success or std::nullopt on failure; call GetError()
+ * for more information.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5433,7 +5475,9 @@
  */
 inline Point GetWindowSizeInPixels(WindowParam window)
 {
-  static_assert(false, "Not implemented");
+  Point p;
+  GetWindowSizeInPixels(window, &p.x, &p.y);
+  return p;
 }
 
 inline void Window::GetSizeInPixels(int* w, int* h) const
@@ -5450,8 +5494,7 @@
  * Set the minimum size of a window's client area.
  *
  * @param window the window to change.
- * @param min_w the minimum width of the window, or 0 for no limit.
- * @param min_h the minimum height of the window, or 0 for no limit.
+ * @param p the minimum width and height of the window, or 0 for no limit.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5463,7 +5506,7 @@
  */
 inline void SetWindowMinimumSize(WindowParam window, const PointRaw& p)
 {
-  CheckError(SDL_SetWindowMinimumSize(window, p));
+  CheckError(SDL_SetWindowMinimumSize(window, p.x, p.y));
 }
 
 inline void Window::SetMinimumSize(const PointRaw& p)
@@ -5502,8 +5545,7 @@
  * Set the maximum size of a window's client area.
  *
  * @param window the window to change.
- * @param max_w the maximum width of the window, or 0 for no limit.
- * @param max_h the maximum height of the window, or 0 for no limit.
+ * @param p the maximum width and height of the window, or 0 for no limit.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5515,7 +5557,7 @@
  */
 inline void SetWindowMaximumSize(WindowParam window, const PointRaw& p)
 {
-  CheckError(SDL_SetWindowMaximumSize(window, p));
+  CheckError(SDL_SetWindowMaximumSize(window, p.x, p.y));
 }
 
 inline void Window::SetMaximumSize(const PointRaw& p)
@@ -5928,7 +5970,7 @@
  */
 inline Surface GetWindowSurface(WindowParam window)
 {
-  return SDL_GetWindowSurface(window);
+  return Surface::Borrow(SDL_GetWindowSurface(window));
 }
 
 inline Surface Window::GetSurface()
@@ -5969,16 +6011,16 @@
   SDL::SetWindowSurfaceVSync(m_resource, vsync);
 }
 
+/// Constant to disable vsync
 constexpr int WINDOW_SURFACE_VSYNC_DISABLED = SDL_WINDOW_SURFACE_VSYNC_DISABLED;
 
+/// Constant to enable adaptive vsync
 constexpr int WINDOW_SURFACE_VSYNC_ADAPTIVE = SDL_WINDOW_SURFACE_VSYNC_ADAPTIVE;
 
 /**
  * Get VSync for the window surface.
  *
  * @param window the window to query.
- * @param vsync an int filled with the current vertical refresh sync interval.
- *              See Window.SetSurfaceVSync() for the meaning of the value.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -5989,7 +6031,9 @@
  */
 inline int GetWindowSurfaceVSync(WindowParam window)
 {
-  return CheckError(SDL_GetWindowSurfaceVSync(window));
+  int vsync;
+  CheckError(SDL_GetWindowSurfaceVSync(window, &vsync));
+  return vsync;
 }
 
 inline int Window::GetSurfaceVSync() const
@@ -6038,7 +6082,6 @@
  * @param window the window to update.
  * @param rects an array of Rect structures representing areas of the surface to
  *              copy, in pixels.
- * @param numrects the number of rectangles.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -6051,7 +6094,7 @@
 inline void UpdateWindowSurfaceRects(WindowParam window,
                                      SpanRef<const RectRaw> rects)
 {
-  CheckError(SDL_UpdateWindowSurfaceRects(window, rects));
+  CheckError(SDL_UpdateWindowSurfaceRects(window, rects.data(), rects.size()));
 }
 
 inline void Window::UpdateSurfaceRects(SpanRef<const RectRaw> rects)
@@ -6205,7 +6248,7 @@
  * @sa Window.SetMouseGrab
  * @sa Window.SetKeyboardGrab
  */
-inline WindowRef GetGrabbedWindow() { return SDL_GetGrabbedWindow(); }
+inline WindowRef GetGrabbedWindow() { return {SDL_GetGrabbedWindow()}; }
 
 inline WindowRef Window::GetGrabbed() { return SDL::GetGrabbedWindow(); }
 
@@ -6230,7 +6273,7 @@
  */
 inline void SetWindowMouseRect(WindowParam window, const RectRaw& rect)
 {
-  CheckError(SDL_SetWindowMouseRect(window, rect));
+  CheckError(SDL_SetWindowMouseRect(window, &rect));
 }
 
 inline void Window::SetMouseRect(const RectRaw& rect)
@@ -6418,9 +6461,7 @@
  * nothing.
  *
  * @param window the window for which the menu will be displayed.
- * @param x the x coordinate of the menu, relative to the origin (top-left) of
- *          the client area.
- * @param y the y coordinate of the menu, relative to the origin (top-left) of
+ * @param p the coordinates of the menu, relative to the origin (top-left) of
  *          the client area.
  * @throws Error on failure.
  *
@@ -6430,7 +6471,7 @@
  */
 inline void ShowWindowSystemMenu(WindowParam window, const PointRaw& p)
 {
-  CheckError(SDL_ShowWindowSystemMenu(window, p));
+  CheckError(SDL_ShowWindowSystemMenu(window, p.x, p.y));
 }
 
 inline void Window::ShowSystemMenu(const PointRaw& p)
@@ -6483,6 +6524,8 @@
                              HitTest callback,
                              void* callback_data)
 {
+  using Wrapper = KeyValueCallbackWrapper<WindowRaw, HitTestCB>;
+  Wrapper::erase(window);
   CheckError(SDL_SetWindowHitTest(window, callback, callback_data));
 }
 
@@ -6520,7 +6563,6 @@
  *
  * @param window the window to set hit-testing on.
  * @param callback the function to call when doing a hit-test.
- * @param callback_data an app-defined void pointer passed to **callback**.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
@@ -6529,7 +6571,14 @@
  */
 inline void SetWindowHitTest(WindowParam window, HitTestCB callback)
 {
-  static_assert(false, "Not implemented");
+  using Wrapper = KeyValueCallbackWrapper<WindowRaw, HitTestCB>;
+  void* cbHandle = Wrapper::Wrap(window, std::move(callback));
+  SetWindowHitTest(
+    window,
+    [](SDL_Window* win, const SDL_Point* area, void* data) {
+      return Wrapper::Call(data, win, Point(*area));
+    },
+    cbHandle);
 }
 
 inline void Window::SetHitTest(HitTest callback, void* callback_data)
@@ -6616,7 +6665,12 @@
  * @sa Window.Window
  * @sa Window.Window
  */
-inline void DestroyWindow(WindowRaw window) { SDL_DestroyWindow(window); }
+inline void DestroyWindow(WindowRaw window)
+{
+  using Wrapper = KeyValueCallbackWrapper<WindowRaw, HitTestCB>;
+  Wrapper::erase(window);
+  SDL_DestroyWindow(window);
+}
 
 inline void Window::Destroy() { DestroyWindow(release()); }
 
@@ -6915,7 +6969,7 @@
  */
 inline void GL_MakeCurrent(WindowParam window, GLContext context)
 {
-  CheckError(SDL_GL_MakeCurrent(window, context));
+  CheckError(SDL_GL_MakeCurrent(window, context.get()));
 }
 
 inline void Window::MakeCurrent(GLContext context)
@@ -6925,7 +6979,7 @@
 
 inline void GLContext::MakeCurrent(WindowParam window)
 {
-  SDL::Window::MakeCurrent(m_resource, window);
+  SDL::GL_MakeCurrent(window, m_resource);
 }
 
 /**
@@ -6941,7 +6995,7 @@
  */
 inline WindowRef GL_GetCurrentWindow()
 {
-  return CheckError(SDL_GL_GetCurrentWindow());
+  return {CheckError(SDL_GL_GetCurrentWindow())};
 }
 
 /**
