--- build/generated/SDL3pp_ttf.h
+++ include/SDL3pp/SDL3pp_ttf.h
@@ -1,17 +1,20 @@
 #ifndef SDL3PP_TTF_H_
 #define SDL3PP_TTF_H_
 
-#include <SDL3/SDL_ttf.h>
 #include "SDL3pp_gpu.h"
 #include "SDL3pp_render.h"
 #include "SDL3pp_surface.h"
 #include "SDL3pp_version.h"
 #include "SDL3pp_video.h"
 
+#if defined(SDL3PP_ENABLE_TTF) || defined(SDL3PP_DOC)
+
+#include <SDL3_ttf/SDL_ttf.h>
+
 namespace SDL {
 
 /**
- * @defgroup CategorySDLTTF Category SDLTTF
+ * @defgroup CategorySDLTTF True type font support
  *
  * Header file for SDL_ttf library
  *
@@ -78,25 +81,39 @@
 /// Safely wrap Text for non owning const parameters
 using TextConstRef = ObjConstParam<TextRaw, TextRawConst>;
 
-/// Printable format: "%d.%d.%d", MAJOR, MINOR, MICRO
+#ifdef SDL3PP_DOC
+
+/**
+ * @name TTF version
+ * @{
+ * Printable format: "%d.%d.%d", MAJOR, MINOR, MICRO
+ */
 #define SDL_TTF_MAJOR_VERSION
 
 #define SDL_TTF_MINOR_VERSION
 
 #define SDL_TTF_MICRO_VERSION
 
-/// This is the version number macro for the current SDL_ttf version.
+///@}
+
+/**
+ * This is the version number macro for the current SDL_ttf version.
+ */
 #define SDL_TTF_VERSION                                                        \
   SDL_VERSIONNUM(                                                              \
     SDL_TTF_MAJOR_VERSION, SDL_TTF_MINOR_VERSION, SDL_TTF_MICRO_VERSION)
 
-/// This macro will evaluate to true if compiled with SDL_ttf at least X.Y.Z.
+/**
+ * This macro will evaluate to true if compiled with SDL_ttf at least X.Y.Z.
+ */
 #define SDL_TTF_VERSION_ATLEAST(X, Y, Z)                                       \
   ((SDL_TTF_MAJOR_VERSION >= X) &&                                             \
    (SDL_TTF_MAJOR_VERSION > X || SDL_TTF_MINOR_VERSION >= Y) &&                \
    (SDL_TTF_MAJOR_VERSION > X || SDL_TTF_MINOR_VERSION > Y ||                  \
     SDL_TTF_MICRO_VERSION >= Z))
 
+#endif // SDL3PP_DOC
+
 namespace TTF {
 
 /**
@@ -301,7 +318,7 @@
  *
  * The values here are chosen to match
  * [hb_direction_t](https://harfbuzz.github.io/harfbuzz-hb-common.html#hb-direction-t)
- * .
+ *
  *
  * @since This enum is available since SDL_ttf 3.0.0.
  *
@@ -387,8 +404,6 @@
    * size becomes the index of choosing which size. If the value is too high,
    * the last indexed size will be the default.
    *
-   * When done with the returned Font, use Font.Close() to dispose of it.
-   *
    * @param file path to font file.
    * @param ptsize point size to use for the newly-opened font.
    * @post a valid Font on success.
@@ -410,14 +425,12 @@
    * the last indexed size will be the default.
    *
    * If `closeio` is true, `src` will be automatically closed once the font is
-   * closed. Otherwise you should close `src` yourself after closing the font.
-   *
-   * When done with the returned Font, use Font.Close() to dispose of it.
+   * closed. Otherwise you should keep `src` open until the font is closed.
    *
    * @param src an IOStream to provide a font file's data.
+   * @param ptsize point size to use for the newly-opened font.
    * @param closeio true to close `src` when the font is closed, false to leave
    *                it open.
-   * @param ptsize point size to use for the newly-opened font.
    * @post a valid Font on success.
    * @throws Error on failure.
    *
@@ -437,11 +450,11 @@
    * - `prop::Font.CREATE_FILENAME_STRING`: the font file to open, if an
    *   IOStream isn't being used. This is required if
    *   `prop::Font.CREATE_IOSTREAM_POINTER` and
-   *   `prop::Font.CREATE_EXISTING_FONT` aren't set.
+   *   `prop::Font.CREATE_EXISTING_FONT_POINTER` aren't set.
    * - `prop::Font.CREATE_IOSTREAM_POINTER`: an IOStream containing the font to
    *   be opened. This should not be closed until the font is closed. This is
    *   required if `prop::Font.CREATE_FILENAME_STRING` and
-   *   `prop::Font.CREATE_EXISTING_FONT` aren't set.
+   *   `prop::Font.CREATE_EXISTING_FONT_POINTER` aren't set.
    * - `prop::Font.CREATE_IOSTREAM_OFFSET_NUMBER`: the offset in the iostream
    *   for the beginning of the font, defaults to 0.
    * - `prop::Font.CREATE_IOSTREAM_AUTOCLOSE_BOOLEAN`: true if closing the font
@@ -458,9 +471,9 @@
    * - `prop::Font.CREATE_VERTICAL_DPI_NUMBER`: the vertical DPI to use for font
    *   rendering, defaults to `prop::Font.CREATE_HORIZONTAL_DPI_NUMBER` if set,
    *   or 72 otherwise.
-   * - `prop::Font.CREATE_EXISTING_FONT`: an optional Font that, if set, will be
-   *   used as the font data source and the initial size and style of the new
-   *   font.
+   * - `prop::Font.CREATE_EXISTING_FONT_POINTER`: an optional Font that, if set,
+   *   will be used as the font data source and the initial size and style of
+   *   the new font.
    *
    * @param props the properties to use.
    * @post a valid Font on success.
@@ -522,7 +535,6 @@
    * @since This function is available since SDL_ttf 3.0.0.
    *
    * @sa Font.Font
-   * @sa Font.Font
    */
   void Close();
 
@@ -532,8 +544,6 @@
    * The copy will be distinct from the original, but will share the font file
    * and have the same size and style as the original.
    *
-   * When done with the returned Font, use Font.Close() to dispose of it.
-   *
    * @returns a valid Font on success.
    * @throws Error on failure.
    *
@@ -895,7 +905,7 @@
    * @threadsafety This function should be called on the thread that created the
    *               font.
    *
-   * @since This function is available since SDL_ttf 3.4.0.
+   * @since This function is available since SDL_ttf 3.2.2.
    */
   int GetWeight() const;
 
@@ -1138,7 +1148,7 @@
    * This updates any Text objects using this font.
    *
    * @param script an [ISO 15924
-   *               code](https://unicode.org/iso15924/iso15924-codes.html) .
+   *               code](https://unicode.org/iso15924/iso15924-codes.html).
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -1290,8 +1300,7 @@
    *
    * @param previous_ch the previous codepoint.
    * @param ch the current codepoint.
-   * @param kerning a pointer filled in with the kerning size between the two
-   *                glyphs, in pixels, may be nullptr.
+   * @returns the kerning size between the two glyphs, in pixels.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -1301,9 +1310,26 @@
    */
   int GetGlyphKerning(Uint32 previous_ch, Uint32 ch) const;
 
+  /**
+   * Calculate the dimensions of a rendered string of UTF-8 text.
+   *
+   * This will report the width and height, in pixels, of the space that the
+   * specified string will take to fully render.
+   *
+   * @param text text to calculate, in UTF-8 encoding.
+   * @returns return a Point with the width, height.
+   * @throws Error on failure.
+   *
+   * @threadsafety This function should be called on the thread that created the
+   *               font.
+   *
+   * @since This function is available since SDL_ttf 3.0.0.
+   */
   Point GetStringSize(std::string_view text) const
   {
-    static_assert(false, "Not implemented");
+    Point p;
+    GetStringSize(text, &p.x, &p.y);
+    return p;
   }
 
   /**
@@ -1312,10 +1338,7 @@
    * This will report the width and height, in pixels, of the space that the
    * specified string will take to fully render.
    *
-   * @param font the font to query.
    * @param text text to calculate, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param w will be filled with width, in pixels, on return.
    * @param h will be filled with height, in pixels, on return.
    * @throws Error on failure.
@@ -1327,9 +1350,32 @@
    */
   void GetStringSize(std::string_view text, int* w, int* h) const;
 
+  /**
+   * Calculate the dimensions of a rendered string of UTF-8 text.
+   *
+   * This will report the width and height, in pixels, of the space that the
+   * specified string will take to fully render.
+   *
+   * Text is wrapped to multiple lines on line endings and on word boundaries if
+   * it extends beyond `wrap_width` in pixels.
+   *
+   * If wrap_width is 0, this function will only wrap on newline characters.
+   *
+   * @param text text to calculate, in UTF-8 encoding.
+   * @param wrap_width the maximum width or 0 to wrap on newline characters.
+   * @returns return a Point with the width, height on success.
+   * @throws Error on failure.
+   *
+   * @threadsafety This function should be called on the thread that created the
+   *               font.
+   *
+   * @since This function is available since SDL_ttf 3.0.0.
+   */
   Point GetStringSizeWrapped(std::string_view text, int wrap_width) const
   {
-    static_assert(false, "Not implemented");
+    Point p;
+    GetStringSizeWrapped(text, wrap_width, &p.x, &p.y);
+    return p;
   }
 
   /**
@@ -1343,10 +1389,7 @@
    *
    * If wrap_width is 0, this function will only wrap on newline characters.
    *
-   * @param font the font to query.
    * @param text text to calculate, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param wrap_width the maximum width or 0 to wrap on newline characters.
    * @param w will be filled with width, in pixels, on return.
    * @param h will be filled with height, in pixels, on return.
@@ -1370,10 +1413,7 @@
    *
    * This does not need to render the string to do this calculation.
    *
-   * @param font the font to query.
    * @param text text to calculate, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param max_width maximum width, in pixels, available for the string, or 0
    *                  for unbounded width.
    * @param measured_width a pointer filled in with the width, in pixels, of the
@@ -1409,10 +1449,7 @@
    * You can render at other quality levels with Font.RenderText_Shaded,
    * Font.RenderText_Blended, and Font.RenderText_LCD.
    *
-   * @param font the font to render with.
    * @param text text to render, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param fg the foreground color for the text.
    * @returns a new 8-bit, palettized surface, or nullptr if there was an error.
    *
@@ -1444,10 +1481,7 @@
    * You can render at other quality levels with Font.RenderText_Shaded_Wrapped,
    * Font.RenderText_Blended_Wrapped, and Font.RenderText_LCD_Wrapped.
    *
-   * @param font the font to render with.
    * @param text text to render, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param fg the foreground color for the text.
    * @param wrapLength the maximum width of the text surface or 0 to wrap on
    *                   newline characters.
@@ -1513,10 +1547,7 @@
    * You can render at other quality levels with Font.RenderText_Solid,
    * Font.RenderText_Blended, and Font.RenderText_LCD.
    *
-   * @param font the font to render with.
    * @param text text to render, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param fg the foreground color for the text.
    * @param bg the background color for the text.
    * @returns a new 8-bit, palettized surface, or nullptr if there was an error.
@@ -1549,10 +1580,7 @@
    * You can render at other quality levels with Font.RenderText_Solid_Wrapped,
    * Font.RenderText_Blended_Wrapped, and Font.RenderText_LCD_Wrapped.
    *
-   * @param font the font to render with.
    * @param text text to render, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param fg the foreground color for the text.
    * @param bg the background color for the text.
    * @param wrap_width the maximum width of the text surface or 0 to wrap on
@@ -1621,10 +1649,7 @@
    * You can render at other quality levels with Font.RenderText_Solid,
    * Font.RenderText_Shaded, and Font.RenderText_LCD.
    *
-   * @param font the font to render with.
    * @param text text to render, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param fg the foreground color for the text.
    * @returns a new 32-bit, ARGB surface, or nullptr if there was an error.
    *
@@ -1655,10 +1680,7 @@
    * You can render at other quality levels with Font.RenderText_Solid_Wrapped,
    * Font.RenderText_Shaded_Wrapped, and Font.RenderText_LCD_Wrapped.
    *
-   * @param font the font to render with.
    * @param text text to render, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param fg the foreground color for the text.
    * @param wrap_width the maximum width of the text surface or 0 to wrap on
    *                   newline characters.
@@ -1723,10 +1745,7 @@
    * You can render at other quality levels with Font.RenderText_Solid,
    * Font.RenderText_Shaded, and Font.RenderText_Blended.
    *
-   * @param font the font to render with.
    * @param text text to render, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param fg the foreground color for the text.
    * @param bg the background color for the text.
    * @returns a new 32-bit, ARGB surface, or nullptr if there was an error.
@@ -1759,10 +1778,7 @@
    * You can render at other quality levels with Font.RenderText_Solid_Wrapped,
    * Font.RenderText_Shaded_Wrapped, and Font.RenderText_Blended_Wrapped.
    *
-   * @param font the font to render with.
    * @param text text to render, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @param fg the foreground color for the text.
    * @param bg the background color for the text.
    * @param wrap_width the maximum width of the text surface or 0 to wrap on
@@ -1920,14 +1936,14 @@
  * last indexed size will be the default.
  *
  * If `closeio` is true, `src` will be automatically closed once the font is
- * closed. Otherwise you should close `src` yourself after closing the font.
+ * closed. Otherwise you should keep `src` open until the font is closed.
  *
  * When done with the returned Font, use Font.Close() to dispose of it.
  *
  * @param src an IOStream to provide a font file's data.
+ * @param ptsize point size to use for the newly-opened font.
  * @param closeio true to close `src` when the font is closed, false to leave it
  *                open.
- * @param ptsize point size to use for the newly-opened font.
  * @returns a valid Font on success.
  * @throws Error on failure.
  *
@@ -1948,7 +1964,7 @@
 }
 
 inline Font::Font(IOStreamRef src, float ptsize, bool closeio)
-  : m_resource(CheckError(TTF_OpenFontIO(src, ptsize, closeio)))
+  : m_resource(CheckError(TTF_OpenFontIO(src, closeio, ptsize)))
 {
 }
 
@@ -1964,11 +1980,11 @@
  *
  * - `prop::Font.CREATE_FILENAME_STRING`: the font file to open, if an IOStream
  *   isn't being used. This is required if `prop::Font.CREATE_IOSTREAM_POINTER`
- *   and `prop::Font.CREATE_EXISTING_FONT` aren't set.
+ *   and `prop::Font.CREATE_EXISTING_FONT_POINTER` aren't set.
  * - `prop::Font.CREATE_IOSTREAM_POINTER`: an IOStream containing the font to be
  *   opened. This should not be closed until the font is closed. This is
  *   required if `prop::Font.CREATE_FILENAME_STRING` and
- *   `prop::Font.CREATE_EXISTING_FONT` aren't set.
+ *   `prop::Font.CREATE_EXISTING_FONT_POINTER` aren't set.
  * - `prop::Font.CREATE_IOSTREAM_OFFSET_NUMBER`: the offset in the iostream for
  *   the beginning of the font, defaults to 0.
  * - `prop::Font.CREATE_IOSTREAM_AUTOCLOSE_BOOLEAN`: true if closing the font
@@ -1985,9 +2001,9 @@
  * - `prop::Font.CREATE_VERTICAL_DPI_NUMBER`: the vertical DPI to use for font
  *   rendering, defaults to `prop::Font.CREATE_HORIZONTAL_DPI_NUMBER` if set, or
  *   72 otherwise.
- * - `prop::Font.CREATE_EXISTING_FONT`: an optional Font that, if set, will be
- *   used as the font data source and the initial size and style of the new
- *   font.
+ * - `prop::Font.CREATE_EXISTING_FONT_POINTER`: an optional Font that, if set,
+ *   will be used as the font data source and the initial size and style of the
+ *   new font.
  *
  * @param props the properties to use.
  * @returns a valid Font on success.
@@ -2025,7 +2041,8 @@
 
 #if SDL_TTF_VERSION_ATLEAST(3, 2, 2)
 
-constexpr auto CREATE_EXISTING_FONT = TTF_PROP_FONT_CREATE_EXISTING_FONT;
+constexpr auto CREATE_EXISTING_FONT_POINTER =
+  TTF_PROP_FONT_CREATE_EXISTING_FONT;
 
 #endif // SDL_TTF_VERSION_ATLEAST(3, 2, 2)
 
@@ -2060,7 +2077,7 @@
  */
 inline Font CopyFont(FontRef existing_font)
 {
-  return CheckError(TTF_CopyFont(existing_font));
+  return Font(CheckError(TTF_CopyFont(existing_font)));
 }
 
 inline Font Font::Copy() const { return SDL::CopyFont(m_resource); }
@@ -2087,7 +2104,7 @@
  */
 inline PropertiesRef GetFontProperties(FontRef font)
 {
-  return CheckError(TTF_GetFontProperties(font));
+  return {CheckError(TTF_GetFontProperties(font))};
 }
 
 inline PropertiesRef Font::GetProperties()
@@ -2548,7 +2565,7 @@
  * @threadsafety This function should be called on the thread that created the
  *               font.
  *
- * @since This function is available since SDL_ttf 3.4.0.
+ * @since This function is available since SDL_ttf 3.2.2.
  */
 inline int GetFontWeight(FontRef font) { return TTF_GetFontWeight(font); }
 
@@ -2963,7 +2980,7 @@
  *
  * @param font the font to modify.
  * @param script an [ISO 15924
- *               code](https://unicode.org/iso15924/iso15924-codes.html) .
+ *               code](https://unicode.org/iso15924/iso15924-codes.html).
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -3092,7 +3109,7 @@
  */
 inline Surface GetGlyphImage(FontRef font, Uint32 ch, ImageType* image_type)
 {
-  return CheckError(TTF_GetGlyphImage(font, ch, image_type));
+  return Surface{CheckError(TTF_GetGlyphImage(font, ch, image_type))};
 }
 
 inline Surface Font::GetGlyphImage(Uint32 ch, ImageType* image_type) const
@@ -3122,7 +3139,8 @@
                                      Uint32 glyph_index,
                                      ImageType* image_type)
 {
-  return CheckError(TTF_GetGlyphImageForIndex(font, glyph_index, image_type));
+  return Surface(
+    CheckError(TTF_GetGlyphImageForIndex(font, glyph_index, image_type)));
 }
 
 inline Surface Font::GetGlyphImageForIndex(Uint32 glyph_index,
@@ -3186,8 +3204,7 @@
  * @param font the font to query.
  * @param previous_ch the previous codepoint.
  * @param ch the current codepoint.
- * @param kerning a pointer filled in with the kerning size between the two
- *                glyphs, in pixels, may be nullptr.
+ * @returns the kerning size between the two glyphs, in pixels.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -3197,7 +3214,8 @@
  */
 inline int GetGlyphKerning(FontRef font, Uint32 previous_ch, Uint32 ch)
 {
-  return CheckError(TTF_GetGlyphKerning(font, previous_ch, ch));
+  if (int r; TTF_GetGlyphKerning(font, previous_ch, ch, &r)) return r;
+  throw Error();
 }
 
 inline int Font::GetGlyphKerning(Uint32 previous_ch, Uint32 ch) const
@@ -3213,8 +3231,6 @@
  *
  * @param font the font to query.
  * @param text text to calculate, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param w will be filled with width, in pixels, on return.
  * @param h will be filled with height, in pixels, on return.
  * @throws Error on failure.
@@ -3226,7 +3242,7 @@
  */
 inline void GetStringSize(FontRef font, std::string_view text, int* w, int* h)
 {
-  CheckError(TTF_GetStringSize(font, text, w, h));
+  CheckError(TTF_GetStringSize(font, text.data(), text.size(), w, h));
 }
 
 inline void Font::GetStringSize(std::string_view text, int* w, int* h) const
@@ -3247,8 +3263,6 @@
  *
  * @param font the font to query.
  * @param text text to calculate, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param wrap_width the maximum width or 0 to wrap on newline characters.
  * @param w will be filled with width, in pixels, on return.
  * @param h will be filled with height, in pixels, on return.
@@ -3265,7 +3279,8 @@
                                  int* w,
                                  int* h)
 {
-  CheckError(TTF_GetStringSizeWrapped(font, text, wrap_width, w, h));
+  CheckError(
+    TTF_GetStringSizeWrapped(font, text.data(), text.size(), wrap_width, w, h));
 }
 
 inline void Font::GetStringSizeWrapped(std::string_view text,
@@ -3286,8 +3301,6 @@
  *
  * @param font the font to query.
  * @param text text to calculate, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param max_width maximum width, in pixels, available for the string, or 0 for
  *                  unbounded width.
  * @param measured_width a pointer filled in with the width, in pixels, of the
@@ -3307,8 +3320,12 @@
                           int* measured_width,
                           size_t* measured_length)
 {
-  CheckError(
-    TTF_MeasureString(font, text, max_width, measured_width, measured_length));
+  CheckError(TTF_MeasureString(font,
+                               text.data(),
+                               text.size(),
+                               max_width,
+                               measured_width,
+                               measured_length));
 }
 
 inline void Font::MeasureString(std::string_view text,
@@ -3339,8 +3356,6 @@
  *
  * @param font the font to render with.
  * @param text text to render, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param fg the foreground color for the text.
  * @returns a new 8-bit, palettized surface, or nullptr if there was an error.
  *
@@ -3357,7 +3372,7 @@
  */
 inline Surface RenderText_Solid(FontRef font, std::string_view text, Color fg)
 {
-  return TTF_RenderText_Solid(font, text, fg);
+  return Surface{TTF_RenderText_Solid(font, text.data(), text.size(), fg)};
 }
 
 inline Surface Font::RenderText_Solid(std::string_view text, Color fg) const
@@ -3382,8 +3397,6 @@
  *
  * @param font the font to render with.
  * @param text text to render, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param fg the foreground color for the text.
  * @param wrapLength the maximum width of the text surface or 0 to wrap on
  *                   newline characters.
@@ -3404,7 +3417,8 @@
                                         Color fg,
                                         int wrapLength)
 {
-  return TTF_RenderText_Solid_Wrapped(font, text, fg, wrapLength);
+  return Surface(TTF_RenderText_Solid_Wrapped(
+    font, text.data(), text.size(), fg, wrapLength));
 }
 
 inline Surface Font::RenderText_Solid_Wrapped(std::string_view text,
@@ -3443,7 +3457,7 @@
  */
 inline Surface RenderGlyph_Solid(FontRef font, Uint32 ch, ColorRaw fg)
 {
-  return TTF_RenderGlyph_Solid(font, ch, fg);
+  return Surface(TTF_RenderGlyph_Solid(font, ch, fg));
 }
 
 inline Surface Font::RenderGlyph_Solid(Uint32 ch, ColorRaw fg) const
@@ -3471,8 +3485,6 @@
  *
  * @param font the font to render with.
  * @param text text to render, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param fg the foreground color for the text.
  * @param bg the background color for the text.
  * @returns a new 8-bit, palettized surface, or nullptr if there was an error.
@@ -3492,7 +3504,7 @@
                                  Color fg,
                                  Color bg)
 {
-  return TTF_RenderText_Shaded(font, text, fg, bg);
+  return Surface(TTF_RenderText_Shaded(font, text.data(), text.size(), fg, bg));
 }
 
 inline Surface Font::RenderText_Shaded(std::string_view text,
@@ -3520,8 +3532,6 @@
  *
  * @param font the font to render with.
  * @param text text to render, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param fg the foreground color for the text.
  * @param bg the background color for the text.
  * @param wrap_width the maximum width of the text surface or 0 to wrap on
@@ -3544,7 +3554,8 @@
                                          Color bg,
                                          int wrap_width)
 {
-  return TTF_RenderText_Shaded_Wrapped(font, text, fg, bg, wrap_width);
+  return Surface(TTF_RenderText_Shaded_Wrapped(
+    font, text.data(), text.size(), fg, bg, wrap_width));
 }
 
 inline Surface Font::RenderText_Shaded_Wrapped(std::string_view text,
@@ -3589,7 +3600,7 @@
                                   ColorRaw fg,
                                   ColorRaw bg)
 {
-  return TTF_RenderGlyph_Shaded(font, ch, fg, bg);
+  return Surface(TTF_RenderGlyph_Shaded(font, ch, fg, bg));
 }
 
 inline Surface Font::RenderGlyph_Shaded(Uint32 ch,
@@ -3618,8 +3629,6 @@
  *
  * @param font the font to render with.
  * @param text text to render, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param fg the foreground color for the text.
  * @returns a new 32-bit, ARGB surface, or nullptr if there was an error.
  *
@@ -3635,7 +3644,7 @@
  */
 inline Surface RenderText_Blended(FontRef font, std::string_view text, Color fg)
 {
-  return TTF_RenderText_Blended(font, text, fg);
+  return Surface(TTF_RenderText_Blended(font, text.data(), text.size(), fg));
 }
 
 inline Surface Font::RenderText_Blended(std::string_view text, Color fg) const
@@ -3660,8 +3669,6 @@
  *
  * @param font the font to render with.
  * @param text text to render, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param fg the foreground color for the text.
  * @param wrap_width the maximum width of the text surface or 0 to wrap on
  *                   newline characters.
@@ -3682,7 +3689,8 @@
                                           Color fg,
                                           int wrap_width)
 {
-  return TTF_RenderText_Blended_Wrapped(font, text, fg, wrap_width);
+  return Surface(TTF_RenderText_Blended_Wrapped(
+    font, text.data(), text.size(), fg, wrap_width));
 }
 
 inline Surface Font::RenderText_Blended_Wrapped(std::string_view text,
@@ -3721,7 +3729,7 @@
  */
 inline Surface RenderGlyph_Blended(FontRef font, Uint32 ch, ColorRaw fg)
 {
-  return TTF_RenderGlyph_Blended(font, ch, fg);
+  return Surface(TTF_RenderGlyph_Blended(font, ch, fg));
 }
 
 inline Surface Font::RenderGlyph_Blended(Uint32 ch, ColorRaw fg) const
@@ -3748,8 +3756,6 @@
  *
  * @param font the font to render with.
  * @param text text to render, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param fg the foreground color for the text.
  * @param bg the background color for the text.
  * @returns a new 32-bit, ARGB surface, or nullptr if there was an error.
@@ -3769,7 +3775,7 @@
                               Color fg,
                               Color bg)
 {
-  return TTF_RenderText_LCD(font, text, fg, bg);
+  return Surface(TTF_RenderText_LCD(font, text.data(), text.size(), fg, bg));
 }
 
 inline Surface Font::RenderText_LCD(std::string_view text,
@@ -3796,8 +3802,6 @@
  *
  * @param font the font to render with.
  * @param text text to render, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @param fg the foreground color for the text.
  * @param bg the background color for the text.
  * @param wrap_width the maximum width of the text surface or 0 to wrap on
@@ -3820,7 +3824,8 @@
                                       Color bg,
                                       int wrap_width)
 {
-  return TTF_RenderText_LCD_Wrapped(font, text, fg, bg, wrap_width);
+  return Surface(TTF_RenderText_LCD_Wrapped(
+    font, text.data(), text.size(), fg, bg, wrap_width));
 }
 
 inline Surface Font::RenderText_LCD_Wrapped(std::string_view text,
@@ -3865,7 +3870,7 @@
                                ColorRaw fg,
                                ColorRaw bg)
 {
-  return TTF_RenderGlyph_LCD(font, ch, fg, bg);
+  return Surface(TTF_RenderGlyph_LCD(font, ch, fg, bg));
 }
 
 inline Surface Font::RenderGlyph_LCD(Uint32 ch, ColorRaw fg, ColorRaw bg) const
@@ -3966,7 +3971,7 @@
   }
 
   /// Destructor
-  ~TextEngine() { Destroy(m_resource); }
+  virtual ~TextEngine() { SDL_assert_paranoid(!m_resource); }
 
   /// Assignment operator.
   constexpr TextEngine& operator=(TextEngine&& other) noexcept
@@ -3995,16 +4000,14 @@
   /// Converts to bool
   constexpr explicit operator bool() const noexcept { return !!m_resource; }
 
-  /// frees up textEngine.
-  void Destroy() { static_assert(false, "Not implemented"); }
+  /// frees up textEngine. Pure virtual
+  virtual void Destroy() = 0;
 
   /**
    * Create a text object from UTF-8 text and a text engine.
    *
    * @param font the font to render with.
    * @param text the text to use, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @returns a Text object or nullptr on failure; call GetError() for more
    *          information.
    *
@@ -4018,6 +4021,7 @@
   Text CreateText(FontRef font, std::string_view text);
 };
 
+/// A surface based text engine
 struct SurfaceTextEngine : TextEngine
 {
   /**
@@ -4035,7 +4039,11 @@
    */
   SurfaceTextEngine();
 
-  ~SurfaceTextEngine() { Destroy(); }
+  SurfaceTextEngine(const SurfaceTextEngine&) = delete;
+
+  SurfaceTextEngine& operator=(const SurfaceTextEngine&) = delete;
+
+  ~SurfaceTextEngine() final { Destroy(); }
 
   /**
    * Destroy a text engine created for drawing text on SDL surfaces.
@@ -4043,9 +4051,6 @@
    * All text created by this engine should be destroyed before calling this
    * function.
    *
-   * @param engine a TextEngine object created with
-   *               SurfaceTextEngine.SurfaceTextEngine().
-   *
    * @threadsafety This function should be called on the thread that created the
    *               engine.
    *
@@ -4053,9 +4058,10 @@
    *
    * @sa SurfaceTextEngine.SurfaceTextEngine
    */
-  void Destroy();
+  void Destroy() final;
 };
 
+/// A renderer based text engine
 struct RendererTextEngine : TextEngine
 {
   /**
@@ -4082,10 +4088,10 @@
    *
    * These are the supported properties:
    *
-   * - `prop::RendererTextEngine.RENDERER`: the renderer to use for creating
-   *   textures and drawing text
-   * - `prop::RendererTextEngine.ATLAS_TEXTURE_SIZE`: the size of the texture
-   *   atlas
+   * - `prop::RendererTextEngine.RENDERER_POINTER`: the renderer to use for
+   *   creating textures and drawing text
+   * - `prop::RendererTextEngine.ATLAS_TEXTURE_SIZE_NUMBER`: the size of the
+   *   texture atlas
    *
    * @param props the properties to use.
    * @post a TextEngine object or nullptr on failure; call GetError() for more
@@ -4102,7 +4108,11 @@
    */
   RendererTextEngine(PropertiesRef props);
 
-  ~RendererTextEngine() { Destroy(); }
+  RendererTextEngine(const RendererTextEngine&) = delete;
+
+  RendererTextEngine& operator=(const RendererTextEngine&) = delete;
+
+  ~RendererTextEngine() final { Destroy(); }
 
   /**
    * Destroy a text engine created for drawing text on an SDL renderer.
@@ -4110,9 +4120,6 @@
    * All text created by this engine should be destroyed before calling this
    * function.
    *
-   * @param engine a TextEngine object created with
-   *               RendererTextEngine.RendererTextEngine().
-   *
    * @threadsafety This function should be called on the thread that created the
    *               engine.
    *
@@ -4120,9 +4127,10 @@
    *
    * @sa RendererTextEngine.RendererTextEngine
    */
-  void Destroy();
+  void Destroy() final;
 };
 
+/// A GPU based text engine
 struct GPUTextEngine : TextEngine
 {
   /**
@@ -4149,9 +4157,10 @@
    *
    * These are the supported properties:
    *
-   * - `prop::GpuTextEngine.DEVICE`: the GPUDevice to use for creating textures
-   *   and drawing text.
-   * - `prop::GpuTextEngine.ATLAS_TEXTURE_SIZE`: the size of the texture atlas
+   * - `prop::GpuTextEngine.DEVICE_POINTER`: the GPUDevice to use for creating
+   *   textures and drawing text.
+   * - `prop::GpuTextEngine.ATLAS_TEXTURE_SIZE_NUMBER`: the size of the texture
+   *   atlas
    *
    * @param props the properties to use.
    * @post a TextEngine object or nullptr on failure; call GetError() for more
@@ -4168,14 +4177,16 @@
    */
   GPUTextEngine(PropertiesRef props);
 
-  ~GPUTextEngine() { Destroy(); }
+  GPUTextEngine(const GPUTextEngine&) = delete;
+
+  GPUTextEngine& operator=(const GPUTextEngine&) = delete;
+
+  ~GPUTextEngine() final { Destroy(); }
 
   /**
    * Sets the winding order of the vertices returned by Text.GetGPUDrawData for
    * a particular GPU text engine.
    *
-   * @param engine a TextEngine object created with
-   *               GPUTextEngine.GPUTextEngine().
    * @param winding the new winding order option.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4185,14 +4196,12 @@
    *
    * @sa GPUTextEngine.GetGPUWinding
    */
-  static void SetGPUWinding(TextEngineRef engine, GPUTextEngineWinding winding);
+  void SetGPUWinding(GPUTextEngineWinding winding);
 
   /**
    * Get the winding order of the vertices returned by Text.GetGPUDrawData for a
    * particular GPU text engine
    *
-   * @param engine a TextEngine object created with
-   *               GPUTextEngine.GPUTextEngine().
    * @returns the winding order used by the GPU text engine or
    *          GPU_TEXTENGINE_WINDING_INVALID in case of error.
    *
@@ -4203,7 +4212,7 @@
    *
    * @sa GPUTextEngine.SetGPUWinding
    */
-  GPUTextEngineWinding GetGPUWinding(TextEngineRef engine) const;
+  GPUTextEngineWinding GetGPUWinding() const;
 
   /**
    * Destroy a text engine created for drawing text with the SDL GPU API.
@@ -4211,9 +4220,6 @@
    * All text created by this engine should be destroyed before calling this
    * function.
    *
-   * @param engine a TextEngine object created with
-   *               GPUTextEngine.GPUTextEngine().
-   *
    * @threadsafety This function should be called on the thread that created the
    *               engine.
    *
@@ -4221,7 +4227,7 @@
    *
    * @sa GPUTextEngine.GPUTextEngine
    */
-  void Destroy();
+  void Destroy() final;
 };
 
 /**
@@ -4311,8 +4317,6 @@
    *               nullptr.
    * @param font the font to render with.
    * @param text the text to use, in UTF-8 encoding.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @post a Text object or nullptr on failure; call GetError() for more
    *       information.
    *
@@ -4382,11 +4386,8 @@
    * `text` must have been created using a TextEngine from
    * SurfaceTextEngine.SurfaceTextEngine().
    *
-   * @param text the text to draw.
-   * @param x the x coordinate in pixels, positive from the left edge towards
-   *          the right.
-   * @param y the y coordinate in pixels, positive from the top edge towards the
-   *          bottom.
+   * @param p the x,y coordinates in pixels, positive from the top-left edge
+   *          towards the bottom-right.
    * @param surface the surface to draw on.
    * @throws Error on failure.
    *
@@ -4407,11 +4408,8 @@
    * RendererTextEngine.RendererTextEngine(), and will draw using the renderer
    * passed to that function.
    *
-   * @param text the text to draw.
-   * @param x the x coordinate in pixels, positive from the left edge towards
-   *          the right.
-   * @param y the y coordinate in pixels, positive from the top edge towards the
-   *          bottom.
+   * @param p the x,y coordinates in pixels, positive from the top-left edge
+   *          towards the bottom-right.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4570,7 +4568,7 @@
    * This returns false if SDL_ttf isn't built with HarfBuzz support.
    *
    * @param script an [ISO 15924
-   *               code](https://unicode.org/iso15924/iso15924-codes.html) .
+   *               code](https://unicode.org/iso15924/iso15924-codes.html).
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4605,11 +4603,7 @@
    *
    * The default text color is white (255, 255, 255, 255).
    *
-   * @param text the Text to modify.
-   * @param r the red color value in the range of 0-255.
-   * @param g the green color value in the range of 0-255.
-   * @param b the blue color value in the range of 0-255.
-   * @param a the alpha value in the range of 0-255.
+   * @param c the color values in the range of 0-255.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4627,11 +4621,7 @@
    *
    * The default text color is white (1.0f, 1.0f, 1.0f, 1.0f).
    *
-   * @param text the Text to modify.
-   * @param r the red color value, normally in the range of 0-1.
-   * @param g the green color value, normally in the range of 0-1.
-   * @param b the blue color value, normally in the range of 0-1.
-   * @param a the alpha value in the range of 0-1.
+   * @param c the color value, normally in the range of 0-1.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4670,14 +4660,7 @@
   /**
    * Get the color of a text object.
    *
-   * @param r a pointer filled in with the red color value in the range of
-   *          0-255, may be nullptr.
-   * @param g a pointer filled in with the green color value in the range of
-   *          0-255, may be nullptr.
-   * @param b a pointer filled in with the blue color value in the range of
-   *          0-255, may be nullptr.
-   * @param a a pointer filled in with the alpha value in the range of 0-255,
-   *          may be nullptr.
+   * @return The color in the range of 0-255 on success.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4688,7 +4671,7 @@
    * @sa Text.GetColorFloat
    * @sa Text.SetColor
    */
-  Color GetColor();
+  Color GetColor() const;
 
   /**
    * Get the color of a text object.
@@ -4716,14 +4699,7 @@
   /**
    * Get the color of a text object.
    *
-   * @param r a pointer filled in with the red color value, normally in the
-   *          range of 0-1, may be nullptr.
-   * @param g a pointer filled in with the green color value, normally in the
-   *          range of 0-1, may be nullptr.
-   * @param b a pointer filled in with the blue color value, normally in the
-   *          range of 0-1, may be nullptr.
-   * @param a a pointer filled in with the alpha value in the range of 0-1, may
-   *          be nullptr.
+   * @return The color in the range of 0-1 on success.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4734,7 +4710,7 @@
    * @sa Text.GetColor
    * @sa Text.SetColorFloat
    */
-  FColor GetColorFloat();
+  FColor GetColorFloat() const;
 
   /**
    * Set the position of a text object.
@@ -4744,8 +4720,8 @@
    *
    * This function may cause the internal text representation to be rebuilt.
    *
-   * @param x the x offset of the upper left corner of this text in pixels.
-   * @param y the y offset of the upper left corner of this text in pixels.
+   * @param p the x, y offset of the upper left corner of this text in pixels.
+   * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
    *               text.
@@ -4763,6 +4739,7 @@
    *          this text in pixels, may be nullptr.
    * @param y a pointer filled in with the y offset of the upper left corner of
    *          this text in pixels, may be nullptr.
+   * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
    *               text.
@@ -4776,10 +4753,9 @@
   /**
    * Get the position of a text object.
    *
-   * @param x a pointer filled in with the x offset of the upper left corner of
-   *          this text in pixels, may be nullptr.
-   * @param y a pointer filled in with the y offset of the upper left corner of
-   *          this text in pixels, may be nullptr.
+   * @returns a Point with the offset of the upper left corner of this text in
+   *          pixels on success.
+   * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
    *               text.
@@ -4788,7 +4764,7 @@
    *
    * @sa Text.SetPosition
    */
-  Point GetPosition();
+  Point GetPosition() const;
 
   /**
    * Set whether wrapping is enabled on a text object.
@@ -4811,9 +4787,8 @@
   /**
    * Get whether wrapping is enabled on a text object.
    *
-   * @param text the Text to query.
-   * @param wrap_width a pointer filled in with the maximum width in pixels or 0
-   *                   if the text is being wrapped on newline characters.
+   * @returns an int with the maximum width in pixels or 0 if the text is being
+   *          wrapped on newline characters on success.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4868,10 +4843,7 @@
    *
    * This function may cause the internal text representation to be rebuilt.
    *
-   * @param text the Text to modify.
-   * @param string the UTF-8 text to use, may be nullptr.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
+   * @param string the UTF-8 text to use.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4890,14 +4862,11 @@
    *
    * This function may cause the internal text representation to be rebuilt.
    *
-   * @param text the Text to modify.
    * @param offset the offset, in bytes, from the beginning of the string if >=
    *               0, the offset from the end of the string if < 0. Note that
    *               this does not do UTF-8 validation, so you should only insert
    *               at UTF-8 sequence boundaries.
    * @param string the UTF-8 text to insert.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4916,10 +4885,7 @@
    *
    * This function may cause the internal text representation to be rebuilt.
    *
-   * @param text the Text to modify.
    * @param string the UTF-8 text to insert.
-   * @param length the length of the text, in bytes, or 0 for null terminated
-   *               text.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -4982,18 +4948,17 @@
    * The size of the text may change when the font or font style and size
    * change.
    *
-   * @param w a pointer filled in with the width of the text, in pixels, may be
-   *          nullptr.
-   * @param h a pointer filled in with the height of the text, in pixels, may be
-   *          nullptr.
+   * @returns Point containing the size on x and y axis on success.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
    *               text.
    *
    * @since This function is available since SDL_ttf 3.0.0.
+   *
+   * @sa GetSize(int*,int*) if might want only one of the coordinates
    */
-  Point GetSize();
+  Point GetSize() const;
 
   /**
    * Get the substring of a text object that surrounds a text offset.
@@ -5016,14 +4981,32 @@
    */
   void GetSubString(int offset, SubString* substring) const;
 
-  SubStringIterator begin() const { static_assert(false, "Not implemented"); }
+  /**
+   * Get iterator to first substring
+   */
+  SubStringIterator begin() const;
 
-  SubStringIterator end() const { static_assert(false, "Not implemented"); }
+  /**
+   * Get iterator to one past last substring
+   */
+  SubStringIterator end() const;
 
-  SubStringIterator GetSubStringForLine(int line) const
-  {
-    static_assert(false, "Not implemented");
-  }
+  /**
+   * Get iterator to substring of a text object that contains the given line.
+   *
+   * If `line` is greater than or equal to `text->num_lines` this will return an
+   * iterator equal to end().
+   *
+   * @param line a zero-based line index, in the range [0 .. text->num_lines-1].
+   * @returns iterator on success or false on failure; call GetError() for more
+   *          information.
+   *
+   * @threadsafety This function should be called on the thread that created the
+   *               text.
+   *
+   * @since This function is available since SDL_ttf 3.0.0.
+   */
+  SubStringIterator GetSubStringForLine(int line) const;
 
   /**
    * Get the substring of a text object that contains the given line.
@@ -5045,9 +5028,20 @@
    */
   void GetSubStringForLine(int line, SubString* substring) const;
 
+  /**
+   * Get all substrings of a text object.
+   *
+   * @returns a nullptr terminated array of substring pointers or nullptr on
+   *          failure; call GetError() for more information.
+   *
+   * @threadsafety This function should be called on the thread that created the
+   *               text.
+   *
+   * @since This function is available since SDL_ttf 3.0.0.
+   */
   OwnArray<SubString*> GetSubStrings() const
   {
-    static_assert(false, "Not implemented");
+    return GetSubStringsForRange(0);
   }
 
   /**
@@ -5056,8 +5050,6 @@
    * @param offset a byte offset into the text string.
    * @param length the length of the range being queried, in bytes, or -1 for
    *               the remainder of the string.
-   * @param count a pointer filled in with the number of substrings returned,
-   *              may be nullptr.
    * @returns a nullptr terminated array of substring pointers or nullptr on
    *          failure; call GetError() for more information. This is a single
    *          allocation that should be freed with free() when it is no longer
@@ -5068,22 +5060,31 @@
    *
    * @since This function is available since SDL_ttf 3.0.0.
    */
-  OwnArray<SubString*> GetSubStringsForRange(int offset, int length) const;
-
-  SubStringIterator GetSubStringForPoint(Point p) const
-  {
-    static_assert(false, "Not implemented");
-  }
+  OwnArray<SubString*> GetSubStringsForRange(int offset, int length = -1) const;
 
   /**
    * Get the portion of a text string that is closest to a point.
    *
    * This will return the closest substring of text to the given point.
    *
-   * @param text the Text to query.
-   * @param x the x coordinate relative to the left side of the text, may be
+   * @param p the coordinates relative to the top-left side of the text, may be
    *          outside the bounds of the text area.
-   * @param y the y coordinate relative to the top side of the text, may be
+   * @returns the iterator on success or false on failure; call GetError() for
+   *          more information.
+   *
+   * @threadsafety This function should be called on the thread that created the
+   *               text.
+   *
+   * @since This function is available since SDL_ttf 3.0.0.
+   */
+  SubStringIterator GetSubStringForPoint(Point p) const;
+
+  /**
+   * Get the portion of a text string that is closest to a point.
+   *
+   * This will return the closest substring of text to the given point.
+   *
+   * @param p the coordinates relative to the top-left side of the text, may be
    *          outside the bounds of the text area.
    * @param substring a pointer filled in with the closest substring of text to
    *                  the given point.
@@ -5103,6 +5104,8 @@
    * substring with the SUBSTRING_TEXT_START flag set.
    *
    * @param substring the SubString to query.
+   * @param previous a pointer filled in with the previous substring in the text
+   *                 object.
    * @throws Error on failure.
    *
    * @threadsafety This function should be called on the thread that created the
@@ -5150,10 +5153,10 @@
    * A copy of the UTF-8 string that this text object represents, useful for
    * layout, debugging and retrieving substring text
    */
-  const char* GetText() const { static_assert(false, "Not implemented"); }
+  const char* GetText() const { return m_resource->text; }
 
   /// The number of lines in the text, 0 if it's empty
-  int GetNumLines() const { static_assert(false, "Not implemented"); }
+  int GetNumLines() const { return m_resource->num_lines; }
 };
 
 /**
@@ -5228,53 +5231,75 @@
   constexpr operator TextRaw() const noexcept { return get(); }
 };
 
+/**
+ * Iterator for substrings
+ *
+ */
 class SubStringIterator
 {
   TextRef m_text;
 
   SubString m_subString;
 
-  constexpr SubStringIterator(TextRef text) {}
-
-public:
-  constexpr SubStringIterator() {}
-
-  constexpr operator bool() const { static_assert(false, "Not implemented"); }
-
-  constexpr const SubString& operator*() const
+  SubStringIterator(TextRef text)
+    : m_text(text)
+    , m_subString()
   {
-    static_assert(false, "Not implemented");
   }
 
-  constexpr const SubString* operator->() const
+public:
+  /// Default constructor.
+  SubStringIterator()
+    : SubStringIterator(TextRef{})
   {
-    static_assert(false, "Not implemented");
   }
 
+  /// True if pointing to valid SubString.
+  constexpr operator bool() const { return bool(m_text); }
+
+  /// Retrieve SubString
+  constexpr const SubString& operator*() const { return m_subString; }
+
+  /// Retrieve SubString.
+  constexpr const SubString* operator->() const { return &m_subString; }
+
+  /// Comparison.
   constexpr bool operator==(const SubStringIterator& other) const
   {
-    static_assert(false, "Not implemented");
+    return m_subString.offset == other.m_subString.offset;
   }
 
-  constexpr SubStringIterator& operator++()
+  /// Increment operator.
+  SubStringIterator& operator++()
   {
-    static_assert(false, "Not implemented");
+    m_text.GetNextSubString(m_subString, &m_subString);
+    return *this;
   }
 
-  constexpr SubStringIterator operator++(int)
+  /// Increment operator.
+  SubStringIterator operator++(int)
   {
-    static_assert(false, "Not implemented");
+    auto curr = *this;
+    m_text.GetNextSubString(m_subString, &m_subString);
+    return curr;
   }
 
-  constexpr SubStringIterator& operator--()
+  /// Decrement operator.
+  SubStringIterator& operator--()
   {
-    static_assert(false, "Not implemented");
+    m_text.GetPreviousSubString(m_subString, &m_subString);
+    return *this;
   }
 
-  constexpr SubStringIterator operator--(int)
+  /// Decrement operator.
+  SubStringIterator operator--(int)
   {
-    static_assert(false, "Not implemented");
+    auto curr = *this;
+    m_text.GetPreviousSubString(m_subString, &m_subString);
+    return curr;
   }
+
+  friend class Text;
 };
 
 /**
@@ -5296,7 +5321,7 @@
 }
 
 inline SurfaceTextEngine::SurfaceTextEngine()
-  : T(TTF_CreateSurfaceTextEngine())
+  : TextEngine(TTF_CreateSurfaceTextEngine())
 {
 }
 
@@ -5307,10 +5332,8 @@
  * SurfaceTextEngine.SurfaceTextEngine().
  *
  * @param text the text to draw.
- * @param x the x coordinate in pixels, positive from the left edge towards the
- *          right.
- * @param y the y coordinate in pixels, positive from the top edge towards the
- *          bottom.
+ * @param p the (x, y) coordinate in pixels, positive from the left edge
+ *          towards the right and from the top edge towards the bottom.
  * @param surface the surface to draw on.
  * @throws Error on failure.
  *
@@ -5324,7 +5347,7 @@
  */
 inline void DrawSurfaceText(TextConstRef text, Point p, SurfaceRef surface)
 {
-  CheckError(TTF_DrawSurfaceText(text, p, surface));
+  CheckError(TTF_DrawSurfaceText(text, p.x, p.y, surface));
 }
 
 inline void Text::DrawSurface(Point p, SurfaceRef surface) const
@@ -5380,12 +5403,12 @@
 }
 
 inline RendererTextEngine::RendererTextEngine(RendererRef renderer)
-  : T(TTF_CreateRendererTextEngine(renderer))
+  : TextEngine(TTF_CreateRendererTextEngine(renderer))
 {
 }
 
 inline RendererTextEngine::RendererTextEngine(PropertiesRef props)
-  : T(TTF_CreateRendererTextEngineWithProperties(props))
+  : TextEngine(TTF_CreateRendererTextEngineWithProperties(props))
 {
 }
 
@@ -5395,10 +5418,10 @@
  *
  * These are the supported properties:
  *
- * - `prop::RendererTextEngine.RENDERER`: the renderer to use for creating
- *   textures and drawing text
- * - `prop::RendererTextEngine.ATLAS_TEXTURE_SIZE`: the size of the texture
- *   atlas
+ * - `prop::RendererTextEngine.RENDERER_POINTER`: the renderer to use for
+ *   creating textures and drawing text
+ * - `prop::RendererTextEngine.ATLAS_TEXTURE_SIZE_NUMBER`: the size of the
+ *   texture atlas
  *
  * @param props the properties to use.
  * @returns a TextEngine object or nullptr on failure; call GetError() for more
@@ -5423,9 +5446,9 @@
 
 #if SDL_TTF_VERSION_ATLEAST(3, 2, 2)
 
-constexpr auto RENDERER = TTF_PROP_RENDERER_TEXT_ENGINE_RENDERER;
+constexpr auto RENDERER_POINTER = TTF_PROP_RENDERER_TEXT_ENGINE_RENDERER;
 
-constexpr auto ATLAS_TEXTURE_SIZE =
+constexpr auto ATLAS_TEXTURE_SIZE_NUMBER =
   TTF_PROP_RENDERER_TEXT_ENGINE_ATLAS_TEXTURE_SIZE;
 
 #endif // SDL_TTF_VERSION_ATLEAST(3, 2, 2)
@@ -5440,10 +5463,8 @@
  * passed to that function.
  *
  * @param text the text to draw.
- * @param x the x coordinate in pixels, positive from the left edge towards the
- *          right.
- * @param y the y coordinate in pixels, positive from the top edge towards the
- *          bottom.
+ * @param p the (x, y) coordinate in pixels, positive from the left edge
+ *          towards the right and from the top edge towards the bottom.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -5452,11 +5473,11 @@
  * @since This function is available since SDL_ttf 3.0.0.
  *
  * @sa RendererTextEngine.RendererTextEngine
- * @sa Text.Text
+ * @sa GPUTextEngine.GPUTextEngine
  */
 inline void DrawRendererText(TextConstRef text, FPoint p)
 {
-  CheckError(TTF_DrawRendererText(text, p));
+  CheckError(TTF_DrawRendererText(text, p.x, p.y));
 }
 
 inline void Text::DrawRenderer(FPoint p) const
@@ -5512,12 +5533,12 @@
 }
 
 inline GPUTextEngine::GPUTextEngine(GPUDeviceRef device)
-  : T(TTF_CreateGPUTextEngine(device))
+  : TextEngine(TTF_CreateGPUTextEngine(device))
 {
 }
 
 inline GPUTextEngine::GPUTextEngine(PropertiesRef props)
-  : T(TTF_CreateGPUTextEngineWithProperties(props))
+  : TextEngine(TTF_CreateGPUTextEngineWithProperties(props))
 {
 }
 
@@ -5527,9 +5548,10 @@
  *
  * These are the supported properties:
  *
- * - `prop::GpuTextEngine.DEVICE`: the GPUDevice to use for creating textures
- *   and drawing text.
- * - `prop::GpuTextEngine.ATLAS_TEXTURE_SIZE`: the size of the texture atlas
+ * - `prop::GpuTextEngine.DEVICE_POINTER`: the GPUDevice to use for creating
+ *   textures and drawing text.
+ * - `prop::GpuTextEngine.ATLAS_TEXTURE_SIZE_NUMBER`: the size of the texture
+ *   atlas
  *
  * @param props the properties to use.
  * @returns a TextEngine object or nullptr on failure; call GetError() for more
@@ -5553,9 +5575,10 @@
 
 #if SDL_TTF_VERSION_ATLEAST(3, 2, 2)
 
-constexpr auto DEVICE = TTF_PROP_GPU_TEXT_ENGINE_DEVICE;
+constexpr auto DEVICE_POINTER = TTF_PROP_GPU_TEXT_ENGINE_DEVICE;
 
-constexpr auto ATLAS_TEXTURE_SIZE = TTF_PROP_GPU_TEXT_ENGINE_ATLAS_TEXTURE_SIZE;
+constexpr auto ATLAS_TEXTURE_SIZE_NUMBER =
+  TTF_PROP_GPU_TEXT_ENGINE_ATLAS_TEXTURE_SIZE;
 
 #endif // SDL_TTF_VERSION_ATLEAST(3, 2, 2)
 
@@ -5585,7 +5608,7 @@
  * @since This function is available since SDL_ttf 3.0.0.
  *
  * @sa GPUTextEngine.GPUTextEngine
- * @sa Text.Text
+ * @sa GPUTextEngine.GPUTextEngine
  */
 inline GPUAtlasDrawSequence* GetGPUTextDrawData(TextConstRef text)
 {
@@ -5619,16 +5642,56 @@
 
 inline void GPUTextEngine::Destroy() { DestroyGPUTextEngine(release()); }
 
-inline void GPUTextEngine::SetGPUWinding(TextEngineRef engine,
-                                         GPUTextEngineWinding winding)
+/**
+ * Sets the winding order of the vertices returned by Text.GetGPUDrawData
+ * for a particular GPU text engine.
+ *
+ * @param engine a TextEngine object created with
+ *               GPUTextEngine.GPUTextEngine().
+ * @param winding the new winding order option.
+ *
+ * @threadsafety This function should be called on the thread that created the
+ *               engine.
+ *
+ * @since This function is available since SDL_ttf 3.0.0.
+ *
+ * @sa GPUTextEngine.GetGPUWinding
+ */
+inline void SetGPUTextEngineWinding(TextEngineRef engine,
+                                    GPUTextEngineWinding winding)
+{
+  TTF_SetGPUTextEngineWinding(engine, winding);
+}
+
+inline void GPUTextEngine::SetGPUWinding(GPUTextEngineWinding winding)
+{
+  SDL::SetGPUTextEngineWinding(get(), winding);
+}
+
+/**
+ * Get the winding order of the vertices returned by Text.GetGPUDrawData
+ * for a particular GPU text engine
+ *
+ * @param engine a TextEngine object created with
+ *               GPUTextEngine.GPUTextEngine().
+ * @returns the winding order used by the GPU text engine or
+ *          GPU_TEXTENGINE_WINDING_INVALID in case of error.
+ *
+ * @threadsafety This function should be called on the thread that created the
+ *               engine.
+ *
+ * @since This function is available since SDL_ttf 3.0.0.
+ *
+ * @sa GPUTextEngine.SetGPUWinding
+ */
+inline GPUTextEngineWinding GetGPUTextEngineWinding(TextEngineRef engine)
 {
-  SDL::SetGPUTextEngineWinding(engine, winding);
+  return TTF_GetGPUTextEngineWinding(engine);
 }
 
-inline GPUTextEngineWinding GPUTextEngine::GetGPUWinding(
-  TextEngineRef engine) const
+inline GPUTextEngineWinding GPUTextEngine::GetGPUWinding() const
 {
-  return SDL::GetGPUTextEngineWinding(engine);
+  return SDL::GetGPUTextEngineWinding(get());
 }
 
 /**
@@ -5638,8 +5701,6 @@
  *               nullptr.
  * @param font the font to render with.
  * @param text the text to use, in UTF-8 encoding.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @returns a Text object or nullptr on failure; call GetError() for more
  *          information.
  *
@@ -5663,7 +5724,7 @@
 }
 
 inline Text::Text(TextEngineRef engine, FontRef font, std::string_view text)
-  : m_resource(TTF_CreateText(engine, font, text))
+  : m_resource(TTF_CreateText(engine, font, text.data(), text.size()))
 {
 }
 
@@ -5681,7 +5742,7 @@
  */
 inline PropertiesRef GetTextProperties(TextConstRef text)
 {
-  return CheckError(TTF_GetTextProperties(text));
+  return {CheckError(TTF_GetTextProperties(text))};
 }
 
 inline PropertiesRef Text::GetProperties() const
@@ -5786,7 +5847,7 @@
  */
 inline FontRef GetTextFont(TextConstRef text)
 {
-  return CheckError(TTF_GetTextFont(text));
+  return {CheckError(TTF_GetTextFont(text))};
 }
 
 inline FontRef Text::GetFont() const { return SDL::GetTextFont(m_resource); }
@@ -5846,7 +5907,7 @@
  *
  * @param text the text to modify.
  * @param script an [ISO 15924
- *               code](https://unicode.org/iso15924/iso15924-codes.html) .
+ *               code](https://unicode.org/iso15924/iso15924-codes.html).
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -5896,10 +5957,7 @@
  * The default text color is white (255, 255, 255, 255).
  *
  * @param text the Text to modify.
- * @param r the red color value in the range of 0-255.
- * @param g the green color value in the range of 0-255.
- * @param b the blue color value in the range of 0-255.
- * @param a the alpha value in the range of 0-255.
+ * @param c the color value in the range of 0-255.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -5912,7 +5970,7 @@
  */
 inline void SetTextColor(TextRef text, Color c)
 {
-  CheckError(TTF_SetTextColor(text, c));
+  CheckError(TTF_SetTextColor(text, c.r, c.g, c.b, c.a));
 }
 
 inline void Text::SetColor(Color c) { SDL::SetTextColor(m_resource, c); }
@@ -5923,10 +5981,7 @@
  * The default text color is white (1.0f, 1.0f, 1.0f, 1.0f).
  *
  * @param text the Text to modify.
- * @param r the red color value, normally in the range of 0-1.
- * @param g the green color value, normally in the range of 0-1.
- * @param b the blue color value, normally in the range of 0-1.
- * @param a the alpha value in the range of 0-1.
+ * @param c the color value, normally in the range of 0-1.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -5939,7 +5994,7 @@
  */
 inline void SetTextColorFloat(TextRef text, FColor c)
 {
-  CheckError(TTF_SetTextColorFloat(text, c));
+  CheckError(TTF_SetTextColorFloat(text, c.r, c.g, c.b, c.a));
 }
 
 inline void Text::SetColorFloat(FColor c)
@@ -5982,14 +6037,7 @@
  * Get the color of a text object.
  *
  * @param text the Text to query.
- * @param r a pointer filled in with the red color value in the range of 0-255,
- *          may be nullptr.
- * @param g a pointer filled in with the green color value in the range of
- *          0-255, may be nullptr.
- * @param b a pointer filled in with the blue color value in the range of 0-255,
- *          may be nullptr.
- * @param a a pointer filled in with the alpha value in the range of 0-255, may
- *          be nullptr.
+ * @return The color in the range of 0-255 on success.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -6002,7 +6050,9 @@
  */
 inline Color GetTextColor(TextRef text)
 {
-  static_assert(false, "Not implemented");
+  Color c;
+  GetTextColor(text, &c.r, &c.g, &c.b, &c.a);
+  return c;
 }
 
 inline void Text::GetColor(Uint8* r, Uint8* g, Uint8* b, Uint8* a) const
@@ -6010,7 +6060,7 @@
   SDL::GetTextColor(m_resource, r, g, b, a);
 }
 
-inline Color Text::GetColor() { return SDL::GetTextColor(m_resource); }
+inline Color Text::GetColor() const { return SDL::GetTextColor(m_resource); }
 
 /**
  * Get the color of a text object.
@@ -6047,14 +6097,7 @@
  * Get the color of a text object.
  *
  * @param text the Text to query.
- * @param r a pointer filled in with the red color value, normally in the range
- *          of 0-1, may be nullptr.
- * @param g a pointer filled in with the green color value, normally in the
- *          range of 0-1, may be nullptr.
- * @param b a pointer filled in with the blue color value, normally in the range
- *          of 0-1, may be nullptr.
- * @param a a pointer filled in with the alpha value in the range of 0-1, may be
- *          nullptr.
+ * @return The color in the range of 0-1 on success.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -6067,7 +6110,9 @@
  */
 inline FColor GetTextColorFloat(TextRef text)
 {
-  static_assert(false, "Not implemented");
+  FColor c;
+  GetTextColorFloat(text, &c.r, &c.g, &c.b, &c.a);
+  return c;
 }
 
 inline void Text::GetColorFloat(float* r, float* g, float* b, float* a) const
@@ -6075,7 +6120,7 @@
   SDL::GetTextColorFloat(m_resource, r, g, b, a);
 }
 
-inline FColor Text::GetColorFloat()
+inline FColor Text::GetColorFloat() const
 {
   return SDL::GetTextColorFloat(m_resource);
 }
@@ -6089,8 +6134,8 @@
  * This function may cause the internal text representation to be rebuilt.
  *
  * @param text the Text to modify.
- * @param x the x offset of the upper left corner of this text in pixels.
- * @param y the y offset of the upper left corner of this text in pixels.
+ * @param p the x, y offset of the upper left corner of this text in pixels.
+ * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
  *               text.
@@ -6101,7 +6146,7 @@
  */
 inline void SetTextPosition(TextRef text, const PointRaw& p)
 {
-  CheckError(TTF_SetTextPosition(text, p));
+  CheckError(TTF_SetTextPosition(text, p.x, p.y));
 }
 
 inline void Text::SetPosition(const PointRaw& p)
@@ -6117,6 +6162,7 @@
  *          this text in pixels, may be nullptr.
  * @param y a pointer filled in with the y offset of the upper left corner of
  *          this text in pixels, may be nullptr.
+ * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
  *               text.
@@ -6134,10 +6180,9 @@
  * Get the position of a text object.
  *
  * @param text the Text to query.
- * @param x a pointer filled in with the x offset of the upper left corner of
- *          this text in pixels, may be nullptr.
- * @param y a pointer filled in with the y offset of the upper left corner of
- *          this text in pixels, may be nullptr.
+ * @returns a Point with the offset of the upper left corner of this text in
+ *          pixels on success.
+ * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
  *               text.
@@ -6148,7 +6193,9 @@
  */
 inline Point GetTextPosition(TextRef text)
 {
-  static_assert(false, "Not implemented");
+  Point p;
+  GetTextPosition(text, &p.x, &p.y);
+  return p;
 }
 
 inline void Text::GetPosition(int* x, int* y) const
@@ -6156,7 +6203,10 @@
   SDL::GetTextPosition(m_resource, x, y);
 }
 
-inline Point Text::GetPosition() { return SDL::GetTextPosition(m_resource); }
+inline Point Text::GetPosition() const
+{
+  return SDL::GetTextPosition(m_resource);
+}
 
 /**
  * Set whether wrapping is enabled on a text object.
@@ -6189,8 +6239,8 @@
  * Get whether wrapping is enabled on a text object.
  *
  * @param text the Text to query.
- * @param wrap_width a pointer filled in with the maximum width in pixels or 0
- *                   if the text is being wrapped on newline characters.
+ * @returns maximum width in pixels or 0 if the text is being wrapped on newline
+ *          characters.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -6202,7 +6252,9 @@
  */
 inline int GetTextWrapWidth(TextConstRef text)
 {
-  return CheckError(TTF_GetTextWrapWidth(text));
+  int w;
+  CheckError(TTF_GetTextWrapWidth(text, &w));
+  return w;
 }
 
 inline int Text::GetWrapWidth() const
@@ -6270,8 +6322,6 @@
  *
  * @param text the Text to modify.
  * @param string the UTF-8 text to use, may be nullptr.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -6285,7 +6335,7 @@
  */
 inline void SetTextString(TextRef text, std::string_view string)
 {
-  CheckError(TTF_SetTextString(text, string));
+  CheckError(TTF_SetTextString(text, string.data(), string.size()));
 }
 
 inline void Text::SetString(std::string_view string)
@@ -6304,8 +6354,6 @@
  *               does not do UTF-8 validation, so you should only insert at
  *               UTF-8 sequence boundaries.
  * @param string the UTF-8 text to insert.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -6319,7 +6367,7 @@
  */
 inline void InsertTextString(TextRef text, int offset, std::string_view string)
 {
-  CheckError(TTF_InsertTextString(text, offset, string));
+  CheckError(TTF_InsertTextString(text, offset, string.data(), string.size()));
 }
 
 inline void Text::InsertString(int offset, std::string_view string)
@@ -6334,8 +6382,6 @@
  *
  * @param text the Text to modify.
  * @param string the UTF-8 text to insert.
- * @param length the length of the text, in bytes, or 0 for null terminated
- *               text.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -6349,7 +6395,7 @@
  */
 inline void AppendTextString(TextRef text, std::string_view string)
 {
-  CheckError(TTF_AppendTextString(text, string));
+  CheckError(TTF_AppendTextString(text, string.data(), string.size()));
 }
 
 inline void Text::AppendString(std::string_view string)
@@ -6418,10 +6464,8 @@
  * The size of the text may change when the font or font style and size change.
  *
  * @param text the Text to query.
- * @param w a pointer filled in with the width of the text, in pixels, may be
- *          nullptr.
- * @param h a pointer filled in with the height of the text, in pixels, may be
- *          nullptr.
+ * @returns Point containing the size on x and y axis on success or
+ *          std::nullopt on failure; call GetError() for more information.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -6431,7 +6475,9 @@
  */
 inline Point GetTextSize(TextRef text)
 {
-  static_assert(false, "Not implemented");
+  Point p;
+  GetTextSize(text, &p.x, &p.y);
+  return p;
 }
 
 inline void Text::GetSize(int* w, int* h) const
@@ -6439,7 +6485,7 @@
   SDL::GetTextSize(m_resource, w, h);
 }
 
-inline Point Text::GetSize() { return SDL::GetTextSize(m_resource); }
+inline Point Text::GetSize() const { return SDL::GetTextSize(m_resource); }
 
 /**
  * Get the substring of a text object that surrounds a text offset.
@@ -6511,12 +6557,8 @@
  * @param offset a byte offset into the text string.
  * @param length the length of the range being queried, in bytes, or -1 for the
  *               remainder of the string.
- * @param count a pointer filled in with the number of substrings returned, may
- *              be nullptr.
  * @returns a nullptr terminated array of substring pointers or nullptr on
- *          failure; call GetError() for more information. This is a single
- *          allocation that should be freed with free() when it is no longer
- *          needed.
+ *          failure; call GetError() for more information.
  *
  * @threadsafety This function should be called on the thread that created the
  *               text.
@@ -6527,7 +6569,9 @@
                                                       int offset,
                                                       int length)
 {
-  return TTF_GetTextSubStringsForRange(text, offset, length);
+  int count = 0;
+  auto data = TTF_GetTextSubStringsForRange(text, offset, length, &count);
+  return OwnArray<SubString*>{data, size_t(count)};
 }
 
 inline OwnArray<SubString*> Text::GetSubStringsForRange(int offset,
@@ -6542,9 +6586,7 @@
  * This will return the closest substring of text to the given point.
  *
  * @param text the Text to query.
- * @param x the x coordinate relative to the left side of the text, may be
- *          outside the bounds of the text area.
- * @param y the y coordinate relative to the top side of the text, may be
+ * @param p the coordinates relative to the top-left side of the text, may be
  *          outside the bounds of the text area.
  * @param substring a pointer filled in with the closest substring of text to
  *                  the given point.
@@ -6559,7 +6601,7 @@
                                      Point p,
                                      SubString* substring)
 {
-  CheckError(TTF_GetTextSubStringForPoint(text, p, substring));
+  CheckError(TTF_GetTextSubStringForPoint(text, p.x, p.y, substring));
 }
 
 inline void Text::GetSubStringForPoint(Point p, SubString* substring) const
@@ -6575,6 +6617,8 @@
  *
  * @param text the Text to query.
  * @param substring the SubString to query.
+ * @param previous a pointer filled in with the previous substring in the text
+ *                 object.
  * @throws Error on failure.
  *
  * @threadsafety This function should be called on the thread that created the
@@ -6653,7 +6697,7 @@
  *
  * @since This function is available since SDL_ttf 3.0.0.
  *
- * @sa Text.Text
+ * @sa RendererTextEngine.RendererTextEngine
  */
 inline void DestroyText(TextRaw text) { TTF_DestroyText(text); }
 
@@ -6689,4 +6733,6 @@
 
 } // namespace SDL
 
+#endif // defined(SDL3PP_ENABLE_TTF) || defined(SDL3PP_DOC)
+
 #endif /* SDL3PP_TTF_H_ */
