--- build/generated/SDL3pp_gpu.h
+++ include/SDL3pp/SDL3pp_gpu.h
@@ -12,7 +12,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryGPU Category GPU
+ * @defgroup CategoryGPU 3D Rendering and GPU Compute
  *
  * The GPU API offers a cross-platform way for apps to talk to modern graphics
  * hardware. It offers both 3D graphics and compute support, in the style of
@@ -138,7 +138,7 @@
  * limited hardware support, this API is probably not for you.
  *
  * Examples demonstrating proper usage of this API can be found
- * [here](https://github.com/TheSpydog/SDL_gpu_examples) .
+ * [here](https://github.com/TheSpydog/SDL_gpu_examples).
  *
  * ## Performance considerations
  *
@@ -1252,7 +1252,6 @@
    * @param first_slot the vertex buffer slot to begin binding from.
    * @param bindings an array of GPUBufferBinding structs containing vertex
    *                 buffers and offset values.
-   * @param num_bindings the number of bindings in the bindings array.
    *
    * @since This function is available since SDL 3.2.0.
    */
@@ -1283,8 +1282,6 @@
    * @param first_slot the vertex sampler slot to begin binding from.
    * @param texture_sampler_bindings an array of texture-sampler binding
    *                                 structs.
-   * @param num_bindings the number of texture-sampler pairs to bind from the
-   *                     array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1305,7 +1302,6 @@
    *
    * @param first_slot the vertex storage texture slot to begin binding from.
    * @param storage_textures an array of storage textures.
-   * @param num_bindings the number of storage texture to bind from the array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1325,7 +1321,6 @@
    *
    * @param first_slot the vertex storage buffer slot to begin binding from.
    * @param storage_buffers an array of buffers.
-   * @param num_bindings the number of buffers to bind from the array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1345,8 +1340,6 @@
    * @param first_slot the fragment sampler slot to begin binding from.
    * @param texture_sampler_bindings an array of texture-sampler binding
    *                                 structs.
-   * @param num_bindings the number of texture-sampler pairs to bind from the
-   *                     array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1367,7 +1360,6 @@
    *
    * @param first_slot the fragment storage texture slot to begin binding from.
    * @param storage_textures an array of storage textures.
-   * @param num_bindings the number of storage textures to bind from the array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1388,7 +1380,6 @@
    *
    * @param first_slot the fragment storage buffer slot to begin binding from.
    * @param storage_buffers an array of storage buffers.
-   * @param num_bindings the number of storage buffers to bind from the array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1554,8 +1545,6 @@
    * @param first_slot the compute sampler slot to begin binding from.
    * @param texture_sampler_bindings an array of texture-sampler binding
    *                                 structs.
-   * @param num_bindings the number of texture-sampler bindings to bind from the
-   *                     array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1576,7 +1565,6 @@
    *
    * @param first_slot the compute storage texture slot to begin binding from.
    * @param storage_textures an array of storage textures.
-   * @param num_bindings the number of storage textures to bind from the array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1596,7 +1584,6 @@
    *
    * @param first_slot the compute storage buffer slot to begin binding from.
    * @param storage_buffers an array of storage buffer binding structs.
-   * @param num_bindings the number of storage buffers to bind from the array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -2140,7 +2127,6 @@
    *
    * @param slot_index the vertex uniform slot to push data to.
    * @param data client data to write.
-   * @param length the length of the data to write.
    *
    * @since This function is available since SDL 3.2.0.
    */
@@ -2157,7 +2143,6 @@
    *
    * @param slot_index the fragment uniform slot to push data to.
    * @param data client data to write.
-   * @param length the length of the data to write.
    *
    * @since This function is available since SDL 3.2.0.
    */
@@ -2174,7 +2159,6 @@
    *
    * @param slot_index the uniform slot to push data to.
    * @param data client data to write.
-   * @param length the length of the data to write.
    *
    * @since This function is available since SDL 3.2.0.
    */
@@ -2201,8 +2185,6 @@
    *
    * @param color_target_infos an array of texture subresources with
    *                           corresponding clear values and load/store ops.
-   * @param num_color_targets the number of color targets in the
-   *                          color_target_infos array.
    * @param depth_stencil_target_info a texture subresource with corresponding
    *                                  clear value and load/store ops, may be
    *                                  nullptr.
@@ -2240,12 +2222,8 @@
    *
    * @param storage_texture_bindings an array of writeable storage texture
    *                                 binding structs.
-   * @param num_storage_texture_bindings the number of storage textures to bind
-   *                                     from the array.
    * @param storage_buffer_bindings an array of writeable storage buffer binding
    *                                structs.
-   * @param num_storage_buffer_bindings the number of storage buffers to bind
-   *                                    from the array.
    * @returns a compute pass handle.
    *
    * @since This function is available since SDL 3.2.0.
@@ -2317,12 +2295,11 @@
    * than the one that created the window.
    *
    * @param window a window that has been claimed.
-   * @param swapchain_texture a pointer filled in with a swapchain texture
-   *                          handle.
    * @param swapchain_texture_width a pointer filled in with the swapchain
    *                                texture width, may be nullptr.
    * @param swapchain_texture_height a pointer filled in with the swapchain
    *                                 texture height, may be nullptr.
+   * @returns a swapchain texture handle.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called from the thread that
@@ -2367,12 +2344,11 @@
    * another reading operation.
    *
    * @param window a window that has been claimed.
-   * @param swapchain_texture a pointer filled in with a swapchain texture
-   *                          handle.
    * @param swapchain_texture_width a pointer filled in with the swapchain
    *                                texture width, may be nullptr.
    * @param swapchain_texture_height a pointer filled in with the swapchain
    *                                 texture height, may be nullptr.
+   * @returns a swapchain texture handle.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called from the thread that
@@ -2573,9 +2549,6 @@
  * format is supported before using it. However, there are a few guaranteed
  * formats.
  *
- * FIXME: Check universal support for 32-bit component formats FIXME: Check
- * universal support for SIMULTANEOUS_READ_WRITE
- *
  * For SAMPLER usage, the following formats are universally supported:
  *
  * - R8G8B8A8_UNORM
@@ -4021,7 +3994,6 @@
    * @param wait_all if 0, wait for any fence to be signaled, if 1, wait for all
    *                 fences to be signaled.
    * @param fences an array of fences to wait on.
-   * @param num_fences the number of fences in the fences array.
    * @throws Error on failure.
    *
    * @since This function is available since SDL 3.2.0.
@@ -4083,6 +4055,8 @@
   bool TextureSupportsSampleCount(GPUTextureFormat format,
                                   GPUSampleCount sample_count);
 
+#ifdef SDL_PLATFORM_GDK
+
   /**
    * Call this to suspend GPU operation on Xbox when you receive the
    * EVENT_DID_ENTER_BACKGROUND event.
@@ -4108,6 +4082,8 @@
    * @sa AddEventWatch
    */
   void GDKResumeGPU();
+
+#endif /* SDL_PLATFORM_GDK */
 };
 
 /**
@@ -5155,6 +5131,17 @@
   return GPUDevice(props);
 }
 
+/**
+ * Properties for GPUDevice creation and information.
+ *
+ * The properties prefixed with `CREATE_` can be used in the properties object
+ * passed to CreateGPUDeviceWithProperties to specify the desired GPU device and
+ * driver features, while the remainder can be queried from an existing GPU
+ * device via GPUDevice.GetProperties.
+ *
+ * @sa CreateGPUDeviceWithProperties
+ * @sa GPUDevice.GetProperties
+ */
 namespace prop::GPUDevice {
 
 constexpr auto CREATE_DEBUGMODE_BOOLEAN =
@@ -5545,6 +5532,11 @@
 {
 }
 
+/**
+ * Properties for GPUComputePipeline creation.
+ *
+ * @sa GPUComputePipeline.CreateComputePipeline
+ */
 namespace prop::GPUComputePipeline {
 
 constexpr auto CREATE_NAME_STRING =
@@ -5594,6 +5586,11 @@
 {
 }
 
+/**
+ * Properties for GPUGraphicsPipeline creation.
+ *
+ * @sa GPUGraphicsPipeline.CreateGraphicsPipeline
+ */
 namespace prop::GPUGraphicsPipeline {
 
 constexpr auto CREATE_NAME_STRING =
@@ -5640,6 +5637,11 @@
 {
 }
 
+/**
+ * Properties for GPUSampler creation.
+ *
+ * @sa GPUDevice.CreateSampler
+ */
 namespace prop::GPUSampler {
 
 constexpr auto CREATE_NAME_STRING = SDL_PROP_GPU_SAMPLER_CREATE_NAME_STRING;
@@ -5734,6 +5736,11 @@
 {
 }
 
+/**
+ * Properties for GPUShader creation.
+ *
+ * @sa GPUDevice.CreateShader
+ */
 namespace prop::GPUShader {
 
 constexpr auto CREATE_NAME_STRING = SDL_PROP_GPU_SHADER_CREATE_NAME_STRING;
@@ -5816,6 +5823,11 @@
 {
 }
 
+/**
+ * Properties for GPUTexture creation.
+ *
+ * @sa GPUDevice.CreateTexture
+ */
 namespace prop::GPUTexture {
 
 constexpr auto CREATE_D3D12_CLEAR_R_FLOAT =
@@ -5904,6 +5916,11 @@
 {
 }
 
+/**
+ * Properties for GPUBuffer creation.
+ *
+ * @sa GPUDevice.CreateBuffer
+ */
 namespace prop::GPUBuffer {
 
 constexpr auto CREATE_NAME_STRING = SDL_PROP_GPU_BUFFER_CREATE_NAME_STRING;
@@ -5958,6 +5975,11 @@
 {
 }
 
+/**
+ * Properties for GPUTransferBuffer creation.
+ *
+ * @sa GPUDevice.CreateTransferBuffer
+ */
 namespace prop::GPUTransferBuffer {
 
 constexpr auto CREATE_NAME_STRING =
@@ -6307,7 +6329,6 @@
  * @param command_buffer a command buffer.
  * @param slot_index the vertex uniform slot to push data to.
  * @param data client data to write.
- * @param length the length of the data to write.
  *
  * @since This function is available since SDL 3.2.0.
  */
@@ -6316,7 +6337,7 @@
                                      SourceBytes data)
 {
   SDL_PushGPUVertexUniformData(
-    command_buffer, slot_index, data.data(), data.size_bytes());
+    command_buffer, slot_index, data.data(), narrowU32(data.size_bytes()));
 }
 
 inline void GPUCommandBuffer::PushVertexUniformData(Uint32 slot_index,
@@ -6338,7 +6359,6 @@
  * @param command_buffer a command buffer.
  * @param slot_index the fragment uniform slot to push data to.
  * @param data client data to write.
- * @param length the length of the data to write.
  *
  * @since This function is available since SDL 3.2.0.
  */
@@ -6347,7 +6367,7 @@
                                        SourceBytes data)
 {
   SDL_PushGPUFragmentUniformData(
-    command_buffer, slot_index, data.data(), data.size_bytes());
+    command_buffer, slot_index, data.data(), narrowU32(data.size_bytes()));
 }
 
 inline void GPUCommandBuffer::PushFragmentUniformData(Uint32 slot_index,
@@ -6369,7 +6389,6 @@
  * @param command_buffer a command buffer.
  * @param slot_index the uniform slot to push data to.
  * @param data client data to write.
- * @param length the length of the data to write.
  *
  * @since This function is available since SDL 3.2.0.
  */
@@ -6378,7 +6397,7 @@
                                       SourceBytes data)
 {
   SDL_PushGPUComputeUniformData(
-    command_buffer, slot_index, data.data(), data.size_bytes());
+    command_buffer, slot_index, data.data(), narrowU32(data.size_bytes()));
 }
 
 inline void GPUCommandBuffer::PushComputeUniformData(Uint32 slot_index,
@@ -6410,8 +6429,6 @@
  * @param command_buffer a command buffer.
  * @param color_target_infos an array of texture subresources with corresponding
  *                           clear values and load/store ops.
- * @param num_color_targets the number of color targets in the
- *                          color_target_infos array.
  * @param depth_stencil_target_info a texture subresource with corresponding
  *                                  clear value and load/store ops, may be
  *                                  nullptr.
@@ -6426,8 +6443,10 @@
   std::span<const GPUColorTargetInfo> color_target_infos,
   OptionalRef<const GPUDepthStencilTargetInfo> depth_stencil_target_info)
 {
-  return SDL_BeginGPURenderPass(
-    command_buffer, color_target_infos, depth_stencil_target_info);
+  return SDL_BeginGPURenderPass(command_buffer,
+                                color_target_infos.data(),
+                                narrowU32(color_target_infos.size()),
+                                depth_stencil_target_info);
 }
 
 inline GPURenderPass GPUCommandBuffer::BeginRenderPass(
@@ -6488,7 +6507,7 @@
  */
 inline void SetGPUScissor(GPURenderPass render_pass, const RectRaw& scissor)
 {
-  SDL_SetGPUScissor(render_pass, scissor);
+  SDL_SetGPUScissor(render_pass, &scissor);
 }
 
 inline void GPURenderPass::SetScissor(const RectRaw& scissor)
@@ -6543,7 +6562,6 @@
  * @param first_slot the vertex buffer slot to begin binding from.
  * @param bindings an array of GPUBufferBinding structs containing vertex
  *                 buffers and offset values.
- * @param num_bindings the number of bindings in the bindings array.
  *
  * @since This function is available since SDL 3.2.0.
  */
@@ -6551,7 +6569,8 @@
                                  Uint32 first_slot,
                                  std::span<const GPUBufferBinding> bindings)
 {
-  SDL_BindGPUVertexBuffers(render_pass, first_slot, bindings);
+  SDL_BindGPUVertexBuffers(
+    render_pass, first_slot, bindings.data(), narrowU32(bindings.size()));
 }
 
 inline void GPURenderPass::BindVertexBuffers(
@@ -6596,8 +6615,6 @@
  * @param render_pass a render pass handle.
  * @param first_slot the vertex sampler slot to begin binding from.
  * @param texture_sampler_bindings an array of texture-sampler binding structs.
- * @param num_bindings the number of texture-sampler pairs to bind from the
- *                     array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -6608,7 +6625,10 @@
   Uint32 first_slot,
   std::span<const GPUTextureSamplerBinding> texture_sampler_bindings)
 {
-  SDL_BindGPUVertexSamplers(render_pass, first_slot, texture_sampler_bindings);
+  SDL_BindGPUVertexSamplers(render_pass,
+                            first_slot,
+                            texture_sampler_bindings.data(),
+                            narrowU32(texture_sampler_bindings.size()));
 }
 
 inline void GPURenderPass::BindVertexSamplers(
@@ -6631,7 +6651,6 @@
  * @param render_pass a render pass handle.
  * @param first_slot the vertex storage texture slot to begin binding from.
  * @param storage_textures an array of storage textures.
- * @param num_bindings the number of storage texture to bind from the array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -6642,7 +6661,10 @@
   Uint32 first_slot,
   SpanRef<const GPUTextureRaw> storage_textures)
 {
-  SDL_BindGPUVertexStorageTextures(render_pass, first_slot, storage_textures);
+  SDL_BindGPUVertexStorageTextures(render_pass,
+                                   first_slot,
+                                   storage_textures.data(),
+                                   narrowU32(storage_textures.size()));
 }
 
 inline void GPURenderPass::BindVertexStorageTextures(
@@ -6665,7 +6687,6 @@
  * @param render_pass a render pass handle.
  * @param first_slot the vertex storage buffer slot to begin binding from.
  * @param storage_buffers an array of buffers.
- * @param num_bindings the number of buffers to bind from the array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -6676,7 +6697,10 @@
   Uint32 first_slot,
   SpanRef<const GPUBufferRaw> storage_buffers)
 {
-  SDL_BindGPUVertexStorageBuffers(render_pass, first_slot, storage_buffers);
+  SDL_BindGPUVertexStorageBuffers(render_pass,
+                                  first_slot,
+                                  storage_buffers.data(),
+                                  narrowU32(storage_buffers.size()));
 }
 
 inline void GPURenderPass::BindVertexStorageBuffers(
@@ -6698,8 +6722,6 @@
  * @param render_pass a render pass handle.
  * @param first_slot the fragment sampler slot to begin binding from.
  * @param texture_sampler_bindings an array of texture-sampler binding structs.
- * @param num_bindings the number of texture-sampler pairs to bind from the
- *                     array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -6710,8 +6732,10 @@
   Uint32 first_slot,
   std::span<const GPUTextureSamplerBinding> texture_sampler_bindings)
 {
-  SDL_BindGPUFragmentSamplers(
-    render_pass, first_slot, texture_sampler_bindings);
+  SDL_BindGPUFragmentSamplers(render_pass,
+                              first_slot,
+                              texture_sampler_bindings.data(),
+                              narrowU32(texture_sampler_bindings.size()));
 }
 
 inline void GPURenderPass::BindFragmentSamplers(
@@ -6734,7 +6758,6 @@
  * @param render_pass a render pass handle.
  * @param first_slot the fragment storage texture slot to begin binding from.
  * @param storage_textures an array of storage textures.
- * @param num_bindings the number of storage textures to bind from the array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -6745,7 +6768,10 @@
   Uint32 first_slot,
   SpanRef<const GPUTextureRaw> storage_textures)
 {
-  SDL_BindGPUFragmentStorageTextures(render_pass, first_slot, storage_textures);
+  SDL_BindGPUFragmentStorageTextures(render_pass,
+                                     first_slot,
+                                     storage_textures.data(),
+                                     narrowU32(storage_textures.size()));
 }
 
 inline void GPURenderPass::BindFragmentStorageTextures(
@@ -6768,7 +6794,6 @@
  * @param render_pass a render pass handle.
  * @param first_slot the fragment storage buffer slot to begin binding from.
  * @param storage_buffers an array of storage buffers.
- * @param num_bindings the number of storage buffers to bind from the array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -6779,7 +6804,10 @@
   Uint32 first_slot,
   SpanRef<const GPUBufferRaw> storage_buffers)
 {
-  SDL_BindGPUFragmentStorageBuffers(render_pass, first_slot, storage_buffers);
+  SDL_BindGPUFragmentStorageBuffers(render_pass,
+                                    first_slot,
+                                    storage_buffers.data(),
+                                    narrowU32(storage_buffers.size()));
 }
 
 inline void GPURenderPass::BindFragmentStorageBuffers(
@@ -6986,12 +7014,8 @@
  * @param command_buffer a command buffer.
  * @param storage_texture_bindings an array of writeable storage texture binding
  *                                 structs.
- * @param num_storage_texture_bindings the number of storage textures to bind
- *                                     from the array.
  * @param storage_buffer_bindings an array of writeable storage buffer binding
  *                                structs.
- * @param num_storage_buffer_bindings the number of storage buffers to bind from
- *                                    the array.
  * @returns a compute pass handle.
  *
  * @since This function is available since SDL 3.2.0.
@@ -7003,8 +7027,11 @@
   std::span<const GPUStorageTextureReadWriteBinding> storage_texture_bindings,
   std::span<const GPUStorageBufferReadWriteBinding> storage_buffer_bindings)
 {
-  return SDL_BeginGPUComputePass(
-    command_buffer, storage_texture_bindings, storage_buffer_bindings);
+  return SDL_BeginGPUComputePass(command_buffer,
+                                 storage_texture_bindings.data(),
+                                 narrowU32(storage_texture_bindings.size()),
+                                 storage_buffer_bindings.data(),
+                                 narrowU32(storage_buffer_bindings.size()));
 }
 
 inline GPUComputePass GPUCommandBuffer::BeginComputePass(
@@ -7045,8 +7072,6 @@
  * @param compute_pass a compute pass handle.
  * @param first_slot the compute sampler slot to begin binding from.
  * @param texture_sampler_bindings an array of texture-sampler binding structs.
- * @param num_bindings the number of texture-sampler bindings to bind from the
- *                     array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -7057,8 +7082,10 @@
   Uint32 first_slot,
   std::span<const GPUTextureSamplerBinding> texture_sampler_bindings)
 {
-  SDL_BindGPUComputeSamplers(
-    compute_pass, first_slot, texture_sampler_bindings);
+  SDL_BindGPUComputeSamplers(compute_pass,
+                             first_slot,
+                             texture_sampler_bindings.data(),
+                             narrowU32(texture_sampler_bindings.size()));
 }
 
 inline void GPUComputePass::BindSamplers(
@@ -7081,7 +7108,6 @@
  * @param compute_pass a compute pass handle.
  * @param first_slot the compute storage texture slot to begin binding from.
  * @param storage_textures an array of storage textures.
- * @param num_bindings the number of storage textures to bind from the array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -7092,7 +7118,10 @@
   Uint32 first_slot,
   SpanRef<const GPUTextureRaw> storage_textures)
 {
-  SDL_BindGPUComputeStorageTextures(compute_pass, first_slot, storage_textures);
+  SDL_BindGPUComputeStorageTextures(compute_pass,
+                                    first_slot,
+                                    storage_textures.data(),
+                                    narrowU32(storage_textures.size()));
 }
 
 inline void GPUComputePass::BindStorageTextures(
@@ -7115,7 +7144,6 @@
  * @param compute_pass a compute pass handle.
  * @param first_slot the compute storage buffer slot to begin binding from.
  * @param storage_buffers an array of storage buffer binding structs.
- * @param num_bindings the number of storage buffers to bind from the array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -7126,7 +7154,10 @@
   Uint32 first_slot,
   SpanRef<const GPUBufferRaw> storage_buffers)
 {
-  SDL_BindGPUComputeStorageBuffers(compute_pass, first_slot, storage_buffers);
+  SDL_BindGPUComputeStorageBuffers(compute_pass,
+                                   first_slot,
+                                   storage_buffers.data(),
+                                   narrowU32(storage_buffers.size()));
 }
 
 inline void GPUComputePass::BindStorageBuffers(
@@ -7658,7 +7689,7 @@
  * supported via GPUDevice.WindowSupportsPresentMode /
  * GPUDevice.WindowSupportsSwapchainComposition prior to calling this function.
  *
- * GPU_PRESENTMODE_VSYNC with GPU_SWAPCHAINCOMPOSITION_SDR is always supported.
+ * GPU_PRESENTMODE_VSYNC with GPU_SWAPCHAINCOMPOSITION_SDR are always supported.
  *
  * @param device a GPU context.
  * @param window an Window that has been claimed.
@@ -7774,11 +7805,11 @@
  *
  * @param command_buffer a command buffer.
  * @param window a window that has been claimed.
- * @param swapchain_texture a pointer filled in with a swapchain texture handle.
  * @param swapchain_texture_width a pointer filled in with the swapchain texture
  *                                width, may be nullptr.
  * @param swapchain_texture_height a pointer filled in with the swapchain
  *                                 texture height, may be nullptr.
+ * @returns a swapchain texture handle.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called from the thread that
@@ -7801,8 +7832,13 @@
   Uint32* swapchain_texture_width = nullptr,
   Uint32* swapchain_texture_height = nullptr)
 {
-  return CheckError(SDL_AcquireGPUSwapchainTexture(
-    command_buffer, window, swapchain_texture_width, swapchain_texture_height));
+  GPUTextureRaw texture;
+  CheckError(SDL_AcquireGPUSwapchainTexture(command_buffer,
+                                            window,
+                                            &texture,
+                                            swapchain_texture_width,
+                                            swapchain_texture_height));
+  return texture;
 }
 
 inline GPUTexture GPUCommandBuffer::AcquireSwapchainTexture(
@@ -7866,11 +7902,11 @@
  *
  * @param command_buffer a command buffer.
  * @param window a window that has been claimed.
- * @param swapchain_texture a pointer filled in with a swapchain texture handle.
  * @param swapchain_texture_width a pointer filled in with the swapchain texture
  *                                width, may be nullptr.
  * @param swapchain_texture_height a pointer filled in with the swapchain
  *                                 texture height, may be nullptr.
+ * @returns a swapchain texture handle.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called from the thread that
@@ -7888,8 +7924,13 @@
   Uint32* swapchain_texture_width = nullptr,
   Uint32* swapchain_texture_height = nullptr)
 {
-  return CheckError(SDL_WaitAndAcquireGPUSwapchainTexture(
-    command_buffer, window, swapchain_texture_width, swapchain_texture_height));
+  GPUTextureRaw texture;
+  CheckError(SDL_WaitAndAcquireGPUSwapchainTexture(command_buffer,
+                                                   window,
+                                                   &texture,
+                                                   swapchain_texture_width,
+                                                   swapchain_texture_height));
+  return texture;
 }
 
 inline GPUTexture GPUCommandBuffer::WaitAndAcquireSwapchainTexture(
@@ -8023,7 +8064,6 @@
  * @param wait_all if 0, wait for any fence to be signaled, if 1, wait for all
  *                 fences to be signaled.
  * @param fences an array of fences to wait on.
- * @param num_fences the number of fences in the fences array.
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
@@ -8035,7 +8075,8 @@
                              bool wait_all,
                              std::span<GPUFence* const> fences)
 {
-  CheckError(SDL_WaitForGPUFences(device, wait_all, fences));
+  CheckError(SDL_WaitForGPUFences(
+    device, wait_all, fences.data(), narrowU32(fences.size())));
 }
 
 inline void GPUDevice::WaitForFences(bool wait_all,
@@ -8203,6 +8244,8 @@
 
 #endif // SDL_VERSION_ATLEAST(3, 4, 0)
 
+#ifdef SDL_PLATFORM_GDK
+
 /**
  * Call this to suspend GPU operation on Xbox when you receive the
  * EVENT_DID_ENTER_BACKGROUND event.
@@ -8237,6 +8280,8 @@
 
 inline void GPUDevice::GDKResumeGPU() { SDL::GDKResumeGPU(get()); }
 
+#endif /* SDL_PLATFORM_GDK */
+
 /// @}
 
 } // namespace SDL
