--- src/generated/SDL3pp_camera.h
+++ include/SDL3pp/SDL3pp_camera.h
@@ -11,7 +11,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryCamera Category Camera
+ * @defgroup CategoryCamera Camera Support
  *
  * Video capture for the SDL library.
  *
@@ -211,9 +211,9 @@
    *
    * @param instance_id the camera device instance ID.
    * @param spec the desired format for data the device will provide. Can be
-   *             nullptr.
-   * @post an Camera object or nullptr on failure; call GetError() for more
-   *       information.
+   *             std::nullopt.
+   * @post an Camera object on success..
+   * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
@@ -352,8 +352,7 @@
    * (or EVENT_CAMERA_DEVICE_DENIED) event, or poll Camera.GetPermissionState()
    * occasionally until it returns non-zero.
    *
-   * @param spec the CameraSpec to be initialized by this function.
-   * @throws Error on failure.
+   * @returns the CameraSpec or std::nullopt if waiting for user approval.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
@@ -394,8 +393,8 @@
    *
    * @param timestampNS a pointer filled in with the frame's timestamp, or 0 on
    *                    error. Can be nullptr.
-   * @returns a new frame of video on success.
-   * @throws Error on failure.
+   * @returns a new frame of video on success, nullptr if none is currently
+   *          available.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
@@ -403,7 +402,7 @@
    *
    * @sa Camera.ReleaseFrame
    */
-  Surface AcquireFrame(Uint64* timestampNS);
+  Surface AcquireFrame(Uint64* timestampNS = nullptr);
 
   /**
    * Release a frame of video acquired from a camera.
@@ -544,11 +543,8 @@
 /**
  * Get a list of currently connected camera devices.
  *
- * @param count a pointer filled in with the number of cameras returned, may be
- *              nullptr.
  * @returns a 0 terminated array of camera instance IDs or nullptr on failure;
- *          call GetError() for more information. This should be freed with
- *          free() when it is no longer needed.
+ *          call GetError() for more information.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -556,7 +552,12 @@
  *
  * @sa Camera.Camera
  */
-inline OwnArray<CameraID> GetCameras() { return SDL_GetCameras(); }
+inline OwnArray<CameraID> GetCameras()
+{
+  int count;
+  auto data = SDL_GetCameras(&count);
+  return OwnArray<CameraID>(data, count);
+}
 
 /**
  * Get the list of native formats/sizes a camera supports.
@@ -580,12 +581,8 @@
  * warning popup.
  *
  * @param instance_id the camera device instance ID.
- * @param count a pointer filled in with the number of elements in the list, may
- *              be nullptr.
  * @returns a nullptr terminated array of pointers to CameraSpec or nullptr on
- *          failure; call GetError() for more information. This is a single
- *          allocation that should be freed with free() when it is no longer
- *          needed.
+ *          failure; call GetError() for more information.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -596,7 +593,9 @@
  */
 inline OwnArray<CameraSpec*> GetCameraSupportedFormats(CameraID instance_id)
 {
-  return SDL_GetCameraSupportedFormats(instance_id);
+  int count;
+  auto data = SDL_GetCameraSupportedFormats(instance_id, &count);
+  return OwnArray<CameraSpec*>(data, count);
 }
 
 /**
@@ -672,9 +671,9 @@
  *
  * @param instance_id the camera device instance ID.
  * @param spec the desired format for data the device will provide. Can be
- *             nullptr.
- * @returns an Camera object or nullptr on failure; call GetError() for more
- *          information.
+ *             std::nullopt.
+ * @returns an Camera object on success.
+ * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -763,7 +762,7 @@
  */
 inline PropertiesRef GetCameraProperties(CameraParam camera)
 {
-  return CheckError(SDL_GetCameraProperties(camera));
+  return {CheckError(SDL_GetCameraProperties(camera))};
 }
 
 inline PropertiesRef Camera::GetProperties()
@@ -784,8 +783,7 @@
  * occasionally until it returns non-zero.
  *
  * @param camera opened camera device.
- * @param spec the CameraSpec to be initialized by this function.
- * @throws Error on failure.
+ * @returns the CameraSpec or std::nullopt if waiting for user approval.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -795,7 +793,8 @@
  */
 inline std::optional<CameraSpec> GetCameraFormat(CameraParam camera)
 {
-  return CheckError(SDL_GetCameraFormat(camera));
+  if (CameraSpec spec; SDL_GetCameraFormat(camera, &spec)) return spec;
+  return std::nullopt;
 }
 
 inline std::optional<CameraSpec> Camera::GetFormat()
@@ -834,8 +833,8 @@
  * @param camera opened camera device.
  * @param timestampNS a pointer filled in with the frame's timestamp, or 0 on
  *                    error. Can be nullptr.
- * @returns a new frame of video on success.
- * @throws Error on failure.
+ * @returns a new frame of video on success, nullptr if none is currently
+ *          available.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -843,9 +842,10 @@
  *
  * @sa Camera.ReleaseFrame
  */
-inline Surface AcquireCameraFrame(CameraParam camera, Uint64* timestampNS)
+inline Surface AcquireCameraFrame(CameraParam camera,
+                                  Uint64* timestampNS = nullptr)
 {
-  return CheckError(SDL_AcquireCameraFrame(camera, timestampNS));
+  return Surface::Borrow(SDL_AcquireCameraFrame(camera, timestampNS));
 }
 
 inline Surface Camera::AcquireFrame(Uint64* timestampNS)
