--- src/generated/SDL3pp_iostream.h
+++ include/SDL3pp/SDL3pp_iostream.h
@@ -8,7 +8,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryIOStream Category IOStream
+ * @defgroup CategoryIOStream I/O Streams
  *
  * SDL provides an abstract interface for reading and writing data streams. It
  * offers implementations for files, memory, etc, and the app can provide their
@@ -266,10 +266,9 @@
    * - `prop::IOStream.MEMORY_SIZE_NUMBER`: this will be the `size` parameter
    *   that was passed to this function.
    *
-   * @param mem a pointer to a buffer to feed an IOStream stream.
-   * @param size the buffer size, in bytes.
-   * @returns a pointer to a new IOStream structure or nullptr on failure; call
-   *          GetError() for more information.
+   * @param mem a buffer to feed an IOStream stream.
+   * @returns a valid IOStream on success.
+   * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
@@ -309,10 +308,9 @@
    * - `prop::IOStream.MEMORY_SIZE_NUMBER`: this will be the `size` parameter
    *   that was passed to this function.
    *
-   * @param mem a pointer to a read-only buffer to feed an IOStream stream.
-   * @param size the buffer size, in bytes.
-   * @returns a pointer to a new IOStream structure or nullptr on failure; call
-   *          GetError() for more information.
+   * @param mem a read-only buffer to feed an IOStreamRef stream.
+   * @returns a valid IOStream on success.
+   * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
@@ -340,8 +338,8 @@
    * - `prop::IOStream.DYNAMIC_CHUNKSIZE_NUMBER`: memory will be allocated in
    *   multiples of this size, defaulting to 1024.
    *
-   * @returns a pointer to a new IOStream structure or nullptr on failure; call
-   *          GetError() for more information.
+   * @returns a valid IOStream on success.
+   * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread.
    *
@@ -370,7 +368,7 @@
    * @param iface the interface that implements this IOStream, initialized using
    *              InitInterface().
    * @param userdata the pointer that will be passed to the interface functions.
-   * @returns a pointer to the allocated memory on success.
+   * @returns a valid stream on success.
    * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread.
@@ -439,8 +437,7 @@
    * and makes the system and your app operate less efficiently, so do so
    * sparingly.
    *
-   * @returns true on success or false on failure; call GetError() for more
-   *          information.
+   * @throws Error on failure.
    *
    * @threadsafety This function is not thread safe.
    *
@@ -538,9 +535,43 @@
    */
   Sint64 Tell() const;
 
+  /**
+   * Read from a data source.
+   *
+   * This function reads up `size` bytes from the data source to the area
+   * pointed at by `ptr`. This function may read less bytes than requested.
+   *
+   * This function will return zero when the data stream is completely read, and
+   * IOStreamRef.GetStatus() will return IO_STATUS_EOF. If zero is returned and
+   * the stream is not at EOF, IOStreamRef.GetStatus() will return a different
+   * error value and GetError() will offer a human-readable message.
+   *
+   * @param size the number of bytes to read from the data source.
+   * @returns the bytes, or empty string on end of file or other failure;
+   *          call GetError() for more information.
+   *
+   * @threadsafety This function is not thread safe.
+   *
+   * @since This function is available since SDL 3.2.0.
+   *
+   * @sa IOStream.Write
+   * @sa IOStream.GetStatus
+   */
   std::string Read(size_t size = -1)
   {
-    static_assert(false, "Not implemented");
+    Sint64 pos = Tell();
+    auto curSize = SDL_GetIOSize(get());
+    if ((curSize < 0 || pos < 0)) {
+      if (size == size_t(-1)) return {};
+    } else if (curSize - pos <= 0) {
+      return {};
+    } else if (curSize - pos < size) {
+      size = curSize - pos;
+    }
+    std::string result(size, 0);
+    auto actualSize = Read(result);
+    if (actualSize < size) result.resize(actualSize);
+    return result;
   }
 
   /**
@@ -554,8 +585,7 @@
    * stream is not at EOF, IOStream.GetStatus() will return a different error
    * value and GetError() will offer a human-readable message.
    *
-   * @param ptr a pointer to a buffer to read data into.
-   * @param size the number of bytes to read from the data source.
+   * @param buf a pointer to a buffer to read data into.
    * @returns the number of bytes read, or 0 on end of file or other failure;
    *          call GetError() for more information.
    *
@@ -582,8 +612,7 @@
    * recoverable, such as a non-blocking write that can simply be retried later,
    * or a fatal error.
    *
-   * @param ptr a pointer to a buffer containing data to write.
-   * @param size the number of bytes to write.
+   * @param buf the bytes to write to
    * @returns the number of bytes written, which will be less than `size` on
    *          failure; call GetError() for more information.
    *
@@ -599,21 +628,37 @@
    */
   size_t Write(SourceBytes buf);
 
-  /// @cat formatted-string
+  /**
+   * Prints formatted string.
+   *
+   * @param fmt a std::format like format string
+   * @param args the arguments to be formatted
+   * @cat formatted-string
+   */
   size_t print(std::string_view fmt, auto... args)
   {
-    static_assert(false, "Not implemented");
+    return Write(std::vformat(fmt, std::make_format_args(args...)));
   }
 
-  /// @cat formatted-string
+  /**
+   * Prints formatted string.
+   *
+   * @param fmt a std::format like format string
+   * @param args the arguments to be formatted
+   * @cat formatted-string
+   */
   size_t println(std::string_view fmt, auto... args)
   {
-    static_assert(false, "Not implemented");
+    std::string result =
+      std::vformat(fmt, std::make_format_args(args...)) + "\n";
+    return Write(result);
   }
 
   /**
    * Print to an IOStream data stream.
    *
+   * @warning this is not typesafe! Prefer using print() and println()
+   *
    * This function does formatted printing to the stream.
    *
    * @param fmt a printf() style format string.
@@ -631,12 +676,21 @@
    */
   size_t printf(SDL_PRINTF_FORMAT_STRING const char* fmt, ...)
   {
-    return SDL_IOprintf(m_resource, fmt, ...);
+    va_list ap;
+    size_t result;
+
+    va_start(ap, fmt);
+    result = vprintf(fmt, ap);
+    va_end(ap);
+
+    return result;
   }
 
   /**
    * Print to an IOStream data stream.
    *
+   * @warning this is not typesafe! Prefer using print() and println()
+   *
    * This function does formatted printing to the stream.
    *
    * @param fmt a printf() style format string.
@@ -675,17 +729,10 @@
    * Load all the data from an SDL data stream.
    *
    * The data is allocated with a zero byte at the end (null terminated) for
-   * convenience. This extra byte is not included in the value reported via
-   * `datasize`.
-   *
-   * The data should be freed with free().
+   * convenience.
    *
-   * @param datasize a pointer filled in with the number of bytes read, may be
-   *                 nullptr.
-   * @param closeio if true, calls IOStream.Close() on `src` before returning,
-   *                even in the case of an error.
-   * @returns the data or nullptr on failure; call GetError() for more
-   *          information.
+   * @returns the data in bytes
+   * @throws Error on failure.
    *
    * @threadsafety This function is not thread safe.
    *
@@ -696,10 +743,29 @@
    */
   StringResult LoadFile();
 
+  /**
+   * Load all the data from an SDL data stream.
+   *
+   * The data is allocated with a zero byte at the end (null terminated) for
+   * convenience. This extra byte is not included in the value reported on
+   * the returned string.
+   *
+   * @returns the data in bytes
+   * @throws Error on failure.
+   *
+   * @threadsafety This function is not thread safe.
+   *
+   * @since This function is available since SDL 3.2.0.
+   *
+   * @sa LoadFile
+   * @sa IOStream.SaveFile
+   */
   template<class T>
   OwnArray<T> LoadFileAs()
   {
-    static_assert(false, "Not implemented");
+    size_t datasize = 0;
+    auto data = static_cast<T*>(SDL_LoadFile_IO(get(), &datasize, false));
+    return OwnArray<T>{CheckError(data), datasize / sizeof(T)};
   }
 
   /**
@@ -707,9 +773,6 @@
    *
    * @param data the data to be written. If datasize is 0, may be nullptr or a
    *             invalid pointer.
-   * @param datasize the number of bytes to be written.
-   * @param closeio if true, calls IOStream.Close() on `src` before returning,
-   *                even in the case of an error.
    * @throws Error on failure.
    *
    * @threadsafety This function is not thread safe.
@@ -729,8 +792,7 @@
    * and the stream is not at EOF, IOStream.GetStatus() will return a different
    * error value and GetError() will offer a human-readable message.
    *
-   * @param src the IOStream to read from.
-   * @param value a pointer filled in with the data read.
+   * @returns the data read on success.
    * @throws Error on failure.
    *
    * @threadsafety This function is not thread safe.
@@ -747,8 +809,7 @@
    * and the stream is not at EOF, IOStream.GetStatus() will return a different
    * error value and GetError() will offer a human-readable message.
    *
-   * @param src the IOStream to read from.
-   * @param value a pointer filled in with the data read.
+   * @returns the data read on success.
    * @throws Error on failure.
    *
    * @threadsafety This function is not thread safe.
@@ -769,8 +830,7 @@
    * and the stream is not at EOF, IOStream.GetStatus() will return a different
    * error value and GetError() will offer a human-readable message.
    *
-   * @param src the stream from which to read data.
-   * @param value a pointer filled in with the data read.
+   * @returns the data read on success.
    * @throws Error on failure.
    *
    * @threadsafety This function is not thread safe.
@@ -791,8 +851,7 @@
    * and the stream is not at EOF, IOStream.GetStatus() will return a different
    * error value and GetError() will offer a human-readable message.
    *
-   * @param src the stream from which to read data.
-   * @param value a pointer filled in with the data read.
+   * @returns the data read on success.
    * @throws Error on failure.
    *
    * @threadsafety This function is not thread safe.
@@ -813,8 +872,7 @@
    * and the stream is not at EOF, IOStream.GetStatus() will return a different
    * error value and GetError() will offer a human-readable message.
    *
-   * @param src the stream from which to read data.
-   * @param value a pointer filled in with the data read.
+   * @returns the data read on success.
    * @throws Error on failure.
    *
    * @threadsafety This function is not thread safe.
@@ -835,8 +893,7 @@
    * and the stream is not at EOF, IOStream.GetStatus() will return a different
    * error value and GetError() will offer a human-readable message.
    *
-   * @param src the stream from which to read data.
-   * @param value a pointer filled in with the data read.
+   * @returns the data read on success.
    * @throws Error on failure.
    *
    * @threadsafety This function is not thread safe.
@@ -857,8 +914,7 @@
    * and the stream is not at EOF, IOStream.GetStatus() will return a different
    * error value and GetError() will offer a human-readable message.
    *
-   * @param src the stream from which to read data.
-   * @param value a pointer filled in with the data read.
+   * @returns the data read on success.
    * @throws Error on failure.
    *
    * @threadsafety This function is not thread safe.
@@ -879,8 +935,7 @@
    * and the stream is not at EOF, IOStream.GetStatus() will return a different
    * error value and GetError() will offer a human-readable message.
    *
-   * @param src the stream from which to read data.
-   * @param value a pointer filled in with the data read.
+   * @returns the data read on success.
    * @throws Error on failure.
    *
    * @threadsafety This function is not thread safe.
@@ -901,8 +956,7 @@
    * and the stream is not at EOF, IOStream.GetStatus() will return a different
    * error value and GetError() will offer a human-readable message.
    *
-   * @param src the stream from which to read data.
-   * @param value a pointer filled in with the data read.
+   * @returns the data read on success.
    * @throws Error on failure.
    *
    * @threadsafety This function is not thread safe.
@@ -923,8 +977,7 @@
    * and the stream is not at EOF, IOStream.GetStatus() will return a different
    * error value and GetError() will offer a human-readable message.
    *
-   * @param src the stream from which to read data.
-   * @param value a pointer filled in with the data read.
+   * @returns the data read on success.
    * @throws Error on failure.
    *
    * @threadsafety This function is not thread safe.
@@ -945,8 +998,7 @@
    * and the stream is not at EOF, IOStream.GetStatus() will return a different
    * error value and GetError() will offer a human-readable message.
    *
-   * @param src the stream from which to read data.
-   * @param value a pointer filled in with the data read.
+   * @returns the data read on success.
    * @throws Error on failure.
    *
    * @threadsafety This function is not thread safe.
@@ -967,8 +1019,7 @@
    * and the stream is not at EOF, IOStream.GetStatus() will return a different
    * error value and GetError() will offer a human-readable message.
    *
-   * @param src the stream from which to read data.
-   * @param value a pointer filled in with the data read.
+   * @returns the data read on success.
    * @throws Error on failure.
    *
    * @threadsafety This function is not thread safe.
@@ -989,8 +1040,7 @@
    * and the stream is not at EOF, IOStream.GetStatus() will return a different
    * error value and GetError() will offer a human-readable message.
    *
-   * @param src the stream from which to read data.
-   * @param value a pointer filled in with the data read.
+   * @returns the data read on success.
    * @throws Error on failure.
    *
    * @threadsafety This function is not thread safe.
@@ -1011,8 +1061,7 @@
    * and the stream is not at EOF, IOStream.GetStatus() will return a different
    * error value and GetError() will offer a human-readable message.
    *
-   * @param src the stream from which to read data.
-   * @param value a pointer filled in with the data read.
+   * @returns the data read on success.
    * @throws Error on failure.
    *
    * @threadsafety This function is not thread safe.
@@ -1021,68 +1070,332 @@
    */
   Sint64 ReadS64BE();
 
-  std::optional<Uint8> TryReadU8() { static_assert(false, "Not implemented"); }
+  /**
+   * Use this function to read a byte from an IOStreamRef.
+   *
+   * This function will return false when the data stream is completely read,
+   * and IOStreamRef.GetStatus() will return IO_STATUS_EOF. If false is
+   * returned and the stream is not at EOF, IOStreamRef.GetStatus() will return
+   * a different error value and GetError() will offer a human-readable message.
+   *
+   * @returns the data read on success, std::nullopt on failure.
+   *
+   * @threadsafety This function is not thread safe.
+   *
+   * @since This function is available since SDL 3.2.0.
+   */
+  std::optional<Uint8> TryReadU8()
+  {
+    if (Uint8 value; SDL_ReadU8(get(), &value)) return value;
+    return {};
+  }
 
-  std::optional<Sint8> TryReadS8() { static_assert(false, "Not implemented"); }
+  /**
+   * Use this function to read a byte from an IOStreamRef.
+   *
+   * This function will return false when the data stream is completely read,
+   * and IOStreamRef.GetStatus() will return IO_STATUS_EOF. If false is
+   * returned and the stream is not at EOF, IOStreamRef.GetStatus() will return
+   * a different error value and GetError() will offer a human-readable message.
+   *
+   * @returns the data read on success, std::nullopt on failure.
+   *
+   * @threadsafety This function is not thread safe.
+   *
+   * @since This function is available since SDL 3.2.0.
+   */
+  std::optional<Sint8> TryReadS8()
+  {
+    if (Sint8 value; SDL_ReadS8(get(), &value)) return value;
+    return {};
+  }
 
+  /**
+   * Use this function to read 16 bits of little-endian data from an
+   * IOStreamRef and return in native format.
+   *
+   * SDL byteswaps the data only if necessary, so the data returned will be in
+   * the native byte order.
+   *
+   * This function will return false when the data stream is completely read,
+   * and IOStreamRef.GetStatus() will return IO_STATUS_EOF. If false is
+   * returned and the stream is not at EOF, IOStreamRef.GetStatus() will return
+   * a different error value and GetError() will offer a human-readable message.
+   *
+   * @returns the data read on success, std::nullopt on failure.
+   *
+   * @threadsafety This function is not thread safe.
+   *
+   * @since This function is available since SDL 3.2.0.
+   */
   std::optional<Uint16> TryReadU16LE()
   {
-    static_assert(false, "Not implemented");
+    if (Uint16 value; SDL_ReadU16LE(get(), &value)) return value;
+    return {};
   }
 
+  /**
+   * Use this function to read 16 bits of little-endian data from an
+   * IOStreamRef and return in native format.
+   *
+   * SDL byteswaps the data only if necessary, so the data returned will be in
+   * the native byte order.
+   *
+   * This function will return false when the data stream is completely read,
+   * and IOStreamRef.GetStatus() will return IO_STATUS_EOF. If false is
+   * returned and the stream is not at EOF, IOStreamRef.GetStatus() will return
+   * a different error value and GetError() will offer a human-readable message.
+   *
+   * @returns the data read on success, std::nullopt on failure.
+   *
+   * @threadsafety This function is not thread safe.
+   *
+   * @since This function is available since SDL 3.2.0.
+   */
   std::optional<Sint16> TryReadS16LE()
   {
-    static_assert(false, "Not implemented");
+    if (Sint16 value; SDL_ReadS16LE(get(), &value)) return value;
+    return {};
   }
 
+  /**
+   * Use this function to read 16 bits of big-endian data from an IOStreamRef
+   * and return in native format.
+   *
+   * SDL byteswaps the data only if necessary, so the data returned will be in
+   * the native byte order.
+   *
+   * This function will return false when the data stream is completely read,
+   * and IOStreamRef.GetStatus() will return IO_STATUS_EOF. If false is
+   * returned and the stream is not at EOF, IOStreamRef.GetStatus() will return
+   * a different error value and GetError() will offer a human-readable message.
+   *
+   * @returns the data read on success, std::nullopt on failure.
+   *
+   * @threadsafety This function is not thread safe.
+   *
+   * @since This function is available since SDL 3.2.0.
+   */
   std::optional<Uint16> TryReadU16BE()
   {
-    static_assert(false, "Not implemented");
+    if (Uint16 value; SDL_ReadU16BE(get(), &value)) return value;
+    return {};
   }
 
+  /**
+   * Use this function to read 16 bits of big-endian data from an IOStreamRef
+   * and return in native format.
+   *
+   * SDL byteswaps the data only if necessary, so the data returned will be in
+   * the native byte order.
+   *
+   * This function will return false when the data stream is completely read,
+   * and IOStreamRef.GetStatus() will return IO_STATUS_EOF. If false is
+   * returned and the stream is not at EOF, IOStreamRef.GetStatus() will return
+   * a different error value and GetError() will offer a human-readable message.
+   *
+   * @returns the data read on success, std::nullopt on failure.
+   *
+   * @threadsafety This function is not thread safe.
+   *
+   * @since This function is available since SDL 3.2.0.
+   */
   std::optional<Sint16> TryReadS16BE()
   {
-    static_assert(false, "Not implemented");
+    if (Sint16 value; SDL_ReadS16BE(get(), &value)) return value;
+    return {};
   }
 
+  /**
+   * Use this function to read 32 bits of little-endian data from an
+   * IOStreamRef and return in native format.
+   *
+   * SDL byteswaps the data only if necessary, so the data returned will be in
+   * the native byte order.
+   *
+   * This function will return false when the data stream is completely read,
+   * and IOStreamRef.GetStatus() will return IO_STATUS_EOF. If false is
+   * returned and the stream is not at EOF, IOStreamRef.GetStatus() will return
+   * a different error value and GetError() will offer a human-readable message.
+   *
+   * @returns the data read on success, std::nullopt on failure.
+   *
+   * @threadsafety This function is not thread safe.
+   *
+   * @since This function is available since SDL 3.2.0.
+   */
   std::optional<Uint32> TryReadU32LE()
   {
-    static_assert(false, "Not implemented");
+    if (Uint32 value; SDL_ReadU32LE(get(), &value)) return value;
+    return {};
   }
 
+  /**
+   * Use this function to read 32 bits of little-endian data from an
+   * IOStreamRef and return in native format.
+   *
+   * SDL byteswaps the data only if necessary, so the data returned will be in
+   * the native byte order.
+   *
+   * This function will return false when the data stream is completely read,
+   * and IOStreamRef.GetStatus() will return IO_STATUS_EOF. If false is
+   * returned and the stream is not at EOF, IOStreamRef.GetStatus() will return
+   * a different error value and GetError() will offer a human-readable message.
+   *
+   * @returns the data read on success, std::nullopt on failure.
+   *
+   * @threadsafety This function is not thread safe.
+   *
+   * @since This function is available since SDL 3.2.0.
+   */
   std::optional<Sint32> TryReadS32LE()
   {
-    static_assert(false, "Not implemented");
+    if (Sint32 value; SDL_ReadS32LE(get(), &value)) return value;
+    return {};
   }
 
+  /**
+   * Use this function to read 32 bits of big-endian data from an IOStreamRef
+   * and return in native format.
+   *
+   * SDL byteswaps the data only if necessary, so the data returned will be in
+   * the native byte order.
+   *
+   * This function will return false when the data stream is completely read,
+   * and IOStreamRef.GetStatus() will return IO_STATUS_EOF. If false is
+   * returned and the stream is not at EOF, IOStreamRef.GetStatus() will return
+   * a different error value and GetError() will offer a human-readable message.
+   *
+   * @returns the data read on success, std::nullopt on failure.
+   *
+   * @threadsafety This function is not thread safe.
+   *
+   * @since This function is available since SDL 3.2.0.
+   */
   std::optional<Uint32> TryReadU32BE()
   {
-    static_assert(false, "Not implemented");
+    if (Uint32 value; SDL_ReadU32BE(get(), &value)) return value;
+    return {};
   }
 
+  /**
+   * Use this function to read 32 bits of big-endian data from an IOStreamRef
+   * and return in native format.
+   *
+   * SDL byteswaps the data only if necessary, so the data returned will be in
+   * the native byte order.
+   *
+   * This function will return false when the data stream is completely read,
+   * and IOStreamRef.GetStatus() will return IO_STATUS_EOF. If false is
+   * returned and the stream is not at EOF, IOStreamRef.GetStatus() will return
+   * a different error value and GetError() will offer a human-readable message.
+   *
+   * @returns the data read on success, std::nullopt on failure.
+   *
+   * @threadsafety This function is not thread safe.
+   *
+   * @since This function is available since SDL 3.2.0.
+   */
   std::optional<Sint32> TryReadS32BE()
   {
-    static_assert(false, "Not implemented");
+    if (Sint32 value; SDL_ReadS32BE(get(), &value)) return value;
+    return {};
   }
 
+  /**
+   * Use this function to read 64 bits of little-endian data from an
+   * IOStreamRef and return in native format.
+   *
+   * SDL byteswaps the data only if necessary, so the data returned will be in
+   * the native byte order.
+   *
+   * This function will return false when the data stream is completely read,
+   * and IOStreamRef.GetStatus() will return IO_STATUS_EOF. If false is
+   * returned and the stream is not at EOF, IOStreamRef.GetStatus() will return
+   * a different error value and GetError() will offer a human-readable message.
+   *
+   * @returns the data read on success, std::nullopt on failure.
+   *
+   * @threadsafety This function is not thread safe.
+   *
+   * @since This function is available since SDL 3.2.0.
+   */
   std::optional<Uint64> TryReadU64LE()
   {
-    static_assert(false, "Not implemented");
+    if (Uint64 value; SDL_ReadU64LE(get(), &value)) return value;
+    return {};
   }
 
+  /**
+   * Use this function to read 64 bits of little-endian data from an
+   * IOStreamRef and return in native format.
+   *
+   * SDL byteswaps the data only if necessary, so the data returned will be in
+   * the native byte order.
+   *
+   * This function will return false when the data stream is completely read,
+   * and IOStreamRef.GetStatus() will return IO_STATUS_EOF. If false is
+   * returned and the stream is not at EOF, IOStreamRef.GetStatus() will return
+   * a different error value and GetError() will offer a human-readable message.
+   *
+   * @returns the data read on success, std::nullopt on failure.
+   *
+   * @threadsafety This function is not thread safe.
+   *
+   * @since This function is available since SDL 3.2.0.
+   */
   std::optional<Sint64> TryReadS64LE()
   {
-    static_assert(false, "Not implemented");
+    if (Sint64 value; SDL_ReadS64LE(get(), &value)) return value;
+    return {};
   }
 
+  /**
+   * Use this function to read 64 bits of big-endian data from an IOStreamRef
+   * and return in native format.
+   *
+   * SDL byteswaps the data only if necessary, so the data returned will be in
+   * the native byte order.
+   *
+   * This function will return false when the data stream is completely read,
+   * and IOStreamRef.GetStatus() will return IO_STATUS_EOF. If false is
+   * returned and the stream is not at EOF, IOStreamRef.GetStatus() will return
+   * a different error value and GetError() will offer a human-readable message.
+   *
+   * @returns the data read on success, std::nullopt on failure.
+   *
+   * @threadsafety This function is not thread safe.
+   *
+   * @since This function is available since SDL 3.2.0.
+   */
   std::optional<Uint64> TryReadU64BE()
   {
-    static_assert(false, "Not implemented");
+    if (Uint64 value; SDL_ReadU64BE(get(), &value)) return value;
+    return {};
   }
 
+  /**
+   * Use this function to read 64 bits of big-endian data from an IOStreamRef
+   * and return in native format.
+   *
+   * SDL byteswaps the data only if necessary, so the data returned will be in
+   * the native byte order.
+   *
+   * This function will return false when the data stream is completely read,
+   * and IOStreamRef.GetStatus() will return IO_STATUS_EOF. If false is
+   * returned and the stream is not at EOF, IOStreamRef.GetStatus() will return
+   * a different error value and GetError() will offer a human-readable message.
+   *
+   * @returns the data read on success, std::nullopt on failure.
+   *
+   * @threadsafety This function is not thread safe.
+   *
+   * @since This function is available since SDL 3.2.0.
+   */
   std::optional<Sint64> TryReadS64BE()
   {
-    static_assert(false, "Not implemented");
+    if (Sint64 value; SDL_ReadS64BE(get(), &value)) return value;
+    return {};
   }
 
   /**
@@ -1467,10 +1780,9 @@
  * - `prop::IOStream.MEMORY_SIZE_NUMBER`: this will be the `size` parameter that
  *   was passed to this function.
  *
- * @param mem a pointer to a buffer to feed an IOStream stream.
- * @param size the buffer size, in bytes.
- * @returns a pointer to a new IOStream structure or nullptr on failure; call
- *          GetError() for more information.
+ * @param mem a buffer to feed an IOStream stream.
+ * @returns a valid IOStream on success.
+ * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -1484,9 +1796,9 @@
  * @sa IOStream.Tell
  * @sa IOStream.Write
  */
-inline IOStream IOFromMem(void* mem, size_t size)
+inline IOStream IOFromMem(TargetBytes mem)
 {
-  return IOStream(SDL_IOFromMem(mem, size));
+  return IOStream(CheckError(SDL_IOFromMem(mem.data(), mem.size_bytes())));
 }
 
 inline IOStream IOStream::FromMem(TargetBytes mem)
@@ -1517,10 +1829,9 @@
  * - `prop::IOStream.MEMORY_SIZE_NUMBER`: this will be the `size` parameter that
  *   was passed to this function.
  *
- * @param mem a pointer to a read-only buffer to feed an IOStream stream.
- * @param size the buffer size, in bytes.
- * @returns a pointer to a new IOStream structure or nullptr on failure; call
- *          GetError() for more information.
+ * @param mem a read-only buffer to feed an IOStreamRef stream.
+ * @returns a valid IOStream on success.
+ * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
  *
@@ -1532,9 +1843,9 @@
  * @sa IOStream.Seek
  * @sa IOStream.Tell
  */
-inline IOStream IOFromConstMem(const void* mem, size_t size)
+inline IOStream IOFromConstMem(SourceBytes mem)
 {
-  return IOStream(SDL_IOFromConstMem(mem, size));
+  return IOStream(CheckError(SDL_IOFromConstMem(mem.data(), mem.size_bytes())));
 }
 
 inline IOStream IOStream::FromConstMem(SourceBytes mem)
@@ -1656,7 +1967,7 @@
  */
 inline PropertiesRef GetIOProperties(IOStreamParam context)
 {
-  return CheckError(SDL_GetIOProperties(context));
+  return {CheckError(SDL_GetIOProperties(context))};
 }
 
 inline PropertiesRef IOStream::GetProperties() const
@@ -1780,8 +2091,7 @@
  * value and GetError() will offer a human-readable message.
  *
  * @param context a pointer to an IOStream structure.
- * @param ptr a pointer to a buffer to read data into.
- * @param size the number of bytes to read from the data source.
+ * @param buf a pointer to a buffer to read data into.
  * @returns the number of bytes read, or 0 on end of file or other failure; call
  *          GetError() for more information.
  *
@@ -1817,8 +2127,7 @@
  * or a fatal error.
  *
  * @param context a pointer to an IOStream structure.
- * @param ptr a pointer to a buffer containing data to write.
- * @param size the number of bytes to write.
+ * @param buf a pointer to a buffer containing data to write.
  * @returns the number of bytes written, which will be less than `size` on
  *          failure; call GetError() for more information.
  *
@@ -1845,6 +2154,8 @@
 /**
  * Print to an IOStream data stream.
  *
+ * @warning this is not typesafe! Prefer using print() and println()
+ *
  * This function does formatted printing to the stream.
  *
  * @param context a pointer to an IOStream structure.
@@ -1862,9 +2173,17 @@
  * @sa IOStream.Write
  */
 inline size_t IOprintf(IOStreamParam context,
-                       SDL_PRINTF_FORMAT_STRING const char* fmt)
+                       SDL_PRINTF_FORMAT_STRING const char* fmt,
+                       ...)
 {
-  return SDL_IOprintf(context, fmt);
+  va_list ap;
+  size_t result;
+
+  va_start(ap, fmt);
+  result = SDL_IOvprintf(context, fmt, ap);
+  va_end(ap);
+
+  return result;
 }
 
 /**
@@ -1923,14 +2242,12 @@
  * Load all the data from an SDL data stream.
  *
  * The data is allocated with a zero byte at the end (null terminated) for
- * convenience. This extra byte is not included in the value reported via
- * `datasize`.
+ * convenience. This extra byte is not included in the value reported on
+ * the returned string.
  *
  * The data should be freed with free().
  *
  * @param src the IOStream to read all available data from.
- * @param datasize a pointer filled in with the number of bytes read, may be
- *                 nullptr.
  * @param closeio if true, calls IOStream.Close() on `src` before returning,
  *                even in the case of an error.
  * @returns the data or nullptr on failure; call GetError() for more
@@ -1945,22 +2262,21 @@
  */
 inline StringResult LoadFile(IOStreamParam src, bool closeio = true)
 {
-  return SDL_LoadFile_IO(src, closeio);
+  size_t datasize = 0;
+  auto data = static_cast<char*>(SDL_LoadFile_IO(src, &datasize, closeio));
+  return StringResult{CheckError(data), datasize};
 }
 
 /**
  * Load all the data from a file path.
  *
  * The data is allocated with a zero byte at the end (null terminated) for
- * convenience. This extra byte is not included in the value reported via
- * `datasize`.
- *
- * The data should be freed with free().
+ * convenience. This extra byte is not included in the value reported on
+ * the returned string.
  *
  * @param file the path to read all available data from.
- * @param datasize if not nullptr, will store the number of bytes read.
- * @returns the data or nullptr on failure; call GetError() for more
- *          information.
+ * @returns the data.
+ * @throws Error on failure.
  *
  * @threadsafety This function is not thread safe.
  *
@@ -1969,23 +2285,49 @@
  * @sa IOStream.LoadFile
  * @sa SaveFile
  */
-inline StringResult LoadFile(StringParam file) { return SDL_LoadFile(file); }
+inline StringResult LoadFile(StringParam file)
+{
+  size_t datasize = 0;
+  auto data = static_cast<char*>(SDL_LoadFile(file, &datasize));
+  return StringResult{CheckError(data), datasize};
+}
 
-inline StringResult IOStream::LoadFile() { return SDL::LoadFile(m_resource); }
+inline StringResult IOStream::LoadFile()
+{
+  return SDL::LoadFile(m_resource, false);
+}
 
+/**
+ * Load all the data from a file path.
+ *
+ * The data is allocated with a zero byte at the end (null terminated) for
+ * convenience. This extra byte is not included in the value reported on
+ * the returned string.
+ *
+ * @param file the path to read all available data from.
+ * @returns the data.
+ * @throws Error on failure.
+ *
+ * @threadsafety This function is not thread safe.
+ *
+ * @since This function is available since SDL 3.2.0.
+ *
+ * @sa IOStream.LoadFile
+ * @sa SaveFile
+ */
 template<class T>
 inline OwnArray<T> LoadFileAs(StringParam file)
 {
-  static_assert(false, "Not implemented");
+  size_t datasize = 0;
+  auto data = static_cast<T*>(SDL_LoadFile(file, &datasize));
+  return OwnArray<T>{CheckError(data), datasize / sizeof(T)};
 }
 
 /**
  * Save all the data into an SDL data stream.
  *
  * @param src the IOStream to write all data to.
- * @param data the data to be written. If datasize is 0, may be nullptr or a
- *             invalid pointer.
- * @param datasize the number of bytes to be written.
+ * @param data the data to be written.
  * @param closeio if true, calls IOStream.Close() on `src` before returning,
  *                even in the case of an error.
  * @throws Error on failure.
@@ -1997,7 +2339,7 @@
  * @sa SaveFile
  * @sa IOStream.LoadFile
  */
-inline void SaveFile(IOStreamParam src, SourceBytes data, bool closeio = true)
+inline void SaveFile(IOStreamParam src, SourceBytes data, bool closeio = false)
 {
   CheckError(SDL_SaveFile_IO(src, data.data(), data.size_bytes(), closeio));
 }
@@ -2006,9 +2348,7 @@
  * Save all the data into a file path.
  *
  * @param file the path to write all available data into.
- * @param data the data to be written. If datasize is 0, may be nullptr or a
- *             invalid pointer.
- * @param datasize the number of bytes to be written.
+ * @param data the data to be written.
  * @throws Error on failure.
  *
  * @threadsafety This function is not thread safe.
@@ -2037,16 +2377,19 @@
  * value and GetError() will offer a human-readable message.
  *
  * @param src the IOStream to read from.
- * @param value a pointer filled in with the data read.
+ * @return the  data read.
  * @throws Error on failure.
  *
  * @threadsafety This function is not thread safe.
  *
  * @since This function is available since SDL 3.2.0.
  */
-inline Uint8 ReadU8(IOStreamParam src) { return CheckError(SDL_ReadU8(src)); }
-
-inline Uint8 IOStream::ReadU8() { return SDL::ReadU8(m_resource); }
+inline Uint8 ReadU8(IOStreamParam src)
+{
+  Uint8 value;
+  CheckError(SDL_ReadU8(src, &value));
+  return value;
+}
 
 /**
  * Use this function to read a signed byte from an IOStream.
@@ -2057,16 +2400,19 @@
  * value and GetError() will offer a human-readable message.
  *
  * @param src the IOStream to read from.
- * @param value a pointer filled in with the data read.
+ * @return the  data read.
  * @throws Error on failure.
  *
  * @threadsafety This function is not thread safe.
  *
  * @since This function is available since SDL 3.2.0.
  */
-inline Sint8 ReadS8(IOStreamParam src) { return CheckError(SDL_ReadS8(src)); }
-
-inline Sint8 IOStream::ReadS8() { return SDL::ReadS8(m_resource); }
+inline Sint8 ReadS8(IOStreamParam src)
+{
+  Sint8 value;
+  CheckError(SDL_ReadS8(src, &value));
+  return value;
+}
 
 /**
  * Use this function to read 16 bits of little-endian data from an IOStream and
@@ -2081,7 +2427,7 @@
  * value and GetError() will offer a human-readable message.
  *
  * @param src the stream from which to read data.
- * @param value a pointer filled in with the data read.
+ * @return the  data read.
  * @throws Error on failure.
  *
  * @threadsafety This function is not thread safe.
@@ -2090,7 +2436,9 @@
  */
 inline Uint16 ReadU16LE(IOStreamParam src)
 {
-  return CheckError(SDL_ReadU16LE(src));
+  Uint16 value;
+  CheckError(SDL_ReadU16LE(src, &value));
+  return value;
 }
 
 inline Uint16 IOStream::ReadU16LE() { return SDL::ReadU16LE(m_resource); }
@@ -2108,7 +2456,7 @@
  * value and GetError() will offer a human-readable message.
  *
  * @param src the stream from which to read data.
- * @param value a pointer filled in with the data read.
+ * @return the  data read.
  * @throws Error on failure.
  *
  * @threadsafety This function is not thread safe.
@@ -2117,7 +2465,9 @@
  */
 inline Sint16 ReadS16LE(IOStreamParam src)
 {
-  return CheckError(SDL_ReadS16LE(src));
+  Sint16 value;
+  CheckError(SDL_ReadS16LE(src, &value));
+  return value;
 }
 
 inline Sint16 IOStream::ReadS16LE() { return SDL::ReadS16LE(m_resource); }
@@ -2135,7 +2485,7 @@
  * value and GetError() will offer a human-readable message.
  *
  * @param src the stream from which to read data.
- * @param value a pointer filled in with the data read.
+ * @return the  data read.
  * @throws Error on failure.
  *
  * @threadsafety This function is not thread safe.
@@ -2144,7 +2494,9 @@
  */
 inline Uint16 ReadU16BE(IOStreamParam src)
 {
-  return CheckError(SDL_ReadU16BE(src));
+  Uint16 value;
+  CheckError(SDL_ReadU16BE(src, &value));
+  return value;
 }
 
 inline Uint16 IOStream::ReadU16BE() { return SDL::ReadU16BE(m_resource); }
@@ -2162,7 +2514,7 @@
  * value and GetError() will offer a human-readable message.
  *
  * @param src the stream from which to read data.
- * @param value a pointer filled in with the data read.
+ * @return the  data read.
  * @throws Error on failure.
  *
  * @threadsafety This function is not thread safe.
@@ -2171,7 +2523,9 @@
  */
 inline Sint16 ReadS16BE(IOStreamParam src)
 {
-  return CheckError(SDL_ReadS16BE(src));
+  Sint16 value;
+  CheckError(SDL_ReadS16BE(src, &value));
+  return value;
 }
 
 inline Sint16 IOStream::ReadS16BE() { return SDL::ReadS16BE(m_resource); }
@@ -2189,7 +2543,7 @@
  * value and GetError() will offer a human-readable message.
  *
  * @param src the stream from which to read data.
- * @param value a pointer filled in with the data read.
+ * @return the  data read.
  * @throws Error on failure.
  *
  * @threadsafety This function is not thread safe.
@@ -2198,7 +2552,9 @@
  */
 inline Uint32 ReadU32LE(IOStreamParam src)
 {
-  return CheckError(SDL_ReadU32LE(src));
+  Uint32 value;
+  CheckError(SDL_ReadU32LE(src, &value));
+  return value;
 }
 
 inline Uint32 IOStream::ReadU32LE() { return SDL::ReadU32LE(m_resource); }
@@ -2216,7 +2572,7 @@
  * value and GetError() will offer a human-readable message.
  *
  * @param src the stream from which to read data.
- * @param value a pointer filled in with the data read.
+ * @return the  data read.
  * @throws Error on failure.
  *
  * @threadsafety This function is not thread safe.
@@ -2225,7 +2581,9 @@
  */
 inline Sint32 ReadS32LE(IOStreamParam src)
 {
-  return CheckError(SDL_ReadS32LE(src));
+  Sint32 value;
+  CheckError(SDL_ReadS32LE(src, &value));
+  return value;
 }
 
 inline Sint32 IOStream::ReadS32LE() { return SDL::ReadS32LE(m_resource); }
@@ -2243,7 +2601,7 @@
  * value and GetError() will offer a human-readable message.
  *
  * @param src the stream from which to read data.
- * @param value a pointer filled in with the data read.
+ * @return the  data read.
  * @throws Error on failure.
  *
  * @threadsafety This function is not thread safe.
@@ -2252,7 +2610,9 @@
  */
 inline Uint32 ReadU32BE(IOStreamParam src)
 {
-  return CheckError(SDL_ReadU32BE(src));
+  Uint32 value;
+  CheckError(SDL_ReadU32BE(src, &value));
+  return value;
 }
 
 inline Uint32 IOStream::ReadU32BE() { return SDL::ReadU32BE(m_resource); }
@@ -2270,7 +2630,7 @@
  * value and GetError() will offer a human-readable message.
  *
  * @param src the stream from which to read data.
- * @param value a pointer filled in with the data read.
+ * @return the  data read.
  * @throws Error on failure.
  *
  * @threadsafety This function is not thread safe.
@@ -2279,7 +2639,9 @@
  */
 inline Sint32 ReadS32BE(IOStreamParam src)
 {
-  return CheckError(SDL_ReadS32BE(src));
+  Sint32 value;
+  CheckError(SDL_ReadS32BE(src, &value));
+  return value;
 }
 
 inline Sint32 IOStream::ReadS32BE() { return SDL::ReadS32BE(m_resource); }
@@ -2297,7 +2659,7 @@
  * value and GetError() will offer a human-readable message.
  *
  * @param src the stream from which to read data.
- * @param value a pointer filled in with the data read.
+ * @return the  data read.
  * @throws Error on failure.
  *
  * @threadsafety This function is not thread safe.
@@ -2306,7 +2668,9 @@
  */
 inline Uint64 ReadU64LE(IOStreamParam src)
 {
-  return CheckError(SDL_ReadU64LE(src));
+  Uint64 value;
+  CheckError(SDL_ReadU64LE(src, &value));
+  return value;
 }
 
 inline Uint64 IOStream::ReadU64LE() { return SDL::ReadU64LE(m_resource); }
@@ -2324,7 +2688,7 @@
  * value and GetError() will offer a human-readable message.
  *
  * @param src the stream from which to read data.
- * @param value a pointer filled in with the data read.
+ * @return the  data read.
  * @throws Error on failure.
  *
  * @threadsafety This function is not thread safe.
@@ -2333,7 +2697,9 @@
  */
 inline Sint64 ReadS64LE(IOStreamParam src)
 {
-  return CheckError(SDL_ReadS64LE(src));
+  Sint64 value;
+  CheckError(SDL_ReadS64LE(src, &value));
+  return value;
 }
 
 inline Sint64 IOStream::ReadS64LE() { return SDL::ReadS64LE(m_resource); }
@@ -2351,7 +2717,7 @@
  * value and GetError() will offer a human-readable message.
  *
  * @param src the stream from which to read data.
- * @param value a pointer filled in with the data read.
+ * @return the  data read.
  * @throws Error on failure.
  *
  * @threadsafety This function is not thread safe.
@@ -2360,7 +2726,9 @@
  */
 inline Uint64 ReadU64BE(IOStreamParam src)
 {
-  return CheckError(SDL_ReadU64BE(src));
+  Uint64 value;
+  CheckError(SDL_ReadU64BE(src, &value));
+  return value;
 }
 
 inline Uint64 IOStream::ReadU64BE() { return SDL::ReadU64BE(m_resource); }
@@ -2378,7 +2746,7 @@
  * value and GetError() will offer a human-readable message.
  *
  * @param src the stream from which to read data.
- * @param value a pointer filled in with the data read.
+ * @return the  data read.
  * @throws Error on failure.
  *
  * @threadsafety This function is not thread safe.
@@ -2387,7 +2755,9 @@
  */
 inline Sint64 ReadS64BE(IOStreamParam src)
 {
-  return CheckError(SDL_ReadS64BE(src));
+  Sint64 value;
+  CheckError(SDL_ReadS64BE(src, &value));
+  return value;
 }
 
 inline Sint64 IOStream::ReadS64BE() { return SDL::ReadS64BE(m_resource); }
