--- src/generated/SDL3pp_image.h
+++ include/SDL3pp/SDL3pp_image.h
@@ -1,7 +1,6 @@
 #ifndef SDL3PP_IMAGE_H_
 #define SDL3PP_IMAGE_H_
 
-#include <SDL3/SDL_image.h>
 #include "SDL3pp_error.h"
 #include "SDL3pp_gpu.h"
 #include "SDL3pp_mouse.h"
@@ -10,10 +9,14 @@
 #include "SDL3pp_version.h"
 #include "SDL3pp_video.h"
 
+#if defined(SDL3PP_ENABLE_IMAGE) || defined(SDL3PP_DOC)
+
+#include <SDL3_image/SDL_image.h>
+
 namespace SDL {
 
 /**
- * @defgroup CategorySDLImage Category SDLImage
+ * @defgroup CategorySDLImage Load images from several formats
  *
  * Header file for SDL_image library
  *
@@ -97,6 +100,8 @@
   constexpr auto operator->() { return value; }
 };
 
+#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
+
 // Forward decl
 struct AnimationEncoder;
 
@@ -170,15 +175,26 @@
   /// Converts to underlying AnimationDecoderRaw
   constexpr operator AnimationDecoderRaw() const { return value; }
 };
+#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
+
+#ifdef SDL3PP_DOC
 
-/// Printable format: "%d.%d.%d", MAJOR, MINOR, MICRO
+/**
+ * @name Image version
+ * @{
+ * Printable format: "%d.%d.%d", MAJOR, MINOR, MICRO
+ */
 #define SDL_IMAGE_MAJOR_VERSION
 
 #define SDL_IMAGE_MINOR_VERSION
 
 #define SDL_IMAGE_MICRO_VERSION
 
-/// This is the version number macro for the current SDL_image version.
+/// @}
+
+/**
+ * This is the version number macro for the current SDL_image version.
+ */
 #define SDL_IMAGE_VERSION                                                      \
   SDL_VERSIONNUM(                                                              \
     SDL_IMAGE_MAJOR_VERSION, SDL_IMAGE_MINOR_VERSION, SDL_IMAGE_MICRO_VERSION)
@@ -190,6 +206,8 @@
    (SDL_IMAGE_MAJOR_VERSION > X || SDL_IMAGE_MINOR_VERSION > Y ||              \
     SDL_IMAGE_MICRO_VERSION >= Z))
 
+#endif // SDL3PP_DOC
+
 namespace IMG {
 
 /**
@@ -199,12 +217,10 @@
  *
  * @since This function is available since SDL_image 3.0.0.
  */
-inline int Version();
+inline int Version() { return IMG_Version(); }
 
 } // namespace IMG
 
-inline int IMG::Version() { return SDL::Version(); }
-
 /**
  * Load an image from a filesystem path into a software surface.
  *
@@ -229,10 +245,7 @@
  *
  * If you are using SDL's 2D rendering API, there is an equivalent call to load
  * images directly into an Texture for use by the GPU without using a software
- * surface: call Texture.Texture() instead.
- *
- * When done with the returned surface, the app should dispose of it with a call
- * to [Surface.Destroy](https://wiki.libsdl.org/SDL3/Surface.Destroy) ().
+ * surface: call LoadTexture() instead.
  *
  * @param file a path on the filesystem to load an image from.
  * @returns a new SDL surface, or nullptr on error.
@@ -242,7 +255,7 @@
  * @sa LoadSurfaceTyped
  * @sa Surface.Surface
  */
-inline Surface LoadSurface(StringParam file) { return IMG_Load(file); }
+inline Surface LoadSurface(StringParam file) { return Surface{IMG_Load(file)}; }
 
 /**
  * Load an image from an SDL data source into a software surface.
@@ -267,7 +280,7 @@
  * during this call in any case.
  *
  * There is a separate function to read files from disk without having to deal
- * with IOStream: `Surface.Surface("filename.jpg")` will call this function and
+ * with IOStream: `LoadSurface("filename.jpg")` will call this function and
  * manage those details for you, determining the file type from the filename's
  * extension.
  *
@@ -277,10 +290,7 @@
  *
  * If you are using SDL's 2D rendering API, there is an equivalent call to load
  * images directly into an Texture for use by the GPU without using a software
- * surface: call Texture.Texture() instead.
- *
- * When done with the returned surface, the app should dispose of it with a call
- * to Surface.Destroy().
+ * surface: call LoadSurface() instead.
  *
  * @param src an IOStream that data will be read from.
  * @param closeio true to close/free the IOStream before returning, false to
@@ -294,7 +304,7 @@
  */
 inline Surface LoadSurface(IOStreamParam src, bool closeio)
 {
-  return IMG_Load_IO(src, closeio);
+  return Surface{IMG_Load_IO(src, closeio)};
 }
 
 inline Surface::Surface(StringParam file)
@@ -336,11 +346,11 @@
  * ability to guess the format.
  *
  * There is a separate function to read files from disk without having to deal
- * with IOStream: `Surface.Surface("filename.jpg")` will call this function and
+ * with IOStream: `LoadSurface("filename.jpg")` will call this function and
  * manage those details for you, determining the file type from the filename's
  * extension.
  *
- * There is also Surface.Surface(), which is equivalent to this function except
+ * There is also LoadSurface(), which is equivalent to this function except
  * that it will rely on SDL_image to determine what type of data it is loading,
  * much like passing a nullptr for type.
  *
@@ -348,14 +358,11 @@
  * images directly into an Texture for use by the GPU without using a software
  * surface: call LoadTextureTyped() instead.
  *
- * When done with the returned surface, the app should dispose of it with a call
- * to Surface.Destroy().
- *
  * @param src an IOStream that data will be read from.
- * @param closeio true to close/free the IOStream before returning, false to
- *                leave it open.
  * @param type a filename extension that represent this data ("BMP", "GIF",
  *             "PNG", etc).
+ * @param closeio true to close/free the IOStream before returning, false
+ *                to leave it open.
  * @returns a new SDL surface, or nullptr on error.
  *
  * @since This function is available since SDL_image 3.0.0.
@@ -367,7 +374,7 @@
                                 StringParam type,
                                 bool closeio = false)
 {
-  return IMG_LoadTyped_IO(src, type, closeio);
+  return Surface(IMG_LoadTyped_IO(src, closeio, type));
 }
 
 /**
@@ -387,10 +394,7 @@
  * read; that function is Texture.Texture().
  *
  * If you would rather decode an image to an Surface (a buffer of pixels in CPU
- * memory), call Surface.Surface() instead.
- *
- * When done with the returned texture, the app should dispose of it with a call
- * to Texture.Destroy().
+ * memory), call LoadSurface() instead.
  *
  * @param renderer the Renderer to use to create the texture.
  * @param file a path on the filesystem to load an image from.
@@ -403,7 +407,7 @@
  */
 inline Texture LoadTexture(RendererParam renderer, StringParam file)
 {
-  return IMG_LoadTexture(renderer, file);
+  return Texture(IMG_LoadTexture(renderer, file));
 }
 
 /**
@@ -432,10 +436,7 @@
  * cannot autodetect the file format.
  *
  * If you would rather decode an image to an Surface (a buffer of pixels in CPU
- * memory), call Surface.Surface() instead.
- *
- * When done with the returned texture, the app should dispose of it with a call
- * to Texture.Destroy().
+ * memory), call LoadSurface() instead.
  *
  * @param renderer the Renderer to use to create the texture.
  * @param src an IOStream that data will be read from.
@@ -450,9 +451,9 @@
  */
 inline Texture LoadTexture(RendererParam renderer,
                            IOStreamParam src,
-                           bool closeio)
+                           bool closeio = false)
 {
-  return IMG_LoadTexture_IO(renderer, src, closeio);
+  return Texture(IMG_LoadTexture_IO(renderer, src, closeio));
 }
 
 inline Texture::Texture(RendererParam renderer, StringParam file)
@@ -499,28 +500,25 @@
  * If you would rather decode an image to an Surface (a buffer of pixels in CPU
  * memory), call LoadSurfaceTyped() instead.
  *
- * When done with the returned texture, the app should dispose of it with a call
- * to Texture.Destroy().
- *
  * @param renderer the Renderer to use to create the texture.
  * @param src an IOStream that data will be read from.
- * @param closeio true to close/free the IOStream before returning, false to
- *                leave it open.
  * @param type a filename extension that represent this data ("BMP", "GIF",
  *             "PNG", etc).
+ * @param closeio true to close/free the IOStream before returning, false to
+ *                leave it open.
  * @returns a new texture, or nullptr on error.
  *
  * @since This function is available since SDL_image 3.0.0.
  *
  * @sa Texture.Texture
- * @sa Texture.Texture
+ * @sa Texture.Destroy
  */
 inline Texture LoadTextureTyped(RendererParam renderer,
                                 IOStreamParam src,
                                 StringParam type,
                                 bool closeio = false)
 {
-  return IMG_LoadTextureTyped_IO(renderer, src, type, closeio);
+  return Texture(IMG_LoadTextureTyped_IO(renderer, src, closeio, type));
 }
 
 #if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
@@ -565,10 +563,6 @@
   return IMG_LoadGPUTexture(device, copy_pass, file, width, height);
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Load an image from an SDL data source into a GPU texture.
  *
@@ -621,10 +615,6 @@
   return IMG_LoadGPUTexture_IO(device, copy_pass, src, closeio, width, height);
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Load an image from an SDL data source into a GPU texture.
  *
@@ -660,10 +650,10 @@
  * @param copy_pass the GPUCopyPass to use to upload the loaded image to the GPU
  *                  texture.
  * @param src an IOStream that data will be read from.
- * @param closeio true to close/free the IOStream before returning, false to
- *                leave it open.
  * @param type a filename extension that represent this data ("BMP", "GIF",
  *             "PNG", etc).
+ * @param closeio true to close/free the IOStream before returning, false to
+ *                leave it open.
  * @param width a pointer filled in with the width of the GPU texture. may be
  *              nullptr.
  * @param height a pointer filled in with the width of the GPU texture. may be
@@ -684,13 +674,9 @@
                                       int* height = nullptr)
 {
   return IMG_LoadGPUTextureTyped_IO(
-    device, copy_pass, src, type, closeio, width, height);
+    device, copy_pass, src, closeio, type, width, height);
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Get the image currently in the clipboard.
  *
@@ -701,11 +687,7 @@
  *
  * @since This function is available since SDL_image 3.4.0.
  */
-inline Surface GetClipboardImage() { return IMG_GetClipboardImage(); }
-
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
+inline Surface GetClipboardImage() { return Surface{IMG_GetClipboardImage()}; }
 
 /**
  * Detect ANI animated cursor data on a readable/seekable IOStream.
@@ -727,7 +709,7 @@
  * @param src a seekable/readable IOStream to provide image data.
  * @returns true if this is ANI animated cursor data, false otherwise.
  *
- * @since This function is available since SDL_image 3.0.0.
+ * @since This function is available since SDL_image 3.4.0.
  *
  * @sa isAVIF
  * @sa isBMP
@@ -1558,7 +1540,10 @@
  * @sa LoadXPM
  * @sa LoadXV
  */
-inline Surface LoadAVIF(IOStreamParam src) { return IMG_LoadAVIF_IO(src); }
+inline Surface LoadAVIF(IOStreamParam src)
+{
+  return Surface(IMG_LoadAVIF_IO(src));
+}
 
 /**
  * Load a BMP image directly.
@@ -1592,7 +1577,10 @@
  * @sa LoadXPM
  * @sa LoadXV
  */
-inline Surface LoadBMP(IOStreamParam src) { return IMG_LoadBMP_IO(src); }
+inline Surface LoadBMP(IOStreamParam src)
+{
+  return Surface{IMG_LoadBMP_IO(src)};
+}
 
 /**
  * Load a CUR image directly.
@@ -1626,7 +1614,10 @@
  * @sa LoadXPM
  * @sa LoadXV
  */
-inline Surface LoadCUR(IOStreamParam src) { return IMG_LoadCUR_IO(src); }
+inline Surface LoadCUR(IOStreamParam src)
+{
+  return Surface{IMG_LoadCUR_IO(src)};
+}
 
 /**
  * Load a GIF image directly.
@@ -1660,7 +1651,10 @@
  * @sa LoadXPM
  * @sa LoadXV
  */
-inline Surface LoadGIF(IOStreamParam src) { return IMG_LoadGIF_IO(src); }
+inline Surface LoadGIF(IOStreamParam src)
+{
+  return Surface{IMG_LoadGIF_IO(src)};
+}
 
 /**
  * Load a ICO image directly.
@@ -1694,7 +1688,10 @@
  * @sa LoadXPM
  * @sa LoadXV
  */
-inline Surface LoadICO(IOStreamParam src) { return IMG_LoadICO_IO(src); }
+inline Surface LoadICO(IOStreamParam src)
+{
+  return Surface{IMG_LoadICO_IO(src)};
+}
 
 /**
  * Load a JPG image directly.
@@ -1728,7 +1725,10 @@
  * @sa LoadXPM
  * @sa LoadXV
  */
-inline Surface LoadJPG(IOStreamParam src) { return IMG_LoadJPG_IO(src); }
+inline Surface LoadJPG(IOStreamParam src)
+{
+  return Surface{IMG_LoadJPG_IO(src)};
+}
 
 /**
  * Load a JXL image directly.
@@ -1762,7 +1762,10 @@
  * @sa LoadXPM
  * @sa LoadXV
  */
-inline Surface LoadJXL(IOStreamParam src) { return IMG_LoadJXL_IO(src); }
+inline Surface LoadJXL(IOStreamParam src)
+{
+  return Surface{IMG_LoadJXL_IO(src)};
+}
 
 /**
  * Load a LBM image directly.
@@ -1796,7 +1799,10 @@
  * @sa LoadXPM
  * @sa LoadXV
  */
-inline Surface LoadLBM(IOStreamParam src) { return IMG_LoadLBM_IO(src); }
+inline Surface LoadLBM(IOStreamParam src)
+{
+  return Surface{IMG_LoadLBM_IO(src)};
+}
 
 /**
  * Load a PCX image directly.
@@ -1830,7 +1836,10 @@
  * @sa LoadXPM
  * @sa LoadXV
  */
-inline Surface LoadPCX(IOStreamParam src) { return IMG_LoadPCX_IO(src); }
+inline Surface LoadPCX(IOStreamParam src)
+{
+  return Surface{IMG_LoadPCX_IO(src)};
+}
 
 /**
  * Load a PNG image directly.
@@ -1864,7 +1873,10 @@
  * @sa LoadXPM
  * @sa LoadXV
  */
-inline Surface LoadPNG(IOStreamParam src) { return IMG_LoadPNG_IO(src); }
+inline Surface LoadPNG(IOStreamParam src)
+{
+  return Surface{IMG_LoadPNG_IO(src)};
+}
 
 /**
  * Load a PNM image directly.
@@ -1898,7 +1910,10 @@
  * @sa LoadXPM
  * @sa LoadXV
  */
-inline Surface LoadPNM(IOStreamParam src) { return IMG_LoadPNM_IO(src); }
+inline Surface LoadPNM(IOStreamParam src)
+{
+  return Surface{IMG_LoadPNM_IO(src)};
+}
 
 /**
  * Load a SVG image directly.
@@ -1933,7 +1948,10 @@
  * @sa LoadXPM
  * @sa LoadXV
  */
-inline Surface LoadSVG(IOStreamParam src) { return IMG_LoadSVG_IO(src); }
+inline Surface LoadSVG(IOStreamParam src)
+{
+  return Surface{IMG_LoadSVG_IO(src)};
+}
 
 /**
  * Load an SVG image, scaled to a specific size.
@@ -1948,8 +1966,7 @@
  * to Surface.Destroy().
  *
  * @param src an IOStream to load SVG data from.
- * @param width desired width of the generated surface, in pixels.
- * @param height desired height of the generated surface, in pixels.
+ * @param size desired width and height of the generated surface, in pixels.
  * @returns a new SDL surface, or nullptr on error.
  *
  * @since This function is available since SDL_image 3.0.0.
@@ -1958,7 +1975,7 @@
  */
 inline Surface LoadSizedSVG(IOStreamParam src, const PointRaw& size)
 {
-  return IMG_LoadSizedSVG_IO(src, size);
+  return Surface{IMG_LoadSizedSVG_IO(src, size.x, size.y)};
 }
 
 /**
@@ -1993,7 +2010,10 @@
  * @sa LoadXPM
  * @sa LoadXV
  */
-inline Surface LoadQOI(IOStreamParam src) { return IMG_LoadQOI_IO(src); }
+inline Surface LoadQOI(IOStreamParam src)
+{
+  return Surface{IMG_LoadQOI_IO(src)};
+}
 
 /**
  * Load a TGA image directly.
@@ -2027,7 +2047,10 @@
  * @sa LoadXPM
  * @sa LoadXV
  */
-inline Surface LoadTGA(IOStreamParam src) { return IMG_LoadTGA_IO(src); }
+inline Surface LoadTGA(IOStreamParam src)
+{
+  return Surface{IMG_LoadTGA_IO(src)};
+}
 
 /**
  * Load a TIFF image directly.
@@ -2061,7 +2084,10 @@
  * @sa LoadXPM
  * @sa LoadXV
  */
-inline Surface LoadTIF(IOStreamParam src) { return IMG_LoadTIF_IO(src); }
+inline Surface LoadTIF(IOStreamParam src)
+{
+  return Surface{IMG_LoadTIF_IO(src)};
+}
 
 /**
  * Load a WEBP image directly.
@@ -2095,7 +2121,10 @@
  * @sa LoadXPM
  * @sa LoadXV
  */
-inline Surface LoadWEBP(IOStreamParam src) { return IMG_LoadWEBP_IO(src); }
+inline Surface LoadWEBP(IOStreamParam src)
+{
+  return Surface{IMG_LoadWEBP_IO(src)};
+}
 
 /**
  * Load a XCF image directly.
@@ -2129,7 +2158,10 @@
  * @sa LoadXPM
  * @sa LoadXV
  */
-inline Surface LoadXCF(IOStreamParam src) { return IMG_LoadXCF_IO(src); }
+inline Surface LoadXCF(IOStreamParam src)
+{
+  return Surface{IMG_LoadXCF_IO(src)};
+}
 
 /**
  * Load a XPM image directly.
@@ -2163,7 +2195,10 @@
  * @sa LoadXCF
  * @sa LoadXV
  */
-inline Surface LoadXPM(IOStreamParam src) { return IMG_LoadXPM_IO(src); }
+inline Surface LoadXPM(IOStreamParam src)
+{
+  return Surface{IMG_LoadXPM_IO(src)};
+}
 
 /**
  * Load a XV image directly.
@@ -2197,7 +2232,7 @@
  * @sa LoadXCF
  * @sa LoadXPM
  */
-inline Surface LoadXV(IOStreamParam src) { return IMG_LoadXV_IO(src); }
+inline Surface LoadXV(IOStreamParam src) { return Surface{IMG_LoadXV_IO(src)}; }
 
 /**
  * Load an XPM image from a memory array.
@@ -2206,9 +2241,6 @@
  * it will be 32bpp. If you always want 32-bit data, use
  * ReadXPMFromArrayToRGB888() instead.
  *
- * When done with the returned surface, the app should dispose of it with a call
- * to Surface.Destroy().
- *
  * @param xpm a null-terminated array of strings that comprise XPM data.
  * @returns a new SDL surface, or nullptr on error.
  *
@@ -2218,7 +2250,7 @@
  */
 inline Surface ReadXPMFromArray(char** xpm)
 {
-  return IMG_ReadXPMFromArray(xpm);
+  return Surface{IMG_ReadXPMFromArray(xpm)};
 }
 
 /**
@@ -2227,9 +2259,6 @@
  * The returned surface will always be a 32-bit RGB surface. If you want 8-bit
  * indexed colors (and the XPM data allows it), use ReadXPMFromArray() instead.
  *
- * When done with the returned surface, the app should dispose of it with a call
- * to Surface.Destroy().
- *
  * @param xpm a null-terminated array of strings that comprise XPM data.
  * @returns a new SDL surface, or nullptr on error.
  *
@@ -2239,7 +2268,7 @@
  */
 inline Surface ReadXPMFromArrayToRGB888(char** xpm)
 {
-  return IMG_ReadXPMFromArrayToRGB888(xpm);
+  return Surface{IMG_ReadXPMFromArrayToRGB888(xpm)};
 }
 
 #if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
@@ -2273,10 +2302,6 @@
   CheckError(IMG_Save(surface, file));
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Save an Surface into formatted image data, via an IOStream.
  *
@@ -2289,10 +2314,10 @@
  *
  * @param surface the SDL surface to save.
  * @param dst the IOStream to save the image data to.
- * @param closeio true to close/free the IOStream before returning, false to
- *                leave it open.
  * @param type a filename extension that represent this data ("BMP", "GIF",
  *             "PNG", etc).
+ * @param closeio true to close/free the IOStream before returning, false to
+ *                leave it open.
  * @throws Error on failure.
  *
  * @since This function is available since SDL_image 3.4.0.
@@ -2313,7 +2338,7 @@
                       StringParam type,
                       bool closeio = false)
 {
-  CheckError(IMG_SaveTyped_IO(surface, dst, type, closeio));
+  CheckError(IMG_SaveTyped_IO(surface, dst, closeio, type));
 }
 
 #endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
@@ -2348,10 +2373,10 @@
  *
  * @param surface the SDL surface to save.
  * @param dst the IOStream to save the image data to.
- * @param closeio true to close/free the IOStream before returning, false to
- *                leave it open.
  * @param quality the desired quality, ranging between 0 (lowest) and 100
  *                (highest).
+ * @param closeio true to close/free the IOStream before returning, false to
+ *                leave it open.
  * @throws Error on failure.
  *
  * @since This function is available since SDL_image 3.0.0.
@@ -2363,7 +2388,7 @@
                      int quality,
                      bool closeio = false)
 {
-  CheckError(IMG_SaveAVIF_IO(surface, dst, quality, closeio));
+  CheckError(IMG_SaveAVIF_IO(surface, dst, closeio, quality));
 }
 
 #if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
@@ -2386,10 +2411,6 @@
   CheckError(IMG_SaveBMP(surface, file));
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Save an Surface into BMP image data, via an IOStream.
  *
@@ -2415,10 +2436,6 @@
   CheckError(IMG_SaveBMP_IO(surface, dst, closeio));
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Save an Surface into a CUR image file.
  *
@@ -2437,10 +2454,6 @@
   CheckError(IMG_SaveCUR(surface, file));
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Save an Surface into CUR image data, via an IOStream.
  *
@@ -2466,10 +2479,6 @@
   CheckError(IMG_SaveCUR_IO(surface, dst, closeio));
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Save an Surface into a GIF image file.
  *
@@ -2488,10 +2497,6 @@
   CheckError(IMG_SaveGIF(surface, file));
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Save an Surface into GIF image data, via an IOStream.
  *
@@ -2517,10 +2522,6 @@
   CheckError(IMG_SaveGIF_IO(surface, dst, closeio));
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Save an Surface into a ICO image file.
  *
@@ -2539,10 +2540,6 @@
   CheckError(IMG_SaveICO(surface, file));
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Save an Surface into ICO image data, via an IOStream.
  *
@@ -2600,10 +2597,10 @@
  *
  * @param surface the SDL surface to save.
  * @param dst the IOStream to save the image data to.
- * @param closeio true to close/free the IOStream before returning, false to
- *                leave it open.
  * @param quality [0; 33] is Lowest quality, [34; 66] is Middle quality, [67;
  *                100] is Highest quality.
+ * @param closeio true to close/free the IOStream before returning, false to
+ *                leave it open.
  * @throws Error on failure.
  *
  * @since This function is available since SDL_image 3.0.0.
@@ -2615,7 +2612,7 @@
                     int quality,
                     bool closeio = false)
 {
-  CheckError(IMG_SaveJPG_IO(surface, dst, quality, closeio));
+  CheckError(IMG_SaveJPG_IO(surface, dst, closeio, quality));
 }
 
 /**
@@ -2681,10 +2678,6 @@
   CheckError(IMG_SaveTGA(surface, file));
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Save an Surface into TGA image data, via an IOStream.
  *
@@ -2710,10 +2703,6 @@
   CheckError(IMG_SaveTGA_IO(surface, dst, closeio));
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Save an Surface into a WEBP image file.
  *
@@ -2736,10 +2725,6 @@
   CheckError(IMG_SaveWEBP(surface, file, quality));
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Save an Surface into WEBP image data, via an IOStream.
  *
@@ -2750,12 +2735,12 @@
  *
  * @param surface the SDL surface to save.
  * @param dst the IOStream to save the image data to.
- * @param closeio true to close/free the IOStream before returning, false to
- *                leave it open.
  * @param quality between 0 and 100. For lossy, 0 gives the smallest size and
  *                100 the largest. For lossless, this parameter is the amount of
  *                effort put into the compression: 0 is the fastest but gives
  *                larger files compared to the slowest, but best, 100.
+ * @param closeio true to close/free the IOStream before returning, false to
+ *                leave it open.
  * @throws Error on failure.
  *
  * @since This function is available since SDL_image 3.4.0.
@@ -2767,7 +2752,7 @@
                      float quality,
                      bool closeio = false)
 {
-  CheckError(IMG_SaveWEBP_IO(surface, dst, quality, closeio));
+  CheckError(IMG_SaveWEBP_IO(surface, dst, closeio, quality));
 }
 
 #endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
@@ -2868,7 +2853,7 @@
    * @sa LoadWEBPAnimation
    * @sa Animation.Free
    */
-  Animation(IOStreamParam src, bool closeio)
+  Animation(IOStreamParam src, bool closeio = false)
     : m_resource(IMG_LoadAnimation_IO(src, closeio))
   {
   }
@@ -2935,16 +2920,30 @@
    */
   void Free();
 
+  /// Get the width in pixels.
   int GetWidth() const;
 
+  /// Get the height in pixels.
   int GetHeight() const;
 
+  /// Get the size in pixels.
   Point GetSize() const;
 
+  /// Return the number of frames.
   int GetCount() const;
 
+  /**
+   * Return the frame image under given index.
+   *
+   * @param index the index to get frame, within [0, GetCount() - 1]
+   */
   Surface GetFrame(int index) const;
 
+  /**
+   * Return the frame delay under given index.
+   *
+   * @param index the index to get frame, within [0, GetCount() - 1]
+   */
   int GetDelay(int index) const;
 
 #if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
@@ -2968,10 +2967,6 @@
    */
   void Save(StringParam file);
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
   /**
    * Save an animation to an IOStream.
    *
@@ -2984,9 +2979,9 @@
    * For formats that accept a quality, a default quality of 90 will be used.
    *
    * @param dst an IOStream that data will be written to.
+   * @param type a filename extension that represent this data ("GIF", etc).
    * @param closeio true to close/free the IOStream before returning, false to
    *                leave it open.
-   * @param type a filename extension that represent this data ("GIF", etc).
    * @throws Error on failure.
    *
    * @since This function is available since SDL_image 3.4.0.
@@ -3000,10 +2995,6 @@
    */
   void SaveTyped(IOStreamParam dst, StringParam type, bool closeio = false);
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
   /**
    * Save an animation in ANI format to an IOStream.
    *
@@ -3026,10 +3017,6 @@
    */
   void SaveANI(IOStreamParam dst, bool closeio = false);
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
   /**
    * Save an animation in APNG format to an IOStream.
    *
@@ -3052,10 +3039,6 @@
    */
   void SaveAPNG(IOStreamParam dst, bool closeio = false);
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
   /**
    * Save an animation in AVIF format to an IOStream.
    *
@@ -3063,10 +3046,10 @@
    * function succeeds or not.
    *
    * @param dst an IOStream from which data will be written to.
-   * @param closeio true to close/free the IOStream before returning, false to
-   *                leave it open.
    * @param quality the desired quality, ranging between 0 (lowest) and 100
    *                (highest).
+   * @param closeio true to close/free the IOStream before returning, false to
+   *                leave it open.
    * @throws Error on failure.
    *
    * @since This function is available since SDL_image 3.4.0.
@@ -3080,10 +3063,6 @@
    */
   void SaveAVIF(IOStreamParam dst, int quality, bool closeio = false);
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
   /**
    * Save an animation in GIF format to an IOStream.
    *
@@ -3106,10 +3085,6 @@
    */
   void SaveGIF(IOStreamParam dst, bool closeio = false);
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
   /**
    * Save an animation in WEBP format to an IOStream.
    *
@@ -3117,12 +3092,12 @@
    * function succeeds or not.
    *
    * @param dst an IOStream from which data will be written to.
-   * @param closeio true to close/free the IOStream before returning, false to
-   *                leave it open.
    * @param quality between 0 and 100. For lossy, 0 gives the smallest size and
    *                100 the largest. For lossless, this parameter is the amount
    *                of effort put into the compression: 0 is the fastest but
    *                gives larger files compared to the slowest, but best, 100.
+   * @param closeio true to close/free the IOStream before returning, false to
+   *                leave it open.
    * @throws Error on failure.
    *
    * @since This function is available since SDL_image 3.4.0.
@@ -3136,15 +3111,10 @@
    */
   void SaveWEBP(IOStreamParam dst, int quality, bool closeio = false);
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
   /**
    * Create an animated cursor from an animation.
    *
-   * @param hot_x the x position of the cursor hot spot.
-   * @param hot_y the y position of the cursor hot spot.
+   * @param hotspot the x, y position of the cursor hot spot.
    * @returns the new cursor on success.
    * @throws Error on failure.
    *
@@ -3195,29 +3165,26 @@
   ~AnimationRef() { release(); }
 };
 
-inline int GetAnimationWidth(AnimationConstParam anim)
-{
-  static_assert(false, "Not implemented");
-}
+/// Get the width in pixels.
+inline int GetAnimationWidth(AnimationConstParam anim) { return anim->w; }
 
 inline int Animation::GetWidth() const
 {
   return SDL::GetAnimationWidth(m_resource);
 }
 
-inline int GetAnimationHeight(AnimationConstParam anim)
-{
-  static_assert(false, "Not implemented");
-}
+/// Get the height in pixels.
+inline int GetAnimationHeight(AnimationConstParam anim) { return anim->h; }
 
 inline int Animation::GetHeight() const
 {
   return SDL::GetAnimationHeight(m_resource);
 }
 
+/// Get the size in pixels.
 inline Point GetAnimationSize(AnimationConstParam anim)
 {
-  static_assert(false, "Not implemented");
+  return {anim->w, anim->h};
 }
 
 inline Point Animation::GetSize() const
@@ -3225,19 +3192,23 @@
   return SDL::GetAnimationSize(m_resource);
 }
 
-inline int GetAnimationCount(AnimationConstParam anim)
-{
-  static_assert(false, "Not implemented");
-}
+/// Return the number of frames.
+inline int GetAnimationCount(AnimationConstParam anim) { return anim->count; }
 
 inline int Animation::GetCount() const
 {
   return SDL::GetAnimationCount(m_resource);
 }
 
+/**
+ * Return the frame image under given index.
+ *
+ * @param anim Animation to dispose of.
+ * @param index the index to get frame, within [0, GetCount() - 1]
+ */
 inline Surface GetAnimationFrame(AnimationConstParam anim, int index)
 {
-  static_assert(false, "Not implemented");
+  return Surface::Borrow(anim->frames[index]);
 }
 
 inline Surface Animation::GetFrame(int index) const
@@ -3245,9 +3216,15 @@
   return SDL::GetAnimationFrame(m_resource, index);
 }
 
+/**
+ * Return the frame delay under given index.
+ *
+ * @param anim Animation to dispose of.
+ * @param index the index to get frame, within [0, GetCount() - 1]
+ */
 inline int GetAnimationDelay(AnimationConstParam anim, int index)
 {
-  static_assert(false, "Not implemented");
+  return anim->delays[index];
 }
 
 inline int Animation::GetDelay(int index) const
@@ -3308,7 +3285,7 @@
  * @sa LoadWEBPAnimation
  * @sa Animation.Free
  */
-inline Animation LoadAnimation(IOStreamParam src, bool closeio)
+inline Animation LoadAnimation(IOStreamParam src, bool closeio = false)
 {
   return Animation(src, closeio);
 }
@@ -3330,9 +3307,9 @@
  * call to Animation.Free().
  *
  * @param src an IOStream that data will be read from.
+ * @param type a filename extension that represent this data ("GIF", etc).
  * @param closeio true to close/free the IOStream before returning, false to
  *                leave it open.
- * @param type a filename extension that represent this data ("GIF", etc).
  * @returns a new Animation, or nullptr on error.
  *
  * @since This function is available since SDL_image 3.0.0.
@@ -3351,7 +3328,7 @@
                                     StringParam type,
                                     bool closeio = false)
 {
-  return IMG_LoadAnimationTyped_IO(src, type, closeio);
+  return Animation(IMG_LoadAnimationTyped_IO(src, closeio, type));
 }
 
 #if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
@@ -3384,13 +3361,9 @@
  */
 inline Animation LoadANIAnimation(IOStreamParam src)
 {
-  return IMG_LoadANIAnimation_IO(src);
+  return Animation(IMG_LoadANIAnimation_IO(src));
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Load an APNG animation directly from an IOStream.
  *
@@ -3419,13 +3392,9 @@
  */
 inline Animation LoadAPNGAnimation(IOStreamParam src)
 {
-  return IMG_LoadAPNGAnimation_IO(src);
+  return Animation(IMG_LoadAPNGAnimation_IO(src));
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Load an AVIF animation directly from an IOStream.
  *
@@ -3454,7 +3423,7 @@
  */
 inline Animation LoadAVIFAnimation(IOStreamParam src)
 {
-  return IMG_LoadAVIFAnimation_IO(src);
+  return Animation(IMG_LoadAVIFAnimation_IO(src));
 }
 
 #endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
@@ -3484,7 +3453,7 @@
  */
 inline Animation LoadGIFAnimation(IOStreamParam src)
 {
-  return IMG_LoadGIFAnimation_IO(src);
+  return Animation(IMG_LoadGIFAnimation_IO(src));
 }
 
 /**
@@ -3512,7 +3481,7 @@
  */
 inline Animation LoadWEBPAnimation(IOStreamParam src)
 {
-  return IMG_LoadWEBPAnimation_IO(src);
+  return Animation{IMG_LoadWEBPAnimation_IO(src)};
 }
 
 #if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
@@ -3540,19 +3509,11 @@
   CheckError(IMG_SaveAnimation(anim, file));
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 inline void Animation::Save(StringParam file)
 {
   SDL::SaveAnimation(m_resource, std::move(file));
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Save an animation to an IOStream.
  *
@@ -3584,13 +3545,9 @@
                                StringParam type,
                                bool closeio = false)
 {
-  CheckError(IMG_SaveAnimationTyped_IO(anim, dst, type, closeio));
+  CheckError(IMG_SaveAnimationTyped_IO(anim, dst, closeio, type));
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 inline void Animation::SaveTyped(IOStreamParam dst,
                                  StringParam type,
                                  bool closeio)
@@ -3598,10 +3555,6 @@
   SDL::SaveAnimationTyped(m_resource, dst, std::move(type), closeio);
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Save an animation in ANI format to an IOStream.
  *
@@ -3630,19 +3583,11 @@
   CheckError(IMG_SaveANIAnimation_IO(anim, dst, closeio));
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 inline void Animation::SaveANI(IOStreamParam dst, bool closeio)
 {
   SDL::SaveANIAnimation(m_resource, dst, closeio);
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Save an animation in APNG format to an IOStream.
  *
@@ -3671,19 +3616,11 @@
   CheckError(IMG_SaveAPNGAnimation_IO(anim, dst, closeio));
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 inline void Animation::SaveAPNG(IOStreamParam dst, bool closeio)
 {
   SDL::SaveAPNGAnimation(m_resource, dst, closeio);
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Save an animation in AVIF format to an IOStream.
  *
@@ -3692,10 +3629,10 @@
  *
  * @param anim the animation to save.
  * @param dst an IOStream from which data will be written to.
- * @param closeio true to close/free the IOStream before returning, false to
- *                leave it open.
  * @param quality the desired quality, ranging between 0 (lowest) and 100
  *                (highest).
+ * @param closeio true to close/free the IOStream before returning, false to
+ *                leave it open.
  * @throws Error on failure.
  *
  * @since This function is available since SDL_image 3.4.0.
@@ -3715,19 +3652,11 @@
   CheckError(IMG_SaveAVIFAnimation_IO(anim, dst, quality, closeio));
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 inline void Animation::SaveAVIF(IOStreamParam dst, int quality, bool closeio)
 {
   SDL::SaveAVIFAnimation(m_resource, dst, quality, closeio);
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Save an animation in GIF format to an IOStream.
  *
@@ -3756,19 +3685,11 @@
   CheckError(IMG_SaveGIFAnimation_IO(anim, dst, closeio));
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 inline void Animation::SaveGIF(IOStreamParam dst, bool closeio)
 {
   SDL::SaveGIFAnimation(m_resource, dst, closeio);
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Save an animation in WEBP format to an IOStream.
  *
@@ -3777,12 +3698,12 @@
  *
  * @param anim the animation to save.
  * @param dst an IOStream from which data will be written to.
- * @param closeio true to close/free the IOStream before returning, false to
- *                leave it open.
  * @param quality between 0 and 100. For lossy, 0 gives the smallest size and
  *                100 the largest. For lossless, this parameter is the amount of
  *                effort put into the compression: 0 is the fastest but gives
  *                larger files compared to the slowest, but best, 100.
+ * @param closeio true to close/free the IOStream before returning, false to
+ *                leave it open.
  * @throws Error on failure.
  *
  * @since This function is available since SDL_image 3.4.0.
@@ -3802,25 +3723,16 @@
   CheckError(IMG_SaveWEBPAnimation_IO(anim, dst, quality, closeio));
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 inline void Animation::SaveWEBP(IOStreamParam dst, int quality, bool closeio)
 {
   SDL::SaveWEBPAnimation(m_resource, dst, quality, closeio);
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Create an animated cursor from an animation.
  *
  * @param anim an animation to use to create an animated cursor.
- * @param hot_x the x position of the cursor hot spot.
- * @param hot_y the y position of the cursor hot spot.
+ * @param hotspot the x, y position of the cursor hot spot.
  * @returns the new cursor on success.
  * @throws Error on failure.
  *
@@ -3832,13 +3744,10 @@
  */
 inline Cursor CreateAnimatedCursor(AnimationParam anim, const PointRaw& hotspot)
 {
-  return CheckError(IMG_CreateAnimatedCursor(anim, hotspot));
+  return Cursor{
+    CheckError(IMG_CreateAnimatedCursor(anim, hotspot.x, hotspot.y))};
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 inline Cursor Animation::CreateCursor(const PointRaw& hotspot)
 {
   return SDL::CreateAnimatedCursor(m_resource, hotspot);
@@ -3856,7 +3765,6 @@
  * @since This function is available since SDL_image 3.0.0.
  *
  * @sa Animation.Animation
- * @sa Animation.Animation
  * @sa LoadAnimationTyped
  * @sa LoadANIAnimation
  * @sa LoadAPNGAnimation
@@ -3868,6 +3776,8 @@
 
 inline void Animation::Free() { FreeAnimation(release()); }
 
+#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
+
 /**
  * An object representing the encoder context.
  *
@@ -3912,8 +3822,6 @@
 
   constexpr AnimationEncoder(AnimationEncoderRef&& other) = delete;
 
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
   /**
    * Create an encoder to save a series of images to a file.
    *
@@ -3944,10 +3852,6 @@
   {
   }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
   /**
    * Create an encoder to save a series of images to an IOStream.
    *
@@ -3964,9 +3868,9 @@
    * succeeds.
    *
    * @param dst an IOStream that will be used to save the stream.
+   * @param type a filename extension that represent this data ("WEBP", etc).
    * @param closeio true to close the IOStream when done, false to leave it
    *                open.
-   * @param type a filename extension that represent this data ("WEBP", etc).
    * @post a new AnimationEncoder, or nullptr on failure; call GetError() for
    *       more information.
    *
@@ -3978,14 +3882,10 @@
    * @sa AnimationEncoder.Close
    */
   AnimationEncoder(IOStreamParam dst, StringParam type, bool closeio = false)
-    : m_resource(IMG_CreateAnimationEncoder_IO(dst, type, closeio))
+    : m_resource(IMG_CreateAnimationEncoder_IO(dst, closeio, type))
   {
   }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
   /**
    * Create an animation encoder with the specified properties.
    *
@@ -4038,8 +3938,6 @@
   {
   }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
   /// Destructor
   ~AnimationEncoder() { IMG_CloseAnimationEncoder(m_resource); }
 
@@ -4080,16 +3978,13 @@
     return {m_resource};
   }
 
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
   /**
    * Close an animation encoder, finishing any encoding.
    *
    * Calling this function frees the animation encoder, and returns the final
    * status of the encoding process.
    *
-   * @returns true on success or false on failure; call GetError() for more
-   *          information.
+   * @throws Error on failure.
    *
    * @since This function is available since SDL_image 3.4.0.
    *
@@ -4097,11 +3992,7 @@
    * @sa AnimationEncoder.AnimationEncoder
    * @sa AnimationEncoder.AnimationEncoder
    */
-  bool Close();
-
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
+  void Close();
 
   /**
    * Add a frame to an animation encoder.
@@ -4121,8 +4012,6 @@
    * @sa AnimationEncoder.Close
    */
   void AddFrame(SurfaceParam surface, Uint64 duration);
-
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
 };
 
 /// Semi-safe reference for AnimationEncoder.
@@ -4162,8 +4051,6 @@
   ~AnimationEncoderRef() { release(); }
 };
 
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Create an encoder to save a series of images to a file.
  *
@@ -4194,10 +4081,6 @@
   return AnimationEncoder(std::move(file));
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Create an encoder to save a series of images to an IOStream.
  *
@@ -4213,8 +4096,8 @@
  * fails, or when the animation encoder is closed if this function succeeds.
  *
  * @param dst an IOStream that will be used to save the stream.
- * @param closeio true to close the IOStream when done, false to leave it open.
  * @param type a filename extension that represent this data ("WEBP", etc).
+ * @param closeio true to close the IOStream when done, false to leave it open.
  * @returns a new AnimationEncoder, or nullptr on failure; call GetError() for
  *          more information.
  *
@@ -4232,10 +4115,6 @@
   return AnimationEncoder(dst, std::move(type), closeio);
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Create an animation encoder with the specified properties.
  *
@@ -4288,8 +4167,6 @@
   return AnimationEncoder(props);
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 namespace prop::AnimationEncoder {
 
 constexpr auto CREATE_FILENAME_STRING =
@@ -4324,8 +4201,6 @@
 
 } // namespace prop::AnimationEncoder
 
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Add a frame to an animation encoder.
  *
@@ -4351,19 +4226,11 @@
   CheckError(IMG_AddAnimationEncoderFrame(encoder, surface, duration));
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 inline void AnimationEncoder::AddFrame(SurfaceParam surface, Uint64 duration)
 {
   SDL::AddAnimationEncoderFrame(m_resource, surface, duration);
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Close an animation encoder, finishing any encoding.
  *
@@ -4384,18 +4251,7 @@
   CheckError(IMG_CloseAnimationEncoder(encoder));
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-inline bool AnimationEncoder::Close()
-{
-  return CloseAnimationEncoder(release());
-}
-
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
+inline void AnimationEncoder::Close() { CloseAnimationEncoder(release()); }
 
 /**
  * An enum representing the status of an animation decoder.
@@ -4404,8 +4260,6 @@
  */
 using AnimationDecoderStatus = IMG_AnimationDecoderStatus;
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 constexpr AnimationDecoderStatus DECODER_STATUS_INVALID =
   IMG_DECODER_STATUS_INVALID; ///< The decoder is invalid
 
@@ -4463,8 +4317,6 @@
 
   constexpr AnimationDecoder(AnimationDecoderRef&& other) = delete;
 
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
   /**
    * Create a decoder to read a series of images from a file.
    *
@@ -4496,10 +4348,6 @@
   {
   }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
   /**
    * Create a decoder to read a series of images from an IOStream.
    *
@@ -4531,14 +4379,10 @@
    * @sa AnimationDecoder.Close
    */
   AnimationDecoder(IOStreamParam src, StringParam type, bool closeio = false)
-    : m_resource(IMG_CreateAnimationDecoder_IO(src, type, closeio))
+    : m_resource(IMG_CreateAnimationDecoder_IO(src, closeio, type))
   {
   }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
   /**
    * Create an animation decoder with the specified properties.
    *
@@ -4582,8 +4426,6 @@
   {
   }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
   /// Destructor
   ~AnimationDecoder() { IMG_CloseAnimationDecoder(m_resource); }
 
@@ -4624,16 +4466,13 @@
     return {m_resource};
   }
 
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
   /**
    * Close an animation decoder, finishing any decoding.
    *
    * Calling this function frees the animation decoder, and returns the final
    * status of the decoding process.
    *
-   * @returns true on success or false on failure; call GetError() for more
-   *          information.
+   * @throws Error on failure.
    *
    * @since This function is available since SDL_image 3.4.0.
    *
@@ -4641,11 +4480,7 @@
    * @sa AnimationDecoder.AnimationDecoder
    * @sa AnimationDecoder.AnimationDecoder
    */
-  bool Close();
-
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
+  void Close();
 
   /**
    * Get the properties of an animation decoder.
@@ -4665,10 +4500,6 @@
    */
   PropertiesRef GetProperties();
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
   /**
    * Get the next frame in an animation decoder.
    *
@@ -4676,18 +4507,11 @@
    * as an Surface. The returned surface should be freed with SDL_FreeSurface()
    * when no longer needed.
    *
-   * If the animation decoder has no more frames or an error occurred while
-   * decoding the frame, this function returns false. In that case, please call
-   * GetError() for more information. If GetError() returns an empty string,
-   * that means there are no more available frames. If GetError() returns a
-   * valid string, that means the decoding failed.
-   *
-   * @param frame a pointer filled in with the Surface for the next frame in the
-   *              animation.
    * @param duration the duration of the frame, usually in milliseconds but can
    *                 be other units if the
    *                 `prop::AnimationDecoder.CREATE_TIMEBASE_DENOMINATOR_NUMBER`
    *                 property is set when creating the decoder.
+   * @returns the Surface for the next frame in the animation.
    * @throws Error on failure.
    *
    * @since This function is available since SDL_image 3.4.0.
@@ -4701,10 +4525,6 @@
    */
   Surface GetFrame(Uint64* duration);
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
   /**
    * Get the decoder status indicating the current state of the decoder.
    *
@@ -4717,10 +4537,6 @@
    */
   AnimationDecoderStatus GetStatus();
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
   /**
    * Reset an animation decoder.
    *
@@ -4739,8 +4555,6 @@
    * @sa AnimationDecoder.Close
    */
   void Reset();
-
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
 };
 
 /// Semi-safe reference for AnimationDecoder.
@@ -4780,8 +4594,6 @@
   ~AnimationDecoderRef() { release(); }
 };
 
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Create a decoder to read a series of images from a file.
  *
@@ -4813,10 +4625,6 @@
   return AnimationDecoder(std::move(file));
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Create a decoder to read a series of images from an IOStream.
  *
@@ -4852,10 +4660,6 @@
   return AnimationDecoder(src, std::move(type), closeio);
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Create an animation decoder with the specified properties.
  *
@@ -4900,8 +4704,6 @@
   return AnimationDecoder(props);
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 namespace prop::AnimationDecoder {
 
 constexpr auto CREATE_FILENAME_STRING =
@@ -4939,8 +4741,6 @@
 
 } // namespace prop::AnimationDecoder
 
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Get the properties of an animation decoder.
  *
@@ -4964,17 +4764,11 @@
   return IMG_GetAnimationDecoderProperties(decoder);
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 inline PropertiesRef AnimationDecoder::GetProperties()
 {
   return SDL::GetAnimationDecoderProperties(m_resource);
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 namespace prop::Metadata {
 
 constexpr auto IGNORE_PROPS_BOOLEAN = IMG_PROP_METADATA_IGNORE_PROPS_BOOLEAN;
@@ -4995,28 +4789,18 @@
 
 } // namespace prop::Metadata
 
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Get the next frame in an animation decoder.
  *
  * This function decodes the next frame in the animation decoder, returning it
- * as an Surface. The returned surface should be freed with SDL_FreeSurface()
- * when no longer needed.
- *
- * If the animation decoder has no more frames or an error occurred while
- * decoding the frame, this function returns false. In that case, please call
- * GetError() for more information. If GetError() returns an empty string, that
- * means there are no more available frames. If GetError() returns a valid
- * string, that means the decoding failed.
+ * as an Surface.
  *
  * @param decoder the animation decoder.
- * @param frame a pointer filled in with the Surface for the next frame in the
- *              animation.
  * @param duration the duration of the frame, usually in milliseconds but can be
  *                 other units if the
  *                 `prop::AnimationDecoder.CREATE_TIMEBASE_DENOMINATOR_NUMBER`
  *                 property is set when creating the decoder.
+ * @returns the Surface for the next frame in the animation.
  * @throws Error on failure.
  *
  * @since This function is available since SDL_image 3.4.0.
@@ -5031,22 +4815,16 @@
 inline Surface GetAnimationDecoderFrame(AnimationDecoderParam decoder,
                                         Uint64* duration)
 {
-  return CheckError(IMG_GetAnimationDecoderFrame(decoder, duration));
+  SDL_Surface* frame = nullptr;
+  CheckError(IMG_GetAnimationDecoderFrame(decoder, &frame, duration));
+  return Surface::Borrow(frame);
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 inline Surface AnimationDecoder::GetFrame(Uint64* duration)
 {
   return SDL::GetAnimationDecoderFrame(m_resource, duration);
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Get the decoder status indicating the current state of the decoder.
  *
@@ -5064,19 +4842,11 @@
   return IMG_GetAnimationDecoderStatus(decoder);
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 inline AnimationDecoderStatus AnimationDecoder::GetStatus()
 {
   return SDL::GetAnimationDecoderStatus(m_resource);
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Reset an animation decoder.
  *
@@ -5100,19 +4870,11 @@
   CheckError(IMG_ResetAnimationDecoder(decoder));
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 inline void AnimationDecoder::Reset()
 {
   SDL::ResetAnimationDecoder(m_resource);
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
-
 /**
  * Close an animation decoder, finishing any decoding.
  *
@@ -5133,19 +4895,40 @@
   CheckError(IMG_CloseAnimationDecoder(decoder));
 }
 
+inline void AnimationDecoder::Close() { CloseAnimationDecoder(release()); }
+
 #endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
 
-#if SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
+/// @}
+
+} // namespace SDL
+
+#else // defined(SDL3PP_ENABLE_IMAGE) || defined(SDL3PP_DOC)
+
+namespace SDL {
 
-inline bool AnimationDecoder::Close()
+inline Surface::Surface(StringParam file)
+  : Surface(LoadBMP(std::move(file)))
 {
-  return CloseAnimationDecoder(release());
 }
 
-#endif // SDL_IMAGE_VERSION_ATLEAST(3, 4, 0)
+inline Surface::Surface(IOStreamParam src, bool closeio)
+  : Surface(LoadBMP(std::move(src), closeio))
+{
+}
 
-/// @}
+inline Texture::Texture(RendererParam renderer, StringParam file)
+  : Texture(std::move(renderer), Surface(std::move(file)))
+{
+}
+
+inline Texture::Texture(RendererParam renderer, IOStreamParam src, bool closeio)
+  : Texture(std::move(renderer), Surface(std::move(src), closeio))
+{
+}
 
 } // namespace SDL
 
+#endif // defined(SDL3PP_ENABLE_IMAGE) || defined(SDL3PP_DOC)
+
 #endif /* SDL3PP_IMAGE_H_ */
