--- src/generated/SDL3pp_dialog.h
+++ include/SDL3pp/SDL3pp_dialog.h
@@ -9,9 +9,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryDialog Category Dialog
- *
- * File dialog support.
+ * @defgroup CategoryDialog File Dialogs
  *
  * SDL offers file dialogs, to let users select files with native GUI
  * interfaces. There are "open" dialogs, "save" dialogs, and folder selection
@@ -159,7 +157,6 @@
  *                option, and platforms that do support it may allow the user to
  *                ignore the filters. If non-nullptr, it must remain valid at
  *                least until the callback is invoked.
- * @param nfilters the number of filters. Ignored if filters is nullptr.
  * @param default_location the default folder or file to start the dialog at,
  *                         may be nullptr. Not all platforms support this
  *                         option.
@@ -181,12 +178,17 @@
 inline void ShowOpenFileDialog(DialogFileCallback callback,
                                void* userdata,
                                WindowParam window,
-                               std::span<const DialogFileFilter> filters,
-                               StringParam default_location,
-                               bool allow_many)
+                               std::span<const DialogFileFilter> filters = {},
+                               StringParam default_location = {},
+                               bool allow_many = false)
 {
-  SDL_ShowOpenFileDialog(
-    callback, userdata, window, filters, default_location, allow_many);
+  SDL_ShowOpenFileDialog(callback,
+                         userdata,
+                         window,
+                         filters.data(),
+                         filters.size(),
+                         default_location,
+                         allow_many);
 }
 
 /**
@@ -211,8 +213,6 @@
  *
  * @param callback a function pointer to be invoked when the user selects a file
  *                 and accepts, or cancels the dialog, or an error occurs.
- * @param userdata an optional pointer to pass extra data to the callback when
- *                 it will be invoked.
  * @param window the window that the dialog should be modal for, may be nullptr.
  *               Not all platforms support this option.
  * @param filters a list of filters, may be nullptr. See the
@@ -221,7 +221,6 @@
  *                option, and platforms that do support it may allow the user to
  *                ignore the filters. If non-nullptr, it must remain valid at
  *                least until the callback is invoked.
- * @param nfilters the number of filters. Ignored if filters is nullptr.
  * @param default_location the default folder or file to start the dialog at,
  *                         may be nullptr. Not all platforms support this
  *                         option.
@@ -242,11 +241,17 @@
  */
 inline void ShowOpenFileDialog(DialogFileCB callback,
                                WindowParam window,
-                               std::span<const DialogFileFilter> filters,
-                               StringParam default_location,
-                               bool allow_many)
+                               std::span<const DialogFileFilter> filters = {},
+                               StringParam default_location = {},
+                               bool allow_many = false)
 {
-  static_assert(false, "Not implemented");
+  using Wrapper = CallbackWrapper<DialogFileCB>;
+  ShowOpenFileDialog(&Wrapper::CallOnce,
+                     Wrapper::Wrap(std::move(callback)),
+                     window,
+                     filters,
+                     std::move(default_location),
+                     allow_many);
 }
 
 /**
@@ -279,7 +284,6 @@
  *                this option, and platforms that do support it may allow the
  *                user to ignore the filters. If non-nullptr, it must remain
  *                valid at least until the callback is invoked.
- * @param nfilters the number of filters. Ignored if filters is nullptr.
  * @param default_location the default folder or file to start the dialog at,
  *                         may be nullptr. Not all platforms support this
  *                         option.
@@ -298,11 +302,16 @@
  */
 inline void ShowSaveFileDialog(DialogFileCallback callback,
                                void* userdata,
-                               WindowParam window,
-                               std::span<const DialogFileFilter> filters,
-                               StringParam default_location)
+                               WindowParam window = {},
+                               std::span<const DialogFileFilter> filters = {},
+                               StringParam default_location = {})
 {
-  SDL_ShowSaveFileDialog(callback, userdata, window, filters, default_location);
+  SDL_ShowSaveFileDialog(callback,
+                         userdata,
+                         window,
+                         filters.data(),
+                         filters.size(),
+                         default_location);
 }
 
 /**
@@ -327,15 +336,12 @@
  *
  * @param callback a function pointer to be invoked when the user selects a file
  *                 and accepts, or cancels the dialog, or an error occurs.
- * @param userdata an optional pointer to pass extra data to the callback when
- *                 it will be invoked.
  * @param window the window that the dialog should be modal for, may be nullptr.
  *               Not all platforms support this option.
  * @param filters a list of filters, may be nullptr. Not all platforms support
  *                this option, and platforms that do support it may allow the
  *                user to ignore the filters. If non-nullptr, it must remain
  *                valid at least until the callback is invoked.
- * @param nfilters the number of filters. Ignored if filters is nullptr.
  * @param default_location the default folder or file to start the dialog at,
  *                         may be nullptr. Not all platforms support this
  *                         option.
@@ -353,11 +359,16 @@
  * @sa ShowFileDialogWithProperties
  */
 inline void ShowSaveFileDialog(DialogFileCB callback,
-                               WindowParam window,
-                               std::span<const DialogFileFilter> filters,
-                               StringParam default_location)
+                               WindowParam window = {},
+                               std::span<const DialogFileFilter> filters = {},
+                               StringParam default_location = {})
 {
-  static_assert(false, "Not implemented");
+  using Wrapper = CallbackWrapper<DialogFileCB>;
+  ShowSaveFileDialog(&Wrapper::CallOnce,
+                     Wrapper::Wrap(std::move(callback)),
+                     window,
+                     filters,
+                     std::move(default_location));
 }
 
 /**
@@ -405,9 +416,9 @@
  */
 inline void ShowOpenFolderDialog(DialogFileCallback callback,
                                  void* userdata,
-                                 WindowParam window,
-                                 StringParam default_location,
-                                 bool allow_many)
+                                 WindowParam window = {},
+                                 StringParam default_location = {},
+                                 bool allow_many = false)
 {
   SDL_ShowOpenFolderDialog(
     callback, userdata, window, default_location, allow_many);
@@ -435,8 +446,6 @@
  *
  * @param callback a function pointer to be invoked when the user selects a file
  *                 and accepts, or cancels the dialog, or an error occurs.
- * @param userdata an optional pointer to pass extra data to the callback when
- *                 it will be invoked.
  * @param window the window that the dialog should be modal for, may be nullptr.
  *               Not all platforms support this option.
  * @param default_location the default folder or file to start the dialog at,
@@ -457,11 +466,16 @@
  * @sa ShowFileDialogWithProperties
  */
 inline void ShowOpenFolderDialog(DialogFileCB callback,
-                                 WindowParam window,
-                                 StringParam default_location,
-                                 bool allow_many)
+                                 WindowParam window = {},
+                                 StringParam default_location = {},
+                                 bool allow_many = false)
 {
-  static_assert(false, "Not implemented");
+  using Wrapper = CallbackWrapper<DialogFileCB>;
+  ShowOpenFolderDialog(&Wrapper::CallOnce,
+                       Wrapper::Wrap(std::move(callback)),
+                       std::move(window),
+                       std::move(default_location),
+                       allow_many);
 }
 
 /**
@@ -566,8 +580,6 @@
  * @param type the type of file dialog.
  * @param callback a function pointer to be invoked when the user selects a file
  *                 and accepts, or cancels the dialog, or an error occurs.
- * @param userdata an optional pointer to pass extra data to the callback when
- *                 it will be invoked.
  * @param props the properties to use.
  *
  * @threadsafety This function should be called only from the main thread. The
@@ -587,7 +599,9 @@
                                          DialogFileCB callback,
                                          PropertiesID props)
 {
-  static_assert(false, "Not implemented");
+  using Wrapper = CallbackWrapper<DialogFileCB>;
+  ShowFileDialogWithProperties(
+    type, &Wrapper::CallOnce, Wrapper::Wrap(std::move(callback)), props);
 }
 
 namespace prop::FileDialog {
