--- src/generated/SDL3pp_vulkan.h
+++ include/SDL3pp/SDL3pp_vulkan.h
@@ -7,7 +7,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryVulkan Category Vulkan
+ * @defgroup CategoryVulkan Vulkan Support
  *
  * Functions for creating Vulkan surfaces on SDL windows.
  *
@@ -139,9 +139,6 @@
  * used with VkInstanceCreateInfo::ppEnabledExtensionNames, for calling Vulkan's
  * vkCreateInstance API.
  *
- * You should not free the returned array; it is owned by SDL.
- *
- * @param count a pointer filled in with the number of extensions returned.
  * @returns an array of extension name strings on success.
  * @throws Error on failure.
  *
@@ -151,7 +148,9 @@
  */
 inline std::span<char const* const> Vulkan_GetInstanceExtensions()
 {
-  return CheckError(SDL_Vulkan_GetInstanceExtensions());
+  Uint32 count;
+  auto data = CheckError(SDL_Vulkan_GetInstanceExtensions(&count));
+  return std::span<char const* const>(data, size_t(count));
 }
 
 /**
