--- src/generated/SDL3pp_hidapi.h
+++ include/SDL3pp/SDL3pp_hidapi.h
@@ -9,7 +9,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryHIDAPI Category HIDAPI
+ * @defgroup CategoryHIDAPI HIDAPI
  *
  * Header file for SDL HIDAPI functions.
  *
@@ -84,26 +84,33 @@
 constexpr hid_bus_type HID_API_BUS_UNKNOWN =
   SDL_HID_API_BUS_UNKNOWN; ///< Unknown bus type
 
-constexpr hid_bus_type HID_API_BUS_USB =
-  SDL_HID_API_BUS_USB; ///< USB bus Specifications: https://usb.org/hid
+/**
+ * USB bus Specifications:
+ *
+ * - https://usb.org/hid
+ */
+constexpr hid_bus_type HID_API_BUS_USB = SDL_HID_API_BUS_USB;
 
 /**
  * Bluetooth or Bluetooth LE bus Specifications:
- * https://www.bluetooth.com/specifications/specs/human-interface-device-profile-1-1-1/
- * https://www.bluetooth.com/specifications/specs/hid-service-1-0/
- * https://www.bluetooth.com/specifications/specs/hid-over-gatt-profile-1-0/
+ *
+ * * https://www.bluetooth.com/specifications/specs/human-interface-device-profile-1-1-1/
+ * * https://www.bluetooth.com/specifications/specs/hid-service-1-0/
+ * * https://www.bluetooth.com/specifications/specs/hid-over-gatt-profile-1-0/
  */
 constexpr hid_bus_type HID_API_BUS_BLUETOOTH = SDL_HID_API_BUS_BLUETOOTH;
 
 /**
  * I2C bus Specifications:
- * https://docs.microsoft.com/previous-versions/windows/hardware/design/dn642101(v=vs.85)
+ *
+ * * https://docs.microsoft.com/previous-versions/windows/hardware/design/dn642101(v=vs.85)
  */
 constexpr hid_bus_type HID_API_BUS_I2C = SDL_HID_API_BUS_I2C;
 
 /**
  * SPI bus Specifications:
- * https://www.microsoft.com/download/details.aspx?id=103325
+ *
+ * - https://www.microsoft.com/download/details.aspx?id=103325
  */
 constexpr hid_bus_type HID_API_BUS_SPI = SDL_HID_API_BUS_SPI;
 
@@ -168,7 +175,6 @@
    * @param product_id the Product ID (PID) of the device to open.
    * @param serial_number the Serial Number of the device to open (Optionally
    *                      nullptr).
-   * @post a pointer to a HidDevice object on success.
    * @throws Error on failure.
    *
    * @since This function is available since SDL 3.2.0.
@@ -187,7 +193,6 @@
    * platform-specific path name can be used (eg: /dev/hidraw0 on Linux).
    *
    * @param path the path name of the device to open.
-   * @post a pointer to a HidDevice object on success.
    * @throws Error on failure.
    *
    * @since This function is available since SDL 3.2.0.
@@ -235,8 +240,7 @@
   /**
    * Close a HID device.
    *
-   * @returns 0 on success or a negative error code on failure; call GetError()
-   *          for more information.
+   * @throws Error on failure.
    *
    * @since This function is available since SDL 3.2.0.
    */
@@ -279,7 +283,6 @@
    *
    * @param data the data to send, including the report number as the first
    *             byte.
-   * @param length the length in bytes of the data to send.
    * @returns the actual number of bytes written and -1 on on failure; call
    *          GetError() for more information.
    *
@@ -295,10 +298,7 @@
    * reports.
    *
    * @param data a buffer to put the read data into.
-   * @param length the number of bytes to read. For devices with multiple
-   *               reports, make sure to read an extra byte for the report
-   *               number.
-   * @param milliseconds timeout in milliseconds or -1 for blocking wait.
+   * @param timeout timeout in milliseconds
    * @returns the actual number of bytes read and -1 on on failure; call
    *          GetError() for more information. If no packet was available to be
    *          read within the timeout period, this function returns 0.
@@ -315,9 +315,6 @@
    * reports.
    *
    * @param data a buffer to put the read data into.
-   * @param length the number of bytes to read. For devices with multiple
-   *               reports, make sure to read an extra byte for the report
-   *               number.
    * @returns the actual number of bytes read and -1 on failure; call GetError()
    *          for more information. If no packet was available to be read and
    *          the handle is in non-blocking mode, this function returns 0.
@@ -336,9 +333,8 @@
    *
    * Nonblocking can be turned on and off at any time.
    *
-   * @param nonblock enable or not the nonblocking reads - 1 to enable
-   *                 nonblocking - 0 to disable nonblocking.
-   * @returns 0 on success.
+   * @param nonblock enable or not the nonblocking reads - true to enable
+   *                 nonblocking - false to disable nonblocking.
    * @throws Error on failure.
    *
    * @since This function is available since SDL 3.2.0.
@@ -360,8 +356,6 @@
    *
    * @param data the data to send, including the report number as the first
    *             byte.
-   * @param length the length in bytes of the data to send, including the report
-   *               number.
    * @returns the actual number of bytes written and -1 on failure; call
    *          GetError() for more information.
    *
@@ -381,8 +375,6 @@
    *             Set the first byte of `data` to the Report ID of the report to
    *             be read, or set it to zero if your device does not use numbered
    *             reports.
-   * @param length the number of bytes to read, including an extra byte for the
-   *               report ID. The buffer can be longer than the actual report.
    * @returns the number of bytes read plus one for the report ID (which is
    *          still in the first byte), or -1 on on failure; call GetError() for
    *          more information.
@@ -403,8 +395,6 @@
    *             Set the first byte of `data` to the Report ID of the report to
    *             be read, or set it to zero if your device does not use numbered
    *             reports.
-   * @param length the number of bytes to read, including an extra byte for the
-   *               report ID. The buffer can be longer than the actual report.
    * @returns the number of bytes read plus one for the report ID (which is
    *          still in the first byte), or -1 on on failure; call GetError() for
    *          more information.
@@ -418,7 +408,6 @@
    *
    * @param string a wide string buffer to put the data into.
    * @param maxlen the length of the buffer in multiples of wchar_t.
-   * @returns 0 on success.
    * @throws Error on failure.
    *
    * @since This function is available since SDL 3.2.0.
@@ -430,7 +419,6 @@
    *
    * @param string a wide string buffer to put the data into.
    * @param maxlen the length of the buffer in multiples of wchar_t.
-   * @returns 0 on success.
    * @throws Error on failure.
    *
    * @since This function is available since SDL 3.2.0.
@@ -442,7 +430,6 @@
    *
    * @param string a wide string buffer to put the data into.
    * @param maxlen the length of the buffer in multiples of wchar_t.
-   * @returns 0 on success.
    * @throws Error on failure.
    *
    * @since This function is available since SDL 3.2.0.
@@ -455,7 +442,6 @@
    * @param string_index the index of the string to get.
    * @param string a wide string buffer to put the data into.
    * @param maxlen the length of the buffer in multiples of wchar_t.
-   * @returns 0 on success.
    * @throws Error on failure.
    *
    * @since This function is available since SDL 3.2.0.
@@ -465,9 +451,10 @@
   /**
    * Get the device info from a HID device.
    *
-   * @returns a pointer to the hid_device_info for this hid_device or nullptr on
-   *          failure; call GetError() for more information. This struct is
-   *          valid until the device is closed with HidDevice.close().
+   * @returns a pointer to the hid_device_info for this hid_device on success.
+   *          This struct is valid until the device is closed with
+   *          HidDevice.close().
+   * @throws Error on failure.
    *
    * @since This function is available since SDL 3.2.0.
    */
@@ -480,7 +467,6 @@
    * to. The recommended size for a preallocated buffer is 4096 bytes.
    *
    * @param buf the buffer to copy descriptor into.
-   * @param buf_size the size of the buffer in bytes.
    * @returns the number of bytes actually copied or -1 on failure; call
    *          GetError() for more information.
    *
@@ -539,8 +525,6 @@
  *
  * Each call to this function should have a matching call to hid_exit()
  *
- * @returns 0 on success.
- *
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
@@ -555,8 +539,6 @@
  * This function frees all of the static data associated with HIDAPI. It should
  * be called at the end of execution to avoid memory leaks.
  *
- * @returns 0 on success.
- *
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
@@ -607,9 +589,9 @@
  * @param product_id the Product ID (PID) of the types of device to open, or 0
  *                   to match any product.
  * @returns a pointer to a linked list of type hid_device_info, containing
- *          information about the HID devices attached to the system, or nullptr
- *          in the case of failure. Free this linked list by calling
- *          hid_free_enumeration().
+ *          information about the HID devices attached to the system on success.
+ *          Free this linked list by calling hid_free_enumeration().
+ * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -646,7 +628,7 @@
  * @param product_id the Product ID (PID) of the device to open.
  * @param serial_number the Serial Number of the device to open (Optionally
  *                      nullptr).
- * @returns a pointer to a HidDevice object on success.
+ * @returns a HidDevice object on success.
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
@@ -665,7 +647,7 @@
  * platform-specific path name can be used (eg: /dev/hidraw0 on Linux).
  *
  * @param path the path name of the device to open.
- * @returns a pointer to a HidDevice object on success.
+ * @returns a HidDevice object on success.
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
@@ -696,28 +678,20 @@
   return CheckError(SDL_hid_get_properties(dev));
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 inline PropertiesRef HidDevice::hid_get_properties()
 {
   return SDL::hid_get_properties(m_resource);
 }
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
 namespace prop::Hidapi {
 
-#if SDL_VERSION_ATLEAST(3, 4, 0)
-
 constexpr auto LIBUSB_DEVICE_HANDLE_POINTER =
   SDL_PROP_HIDAPI_LIBUSB_DEVICE_HANDLE_POINTER;
 
-#endif // SDL_VERSION_ATLEAST(3, 4, 0)
-
 } // namespace prop::Hidapi
 
+#endif // SDL_VERSION_ATLEAST(3, 4, 0)
+
 /**
  * Write an Output report to a HID device.
  *
@@ -735,7 +709,6 @@
  *
  * @param dev a device handle returned from HidDevice.HidDevice().
  * @param data the data to send, including the report number as the first byte.
- * @param length the length in bytes of the data to send.
  * @returns the actual number of bytes written and -1 on on failure; call
  *          GetError() for more information.
  *
@@ -743,7 +716,7 @@
  */
 inline int hid_write(HidDeviceParam dev, SourceBytes data)
 {
-  return SDL_hid_write(dev, data.data(), data.size_bytes());
+  return SDL_hid_write(dev, data.data_as<Uint8>(), data.size_bytes());
 }
 
 inline int HidDevice::write(SourceBytes data)
@@ -760,9 +733,7 @@
  *
  * @param dev a device handle returned from HidDevice.HidDevice().
  * @param data a buffer to put the read data into.
- * @param length the number of bytes to read. For devices with multiple reports,
- *               make sure to read an extra byte for the report number.
- * @param milliseconds timeout in milliseconds or -1 for blocking wait.
+ * @param timeout timeout in milliseconds
  * @returns the actual number of bytes read and -1 on on failure; call
  *          GetError() for more information. If no packet was available to be
  *          read within the timeout period, this function returns 0.
@@ -773,7 +744,8 @@
                             TargetBytes data,
                             Milliseconds timeout)
 {
-  return SDL_hid_read_timeout(dev, data.data(), data.size_bytes(), timeout);
+  return SDL_hid_read_timeout(
+    dev, data.data_as<Uint8>(), data.size_bytes(), timeout.count());
 }
 
 inline int HidDevice::read_timeout(TargetBytes data, Milliseconds timeout)
@@ -790,8 +762,6 @@
  *
  * @param dev a device handle returned from HidDevice.HidDevice().
  * @param data a buffer to put the read data into.
- * @param length the number of bytes to read. For devices with multiple reports,
- *               make sure to read an extra byte for the report number.
  * @returns the actual number of bytes read and -1 on failure; call GetError()
  *          for more information. If no packet was available to be read and the
  *          handle is in non-blocking mode, this function returns 0.
@@ -800,7 +770,7 @@
  */
 inline int hid_read(HidDeviceParam dev, TargetBytes data)
 {
-  return SDL_hid_read(dev, data.data(), data.size_bytes());
+  return SDL_hid_read(dev, data.data_as<Uint8>(), data.size_bytes());
 }
 
 inline int HidDevice::read(TargetBytes data)
@@ -818,9 +788,8 @@
  * Nonblocking can be turned on and off at any time.
  *
  * @param dev a device handle returned from HidDevice.HidDevice().
- * @param nonblock enable or not the nonblocking reads - 1 to enable nonblocking
- *                 - 0 to disable nonblocking.
- * @returns 0 on success.
+ * @param nonblock enable or not the nonblocking reads - true to enable
+ *                 nonblocking - false to disable nonblocking.
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
@@ -850,8 +819,6 @@
  *
  * @param dev a device handle returned from HidDevice.HidDevice().
  * @param data the data to send, including the report number as the first byte.
- * @param length the length in bytes of the data to send, including the report
- *               number.
  * @returns the actual number of bytes written and -1 on failure; call
  *          GetError() for more information.
  *
@@ -859,7 +826,8 @@
  */
 inline int hid_send_feature_report(HidDeviceParam dev, SourceBytes data)
 {
-  return SDL_hid_send_feature_report(dev, data.data(), data.size_bytes());
+  return SDL_hid_send_feature_report(
+    dev, data.data_as<Uint8>(), data.size_bytes());
 }
 
 inline int HidDevice::send_feature_report(SourceBytes data)
@@ -880,8 +848,6 @@
  *             the first byte of `data` to the Report ID of the report to be
  *             read, or set it to zero if your device does not use numbered
  *             reports.
- * @param length the number of bytes to read, including an extra byte for the
- *               report ID. The buffer can be longer than the actual report.
  * @returns the number of bytes read plus one for the report ID (which is still
  *          in the first byte), or -1 on on failure; call GetError() for more
  *          information.
@@ -890,7 +856,8 @@
  */
 inline int hid_get_feature_report(HidDeviceParam dev, TargetBytes data)
 {
-  return SDL_hid_get_feature_report(dev, data.data(), data.size_bytes());
+  return SDL_hid_get_feature_report(
+    dev, data.data_as<Uint8>(), data.size_bytes());
 }
 
 inline int HidDevice::get_feature_report(TargetBytes data)
@@ -911,8 +878,6 @@
  *             the first byte of `data` to the Report ID of the report to be
  *             read, or set it to zero if your device does not use numbered
  *             reports.
- * @param length the number of bytes to read, including an extra byte for the
- *               report ID. The buffer can be longer than the actual report.
  * @returns the number of bytes read plus one for the report ID (which is still
  *          in the first byte), or -1 on on failure; call GetError() for more
  *          information.
@@ -921,7 +886,8 @@
  */
 inline int hid_get_input_report(HidDeviceParam dev, TargetBytes data)
 {
-  return SDL_hid_get_input_report(dev, data.data(), data.size_bytes());
+  return SDL_hid_get_input_report(
+    dev, data.data_as<Uint8>(), data.size_bytes());
 }
 
 inline int HidDevice::get_input_report(TargetBytes data)
@@ -933,14 +899,13 @@
  * Close a HID device.
  *
  * @param dev a device handle returned from HidDevice.HidDevice().
- * @returns 0 on success.
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
  */
-inline int hid_close(HidDeviceRaw dev)
+inline void hid_close(HidDeviceRaw dev)
 {
-  return CheckError(SDL_hid_close(dev));
+  CheckErrorIfNot(SDL_hid_close(dev), 0);
 }
 
 inline void HidDevice::close() { hid_close(release()); }
@@ -951,7 +916,6 @@
  * @param dev a device handle returned from HidDevice.HidDevice().
  * @param string a wide string buffer to put the data into.
  * @param maxlen the length of the buffer in multiples of wchar_t.
- * @returns 0 on success.
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
@@ -974,7 +938,6 @@
  * @param dev a device handle returned from HidDevice.HidDevice().
  * @param string a wide string buffer to put the data into.
  * @param maxlen the length of the buffer in multiples of wchar_t.
- * @returns 0 on success.
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
@@ -997,7 +960,6 @@
  * @param dev a device handle returned from HidDevice.HidDevice().
  * @param string a wide string buffer to put the data into.
  * @param maxlen the length of the buffer in multiples of wchar_t.
- * @returns 0 on success.
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
@@ -1021,7 +983,6 @@
  * @param string_index the index of the string to get.
  * @param string a wide string buffer to put the data into.
  * @param maxlen the length of the buffer in multiples of wchar_t.
- * @returns 0 on success.
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
@@ -1046,9 +1007,10 @@
  * Get the device info from a HID device.
  *
  * @param dev a device handle returned from HidDevice.HidDevice().
- * @returns a pointer to the hid_device_info for this hid_device or nullptr on
- *          failure; call GetError() for more information. This struct is valid
- *          until the device is closed with HidDevice.close().
+ * @returns a pointer to the hid_device_info for this hid_device on success.
+ *          This struct is valid until the device is closed with
+ *          HidDevice.close().
+ * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
  */
@@ -1070,7 +1032,6 @@
  *
  * @param dev a device handle returned from HidDevice.HidDevice().
  * @param buf the buffer to copy descriptor into.
- * @param buf_size the size of the buffer in bytes.
  * @returns the number of bytes actually copied or -1 on failure; call
  *          GetError() for more information.
  *
@@ -1078,7 +1039,8 @@
  */
 inline int hid_get_report_descriptor(HidDeviceParam dev, TargetBytes buf)
 {
-  return SDL_hid_get_report_descriptor(dev, buf.data(), buf.size_bytes());
+  return SDL_hid_get_report_descriptor(
+    dev, buf.data_as<Uint8>(), buf.size_bytes());
 }
 
 inline int HidDevice::get_report_descriptor(TargetBytes buf)
