--- build/generated/SDL3pp_init.h
+++ include/SDL3pp/SDL3pp_init.h
@@ -12,7 +12,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryInit Category Init
+ * @defgroup CategoryInit Initialization and Shutdown
  *
  * All SDL programs need to initialize the library before starting to work with
  * it.
@@ -41,6 +41,12 @@
  */
 
 /**
+ * @defgroup InitFlags Initialization flags
+ *
+ * @{
+ */
+
+/**
  * Initialization flags for Init and/or InitSubSystem
  *
  * These are the flags which may be passed to Init(). You should specify the
@@ -80,6 +86,14 @@
 constexpr InitFlags INIT_CAMERA =
   SDL_INIT_CAMERA; ///< `INIT_CAMERA` implies `INIT_EVENTS`
 
+/// @}
+
+/**
+ * @name AppResult
+ * App result for Main callback
+ * @{
+ */
+
 /**
  * Return values for optional main callbacks.
  *
@@ -109,6 +123,14 @@
 /// Value that requests termination with error from the main callbacks.
 constexpr AppResult APP_FAILURE = SDL_APP_FAILURE;
 
+/// @}
+
+/**
+ * @name Callbacks for EnterAppMainCallbacks()
+ *
+ * @{
+ */
+
 /**
  * Function pointer typedef for SDL_AppInit.
  *
@@ -174,6 +196,8 @@
  */
 using AppQuit_func = void(SDLCALL*)(void* appstate, AppResult result);
 
+/// @}
+
 /**
  * Initialize the SDL library.
  *
@@ -324,6 +348,11 @@
 inline bool IsMainThread() { return SDL_IsMainThread(); }
 
 /**
+ * @name Callbacks for RunOnMainThread()
+ * @{
+ */
+
+/**
  * Callback run on the main thread.
  *
  * @param userdata an app-controlled pointer that is passed to the callback.
@@ -342,9 +371,14 @@
  * @sa RunOnMainThread
  *
  * @sa MainThreadCallback
+ *
+ * @cat result-callback
+ *
  */
 using MainThreadCB = std::function<void()>;
 
+/// @}
+
 /**
  * Call a function on the main thread during event processing.
  *
@@ -387,7 +421,6 @@
  * called with `wait_complete` true.
  *
  * @param callback the callback to call on the main thread.
- * @param userdata a pointer that is passed to `callback`.
  * @param wait_complete true to wait for the callback to complete, false to
  *                      return immediately.
  * @throws Error on failure.
@@ -397,10 +430,15 @@
  * @since This function is available since SDL 3.2.0.
  *
  * @sa IsMainThread
+ * @sa result-callback
+ *
+ * @cat result-callback
  */
 inline void RunOnMainThread(MainThreadCB callback, bool wait_complete)
 {
-  static_assert(false, "Not implemented");
+  using Wrapper = CallbackWrapper<MainThreadCB>;
+  void* wrapped = Wrapper::Wrap(std::move(callback));
+  RunOnMainThread(&Wrapper::CallOnce, wrapped, wait_complete);
 }
 
 /**
@@ -506,6 +544,12 @@
   CheckError(SDL_SetAppMetadataProperty(name, value));
 }
 
+/**
+ * Metadata property names for application.
+ *
+ * @sa SetAppMetadataProperty
+ * @sa GetAppMetadataProperty
+ */
 namespace prop::appMetaData {
 
 constexpr auto NAME_STRING =
@@ -557,6 +601,13 @@
   return SDL_GetAppMetadataProperty(name);
 }
 
+#ifndef SDL3PP_APPCLASS_LOG_PRIORITY
+/**
+ * The default log priority for app class.
+ */
+#define SDL3PP_APPCLASS_LOG_PRIORITY SDL::LOG_PRIORITY_CRITICAL
+#endif // SDL3PP_APPCLASS_LOG_PRIORITY
+
 /// @}
 
 } // namespace SDL
