--- build/generated/SDL3pp_properties.h
+++ include/SDL3pp/SDL3pp_properties.h
@@ -13,14 +13,15 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryProperties Category Properties
+ * @defgroup CategoryProperties Object Properties
  *
  * A property is a variable that can be created and retrieved by name at
  * runtime.
  *
  * All properties are part of a property group (Properties). A property group
  * can be created with the CreateProperties function and destroyed with the
- * DestroyProperties function.
+ * DestroyProperties function, but the Properties destructor probably will do
+ * what you want to, automatically.
  *
  * Properties can be added to and retrieved from a property group through the
  * following functions:
@@ -89,6 +90,13 @@
 // Forward decl
 struct PropertyIterator;
 
+/// A value assignable to a property.
+template<class T>
+concept PropertyValue =
+  std::convertible_to<T, const PropertyProxy&> ||
+  std::convertible_to<T, void*> || std::convertible_to<T, StringParam> ||
+  std::convertible_to<T, Sint64> || std::convertible_to<T, float>;
+
 /**
  * A callback used to enumerate all the properties in a group of properties.
  *
@@ -205,13 +213,37 @@
    */
   void Destroy();
 
+  /**
+   * Access a property from a group of properties.
+   *
+   * @param name the name of the property to query.
+   * @returns A proxy that is convertible to any valid property type.
+   */
   PropertyProxy operator[](StringParam name) const;
 
+  /**
+   * Gets an iterator to the first property in a group of properties.
+   *
+   * @return PropertyIterator No particular order is guaranteed.
+   */
   PropertyIterator begin() const;
 
+  /**
+   * Gets sentinel iterator for a group of properties.
+   *
+   * @return a sentinel, which is always a nullptr.
+   */
   std::nullptr_t end() const;
 
   /**
+   * Access a property from a group of properties.
+   *
+   * @param name the name of the property to query.
+   * @returns A proxy that is convertible to any valid property type.
+   */
+  PropertyMutableProxy operator[](StringParam name);
+
+  /**
    * Copy a group of properties.
    *
    * Copy all the properties from one group of properties to another, with the
@@ -263,6 +295,14 @@
    */
   void Unlock(PropertiesLock&& lock);
 
+  /**
+   * Set a property in a group of properties.
+   *
+   * @param name the name of the property to modify.
+   * @param value the new value of the property, or nullptr to delete the
+   *              property.
+   * @throws Error on failure.
+   */
   template<PropertyValue V>
   void Set(StringParam name, V&& value);
 
@@ -316,7 +356,6 @@
    *              property.
    * @param cleanup the function to call when this property is deleted, or
    *                nullptr if no cleanup is necessary.
-   * @param userdata a pointer that is passed to the cleanup function.
    * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread.
@@ -446,6 +485,12 @@
    */
   PropertyType GetPropertyType(StringParam name) const;
 
+  /**
+   * Get a property from a group of properties.
+   *
+   * @param name the name of the property to query.
+   * @returns A proxy that is convertible to any valid property type.
+   */
   PropertyProxy Get(StringParam name) const;
 
   /**
@@ -601,7 +646,6 @@
    * properties. The properties are locked during enumeration.
    *
    * @param callback the function to call for each property.
-   * @param userdata a pointer that is passed to `callback`.
    * @throws Error on failure.
    *
    * @threadsafety It is safe to call this function from any thread.
@@ -610,6 +654,13 @@
    */
   void Enumerate(EnumeratePropertiesCB callback) const;
 
+  /**
+   * Returns the number of properties this has
+   *
+   * This uses EnumerateProperties() internally, so might not be so fast
+   *
+   * @return Uint64
+   */
   Uint64 GetCount() const;
 };
 
@@ -619,6 +670,9 @@
  * @since This datatype is available since SDL 3.2.0.
  *
  * @cat resource
+ *
+ * @sa Properties.Create
+ * @sa prop
  */
 struct Properties : PropertiesBase
 {
@@ -659,6 +713,43 @@
    */
   Properties();
 
+  /**
+   * Create a group of properties.
+   *
+   * All properties are automatically destroyed when Quit() is called.
+   *
+   * @param entries the properties to set on this group of properties, as a list
+   *                of name-value pairs.
+   * @post an ID for a new group of properties on success.
+   *
+   * @throws Error on failure.
+   *
+   * @threadsafety It is safe to call this function from any thread.
+   *
+   * @since This function is available since SDL 3.2.0.
+   *
+   * @sa DestroyProperties
+   */
+  Properties(
+    std::initializer_list<
+      std::pair<StringParam,
+                std::variant<StringParam, void*, Sint64, float, bool>>> entries)
+    : Properties()
+  {
+    for (auto& [name, value] : entries) {
+      std::visit(
+        [this, &name](auto& arg) {
+          if constexpr (std::is_convertible_v<decltype(arg),
+                                              const StringParam&>) {
+            Set(name.c_str(), arg.c_str());
+          } else {
+            Set(name.c_str(), arg);
+          }
+        },
+        value);
+    }
+  }
+
   /// Destructor
   ~Properties() { SDL_DestroyProperties(get()); }
 
@@ -709,7 +800,7 @@
    * or want to guarantee that properties being queried aren't freed in another
    * thread.
    *
-   * @param props the properties to lock.
+   * @param resource the properties to lock.
    * @post true on success or false on failure; call GetError() for more
    *       information.
    *
@@ -733,8 +824,6 @@
   /**
    * Unlock a group of properties.
    *
-   * @param props the properties to unlock.
-   *
    * @threadsafety It is safe to call this function from any thread.
    *
    * @since This function is available since SDL 3.2.0.
@@ -758,8 +847,6 @@
   /**
    * Unlock a group of properties.
    *
-   * @param props the properties to unlock.
-   *
    * @threadsafety It is safe to call this function from any thread.
    *
    * @since This function is available since SDL 3.2.0.
@@ -775,18 +862,82 @@
   void release() { m_lock.release(); }
 };
 
+/// Represent a property name from a given Properties
 class PropertyProxy
 {
+public:
+  /// Constructor
+  PropertyProxy(PropertiesRef props, StringParam name)
+    : m_props(props)
+    , m_name(std::move(name))
+  {
+  }
+
+  /// True if the there is a property with this name.
+  bool IsValid() const { return m_props.HasProperty(m_name.c_str()); }
+
+  /// Get Property type.
+  PropertyType GetType() const
+  {
+    return m_props.GetPropertyType(m_name.c_str());
+  }
+
+  /// Get Property name.
+  const char* GetName() const { return m_name; }
+
+  /// Get the property group this proxy belongs to.
+  PropertiesRef GetProperties() const { return m_props; }
+
+  /// Get property as a void pointer
+  operator void*() const { return m_props.GetPointerProperty(m_name.c_str()); }
+
+  /// Get property as string.
+  operator const char*() const
+  {
+    return m_props.GetStringProperty(m_name.c_str());
+  }
+
+  /// Get property as integral type.
+  template<std::integral T>
+  operator T() const
+  {
+    return T(m_props.GetNumberProperty(m_name.c_str()));
+  }
+
+  /// Get property as float.
+  operator float() const { return m_props.GetFloatProperty(m_name.c_str()); }
+
+  /// Get property as boolean.
+  operator bool() const { return m_props.GetBooleanProperty(m_name.c_str()); }
+
+  /// Call visitor with the correct type based on its GetType().
+  template<class T>
+  auto visit(T visitor) const
+  {
+    switch (GetType()) {
+    case SDL_PROPERTY_TYPE_POINTER: return visitor((void*)(*this));
+    case SDL_PROPERTY_TYPE_STRING: return visitor((const char*)(*this));
+    case SDL_PROPERTY_TYPE_NUMBER: return visitor(Sint64(*this));
+    case SDL_PROPERTY_TYPE_FLOAT: return visitor(float(*this));
+    case SDL_PROPERTY_TYPE_BOOLEAN: return visitor(bool(*this));
+    default: return visitor(std::nullopt);
+    }
+  }
+
 private:
   PropertiesRef m_props;
 
   StringParam m_name;
+
+  friend class PropertyIterator;
 };
 
+/// Represent a mutable property name from a given Properties
 struct PropertyMutableProxy : PropertyProxy
 {
   using PropertyProxy::PropertyProxy;
 
+  /// Change the value of this property.
   template<PropertyValue V>
   PropertyMutableProxy& operator=(V&& value)
   {
@@ -795,13 +946,83 @@
   }
 };
 
+/// Iterator for properties in a Properties set.
 class PropertyIterator
 {
 public:
+  /// Default constructor
   constexpr PropertyIterator() = default;
 
+  /// Constructor
+  PropertyIterator(PropertiesRef props, size_t index = 0)
+    : m_index(index)
+  {
+    props.Enumerate(
+      [this](auto, const char* key) { m_keys.emplace_back(key); });
+    if (m_index < m_keys.size()) m_proxy = {props, m_keys[m_index]};
+  }
+
+  ~PropertyIterator() = default;
+
+  /// Comparison with nullptr, to check if the iterator is at the end.
+  constexpr bool operator==(std::nullptr_t) const
+  {
+    return m_index >= m_keys.size();
+  }
+
+  /// Comparison operator.
+  constexpr bool operator==(const PropertyIterator& other) const
+  {
+    return m_index == other.m_index;
+  }
+
+  /// Dereference operator to get the current property proxy.
+  const PropertyProxy& operator*() const { return m_proxy; }
+
+  /// Arrow operator to get a pointer to the current property proxy.
+  const PropertyProxy* operator->() const { return &m_proxy; }
+
+  /// Pre-increment operator to move to the next property.
+  PropertyIterator& operator++()
+  {
+    ++m_index;
+    update();
+    return *this;
+  }
+
+  /// Post-increment operator to move to the next property.
+  PropertyIterator operator++(int)
+  {
+    auto copy = *this;
+    ++(*this);
+    return copy;
+  }
+
+  /// Pre-decrement operator to move to the previous property.
+  PropertyIterator& operator--()
+  {
+    --m_index;
+    update();
+    return *this;
+  }
+
+  /// Post-decrement operator to move to the previous property.
+  PropertyIterator operator--(int)
+  {
+    auto copy = *this;
+    --(*this);
+    return copy;
+  }
+
 private:
   std::vector<std::string> m_keys;
+  PropertyProxy m_proxy{{}, ""};
+  size_t m_index = 0;
+
+  void update()
+  {
+    if (m_index < m_keys.size()) m_proxy.m_name = m_keys[m_index];
+  }
 };
 
 #if SDL_VERSION_ATLEAST(3, 4, 0)
@@ -828,7 +1049,7 @@
  *
  * @since This constant is available since SDL 3.4.0.
  */
-inline auto PROP_NAME_STRING = SDL_PROP_NAME_STRING;
+constexpr const char* PROP_NAME_STRING = SDL_PROP_NAME_STRING;
 
 #endif // SDL_VERSION_ATLEAST(3, 4, 0)
 
@@ -845,7 +1066,7 @@
  */
 inline PropertiesRef GetGlobalProperties()
 {
-  return CheckError(SDL_GetGlobalProperties());
+  return {CheckError(SDL_GetGlobalProperties())};
 }
 
 /**
@@ -970,16 +1191,63 @@
 
 inline std::nullptr_t PropertiesBase::end() const { return nullptr; }
 
+inline PropertyMutableProxy PropertiesBase::operator[](StringParam name)
+{
+  return {*this, std::move(name)};
+}
+
+/**
+ * Set a property in a group of properties.
+ *
+ * @param props the properties to modify.
+ * @param name the name of the property to modify.
+ * @param value the new value of the property, or nullptr to delete the
+ *              property.
+ * @throws Error on failure.
+ */
 template<PropertyValue V>
 inline void SetProperty(PropertiesRef props, StringParam name, V&& value)
 {
-  static_assert(false, "Not implemented");
+  if constexpr (std::is_same_v<V, bool>) {
+    props.SetBooleanProperty(std::move(name), std::forward<V>(value));
+  } else if constexpr (std::is_integral_v<V>) {
+    props.SetNumberProperty(std::move(name), std::forward<V>(value));
+  } else {
+    struct Visitor
+    {
+      PropertiesRef props;
+      StringParam name;
+
+      void operator()(std::monostate) { props.ClearProperty(std::move(name)); }
+      void operator()(std::nullopt_t) { props.ClearProperty(std::move(name)); }
+      void operator()(void* v) { props.SetPointerProperty(std::move(name), v); }
+
+      void operator()(StringParam&& v)
+      {
+        props.SetStringProperty(std::move(name), std::move(v));
+      }
+      void operator()(const char* v)
+      {
+        props.SetStringProperty(std::move(name), std::move(v));
+      }
+
+      void operator()(Sint64 v) { props.SetNumberProperty(std::move(name), v); }
+      void operator()(float v) { props.SetFloatProperty(std::move(name), v); }
+      void operator()(bool v) { props.SetBooleanProperty(std::move(name), v); }
+
+      void operator()(const PropertyProxy& prop)
+      {
+        prop.visit([this](auto v) { (*this)(v); });
+      }
+    };
+    Visitor{props, std::move(name)}(std::forward<V>(value));
+  }
 }
 
 template<PropertyValue V>
 inline void PropertiesBase::Set(StringParam name, V&& value)
 {
-  SDL::SetProperty(get(), std::move(name), value);
+  SDL::SetProperty(get(), std::move(name), std::forward<V>(value));
 }
 
 /**
@@ -1039,7 +1307,6 @@
  *              property.
  * @param cleanup the function to call when this property is deleted, or nullptr
  *                if no cleanup is necessary.
- * @param userdata a pointer that is passed to the cleanup function.
  * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
@@ -1055,7 +1322,12 @@
                                           void* value,
                                           CleanupPropertyCB cleanup)
 {
-  static_assert(false, "Not implemented");
+  using Wrapper = CallbackWrapper<CleanupPropertyCB>;
+  SDL_SetPointerPropertyWithCleanup(props,
+                                    std::move(name),
+                                    value,
+                                    &Wrapper::CallOnce,
+                                    Wrapper::Wrap(std::move(cleanup)));
 }
 
 inline void PropertiesBase::SetPointerPropertyWithCleanup(
@@ -1073,7 +1345,8 @@
   void* value,
   CleanupPropertyCB cleanup)
 {
-  SDL::SetPointerPropertyWithCleanup(get(), std::move(name), value, cleanup);
+  SDL::SetPointerPropertyWithCleanup(
+    get(), std::move(name), value, std::move(cleanup));
 }
 
 /**
@@ -1262,9 +1535,16 @@
   return SDL::GetPropertyType(get(), std::move(name));
 }
 
+/**
+ * Get a property from a group of properties.
+ *
+ * @param props the properties to query.
+ * @param name the name of the property to query.
+ * @returns A proxy that is convertible to any valid property type.
+ */
 inline PropertyProxy GetProperty(PropertiesRef props, StringParam name)
 {
-  static_assert(false, "Not implemented");
+  return {props, std::move(name)};
 }
 
 inline PropertyProxy PropertiesBase::Get(StringParam name) const
@@ -1501,7 +1781,6 @@
  *
  * @param props the properties to query.
  * @param callback the function to call for each property.
- * @param userdata a pointer that is passed to `callback`.
  * @throws Error on failure.
  *
  * @threadsafety It is safe to call this function from any thread.
@@ -1511,7 +1790,13 @@
 inline void EnumerateProperties(PropertiesRef props,
                                 EnumeratePropertiesCB callback)
 {
-  static_assert(false, "Not implemented");
+  return EnumerateProperties(
+    props,
+    [](void* userdata, PropertiesID props, const char* name) {
+      auto& f = *static_cast<const EnumeratePropertiesCB*>(userdata);
+      f(props, name);
+    },
+    &callback);
 }
 
 inline void PropertiesBase::Enumerate(EnumeratePropertiesCallback callback,
@@ -1522,12 +1807,22 @@
 
 inline void PropertiesBase::Enumerate(EnumeratePropertiesCB callback) const
 {
-  SDL::EnumerateProperties(get(), callback);
+  SDL::EnumerateProperties(get(), std::move(callback));
 }
 
+/**
+ * Returns the number of properties this has
+ *
+ * This uses EnumerateProperties() internally, so might not be so fast
+ *
+ * @param props
+ * @return Uint64
+ */
 inline Uint64 CountProperties(PropertiesRef props)
 {
-  static_assert(false, "Not implemented");
+  Uint64 count = 0;
+  EnumerateProperties(props, [&count](auto, const char*) { count++; });
+  return count;
 }
 
 inline Uint64 PropertiesBase::GetCount() const
@@ -1560,6 +1855,17 @@
 
 /// @}
 
+/**
+ * @namespace SDL::prop
+ * @brief Constants for Properties keys.
+ *
+ * All subnamespaces in `prop` are intended to be used as namespaces for
+ * properties related to a specific type of object. For example, `prop::Surface`
+ * contains properties related to surfaces.
+ *
+ * @sa Properties
+ */
+
 } // namespace SDL
 
 #endif /* SDL3PP_PROPERTIES_H_ */
