--- src/generated/SDL3pp_gpu.h
+++ include/SDL3pp/SDL3pp_gpu.h
@@ -12,7 +12,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryGPU Category GPU
+ * @defgroup CategoryGPU 3D Rendering and GPU Compute
  *
  * The GPU API offers a cross-platform way for apps to talk to modern graphics
  * hardware. It offers both 3D graphics and compute support, in the style of
@@ -138,7 +138,7 @@
  * limited hardware support, this API is probably not for you.
  *
  * Examples demonstrating proper usage of this API can be found
- * [here](https://github.com/TheSpydog/SDL_gpu_examples) .
+ * [here](https://github.com/TheSpydog/SDL_gpu_examples).
  *
  * ## Performance considerations
  *
@@ -1218,7 +1218,6 @@
    * @param first_slot the vertex buffer slot to begin binding from.
    * @param bindings an array of GPUBufferBinding structs containing vertex
    *                 buffers and offset values.
-   * @param num_bindings the number of bindings in the bindings array.
    *
    * @since This function is available since SDL 3.2.0.
    */
@@ -1249,8 +1248,6 @@
    * @param first_slot the vertex sampler slot to begin binding from.
    * @param texture_sampler_bindings an array of texture-sampler binding
    *                                 structs.
-   * @param num_bindings the number of texture-sampler pairs to bind from the
-   *                     array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1271,7 +1268,6 @@
    *
    * @param first_slot the vertex storage texture slot to begin binding from.
    * @param storage_textures an array of storage textures.
-   * @param num_bindings the number of storage texture to bind from the array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1291,7 +1287,6 @@
    *
    * @param first_slot the vertex storage buffer slot to begin binding from.
    * @param storage_buffers an array of buffers.
-   * @param num_bindings the number of buffers to bind from the array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1311,8 +1306,6 @@
    * @param first_slot the fragment sampler slot to begin binding from.
    * @param texture_sampler_bindings an array of texture-sampler binding
    *                                 structs.
-   * @param num_bindings the number of texture-sampler pairs to bind from the
-   *                     array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1333,7 +1326,6 @@
    *
    * @param first_slot the fragment storage texture slot to begin binding from.
    * @param storage_textures an array of storage textures.
-   * @param num_bindings the number of storage textures to bind from the array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1354,7 +1346,6 @@
    *
    * @param first_slot the fragment storage buffer slot to begin binding from.
    * @param storage_buffers an array of storage buffers.
-   * @param num_bindings the number of storage buffers to bind from the array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1520,8 +1511,6 @@
    * @param first_slot the compute sampler slot to begin binding from.
    * @param texture_sampler_bindings an array of texture-sampler binding
    *                                 structs.
-   * @param num_bindings the number of texture-sampler bindings to bind from the
-   *                     array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1542,7 +1531,6 @@
    *
    * @param first_slot the compute storage texture slot to begin binding from.
    * @param storage_textures an array of storage textures.
-   * @param num_bindings the number of storage textures to bind from the array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -1562,7 +1550,6 @@
    *
    * @param first_slot the compute storage buffer slot to begin binding from.
    * @param storage_buffers an array of storage buffer binding structs.
-   * @param num_bindings the number of storage buffers to bind from the array.
    *
    * @since This function is available since SDL 3.2.0.
    *
@@ -2065,7 +2052,6 @@
    *
    * @param slot_index the vertex uniform slot to push data to.
    * @param data client data to write.
-   * @param length the length of the data to write.
    *
    * @since This function is available since SDL 3.2.0.
    */
@@ -2082,7 +2068,6 @@
    *
    * @param slot_index the fragment uniform slot to push data to.
    * @param data client data to write.
-   * @param length the length of the data to write.
    *
    * @since This function is available since SDL 3.2.0.
    */
@@ -2099,7 +2084,6 @@
    *
    * @param slot_index the uniform slot to push data to.
    * @param data client data to write.
-   * @param length the length of the data to write.
    *
    * @since This function is available since SDL 3.2.0.
    */
@@ -2118,8 +2102,6 @@
    *
    * @param color_target_infos an array of texture subresources with
    *                           corresponding clear values and load/store ops.
-   * @param num_color_targets the number of color targets in the
-   *                          color_target_infos array.
    * @param depth_stencil_target_info a texture subresource with corresponding
    *                                  clear value and load/store ops, may be
    *                                  nullptr.
@@ -2157,12 +2139,8 @@
    *
    * @param storage_texture_bindings an array of writeable storage texture
    *                                 binding structs.
-   * @param num_storage_texture_bindings the number of storage textures to bind
-   *                                     from the array.
    * @param storage_buffer_bindings an array of writeable storage buffer binding
    *                                structs.
-   * @param num_storage_buffer_bindings the number of storage buffers to bind
-   *                                    from the array.
    * @returns a compute pass handle.
    *
    * @since This function is available since SDL 3.2.0.
@@ -2230,12 +2208,11 @@
    * than the one that created the window.
    *
    * @param window a window that has been claimed.
-   * @param swapchain_texture a pointer filled in with a swapchain texture
-   *                          handle.
    * @param swapchain_texture_width a pointer filled in with the swapchain
    *                                texture width, may be nullptr.
    * @param swapchain_texture_height a pointer filled in with the swapchain
    *                                 texture height, may be nullptr.
+   * @returns a swapchain texture handle.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called from the thread that
@@ -2280,12 +2257,11 @@
    * another reading operation.
    *
    * @param window a window that has been claimed.
-   * @param swapchain_texture a pointer filled in with a swapchain texture
-   *                          handle.
    * @param swapchain_texture_width a pointer filled in with the swapchain
    *                                texture width, may be nullptr.
    * @param swapchain_texture_height a pointer filled in with the swapchain
    *                                 texture height, may be nullptr.
+   * @returns a swapchain texture handle.
    * @throws Error on failure.
    *
    * @threadsafety This function should only be called from the thread that
@@ -3784,7 +3760,6 @@
    * @param wait_all if 0, wait for any fence to be signaled, if 1, wait for all
    *                 fences to be signaled.
    * @param fences an array of fences to wait on.
-   * @param num_fences the number of fences in the fences array.
    * @throws Error on failure.
    *
    * @since This function is available since SDL 3.2.0.
@@ -3845,6 +3820,7 @@
    */
   bool TextureSupportsSampleCount(GPUTextureFormat format,
                                   GPUSampleCount sample_count);
+#ifdef SDL_PLATFORM_GDK
 
   /**
    * Call this to suspend GPU operation on Xbox when you receive the
@@ -3871,6 +3847,8 @@
    * @sa AddEventWatch
    */
   void GDKResumeGPU();
+
+#endif /* SDL_PLATFORM_GDK */
 };
 
 /// Semi-safe reference for GPUDevice.
@@ -5666,7 +5644,6 @@
  * @param command_buffer a command buffer.
  * @param slot_index the vertex uniform slot to push data to.
  * @param data client data to write.
- * @param length the length of the data to write.
  *
  * @since This function is available since SDL 3.2.0.
  */
@@ -5697,7 +5674,6 @@
  * @param command_buffer a command buffer.
  * @param slot_index the fragment uniform slot to push data to.
  * @param data client data to write.
- * @param length the length of the data to write.
  *
  * @since This function is available since SDL 3.2.0.
  */
@@ -5728,7 +5704,6 @@
  * @param command_buffer a command buffer.
  * @param slot_index the uniform slot to push data to.
  * @param data client data to write.
- * @param length the length of the data to write.
  *
  * @since This function is available since SDL 3.2.0.
  */
@@ -5761,8 +5736,6 @@
  * @param command_buffer a command buffer.
  * @param color_target_infos an array of texture subresources with corresponding
  *                           clear values and load/store ops.
- * @param num_color_targets the number of color targets in the
- *                          color_target_infos array.
  * @param depth_stencil_target_info a texture subresource with corresponding
  *                                  clear value and load/store ops, may be
  *                                  nullptr.
@@ -5777,8 +5750,10 @@
   std::span<const GPUColorTargetInfo> color_target_infos,
   OptionalRef<const GPUDepthStencilTargetInfo> depth_stencil_target_info)
 {
-  return SDL_BeginGPURenderPass(
-    command_buffer, color_target_infos, depth_stencil_target_info);
+  return SDL_BeginGPURenderPass(command_buffer,
+                                color_target_infos.data(),
+                                color_target_infos.size(),
+                                depth_stencil_target_info);
 }
 
 inline GPURenderPass GPUCommandBuffer::BeginRenderPass(
@@ -5839,7 +5814,7 @@
  */
 inline void SetGPUScissor(GPURenderPass render_pass, const RectRaw& scissor)
 {
-  SDL_SetGPUScissor(render_pass, scissor);
+  SDL_SetGPUScissor(render_pass, &scissor);
 }
 
 inline void GPURenderPass::SetScissor(const RectRaw& scissor)
@@ -5894,7 +5869,6 @@
  * @param first_slot the vertex buffer slot to begin binding from.
  * @param bindings an array of GPUBufferBinding structs containing vertex
  *                 buffers and offset values.
- * @param num_bindings the number of bindings in the bindings array.
  *
  * @since This function is available since SDL 3.2.0.
  */
@@ -5902,7 +5876,8 @@
                                  Uint32 first_slot,
                                  std::span<const GPUBufferBinding> bindings)
 {
-  SDL_BindGPUVertexBuffers(render_pass, first_slot, bindings);
+  SDL_BindGPUVertexBuffers(
+    render_pass, first_slot, bindings.data(), bindings.size());
 }
 
 inline void GPURenderPass::BindVertexBuffers(
@@ -5947,8 +5922,6 @@
  * @param render_pass a render pass handle.
  * @param first_slot the vertex sampler slot to begin binding from.
  * @param texture_sampler_bindings an array of texture-sampler binding structs.
- * @param num_bindings the number of texture-sampler pairs to bind from the
- *                     array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -5959,7 +5932,10 @@
   Uint32 first_slot,
   std::span<const GPUTextureSamplerBinding> texture_sampler_bindings)
 {
-  SDL_BindGPUVertexSamplers(render_pass, first_slot, texture_sampler_bindings);
+  SDL_BindGPUVertexSamplers(render_pass,
+                            first_slot,
+                            texture_sampler_bindings.data(),
+                            texture_sampler_bindings.size());
 }
 
 inline void GPURenderPass::BindVertexSamplers(
@@ -5982,7 +5958,6 @@
  * @param render_pass a render pass handle.
  * @param first_slot the vertex storage texture slot to begin binding from.
  * @param storage_textures an array of storage textures.
- * @param num_bindings the number of storage texture to bind from the array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -5993,7 +5968,8 @@
   Uint32 first_slot,
   SpanRef<const GPUTextureRaw> storage_textures)
 {
-  SDL_BindGPUVertexStorageTextures(render_pass, first_slot, storage_textures);
+  SDL_BindGPUVertexStorageTextures(
+    render_pass, first_slot, storage_textures.data(), storage_textures.size());
 }
 
 inline void GPURenderPass::BindVertexStorageTextures(
@@ -6016,7 +5992,6 @@
  * @param render_pass a render pass handle.
  * @param first_slot the vertex storage buffer slot to begin binding from.
  * @param storage_buffers an array of buffers.
- * @param num_bindings the number of buffers to bind from the array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -6027,7 +6002,8 @@
   Uint32 first_slot,
   SpanRef<const GPUBufferRaw> storage_buffers)
 {
-  SDL_BindGPUVertexStorageBuffers(render_pass, first_slot, storage_buffers);
+  SDL_BindGPUVertexStorageBuffers(
+    render_pass, first_slot, storage_buffers.data(), storage_buffers.size());
 }
 
 inline void GPURenderPass::BindVertexStorageBuffers(
@@ -6049,8 +6025,6 @@
  * @param render_pass a render pass handle.
  * @param first_slot the fragment sampler slot to begin binding from.
  * @param texture_sampler_bindings an array of texture-sampler binding structs.
- * @param num_bindings the number of texture-sampler pairs to bind from the
- *                     array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -6061,8 +6035,10 @@
   Uint32 first_slot,
   std::span<const GPUTextureSamplerBinding> texture_sampler_bindings)
 {
-  SDL_BindGPUFragmentSamplers(
-    render_pass, first_slot, texture_sampler_bindings);
+  SDL_BindGPUFragmentSamplers(render_pass,
+                              first_slot,
+                              texture_sampler_bindings.data(),
+                              texture_sampler_bindings.size());
 }
 
 inline void GPURenderPass::BindFragmentSamplers(
@@ -6085,7 +6061,6 @@
  * @param render_pass a render pass handle.
  * @param first_slot the fragment storage texture slot to begin binding from.
  * @param storage_textures an array of storage textures.
- * @param num_bindings the number of storage textures to bind from the array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -6096,7 +6071,8 @@
   Uint32 first_slot,
   SpanRef<const GPUTextureRaw> storage_textures)
 {
-  SDL_BindGPUFragmentStorageTextures(render_pass, first_slot, storage_textures);
+  SDL_BindGPUFragmentStorageTextures(
+    render_pass, first_slot, storage_textures.data(), storage_textures.size());
 }
 
 inline void GPURenderPass::BindFragmentStorageTextures(
@@ -6119,7 +6095,6 @@
  * @param render_pass a render pass handle.
  * @param first_slot the fragment storage buffer slot to begin binding from.
  * @param storage_buffers an array of storage buffers.
- * @param num_bindings the number of storage buffers to bind from the array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -6130,7 +6105,8 @@
   Uint32 first_slot,
   SpanRef<const GPUBufferRaw> storage_buffers)
 {
-  SDL_BindGPUFragmentStorageBuffers(render_pass, first_slot, storage_buffers);
+  SDL_BindGPUFragmentStorageBuffers(
+    render_pass, first_slot, storage_buffers.data(), storage_buffers.size());
 }
 
 inline void GPURenderPass::BindFragmentStorageBuffers(
@@ -6337,12 +6313,8 @@
  * @param command_buffer a command buffer.
  * @param storage_texture_bindings an array of writeable storage texture binding
  *                                 structs.
- * @param num_storage_texture_bindings the number of storage textures to bind
- *                                     from the array.
  * @param storage_buffer_bindings an array of writeable storage buffer binding
  *                                structs.
- * @param num_storage_buffer_bindings the number of storage buffers to bind from
- *                                    the array.
  * @returns a compute pass handle.
  *
  * @since This function is available since SDL 3.2.0.
@@ -6354,8 +6326,11 @@
   std::span<const GPUStorageTextureReadWriteBinding> storage_texture_bindings,
   std::span<const GPUStorageBufferReadWriteBinding> storage_buffer_bindings)
 {
-  return SDL_BeginGPUComputePass(
-    command_buffer, storage_texture_bindings, storage_buffer_bindings);
+  return SDL_BeginGPUComputePass(command_buffer,
+                                 storage_texture_bindings.data(),
+                                 storage_texture_bindings.size(),
+                                 storage_buffer_bindings.data(),
+                                 storage_buffer_bindings.size());
 }
 
 inline GPUComputePass GPUCommandBuffer::BeginComputePass(
@@ -6396,8 +6371,6 @@
  * @param compute_pass a compute pass handle.
  * @param first_slot the compute sampler slot to begin binding from.
  * @param texture_sampler_bindings an array of texture-sampler binding structs.
- * @param num_bindings the number of texture-sampler bindings to bind from the
- *                     array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -6408,8 +6381,10 @@
   Uint32 first_slot,
   std::span<const GPUTextureSamplerBinding> texture_sampler_bindings)
 {
-  SDL_BindGPUComputeSamplers(
-    compute_pass, first_slot, texture_sampler_bindings);
+  SDL_BindGPUComputeSamplers(compute_pass,
+                             first_slot,
+                             texture_sampler_bindings.data(),
+                             texture_sampler_bindings.size());
 }
 
 inline void GPUComputePass::BindSamplers(
@@ -6432,7 +6407,6 @@
  * @param compute_pass a compute pass handle.
  * @param first_slot the compute storage texture slot to begin binding from.
  * @param storage_textures an array of storage textures.
- * @param num_bindings the number of storage textures to bind from the array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -6443,7 +6417,8 @@
   Uint32 first_slot,
   SpanRef<const GPUTextureRaw> storage_textures)
 {
-  SDL_BindGPUComputeStorageTextures(compute_pass, first_slot, storage_textures);
+  SDL_BindGPUComputeStorageTextures(
+    compute_pass, first_slot, storage_textures.data(), storage_textures.size());
 }
 
 inline void GPUComputePass::BindStorageTextures(
@@ -6466,7 +6441,6 @@
  * @param compute_pass a compute pass handle.
  * @param first_slot the compute storage buffer slot to begin binding from.
  * @param storage_buffers an array of storage buffer binding structs.
- * @param num_bindings the number of storage buffers to bind from the array.
  *
  * @since This function is available since SDL 3.2.0.
  *
@@ -6477,7 +6451,8 @@
   Uint32 first_slot,
   SpanRef<const GPUBufferRaw> storage_buffers)
 {
-  SDL_BindGPUComputeStorageBuffers(compute_pass, first_slot, storage_buffers);
+  SDL_BindGPUComputeStorageBuffers(
+    compute_pass, first_slot, storage_buffers.data(), storage_buffers.size());
 }
 
 inline void GPUComputePass::BindStorageBuffers(
@@ -7119,11 +7094,11 @@
  *
  * @param command_buffer a command buffer.
  * @param window a window that has been claimed.
- * @param swapchain_texture a pointer filled in with a swapchain texture handle.
  * @param swapchain_texture_width a pointer filled in with the swapchain texture
  *                                width, may be nullptr.
  * @param swapchain_texture_height a pointer filled in with the swapchain
  *                                 texture height, may be nullptr.
+ * @returns a swapchain texture handle.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called from the thread that
@@ -7146,8 +7121,13 @@
   Uint32* swapchain_texture_width = nullptr,
   Uint32* swapchain_texture_height = nullptr)
 {
-  return CheckError(SDL_AcquireGPUSwapchainTexture(
-    command_buffer, window, swapchain_texture_width, swapchain_texture_height));
+  GPUTextureRaw texture;
+  CheckError(SDL_AcquireGPUSwapchainTexture(command_buffer,
+                                            window,
+                                            &texture,
+                                            swapchain_texture_width,
+                                            swapchain_texture_height));
+  return texture;
 }
 
 inline GPUTexture GPUCommandBuffer::AcquireSwapchainTexture(
@@ -7211,11 +7191,11 @@
  *
  * @param command_buffer a command buffer.
  * @param window a window that has been claimed.
- * @param swapchain_texture a pointer filled in with a swapchain texture handle.
  * @param swapchain_texture_width a pointer filled in with the swapchain texture
  *                                width, may be nullptr.
  * @param swapchain_texture_height a pointer filled in with the swapchain
  *                                 texture height, may be nullptr.
+ * @returns a swapchain texture handle.
  * @throws Error on failure.
  *
  * @threadsafety This function should only be called from the thread that
@@ -7233,8 +7213,13 @@
   Uint32* swapchain_texture_width = nullptr,
   Uint32* swapchain_texture_height = nullptr)
 {
-  return CheckError(SDL_WaitAndAcquireGPUSwapchainTexture(
-    command_buffer, window, swapchain_texture_width, swapchain_texture_height));
+  GPUTextureRaw texture;
+  CheckError(SDL_WaitAndAcquireGPUSwapchainTexture(command_buffer,
+                                                   window,
+                                                   &texture,
+                                                   swapchain_texture_width,
+                                                   swapchain_texture_height));
+  return texture;
 }
 
 inline GPUTexture GPUCommandBuffer::WaitAndAcquireSwapchainTexture(
@@ -7368,7 +7353,6 @@
  * @param wait_all if 0, wait for any fence to be signaled, if 1, wait for all
  *                 fences to be signaled.
  * @param fences an array of fences to wait on.
- * @param num_fences the number of fences in the fences array.
  * @throws Error on failure.
  *
  * @since This function is available since SDL 3.2.0.
@@ -7380,7 +7364,8 @@
                              bool wait_all,
                              std::span<GPUFence* const> fences)
 {
-  CheckError(SDL_WaitForGPUFences(device, wait_all, fences));
+  CheckError(
+    SDL_WaitForGPUFences(device, wait_all, fences.data(), fences.size()));
 }
 
 inline void GPUDevice::WaitForFences(bool wait_all,
@@ -7516,6 +7501,8 @@
     format, width, height, depth_or_layer_count);
 }
 
+#ifdef SDL_PLATFORM_GDK
+
 /**
  * Call this to suspend GPU operation on Xbox when you receive the
  * EVENT_DID_ENTER_BACKGROUND event.
@@ -7550,6 +7537,8 @@
 
 inline void GPUDevice::GDKResumeGPU() { SDL::GDKResumeGPU(m_resource); }
 
+#endif /* SDL_PLATFORM_GDK */
+
 /// @}
 
 } // namespace SDL
