--- src/generated/SDL3pp_keyboard.h
+++ include/SDL3pp/SDL3pp_keyboard.h
@@ -10,7 +10,7 @@
 namespace SDL {
 
 /**
- * @defgroup CategoryKeyboard Category Keyboard
+ * @defgroup CategoryKeyboard Keyboard Support
  *
  * SDL keyboard management.
  *
@@ -55,11 +55,9 @@
  * buttons, etc. You should wait for input from a device before you consider it
  * actively in use.
  *
- * @param count a pointer filled in with the number of keyboards returned, may
- *              be nullptr.
- * @returns a 0 terminated array of keyboards instance IDs or nullptr on
- *          failure; call GetError() for more information. This should be freed
- *          with free() when it is no longer needed.
+ * @returns a 0 terminated array of keyboards instance IDs on success..
+ *
+ * @throws Error on failure.
  *
  * @threadsafety This function should only be called on the main thread.
  *
@@ -68,7 +66,12 @@
  * @sa GetKeyboardNameForID
  * @sa HasKeyboard
  */
-inline OwnArray<KeyboardID> GetKeyboards() { return SDL_GetKeyboards(); }
+inline OwnArray<KeyboardID> GetKeyboards()
+{
+  int count;
+  auto data = CheckError(SDL_GetKeyboards(&count));
+  return OwnArray<KeyboardID>{data, size_t(count)};
+}
 
 /**
  * Get the name of a keyboard.
@@ -99,7 +102,7 @@
  *
  * @since This function is available since SDL 3.2.0.
  */
-inline WindowRef GetKeyboardFocus() { return SDL_GetKeyboardFocus(); }
+inline WindowRef GetKeyboardFocus() { return {SDL_GetKeyboardFocus()}; }
 
 /**
  * Get a snapshot of the current state of the keyboard.
@@ -122,7 +125,6 @@
  * Note: This function doesn't take into account whether shift has been pressed
  * or not.
  *
- * @param numkeys if non-nullptr, receives the length of the returned array.
  * @returns a pointer to an array of key states.
  *
  * @threadsafety It is safe to call this function from any thread.
@@ -134,7 +136,9 @@
  */
 inline std::span<const bool> GetKeyboardState()
 {
-  return SDL_GetKeyboardState();
+  int count;
+  auto data = SDL_GetKeyboardState(&count);
+  return std::span{data, size_t(count)};
 }
 
 /**
@@ -332,7 +336,7 @@
 
 inline void Window::SetTextInputArea(const RectRaw& rect, int cursor)
 {
-  CheckError(SDL_SetTextInputArea(m_resource, rect, cursor));
+  CheckError(SDL_SetTextInputArea(m_resource, &rect, cursor));
 }
 
 inline void Window::GetTextInputArea(RectRaw* rect, int* cursor)
